<!DOCTYPE html>
<html lang="en-us">
  <head>
    <meta charset="UTF-8">
    <title>SoS by bpeng2000</title>
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <link rel="stylesheet" type="text/css" href="stylesheets/normalize.css" media="screen">
    <link href='https://fonts.googleapis.com/css?family=Open+Sans:400,700' rel='stylesheet' type='text/css'>
    <link rel="stylesheet" type="text/css" href="stylesheets/stylesheet.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/github-light.css" media="screen">
  </head>
  <body>
    <section class="page-header">
      <h1 class="project-name">SoS</h1>
      <h2 class="project-tagline">Script of Scripts, a lightweight workflow system for the creation of readable pipelines</h2>
      <a href="https://github.com/bpeng2000/SOS" class="btn">View on GitHub</a>
      <a href="https://github.com/bpeng2000/SOS/zipball/master" class="btn">Download .zip</a>
      <a href="https://github.com/bpeng2000/SOS/tarball/master" class="btn">Download .tar.gz</a>
    </section>

    <section class="main-content">
      



<p><strong>Table of Contents</strong>  <em>generated with <a href="https://github.com/thlorenz/doctoc">DocToc</a></em></p>

<ul>
<li><a href="#design-of-sos">Design of SoS</a></li>
<li>
<a href="#tutorial">Tutorial</a>

<ul>
<li><a href="#organize-your-scripts-as-a-sos-script">Organize your scripts as a SoS script</a></li>
<li><a href="#make-the-script-work-for-other-input-files">Make the script work for other input files</a></li>
<li><a href="#ignore-steps-that-do-not-need-to-be-rerun">Ignore steps that do not need to be rerun</a></li>
<li><a href="#execute-steps-in-parallel">Execute steps in parallel</a></li>
</ul>
</li>
<li><a href="#unique-features-and-limitations">Unique features and limitations</a></li>
<li><a href="#summary">Summary</a></li>
</ul>



<p><strong>Script of Scripts (SoS)</strong> is a lightweight workflow system that helps you oraganize your commands and scripts in different languages into readable workflows that can be easily understood and modified by others. It is an easy-to-use alternative to specialized workflow systems such as <a href="http://common-workflow-language.github.io/draft-3/">CWL</a> which makes it an ideal tool for the creation and maintainance of workflows that need to be frequently updated and shared with others. </p>

<h2>
<a id="design-of-sos" class="anchor" href="#design-of-sos" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Design of SoS</h2>

<p>The core of SoS has mostly been implemented in another project but we are re-designing and re-implementing it to make SoS more user-friendly and powerful. Your involvement and suggestions are very welcome.</p>

<ul>
<li><a href="doc/commandline.md">User interface</a></li>
<li><a href="doc/sos_format_v1.md">Format specification</a></li>
<li><a href="doc/actions.md">Workflow actions</a></li>
<li><a href="doc/implementation.md">Implementation details</a></li>
<li><a href="doc/examples.md">Examples</a></li>
<li><a href="doc/problems.md">Pending problems</a></li>
</ul>

<h2>
<a id="tutorial" class="anchor" href="#tutorial" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Tutorial</h2>

<p>A SoS script consists of one or more scripts, comments and optional SoS directives. In its simplest form, a sos script is simply a series of scripts that can be executed sequentially by different intepreters.</p>

<p>Let us assume that you are a bioinformaticist needed to compare the expression levels between two samples. After reading some online tutorials, you ended up with some working commands</p>

<div class="highlight highlight-source-shell"><pre><span class="pl-c"># index reference genome</span>
STAR --runMode genomeGenerate --genomeFastaFile human38.fasta --genomeDir STAR_index
<span class="pl-c"># align reads to the reference genome</span>
STAR --genomeDir STAR_index --outSAMtype BAM SortedByCoordinate --readFilesIn control.fasta \
    --quantMode GeneCounts --outFileNamePrefix aligned/control
STAR --genomeDir STAR_index --outSAMtype BAM SortedByCoordinate --readFilesIn mutated.fasta \
    --quantMode GeneCounts --outFileNamePrefix aligned/mutated</pre></div>

<p>The first command builds an index of the reference genome in prepration for the latter steps, the second command aligns reads from the first sample to the reference genome, and the third command aligns reads from the second sample to the reference genome. Do not panic if you do not know what these commands are doing, this is just an example.</p>

<p>These commands generate, among other files, two files named <code>aligned/control.out.tab</code> and <code>aligned/mutated.out.tab</code> with expression counts of all genes. You then wrote a <a href="https://www.r-project.org/">R</a> script to analyze the results, something like</p>

<div class="highlight highlight-source-r"><pre><span class="pl-smi">control.count</span> <span class="pl-k">&lt;-</span> read.table(<span class="pl-s"><span class="pl-pds">'</span>aligned/control.out.tab<span class="pl-pds">'</span></span>)
<span class="pl-smi">mutated.count</span> <span class="pl-k">&lt;-</span> read.table(<span class="pl-s"><span class="pl-pds">'</span>aligned/mutated.out.tab<span class="pl-pds">'</span></span>)
<span class="pl-c"># normalize, compare, output etc, ignored.</span>
pdf(<span class="pl-s"><span class="pl-pds">'</span>myfigure.pdf<span class="pl-pds">'</span></span>)
<span class="pl-c"># plot results</span>
dev.off()</pre></div>

<h3>
<a id="organize-your-scripts-as-a-sos-script" class="anchor" href="#organize-your-scripts-as-a-sos-script" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Organize your scripts as a SoS script</h3>

<p>The project completed successfully and you needed to archive the scripts for later reference. Instead of having two files lying around with perhaps another <code>README</code> file to describe what you have done, you can write a single SoS script named <code>myanalysis.sos</code> with content</p>

<div class="highlight highlight-source-python"><pre><span class="pl-c">#!/usr/bin/env sos-runner</span>
<span class="pl-c">#fileformat=SOS1.0</span>

<span class="pl-c"># This script aligns raw reads of a control and a mutated sample </span>
<span class="pl-c"># to the reference genome and compare the expression values</span>
<span class="pl-c"># of the samples at genes A, B and C.</span>

[<span class="pl-c1">1</span>]
<span class="pl-c"># index reference genome</span>
run(<span class="pl-s"><span class="pl-pds">'''</span></span>
<span class="pl-s">STAR --runMode genomeGenerate --genomeFastaFile human38.fasta --genomeDir STAR_index</span>
<span class="pl-s"><span class="pl-pds">'''</span></span>)

[<span class="pl-c1">2</span>]
<span class="pl-c"># align reads to the reference genome</span>
run(<span class="pl-s"><span class="pl-pds">'''</span></span>
<span class="pl-s">STAR --genomeDir STAR_index --outSAMtype BAM SortedByCoordinate  --readFilesIn control.fasta <span class="pl-c1">\</span></span>
<span class="pl-s">    --quantMode GeneCounts --outFileNamePrefix aligned/control</span>
<span class="pl-s">STAR --genomeDir STAR_index --outSAMtype BAM SortedByCoordinate --readFilesIn mutated.fasta <span class="pl-c1">\</span></span>
<span class="pl-s">    --quantMode GeneCounts --outFileNamePrefix aligned/mutated</span>
<span class="pl-s"><span class="pl-pds">'''</span></span>)

[<span class="pl-c1">3</span>]
<span class="pl-c"># compare expression values</span>
R(<span class="pl-s"><span class="pl-pds">'''</span></span>
<span class="pl-s">control.count &lt;- read.table('aligned/control.out.tab')</span>
<span class="pl-s">mutated.count &lt;- read.table('aligned/mutated.out.tab')</span>
<span class="pl-s"># normalize, compare, output etc, ignored.</span>
<span class="pl-s">pdf('myfigure.pdf')</span>
<span class="pl-s"># plot results</span>
<span class="pl-s">dev.off()</span>
<span class="pl-s"><span class="pl-pds">'''</span></span>)
</pre></div>

<p>You can execute two shell scripts and a R script defined in this SoS script sequentially by running command</p>

<div class="highlight highlight-source-shell"><pre>sos run myanalysis.sos</pre></div>

<p>or simply</p>

<div class="highlight highlight-source-shell"><pre>myanalysis.sos</pre></div>

<p>if you give <code>myanalyis.sos</code> executable permission (<code>chmod +x myanalysis.sos</code>). </p>

<h3>
<a id="make-the-script-work-for-other-input-files" class="anchor" href="#make-the-script-work-for-other-input-files" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Make the script work for other input files</h3>

<p>After a while, before you almost forgot about this analysis, you needed to analyze another pair of samples. You could copy <code>myanalysis.sos</code> to <code>myanalysis2.sos</code>, change filenames and run it, but an easier way is to change your SoS file to accommodate other input files. This can be done by
defining a command line argument and passing files name to a <strong>SoS variable</strong>:</p>

<div class="highlight highlight-source-python"><pre><span class="pl-c">#!/usr/bin/env sos-runner</span>
<span class="pl-c">#fileformat=SOS1.0</span>

<span class="pl-c"># This script aligns raw reads of a control and a mutated sample </span>
<span class="pl-c"># to the reference genome and compare the expression values</span>
<span class="pl-c"># of the samples at genes A, B and C.</span>

[parameters]
<span class="pl-c"># Two input files in .fasta formats. The first one for control sample</span>
<span class="pl-c"># and the second one for mutated sample.</span>
fasta_files<span class="pl-k">=</span>[<span class="pl-s"><span class="pl-pds">'</span>control.fasta<span class="pl-pds">'</span></span>, <span class="pl-s"><span class="pl-pds">'</span>mutated.fasta<span class="pl-pds">'</span></span>]

[<span class="pl-c1">1</span>]
<span class="pl-c"># index reference genome</span>
run(<span class="pl-s"><span class="pl-pds">'''</span></span>
<span class="pl-s">STAR --runMode genomeGenerate --genomeFastaFile human38.fasta --genomeDir STAR_index</span>
<span class="pl-s"><span class="pl-pds">'''</span></span>)

[<span class="pl-c1">2</span>]
<span class="pl-c"># align reads to the reference genome</span>
run(<span class="pl-s"><span class="pl-pds">'''</span></span>
<span class="pl-s">STAR --genomeDir STAR_index --outSAMtype BAM SortedByCoordinate --readFilesIn $<span class="pl-c1">{fasta_files[0]}</span>  <span class="pl-c1">\</span></span>
<span class="pl-s">    --quantMode GeneCounts --outFileNamePrefix aligned/control</span>
<span class="pl-s">STAR --genomeDir STAR_index --outSAMtype BAM SortedByCoordinate --readFilesIn $<span class="pl-c1">{fasta_files[1]}</span>  <span class="pl-c1">\</span></span>
<span class="pl-s">    --quantMode GeneCounts --outFileNamePrefix aligned/mutated</span>
<span class="pl-s"><span class="pl-pds">'''</span></span>)

[<span class="pl-c1">3</span>]
<span class="pl-c"># compare expression values</span>
R(<span class="pl-s"><span class="pl-pds">'''</span></span>
<span class="pl-s">control.count &lt;- read.table('aligned/control.out.tab')</span>
<span class="pl-s">mutated.count &lt;- read.table('aligned/mutated.out.tab')</span>
<span class="pl-s"># normalize, compare, output etc, ignored.</span>
<span class="pl-s">pdf('myfigure.pdf')</span>
<span class="pl-s"># plot results</span>
<span class="pl-s">dev.off()</span>
<span class="pl-s"><span class="pl-pds">'''</span></span>)</pre></div>

<p>A command line argument <code>fasta_files</code> is defined in the <code>[parameters]</code>
section of the script. With this definition, you can pass two filenames to
variable <code>fasta_files</code> from command line</p>

<div class="highlight highlight-source-shell"><pre>sos run myanalysis.sos --fasta_files control1.fasta control2.fasta</pre></div>

<p><code>${fasta_files[0]}</code> and <code>${fasta_files[1]}</code> in command <code>STAR --genomeDir
...</code> are replaced with their values before the commands are executed.</p>

<h3>
<a id="ignore-steps-that-do-not-need-to-be-rerun" class="anchor" href="#ignore-steps-that-do-not-need-to-be-rerun" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Ignore steps that do not need to be rerun</h3>

<p>Although the SoS script now accepts command line arguments, it is still no more than a compilation of scripts and you immediately realized that it is a waste of time to execute the first command each time. To solve this problem, you can convert the SoS script to a real workflow by telling SoS the input and output of each step:</p>

<div class="highlight highlight-source-python"><pre><span class="pl-c">#!/usr/bin/env sos-runner</span>
<span class="pl-c">#fileformat=SOS1.0</span>

<span class="pl-c"># This script aligns raw reads of a control and a mutated sample </span>
<span class="pl-c"># to the reference genome and compare the expression values</span>
<span class="pl-c"># of the samples at genes A, B and C.</span>

[parameters]
<span class="pl-c"># Two input files in .fasta formats. The first one for control sample</span>
<span class="pl-c"># and the second one for mutated sample.</span>
fasta_files<span class="pl-k">=</span>[<span class="pl-s"><span class="pl-pds">'</span>control.fasta<span class="pl-pds">'</span></span>, <span class="pl-s"><span class="pl-pds">'</span>mutated.fasta<span class="pl-pds">'</span></span>]

[<span class="pl-c1">1</span>]
<span class="pl-c"># create a index for reference genome</span>
output: <span class="pl-s"><span class="pl-pds">'</span>STAR_index/chrName.txt<span class="pl-pds">'</span></span>

run(<span class="pl-s"><span class="pl-pds">'''</span></span>
<span class="pl-s">STAR --runMode genomeGenerate --genomeFastaFile human38.fasta --genomeDir STAR_index</span>
<span class="pl-s"><span class="pl-pds">'''</span></span>)

[<span class="pl-c1">2</span>]
<span class="pl-c"># align the reads to the reference genome</span>
<span class="pl-c1">input</span>:    fasta_files
depends:  <span class="pl-s"><span class="pl-pds">'</span>STAR_index/chrName.txt<span class="pl-pds">'</span></span>
output:   [<span class="pl-s"><span class="pl-pds">'</span>aligned/control.out.tab<span class="pl-pds">'</span></span>, <span class="pl-s"><span class="pl-pds">'</span>aligned/mutated.out.tab<span class="pl-pds">'</span></span>]

run(<span class="pl-s"><span class="pl-pds">'''</span></span>
<span class="pl-s">STAR --genomeDir STAR_index --outSAMtype BAM SortedByCoordinate  --readFilesIn $<span class="pl-c1">{input[0]}</span>  <span class="pl-c1">\</span></span>
<span class="pl-s">    --quantMode GeneCounts --outFileNamePrefix aligned/control</span>
<span class="pl-s">STAR --genomeDir STAR_index --outSAMtype BAM SortedByCoordinate --readFilesIn $<span class="pl-c1">{input[1]}</span>  <span class="pl-c1">\</span></span>
<span class="pl-s">    --quantMode GeneCounts --outFileNamePrefix aligned/mutated</span>
<span class="pl-s"><span class="pl-pds">'''</span></span>)

[<span class="pl-c1">3</span>]
<span class="pl-c"># compare expression values</span>
output: <span class="pl-s"><span class="pl-pds">'</span>myfigure.pdf<span class="pl-pds">'</span></span>

R(<span class="pl-s"><span class="pl-pds">'''</span></span>
<span class="pl-s">control.count &lt;- read.table('$<span class="pl-c1">{input[0]}</span>')</span>
<span class="pl-s">mutated.count &lt;- read.table('$<span class="pl-c1">{input[1]}</span>')</span>
<span class="pl-s"># normalize, compare, output etc, ignored.</span>
<span class="pl-s">pdf('myfigure.pdf')</span>
<span class="pl-s"># plot results</span>
<span class="pl-s">dev.off()</span>
<span class="pl-s"><span class="pl-pds">'''</span></span>)
</pre></div>

<p>Here we</p>

<ul>
<li>Use <strong>output directive</strong> to specify the expected output of all steps2.</li>
<li>Use <strong>input directive</strong> to specify the input of step 2. Step 1 by default has no input and step 3 by default has the output from its previous step.</li>
<li>Use <strong>depends directive</strong> to let step 2 depend on the output of step 1.</li>
<li>Use <code>${input[0]}</code> and <code>${input[1]}</code> in step 2 and 3 because these steps now have properly-defined <code>input</code>.</li>
</ul>

<p>With such information, when you run the same command</p>

<div class="highlight highlight-source-shell"><pre>sos run myanalysis.sos --input control1.fasta control2.fasta</pre></div>

<p>SoS will ignore step 1 if this step has been run with output <code>STAR_index/chrName.txt</code>. The same happens to step 2 and 3 so all steps will be ignored if you run the script repeatedly with the same input and processing scripts. SoS uses <strong>runtime signature</strong> for each step and will re-run the step if and only if the content or filename of input, output files or the processing scripts are changed.</p>

<h3>
<a id="execute-steps-in-parallel" class="anchor" href="#execute-steps-in-parallel" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Execute steps in parallel</h3>

<p>SoS allows execution of workflows in parallel. Although the three steps of this example have to be
executed sequentially, the second step runs the <code>STAR</code> command twice on two input files, and can
be executed in parallel. You can tell this to SoS by modifying the script as follows</p>

<div class="highlight highlight-source-python"><pre><span class="pl-c">#!/usr/bin/env sos-runner</span>
<span class="pl-c">#fileformat=SOS1.0</span>

<span class="pl-c"># This script aligns raw reads of a control and a mutated sample </span>
<span class="pl-c"># to the reference genome and compare the expression values</span>
<span class="pl-c"># of the samples at genes A, B and C.</span>

[parameters]
<span class="pl-c"># Two input files in .fasta formats. The first one for control sample</span>
<span class="pl-c"># and the second one for mutated sample.</span>
fasta_files<span class="pl-k">=</span>[<span class="pl-s"><span class="pl-pds">'</span>control.fasta<span class="pl-pds">'</span></span>, <span class="pl-s"><span class="pl-pds">'</span>mutated.fasta<span class="pl-pds">'</span></span>]

[<span class="pl-c1">1</span>]
<span class="pl-c"># create a index for reference genome</span>
output: <span class="pl-s"><span class="pl-pds">'</span>STAR_index/chrName.txt<span class="pl-pds">'</span></span>

run(<span class="pl-s"><span class="pl-pds">'''</span></span>
<span class="pl-s">STAR --runMode genomeGenerate --genomeFastaFile human38.fasta --genomeDir STAR_index</span>
<span class="pl-s"><span class="pl-pds">'''</span></span>)


[<span class="pl-c1">2</span>]
<span class="pl-c"># align the reads to the reference genome</span>
sample_type<span class="pl-k">=</span>[<span class="pl-s"><span class="pl-pds">'</span>control<span class="pl-pds">'</span></span>, <span class="pl-s"><span class="pl-pds">'</span>mutated<span class="pl-pds">'</span></span>]

<span class="pl-c1">input</span>:    fasta_files, group_by<span class="pl-k">=</span><span class="pl-s"><span class="pl-pds">'</span>single<span class="pl-pds">'</span></span>, labels<span class="pl-k">=</span><span class="pl-s"><span class="pl-pds">'</span>sample_type<span class="pl-pds">'</span></span>
depends:  <span class="pl-s"><span class="pl-pds">'</span>STAR_index/chrName.txt<span class="pl-pds">'</span></span>
output:   [<span class="pl-s"><span class="pl-pds">'</span>aligned/control.out.tab<span class="pl-pds">'</span></span>, <span class="pl-s"><span class="pl-pds">'</span>aligned/mutated.out.tab<span class="pl-pds">'</span></span>]

run(<span class="pl-s"><span class="pl-pds">'''</span></span>
<span class="pl-s">STAR --genomeDir STAR_index --outSAMtype BAM SortedByCoordinate  --readFilesIn $<span class="pl-c1">{input}</span>  <span class="pl-c1">\</span></span>
<span class="pl-s">    --quantMode GeneCounts --outFileNamePrefix aligned/$<span class="pl-c1">{_sample_type}</span></span>
<span class="pl-s"><span class="pl-pds">'''</span></span>)

[<span class="pl-c1">3</span>]
<span class="pl-c"># compare expression values</span>
output: <span class="pl-s"><span class="pl-pds">'</span>myfigure.pdf<span class="pl-pds">'</span></span>

R(<span class="pl-s"><span class="pl-pds">'''</span></span>
<span class="pl-s">control.count &lt;- read.table('$<span class="pl-c1">{input[0]}</span>')</span>
<span class="pl-s">mutated.count &lt;- read.table('$<span class="pl-c1">{input[1]}</span>')</span>
<span class="pl-s"># normalize, compare, output etc, ignored.</span>
<span class="pl-s">pdf('myfigure.pdf')</span>
<span class="pl-s"># plot results</span>
<span class="pl-s">dev.off()</span>
<span class="pl-s"><span class="pl-pds">'''</span></span>)
</pre></div>

<p>Here we </p>

<ol>
<li>Use option <code>group_by='single'</code> to pass input one by one to action. The action will be executed twice with <code>input</code> set to the first and second input file respectively.</li>
<li>Define a variable <code>sample_type</code> and mark it as the labels for input files (option <code>labels</code>). This will generate a variable <code>_sample_type</code> for each input file so <code>_sample_type</code> will be <code>control</code> for the first input file, and <code>mutated</code> for the second.</li>
</ol>

<p>Now, if you execute the script with option <code>-j 2</code> (2 concurrent processes),</p>

<div class="highlight highlight-source-shell"><pre>sos run myanalysis.sos --input control1.fasta control2.fasta -j 2</pre></div>

<p>the second step would be run in parallel.</p>

<p>We have showed you four versions of the same SoS script, each using more features of SoS. This actually demonstrates one of the advantages of the SoS system, namely you can start using SoS in minutes without knowing any of its advanced features, and gradually improve your script if needed.</p>

<h2>
<a id="unique-features-and-limitations" class="anchor" href="#unique-features-and-limitations" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Unique features and limitations</h2>

<p>The biggest difference between SoS and GNU make style workflow systems such as
<a href="https://bitbucket.org/johanneskoester/snakemake">snakemake</a> is that <strong>SoS is
input driven instead of output driven</strong>. SoS will execute all steps in a
workflow even if some steps do not produce any output (but can for example
check input files and stop the execution if an error occurs). It can also
execute steps conditionally, repeatedly, or generate different outputs with
different input files. In contrast, make-style systems focus on output files
and ignore steps that are irrelevant to the final outcome.</p>

<p>Compared to XML or system-specific workflow lanaguges such as
<a href="http://common-workflow-language.github.io/">CWL</a>, <strong>SoS emphasizes greatly
on readability</strong>. The workflow steps are logically arranged (although actual
execution order might differ) and contain explict commands and scripts.
Most comments are significant and provide usable information in output such
as <code>sos show script</code>. This makes it easy for readers to read the workflow
and modify it if needed. This can be very useful for fields such as
bioinformatics where workflows need to be frequently updated to accomodate
new tools, reference genomes, and annotations.</p>

<p>SoS is essentially a tool to compose and execute commands and scripts. It
uses workflow variables (and their derived forms) and string substitution
to create scripts (from user-provided templates) and execute them with
their own interpreters. Therefore, </p>

<ul>
<li>SoS only supports command line tools. There is no plan to extend it for interactive or GUI usages.</li>
<li>SoS does not understand the underlying tools and scripts and have no
control over exactly what they do. Incorrect use of workflow variables
might result in non-executable scripts.</li>
<li>SoS workflow system is largely file based (although you can check environments such as environmental variables through SoS variables). It does not support features such as piping between steps (neither do systems such as CWL support it).</li>
</ul>

<h2>
<a id="summary" class="anchor" href="#summary" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Summary</h2>

<p>The above example only shows a small fraction of what SoS can offer, but should be enough to demonstrate the unique features of SoS. Compared to maintaining multiple scripts or using more specifilized workflow systems such as <a href="http://www.yawlfoundation.org/">YAWL</a>, <a href="http://common-workflow-language.github.io/">CWL</a>, and <a href="https://galaxyproject.org/">Galaxy</a>,</p>

<ul>
<li>
<strong>SoS offers a way to organize your scripts in a single file</strong>, which makes it easy to execute and maintain. You can include small and freqently changed commands and scripts in SoS and keep large and stable scripts as separate scripts.</li>
<li>
<strong>SoS scripts are human readable and writable</strong>. In comparison, it would take a lot of time and practice to learn its syntax write a <a href="http://common-workflow-language.github.io/">CWL</a> workflow (see <a href="https://github.com/common-workflow-language/workflows/wiki/Tutorial-DRAFT2">this CWL tutorial</a> for an example). </li>
<li>
<strong>SoS help you execute your scripts with advanced workflow features</strong>. The workflow features of SoS is easy to use yet very powerful in helping you execute your pipelines efficiently not only locally, but also on cluster and cloud systems.</li>
</ul>

<p>If you are afraid of being tied to a new workflow tool, rest assured that SoS allows you to <strong><a href="doc/commandline.md#subcommand-export">export SoS scripts</a></strong> to a series of scripts called by a master bash (or windows .bat) script. This would allow you to execute your workflow in an environement without SoS installed.</p>

<p>Please refer to the SoS documentation for more details and feel free to <a href="mailto:ben.bob@gmail.com">contact me</a> if you have any comment on this project.</p>

      <footer class="site-footer">
        <span class="site-footer-owner"><a href="https://github.com/bpeng2000/SOS">SoS</a> is maintained by <a href="https://github.com/bpeng2000">bpeng2000</a>.</span>

        <span class="site-footer-credits">This page was generated by <a href="https://pages.github.com">GitHub Pages</a> using the <a href="https://github.com/jasonlong/cayman-theme">Cayman theme</a> by <a href="https://twitter.com/jasonlong">Jason Long</a>.</span>
      </footer>

    </section>

  
  </body>
</html>
