<!doctype html>
<!--[if lt IE 7]><html class="no-js lt-ie9 lt-ie8 lt-ie7" lang="en"> <![endif]-->
<!--[if (IE 7)&!(IEMobile)]><html class="no-js lt-ie9 lt-ie8" lang="en"><![endif]-->
<!--[if (IE 8)&!(IEMobile)]><html class="no-js lt-ie9" lang="en"><![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en"><!--<![endif]-->
<head>
<meta charset="utf-8">
<title>All Posts &#8211; SOS</title>
<meta name="description" content="An archive of posts.">




<!-- Open Graph -->
<meta property="og:locale" content="en_US">
<meta property="og:type" content="article">
<meta property="og:title" content="All Posts">
<meta property="og:description" content="An archive of posts.">
<meta property="og:url" content="https://bopeng.github.io/posts/">
<meta property="og:site_name" content="SOS">





<link rel="canonical" href="https://bopeng.github.io/posts/">
<link href="https://bopeng.github.io/feed.xml" type="application/atom+xml" rel="alternate" title="SOS Feed">

<!-- http://t.co/dKP3o1e -->
<meta name="HandheldFriendly" content="True">
<meta name="MobileOptimized" content="320">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<!-- For all browsers -->
<link rel="stylesheet" href="/SOS/assets/css/bootstrap.min.css">


<!-- <link rel="stylesheet" href="https://bopeng.github.io/style.css"> -->
<!-- Webfonts -->
    <link rel="stylesheet" href="/SOS/assets/css/font-awesome/css/font-awesome.min.css">
    <link href="https://fonts.googleapis.com/css?family=Montserrat:400,700" rel="stylesheet" type="text/css">
    <link href='https://fonts.googleapis.com/css?family=Kaushan+Script' rel='stylesheet' type='text/css'>
    <link href="https://fonts.googleapis.com/css?family=Lato:400,700,400italic,700italic" rel="stylesheet" type="text/css">
    <link href='https://fonts.googleapis.com/css?family=Droid+Serif:400,700,400italic,700italic' rel='stylesheet' type='text/css'>
    <link href='https://fonts.googleapis.com/css?family=Roboto+Slab:400,100,300,700' rel='stylesheet' type='text/css'>


<link rel="stylesheet" href="/SOS/assets/css/main.css">
<link rel="stylesheet" href="/SOS/assets/css/agency.css">
<link rel="stylesheet" href="/SOS/assets/css/pygmentsSoS.css">

<link href="//fonts.googleapis.com/css?family=Lato:300,400,700,300italic,400italic" rel="stylesheet" type="text/css">

<meta http-equiv="cleartype" content="on">

<!-- Load Modernizr -->
<script src="/SOS/assets/js/vendor/modernizr-2.6.2.custom.min.js"></script>

<!-- Icons -->
<!-- 16x16 -->
<link rel="shortcut icon" href="https://bopeng.github.io/favicon.ico">
<!-- 32x32 -->
<link rel="shortcut icon" href="https://bopeng.github.io/favicon.png">
<!-- 57x57 (precomposed) for iPhone 3GS, pre-2011 iPod Touch and older Android devices -->
<link rel="apple-touch-icon-precomposed" href="https://bopeng.github.io/images/apple-touch-icon-precomposed.png">
<!-- 72x72 (precomposed) for 1st generation iPad, iPad 2 and iPad mini -->
<link rel="apple-touch-icon-precomposed" sizes="72x72" href="https://bopeng.github.io/images/apple-touch-icon-72x72-precomposed.png">
<!-- 114x114 (precomposed) for iPhone 4, 4S, 5 and post-2011 iPod Touch -->
<link rel="apple-touch-icon-precomposed" sizes="114x114" href="https://bopeng.github.io/images/apple-touch-icon-114x114-precomposed.png">
<!-- 144x144 (precomposed) for iPad 3rd and 4th generation -->
<link rel="apple-touch-icon-precomposed" sizes="144x144" href="https://bopeng.github.io/images/apple-touch-icon-144x144-precomposed.png">



</head>

<body id="post-index" >

<!-- Navigation -->
    <nav class="navbar navbar-default navbar-fixed-top">
        <div class="container">
            <!-- Brand and toggle get grouped for better mobile display -->
            <div class="navbar-header page-scroll">
                <button type="button" class="navbar-toggle" data-toggle="collapse" data-target="#bs-example-navbar-collapse-1">
                    <span class="sr-only">Toggle navigation</span>
                    <span class="icon-bar"></span>
                    <span class="icon-bar"></span>
                    <span class="icon-bar"></span>
                </button>
                <a class="navbar-brand page-scroll" href="#page-top">SOS</a>
            </div>

            <!-- Collect the nav links, forms, and other content for toggling -->
            <div class="collapse navbar-collapse" id="bs-example-navbar-collapse-1">
                <ul class="nav navbar-nav navbar-right">
                    <li class="hidden">
                        <a href="#page-top"></a>
                    </li>
                    <li>
                        <a class="page-scroll" href="#intro">Introduction</a>
                    </li>
                    <li>
                        <a class="page-scroll" href="#documentation">Documentation</a>
                    </li>
                      <li>
                        <a class="page-scroll" href="#tutorial">Tutorial</a>
                    </li>
                    <li>
                        <a class="page-scroll" href="#about">About</a>
                    </li>
                    <li>
                        <a class="page-scroll" href="#team">Team</a>
                    </li>
                    <li>
                        <a class="page-scroll" href="#contact">Contact</a>
                    </li>
                </ul>
            </div>
            <!-- /.navbar-collapse -->
        </div>
        <!-- /.container-fluid -->
    </nav>

    <!-- Header -->
   <header style="background-image: url('https://bopeng.github.io//SOS/images/header-bg.jpg');"">
      
        <div class="container">
            <div class="intro-text">
                 <div class="intro-heading">SOS</div>
                <div class="intro-lead-in">Script of Scripts, a lightweight workflow system for the creation of readable workflows</div>
               
                <a href="#services" class="page-scroll btn btn-xl">Tell Me More</a>
            </div>
        </div>
    </header>
<!--
        
          <a class="page-link" href="/SOS/404.html">Page Not Found</a>
        
          <a class="page-link" href="/SOS/posts/">All Posts</a>
        
          <a class="page-link" href="/SOS/"></a>
        
          <a class="page-link" href="/SOS/tags/">Tag Archive</a>
        
          
        
          
        
 -->


<!--[if lt IE 9]><div class="upgrade"><strong><a href="http://whatbrowser.org/">Your browser is quite old!</strong> Why not upgrade to a different browser to better enjoy this site?</a></div><![endif]-->



    <section id="intro" class="bg-medium-gray">
        <div class="container">
            <div class="row">
              <div class="col-lg-12 text-center">
                    <h2 class="section-heading">Introduction</h2>
                </div>
              </div>
            <div class="row" id="main" role="main">
                
    
    

    
    <article>
      <h2 id="2016-ref">2016</h2>
      <ul>
    

        <li class="entry-title"><a href="https://bopeng.github.io/home/" title="Home">Home</a></li>

    
      </ul>
    </article>
    

              </div><!-- /#main -->
          </div>
        </section>  

<!-- Portfolio Grid Section -->
    <section id="documentation" class="bg-light-gray">
        <div class="container">
            <div class="row">
                <div class="col-lg-12 text-center">
                    <h2 class="section-heading">Documentation</h2>
                </div>
            </div>
            <div class="row">
        
            
                <div class="col-md-4 col-sm-6 portfolio-item">
                    <a href="#Introduction" class="portfolio-link" data-toggle="modal">
              <!--           <div class="portfolio-hover">
                            <div class="portfolio-hover-content">
                                <i class="fa fa-plus fa-3x"></i>
                            </div>
                        </div> -->
                        1.Introduction
                     <!--    <img src="img/portfolio/" class="img-responsive img-centered" alt=""> -->
                    </a>
                    <!-- <div class="portfolio-caption">
                        <h4></h4>
                        <p class="text-muted"></p>
                    </div> -->
                </div>
            
                <div class="col-md-4 col-sm-6 portfolio-item">
                    <a href="#SoS-Syntax" class="portfolio-link" data-toggle="modal">
              <!--           <div class="portfolio-hover">
                            <div class="portfolio-hover-content">
                                <i class="fa fa-plus fa-3x"></i>
                            </div>
                        </div> -->
                        2.SoS-Syntax
                     <!--    <img src="img/portfolio/" class="img-responsive img-centered" alt=""> -->
                    </a>
                    <!-- <div class="portfolio-caption">
                        <h4></h4>
                        <p class="text-muted"></p>
                    </div> -->
                </div>
            
                <div class="col-md-4 col-sm-6 portfolio-item">
                    <a href="#File-Structure" class="portfolio-link" data-toggle="modal">
              <!--           <div class="portfolio-hover">
                            <div class="portfolio-hover-content">
                                <i class="fa fa-plus fa-3x"></i>
                            </div>
                        </div> -->
                        3.File-Structure
                     <!--    <img src="img/portfolio/" class="img-responsive img-centered" alt=""> -->
                    </a>
                    <!-- <div class="portfolio-caption">
                        <h4></h4>
                        <p class="text-muted"></p>
                    </div> -->
                </div>
            
                <div class="col-md-4 col-sm-6 portfolio-item">
                    <a href="#Command-line-options-and-configuration-file" class="portfolio-link" data-toggle="modal">
              <!--           <div class="portfolio-hover">
                            <div class="portfolio-hover-content">
                                <i class="fa fa-plus fa-3x"></i>
                            </div>
                        </div> -->
                        4.Command-line-options-and-configuration-file
                     <!--    <img src="img/portfolio/" class="img-responsive img-centered" alt=""> -->
                    </a>
                    <!-- <div class="portfolio-caption">
                        <h4></h4>
                        <p class="text-muted"></p>
                    </div> -->
                </div>
            
                <div class="col-md-4 col-sm-6 portfolio-item">
                    <a href="#Workflow-Definition" class="portfolio-link" data-toggle="modal">
              <!--           <div class="portfolio-hover">
                            <div class="portfolio-hover-content">
                                <i class="fa fa-plus fa-3x"></i>
                            </div>
                        </div> -->
                        5.Workflow-Definition
                     <!--    <img src="img/portfolio/" class="img-responsive img-centered" alt=""> -->
                    </a>
                    <!-- <div class="portfolio-caption">
                        <h4></h4>
                        <p class="text-muted"></p>
                    </div> -->
                </div>
            
                <div class="col-md-4 col-sm-6 portfolio-item">
                    <a href="#SoS-Steps" class="portfolio-link" data-toggle="modal">
              <!--           <div class="portfolio-hover">
                            <div class="portfolio-hover-content">
                                <i class="fa fa-plus fa-3x"></i>
                            </div>
                        </div> -->
                        6.SoS-Steps
                     <!--    <img src="img/portfolio/" class="img-responsive img-centered" alt=""> -->
                    </a>
                    <!-- <div class="portfolio-caption">
                        <h4></h4>
                        <p class="text-muted"></p>
                    </div> -->
                </div>
            
                <div class="col-md-4 col-sm-6 portfolio-item">
                    <a href="#Make-Style-Rules" class="portfolio-link" data-toggle="modal">
              <!--           <div class="portfolio-hover">
                            <div class="portfolio-hover-content">
                                <i class="fa fa-plus fa-3x"></i>
                            </div>
                        </div> -->
                        7.Make-Style-Rules
                     <!--    <img src="img/portfolio/" class="img-responsive img-centered" alt=""> -->
                    </a>
                    <!-- <div class="portfolio-caption">
                        <h4></h4>
                        <p class="text-muted"></p>
                    </div> -->
                </div>
            
                <div class="col-md-4 col-sm-6 portfolio-item">
                    <a href="#SoS-Actions" class="portfolio-link" data-toggle="modal">
              <!--           <div class="portfolio-hover">
                            <div class="portfolio-hover-content">
                                <i class="fa fa-plus fa-3x"></i>
                            </div>
                        </div> -->
                        8.SoS-Actions
                     <!--    <img src="img/portfolio/" class="img-responsive img-centered" alt=""> -->
                    </a>
                    <!-- <div class="portfolio-caption">
                        <h4></h4>
                        <p class="text-muted"></p>
                    </div> -->
                </div>
            
                <div class="col-md-4 col-sm-6 portfolio-item">
                    <a href="#Workflow-Execution" class="portfolio-link" data-toggle="modal">
              <!--           <div class="portfolio-hover">
                            <div class="portfolio-hover-content">
                                <i class="fa fa-plus fa-3x"></i>
                            </div>
                        </div> -->
                        9.Workflow-Execution
                     <!--    <img src="img/portfolio/" class="img-responsive img-centered" alt=""> -->
                    </a>
                    <!-- <div class="portfolio-caption">
                        <h4></h4>
                        <p class="text-muted"></p>
                    </div> -->
                </div>
            
                <div class="col-md-4 col-sm-6 portfolio-item">
                    <a href="#User-Interface" class="portfolio-link" data-toggle="modal">
              <!--           <div class="portfolio-hover">
                            <div class="portfolio-hover-content">
                                <i class="fa fa-plus fa-3x"></i>
                            </div>
                        </div> -->
                        10.User-Interface
                     <!--    <img src="img/portfolio/" class="img-responsive img-centered" alt=""> -->
                    </a>
                    <!-- <div class="portfolio-caption">
                        <h4></h4>
                        <p class="text-muted"></p>
                    </div> -->
                </div>
            
            </div>
        </div>
    </section>

<!-- Portfolio Grid Section -->
    <section id="tutorial" class="bg-medium-gray">
        <div class="container">
            <div class="row">
                <div class="col-lg-12 text-center">
                    <h2 class="section-heading">Tutorial</h2>
                </div>
            </div>
            <div class="row">
            
                <div class="col-md-4 col-sm-6 portfolio-item">
                    <a href="#Quick-start" class="portfolio-link" data-toggle="modal">
              <!--           <div class="portfolio-hover">
                            <div class="portfolio-hover-content">
                                <i class="fa fa-plus fa-3x"></i>
                            </div>
                        </div> -->
                        1.Quick-Start
                     <!--    <img src="img/portfolio/" class="img-responsive img-centered" alt=""> -->
                    </a>
                    <!-- <div class="portfolio-caption">
                        <h4></h4>
                        <p class="text-muted"></p>
                    </div> -->
                </div>
            
                <div class="col-md-4 col-sm-6 portfolio-item">
                    <a href="#Using-SoS-with-iPython" class="portfolio-link" data-toggle="modal">
              <!--           <div class="portfolio-hover">
                            <div class="portfolio-hover-content">
                                <i class="fa fa-plus fa-3x"></i>
                            </div>
                        </div> -->
                        2.Using-SoS-with-iPython
                     <!--    <img src="img/portfolio/" class="img-responsive img-centered" alt=""> -->
                    </a>
                    <!-- <div class="portfolio-caption">
                        <h4></h4>
                        <p class="text-muted"></p>
                    </div> -->
                </div>
            
                <div class="col-md-4 col-sm-6 portfolio-item">
                    <a href="#SoS-Notebook-Using-Jupyter" class="portfolio-link" data-toggle="modal">
              <!--           <div class="portfolio-hover">
                            <div class="portfolio-hover-content">
                                <i class="fa fa-plus fa-3x"></i>
                            </div>
                        </div> -->
                        3.SoS-Notebook-Using-Jupyter
                     <!--    <img src="img/portfolio/" class="img-responsive img-centered" alt=""> -->
                    </a>
                    <!-- <div class="portfolio-caption">
                        <h4></h4>
                        <p class="text-muted"></p>
                    </div> -->
                </div>
            
                <div class="col-md-4 col-sm-6 portfolio-item">
                    <a href="#Using-Spyder-as-SoS-IDE" class="portfolio-link" data-toggle="modal">
              <!--           <div class="portfolio-hover">
                            <div class="portfolio-hover-content">
                                <i class="fa fa-plus fa-3x"></i>
                            </div>
                        </div> -->
                        4.Using-Spyder-as-SoS-IDE
                     <!--    <img src="img/portfolio/" class="img-responsive img-centered" alt=""> -->
                    </a>
                    <!-- <div class="portfolio-caption">
                        <h4></h4>
                        <p class="text-muted"></p>
                    </div> -->
                </div>
            
                <div class="col-md-4 col-sm-6 portfolio-item">
                    <a href="#A-Complete-Example" class="portfolio-link" data-toggle="modal">
              <!--           <div class="portfolio-hover">
                            <div class="portfolio-hover-content">
                                <i class="fa fa-plus fa-3x"></i>
                            </div>
                        </div> -->
                        5.A-Complete-Example
                     <!--    <img src="img/portfolio/" class="img-responsive img-centered" alt=""> -->
                    </a>
                    <!-- <div class="portfolio-caption">
                        <h4></h4>
                        <p class="text-muted"></p>
                    </div> -->
                </div>
            
                <div class="col-md-4 col-sm-6 portfolio-item">
                    <a href="#SoS-Docker-Guide" class="portfolio-link" data-toggle="modal">
              <!--           <div class="portfolio-hover">
                            <div class="portfolio-hover-content">
                                <i class="fa fa-plus fa-3x"></i>
                            </div>
                        </div> -->
                        6.SoS-Docker-Guide
                     <!--    <img src="img/portfolio/" class="img-responsive img-centered" alt=""> -->
                    </a>
                    <!-- <div class="portfolio-caption">
                        <h4></h4>
                        <p class="text-muted"></p>
                    </div> -->
                </div>
            
                <div class="col-md-4 col-sm-6 portfolio-item">
                    <a href="#Generating-Report-with-SoS" class="portfolio-link" data-toggle="modal">
              <!--           <div class="portfolio-hover">
                            <div class="portfolio-hover-content">
                                <i class="fa fa-plus fa-3x"></i>
                            </div>
                        </div> -->
                        7.Generating-Report-with-SoS
                     <!--    <img src="img/portfolio/" class="img-responsive img-centered" alt=""> -->
                    </a>
                    <!-- <div class="portfolio-caption">
                        <h4></h4>
                        <p class="text-muted"></p>
                    </div> -->
                </div>
            
            </div>
        </div>
    </section>
<!-- Team Section -->
    <section id="team" class="bg-light-gray">
        <div class="container">
            <div class="row">
                <div class="text-center">
                    <h2 class="section-heading">Our Amazing Team</h2>
                </div>
            </div>
            <div class="row">
        
                    <div class="team-member">
                        <h4>Bo Peng</h4>
                        <p class="text-muted">Assistant Professor, Department of Bioinformatics and Computational Biology, MD Anderson Cancer Center.</p>
                    </div>
                </div>
            <!-- <div class="row">
                <div class="col-lg-8 col-lg-offset-2 text-center">
                    <p class="large text-muted">Lorem ipsum dolor sit amet, consectetur adipisicing elit. Aut eaque, laboriosam veritatis, quos non quis ad perspiciatis, totam corporis ea, alias ut unde.</p>
                </div>
            </div> -->
        </div>
    </section>
<section id="contact">
        <div class="container">
            <div class="row">
                <div class="col-lg-12 text-center">
                    <h2 class="section-heading">Contact Us</h2>
                </div>
            </div>
            <div class="row">
                <div class="col-lg-12">
                    <form name="sentMessage" id="contactForm" novalidate>
                        <div class="row">
                            <div class="col-md-6">
                                <div class="control-group form-group">
                                    <input type="text" class="form-control" placeholder="Your Name *" id="name" required data-validation-required-message="Please enter your name.">
                                    <p class="help-block text-danger"></p>
                                </div>
                                <div class="control-group form-group">
                                    <input type="email" class="form-control" placeholder="Your Email *" id="email" required data-validation-required-message="Please enter your email address.">
                                    <p class="help-block text-danger"></p>
                                </div>
                                <div class="control-group form-group">
                                    <input type="tel" class="form-control" placeholder="Your Phone *" id="phone" required data-validation-required-message="Please enter your phone number.">
                                    <p class="help-block text-danger"></p>
                                </div>
                            </div>
                            <div class="col-md-6">
                                <div class="control-group form-group">
                                    <textarea class="form-control" placeholder="Your Message *" id="message" required data-validation-required-message="Please enter a message."></textarea>
                                    <p class="help-block text-danger"></p>
                                </div>
                            </div>
                            <div class="clearfix"></div>
                            <div class="col-lg-12 text-center">
                                <div id="success"></div>
                                <button type="submit" class="btn btn-xl">Send Message</button>
                            </div>
                        </div>
                    </form>
                </div>
            </div>
        </div>
    </section>


<div class="footer-wrapper">
  <footer role="contentinfo">
    <span>&copy; 2016 Bo Peng.</span>
  </footer>
</div><!-- /.footer-wrapper -->

 <!-- Portfolio Modals -->
 
   <div class="portfolio-modal modal fade" id="Introduction" tabindex="-1" role="dialog" aria-hidden="true">
        <div class="modal-content">
            <div class="close-modal" data-dismiss="modal">
                <div class="lr">
                    <div class="rl">
                    </div>
                </div>
            </div>
            <div class="container">
                <div class="row">
                 <div class="col-lg-8 col-lg-offset-2">
                        <h1>1.Introduction</h1>
                        <div class="modal-body">
                         
                            <hr class="star-primary">
                            <h2 id="overview-of-sos-concepts">Overview of SoS concepts</h2>

<p>A SoS <strong>script</strong> defines one or more <strong>workflows</strong>, and each workflow consists of one or more <strong>steps</strong>.</p>

<p><img src="/SOS/media/workflow.png" alt="workflow" class="img-responsive" /></p>

<p>Although the input and output can be more general, each step typically has its <strong>input</strong>, <strong>output</strong>, and <strong>dependents</strong> files, it executes a <strong>step process</strong> that consists of one or more Python statements and SoS actions (special python functions). Part or all the step process, called <strong>tasks</strong>, can be executed and monitored externally.</p>

<p><img src="/SOS/media/sos_step.png" alt="sos_step" class="img-responsive" /></p>

<p>A SoS script contains <strong>comments</strong>, <strong>statements</strong>, and one or more SoS <strong>steps</strong>. A SoS <strong>step</strong> consists of a <strong>header</strong>
with one or more step names and optional options. The body of a SoS step consists of optional <strong>comments</strong>, 
<strong>statements</strong>, <strong>input</strong>, <strong>output</strong>, <strong>depends</strong> files, <strong>parameter</strong> definitions, followed by step <strong>process</strong>. The following figure 
shows a sample script that defines a workflow with two steps:</p>

<p><img src="/SOS/media/sample_script.jpg" alt="sample_script" class="img-responsive" /></p>

<h2 id="formal-definitions-of-terminology--grammar">Formal definitions of terminology &amp; grammar</h2>

<ul>
  <li><strong>Script</strong>: A SoS script that defines one or more workflows.</li>
  <li><strong>Workflow</strong>: A sequence of processes that can be executed to complete certain task.</li>
  <li><strong>Step</strong>: A step of a workflow that perform one piece of the workflow.</li>
  <li><strong>Target</strong>: Objects that are input and result of a SoS step, which are usually files, but can also be objects such as an executable command (with variable locations), and a SoS variable.</li>
  <li><strong>Step options</strong>: Options of the step that assist the definition of the workflow.</li>
  <li><strong>Step input</strong>: Specifies the input files of the step.</li>
  <li><strong>Step output</strong>: Specifies the output files and targets of the step.</li>
  <li><strong>Step dependencies</strong>: Specifies the files and targets that are required by the step.</li>
  <li><strong>Step process</strong>: The process that a step executes to complete specified work, specified as one or more Python statements.</li>
  <li><strong>Task</strong>: Part or all step process that will be executed and monitored outside of SoS. These are usually resource intensive jobs that will take long time to complete.</li>
  <li><strong>Action</strong>: SoS or user-defined Python functions. They differ from regular Python functions in that they may behave differently in different running mode of SoS (e.g. ignore when executed in dryrun mode).</li>
</ul>

<p>More formally defined, the SoS syntax obeys the following grammar, given in extended Backus-Naur form (EBNF):</p>

<pre><code class="language-python">Script         = {comment}, {statement}, {step};
comment        = "#", text, NEWLINE
assignment     = name, "=", expression, NEWLINE
</code></pre>

<p>with SoS steps defined as</p>

<pre><code>step           = step_header,
                 {comment}, { {statement}, [input | output | depends ]},
                 [process, NEWLINE, {script} ]
step_header    = "[", section_names, [":", names | options], "]", NEWLINE
parameter      = "parameter", ":", assignment
input          = "input", ":", [expressions], [",", options], NEWLINE
output         = "output", ":", [expressions], [",", options], NEWLINE
depends        = "depends", ":", [expressions], [",", options], NEWLINE
task           = "task", ":",  [options]
action         = func_format | script_format
func_format    = name, "(", [options], ")"
script_format  = name, ":", [options], NEWLINE, script 
section_names  = section_name, ",", section_name
section_name   = name, "(", text, ")"
names          = name, {",", name}
workflow       = name, ['_', steps], {"+", name, ['_', steps}
assignment     = name, "=", expression, NEWLINW
expressions    = expression, {",", expression}
options        = option, {"," option}
option         = name, "=", expression
</code></pre>

<p>Here <code>name</code>, <code>expression</code> and <code>statement</code> are arbitrary <a href="http://www.python.org">Python</a> names, expression and statements with added SoS features. <strong>SoS requires Python 3 and does not support Python 2.x specific syntax</strong></p>

<p>This documentation is a comprehensive reference to all SoS features. Please refer to this <a href="https://github.com/BoPeng/SOS/wiki/1.-Quick-Start">SoS quick start tutorial</a> to learn some basics of SoS before diving into the details.</p>


                            </div>
                            <button type="button" class="btn btn-default" data-dismiss="modal"><i class="fa fa-times"></i> Close</button>
                        </div>
                    </div> 
                    <div class="footer-wrapper">
                      <footer role="contentinfo">
                        <span>&copy; 2016 Bo Peng.</span>
                      </footer>
                    </div>
                </div>
            </div>
        </div>
    </div> 


   <div class="portfolio-modal modal fade" id="User-Interface" tabindex="-1" role="dialog" aria-hidden="true">
        <div class="modal-content">
            <div class="close-modal" data-dismiss="modal">
                <div class="lr">
                    <div class="rl">
                    </div>
                </div>
            </div>
            <div class="container">
                <div class="row">
                 <div class="col-lg-8 col-lg-offset-2">
                        <h1>10.User-Interface</h1>
                        <div class="modal-body">
                         
                            <hr class="star-primary">
                            <h2 id="command-sos">Command <code>sos</code></h2>

<p>Command <code>sos</code> accepts a number of subcommands (similar to <code>svn</code>, <code>git</code> etc). Its syntax follows</p>

<pre><code class="language-python">sos subcommand [subcommand-options]
</code></pre>

<p>You can use command</p>

<pre><code class="language-python">$ sos -h
usage: sos [-h] [--version]
           {run,dryrun,prepare,convert,remove,start,config} ...

A workflow system for the execution of commands and scripts in different
languages.

optional arguments:
  -h, --help            show this help message and exit
  --version             show program's version number and exit

subcommands:
  {run,dryrun,prepare,convert,remove,start,config}
    run                 Execute a SoS script
    dryrun              Execute a SoS script in dryrun mode
    prepare             Execute a SoS script in prepare mode
    convert             Convert between sos and other file formats such as
                        html and Jupyter notebooks
    remove              Remove tracked and/or untracked files with their
                        signatures
    config              Set, unset or get the value of system or local
                        configuration files

Use 'sos cmd -h' for details about each subcommand. Please contact Bo Peng
(bpeng at mdanderson.org) if you have any question.
</code></pre>

<p>to get a list of subcommands with brief descriptions and</p>

<pre><code class="language-python">sos subcommand -h
</code></pre>

<p>to get detailed description of a particular subcommand.</p>

<h2 id="subcommand-run">subcommand <code>run</code></h2>

<pre><code class="language-python">$ sos run -h
usage: sos run [-h] [-j JOBS] [-c CONFIG_FILE] [-t FILE [FILE ...]]
               [-b [BIN_DIR [BIN_DIR ...]]] [-q QUEUE] [-n] [-p] [-f] [-F]
               [-v {0,1,2,3,4}]
               SCRIPT [WORKFLOW]

Execute a workflow defined in script

positional arguments:
  SCRIPT                A SoS script that defines one or more workflows. The
                        script can be a filename or a URL from which the
                        content of a SoS will be read. If a valid file cannot
                        be located or downloaded, SoS will search for the
                        script in a search path specified by variable
                        `sos_path` defined in the global SoS configuration
                        file (~/.sos/config.yaml).
  WORKFLOW              Name of the workflow to execute. This option can be
                        ignored if the script defines a default workflow (with
                        no name or with name `default`) or defines only a
                        single workflow. A subworkflow or a combined workflow
                        can also be specified, where a subworkflow executes a
                        subset of workflow (`name_steps` where `steps` can be
                        `n` (a step `n`), `:n` (up to step `n`), `n:m` (from
                        step `n` to `m`), and `n:` (from step `n`)), and a
                        combined workflow executes to multiple (sub)workflows
                        combined by `+` (e.g. `A_0+B+C`).

optional arguments:
  -h, --help            show this help message and exit
  -j JOBS               Number of concurrent process allowed. A workflow is by
                        default executed sequentially (-j 1). If a greater
                        than 1 number is specified SoS will execute the
                        workflow in parallel mode and execute up to specified
                        processes concurrently. These include looped processes
                        within a step (with runtime option `concurrent=True`)
                        and steps with non-missing required files.
  -c CONFIG_FILE        A configuration file in the format of YAML/JSON. The
                        content of the configuration file will be available as
                        a dictionary CONF in the SoS script being executed.
  -t FILE [FILE ...]    One of more files or alias of other targets that will
                        be the target of execution. If specified, SoS will
                        execute only part of a workflow or multiple workflows
                        or auxiliary steps to generate specified targets.
  -b [BIN_DIR [BIN_DIR ...]]
                        Extra directories in which SoS will look for
                        executables before standard $PATH. This option
                        essentially prefix $PATH with these directories. Note
                        that the default value '~/.sos/bin' is by convention a
                        default directory for commands that are installed by
                        SoS. You can use option '-b' without value to disallow
                        commands under ~/.sos/bin.
  -q QUEUE              Task-processing queue. SoS by default uses a local
                        multiprocessing queue where tasks are executed by
                        different processes. Supported task queues include a
                        'rq' engine where tasks will be distributed to one or
                        more rq-workers with assistance from a redis server,
                        and a 'celery' quque where tasks will be distributed
                        to celery workers.
  -v {0,1,2,3,4}, --verbosity {0,1,2,3,4}
                        Output error (0), warning (1), info (2), debug (3) and
                        trace (4) information to standard output (default to
                        2).

Run mode options:
  SoS scripts are by default executed in run mode where all the script is
  run in dryrun mode to check syntax error, prepare mode to prepare
  resources, and run mode to execute the pipelines. Run mode options allow
  you to execute these steps selectively.

  -n                    Execute a workflow without executing any actions. This
                        can be used to check the syntax of a SoS file.
  -p                    Execute the workflow in preparation mode in which SoS
                        prepare the execution of workflow by, for example,
                        download required resources and docker images.
  -f                    Execute the workflow in a special run mode that
                        ignores saved runtime signatures and re-execute all
                        the steps.
  -F                    Execute the workflow in a special run mode that re-use
                        existing output files and recontruct runtime
                        signatures if output files exist.

Arbitrary parameters defined by the [parameters] step of the script, and
[parameters] steps of other scripts if nested workflows are defined in other
SoS files (option `source`). The name, default and type of the parameters are
specified in the script. Single value parameters should be passed using option
`--name value` and multi-value parameters should be passed using option
`--name value1 value2`.
</code></pre>

<p>Examples of the <code>sos run</code> include</p>

<pre><code class="language-python">sos run -h                       # get help message
sos run myscript                 # run default workflow defined in myscript
sos run myscript align           # run workflow align defined in myscript
sos run myscript align+call      # run a combined workflow align+call
sos run myscript align -h        # help message for the align workflow defined in myscript
sos run -n myscript align        # run align workflow in inspect mode 
</code></pre>

<h2 id="subcommand-dryrun">subcommand <code>dryrun</code></h2>

<p>This command execute the script in dryrun mode. It is alias to command <code>sos run -n</code>.</p>

<h2 id="subcommand-prepare">subcommand <code>prepare</code></h2>

<p>This command execute the script in prepare mode. It is alias to command <code>sos run -p</code>.</p>

<h2 id="subcommand-convert">subcommand <code>convert</code></h2>

<pre><code class="language-python">$ sos convert -h
usage: sos convert [-h] [--html [FILENAME]] [--markdown [FILENAME]] [--term]
                   [--notebook [FILENAME]] [--sos [SCRIPT]] [-v {0,1,2,3,4}]
                   FILENAME [WORKFLOW]

The show command displays details of all workflows defined in a script,
including description of script, workflow, steps, and command line parameters.
The output can be limited to a specified workflow (which can be a subworkflow
or a combined workflow) if a workflow is specified.

positional arguments:
  FILENAME              File to be converted, can be a SoS script or a Jupyter
                        notebook.
  WORKFLOW              Workflow to be converted if the file being converted
                        is a SoS script.

optional arguments:
  -h, --help            show this help message and exit
  --html [FILENAME]     Generate a syntax-highlighted HTML file, write it to a
                        specified file, or view in a browser if no filename is
                        specified. Additional argument --raw can be used to
                        specify a URL to raw file, arguments --linenos and
                        --style can be used to customize style of html output.
                        You can pass an arbitrary name to option --style get a
                        list of available styles.
  --markdown [FILENAME]
                        Convert script or workflow to markdown format and
                        write it to specified file, or standard output if not
                        filename is specified.
  --term                Output syntax-highlighted script or workflow to the
                        terminal. Additional arguments --bg=light|dark
                        --lineno can be used to customized output.
  --notebook [FILENAME]
                        Convert script or workflow to jupyter notebook format
                        and write it to specified file, or standard output if
                        no filename is specified. If the input file is a
                        notebook, it will be converted to .sos (see option
                        --sos) then to notebook, resetting indexes and
                        removing all output cells.
  --sos [SCRIPT]        Convert specified Jupyter notebook to SoS format. The
                        output is the same as you use File -&gt; Download as -&gt;
                        SoS (.sos) from Jupyter with nbconvert version 4.2.0
                        or higher although you can customize output using
                        options --reorder (rearrange notebook cells with
                        execution order), --reset-index (reset indexes to 1,
                        2, 3, ..), --add-header (add section header [index] if
                        the cell does not start with a header), --no-index
                        (does not save cell index), --remove-magic (remove
                        cell magic), and --md-to-report (convert markdown cell
                        to code cell with report.)
  -v {0,1,2,3,4}, --verbosity {0,1,2,3,4}
                        Output error (0), warning (1), info (2), debug (3) and
                        trace (4) information to standard output (default to
                        2).

Extra command line argument could be specified to customize the style of html,
markdown, and terminal output.
</code></pre>

<p>The <code>--html</code> option is very useful in displaying SoS scripts with proper syntax highlighting for different embedded languages. It can be used to generate a nice-looking .html file if you run</p>

<pre><code class="language-python">sos convert myscript.sos --html myscript.sos.html
</code></pre>

<p>or generate a temporary file and display it in a browser if you do not specify an output filename.</p>

<pre><code class="language-python">sos convert myscript.sos --html myscript.sos.html
</code></pre>

<p>For example, the example in our <a href="https://github.com/bpeng2000/SOS/wiki/Quick-Start">QuickStart Guide</a> can be displayed as</p>

<p><img src="/SOS/media/QuickStart.sos.html.png" alt="QuickStart.sos.html" class="img-responsive" /></p>

<p>It is also worth noting that you are a VIM user, it is highly recommended that you install the SoS syntax file so that your SoS scripts can be properly highlighted during editing. This is not only visually appealing but is very helpful in identifying syntax errors as you write. The same script in vim with a dark background would look look like</p>

<p><img src="/SOS/media/QuickStart.sos.vim.png" alt="QuickStart.sos.vim" class="img-responsive" /></p>

<h2 id="subcommand-config">subcommand <code>config</code></h2>

<pre><code class="language-python">usage: sos config [-h] [-g] [-c CONFIG_FILE]
                  (--get [OPTION [OPTION ...]] | --unset OPTION [OPTION ...] | --set KEY VALUE [KEY VALUE ...])
                  [-v {0,1,2,3,4}]

The config command displays, set, and unset configuration variables defined in
global or local configuration files.

optional arguments:
  -h, --help            show this help message and exit
  -g, --global          If set, change global (~/.sos/config.yaml) instead of
                        local (.sos/config.yaml) configuration
  -c CONFIG_FILE, --config CONFIG_FILE
                        User specified configuration file in YAML format. This
                        file will not be automatically loaded by SoS but can
                        be specified using option `-c`
  --get [OPTION [OPTION ...]]
                        Display values of specified configuration. The
                        arguments of this option can be a single configuration
                        option or a list of option. Wildcard characters are
                        allowed to match more options (e.g. '*timeout',
                        quotation is needed to avoid shell expansion). If no
                        option is given, all options will be outputted.
  --unset OPTION [OPTION ...]
                        Unset (remove) settings for specified options. The
                        arguments of this option can be a single configuration
                        option or a list of option. Wildcard characters are
                        allowed to match more options (e.g. '*timeout', or '*'
                        for all options, quotation is needed to avoid shell
                        expansion).
  --set KEY VALUE [KEY VALUE ...]
                        --set KEY VALUE sets VALUE to variable KEY. The value
                        can be any valid python expression (e.g. 5 for integer
                        5 and '{"c": 2, "d": 1}' for a dictionary) with
                        invalid expression (e.g. val without quote) considered
                        as string. Syntax 'A.B=v' can be used to add {'B': v}
                        to dictionary 'A', and --set KEY VALUE1 VALUE2 ...
                        will create a list with multiple values.
  -v {0,1,2,3,4}, --verbosity {0,1,2,3,4}
                        Output error (0), warning (1), info (2), debug (3) and
                        trace (4) information to standard output (default to
                        2).

</code></pre>

<p>This subcommand can read/write three configuration files</p>

<ol>
  <li><strong><code>~/.sos/config.yaml</code></strong> A global SoS configuration file. (option <code>--global</code>)</li>
  <li><strong><code>.sos/config.yaml</code></strong> Project specific SoS configuration file. (default)</li>
  <li><strong>User specified</strong> Any configuration file specified with option <code>-c</code> (<code>--config</code>).</li>
</ol>

<p>The global and local configuration files will be loaded by default. Other configuration files can be specified using option <code>-c</code> of the <code>sos run</code> command. Options defined in these configuration file will be available to the script as dictionary <code>CONFIG</code>.</p>

<p>Note that you can set complex datatypes from command line using Python expressions,</p>

<pre><code class="language-python">sos config --set a '{"b":1}'
</code></pre>

<p>although</p>

<pre><code class="language-python">sos config --set a.b 1
</code></pre>

<p>is easier to write if you only need to set one key <code>b</code> in dictionary <code>a</code>.</p>

<h2 id="subcommand-remove">subcommand <code>remove</code></h2>

<pre><code class="language-python">$ sos remove -h
usage: sos remove [-h] [-t | -u] [-n] [-y] [FILE_OR_DIR [FILE_OR_DIR ...]]

Remove specified files and directories and their signatures (if available).
Optionally, you can remove only tracked files (input, output and intermediate
files of executed workflows) or untracked file from specified files and/or
directories.

positional arguments:
  FILE_OR_DIR  Files and directories to be removed, which should be under the
               current directory (default). All, tracked, or untracked files
               will be removed depending on other options ('-t' or '-u'). For
               safety reasons, files under the current directory have to be
               listed (not as files under .) to be removed.

optional arguments:
  -h, --help   show this help message and exit
  -t           Remove tracked files and their signatures from specified files
               and directories.
  -u           Remove untracked files from specified files and directories.
  -n           List files or directories to be removed, without actually
               removing them.
  -y           Remove files without confirmation, suitable for batch removal
               of files.
</code></pre>

<h2 id="subcommand-pack">subcommand <code>pack</code></h2>

<pre><code class="language-python">$ sos pack -h
usage: sos pack [-h] [-o OUTPUT] [-i [INCLUDE [INCLUDE ...]]]
                [-e [EXCLUDE [EXCLUDE ...]]] [-a] [-m MESSAGE] [-y]
                [-v {0,1,2,3,4}]
                [session]

positional arguments:
  session               ID of the session to be saved, which can be any number
                        of digits as long as it can uniquely determine a
                        workflow session. This parameter can be ignored if
                        only one session is available.

optional arguments:
  -h, --help            show this help message and exit
  -o OUTPUT, --output OUTPUT
                        Output file, which can be a file with extension ".sar"
                        (the extension will be be automatically appended if
                        needed), or "-" for standard output (default).
  -i [INCLUDE [INCLUDE ...]], --include [INCLUDE [INCLUDE ...]]
                        Additional files or directories to be incldued in the
                        archive. SoS will archive all files under specified
                        directories, including hidden directories such as
                        ".git". Option --exclude could be used to exclude
                        these files.
  -e [EXCLUDE [EXCLUDE ...]], --exclude [EXCLUDE [EXCLUDE ...]]
                        Files that should be excluded from archive. The
                        parameter should be one or more patterns that match
                        the whole path (e.g. "output/*.log" or file or
                        directory names such as "tmp" or "*.bam".
  -a, --all             Include all tracked files even if they reside outside
                        of the current working directory.
  -m MESSAGE, --message MESSAGE
                        A short message to be included into the archive.
                        Because the message would be lost during unpacking, it
                        is highly recommended that you create a README file
                        and include it with option --include.
  -y, --yes             Overwrite output file if it already exists
  -v {0,1,2,3,4}, --verbosity {0,1,2,3,4}
                        Output error (0), warning (1), info (2), debug (3) and
                        trace (4) information to standard output (default to
                        2).
</code></pre>

<h2 id="subcommand-unpack">subcommand <code>unpack</code></h2>

<pre><code class="language-python">$ sos unpack -h
usage: sos unpack [-h] [-d DEST] [-l] [-e] [-n] [-y] [-v {0,1,2,3,4}]
                  archive [files [files ...]]

positional arguments:
  archive               SoS archive saved by command sos pack
  files                 An optional list of files to be processed, which can
                        be exact filenames or patterns (e.g. "*.bam")

optional arguments:
  -h, --help            show this help message and exit
  -d DEST, --dest DEST  Directory where a sos archive would be unpacked.
                        Default to current directory.
  -l, --list            List content of the archive instead of extracting it.
                        The names, uncompressed file sizes and modification
                        dates and times of the specified files are printed,
                        along with totals for all files specified.
  -e, --external        Extract files outside of the project to their external
                        destinations. This option can be dangerous because it
                        can overwrite system files silently if accompanied
                        with option -y.
  -n, --no              Do not overwrite existing files without promoting
                        users.
  -y, --yes             Overwrite existing files without promoting users. This
                        option can be dangerous to use. Note that SoS checks
                        file signature and ignores existing files that are
                        identical to those in the archive.
  -v {0,1,2,3,4}, --verbosity {0,1,2,3,4}
                        Output error (0), warning (1), info (2), debug (3) and
                        trace (4) information to standard output (default to
                        2).
</code></pre>

<h2 id="command-sos-runner">Command <code>sos-runner</code></h2>

<p>Command <code>sos-runner</code> is a shortcut for <code>sos run</code> so</p>

<pre><code class="language-python">sos-runner script
</code></pre>

<p>is equivalent to</p>

<pre><code class="language-python">sos run script
</code></pre>

<p>This allows a SoS script to be executed directly if it is executable with shebang line</p>

<pre><code class="language-python">#!/usr/bin/env sos-runner
</code></pre>

                            </div>
                            <button type="button" class="btn btn-default" data-dismiss="modal"><i class="fa fa-times"></i> Close</button>
                        </div>
                    </div> 
                    <div class="footer-wrapper">
                      <footer role="contentinfo">
                        <span>&copy; 2016 Bo Peng.</span>
                      </footer>
                    </div>
                </div>
            </div>
        </div>
    </div> 


   <div class="portfolio-modal modal fade" id="SoS-Syntax" tabindex="-1" role="dialog" aria-hidden="true">
        <div class="modal-content">
            <div class="close-modal" data-dismiss="modal">
                <div class="lr">
                    <div class="rl">
                    </div>
                </div>
            </div>
            <div class="container">
                <div class="row">
                 <div class="col-lg-8 col-lg-offset-2">
                        <h1>2.SoS-Syntax</h1>
                        <div class="modal-body">
                         
                            <hr class="star-primary">
                            <p>SoS uses <a href="http://www.python.org">Python</a> expressions and statements. If you are unfamiliar with Python, you can learn some basics of Python, usually in less than half a day, by reading some Python tutorials (e.g. <a href="https://docs.python.org/3/tutorial/">the official python tutorial</a>). This <a href="https://docs.python.org/3/tutorial/introduction.html">short introduction</a> is good enough for you to use SoS.</p>

<p>The only difference between SoS python syntax and standard Python syntax is that <strong>string literals are format string that will be <a href="#string-interpolation">interpolated</a> using SoS defined sigil</strong>.</p>

<p>and introduces the following SoS-specific syntax:</p>

<ul>
  <li><code>[names: options]</code> line to define steps of workflows</li>
  <li>special directives such as <code>input:</code></li>
  <li>A special script style of function calls for SoS actions</li>
  <li>structural preprocessors such as <code>%include</code>, <code>%if</code>, <code>%cell</code></li>
</ul>

<h2 id="string-interpolation">String interpolation</h2>

<p>On top of python string manipulation functions and similar to recently introduced (Python 3.6) format string, SoS uses string interpolation to replace variables and expressions within a string with their values. For example, expressions <code>resource_path</code>, <code>sample_names[0]</code> and <code>sample_names</code> would be replaced by their values in the following string definitions.</p>

<pre><code class="language-python">resource_path = '~/.sos/resources'
ref_genome    = '${resource_path}/hg19/refGenome.fasta'     # '~/.sos/resources/hg19/refGenome.fasta'

sample_names  = ['A', 'B', 'C']
title         = 'Sample ${sample_names[0]} results'         # 'Sample A results'
all_names     = 'Samples ${sample_names}'                   # 'Samples A B C'
</code></pre>

<p>Depending on the return type of the expression</p>

<ul>
  <li>String representations (<code>repr(obj)</code>) are returned for objects in simple Python types (e.g. string, True, False, None, numeric numbers)</li>
  <li>
    <p>For objects with an iterator interface (e.g. Python <code>list</code>, <code>tuple</code>, <code>dict</code>, and <code>set</code>), SoS join the string representation of each item by a space or comma as specified by conversion flag <code>!,</code> (see <a href="#conversion-and-format">conversion and format</a> for details). More specifically,</p>

    <ul>
      <li>List of strings will be converted to a string by joining strings with a space or comma.</li>
      <li>Dictionary of strings will be converted to a string by joining dictionary keys with no guarantee on the order of values.</li>
    </ul>
  </li>
</ul>

<p>If you are unhappy with the string conversion, you can always format your object using python expressions such as <code>${repr(obj)}</code> and <code>${', '.join(x for x in obj)}</code>.</p>

<p>SoS supports nested interpolation, for example</p>

<pre><code class="language-python">'${_input[${index}]}'
</code></pre>

<p>would evaluate <code>${index}</code> and then <code>${_input[?]}</code> where <code>?</code> is the result of <code>${index}</code>.</p>

<p>Note that you can continue to use Python string functions such as</p>

<pre><code class="language-python">ref_genome = resource_path + '/hg19/refGenome.fasta'
title      = 'Result for the first sample {} is invalid'.format(sample_names[0])
</code></pre>

<p>but string interpolation is recommended for multi-line scripts because it is easier to read.</p>

<p>SoS interpolation also support all string format and conversion specification as in the <a href="https://docs.python.org/2/library/string.html#formatspec">Python string format specifier</a>, that is to say, you can use <code>: specifier</code> at the end of the expression to control the format of the output. For example</p>

<ul>
  <li><code>${1/3. :.2f}</code> in a string literal yields <code>0.33</code> instead of its long representation <code>0.3333333333333333</code>, and</li>
  <li><code>${filename!r}</code> in a string literal yields <code>'Article by "Jane Eyre"'</code> (note the quotation mark) if <code>filename='Article by "Jane Eyre"'</code>.</li>
</ul>

<p>The latter is interesting because SoS uses correct quotation marks for filenames with quotation marks, making</p>

<pre><code class="language-python">R('''
read.csv(${_input!r})
''')
</code></pre>

<p>a safer choice than</p>

<pre><code class="language-python">R('''
read.csv("${_input}")
''')
</code></pre>

<p>because <code>${_input}</code> might contain quotation marks.</p>

<p>SoS also supports a few other conversion operators, for</p>

<pre><code class="language-python">file1='file 1.txt'
file2='~/SoS/test.sos'
files=['a.txt', 'b.txt']
</code></pre>

<table>
  <thead>
    <tr>
      <th style="text-align: left">convertor</th>
      <th style="text-align: left">usage</th>
      <th style="text-align: left">input</th>
      <th style="text-align: left">output</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td style="text-align: left"><code>s</code></td>
      <td style="text-align: left"><code>str()</code></td>
      <td style="text-align: left"><code>${file1!s}</code></td>
      <td style="text-align: left"><code>file 1.txt</code></td>
    </tr>
    <tr>
      <td style="text-align: left"><code>r</code></td>
      <td style="text-align: left"><code>repr()</code></td>
      <td style="text-align: left"><code>${file1!r}</code></td>
      <td style="text-align: left"><code>'file 1.txt'</code></td>
    </tr>
    <tr>
      <td style="text-align: left"><code>q</code></td>
      <td style="text-align: left"><code>quoted()</code></td>
      <td style="text-align: left"><code>${file1!q}</code></td>
      <td style="text-align: left"><code>file\ 1.txt</code></td>
    </tr>
    <tr>
      <td style="text-align: left"><code>e</code></td>
      <td style="text-align: left"><code>expanduser()</code></td>
      <td style="text-align: left"><code>${file2!e}</code></td>
      <td style="text-align: left"><code>/path/to/user/SoS/test.sos</code></td>
    </tr>
    <tr>
      <td style="text-align: left"><code>a</code></td>
      <td style="text-align: left"><code>abspath(expanduser())</code></td>
      <td style="text-align: left"><code>${file2!a}</code></td>
      <td style="text-align: left"><code>/path/to/user/SoS/test.sos</code></td>
    </tr>
    <tr>
      <td style="text-align: left"><code>b</code></td>
      <td style="text-align: left"><code>basename())</code></td>
      <td style="text-align: left"><code>${file2!b}</code></td>
      <td style="text-align: left"><code>test.sos</code></td>
    </tr>
    <tr>
      <td style="text-align: left"><code>d</code></td>
      <td style="text-align: left"><code>dirname())</code></td>
      <td style="text-align: left"><code>${file2!d}</code></td>
      <td style="text-align: left"><code>/path/to/user/SoS/</code></td>
    </tr>
    <tr>
      <td style="text-align: left"><code>,</code></td>
      <td style="text-align: left"><code>', '.join()</code></td>
      <td style="text-align: left"><code>${files!,}</code></td>
      <td style="text-align: left"><code>a.txt, b.txt</code></td>
    </tr>
  </tbody>
</table>

<p>For example, the following command</p>

<pre><code class="language-python">run('cat ${_input!q}`)
</code></pre>

<p>would produce the correct command and execute</p>

<pre><code class="language-python">cat 'Bon Jovi.txt'
</code></pre>

<p>instead of</p>

<pre><code class="language-python">cat Bon Jovi.txt
</code></pre>

<p>for <code>_input=['Bon Jovi.txt']</code>. The conversion flags can also be combined, as shown in</p>

<pre><code class="language-python">files   = ['AB.txt', 'C D.txt']
fname   = '${files!q,}'            # AB.txt,C\D.txt or AB.txt,'C D.txt'
</code></pre>

<p>If the default sigil conflicts with the script used, an alternative sigil can be used. Please see section <a href="#option-sigil">option <code>sigil</code></a> for details.</p>

<h2 id="section-header-for-workflow-steps">Section header for workflow steps</h2>

<p>SoS defines workflows that consists of multiple steps. A step is marked by a section head in the format of</p>

<pre><code class="language-python">[names: options]
</code></pre>

<p>The header should start with a <code>[</code> from the beginning of a line and end with a <code>]</code>. It can contain one or more names with optional section options.</p>

<h2 id="sos-directives">SoS directives</h2>

<p>SoS introduces a number of directives to specify elements of workflows, including</p>

<ul>
  <li><code>parameter:</code> to define command line options</li>
  <li><code>input:</code>, <code>output:</code>, <code>depends:</code> to define input, output, and dependent files of workflow steps, and</li>
  <li><code>task:</code> to define (long) tasks that will be executed as separate tasks</li>
</ul>

<h2 id="script-style-function">Script style function</h2>

<p>SoS allows you to write SoS <code>action</code> (basically a Python function) in a special script format. For example,</p>

<pre><code class="language-python">R('''
pdf('${input}')
plot(0, 0)
dev.off()
''', workdir='result')
</code></pre>

<p>can be written as</p>

<pre><code class="language-python">R:     workdir='result'
pdf('${_input}')
plot(0, 0)
dev.off()
</code></pre>

<p><strong>The script is a string without quotation marks</strong> and the normal string interpolation will take place. You can also indent the script (add leading white spaces to all lines) and write the action as</p>

<pre><code class="language-python">R:  workdir='result'
   pdf('${_input}')
   plot(0, 0)
   dev.off()
</code></pre>

<p>The latter is much preferred because it avoids trouble if your script contains strings such as <code>[1]</code> and <code>option:</code> (and be treated as SoS directives), and more importantly, allows starting a new statement from a non-indented line. For example, <code>check_command('dot')</code> would be considered part of a R script in</p>

<pre><code class="language-python">R:  workdir='result'
pdf('${_input}')
plot(0, 0)
dev.off()

check_command('dot')
</code></pre>

<p>but a separate action in</p>

<pre><code class="language-python">R:  workdir='result'
   pdf('${_input}')
   plot(0, 0)
   dev.off()

check_command('dot')
</code></pre>

<p>Although the script format is more concise and easier to read, it is limited to actions that accept a string as its first parameter and cannot return value or be used within <code>try/except</code> of <code>if/else</code> statements.</p>

<p>One final difference between SoS and regular Python 3 syntax is that SoS is more lenient on the use of mixed tab and spaces for indentation. Although it is highly recommended that you use all spaces for indentation, SoS will give an warning and treat tabs as 4 spaces during execution.</p>

                            </div>
                            <button type="button" class="btn btn-default" data-dismiss="modal"><i class="fa fa-times"></i> Close</button>
                        </div>
                    </div> 
                    <div class="footer-wrapper">
                      <footer role="contentinfo">
                        <span>&copy; 2016 Bo Peng.</span>
                      </footer>
                    </div>
                </div>
            </div>
        </div>
    </div> 


   <div class="portfolio-modal modal fade" id="File-Structure" tabindex="-1" role="dialog" aria-hidden="true">
        <div class="modal-content">
            <div class="close-modal" data-dismiss="modal">
                <div class="lr">
                    <div class="rl">
                    </div>
                </div>
            </div>
            <div class="container">
                <div class="row">
                 <div class="col-lg-8 col-lg-offset-2">
                        <h1>3.File-Structure</h1>
                        <div class="modal-body">
                         
                            <hr class="star-primary">
                            <p>A SoS script usually starts with lines</p>

<pre><code class="language-python">#!/usr/bin/env sos-runner
#fileformat=SOS1.0
</code></pre>

<p>The first line allows the script to be executed by command <code>sos-runner</code> if it is executed as an executable script. The second line tells SoS the format of the script. The <code>#fileformat</code> line does not have to be the first or second line but should be in the first comment block. SOS format 1.0 is assumed if no format line is present.</p>

<h2 id="global-definitions">Global definitions</h2>

<p>Python functions, classes, variables can be defined or imported (using Python <code>import</code> statement) before any SoS step is defined. These definitions usually contains variables such as version and date of the script, paths to various resources, and utility functions that will be used by later steps. <strong>These definitions are visible to all steps of workflows and are assumed to be readonly</strong>.</p>

<p>In addition to user-defined global variables, SoS defines the following variables before any variables are defined</p>

<ul>
  <li><strong><code>SOS_VERSION</code></strong>: version of SoS command.</li>
  <li><strong><code>CONFIG</code></strong>: A dictionary of configurations specified by the global sos configuration file (<code>~/.sos/config.yaml</code>), local configuration file (<code>~/.sos/config.yaml</code>) and command line option <code>-c config_file</code>.</li>
  <li><strong><code>run_mode</code></strong>: A string indicating the run mode in which the script is executed, which can be <code>inspect</code>, <code>prepare</code> or <code>run</code>. Because the script and therefore all Python statements in the script will be executed in all three modes, this variable allows the execution of some statement only in selected modes.</li>
</ul>

<h2 id="include-preprocessor"><code>%include</code> preprocessor</h2>

<p>SoS allows you to include variables and steps of another script into the current script using preprocessor <code>%include</code>. The <code>%include</code> statement should be used before any other SoS statements and follows the same syntax as the python <code>import</code> keyword. For example, the following statements are allowed</p>

<pre><code class="language-python">%include alignment
%include alignment as ag
%include alignment, calling
%include alignment as ag, calling as ca

%from alignment include *
%from alignment include var1, workflow1
%from alignment include workflow1 as wf
</code></pre>

<p>Similar to python <code>import</code> statement, variables and workflows included using the <code>%include ... as ...</code> syntax can be accessed using <code>module.name</code> (e.g. <code>alignment.var1</code>, <code>ag.var1</code> (for <code>include alignment as ag</code>)), whereas variables and workflows included using the <code>from ... include ...</code> syntax are available directly (e.g. <code>var1</code>, <code>workflow1</code>, and <code>wf</code> (for <code>workflow1 as wf</code>).</p>

<p>For example</p>

<pre><code class="language-python">#!/usr/bin/env sos-runner
#fileformat=SOS1.0

%include alignment
%include call
[10]
sos_run('alignment.default + call.default')
</code></pre>

<p>would execute two workflows defined in <code>alignment.sos</code> and <code>call.sos</code>. The files should be in SoS search path to be included.</p>

<h2 id="if--elif-else--endif"><code>%if .. %elif ..%else .. %endif</code></h2>

<p>SoS allows you to conditionally include part of the script using macos <code>%if</code>, <code>%elif</code>, <code>%else</code>, and <code>%endif</code>. The condition in <code>%if cond</code> and <code>elif cond</code> should be valid python expression. The condition should not expand multiple lines and should not have trailing <code>:</code>. The conditions are evaluated at parsing time so no SoS variables are allowed. For example, you can write</p>

<pre><code class="language-python">%if sys.platform == 'darwin'

[step]
Mac OSX step

%else

[step]
Linux step

%endif
</code></pre>


                            </div>
                            <button type="button" class="btn btn-default" data-dismiss="modal"><i class="fa fa-times"></i> Close</button>
                        </div>
                    </div> 
                    <div class="footer-wrapper">
                      <footer role="contentinfo">
                        <span>&copy; 2016 Bo Peng.</span>
                      </footer>
                    </div>
                </div>
            </div>
        </div>
    </div> 


   <div class="portfolio-modal modal fade" id="Command-line-options-and-configuration-file" tabindex="-1" role="dialog" aria-hidden="true">
        <div class="modal-content">
            <div class="close-modal" data-dismiss="modal">
                <div class="lr">
                    <div class="rl">
                    </div>
                </div>
            </div>
            <div class="container">
                <div class="row">
                 <div class="col-lg-8 col-lg-offset-2">
                        <h1>4.Command-line-options-and-configuration-file</h1>
                        <div class="modal-body">
                         
                            <hr class="star-primary">
                            <h2 id="command-line-options">Command line options</h2>

<h3 id="optional-arguments">Optional arguments</h3>

<p>Any variable assignment prefixed with <code>parameter:</code> can accept values from command  line. The format of such lines are</p>

<pre><code class="language-python"># comment
parameter: var_name = default_value
</code></pre>

<p>The default value can be number, string, list of string, or expressions that return values of these types. Other types can be used as long as they can be converted to these types from user-provided values. For example</p>

<pre><code class="language-python"># path to tool gatk
parameter: gatk_path = '~/bin/GATK'
</code></pre>

<p>defines a variable <code>gatk_path</code> with default value <code>'~/bin/GATK'</code>.</p>

<pre><code class="language-python"># A list of sample names
parameter: sample_names=[]
</code></pre>

<p>defines a variable <code>sample_name</code> with default value <code>[]</code>, and</p>

<pre><code class="language-python"># path to gatk
parameter: gatk_path = CONFIG['gatk_path']
</code></pre>
<p>uses <code>gatk_path</code> from a YAML-based configuration file (specified from command line using option <code>-c</code>) as default value. You can set default value for <code>CONFIG</code> inside SoS script via:</p>

<pre><code class="language-python"># path to gatk
parameter: gatk_path = CONFIG.get('gatk_path', '/default/path/to/gatk')
</code></pre>

<p>so that if configuration files are not supplied the default value <code>/default/path/to/gatk</code> will be used.</p>

<p>The default values not only determines the values of variable when they are not specified from command line or configuration files, but also determines the type of input these parameters accept. For example, with the above definitions for command arguments <code>--gatk_path</code> and <code>--sample_names</code>, you can pass values to these variables from command line,</p>

<pre><code class="language-python">sos run myscript --gatk_path /path/to/gatk --sample_names A1 A2 A3
</code></pre>

<p>A list will be passed to <code>sample_names</code> even if only a single value is provided (e.g. <code>sample_names=['A1']</code> for <code>--sample_name A1</code>).
Attempts to pass more than one values (a list) to <code>gatk_path</code> (e.g. <code>--gatk_path /path1 /path2</code>) will trigger an error.</p>

<p>Note that boolean values can be specified from command line as <code>--param</code> for <code>True</code> and <code>--no-param</code> for <code>False</code>. It is also allowed to use <code>--gatk-path</code> in addition to <code>--gatk_path</code> for parameter <code>--gatk_path</code>.</p>

<h3 id="required-arguments">Required arguments</h3>

<p>In cases where there is no suitable default values and/or command line arguments are mandatary, you can list the type of arguments (e.g. <code>int</code>, <code>bool</code>, <code>str</code>, <code>list</code> of strings) in place of default values. For example, if an integer parameter <code>cutoff</code> is required, you can define it as</p>

<pre><code class="language-python"># cutoff value
parameter: cutoff = int
</code></pre>

<p>This will force the users to provide an integer to this parameter. You can do the same for lists but SoS assumes that you need a <strong>list of strings</strong>. For example, the following definition</p>

<pre><code class="language-python"># input bam files
parameter: bam_files = list
</code></pre>

<p>request a list of strings from command line. SoS will return a list even if only one value is provided.</p>

<h3 id="redefinition-of-parameter">Redefinition of <code>parameter</code></h3>

<p>A very important property of the <code>parameter</code> directive is that <strong>parameter definition is only effective for undefined variable</strong>. That is to say, the parameter statement will be ignored if a parameter has been specified in anyway, as shown in the following example:</p>

<pre><code>a = 10
parameter: a = int
</code></pre>

<p>An useful implication of this rule is that you can define a workflow and allow some of its parameters to be specified from command line and from other workflows. For example, the following workflow</p>

<pre><code class="language-python">[default_1]
parameter: cutoff=10
sh:
  do something with ${cutoff}

[default_2]
sh:
  do something else

[batch]
for cutoff in range(10):
    sos_run('default')
</code></pre>

<p>would be called in a number of ways</p>

<pre><code class="language-python">sos run myscript 
sos run myscript --cutoff 5
sos run myscript batch
</code></pre>

<p>The first command uses default value <code>10</code> for parameter <code>cutoff</code>, the second command assigns value <code>5</code> from command line, and the third command uses a series of values <code>0</code>, <code>1</code>, . Because <code>cutoff</code> is already defined for workflow <code>default</code>, the parameter statement will be ignored.</p>

<h2 id="configuration-files">Configuration files</h2>

<p>The configuration files should be in the format of <a href="http://yaml.org/"><code>YAML</code></a> or its subset format <a href="http://json-schema.org/implementations.html"><code>JSON</code></a>. Inside SoS script you can access these variables via, for example, either <code>CONFIG['gatk_path']</code> or <code>CONFIG.gatk_path</code>.</p>

<p>The <code>-c</code> option allows the specification of a configuration file in YAML format. YAML is a superset of JSON so any configuration file in JSON format should also be acceptable. Variables defined in the configuration file are available in SoS script as a dictionary <code>CONFIG</code>. For example</p>

<pre><code class="language-python">gatk_path = CONFIG['gatk_path']
</code></pre>

<p>or equivalently</p>

<pre><code class="language-python">gatk_path = CONFIG.gatk_path
</code></pre>

<p>requires a configuration file with <code>gatk_path</code> defined. If you do not want to require a configuration file, you can define <code>gatk_path</code> as</p>

<pre><code class="language-python">gatk_path = CONFIG.get('gatk_path', '/path/to/default/gatk')
</code></pre>

<p>In this way, a default path would be used if no configuration file is specified (so <code>CONFIG</code> is an empty dictionary) or if <code>gatk_path</code> is not defined in the specified configuration file.</p>

<p>If you would further want to allow modification of this value from command line, you can place this definition in the <code>[parameters]</code> section</p>

<pre><code class="language-python">[parameters]
# path to gatk executable
gatk_path = CONFIG.get('gatk_path', '/path/to/default/gatk')
</code></pre>

<p>In this way, users have the freedom to use the default value, define a value in a configuration file, and provide another value from command line.</p>

                            </div>
                            <button type="button" class="btn btn-default" data-dismiss="modal"><i class="fa fa-times"></i> Close</button>
                        </div>
                    </div> 
                    <div class="footer-wrapper">
                      <footer role="contentinfo">
                        <span>&copy; 2016 Bo Peng.</span>
                      </footer>
                    </div>
                </div>
            </div>
        </div>
    </div> 


   <div class="portfolio-modal modal fade" id="Workflow-Definition" tabindex="-1" role="dialog" aria-hidden="true">
        <div class="modal-content">
            <div class="close-modal" data-dismiss="modal">
                <div class="lr">
                    <div class="rl">
                    </div>
                </div>
            </div>
            <div class="container">
                <div class="row">
                 <div class="col-lg-8 col-lg-offset-2">
                        <h1>5.Workflow-Definition</h1>
                        <div class="modal-body">
                         
                            <hr class="star-primary">
                            <p>A SoS script can specify one or more workflows. Each workflow consists of one or more numbered steps. The numbers (should be non-negative) specify the <strong>logical order</strong> by which the steps are executed, but a later step might be executed before the completion of previous steps if it does not depend on the output of these steps.</p>

<h2 id="single-workflow">Single workflow</h2>

<p>A single workflow can be specified without a name in a SoS script. For example, the following sections specify a workflow with four steps <code>5</code>, <code>10</code>, <code>20</code>, and <code>100</code>.
As you can see, the workflow steps can be specified in any order and do not have to be consecutive (which is actually preferred because it allows easy insertion of extra steps).</p>

<pre><code class="language-python">[5]
[20]
[10]
[100]
</code></pre>

<p>Workflows specified in this way is the <code>default</code> workflow and are actually called <code>default</code> in SoS output. If you want to give it a meaningful name, you can specify the steps as</p>

<pre><code class="language-python">[mapping_5]
[mapping_20]
[mapping_10]
[mapping_100]
</code></pre>

<p>Because this SoS script defines only one workflow (<code>mapping</code>), you do not have to specify the name of workflow from SoS command</p>

<pre><code class="language-python">sos run myscript.sos --input input1.fasta
</code></pre>

<p>Unnumbered workflow steps are assumed to be the first steps (with index <code>0</code>) of a workflow unless they have other meanings (e.g. <a href="auxiliary-workflow-steps-and-makefile-style-dependency-rules">auxiliary step</a>).</p>

<h2 id="multiple-workflows">Multiple workflows</h2>

<p>A SoS script can define multiple workflows. For example, the following sections of SoS script defines two workflows named <code>mouse</code> and <code>human</code>.</p>

<pre><code>[mouse_10]
[mouse_20]
[mouse_30]
[human_10]
[human_20]
[human_30]
</code></pre>

<p>You will have to specify which workflow to execute from the command line, e.g.</p>

<pre><code class="language-python">sos run myscript mouse --input input1.fasta
</code></pre>

<p>If you would like to define a <code>default</code> and a named workflow, you can define them as</p>

<pre><code class="language-python">[10]
[20]
[30]
[test_10]
[test_20]
[test_30]
</code></pre>

<p>The <code>default</code> workflow will be executed by default using command</p>

<pre><code class="language-python">sos run myscript.sos
</code></pre>

<p>The <code>test</code> workflow will be executed if its name is specified from the command line</p>

<pre><code class="language-python">sos run myscript.sos test
</code></pre>

<h2 id="shared-workflow-steps">Shared workflow steps</h2>

<p>The most common motivation of defining multiple workflows in a single SoS script is that they share certain processing steps. If this is the case, you can define sections such as</p>

<pre><code class="language-python">[mouse_10,human_10]
[mouse_20]
[human_20]
[mouse_30,human_30]
</code></pre>

<p>or</p>

<pre><code class="language-python">[*_10]
[mouse_20]
[human_20]
[*_30]
</code></pre>

<p>or</p>

<pre><code class="language-python">[*_10]
[mouse_20,human_20]
[fly_20]
[*_30,fly_50]
[fly_40]
</code></pre>

<p>In the last case, step defined by <code>[*_30,fly_40]</code> will be expanded to <code>mouse_30</code>, <code>human_30</code>, <code>fly_30</code>, and <code>fly_50</code> and will be executed twice for the <code>fly</code> workflow. Note that workflow steps can use variable <code>step_name</code> to act (slightly) differently for different workflows. For example,</p>

<pre><code class="language-python">[mouse_20,human_20]
reference = mouse_reference if step_name.startswith('mouse') else human_reference

input: fasta_files
depends: reference

</code></pre>

<p>Here the variable <code>step_name</code> is <code>mouse_20</code> or <code>human_20</code> depending on the workflow being executed, and expression <code>mouse_reference if step_name.startswith('mouse') else human_reference</code> returns <code>mouse_reference</code> if the workflow <code>mouse</code> is executed, and <code>human_reference</code> otherwise.</p>

<h2 id="subworkflow">Subworkflow</h2>

<p>Although workflows are defined separately with all their steps, they do not have to be executed in their entirety. A <code>subworkflow</code> refers to a workflow that is defined from one or more steps of an existing workflows. It is specified using syntax <code>workflow[_steps]</code> where step can be <code>n</code> (step <code>n</code>), <code>:n</code> (up to <code>n</code>), <code>n:m</code> (step <code>n</code> to <code>m</code>) and <code>m:</code> (from <code>m</code>). For example</p>

<pre><code class="language-python">A              # complete workflow A
A:5-10         # step 5 to 10 of A
A:50-          # step 50 up
A:-10          # up to step 10 of A
A:10           # step 10 of workflow A can be considered a subworkflow
</code></pre>

<h2 id="combined-workflow">Combined workflow</h2>

<p>You can also combine subworkflows to execute multiple workflows one after another. For example,</p>

<pre><code class="language-python">A + B          # workflow A, followed by B
A_0 + B        # step 0 of A, followed by B
A:-50 + B + C  # up to step 50 of workflow A, followed by B, and C
</code></pre>

<p>This syntax can be used from the command line (option <code>workflow</code>, e.g. <code>sos-runner myscript.sos align+call</code>) or used to execute <a href="#nested-workflow">nested workflows</a> inside the SoS script.</p>

<p>It is worth noting that combined workflow might work differently from when they are executed individually (e.g. default input of <code>B</code> is changed from empty to output of <code>A_0</code>), and it is up to the user to resolve conflicts between them.</p>

<h2 id="nested-workflow">Nested workflow</h2>

<p>SoS also supports nested workflow in which a complete workflow is treated as part of a step process.
The workflow is execute by SoS action <code>sos_run</code>, e.g.</p>

<pre><code class="language-python">sos_run('A')                       # execute workflow A
sos_run('A + B')                   # execute workflow B after A
sos_run('D:-10 + C')               # execute up to step 10 of D and workflow C
sos_run('${aligner} + ${caller}')  # execute user-specified aligner and caller workflows
</code></pre>

<p>In its simplest form, nested workflow allows you to define another workflow from existing ones. For example,</p>

<pre><code class="language-python">[default]
sos_run('align+call')
</code></pre>

<p>defines a nested workflow that combines workflows <code>align</code> and <code>call</code> so that the workflow will by default execute two workflows, but can also execute one of them as separate workflows <code>align</code> and <code>call</code>.</p>

<p>Nested workflow also allows you to define multiple mini-workflows and connect them freely. For example</p>

<pre><code class="language-python">[a_1]
[a_2]
[b]
[c]
[d_1]
sos_run('a+b')
[d_2]
sos_run('a+c')
</code></pre>

<p>defines workflows <code>d</code> that will execute steps <code>d_1</code>, <code>a_1</code>, <code>a_2</code>, <code>b_0</code>, <code>d_2</code>,  <code>a_1</code>, <code>a_2</code>, and <code>c_0</code>.</p>

<p>Nested workflows, like other SoS actions, can be executed repeatedly, for example,</p>

<pre><code class="language-python">[b_1]
[b_2]
[b_3]

[a]
parameters = range(20)
input: 'some.txt', for_each='parameters'
output: '${input}_${_parameters}.res'
sos_run('b')
</code></pre>

<p>would execute the complete workflow <code>b</code> 20 times each with a different parameter. Similarly you can let the nested workflow process groups of input files.</p>

<p>Nested workflows can also be used to compose workflows from user-provided options through command line arguments, configuration files, and even results from previous steps. For example, the following example</p>

<pre><code class="language-python"># aligner steps to use to align the reads 
parameter: aligner = CONFIG.get('aligner', 'bwa')

[bwa_1]
[bwa_2]
[novaalign_1]
[novaalign_2]

[align]
sos_run(aligner)
</code></pre>

<p>defines workflows <code>bwa</code> and <code>novaalign</code> to align raw reads. The <code>align</code> workflow is a master workflow that executes <code>bwa</code> or <code>novaalign</code> determined by option <code>aligner</code> defined in a configuration file (command line option <code>-c</code>) and command line option <code>--aligner</code>.</p>


                            </div>
                            <button type="button" class="btn btn-default" data-dismiss="modal"><i class="fa fa-times"></i> Close</button>
                        </div>
                    </div> 
                    <div class="footer-wrapper">
                      <footer role="contentinfo">
                        <span>&copy; 2016 Bo Peng.</span>
                      </footer>
                    </div>
                </div>
            </div>
        </div>
    </div> 


   <div class="portfolio-modal modal fade" id="SoS-Steps" tabindex="-1" role="dialog" aria-hidden="true">
        <div class="modal-content">
            <div class="close-modal" data-dismiss="modal">
                <div class="lr">
                    <div class="rl">
                    </div>
                </div>
            </div>
            <div class="container">
                <div class="row">
                 <div class="col-lg-8 col-lg-offset-2">
                        <h1>6.SoS-Steps</h1>
                        <div class="modal-body">
                         
                            <hr class="star-primary">
                            <p>Although no item is required for a SoS step, a complete SoS step can have the following form</p>

<pre><code class="language-python">[step_name: option1, option2, ...]
#
# description of the step
#

statements
input:    input files, opt1=value1, opt2=value2

statements (executed in SoS)
output:	  output files, opt1=value1, opt2=value2

statements (executed in SoS)
depends:  dependent files, opt1=value1, opt2=value2

statements (executed in SoS)
task:  opt1=value1, opt2=value2 (executed outside of SoS)
statements or script 
</code></pre>

<p>Basically, a step can have <code>input</code>, <code>output</code>, <code>depends</code>, <code>task</code> keywords with their options, with arbitrary Python statements around them.</p>

<h2 id="targets">Targets</h2>

<p>Targets are objects that a SoS step can input, output, or dependent on. They are usually files that are presented by filenames, but can also be other targets.</p>

<h3 id="executable-target"><code>executable</code> target</h3>

<p><code>executable</code> targets are commands that should be accessible and executable by SoS. These targets are usually listed in the <code>depends</code> section of a SoS step. For example, SoS would stop if a command <code>fastqc</code> is not found.</p>

<pre><code class="language-python">[10]
input:     fastq_files
depends:   executable('fastqc')
sh:
    fastqc ${fastq_files}
</code></pre>

<p><code>executable</code> target can also be output of a step but installing executables can be tricky because the commands should be installed to existing <code>$PATH</code> so that they can be immediately accessible by SoS. Because SoS automatically adds <code>~/.sos/bin</code> to <code>$PATH</code> (option <code>-b</code>), an environment-neutral way for on-the-fly installation is to install commands to this directory. For example</p>

<pre><code class="language-python">[10: provides=executable('fastqc')]
sh:
    # download fastqc
    # install fastqc
    # link fastqc to ~/.sos/bin
</code></pre>

<p>would be called if executable <code>fastqc</code> is not found.</p>

<h3 id="sosvariable-target"><code>sos_variable</code> target</h3>

<p><code>sos_variable(name)</code> targets represent SoS variables that are created by a SoS step and shared to other steps. These targets can be used to provide information to other steps. For example,</p>

<pre><code class="language-python">[get_total_reads: shared='total_reads']
input:  'alignment/summary.txt'
with open(input[0]) as ifile:
    ...
    total_reads = int(...)


[100]
depends: sos_variable('total_reads')
output:  'report.txt'

with open(output[0]) as report:
    report.write('total reads: ${total_reads}')

</code></pre>

<p>Step <code>100</code> needed some information extracted from output of another step. You can either parse the information in step <code>100</code> or use another step to provide the information. The latter is recommended because the information could be requested by multiple steps.</p>

<h3 id="envvariable-target"><code>env_variable</code> target</h3>

<p>SoS keeps tract of runtime environment and creates signatures of executed steps so that they do not have to be executed again. Some commands, especially shell scripts, could however behave differently with different environmental variables. To make sure a step would be re-executed with changing environments, you should list these variables as dependencies of the step. For example</p>

<pre><code class="language-python">[10]
depends:   env_variable('DEBUG')
sh:
    echo DEBUG is set to $DEBUG
</code></pre>

<h3 id="dynamic-target"><code>dynamic</code> target</h3>

<p>A <code>dynamic</code> target is a target that can only be determined when the step is actually executed. For example, the input of the following step would not be evaluated until step <code>20</code> is executed.</p>

<pre><code class="language-python">[10]
#
# a step that generate some text with dynamically determined names
#

[20]
input: dynamic('*.txt')

</code></pre>

<p>In this particular case, if you specify</p>

<pre><code class="language-python">[20]
input: '*.txt'

</code></pre>

<p>SoS would find all <code>.txt</code> files at the beginning of the execution of the workflow and feed them to step <code>20</code>.</p>

<h2 id="step-name-and-alias">Step name and alias</h2>

<p>A section can define one or more steps, with step names separated by comma. For example</p>

<pre><code class="language-python">[human_10, mouse_10]
</code></pre>

<p>defines a section with two steps for two different workflows. The step name can be followed by a numeric index, which indicates the relative position of the step in a workflow.</p>

<p>An alias, or description, is allowed for step name. Although the description does not affect the workflow and index of the step, it helps identify the step in the SoS output. For example</p>

<pre><code class="language-python">[human_10 (quality check)]
</code></pre>

<p>define a step with description <code>quality check</code>.</p>

<h2 id="step-options">Step options</h2>

<p><strong>Step options</strong> are specified after step name that assists the specification of workflows. SoS provides the following options</p>

<h3 id="option-skip">Option <code>skip</code></h3>

<p>Option <code>skip</code> takes two formats, the first format has no value</p>

<pre><code class="language-python">[10: skip]
</code></pre>

<p>and is equivalent to</p>

<pre><code class="language-python">[10: skip=True]
</code></pre>

<p>The whole step will be skipped as if it is not defined at all in the script. This option provides a quick method to disable a step.</p>

<p>The second format takes a value, which is usually an expression that will be evaluated when the step is executed. For example, in the following workflow,</p>

<pre><code class="language-python"># tool suite to align the reads
parameter: quality_check = True

[10: skip= not quality_check]
</code></pre>

<p>Step 10 will be skipped if option <code>--quality_check no</code> is specified from command line.</p>

<h3 id="option-sigil">Option <code>sigil</code></h3>

<p>Because a SoS script can include scripts in different languages with different sigils (special characters used to enclose expression), <strong>SoS allows the use of different sigils</strong> in the script. Whereas the default sigil (<code>${ }</code>) has to be used for global variables, other sigils can be used for different sections of the script. For example</p>

<pre><code class="language-python">[step_10: sigil='%( )']
title = 'Sample %(sample_names[0]) results'
run:
	for file in *.fastq
	do
		echo Processing ${file} ...
		echo %(title)
	done
</code></pre>

<p>uses a different sigil style because the embedded shell script uses <code>${ }</code>. In this example <code>${file}</code> keeps its meaning in the shell script while <code>%(sample_names[0])</code> and <code>%(title)</code> are replaced by SoS to their values.</p>

<p>Option <code>sigil</code> accept only a constant string (not a variable or expression).</p>

<h3 id="option-shared">Option <code>shared</code></h3>

<p>SoS executes each step in a separate process and by default does not return any result to the master SoS process. Option <code>shared</code> is used to share variables between steps. This option accepts:</p>

<ul>
  <li>A string (variable name), or</li>
  <li>A map between variable names and expressions (strings) that will be evaluated upon the completion of the step.</li>
  <li>A sequence of strings (variables) or maps.</li>
</ul>

<p>After the completion of a step, local variables defined in the <code>shared</code> option will be sent to the global environment so that later steps can use the updated value. For example,</p>

<pre><code class="language-python">parameter: trim = False
parameter: input_files = list

[10: shared={'input_files':'output'}, skip=trim is False]
input:  input_files, group_by=1
output: '${_input[0]}.trimmed'

sh:
    trim ${input} &gt; ${output}

[20]
input: input_files
# do something
</code></pre>

<p>In this example, <code>trim</code> is an option to trim input files defined in another parameter <code>input_files</code>. If <code>trim</code> is false, step 10 is ignored and step 20 handles <code>input_files</code>. If <code>trim</code> is True, the files are trimmed and produces <code>${_input[0]}.trimmed</code>. Variable <code>input_files</code> are sent back to the global environment so that step 20 would be handling the trimmed version of input files.</p>

<p>A map syntax is recommended to share <code>output</code> of one step with others, because the variable assignment will be evaluated only after the step is complete:
<code>python
[1: shared = {'test_output': 'output'}]
...
[2]
input: test_output
</code>
The map syntax is evaluated as expressions; therefore it is possible to finer control what specific output, or variations of output, to share with others. For example:</p>

<pre><code class="language-python">[1: shared={'test_output_1':'output[0]', 'test_output_2': 'output[1]'}]
...
[2]
input: test_output_1
</code></pre>

<p>to shared the first file in <code>output</code> (filename <code>output[0]</code>) instead of the entire output file list.</p>

<p>NOTE: Because global variables are readonly, they cannot be changed by individual steps through the <code>shared</code> option. Shared variable therefore has to be either new variables or parameters.</p>

<h3 id="option-provides">Option <code>provides</code></h3>

<p>This option lists files or targets a step generates so that it can be called if the target is required but does not exist. Please see section <a href="Format-Specification#auxiliary-workflow-steps-and-makefile-style-dependency-rules">auxiliary step</a> for details. A step can be treated as both a forward step and an auxiliary step despite of the differences on how they are invoked.</p>

<h2 id="step-input">Step input</h2>

<h3 id="input-files">Input files</h3>

<p>The input of SoS step follows the following rules:</p>

<ul>
  <li><strong>the input of a SoS step is by default the output of the previous step</strong>, which is <code>None</code> for the first step.</li>
  <li><strong><code>input</code> option</strong>, which could be <strong>a list</strong> of filenames (string literal, variables, or expressions that return filenames).  Wildcard characters (<code>*</code>, which matches everything and <code>?</code>, which matches any single character) are acceptable. Nested lists are flattened.</li>
</ul>

<p>Examples of input specification are as follows:</p>

<pre><code class="language-python">input: []

input: 'file1.fasta', 'file2.fasta'

input: 'file*.fasta', 'filename with space.fasta'

input:
    'file*.txt',
    'directory/file2.txt'

input: aligned_reads

input: aligned_reads, reference_genome

input: aligned_reads[2:]

input: 'data/*.fastq'

input: '*/GXT*.fastq'

input: func(parameter)
</code></pre>

<p>It is worth noting that</p>

<ul>
  <li>The first examples shows that the step does not need any input file (so it does not depend on any other step).</li>
  <li>It does not matter if <code>aligned_reads</code> and <code>reference_genome</code> are strings or lists of strings because SoS will flatten nested lists to a single list of filenames.</li>
  <li>The <code>input</code> option tries to expand filenames with wildcard characters (<code>*</code> and <code>?</code>). This can be very useful for workflows that, for example, regularly scan a directory and process unprocessed files. However, because the value of this step depends on availability of files, the output of <code>sos show script</code> and the execution path will be unpredictable, and even wrong if there is no available file during the execution of <code>sos show script</code>.</li>
</ul>

<p>The input files will be evaluated and form a list of input files. They are by default sent to the step process all at once as varible <code>_input</code>, but can also be sent in groups, each time
with different <code>_input</code>. Here <code>_input</code> is a temporary variable that is available only within the step.</p>

<h3 id="option-filetype">Option <code>filetype</code></h3>

<p>SoS allows the specification of input options, which are appended to input file list as comma separated <code>name=value</code> pairs.</p>

<p>Option <code>filetype</code> accepts one or more filetypes or a lambda function. For example,</p>

<pre><code class="language-python">[step]
input:
	input_files, filetype='*.fastq'
</code></pre>

<p>passes only files with extension <code>*.fastq</code>.</p>

<pre><code class="language-python">[step]
input:
	input_files,
	filetype=['*.fastq', '*.fastq.gz']

</code></pre>

<p>passes only files with extension <code>.fastq</code> or <code>.fastq.gz</code>. Under the hood SoS treats the pattern as Unix shell-stype wildcard pattern (with <code>*</code>, <code>?</code>, <code>[seq]</code> and <code>[!seq]</code>, see <a href="https://docs.python.org/2/library/fnmatch.html#module-fnmatch">doc</a> for details) so</p>

<ul>
  <li><strong><code>filetype='.txt'</code> does not match <code>file.txt</code></strong></li>
  <li><code>filetype='*.fastq*'</code> matches <code>a.fastq</code>, <code>a.fastq.gz</code> and <code>a.fastq.zip</code></li>
  <li><code>filetype='[!_]*.txt'</code> matches <code>file1.txt</code> but not <code>_file1.txt</code></li>
</ul>

<p>If you need more refined control over the selection of files, you can use lambda functions (a bit python knowledge is required). For example,</p>

<pre><code class="language-python">[step]
input:
	input_files,
	filetype=lambda x: open(x).readline().startswith('##fileformat=VCF4.1')
</code></pre>

<p>passes only files with the first line starting with string <code>##fileformat=VCF4.1</code>`.</p>

<h3 id="option-groupby">Option <code>group_by</code></h3>

<p>SoS by default passes all input files to step process as a single list. Option <code>group_by</code> pass input files in groups, each time with a subset of input files named <code>_input</code>. SoS allows you to group input by <code>single</code> (individual file), <code>pairs</code> (match first half of files with the second half),  <code>combinations</code> (all unordered combinations of 2-sets), <code>pairwise</code> (all adjacent 2-sets), or chunks of size <code>N</code> (integer <code>group_by=3</code> or string <code>group_by='4'</code>). For example, with the following sos script</p>

<pre><code class="language-python"># how to group input files
parameter: group = 'all'

[0]
print('group_by=${group}')
input: 'file1', 'file2', 'file3', 'file4', group_by=group
print('${_index}: ${_input}')
</code></pre>

<p>The output of commands are</p>

<pre><code class="language-python">$ sos run test.sos  -i -v0
group_by=all
0: file1 file2 file3 file4

$ sos run test.sos --group 'single' -i -v0
group_by=single
0: file1
1: file2
2: file3
3: file4

$ sos run test.sos --group 'pairwise' -i -v0
group_by=pairwise
0: file1 file2
1: file2 file3
2: file3 file4

$ sos run test.sos --group 'pairs' -i -v0
group_by=pairs
0: file1 file3
1: file2 file4

$ sos run test.sos --group 'combinations' -i -v0
group_by=combinations
0: file1 file2
1: file1 file3
2: file1 file4
3: file2 file3
4: file2 file4
5: file3 file4

$ sos run test.sos --group 1 -i -v0
group_by=1
0: file1
1: file2
2: file3
3: file4

$ sos run test.sos --group 2 -i -v0
group_by=2
0: file1 file2
1: file3 file4
</code></pre>
<p>Here we are running the script in inspect mode to avoid file not found errors for input files.</p>

<p>Obviously, the output of the <code>pairs</code> cases depends on the order of files. If you need to pair files in any particular order, you can control it in input. For example</p>

<pre><code class="language-python">[step]
input:
	sorted([x for x in fastq_files if '_R1_' in x]),
	sorted([x for x in fastq_files if '_R2_' in x]),
	group_by='pairs'

run('echo ${input}')
</code></pre>

<p>will take all input files and sort them by <code>_R1_</code> and <code>_R2_</code> and by filename, and pair them.</p>

<h3 id="option-foreach">Option <code>for_each</code></h3>

<p>Option <code>for_each</code> allows you to repeat step process for each value of a variable. For example,</p>

<pre><code class="language-python">[0]
method = ['m1', 'm2']
input: 'file1', 'file2', for_each='method'
print('${_index}: ${_input} ${_method}')
</code></pre>

<p>will repeat the step with each item of variable <code>method</code></p>

<pre><code class="language-python">$ sos run test.sos -v0
0: file1 file2 m1
1: file1 file2 m2
</code></pre>

<p>Note that the SoS automatically creates a loop variable <code>_method</code> for variable <code>method</code>. If the variable is an attribute of an object (e.g. <code>aligned.output</code>), the iterator variable will be named without the attribute part.</p>

<p>Nested loops are also allowed. For example,</p>

<pre><code class="language-python">[0]
method = ['m1', 'm2']
pars = [1, 2]
input: 'file1', 'file2', for_each=['method', 'pars']
print('${_index}: _input=${_input} _method=${_method}, _pars=${_pars}')
</code></pre>

<p>would produce</p>

<pre><code class="language-python">$ sos run test.sos -v0
0: _input=file1 file2 _method=m1, _pars=1
1: _input=file1 file2 _method=m2, _pars=1
2: _input=file1 file2 _method=m1, _pars=2
3: _input=file1 file2 _method=m2, _pars=2
</code></pre>

<p>If you would like to loop the process with several parameters, you can put them into the same level by var1,var2. For example,</p>

<pre><code class="language-python">[0]
method = ['m1', 'm2']
pars = [1, 2]
input: 'file1', 'file2', for_each='method,pars'
print('${_index}: _input=${_input} _method=${_method}, _pars=${_pars}')
</code></pre>

<p>would produce</p>

<pre><code class="language-python">$ sos run test.sos -v0
0: _input=file1 file2 _method=m1, _pars=1
1: _input=file1 file2 _method=m2, _pars=2
</code></pre>

<p>The variable passed to option <code>for_each</code> can a sequence, or a Pandas dataframe. In this case, each <code>_loop</code> variable presents a line in the dataframe and you can access single values using format <code>_loop["header"]</code>. For example</p>

<pre><code class="language-python">[0]
import pandas as pd
data = pd.DataFrame([(1, 2, 'Hello'), (2, 4, 'World')], columns=['A', 'B', 'C'])
input: for_each='data'
output: '${_data["A"]}_${_data["B"]}_${_data["C"]}.txt'
sh:
    touch ${_output}
</code></pre>

<p>would produce results with names <code>1_2_Hello.txt</code> and <code>2_4_World.txt</code>.</p>

<h3 id="option-pairedwith">Option <code>paired_with</code></h3>

<p>Input files might might come with additional information such as sample type, and sample name, and you can pair these information to input files using the <code>paired_with</code> option. For example, <code>bam_files</code> in the following example have matched <code>mutated</code> and <code>sample_name</code>, you can atteched these information to groups of input files <code>_input</code> with looped paired with variables.</p>

<pre><code class="language-python">[0]
bam_files = ['case/A1.bam', 'case/A2.bam', 'ctrl/A1.bam', 'ctrl/A2.bam']
mutated = ['case', 'case', 'ctrl', 'ctrl']
sample_name = ['A1', 'A2', 'A1', 'A2']

input: bam_files, paired_with=['mutated', 'sample_name'], group_by='pairs'

print('${_index}: _input=${_input} _mutated=${_mutated}, _sample_name=${_sample_name}')
</code></pre>

<p>Output:</p>

<pre><code class="language-python">$ sos run test.sos -v0
0: _input=case/A1.bam ctrl/A1.bam _mutated=case ctrl, _sample_name=A1 A1
1: _input=case/A2.bam ctrl/A2.bam _mutated=case ctrl, _sample_name=A2 A2
</code></pre>

<p>Similar to option <code>for_each</code>, if the variable is an attribute of an object (e.g. <code>aligned.output</code>), the iterator variable will be named without the attribute part (e.g. <code>_aligned</code> for <code>paired_with='aligned.output'</code>).</p>

<h3 id="option-pattern">Option <code>pattern</code></h3>

<p>This option does the reverse of function <code>expand_pattern</code>. It uses named wildcards to match pattern to all input files, and creates step variables for these wildcard objects. For example,</p>

<pre><code class="language-python">[step]
input:  'a-20.txt', 'b-10.txt', pattern = '{name}-{par}.txt'
output: expand_pattern('{name}-processed-{par}.txt')
run('echo ${output}; touch ${output}')
</code></pre>

<p>will take all input files and extract <code>name</code> and <code>par</code> from each file name as variables <code>name</code> and <code>par</code>. It is then used to create output file names adding the word <code>processed</code> in between these wildcard objects. The outcome of the SoS script above is creation of files <code>a-processed-10.txt</code> and <code>b-processed-20.txt</code>.</p>

<p>When wildcard objects are accessed as step variables, both variable names with and without <code>_</code> prefix is available, e.g. in this example, both <code>_name</code> and <code>name</code>, <code>_par</code> and <code>par</code> are avaiable and are the same. The two conventions will only differ when <a href="Documentation#option-group_by"><code>group_by</code></a> or <a href="Documentation#option-for_each">for_each</a> is also used.</p>

<pre><code class="language-python">[step]
input:  'a-20.txt', 'b-10.txt', extract = '{name}-{par}.txt', group_by='single'
output: expand = '{_name}-processed-{_par}.txt'
run('echo ${_output}; touch ${_output}')
</code></pre>

<p>Please note that the sigil <code>{}</code> is exclusively used for named wildcards and you should not use <code>${}</code> because it will be interpreted as strings.</p>

<h3 id="option-skip-1">Option <code>skip</code></h3>

<p>Option <code>skip</code> takes either a constant (<code>True</code> or <code>False</code>) or a function. Option <code>skip=True</code>  will make SoS skip the execution of the current step. Using <code>skip=True</code> is not very useful so this option is often used with a SoS variable. For example</p>

<pre><code class="language-python">[10]
input:
	fasta_files,
	skip=len(fasta_failes) == 1

output: 'merged.fasta'

run('command to merge multiple fasta files.')

</code></pre>

<p>Here the <code>skip</code> option gets the value of <code>True</code> if there is only one input file. The command to merge multiple input files would then be skipped.</p>

<p>Another use of option <code>skip</code> is to assign it a function. In this case, this function will be applied
to each input group with <code>_input</code> as the first variable and all <code>paired_with</code>, <code>pattern</code>, <code>loop</code> variables (e.g. <code>_loopvar</code>) as keyword arguments. The input group will be skipped if this function returns <code>False</code>. For example,</p>

<pre><code class="language-python">[10]
def filter_group(ifiles, **kwargs):
    return all(os.path.getfile(x) &gt; 0 for x in ifiles)

input: group_by='combinations', skip=filter_group
</code></pre>

<p>will check all input groups and skip groups with one or two empty files (file size = 0).</p>

<h3 id="dynamic-input-files"><code>dynamic</code> input files</h3>

<p>In order to determine the best execution strategy, SoS evaluates all expressions for all steps before the execution of a workflow to figure
out input and output of steps. This works most of the time but sometimes the input of a step can only be determined at runtime. For example,
if you would like your workflow to automatically scan an input directory and process all fasta files under it, or if a previous step produces
files that cannot be determined beforehand, you can specify input files as follows,</p>

<pre><code class="language-python">input: 'input/*.fasta'
</code></pre>

<p>The problem is that no file or a wrong set files might exist during the planing stage so SoS might skip this step or start the step
with a wrong set of files. To address this problem, you can declare the input files as <strong>dynamic</strong> by passing a <code>dynamic</code> object</p>

<pre><code class="language-python">input: dynamic('input/*.fasta')
</code></pre>

<p>This tells SoS that the input of this step can only be determined at runtime and will execute the step only after all its previous
steps have been completed.</p>

<h3 id="summary">Summary</h3>

<p>Options of step <code>input</code> are evaluated in the following orders:</p>

<ol>
  <li>A list of input files, if specified, would replace <code>input</code>, which is by default output from the previous step.</li>
  <li>Option <code>filetype</code> filters input files. <strong>The output becomes <code>input</code></strong>.</li>
  <li>Option <code>group_by</code> groups the files into several groups, named <code>_input</code></li>
  <li>Option <code>for_each</code> repeat <code>_input</code> for each loop var, named <code>_loopvar</code> if <code>for_each='loopvar'</code>.</li>
  <li>Option <code>paired_with</code> pairs one or more variables with <code>input</code>, variable <code>paired</code> is paired with <code>input</code>
 and variable <code>_paired</code> is paired with <code>_input</code> in each loop if <code>paired_with='paired'</code></li>
  <li>Option <code>pattern</code> extract variables from filenames in <code>input</code>. Variable <code>extracted</code> is paired with <code>input</code>
 and variable <code>_extracted</code> is paired with <code>_input</code> in each loop if <code>extract='{extracted}_other_part'</code>.</li>
  <li>Option <code>skip</code> optionally skip all or part of the input groups.</li>
</ol>

<p>The differences between looped and non-loop steps are sumarized in the following figure</p>

<p><img src="/SOS/media/step_loop.jpg" alt="step_loop" class="img-responsive" /></p>

<h2 id="step-output">Step output</h2>

<h3 id="output-files">Output files</h3>

<p>Output files of a step can be specified by step <code>output</code>. Whereas step <code>input</code> override or
change SoS provided variable <code>input</code> to produce one or more variable <code>_input</code>,
the step <code>output</code> specify <code>output</code> which should be <code>[]</code> if no output is generated
(or of interest). Similar to <code>input</code>, step output accepts strings, variables, expressions, and allows wildcard characters. For example, the following are acceptable output files</p>

<pre><code class="language-python">output:  []

output:  'accepted_hits.bam'

output:  aligned_reads, bam_stats

output:  'aligned/*.bam'

output:  expand_pattern('aligned_{samples}.bam')
</code></pre>

<p>In the last example, function <code>expand_pattern</code> is used to contruct list of files from items of a sequence <code>samples</code>. In case of <code>input</code> loop, step <code>output</code> actually determines variable <code>_output</code> for each input loop. For example, the following step accepts one or more bam files and index them using command <code>samtools index</code>. The input files are passed one
by one so generate multiple <code>output</code> files (<code>'${_input}.bai'</code>).</p>

<pre><code class="language-python">[10]
input:
	bamfiles, group_by='single'

output:
	'${_input}.bai'

run('''samtools index ${_input} ''')
</code></pre>

<p>The use of variable <code>output</code> in this scenario is discouraged because <code>output</code>, as the collection of all <code>_output</code> increases with each input group. The accumulation only happens when <code>_output</code> and <code>output</code> are different so <code>output</code> will stay the same if all output files are specified as in the following example</p>

<pre><code class="language-python">output: [x + '.bai' for x in bamfiles]
</code></pre>

<h3 id="dynamic-output-files"><code>dynamic</code> output files</h3>

<p>Similar to the cases with <a href="#dynamically-determined-input-files-function-dynamic">dynamic input files</a>, the
output of some steps could also not be determined beforehand. For example, with the following script that generates <code>html</code> files that cannot be determined during dry run,</p>

<pre><code class="language-python">[0]
run:
    rm -f *.html

[10]
input: []
output: '*.html'

import random
for i in range(5):
    run('touch result_${random.randint(1, 20)}.html')
</code></pre>

<p>SoS will determine that you do not have any output file (no <code>*.html</code> file) and produce the following output</p>

<pre><code class="language-python">$ sos run test.sos
INFO: Execute default_0:
INFO: input:   []
INFO: output:  unspecified
INFO: Execute default_10:
INFO: input:   []
WARNING: *.html does not expand to any valid file.
INFO: output:  []
</code></pre>

<p>In this case, you will need to define the output as <code>dynamic</code> using</p>

<pre><code class="language-python">[0]
run:
    rm -f *.html

[10]
input: []
output: dynamic('*.html')

import random
for i in range(5):
    run('touch result_${random.randint(1, 20)}.html')
</code></pre>

<p>so that SoS knows that the output can only be determined after the completion of the step. Because of this, variable <code>output</code> is unavailable to the step process.</p>

<h2 id="step-dependencies">Step dependencies</h2>

<p>This item specifies files that are required for the step. Although not required, it is a good practice to list resource files and other dependency files for a particular step. For example</p>

<pre><code class="language-python">[10]
input:
	fasta_files

depends:
	reference_seq

</code></pre>

<p>Similar to <code>output</code> options, dependent files can also be defined after <code>input</code> options and consist of
dependent files determined from loop variables.</p>

<p>The following figure summarizes the effect of <code>input</code>
and <code>output</code> options and input options <code>group_by</code> and <code>for_each</code> on the flow
of input and output files and related variables.</p>

<p><img src="/SOS/media/step_options.jpg" alt="step_options" class="img-responsive" /></p>

<h2 id="step-process">Step process</h2>

<p>A step process is the Python statements that perform certain tasks and produce step output from step input. A step process can contain arbitrary Python statements. For example,</p>

<pre><code class="language-python">[10]
output: 'a.txt'
with open(${_output!r}, 'w') as dest:
   dest.write('some text')
</code></pre>

<p>and</p>

<pre><code class="language-python">[10]
output: 'a.txt'
run('echo "some text" &gt; ${_output!q}')
</code></pre>

<p>use inline (interpreted and executed by SoS) python code or shell script to generate <code>a.txt</code>.</p>

<p>Step processes are executed within SoS and are executed sequentially. However, part or all of the step process can be executed externally and potentially in parallel as step <code>task</code>.</p>

<h2 id="external-task">External <code>task</code></h2>

<p>If a job is long and time consuming, it is much preferred to submit them as separate tasks to be executed, for example, on a cluster system. These jobs should be specified using the <code>task</code> keyword, which marks the beginning of a task, with optional runtime options to control its execution. For example,</p>

<pre><code class="language-python">[10]
input: group_by='single'

task: concurrent=True

run('''
samtools index {_input}
''')
</code></pre>

<p>execute a shell script in parallel (with <code>concurrent=True</code>). The step process can consists of arbitrary python statements and execute multiple step actions. For example,</p>

<pre><code class="language-python">task:
try:
   action1()
except RuntimeError:
   action2()
</code></pre>

<p>execute <code>action1</code> and <code>action2</code> if <code>action1</code> raises an error.</p>

<pre><code class="language-python">task:
for par in ['-4', '-6']:
   run('command with ${par}')
</code></pre>

<p>executes commands in a loop. This is similar to</p>

<pre><code class="language-python">pars = ['-4', '-6']
input: for_each=pars
task:
run('command with ${_pars}')
</code></pre>

<p>but the <code>for</code> loop version would not be able to be executed in parallel. Note that SoS actions can be used outside of <code>step process</code> but only statements specified after the <code>process</code> keyword can have runtime options and be executed in separate processes. That is to say,</p>

<pre><code class="language-python">pars = ['-4', '-6']
input: for_each=pars
run('command with ${_pars}')
</code></pre>

<p>is equivalent to</p>

<pre><code class="language-python">pars = ['-4', '-6']
input: for_each=pars
task:
run('command with ${_pars}')
</code></pre>

<p>but the latter can have additional runtime options to run commands in parallel</p>

<pre><code class="language-python">pars = ['-4', '-6']
input: for_each=pars
task: concurrent=True
run('command with ${_pars}')
</code></pre>

<p>Because step tasks are executed outside of SoS, variables assigned in step tasks are not accessible to SoS. For example,</p>

<pre><code class="language-python">[10: shared='res']
res = some_action()
</code></pre>

<p>executes <code>some_action()</code> in step process and return its result as a shared variable <code>res</code>. The following script,</p>

<pre><code class="language-python">[10: shared='res']
task:
res = some_action()
</code></pre>

<p>however, does not work because <code>res</code> is assigned in step task and is not accessible from the step.</p>

<h3 id="option-workdir">Option <code>workdir</code></h3>

<p>Default to current working directory.</p>

<p>Option <code>workdir</code> controls the working directory of the process. For example, the following step downloads a file to the <code>resource_dir</code> using command <code>wget</code>.</p>

<pre><code class="language-python">[10]

run: workdir=resource_dir

  wget a_url -O filename

</code></pre>

<h3 id="option-concurrent">Option <code>concurrent</code></h3>

<p>Default to <code>False</code>.</p>

<p>If the step process is repeated for different input files or parameters (using input options <code>group_by</code> or <code>for_each</code>), the loop process can be execute in parallel, up to the maximum number of concurrent jobs specified by command line option <code>-j</code>.</p>

<h3 id="option-env">Option <code>env</code></h3>

<p>The <code>env</code> option allow you to modify runtime environment, similar to the <code>env</code> parameter of the <code>subprocess.Popen</code> function. For example, you can execute your command with in a specific directory using</p>

<pre><code class="language-python">task:  env={'PATH': '/path/to/mycommand' + os.sep + os.environ['PATH']}
run:
   mycommand 
</code></pre>

<h3 id="option-prependpath">Option <code>prepend_path</code></h3>

<p>Option <code>prepend_path</code> is a shortcut to option <code>env</code> to prepend one (a string) or more (a list of strings) paths to system path. For example, the above example can be shortened to</p>

<pre><code class="language-python">task:  prepend_path='/path/to/mycommand'
run:
   mycommand 
</code></pre>

<h3 id="option-walltime">Option <code>walltime</code></h3>


                            </div>
                            <button type="button" class="btn btn-default" data-dismiss="modal"><i class="fa fa-times"></i> Close</button>
                        </div>
                    </div> 
                    <div class="footer-wrapper">
                      <footer role="contentinfo">
                        <span>&copy; 2016 Bo Peng.</span>
                      </footer>
                    </div>
                </div>
            </div>
        </div>
    </div> 


   <div class="portfolio-modal modal fade" id="Make-Style-Rules" tabindex="-1" role="dialog" aria-hidden="true">
        <div class="modal-content">
            <div class="close-modal" data-dismiss="modal">
                <div class="lr">
                    <div class="rl">
                    </div>
                </div>
            </div>
            <div class="container">
                <div class="row">
                 <div class="col-lg-8 col-lg-offset-2">
                        <h1>7.Make-Style-Rules</h1>
                        <div class="modal-body">
                         
                            <hr class="star-primary">
                            <h2 id="makefile-style-steps">Makefile-style steps</h2>

<p>Auxiliary steps are special steps that are used only when a target is needed but not available. Such steps are defined in the format of</p>

<pre><code class="language-python">[step_name : provides=pattern]
</code></pre>

<p>where <code>pattern</code> can be</p>

<ul>
  <li>A file pattern such as <code>{sample}.bam.idx</code></li>
  <li>A <code>target</code> such as <code>executable("ms")</code></li>
  <li>A list (sequence) of one or more file patterns and targets.</li>
</ul>

<p>When a target is required, SoS will look in the script for steps that provides such a target. It will then</p>

<ol>
  <li>Create one or more variables defined by the pattern. For example, if the step matches to <code>AS123.bam.idx</code>, a variable named <code>sample</code> would be defined with value <code>AS123</code>.</li>
  <li>The default <code>input</code> of the step would be <code>None</code> instead of output of previous step because there is no previous step for auxiliary steps.</li>
  <li>The default <code>output</code> of the step would be the matched target, although you are free to generate more output files.</li>
</ol>

<p>For example, step</p>

<pre><code class="language-python">[index_bam : provides='{sample}.bam.bai']
input: '${sample}.bam'
run:
     samtools index ${input}
</code></pre>

<p>would be called when a target <code>AS123.bam.bai</code> is required. This step would have variable <code>sample</code> defined as <code>AS123</code>, default input <code>None</code>, and default output <code>["AS123.bam.bai"]</code>. The input would be redefined by <code>input: ${sample}.bam</code> as <code>AS123.bam</code> and be passed to the <code>run</code> action.</p>

<p>You might have already realized that an auxiliary step is a makefile style step and you can use this technique to build complete
workflows in a way similar to <a href="https://bitbucket.org/johanneskoester/snakemake">snakemake</a>. That is to say, you can define multiple
auxiliary steps (rules in snakemakes term) and let SoS determine what steps to execute depending on what workflow target to produce.
You can even mix the forward, input-oriented style with backward, output-oriented style in SoS.</p>


                            </div>
                            <button type="button" class="btn btn-default" data-dismiss="modal"><i class="fa fa-times"></i> Close</button>
                        </div>
                    </div> 
                    <div class="footer-wrapper">
                      <footer role="contentinfo">
                        <span>&copy; 2016 Bo Peng.</span>
                      </footer>
                    </div>
                </div>
            </div>
        </div>
    </div> 


   <div class="portfolio-modal modal fade" id="SoS-Actions" tabindex="-1" role="dialog" aria-hidden="true">
        <div class="modal-content">
            <div class="close-modal" data-dismiss="modal">
                <div class="lr">
                    <div class="rl">
                    </div>
                </div>
            </div>
            <div class="container">
                <div class="row">
                 <div class="col-lg-8 col-lg-offset-2">
                        <h1>8.SoS-Actions</h1>
                        <div class="modal-body">
                         
                            <hr class="star-primary">
                            <p>Although arbitrary python functions can be used in SoS step process, SoS defines some <strong><code>actions</code></strong> (e.g. the <code>run</code> function in the aforementioned examples)
that can be used in a SoS script. The only difference between a SoS action and a regular Python function is that they can behave differently
in different <code>run_mode</code>. For example,</p>

<ul>
  <li>Most SoS actions return 0 directly in <code>inspect</code> mode.</li>
  <li>Some SoS actions such as <code>check_command</code> work in both <code>run</code> and <code>inspect</code> mode so that you can check if you script can be executed by running it in inspect mode.</li>
</ul>

<p>Because SoS executes all step processes in <code>inspect</code> and <code>prepare</code> mode for the planning of large workflows, <strong>it is important to enclose data processing steps in SoS actions or execute conditionally (with <code>if run_mode == 'run'</code>)</strong>. For example, running script</p>

<pre><code class="language-python">[0]
import time
time.sleep(5)
</code></pre>

<p>with command <code>sos run</code> will sleep for 15 seconds instead of 5 seconds because the <code>sleep</code> function will be called in inspect, prepare, and run mode. To fix this, you can either put the function in an action</p>

<pre><code class="language-python">[0]
python:
   import time
   time.sleep(5)
</code></pre>

<p>or use <code>run_mode</code> condition</p>

<pre><code class="language-python">[0]
if run_mode == 'run':
    import time
    time.sleep(5)
</code></pre>

<p>It is easy to define your own actions. All you need to do is to define a function and decorate it with a <code>SoS_Action</code> decorator. For example</p>

<pre><code class="language-python">
from pysos import SoS_Action

@SoS_Action(run_mode='run')
def my_action(parameters):
    do_something_with_parameters
	return 1
</code></pre>

<p><code>run_mode='run'</code> can also be <code>run_mode=('inspect', 'run')</code> or <code>run_mode='inspect'</code> indicating in which mode the function will be executed (instead of returning 0 directly).</p>

<h2 id="option-runmode">Option <code>run_mode</code></h2>

<p>All actions have pre-defined run_mode. For example, <code>sh</code> action runs at <code>run</code> mode and <code>download</code> runs at <code>prepare</code> mode. If for some applications you would like to run an action in different mode, you can change it by setting option <code>run_mode</code> to a mode at which an action would be executed. For example, the following step</p>

<pre><code class="language-python">[install_fastq: provides=executable('fastq')]
run: run_mode='prepare'
  script_to_install_fastq
</code></pre>

<p>would be called when an executable <code>fastq</code> is required. The commands would be running in <code>prepare</code> mode to install the program for the execution of the workflow.</p>

<h2 id="option-dockerimage">Option <code>docker_image</code></h2>

<p>If a docker image is specified (either a name, an Id, or a file), the action is assumed to be executed in the specified docker. The image will be automatically downloaded (pulled) or loaded (if a <code>.tar</code> or <code>.tar.gz</code> file is specified<code>) if it is not available locally. Note that this option only affect script executing actions such as </code>run<code>, </code>python<code> and </code>perl<code> so other actions such as </code>check_command` will be executed on the host machine even if it is included in the step process.</p>

<p>For example, executing the following script</p>

<pre><code class="language-python">[10]
python3: docker_image='python'
  set = {'a', 'b'}
  print(set)
</code></pre>

<p>under a docker terminal (that is connected to the docker daemon) will</p>

<ol>
  <li>Pull docker image <code>python</code>,  which is the official docker image for Python 2 and 3.</li>
  <li>Create a python script with the specified content</li>
  <li>Run the docker container <code>python</code> and make the script available inside the container</li>
  <li>Use the <code>python3</code> command inside the container to execute the script.</li>
</ol>

<p>Additional <code>docker_run</code> parameters can be passed to actions when the action
is executed in a docker image. These options include</p>

<ul>
  <li><code>name</code>: name of the container (option <code>--name</code>)</li>
  <li><code>tty</code>: if a tty is attached (default to <code>True</code>, option <code>-t</code>)</li>
  <li><code>stdin_open</code>: if stdin should be open (default to <code>False</code>, option <code>-i</code>)</li>
  <li><code>user</code>: username (default o <code>root</code>, option <code>-u</code>)</li>
  <li><code>environment</code>: Can be a string, a list of string or dictinary of environment variables for docker (option <code>-e</code>)</li>
  <li><code>volumes</code>: string or list of string, extra volumes that need to be link, in addition to SoS mounted (<code>/tmp</code>, <code>/Users</code> (if mac), <code>/Volumes</code> (if <a href="https://github.com/bpeng2000/SOS/wiki/SoS-Docker-guide">properly configured</a> under mac) and script file)</li>
  <li><code>volumes_from</code>: container names or Ids to get volumes from</li>
  <li><code>working_dir</code>: working directory (option <code>-w</code>), default working directory, or working directory set by runtime option <code>workdir</code>.</li>
  <li><code>port</code>: port opened (option <code>-p</code>)</li>
  <li><code>extra_args</code>: If there is any extra arguments you would like to pass to the <code>docker run</code> process (after you check the actual command of <code>docker run</code> of SoS</li>
</ul>

<h2 id="option-dockerfile">Option <code>docker_file</code></h2>

<p>This option allows you to import a docker from specified <code>docker_file</code>, which can be an archive file (<code>.tar</code>, <code>.tar.gz</code>, <code>.tgz</code>, <code>.bzip</code>, <code>.tar.xz</code>, <code>.txz</code>) or a URL to an archive file (e.g. <code>http://example.com/exampleimage.tgz</code>). SoS will use command <code>docker import</code> to import the <code>docker_file</code>. However, because SoS does not know the repository and tag names of the imported docker file, you will still need to use option <code>docker_image</code> to specify the image to use.</p>

<h2 id="action-sosrun">Action <code>sos_run</code></h2>

<p>Action <code>sos_run(workflow, source)</code> executes a specified workflow. The workflow can be a single workflow, a subworkflow (e.g. <code>A_-10</code>), or a combined workflow (e.g. <code>A + B</code>). Because the workflow is executed from a step, it takes step <code>_input</code> as the input of the nested workflow and it can access local step variables. For example,</p>

<pre><code class="language-python">import random

[simulate]
output:   'result_${_reps}.txt'
run:
simulate_experiment --seed=${seed} --output=${_output}

[10]
reps = range(100)
input: for_each='reps'
outout: 'result_${_reps}.txt'

seed = random.randint(1, 2**32)
sos_run('simulate')
</code></pre>

<p>would run the nested pipeline <code>simulate</code> (which is a single step in this example) 100 times with their own <code>seed</code>, and <code>_reps</code>.</p>

<p>The workflow can be defined in the current script, or in other SoS scripts, in which case the name or full path of the SoS script should be provided to parameter <code>source</code>. For example,</p>

<pre><code class="language-python">[myworkflow]
sos_run('A+B', source="AB.sos")
</code></pre>

<p>defines a nested workflow with workflow <code>A</code> and/or <code>B</code> defined in <code>AB.sos</code>. The nested workflow is a combination of two workflows <code>A</code> and <code>B</code> with their own parameters sections. SoS searches the specified files in the current working directory, the directory of the master script, and a search path specified by variable <code>sos_path</code> defined in the SoS global (<code>~/.sos/config.yaml</code>) or local (<code>.sos/config.yaml</code>) configuration files, and will produce an error if no file can be found.</p>

<h2 id="action-run-bash-sh-csh-tcsh-zsh">Action <code>run</code>, <code>bash</code>, <code>sh</code>, <code>csh</code>, <code>tcsh</code>, <code>zsh</code></h2>

<p>Actions <code>run(script)</code> and <code>bash(script)</code> accepts a shell script and execute it using <code>bash</code>. <code>sh</code>, <code>csh</code>, <code>tcsh</code>, <code>zsh</code> uses respective shell to execute the provided script.</p>

<p>These actions, as well as all script-executing actions such as <code>python</code>, also accept an option <code>args</code> and allows you to pass additional arguments to the interpreter. For example</p>

<pre><code class="language-python">run: args='-n'
  echo "a"
</code></pre>

<p>will execute the script with command <code>bash -n</code> (check syntax).</p>

<h2 id="action-python-and-python3">Action <code>python</code> and <code>python3</code></h2>

<p>Action <code>python(script)</code> and <code>python3(script)</code> accepts a Python script and execute it with python or python3, respectively.</p>

<p>Because SoS can include Python statements directly in a SoS script, it is important to note that embedded Python
statements are interpreted by SoS and the <code>python</code> and <code>python3</code> actions are execute in separate processes without
access to the SoS environment.</p>

<p>For example, the following SoS step execute some python statements <strong>within</strong> SoS with direct access to SoS variables
such as <code>input</code>, and with <code>result</code> writing directly to the SoS environment,</p>

<pre><code class="language-python">[10]
for filename in input:
    with open(filename) as data:
        result = filename + '.res'
        ....
</code></pre>

<p>while</p>

<pre><code class="language-python">[10]
input: group_by='single'

python:

with open(${input!r}) as data:
   result = ${input!r} + '.res'
   ...


</code></pre>

<p>composes a Python script for each input file and calls separate Python interpreters to execute them. Whereas
the Python statement in the first example will always be executed, the statements in <code>python</code> will not be executed
in <code>inspect</code> mode.</p>

<h2 id="action-r-and-checkrlibrary">Action <code>R</code> and <code>check_R_library</code></h2>

<p>Action <code>R(script)</code> exexute the passed script using <code>Rscript</code> command. Action <code>check_R_library</code> accept an R library and optionally can check for required versions. If the libraries are not available, it will try to install it from <a href="https://cran.r-project.org/">CRAN</a>, <a href="https://www.bioconductor.org/">bioconductor</a>, or <a href="https://github.com/">github</a>. Github package name should be formatted as <code>repo/pkg</code>. Action <code>check_R_library</code> is available in both <code>run</code> and <code>inspect</code> mode.</p>

<p>For example, <code>check_R_library('edgeR')</code> will check and install (if necessary) <code>edgeR</code> from bioconductor. <code>check_R_library('stephens999/ashr')</code> will check and install <code>ashr</code> package from a github repository https://github.com/stephens999/ashr.</p>

<p><code>check_R_library</code> can also be used to check for required version of packages. For example:</p>

<pre><code class="language-python">check_R_library('edgeR', '3.12.0')
</code></pre>
<p>will result in a warning if edgeR version is not 3.12.0, and</p>

<p>Multiple versions is allowed, for example:
<code>python
check_R_library('edgeR', ['3.12.0', '3.12.1'])
</code>
Or, simply:
<code>python
check_R_library('edgeR', '3.12.0+')
</code>
It is also possible to restrict package to old version, for example:
<code>python
check_R_library('ggplot2', '1.0.0-')
</code></p>

<p>Note that version mismatch triggers a warning message, not an error. If you would like to enforce certain version, use</p>

<pre><code class="language-python">fail_if(check_R_library('ggplot2', '1.0.0+') != 0, 'Version 1.0.0 or newer version of ggplot2 is required')
</code></pre>

<p>The default R library repo is <code>http://cran.us.r-project.org</code>. It is possible to customize the repo, for example:</p>

<pre><code class="language-python">check_R_library('Rmosek', repos = "http://download.mosek.com/R/7")
</code></pre>

<h2 id="action-perl-ruby">Action <code>perl</code>, <code>ruby</code></h2>

<p>Action <code>perl(script)</code> execute the passed script using <code>perl</code> interpreter. Action <code>ruby(script)</code> execute the passed script using <code>ruby</code> interpreter.</p>

<h2 id="action-node-javascript">Action <code>node</code>, <code>JavaScript</code></h2>

<p>Action <code>node(script)</code> and <code>JavaScript(script)</code> execute the passed script using <code>node</code> interpreter.</p>

<h2 id="action-dockerbuild">Action <code>docker_build</code></h2>

<p>Build a docker image from an inline Docker file. The inline version of the action currently does not support adding any file from local machine because the docker file will be saved to a random directory. You can walk around this problem by creating a <code>Dockerfile</code> and pass it to the action through option <code>path</code>. This action accepts all parameters as specified in https://docker-py.readthedocs.org/en/stable/api/#build because SoS simply pass additional parameters to the <code>build</code> function.</p>

<p>For example, the following step builds a docker container for <a href="http://miso.readthedocs.org/en/fastmiso/">MISO</a> based on anaconda python 2.7.</p>

<pre><code class="language-python">[build_1]
# building miso from a Dockerfile
docker_build: tag='mdabioinfo/miso:latest'

	############################################################
	# Dockerfile to build MISO container images
	# Based on Anaconda python
	############################################################

	# Set the base image to anaconda Python 2.7 (miso does not support python 3)
	FROM continuumio/anaconda

	# File Author / Maintainer
	MAINTAINER Bo Peng &lt;bpeng@mdanderson.org&gt;

	# Update the repository sources list
	RUN apt-get update

	# Install compiler and python stuff, samtools and git
	RUN apt-get install --yes \
	 build-essential \
	 gcc-multilib \
	 gfortran \ 
	 apt-utils \
	 libblas3 \ 
	 liblapack3 \
	 libc6 \
	 cython \ 
	 samtools \
	 libbam-dev \
	 bedtools \
	 wget \
	 zlib1g-dev \ 
	 tar \
	 gzip

	WORKDIR /usr/local
	RUN pip install misopy
</code></pre>

<h2 id="action-download">Action <code>download</code></h2>

<p>Action <code>download(URLs, dest_dir='.', dest_file=None, decompress=False)</code> download files from specified URLs, which can be a list of URLs, or a string with tab, space or newline separated URLs.</p>

<ul>
  <li>If <code>dest_file</code> is specified, only one URL is allowed and the URL can have any form.</li>
  <li>Otherwise all files will be downloaded to <code>dest_dir</code>. Filenames are determined from URLs so the URLs must have the last portion as the filename to save.</li>
  <li>If <code>decompress</code> is True, <code>.zip</code> file, compressed or plan <code>tar</code> (e.g. <code>.tar.gz</code>) files, and <code>.gz</code> files will be decompressed to the same directory as the downloaded file.</li>
</ul>

<p>For example,</p>

<pre><code class="language-python">[10]
GATK_RESOURCE_DIR = '/path/to/resource'
GATK_URL = 'ftp://gsapubftp-anonymous@ftp.broadinstitute.org/bundle/2.8/hg19/'

download:   dest=GATK_RESOURCE_DIR
    ${GATK_URL}/1000G_omni2.5.hg19.sites.vcf.gz
    ${GATK_URL}/1000G_omni2.5.hg19.sites.vcf.gz.md5
    ${GATK_URL}/1000G_omni2.5.hg19.sites.vcf.idx.gz
    ${GATK_URL}/1000G_omni2.5.hg19.sites.vcf.idx.gz.md5
    ${GATK_URL}/dbsnp_138.hg19.vcf.gz
    ${GATK_URL}/dbsnp_138.hg19.vcf.gz.md5
    ${GATK_URL}/dbsnp_138.hg19.vcf.idx.gz
    ${GATK_URL}/dbsnp_138.hg19.vcf.idx.gz.md5
    ${GATK_URL}/hapmap_3.3.hg19.sites.vcf.gz
    ${GATK_URL}/hapmap_3.3.hg19.sites.vcf.gz.md5
    ${GATK_URL}/hapmap_3.3.hg19.sites.vcf.idx.gz
    ${GATK_URL}/hapmap_3.3.hg19.sites.vcf.idx.gz.md5
</code></pre>

<p>download the specified files to <code>GATK_RESOURCE_DIR</code>. The <code>.md5</code> files will be automatically used to validate the content of the associated files. Note that</p>

<p>SoS automatically save signature of downloaded and decompressed files so the files will not be re-downloaded if the action is called multiple times. You can however still still specifies input and output of the step to use step signature</p>

<pre><code class="language-python">[10]
GATK_RESOURCE_DIR = '/path/to/resource'
GATK_URL = 'ftp://gsapubftp-anonymous@ftp.broadinstitute.org/bundle/2.8/hg19/'
RESOUCE_FILES =  '''1000G_omni2.5.hg19.sites.vcf.gz
    1000G_omni2.5.hg19.sites.vcf.gz.md5
    1000G_omni2.5.hg19.sites.vcf.idx.gz
    1000G_omni2.5.hg19.sites.vcf.idx.gz.md5
    dbsnp_138.hg19.vcf.gz
    dbsnp_138.hg19.vcf.gz.md5
    dbsnp_138.hg19.vcf.idx.gz
    dbsnp_138.hg19.vcf.idx.gz.md5
    hapmap_3.3.hg19.sites.vcf.gz
    hapmap_3.3.hg19.sites.vcf.gz.md5
    hapmap_3.3.hg19.sites.vcf.idx.gz
    hapmap_3.3.hg19.sites.vcf.idx.gz.md5'''.split() 
input: []
output:  [os.path.join(GATK_RESOURCE_DIR, x) for x in GATK_RESOURCE_FILES]
download(['${GATK_URL}/${x}' for x in GATK_RESOURCE_FILES], dest=GATK_RESOURCE_DIR)
</code></pre>

<p>Note that the <code>download</code> action uses up to 5 processes to download files. You can change this number by adjusting system configuration <code>sos_download_processes</code>.</p>

<h2 id="action-report">Action <code>report</code></h2>

<p>Action <code>report(script, filename=None, mode='a')</code> writes the specified script to the report channel, or to a file with specified <code>filename</code>. It will by default append to the report (<code>mode='a'</code>) but you can start a new report with <code>mode='w'</code>.</p>

<p>Note that:</p>

<ul>
  <li>
    <p>Although markdown is generally used, SoS does not restrict the type of report a script generates. If so desired, you can use this mechanism to generate a log file in plain text format, a HTML file with fancy style, any flavor of markdown that you can post-process, or a <code>.Rmd</code> file that can be loaded into RMarkdown.</p>
  </li>
  <li>
    <p>The report is generated from the steps that are executed so the content can be controlled with parameters and configuration files.</p>
  </li>
</ul>

<h2 id="action-pandoc">Action <code>pandoc</code></h2>

<p>Action <code>pandoc</code> converts the report generated so far to another format. Parameters <code>outputfile</code>, <code>format</code>, <code>to</code>, <code>filter</code> and <code>extra_args</code> can be used to customize the output. Please refer to <a href="https://pypi.python.org/pypi/pypandoc/">pandoc and pypandoc documentation</a> for details about these parameters.</p>

<h2 id="action-checkcommand">Action <code>check_command</code></h2>

<p>Action <code>check_command(cmd, pattern)</code> check existence of command or the output of the command. This action works in both dyrun and run mode so it can be used to check the existence of command in inspect mode.</p>

<p>For example, if a script contains step</p>

<pre><code class="language-python">[100]
check_command('tophat')
run('tophat ...')
</code></pre>

<p>Command sos run script -i would check the existence of command <code>tophat</code> without actually running
the <code>tophat ...</code> command.</p>

<p>Action <code>check_command(cmd, pattern)</code> execute specified <code>cmd</code> and searches specified pattern in its output. <code>pattern</code>  should be one or a list of python regular expressions (see <a href="https://docs.python.org/2/library/re.html">Python re module</a>, especially the <code>search</code> function for details). It raises an exception if the output does not match any of the patterns. This function is usually used to check the version of commands.</p>

<p>For example, action</p>

<pre><code class="language-python">check_command('STAR --version', ['2.4.0', '2.5.0'])
</code></pre>
<p>checks the output of command <code>STAR --version</code> and raises an error if it does not contain string <code>2.4.0</code> or <code>2.5.0</code>.</p>

<p>Note that</p>

<ol>
  <li>
    <p><code>check_command</code> will return a non-zero value if the command exist but returns non-zero value (e.g. because of incorrect command line argument). A warning message will also be printed.</p>
  </li>
  <li>
    <p><code>check_command</code> will timeout after 2 seconds because it is intended to check basic information of specified command. The action will produce a warning message and return 1 in this case.</p>
  </li>
</ol>

<h2 id="action-failif">Action <code>fail_if</code></h2>

<p>Action <code>fail_if(expr, msg='')</code> raises an exception with <code>msg</code> (and terminate the execution of the workflow if the exception is not caught) if <code>expr</code> returns True.</p>

<h2 id="action-warnif">Action <code>warn_if</code></h2>

<p>Action <code>warn_if(expr, msg)</code> yields a warning message <code>msg</code> if <code>expr</code> is evaluate to be true.</p>

<h2 id="action-abortif">Action <code>abort_if</code></h2>

<p>Action <code>abort_if(expr, msg='')</code> stops the execution of the current step (or current processes if within <code>for_each</code> loop) and gives a warning message if <code>msg</code> is specified. For example,</p>

<pre><code class="language-python">[10]
input: '*.txt', for_each=1

abort_if(os.path.getsize(_input[0]) &gt; 10000)
do_something_else
</code></pre>

<p>skips files that are larger than <code>10k</code>.</p>

<h2 id="utility-functions-and-logger">Utility functions and <code>logger</code></h2>

<p>SoS exposes a few utility functions that can be helpful from time to time.</p>

<h2 id="function-getoutput">Function <code>get_output</code></h2>

<p>Function <code>get_output(cmd)</code> returns the output of command (decoded in <code>UTF-8</code>), which is a shortcut for <code>subprocess.check_output(cmd, shell=True).decode()</code>. It is worth noting that SoS kills any function call after 5 seconds in inspect mode so you will need to put this function call inside a step process if it will take more than 5 seconds to execute.</p>

<h2 id="function-expandpattern">Function <code>expand_pattern</code></h2>

<p>Function <code>expand_pattern</code> expands a string to multiple ones using items of variables quoted between <code>{ }</code>. For example,</p>

<pre><code class="language-python">output: expand_pattern('{a}_{b}.txt')
</code></pre>

<p>is equivalent to</p>

<pre><code class="language-python">output: ['{x}_{y}.txt' for x,y in zip(a, b)]
</code></pre>

<p>if <code>a</code> and <code>b</code> are sequences of the same length.</p>

<h2 id="sos-logger-object">SoS <code>logger</code> object</h2>

<p>The SoS logger object is a <code>logging</code> object used by SoS to produce various outputs. You can use this object
to output error, warning, info, debug, and trace messages to terminal. For example,</p>

<pre><code class="language-python">[0]
logger.info('I am at ${step_name}')
</code></pre>

<p>would print a logging message <code>I am at default_0</code> when the first step is execute.</p>

<pre><code class="language-python">$ sos run test.sos
INFO: Execute default_0: 
INFO: input:   []
INFO: I am at default_0
INFO: output:  unspecified
</code></pre>


                            </div>
                            <button type="button" class="btn btn-default" data-dismiss="modal"><i class="fa fa-times"></i> Close</button>
                        </div>
                    </div> 
                    <div class="footer-wrapper">
                      <footer role="contentinfo">
                        <span>&copy; 2016 Bo Peng.</span>
                      </footer>
                    </div>
                </div>
            </div>
        </div>
    </div> 


   <div class="portfolio-modal modal fade" id="Workflow-Execution" tabindex="-1" role="dialog" aria-hidden="true">
        <div class="modal-content">
            <div class="close-modal" data-dismiss="modal">
                <div class="lr">
                    <div class="rl">
                    </div>
                </div>
            </div>
            <div class="container">
                <div class="row">
                 <div class="col-lg-8 col-lg-offset-2">
                        <h1>9.Workflow-Execution</h1>
                        <div class="modal-body">
                         
                            <hr class="star-primary">
                            <h2 id="specification-of-workflow-or-targets">Specification of workflow or targets</h2>

<p>There are two ways to specify a workflow</p>

<ol>
  <li>
    <p>Specify name of a workflow (<code>sos run script workflow</code>), this includes the execution of default workflow, single workflow (e.g. <code>hg18</code>), subworflow (e.g. <code>hg18_:100</code> (up to step 100)), and combined workflows (e.g. <code>hg18+report</code>). SoS will execute steps of the specified workflow, in addition to any auxiliary steps that are needed. The <strong>target</strong> of the execution would be the output of the last step of the workflow.</p>
  </li>
  <li>
    <p>Specify targets that needs to be generated (<code>sos run script -t target</code>). SoS will construct a DAG (Directed Acyclic Graph) with auxiliary steps that are need to generate specified targets and execute the graph.</p>
  </li>
</ol>

<h2 id="dag-and-parallel-execution">DAG and Parallel execution</h2>

<p>Logically speaking, with command</p>

<ul>
  <li>Evaluate parameters with either command line input or their default values</li>
  <li>Evaluate the rest of steps in numeric order</li>
</ul>

<p>Actual execution order depends on step actions, input output options of SoS steps, and option <code>-j</code>.
In the sequential execution mode (by default),
all steps are executed one by one with auxiliary steps called when necessary. If a script is written without any step option and
input and output files, it will be executed sequentially even in parallel execution mode (with <code>-j</code> option).</p>

<p>If the steps are described with necessary input and output information, steps in SoS workflows can be executed in parallel. The
following figure illustrates the impact input/output options on the execution order of workflows. Note that a step with
unknown <code>input</code> (no <code>input</code> at present step and no <code>output</code> at previous step) can only be executed after all its previous steps
are completed can become bottlenecks of the workflow.</p>

<p>[[/SOS/media/workflow.jpg]]</p>

<h2 id="runtime-signature">Runtime signature</h2>

<h2 id="removal-of-intermediate-files">Removal of intermediate files</h2>

<h2 id="targets-of-workflow">Targets of workflow</h2>

<h2 id="configuration-file">Configuration file</h2>

<p>SoS reads a global (<code>~/.sos/config.yaml</code>) and a local (<code>.sos/config.yaml</code>) configuration files before executing any SoS script. The content of these files are available in the <code>CONFIG</code> variable.</p>

<p>SoS currently support the following configuration variables</p>

<ul>
  <li><code>sos_path</code> (default to <code>[]</code>): A list of directories from which sos will try to locate a script if the script is not in the current directory. For example, if you set <code>sos_path</code> to <code>['~/scripts']</code>, <code>sos run myscript.sos</code> will execute <code>~/scripts/myscript.sos</code> if <code>myscript.sos</code> does not exist in the current directory.</li>
  <li><code>sos_download_processes</code> (default to <code>5</code>): Number of download processes to download files specified in action <code>download</code>.</li>
</ul>

<p>The default value of 5 for option <code>sos_download_processes</code> is usually good enough but you can set it to a larger number if you have enough bandwidthonger time if you are working on a machine with, for example, slow disk access. To change this option, you will need to run</p>

<pre><code class="language-python">$ sos config --global --set 'sos_download_processes=10'
</code></pre>

<h2 id="inspect-and-prepare-workflows"><code>inspect</code> and <code>prepare</code> workflows</h2>

<h2 id="alternative-search-path-path">Alternative search path <code>$PATH</code></h2>

<p>SoS executes command in the environment the script is executed.</p>

                            </div>
                            <button type="button" class="btn btn-default" data-dismiss="modal"><i class="fa fa-times"></i> Close</button>
                        </div>
                    </div> 
                    <div class="footer-wrapper">
                      <footer role="contentinfo">
                        <span>&copy; 2016 Bo Peng.</span>
                      </footer>
                    </div>
                </div>
            </div>
        </div>
    </div> 



 
   <div class="portfolio-modal modal fade" id="Quick-start" tabindex="-1" role="dialog" aria-hidden="true">
        <div class="modal-content">
            <div class="close-modal" data-dismiss="modal">
                <div class="lr">
                    <div class="rl">
                    </div>
                </div>
            </div>
            <div class="container">
                <div class="row">
                 <div class="col-lg-8 col-lg-offset-2">
                        <h1>1.Quick-Start</h1>
                        <div class="modal-body">
                         
                            <hr class="star-primary">
                            <p>A SoS script consists of one or more scripts, comments and optional SoS directives. In its simplest form, a sos script is simply a series of scripts that can be executed sequentially by different intepreters.</p>

<p>Let us assume that you are a bioinformaticist needed to compare the expression levels between two samples. After reading some online tutorials, you ended up with some working commands</p>

<pre><code class="language-python"># index reference genome
STAR --runMode genomeGenerate --genomeFastaFile human38.fasta --genomeDir STAR_index
# align reads to the reference genome
STAR --genomeDir STAR_index --outSAMtype BAM SortedByCoordinate --readFilesIn control.fasta \
    --quantMode GeneCounts --outFileNamePrefix aligned/control
STAR --genomeDir STAR_index --outSAMtype BAM SortedByCoordinate --readFilesIn mutated.fasta \
    --quantMode GeneCounts --outFileNamePrefix aligned/mutated
</code></pre>

<p>The first command builds an index of the reference genome in preparation for the latter steps, the second command aligns reads from the first sample to the reference genome, and the third command aligns reads from the second sample to the reference genome. Do not panic if you do not know what these commands are doing, this is just an example.</p>

<p>These commands generate, among other files, two files named <code>aligned/control.out.tab</code> and <code>aligned/mutated.out.tab</code> with expression counts of all genes. You then wrote a <a href="https://www.r-project.org/">R</a> script to analyze the results, something like</p>

<pre><code class="language-R">control.count &lt;- read.table('aligned/control.out.tab')
mutated.count &lt;- read.table('aligned/mutated.out.tab')
# normalize, compare, output etc, ignored.
pdf('myfigure.pdf')
# plot results
dev.off()
</code></pre>

<h1 id="your-first-sos-script">Your first SoS script</h1>

<p>The project completed successfully and you needed to archive the scripts for later reference. Instead of having two files lying around with perhaps another <code>README</code> file to describe what you have done, you can write a single SoS script named <code>myanalysis.sos</code> with content</p>

<pre><code class="language-python">#!/usr/bin/env sos-runner
#fileformat=SOS1.0

# This script aligns raw reads of a control and a mutated sample
# to the reference genome and compare the expression values
# of the samples at genes A, B and C.

run:
# index reference genome
STAR --runMode genomeGenerate --genomeFastaFile human38.fasta --genomeDir STAR_index

# align reads to the reference genome
STAR --genomeDir STAR_index --outSAMtype BAM SortedByCoordinate  --readFilesIn control.fasta \
    --quantMode GeneCounts --outFileNamePrefix aligned/control
STAR --genomeDir STAR_index --outSAMtype BAM SortedByCoordinate --readFilesIn mutated.fasta \
    --quantMode GeneCounts --outFileNamePrefix aligned/mutated

R:
control.count &lt;- read.table('aligned/control.out.tab')
mutated.count &lt;- read.table('aligned/mutated.out.tab')
# normalize, compare, output etc, ignored.
pdf('myfigure.pdf')
# plot results
dev.off()
</code></pre>

<p>Here <strong><code>run</code></strong> and <strong><code>R</code></strong> are SoS actions that executes the following scripts in <code>bash</code> and <code>R</code> respectively. The scripts are included verbatim and end after reaching another SoS action or directive. It is however clearer to indent the script and write your SoS script as:</p>

<pre><code class="language-python">#!/usr/bin/env sos-runner
#fileformat=SOS1.0

# This script aligns raw reads of a control and a mutated sample
# to the reference genome and compare the expression values
# of the samples at genes A, B and C.

run:
    # index reference genome
    STAR --runMode genomeGenerate --genomeFastaFile human38.fasta --genomeDir STAR_index

    # align reads to the reference genome
    STAR --genomeDir STAR_index --outSAMtype BAM SortedByCoordinate  --readFilesIn control.fasta \
        --quantMode GeneCounts --outFileNamePrefix aligned/control
    STAR --genomeDir STAR_index --outSAMtype BAM SortedByCoordinate --readFilesIn mutated.fasta \
        --quantMode GeneCounts --outFileNamePrefix aligned/mutated

# compare expression values
R:
    control.count &lt;- read.table('aligned/control.out.tab')
    mutated.count &lt;- read.table('aligned/mutated.out.tab')
    # normalize, compare, output etc, ignored.
    pdf('myfigure.pdf')
    # plot results
    dev.off()
</code></pre>

<p>The scripts in this case end with the end of indentation so that you can add some comments for the <code>R</code> script without being considered as part of the previous script.</p>

<p>You can execute the shell and R scripts defined in this SoS script sequentially by running command</p>

<pre><code class="language-python">sos run myanalysis
</code></pre>

<p>or simply</p>

<pre><code class="language-python">myanalysis.sos
</code></pre>

<p>if you give <code>myanalyis.sos</code> executable permission (<code>chmod +x myanalysis.sos</code>).</p>

<h1 id="separate-scripts-into-steps">Separate scripts into steps</h1>

<p>Because the scripts perform different tasks, it is logically clearer to separate them into different <strong>steps</strong>. Actually, because the first reference-generating command is not a data processing step, it makes sense to separate it into two scripts. This can be done by inserting step headers to the script as follows:</p>

<pre><code class="language-python">#!/usr/bin/env sos-runner
#fileformat=SOS1.0

# This script aligns raw reads of a control and a mutated sample
# to the reference genome and compare the expression values
# of the samples at genes A, B and C.

[1]
# index reference genome
run:
    STAR --runMode genomeGenerate --genomeFastaFile human38.fasta --genomeDir STAR_index

[2]
# align reads to the reference genome
run:
    STAR --genomeDir STAR_index --outSAMtype BAM SortedByCoordinate  --readFilesIn control.fasta \
        --quantMode GeneCounts --outFileNamePrefix aligned/control
    STAR --genomeDir STAR_index --outSAMtype BAM SortedByCoordinate --readFilesIn mutated.fasta \
        --quantMode GeneCounts --outFileNamePrefix aligned/mutated

[3]
# compare expression values
R:
    control.count &lt;- read.table('aligned/control.out.tab')
    mutated.count &lt;- read.table('aligned/mutated.out.tab')
    # normalize, compare, output etc, ignored.
    pdf('myfigure.pdf')
    # plot results
    dev.off()
</code></pre>

<p>Now, when you execute the script with command</p>

<pre><code class="language-python">sos run myscript
</code></pre>

<p>SoS will display <code>default_1</code>, <code>default_2</code> and <code>default_3</code> to report progress. The comments after section heads are considered step comments and will also be displayed during execution.</p>

<h1 id="make-the-script-work-for-other-input-files">Make the script work for other input files</h1>

<p>After a while, before you almost forgot about this analysis, you needed to analyze another pair of samples. You could copy <code>myanalysis.sos</code> to <code>myanalysis2.sos</code>, change filenames and run it, but an easier way is to change your SoS file to accommodate other input files. This can be done by defining a command line argument and passing files name to a <strong>SoS variable</strong>:</p>

<pre><code class="language-python">#!/usr/bin/env sos-runner
#fileformat=SOS1.0

# This script aligns raw reads of a control and a mutated sample
# to the reference genome and compare the expression values
# of the samples at genes A, B and C.

# Two input files in .fasta formats. The first one for control sample
# and the second one for mutated sample.
parameter: fasta_files=['control.fasta', 'mutated.fasta']

[1]
# index reference genome
run:
    STAR --runMode genomeGenerate --genomeFastaFile human38.fasta --genomeDir STAR_index

[2]
# align reads to the reference genome
run:
    STAR --genomeDir STAR_index --outSAMtype BAM SortedByCoordinate \
        --readFilesIn ${fasta_files[0]} --quantMode GeneCounts \
        --outFileNamePrefix aligned/control
    STAR --genomeDir STAR_index --outSAMtype BAM SortedByCoordinate \
        --readFilesIn ${fasta_files[1]} --quantMode GeneCounts \
        --outFileNamePrefix aligned/mutated

[3]
# compare expression values
R:
    control.count &lt;- read.table('aligned/control.out.tab')
    mutated.count &lt;- read.table('aligned/mutated.out.tab')
    # normalize, compare, output etc, ignored.
    pdf('myfigure.pdf')
    # plot results
    dev.off()
</code></pre>

<p>A command line argument <code>fasta_files</code> is defined in with <code>parameter</code> keyword. With this definition, you can pass two filenames to variable <code>fasta_files</code> from command line</p>

<pre><code class="language-python">sos run myanalysis --fasta_files control1.fasta control2.fasta
</code></pre>

<p><code>${fasta_files[0]}</code> and <code>${fasta_files[1]}</code> in command <code>STAR --genomeDir
...</code> will be replaced with their values before the commands are executed. Here <code>fasta_files[0]</code> and <code>fasta_files[1]</code> are Python expressions that will be evaluated during execution.</p>

<h1 id="ignore-steps-that-do-not-need-to-be-rerun">Ignore steps that do not need to be rerun</h1>

<p>Although the SoS script now accepts command line arguments, it is still no more than a compilation of scripts and you immediately realized that it is a waste of time to execute the first command each time. To solve this problem, you can convert the SoS script to a real workflow by telling SoS the input and output of each step:</p>

<pre><code class="language-python">#!/usr/bin/env sos-runner
#fileformat=SOS1.0

# This script aligns raw reads of a control and a mutated sample
# to the reference genome and compare the expression values
# of the samples at genes A, B and C.

# Two input files in .fasta formats. The first one for control sample
# and the second one for mutated sample.
parameter: fasta_files=['control.fasta', 'mutated.fasta']

[1]
# create a index for reference genome
output: 'STAR_index/chrName.txt'
run:
    STAR --runMode genomeGenerate --genomeFastaFile human38.fasta --genomeDir STAR_index

[2]
# align the reads to the reference genome
input:    fasta_files
depends:  'STAR_index/chrName.txt'
output:   ['aligned/control.out.tab', 'aligned/mutated.out.tab']
run:
    STAR --genomeDir STAR_index --outSAMtype BAM SortedByCoordinate  --readFilesIn ${input[0]}  \
        --quantMode GeneCounts --outFileNamePrefix aligned/control
    STAR --genomeDir STAR_index --outSAMtype BAM SortedByCoordinate --readFilesIn ${input[1]}  \
        --quantMode GeneCounts --outFileNamePrefix aligned/mutated

[3]
# compare expression values
output: 'myfigure.pdf'
R:
    control.count &lt;- read.table('${input[0]}')
    mutated.count &lt;- read.table('${input[1]}')
    # normalize, compare, output etc, ignored.
    pdf('${output}')
    # plot results
    dev.off()
</code></pre>

<p>Here we</p>

<ul>
  <li>Use <strong>output directive</strong> to specify the expected output of all steps.</li>
  <li>Use <strong>input directive</strong> to specify the input of step 2. Step 1 by default has no input and input for step 3 by default is the output of step 2, its previous step.</li>
  <li>Use <strong>depends directive</strong> to let step 2 depend on the output of step 1.</li>
  <li>Use <code>${input[0]}</code> and <code>${input[1]}</code> in step 2 and 3 because these steps now have properly-defined <code>input</code>. This variable is defined by step input as the input file of the step.</li>
</ul>

<p>With such information, when you run the same command</p>

<pre><code class="language-python">sos run myanalysis --input control1.fasta control2.fasta
</code></pre>

<p>SoS will ignore step 1 if this step has been run with output <code>STAR_index/chrName.txt</code>. The same happens to step 2 and 3 so all steps will be ignored if you run the script repeatedly with the same input and processing scripts. SoS uses <strong>runtime signature</strong> for each step and will re-run the step if and only if the content or filename of input, output files or the processing scripts are changed.</p>

<h1 id="use-make-rule-to-define-resource-providing-steps">Use make-rule to define resource-providing steps</h1>

<p>Instead of using runtime signature to avoid re-running the first step, we can also make step 1 an optional step that will be executed only necessary. That is to say, we can define this step as an <code>auxiliary step</code> that will only be called when the file it <strong>provides</strong> does not exist.</p>

<pre><code class="language-python">#!/usr/bin/env sos-runner
#fileformat=SOS1.0

# This script aligns raw reads of a control and a mutated sample
# to the reference genome and compare the expression values
# of the samples at genes A, B and C.

# Two input files in .fasta formats. The first one for control sample
# and the second one for mutated sample.
parameter: fasta_files=['control.fasta', 'mutated.fasta']

[build_index: provides='STAR_index/chrName.txt']
# create a index for reference genome
run:
    STAR --runMode genomeGenerate --genomeFastaFile human38.fasta --genomeDir STAR_index

[1]
# align the reads to the reference genome
input:    fasta_files
depends:  'STAR_index/chrName.txt'
output:   ['aligned/control.out.tab', 'aligned/mutated.out.tab']
run:
    STAR --genomeDir STAR_index --outSAMtype BAM SortedByCoordinate  --readFilesIn ${input[0]}  \
        --quantMode GeneCounts --outFileNamePrefix aligned/control
    STAR --genomeDir STAR_index --outSAMtype BAM SortedByCoordinate --readFilesIn ${input[1]}  \
        --quantMode GeneCounts --outFileNamePrefix aligned/mutated

[2]
# compare expression values
output: 'myfigure.pdf'
R:
    control.count &lt;- read.table('${input[0]}')
    mutated.count &lt;- read.table('${input[1]}')
    # normalize, compare, output etc, ignored.
    pdf('${output}')
    # plot results
    dev.off()
</code></pre>

<p>The new script consists of two steps, and an auxiliary step <code>build_index</code> that will only be executed when <code>STAR_index/chrName.txt</code> is not available. A slight difference between this version and the previous one is that while the previous version will always execute the first step, this version will not execute it if <code>STAR_index/chrName.txt</code> has been generated before in any way, perhaps not by SoS.</p>

<h1 id="execute-long-running-jobs-externally">Execute long-running jobs externally</h1>

<p>The first step will takes a long time to execute. Instead of executing them by SoS, you might want to submit the commands to a job-queue and be executed and monitored externally. This is especially useful when you need to execute multiple SoS workflows on a cluster-based system. Without going through the details on how to set up your job-queue (see another tutorial for details), it is almost trivial to modify your script to define part of a <strong>step process</strong> to a <strong>task</strong>:</p>

<pre><code class="language-python">#!/usr/bin/env sos-runner
#fileformat=SOS1.0

# This script aligns raw reads of a control and a mutated sample
# to the reference genome and compare the expression values
# of the samples at genes A, B and C.

# Two input files in .fasta formats. The first one for control sample
# and the second one for mutated sample.
parameter: fasta_files=['control.fasta', 'mutated.fasta']

[build_index: provides='STAR_index/chrName.txt']
# create a index for reference genome
run:
    STAR --runMode genomeGenerate --genomeFastaFile human38.fasta --genomeDir STAR_index

[1]
# align the reads to the reference genome
input:    fasta_files
depends:  'STAR_index/chrName.txt'
output:   ['aligned/control.out.tab', 'aligned/mutated.out.tab']
task:
run:
    STAR --genomeDir STAR_index --outSAMtype BAM SortedByCoordinate  --readFilesIn ${input[0]}  \
        --quantMode GeneCounts --outFileNamePrefix aligned/control
    STAR --genomeDir STAR_index --outSAMtype BAM SortedByCoordinate --readFilesIn ${input[1]}  \
        --quantMode GeneCounts --outFileNamePrefix aligned/mutated

[2]
# compare expression values
output: 'myfigure.pdf'
R:
    control.count &lt;- read.table('${input[0]}')
    mutated.count &lt;- read.table('${input[1]}')
    # normalize, compare, output etc, ignored.
    pdf('${output}')
    # plot results
    dev.off()
</code></pre>

<p>As you can see, the only difference is the insertion of <code>task:</code> directive before <code>run</code>. Now, when you execute the command, the <code>STAR</code> commands will be executed externally and you can monitor the status of the jobs using your web browser. If no job-queue is set up, the command will be executed in a separate process.</p>

<h1 id="execute-steps-in-parallel">Execute steps in parallel</h1>

<p>Although the two steps of this example have to be executed sequentially, the first step runs the <code>STAR</code> command twice on two input files, and can be executed in parallel. You can tell this to SoS by modifying the script as follows</p>

<pre><code class="language-python">#!/usr/bin/env sos-runner
#fileformat=SOS1.0

# This script aligns raw reads of a control and a mutated sample
# to the reference genome and compare the expression values
# of the samples at genes A, B and C.

# Two input files in .fasta formats. The first one for control sample
# and the second one for mutated sample.
parameter: fasta_files=['control.fasta', 'mutated.fasta']

[build_index: provides='STAR_index/chrName.txt']
# create a index for reference genome
run:
    STAR --runMode genomeGenerate --genomeFastaFile human38.fasta --genomeDir STAR_index

[1]
# align the reads to the reference genome
sample_type = ['control', 'mutated']
input:    fasta_files, group_by='single', paired_with='sample_type'
depends:  'STAR_index/chrName.txt'
output:   'aligned/${_sample_type}.out.tab'

task:     concurrent=True
run:
    STAR --genomeDir STAR_index --outSAMtype BAM SortedByCoordinate  --readFilesIn ${_input!q}  \
        --quantMode GeneCounts --outFileNamePrefix aligned/${_sample_type}

[2]
# compare expression values
output: 'myfigure.pdf'
R:
    control.count &lt;- read.table('${input[0]}')
    mutated.count &lt;- read.table('${input[1]}')
    # normalize, compare, output etc, ignored.
    pdf('${output}')
    # plot results
    dev.off()
</code></pre>

<p>Here we</p>

<ol>
  <li>Use option <code>group_by='single'</code> to pass input one by one to action. The action will be executed twice with <code>_input</code> set to the first and second input file respectively.</li>
  <li>Define a variable <code>sample_type</code> and pair it with input files (option <code>paired_with</code>). This will generate a variable <code>_sample_type</code> for each input file so <code>_sample_type</code> will be <code>control</code> for the first input file, and <code>mutated</code> for the second.</li>
  <li>Use <code>${_input}</code> and <code>${_sample_type}</code> to define partial <code>output</code>.</li>
  <li>Use <code>${_input!q}</code> instead of <code>${_input}</code> in the script. This is a small trick to shell-quote filenames so that filenames with spaces and other special characters can be properly quoted in shell commands.</li>
</ol>

<p>Now, if you execute the script with option <code>-j 2</code> (2 concurrent processes),</p>

<pre><code class="language-python">sos run myanalysis.sos --input control1.fasta control2.fasta -j 2
</code></pre>

<p>the second step submit two jobs to job-queue, or execute them in two separate processes if a job-queue is not setup.</p>

<p>We have showed you multiple versions of the same SoS script, each using more features of SoS. This actually demonstrates one of the advantages of the SoS system, namely you can start using SoS in minutes without knowing any of its advanced features, and gradually improve your script if needed.</p>

                            </div>
                            <button type="button" class="btn btn-default" data-dismiss="modal"><i class="fa fa-times"></i> Close</button>
                        </div>
                    </div> 
                    <div class="footer-wrapper">
                      <footer role="contentinfo">
                        <span>&copy; 2016 Bo Peng.</span>
                      </footer>
                    </div>
                </div>
            </div>
        </div>
    </div> 


   <div class="portfolio-modal modal fade" id="Using-SoS-with-iPython" tabindex="-1" role="dialog" aria-hidden="true">
        <div class="modal-content">
            <div class="close-modal" data-dismiss="modal">
                <div class="lr">
                    <div class="rl">
                    </div>
                </div>
            </div>
            <div class="container">
                <div class="row">
                 <div class="col-lg-8 col-lg-offset-2">
                        <h1>2.Using-SoS-with-iPython</h1>
                        <div class="modal-body">
                         
                            <hr class="star-primary">
                            <p>Using sos magic within iPython allows you to run sos within an iPython session. It does not provide a full-blown SoS system but on the other hand you can use all the iPython features that you are familiar with.</p>

<h2 id="setting-up-ipython">Setting up ipython</h2>

<p>sos magic is installed by default when you install sos. A profile has been created for you so that you can use it by command</p>

<pre><code class="language-python">ipython --profile sos 
</code></pre>

<p>You can also load the extension using command <code>%load_ext sos_magic</code> after ipython starts in its default profile, or edit <code>~/.ipython/profile_default/ipython_config.py</code> and add <code>sos_magic</code> to <code>c.InteractiveShellApp.extensions</code> so that the extension is loaded automatically to your default profile. The latter is useful if you would like to use the ipython mode in Jupyter notebook (<code>jupyter notebook</code> and select python as your kernel).</p>

<h2 id="sos-ipython-magics">SoS ipython magics</h2>

<h3 id="magic-sosdict">Magic <code>sosdict</code></h3>

<p>Let us starts with the basic. When you start an <code>ipython</code> interactive shell through <code>ipython</code> command or <a href="https://ipython.org/notebook.html">ipython/Jupyter notebook</a>, you are given a cell that you can enter python expression or statements, or iPython magic. There are two kinds of magics</p>

<ul>
  <li><strong>Line magic</strong> that starts with <code>%</code> or starts at the beginning of line without <code>%</code> if <code>automagic</code> is set to <code>True</code> (default). Line magic takes the words after it as parameter.</li>
  <li><strong>Cell magic</strong> that starts with <code>%%</code> that takes both the words and lines after the magic word.</li>
</ul>

<p>If the <code>sos_magic</code> extension is loaded, it already has a SoS dictionary. You can get the dictionary using line magic <code>%sosdict</code>, for example</p>

<pre><code class="language-python">In [1]: sosdict
Out[1]: {}
</code></pre>

<p>The dictionary is empty because we have not assigned anything to it. Let us run a sos statement</p>

<pre><code class="language-python">In [2]: sos a = 1

In [3]: sosdict
Out[3]: {'a': 1}
</code></pre>
<p>and you can see the sos dictionary contains one item. There are other usages of the <code>%sosdict</code> magic, you can get the dictionary by assigning the dictionary to a variable</p>

<pre><code class="language-python">In [4]: d = %sosdict

In [5]: d.keys()
Out[5]: dict_keys(['a'])
</code></pre>

<p>If you are interested in only a subset of variables, you can list them after <code>%sosdict</code></p>

<pre><code class="language-python">In [6]: d = %sosdict a

In [7]: d.keys()
Out[7]: dict_keys(['a'])
</code></pre>

<p>You can get the keys of the dictionary easier using</p>

<pre><code class="language-python">In [8]: %sosdict --keys
Out[8]: {'a'}
</code></pre>

<p>If after a while you would like to reset the dictionary and run a SoS script from fresh, you can reset the dictionary using option <code>--reset</code></p>

<pre><code class="language-python">In [9]: sosdict --reset

In [10]: sosdict
Out[10]: {}
</code></pre>

<p>The SoS dictionary actually contains other items such as all the SoS actions and functions. If you would like to see all of them, use option <code>--all</code>. For example,</p>

<pre><code class="language-python">In [11]: sosdict --keys --all
Out[11]: dict_keys(['fail_if', 'expand_pattern', 'SoS_Script', 'zsh', 'download', 'perl', '__builtins__', 'R', 'logger', 'ruby', 'docker_commit', 'Rmarkdown', 'python3', 'report', 'env_variable', 'stop_if', 'SoS_Action', 'sos_variable', 'tcsh', 'warn_if', 'dynamic', 'executable', '__interactive__', 'check_R_library', 'JavaScript', 'run', 'sos_handle_parameter_', 'get_output', 'node', 'docker_build', 'interpolate', 'python', 'sos_run', 'csh', 'bash', 'pandoc', 'sos_namespace_', 'execute_script', 'check_command', 'sh'])
</code></pre>

<p>In summary, the <code>%sosdict</code> magic accepts</p>

<pre><code class="language-python">%sosdict [-a|-all] [-k|--keys] [-r|--reset] [var1] [var2] ...
</code></pre>

<p>where</p>

<ul>
  <li><code>var1</code>, <code>var2</code> etc are name of variables. All variables will be displayed if no variable is specified.</li>
  <li><code>-a|-all</code>: list all dictionary keys, including SoS functions and variables.</li>
  <li><code>-k|--keys</code>: list only keys, not their values</li>
  <li><code>-r|--reset</code>: reset the dictionary to its original content (with only SoS internal values)</li>
</ul>

<h3 id="magic-sos">Magic <code>sos</code></h3>

<p>Magic <code>sos</code> can be either a line magic or cell magic. Using <code>sos</code> as a line magic, it simply executes the SoS (python) expression or statement in SoS. For example,</p>

<pre><code class="language-python">In [5]: sos a=10

In [6]: sos 'a + 100 = ${a+100}'
Out[6]: 'a + 100 = 110'

In [7]: sos b=['file1.txt', 'file2.txt']

In [8]: sos '${b!r,}'
Out[8]: "'file1.txt','file2.txt'"
</code></pre>

<p>If you would like to execute multi-line SoS statements or scripts, you will need to use the magic in cell mode (with <code>%%</code> prefix), for example,</p>

<pre><code class="language-python">In [9]: %%sos
run:
    echo "something"
   ....: 
something
</code></pre>
<p>runs a shell script within iPython. Similarly, you can run arbitrary shell, R, perl, python code in ipython/SoS, with string interpolation. Note that ipython already has a magic called <a href="https://ipython.org/ipython-doc/3/interactive/magics.html"><code>%%script</code></a> that allows you to execute scripts in a cell. The difference here is that SoS process passed script with string interpolation before executing it.</p>

<p>You can also input one or more SoS steps and run the workflow</p>

<pre><code class="language-python">In [10]: %%sos
   ...: resource   = '~/resources'
   ...: ref_genome = '${resource}/hg19'
   ...: parameter: rep=5
   ...: 

</code></pre>

<p>Here a SoS script with a parameter section is executed, but without parameter. The <code>%%sos</code> cell magic allows you to pass command line arguments after <code>%%sos</code>, for example,</p>

<pre><code class="language-python">In [11]: %%sos --rep 3
resource   = '~/resources'
ref_genome = '${resource}/hg19'
parameter: rep=5
   ...: 

In [12]: sos rep
Out[12]: 3
</code></pre>

<h3 id="magic-sospaste">Magic <code>sospaste</code></h3>

<p>You will soon notice a problem with the cell magic <code>%%sos</code> in that it does not accept blank new lines, which is problematic for large piece of code. In this case, you can use line magic <code>%sospaste</code> to read the content directly from clipboard. Using the same example, you can select the text (with newline), and run</p>

<pre><code class="language-python">In [5]: sospaste --rep 4
resource   = '~/resources'
ref_genome = '${resource}/hg19'

parameter: rep = 5
## -- End pasted text --
</code></pre>

<p>This is the most convenient way to execute pieces of SoS script in iPython and is used most frequently.</p>

<h3 id="magics-sosget-and-sosput">Magics <code>sosget</code> and <code>sosput</code></h3>

<p>Magics <code>sosget</code> and <code>sosput</code> are used to exchange variables between ipython and sos. For example, if you defined a variable in SoS and would like to use it in ipython, you can use magic <code>%sosget var1 var2</code> to copy the value of <code>var1</code> and <code>var2</code> from SoS dictionary to ipython namespace. Similarly, <code>%sosput</code> copies variables from the ipython namespace to SoS dictionary. For example,</p>

<pre><code class="language-python">In [1]: %sos a = 20

In [2]: %sos b = 'a**2 = ${a**2}'

In [3]: %sosget a b

In [4]: a
Out[4]: 20

In [5]: b
Out[5]: 'a**2 = 400'

In [6]: b = 'something else'

In [7]: %sosput b

In [8]: %sosdict
Out[8]: {'a': 20, 'b': 'something else'}
</code></pre>

<h3 id="magic-sosset">Magic <code>sosset</code></h3>

<p>If you are tired of entering certain SoS options after magic <code>%%sos</code> or <code>%sospaste</code>, you can set these options persistently using magic <code>sosset</code>. For example, if you set</p>

<pre><code class="language-python">sosset -v 3
</code></pre>

<p>SoS will be running in a more verbose mode and print more messages. Other acceptable parameters include</p>

<ul>
  <li><code>-c CONFIG</code>: read from configuration file</li>
  <li><code>-f</code>: force execution, regardless of signature</li>
  <li><code>-F</code>: construct signature from existing files</li>
</ul>

<h2 id="a-complete-example">A complete example</h2>

<p>Suppose you are working on a script</p>

<pre><code class="language-python">resource   = '~/resources'
ref_genome = '${resource}/hg19'

parameter: rep = 5

[1]
print(rep)
seq = range(rep)
input:  for_each='seq'

python:
  import time
  print('sleep {} seconds.'.format(_seq))
  time.sleep(_seq)
</code></pre>

<p>To check the global definition, you can copy and paste the definitions to ipython as</p>

<pre><code class="language-python">In [1]: %%sos
   ...: resource   = '~/resources'
   ...: ref_genome = '${resource}/hg19'
   ...: 

</code></pre>

<p>The statements are executed and you can check the result using</p>

<pre><code class="language-python">In [2]: sos ref_genome
Out[2]: '~/resources/hg19'
</code></pre>

<p>Now we select the global definitions and execute them</p>

<pre><code class="language-python">In [3]: sospaste
resource   = '~/resources'
ref_genome = '${resource}/hg19'

parameter: rep = 5
## -- End pasted text --

In [4]: sos rep
Out[4]: 5
</code></pre>

<p>We can see the default parameter is passed. If you want, you can set the parameter to 10,</p>

<pre><code class="language-python">In [5]: sospaste --rep 10
resource   = '~/resources'
ref_genome = '${resource}/hg19'

parameter: rep = 5
## -- End pasted text --

In [6]: sos rep
Out[6]: 10
</code></pre>

<p>Now, let us continue, select the next step, run <code>sospaste</code> and .  a big block of errors!</p>

<pre><code class="language-python">In [7]: sospaste 

[1]
print(rep)
seq = range(rep)
input:  for_each='seq'

python:
  import time
  print('sleep {} seconds.'.format(_seq))
  time.sleep(_seq)
## -- End pasted text --
INFO: Execute default_1: 
10
INFO: input:   []
Traceback (most recent call last):
  File "/var/folders/ys/gnzk0qbx5wbdgm531v82xxljv5yqy8/T/tmpm5g6f7k7.py", line 2, in &lt;module&gt;
    print('sleep {} seconds.'.format(_seq))
NameError: name '_seq' is not defined

# IGNORED

RuntimeError: Failed to execute process
"python(r'''import time
print('sleep {} seconds.'.format(_seq))
time.sleep(_seq)
''')
"
Failed to execute script. The script is saved to .sos/default_1.py. Please use command "python .sos/default_1.py" under /Users/bpeng1/SOS to test it.
</code></pre>

<p>Here 10 is printed so <code>rep</code> is valid. This means the <code>rep</code> we set last time is available and correct, then what might be the problem? Let us see what is saved in <code>.sos/default_1.py</code> using a bit magic of ipython</p>

<pre><code class="language-python">In [8]: cat .sos/default_1.py
import time
print('sleep {} seconds.'.format(_seq))
time.sleep(_seq)
</code></pre>

<p>The error seems to be obvious, we need to use <code>${_rep}</code> for the value to be passed through string interpolation. Let us make some changes to the script, selct and run <code>sospaste</code> again:</p>

<pre><code class="language-python">In [10]: sospaste

[1]
print(rep)
seq = range(rep)
input:  for_each='seq'

python:
  import time
  print('sleep ${_seq} seconds.')
  time.sleep(${_seq})
## -- End pasted text --
INFO: Execute default_1: 
10
INFO: input:   []
sleep 0 seconds.
sleep 1 seconds.
sleep 2 seconds.
sleep 3 seconds.
sleep 4 seconds.
sleep 5 seconds.
sleep 6 seconds.
sleep 7 seconds.
sleep 8 seconds.
sleep 9 seconds.
</code></pre>

<p>Using this method, you can execute your SoS script step by step and make sure everything works, before you execute it from a command line.</p>


                            </div>
                            <button type="button" class="btn btn-default" data-dismiss="modal"><i class="fa fa-times"></i> Close</button>
                        </div>
                    </div> 
                    <div class="footer-wrapper">
                      <footer role="contentinfo">
                        <span>&copy; 2016 Bo Peng.</span>
                      </footer>
                    </div>
                </div>
            </div>
        </div>
    </div> 


   <div class="portfolio-modal modal fade" id="SoS-Notebook-Using-Jupyter" tabindex="-1" role="dialog" aria-hidden="true">
        <div class="modal-content">
            <div class="close-modal" data-dismiss="modal">
                <div class="lr">
                    <div class="rl">
                    </div>
                </div>
            </div>
            <div class="container">
                <div class="row">
                 <div class="col-lg-8 col-lg-offset-2">
                        <h1>3.SoS-Notebook-Using-Jupyter</h1>
                        <div class="modal-body">
                         
                            <hr class="star-primary">
                            <p>Jupyter notebook is a web application that allows you to create documents with live code and results. The underlying interpreters of the cells are called <strong>kernel</strong>. For example, you can use a <code>python</code> kernel to interpret python expressions, or a <code>ir</code> kernel to interpret <code>R</code> commands. You can use sos within a <code>python</code> kernel by loading the <code>sos_magic</code> ipython extension, or use a <code>sos</code> kernel to create a complete SoS environment. In addition to the ability to execute SoS workflows within Jupyter, the SoS kernel allows you to start subkernels (e.g. <code>ir</code>) and switch between the subkernels in the same notebook. That allows you to freely mix code in different languages in the same notebook.</p>

<h2 id="installing-the-sos-kernel">Installing the SoS kernel</h2>

<p>The SoS installation process will install a SoS kernel by default. To verify if you have the sos kernel installed, type</p>

<pre><code class="language-python">$ jupyter kernelspec list
Available kernels:
  python3    /Users/bpeng1/bin/anaconda/lib/python3.5/site-packages/ipykernel/resources
  sos        /usr/local/share/jupyter/kernels/sos
</code></pre>
<p>and check if you have <code>sos</code> listed as one of the kernels. If not, try to re-install SoS using <code>pip3 install sos --upgrade</code> or clone SoS locally and run <code>python setup.py install</code>. At the end of the installation process, you should see</p>

<pre><code class="language-python">IPython kernel named "sos" is installed.
</code></pre>

<p>If the kernel is installed, fire Jupyter using command</p>

<pre><code class="language-python">$ jupyter notebook
</code></pre>

<p>from the <code>New</code> button to the top right corner, select <code>SoS</code>, and create a SoS notebook.</p>

<p>SoS uses <a href="http://www.imagemagick.org/script/index.php">imagemagick</a> and python package <a href="http://docs.wand-py.org/en/0.4.2/"><code>wand</code></a> to preview output files so it is recommended that you install imagemagick.</p>

<h2 id="convert-a-sos-workflow-to-jupyter-notebook">Convert a SoS workflow to Jupyter notebook</h2>

<p>You can convert an existing SoS script to the <code>.ipynb</code> format using command</p>

<pre><code class="language-python">$ sos convert myscript.sos --notebook myscript.ipynb
</code></pre>

<p>and open the resulting notebook from the web interface. The converter uses lines that starts with <code>%cell</code> to split SoS a script into cells in Jupyter notebook.</p>

<h2 id="using-the-sos-kernel">Using the SoS kernel</h2>

<p>You can enter any SoS expression and statements in a Jupyter notebook but each cell needs to be a valid SoS script. That is to say,</p>

<ul>
  <li>
    <p>You can enter any Python expression and statements. For example, you can write pieces of SoS code</p>

    <p><code>python
  res_path = '/path/to/resource'
  ref_genome = '${res_path}/hg19'
 </code>
  and evaluate them.</p>
  </li>
  <li>
    <p>You can enter any SoS actions in function or script format. For example, you can execute a shell script using</p>

    <p><code>python
  run:
      echo Reference genome is located at ${ref_genome}
 </code></p>
  </li>
  <li>
    <p>You can create markdown cells to document your workflow. These cells are not evaluated by SoS so string interpolations are not available to these cells.</p>
  </li>
  <li>
    <p>Enter complete SoS steps with or without headers. Section headers are usually ignored because the cells are executed interactively regardless of the specified step indexes.</p>
  </li>
</ul>

<h2 id="sos-magics">SOS Magics</h2>

<p>In addition to SoS statements, you can use a few SoS magics in Jupyter notebook. They are similar to ipython sos magics but the syntax are a bit different. More specifically,</p>

<ul>
  <li>
    <p><code>%dict [reset|keys|all]</code></p>

    <p>The <code>%dict</code> magic lists or rests the content of SoS dict, using syntax</p>

    <p><code>
  %dict [-a|-all] [-k|--keys] [-r|--reset] [var1] [var2] ...
 </code></p>

    <p>where</p>

    <ul>
      <li><code>var1</code>, <code>var2</code> etc are name of variables. All variables will be displayed if no variable is specified.</li>
      <li><code>-a|-all</code>: list all dictionary keys, including SoS functions and variables.</li>
      <li><code>-k|--keys</code>: list only keys, not their values</li>
      <li><code>-r|--reset</code>: reset the dictionary to its original content (with only SoS internal values)</li>
    </ul>
  </li>
  <li>
    <p><code>%run [command-options] [workflow-options]</code></p>

    <p>The <code>%sos</code> magic allows you to specify SoS options such as <code>-v</code> (verbosity), <code>-j</code> (max number of jobs), and workflow options as defined by <code>parameter:</code> keyword.</p>
  </li>
  <li>
    <p><code>%paste [command-options] [workflow-options]</code></p>

    <p>This magic is used only for SoS kernel on a console (e.g. <code>ipython qtconsole --kernel sos</code>) where you can grab and execute content from the clipboard.</p>
  </li>
  <li>
    <p><code>%set [command-options] [workflow-options]</code></p>

    <p>The <code>%set</code> magic sets a persistent sos options so you do not have to enter them each time after <code>%run</code> or <code>%paste</code>. For example, if you set <code>%set -v3</code>, you can execute all cells in the notebook at verbosity level 3 (<code>DEBUG</code>).</p>
  </li>
  <li>
    <p><code>%matplotlib [GUI]</code></p>

    <p>Similar to ipythons <code>matplotlib</code> magic, the <code>%matplotlib inline</code> magic allows the display of matplotlib figures inline in Jupyter notebook or qtconsole.</p>
  </li>
  <li>
    <p><code>%use kernel</code> 
 This magic starts (or switches to) an alternative jupyter kernel to which expanded code (with string interpolation) will be sent. Kernel set by <code>%use kernel</code> will be persistent, affecting all subsequent input, including text pasted using magic <code>%paste</code>.</p>

    <p><code>kernel</code> can be any <a href="https://github.com/ipython/ipython/wiki/IPython-kernels-for-other-languages">Jupyter supported kernels</a>. In particular, you can use <code>ir</code> for R kernel (alias <code>R</code> is also acceptable), <code>python3</code> for python3, <code>iperl</code> for perl, <code>bash</code> for bash. Please refer to <a href="https://github.com/IRkernel/IRkernel">IRKernel</a> for instructions on how to install the <code>ir</code> kernel for <code>R</code>.</p>

    <p><code>%use</code> magic accepts parameters <code>--in (-i)</code> and <code>--out (-o)</code>, which are variables that will be transferred to the subkernel after starting (or switching to) the subkernel, and variables that will be transferred back to SoS before magic <code>%use sos</code>. More specifically,</p>
  </li>
</ul>

<pre><code class="language-python">  %use R -i var1 var2 -o var3
  do_something
  --- (another cell)
  %use sos
  ```  

  is equivalent to

```python
  %use R
  %get var1 var2
  do_something
  --- (another cell)
  %put var3
  %use sos
  ```  

* `%with kernel`

  `%with kernel` also starts (or switches to) a subkernel, but kernel specified by `%with kernel` will be reset as soon as the cell is executed. This magic is therefore suitable for a quick errand to another kernel. This magic also accepts options `--in` (or `-i`) and `--out` (or `-o`), so you could use

```python
  %with R -i n -o rn
  rn &lt;- rnorm(n)
  ```

  to get a list of normally distributed numbers using R's `rnorm` function.

* `%restart kernel`

  `%restart` specified kernel without affecting the current working kernel. 

* `%put` (`ir` and `python` kernels only)

    Magics `%put` are used to return results from subkernel to SoS kernel when the current kernel is a subkernel started by magic `%use`. Only the `ir` and `python` kernels are supported now. Whereas there is virtually no limit on exchangeable datatypes between `SoS` and `python` kernels, there are no perfect map between Python (SoS) and `R` datatypes so SoS tries to find the best match between data types. More specifically,
  
  | R  |  length (n) |   Python |
  | --- | --- |---|
  | NULL | |    None |
  | logical |  1 |  boolean |
  | integer |  1 |  integer |
  | numeric |  1 |  double |
  | character |  1 |  unicode |
  | logical |  n &gt; 1 |  list |
  | integer |  n &gt; 1 |  list |
  | numeric |  n &gt; 1 |  list |
  | character |  n &gt; 1 |  list |
  | list without names |  n &gt; 0 |  list |
  | list with names |  n &gt; 0 |  dict |
  | matrix |  n &gt; 0 |  numpy.array |
  | data.frame |  n &gt; 0 |  DataFrame |

* `%get` (`ir` and `python` kernels only)

    Magics `%get` retrieve variables from the SoS to the current subkernel started by magic `%use`. Only the `ir` and `python` kernels are supported now. SoS tries to use the best matching data type for the conversion. More specifically,

  
  | Python  |  cond |   R |
  | --- | --- |---|
  | None | |    NULL |
  | boolean |   | logical |
  | integer |  |  integer |
  | float |  |  float |
  | str |  | character |
  | Sequence (list, tuple, ...) |  homogenous type |  c() |
  | Sequence (list, tuple, ...) |  multiple types |  list |
  | set |  |  list |
  | dictionary |  |  list with names |
  | numpy.ndarray |  | c() array |
  | numpy.matrix |  | matrix |
  | pandas.DataFrame |  |  R data.frame |

  Python objects in other datatypes are transferred as string `"Unsupported datatype"`.

* `!any-shell-command` such as `!ls` and `!pwd`

   If any other command is entered after `!`, sos will treat the rest of the line as a shell command and execute it. Only single-line commands are supported. String interpolation is supported. Most commands are executed in a separate process and have no impact on SoS, except for command `!cd` that changes the current working directory of SoS.

## Convert `.ipynb` files to `.sos`

You can save the notebook to a SoS script using `File` -&gt; `Download As` -&gt; `SoS`, or using command

```python
$ sos convert myscript.ipynb --sos myscript.sos
</code></pre>

<p>The conversion process will strip all results (which can be useful if you would like to version control your notebook without results) but keep cell information and SoS magics. However, <strong>SoS files converted from Jupyter notebook cannot be executed by the <code>sos</code> command if they contain SoS magics such as <code>%use R</code></strong> because of non-interactive execution of SoS steps and more rigorous structure required for SoS scripts. You will generally need to</p>

<ol>
  <li>Remove <code>%</code> magic from the notebook and replace cells with</li>
</ol>

<p><code>python
    %with R
      R code
   </code>
    to
<code>python
    R:
      R code
   </code>
    The latter quires the <code>R code</code> to be self-contained and does not require definitions from other cells.</p>

<ol>
  <li>Add section headers such as <code>[10]</code>, and <code>[human_10]</code> to the SoS steps if they are executed separately.</li>
</ol>

<p>These can be done either in Jupyter notebook or a text editor. Note that Jupyter notebooks converted from a working <code>.sos</code> file already have these features so they will work just fine if you convert back from Jupyter notebook, and you can keep your Jupyter notebook executable in batch mode by removing <code>%</code> magics after using them for debug, and use section headers for different steps.</p>

                            </div>
                            <button type="button" class="btn btn-default" data-dismiss="modal"><i class="fa fa-times"></i> Close</button>
                        </div>
                    </div> 
                    <div class="footer-wrapper">
                      <footer role="contentinfo">
                        <span>&copy; 2016 Bo Peng.</span>
                      </footer>
                    </div>
                </div>
            </div>
        </div>
    </div> 


   <div class="portfolio-modal modal fade" id="Using-Spyder-as-SoS-IDE" tabindex="-1" role="dialog" aria-hidden="true">
        <div class="modal-content">
            <div class="close-modal" data-dismiss="modal">
                <div class="lr">
                    <div class="rl">
                    </div>
                </div>
            </div>
            <div class="container">
                <div class="row">
                 <div class="col-lg-8 col-lg-offset-2">
                        <h1>4.Using-Spyder-as-SoS-IDE</h1>
                        <div class="modal-body">
                         
                            <hr class="star-primary">
                            <h2 id="using-sos-with-spyder">Using SoS with spyder</h2>

<table>
  <tbody>
    <tr>
      <td>[spyder</td>
      <td>https://pythonhosted.org/spyder/) is a Python GUI that works well with ipython, and therefore SoS if you configure it properly. Spyder should be readily available if you use Anaconda python, or you can install spyder according to instructions on the <a href="https://pythonhosted.org/spyder/">spyder website</a>.</td>
    </tr>
  </tbody>
</table>

<p>Disclaimer: Spyder does not support third-party kernels and does not yet support <code>.sos</code> files. The SoS installer tries to patch Spyder so that it can recognize <code>.sos</code> files and the <code>%cell</code> syntax. These are supposed to change after Spyder provides native support for SoS in its next major release.</p>

<p>Therefore two ways to use spyder with SoS:</p>

<ol>
  <li>
    <p>Connect spyder to an existing qtconsole with sos kernel</p>

    <ul>
      <li>Start a qtconsole with sos kernel using command <code>jupyter qtconsole --kernel sos</code></li>
      <li>Record connection file id from command line, or by running <code>%connect_info</code> from the console if you cannot find it.</li>
      <li>Start <code>spyder</code>, select <code>consoles</code> -&gt; <code>Connect an existing kernel</code>. Put connection id and connect.</li>
    </ul>

    <p>It is certainly possible to set up a remote Jupyter server and connect to a remote SoS kernel but this usage is beyond the scope of this tutorial.</p>
  </li>
  <li>
    <p>Make sos your default kernel by adding</p>
  </li>
</ol>

<p><code>python
    c.IPKernelApp.kernel_class =  'pysos.kernel.SoS_Kernel'
   </code>
    to <code>~/.ipython/profile_default/ipython_config.py</code> . You can then start <code>spyder</code> and use spyder as usual.</p>


                            </div>
                            <button type="button" class="btn btn-default" data-dismiss="modal"><i class="fa fa-times"></i> Close</button>
                        </div>
                    </div> 
                    <div class="footer-wrapper">
                      <footer role="contentinfo">
                        <span>&copy; 2016 Bo Peng.</span>
                      </footer>
                    </div>
                </div>
            </div>
        </div>
    </div> 


   <div class="portfolio-modal modal fade" id="A-Complete-Example" tabindex="-1" role="dialog" aria-hidden="true">
        <div class="modal-content">
            <div class="close-modal" data-dismiss="modal">
                <div class="lr">
                    <div class="rl">
                    </div>
                </div>
            </div>
            <div class="container">
                <div class="row">
                 <div class="col-lg-8 col-lg-offset-2">
                        <h1>5.A-Complete-Example</h1>
                        <div class="modal-body">
                         
                            <hr class="star-primary">
                            <h2 id="a-rna-seq-data-processing-workflow">A RNA Seq data processing workflow</h2>

<p>If you are curious about how to write large workflows using SoS or just how a long SoS script looks like, here is a RNA Seq data processing workflow written in SoS. The workflow consists of many steps and was written in a mixed style, namely a spine of forward-moving steps and several auxiliary steps that supply required information. The DAG of the workflow looks like</p>

<p><img src="/SOS/media/example_dag.png" alt="example_dag.png" class="img-responsive" /></p>

<p>although this DAG does not include many auxiliary steps (e.g. those that download required resources) because these steps were called once and would not be called once the resources are available. The DAG is generated using command</p>

<pre><code class="language-python">sos prepare RNASeq_hg10.sos
cat .sos/human_hg19.dot | dot -T png &gt; example_dag.png
</code></pre>

<p>This tutorial will divide the workflow into several parts and briefly describe each of them.</p>

<h3 id="header">Header</h3>

<p>The header of script contains description of the workflow.</p>

<pre><code class="language-python">
#!/usr/bin/env sos-runner
#fileformat=SOS1.0

# Workflow for a comprehensive analysis of paired-end RNA seq data
# that performs read-level quality assessment, alignment, alignment quality
# assessment, summary statistics, fusion detection and filtering, gene and exon
# level expression counts, and variant calling. The input of this pipeline
# should be a list of fastq or fastq.gz files with paired reads differentiated
# by _R1_ and _R2_ in filenames. &lt;p&gt;The output of the pipeline should be a
# directory to which all results will be written. The directory will be 
# created if it does not exist. Note that it is highly recommended that
# you execute the pipeline with a subset of samples (using option --sampling
# 1000000) to test the pipeline before you apply it to the full dataset. 
# Network connection will be needed for the first run (to download resources)
# but not for subsequent runs. For details of the pipeline including the
# accepted parameters, please run command "vtools show pipeline RNASeq_hg19_v3".

# human_hg19
# 
# This pipeline uses the hg19 version of the reference
# genome, genes downloaded from Illumina iGenome, and existing variants from
# the GATK resource bundle. This pipeline uses &lt;ul&gt;
#   &lt;li&gt;tophat 2.0.13 
#   &lt;li&gt;bowtie 1.1.1
#   &lt;li&gt;samtools 0.1.19
#   &lt;li&gt;picard 1.82
#   &lt;li&gt;GATK 3.3
#   &lt;li&gt;fastqc 0.11.2
#   &lt;li&gt;BEDTools 2.17.0
#   &lt;li&gt;RSeQC 2.4, and
#   &lt;li&gt;Oncofuse 1.0.7
#   &lt;/ul&gt;
# Other versions of the tools can be used if you execute the pipeline with option
# --strict_version False, but the pipeline might or might not work as expected. 
# Please download all these tools and make them available to the pipeline (set $PATH),
# and start the pipeline using command "vtools execute /path/to/this_file options".
# Noted that the precompiled version of tophat 2.0.13 can fail under mac so it is
# recommended that you compile tophat2 from source.

</code></pre>

<h3 id="inclusion-of-auxiliary-steps">inclusion of auxiliary steps</h3>

<pre><code class="language-python">%from install_ngs include *
</code></pre>

<h3 id="global-definitions">Global definitions</h3>

<pre><code class="language-python">
if os.path.isdir(os.path.expanduser('~/rsrch1/variant_tools_resource/pipeline_resource')):
    resource_dir  = os.path.expanduser('~/rsrch1/variant_tools_resource/pipeline_resource')
elif os.path.isdir(os.path.expanduser('~/.variant_tools/pipeline_resource')):
    resource_dir  = os.path.expanduser('~/.variant_tools/pipeline_resource')
else:
    fail_if(True, 'Unable to locate a resource directory')

# Path to picard jar files
parameter: picard_path = '~/bin/Picard'

# Path to GenomeAnalysisTK.jar
parameter: gatk_path = '~/bin/GATK'

# Path to Oncofuse.jar
parameter: oncofuse_path = '~/bin/Oncofuse'

# input fastq files
parameter: fastq_files = list

fail_if(len([x for x in fastq_files if '_R1_' in x]) == 0  \
    or len([x for x in fastq_files if '_R1_' in x]) != len([x for x in fastq_files if '_R2_' in x]),
    'Input file names are not paired fastq files differentiated by _R1_ and _R2_ in filename.')

# Name of the sample being processed, which will be used for filename
# and read group of aligned reads.
parameter: samplename = str

# If set to a positive number, this pipeline will sample specified
# number of reads (pairs) from the input. This is useful to test if the pipeline
# works for your environment, download all needed resources, created indexes,
# and check the quality of data.
parameter: sampling = 0


# An optional filename that contains detailed information
# about the sample. The sample sheet should be a tab or comma-delimited 
# file with a header, and one or more lines with one of which contain the 
# information of the sample. The information from the sample sheet will be
# written to the final report (summary.html).
parameter: sample_sheet = ''

#
# CONSTANT VALUES: not configurable in this pipeline
# 
pipeline_name     = 'RNASeq_hg19'
pipeline_version  = 'ver1.0 (2016/8/29'
reference_genome  = 'hg19'
reference_genes   = 'UCSC/Illumina iGenome'
# HG19 version of the resource files
gatk_resource_dir = '${resource_dir}/GATK/2.8/hg19/'
igenome_resource_dir = '${resource_dir}/iGenome'
ncbi_resource_dir = '${resource_dir}/NCBI'

igenome_url       = 'ftp://igenome:G3nom3s4u@ussd-ftp.illumina.com/Homo_sapiens/UCSC/hg19/Homo_sapiens_UCSC_hg19.tar.gz'
gene_info_url     = 'ftp://ftp.ncbi.nih.gov/gene/DATA/GENE_INFO/Mammalia/Homo_sapiens.gene_info.gz'
reference_dir     = '${igenome_resource_dir}/Homo_sapiens/UCSC/hg19/Sequence'
transcriptome_dir = '${igenome_resource_dir}/Homo_sapiens/UCSC/hg19/Annotation/transcriptome_index'
annotation_dir    = '${igenome_resource_dir}/Homo_sapiens/UCSC/hg19/Annotation'
refgene_txt       = '${annotation_dir}/Genes/refGene.txt'
genes_gtf         = '${annotation_dir}/Genes/genes.gtf'
# generated files'
refgene_bed       = '${annotation_dir}/Genes/refGene.bed'
genes_exon_bed    = '${annotation_dir}/Genes/genes_exon.bed'
genes_with_chr_gtf= '${annotation_dir}/Genes/genes_chr.gtf'
genesize_txt      = '${annotation_dir}/Genes/geneSize.txt'
# Download from UCSC'
ensgene_url       = 'http://hgdownload.cse.ucsc.edu/goldenpath/hg19/database/ensGene.txt.gz'
ensgene_txt       = '${annotation_dir}/Genes/ensGene.txt'
chromsize_url     = 'ftp://hgdownload.cse.ucsc.edu/goldenPath/hg19/database/chromInfo.txt.gz'
chromsize_txt     = '${annotation_dir}/chromsize.txt'
# GATK resource bundle'
gatk_url          = 'ftp://gsapubftp-anonymous@ftp.broadinstitute.org/bundle/2.8/hg19/'
# '
# Output directories'
output_dir        = samplename
temp_dir          = '${output_dir}/tmp'
alignment_out     = '${output_dir}/alignment'
alignmentqc_out   = '${output_dir}/alignment_qc'
fusion_out        = '${output_dir}/fusion'
readqc_out        = '${output_dir}/read_qc'
count_out         = '${output_dir}/counts'
variant_out       = '${output_dir}/variants'



[human_hg19_100: skip=sampling == 0, shared='fastq_files']
# Select a subset of reads if parameter --sampling is specified. Output will be 
# written to the project cache directory.
input:
    fastq_files

depends:
    executable('gunzip')

output:
    '${temp_dir}/${samplename}_R1_${sampling}.fastq',
    '${temp_dir}/${samplename}_R2_${sampling}.fastq'

run:
    gunzip -c ${' '.join(sorted([x for x in input if '_R1_' in x]))} | head -n ${4*sampling} &gt; ${output[0]}
    gunzip -c ${' '.join(sorted([x for x in input if '_R2_' in x]))} | head -n ${4*sampling} &gt; ${output[1]}

fastq_files = output


[human_hg19_110]
# Use fastqc to check the quality of input reads
#
input:
    fastq_files

depends:
    executable('fastqc'),
    executable('gunzip')

output:
    '${readqc_out}/${samplename}_fastqc.html',
    '${readqc_out}/${samplename}_fastqc.zip'

task:

run:
    # the input can be original .fast.gz file or extracted 
    ${"gunzip -c" if input[0].endswith(".gz") else "cat"} ${input} | fastqc stdin --outdir=${readqc_out}
    mv ${readqc_out}/stdin_fastqc.html  ${output[0]}
    mv ${readqc_out}/stdin_fastqc.zip ${output[1]}

[extract_fastqc_figures: provides=[os.path.join(readqc_out, x) for x in ('per_base_quality.png', 'per_sequence_gc_content.png', 'per_base_sequence_content.png')]]
input:
    '${readqc_out}/${samplename}_fastqc.zip'

import zipfile
import os
with zipfile.ZipFile(input[0]) as qczip:
    for name in qczip.namelist():
        if os.path.basename(name) in ['per_base_quality.png', 'per_sequence_gc_content.png', 'per_base_sequence_content.png']:
            qczip.extract(name, readqc_out)
            # move the file to top directory
            os.rename(os.path.join(readqc_out, name), os.path.join(readqc_out, os.path.basename(name)))


[download_ensgene: provides=genes_gtf]
download:  '${annotation_dir}/Genes/'
    ${ensgene_url}

[chr_gtf: provides = genes_with_chr_gtf]
# takes gtf and and chromosome names to gene name to avoid problems
# with genes with the same names on different chromosomes
input:
    genes_gtf

output:
    genes_with_chr_gtf

python:
    #
    #    Input format:
    #
    #        chr1 unknown exon 17233 17368 . - . gene_id "WASH7P"; gene_name "WASH7P"; transcript_id "NR_024540"; tss_id "TSS8151";
    #
    #    Output format:
    #           chr1 unknown exon 17233 17368 . - . gene_id "WASH7P"; gene_name "WASH7P.chr1"; transcript_id "NR_024540"; tss_id "TSS8151";
    # 
    import re
    pattern = re.compile(r'''(gene_id\s+['"]*)([\w.-]+)''')
    # only work on selected chromosomes
    allowed_chromosomes = set([str(x) for x in range(25)] + ['chr{}'.format(x) for x in range(25)] + \
                ['X', 'Y', 'M', 'chrX', 'chrY', 'chrM'])
    with open(input[0], 'rU') as ifile, open(output[0], 'w') as ofile:
       for line in ifile:
           if line.startswith('#'):
               continue
           ls = line.strip().split('\t')
           if ls[0] not in allowed_chromosomes:
               continue
           ofile.write(re.sub(pattern, r'\1\2.{}'.format(ls[0]), line)) 

[download_igenome: provides=refgene_bed]
download: dest_dir=igenome_resource_dir
    ${igenome_url}

[human_hg19_400: shared={'param': 'output[0]'}]
#
# Align a subset of reads to obtain two key parameters parameters
# --mate-inner-dist and --mate-std-dev for full scale alignment
#
input:
    fastq_files

depends:
    genes_with_chr_gtf,
    refgene_bed,
    executable('tophat2'),
    executable('bowtie'),
    executable('inner_distance.py')
    
output:
    '${temp_dir}/${samplename}.inner_distance_freq.txt'

test_files = '${temp_dir}/${samplename}_R1_test.fastq', '${temp_dir}/${samplename}_R2_test.fastq'

task:

run:
    ${'gunzip -c' if input[0].endswith('.gz') else 'cat'}  ${' '.join(sorted([x for x in input if '_R1_' in x]))} \
        | head -n 40000 &gt; ${test_files[0]}

    ${'gunzip -c' if input[0].endswith('.gz') else 'cat'}  ${' '.join(sorted([x for x in input if '_R2_' in x]))} \
        | head -n 40000 &gt; ${test_files[1]}

run:
    tophat2 --zpacker 0 -no-coverage-search --library-type fr-unstranded \
        --max-multihits 20 --prefilter-multihits --num-threads 8 \
        --GTF ${genes_with_chr_gtf} --segment-length 25 \
        --transcriptome-index ${transcriptome_dir}/genes --bowtie1 \
        --output-dir ${temp_dir} ${reference_dir}/BowtieIndex/genome \
        ${test_files[0]} ${test_files[1]}

run:
    inner_distance.py -r ${refgene_bed} -i ${temp_dir}/accepted_hits.bam -o ${temp_dir}/${samplename}


[human_hg19_440: shared={'accepted_hits':'output[0]', 'mate_inner_dist':'mate_inner_dist', 'mate_std_dev':'mate_std_dev'}]
#Alignment (Tophat2): Running tophat2 with fusion search.
#Alignment Preparation: Calculate mean and standard devision of inner distance and set parameters
# --mate-inner-dist and --mate-std-dev
import math
values = []
with open(param) as freq:
    for line in freq:
        s, e, c = [int(x) for x in line.split()]
        values.extend([int(s+e)/2.] * c)
if not values:
    # no mapped reads...
    mate_inner_dist = 50
    mate_std_dev = 20
    logger.info('Mean and standard deviation of inner distance is set to default values 50 and 20')
else:
    mean = sum(values)/len(values)
    mean_sq = sum([x*x for x in values])/len(values)
    sd = math.sqrt(mean_sq - mean*mean)
    mate_inner_dist = int(mean)
    mate_std_dev = int(sd)
    logger.info('Mean and standard deviation of inner distance is set to {} and {}'.format(int(mean), int(sd)))

input:
    fastq_files

depends:
    genes_with_chr_gtf

output:
    '${alignment_out}/accepted_hits.bam'

# use --zpacker 0 to disable compression of temporary file
# which makes the program run faster
run:
    if [[ -f ${alignment_out}/logs/run.log ]] &amp;&amp; [[ ! -f ${alignment_out}/accepted_hits.bam ]];  \
        then  \
        tophat2 --resume ${alignment_out} ||  \
        tophat2  \
        -no-coverage-search  \
        --library-type fr-unstranded  \
        --max-multihits 20  \
        --prefilter-multihits  \
        --num-threads 8  \
        --GTF ${genes_with_chr_gtf}  \
        --segment-length 25  \
        --keep-fasta-order  \
        --transcriptome-index ${transcriptome_dir}/genes  \
        --fusion-search --fusion-ignore-chromosomes chrM,M  \
        --fusion-min-dist 50000  \
        --bowtie1  \
        --mate-inner-dist ${mate_inner_dist}  \
        --mate-std-dev ${mate_std_dev}  \
        --output-dir ${alignment_out}  \
        ${reference_dir}/BowtieIndex/genome  \
        ${','.join(sorted([x for x in input if '_R1_' in x]))}  \
        ${','.join(sorted([x for x in input if '_R2_' in x]))}  \
        ;  \
    else   \
        tophat2  \
        -no-coverage-search  \
        --library-type fr-unstranded   \
        --max-multihits 20   \
        --prefilter-multihits  \
        --num-threads 8   \
        --GTF ${genes_with_chr_gtf}  \
        --segment-length 25   \
        --keep-fasta-order   \
        --transcriptome-index ${transcriptome_dir}/genes  \
        --fusion-search --fusion-ignore-chromosomes chrM,M   \
        --fusion-min-dist 50000  \
        --bowtie1   \
        --mate-inner-dist ${mate_inner_dist}  \
        --mate-std-dev ${mate_std_dev}  \
        --output-dir ${alignment_out}  \
        ${reference_dir}/BowtieIndex/genome  \
        ${','.join(sorted([x for x in input if '_R1_' in x]))}  \
        ${','.join(sorted([x for x in input if '_R2_' in x]))};  \
        fi 

 
[index_bam: provides='{filename}.bam.bai']
# Index generated bam file
input:
    '${filename}.bam'

run:
    samtools index ${input}


[human_hg19_460]
# Add ReadGroup information to sorted bam file.

# Read group identifier, which should be an unique identifier for each
# instrument run (e.g. flowcell number + lane name + number). The filename before
# the '_R1_' part will be used if this option is left unspecified.
parameter: rgid = 'NA'
warn_if(rgid == 'NA',
    'A default (NA) value is used for parameter --rgid (intrument run ID)')

# Platform/technology used to produce the read. Valid values include
# ILLUMINA, SOLID, LS454, HELICOS and PACBIO
parameter: rgpl = 'ILLUMINA'
# check input parameters
warn_if(not rgpl in ['ILLUMINA', 'SOLID', 'LS454', 'HELICOS', 'PACBIO'],
    'PL (platform) value if not one of ILLUMINA, SOLID, LS454, HELICOS and PACBIO')

# DNA preparation library identity.
parameter: rglb = 'NA'
warn_if(rglb == 'NA',
    'A default (NA) value is used for parameter --rglb (library identification string)')
#
# Platform unit (e.g. barcode). This field is not used by GATK.
parameter: rgpu = 'NA'
warn_if(rgpu == 'NA',
    'A default (NA) value is used for parameter --rgpu (platform unit, barcode)')


input:
     accepted_hits

output:
    '${temp_dir}/${samplename}-with_rg.bam'

run:
    java -Xmx16g -Xms512m   \
        -Djava.io.tmpdir=${temp_dir}   \
        -jar ${picard_path}/AddOrReplaceReadGroups.jar  \
        input=${input}  \
        OUTPUT=${temp_dir}/${samplename}-with_rg.bam  \
        VALIDATION_STRINGENCY=LENIENT  \
        RGID=${rgid}  \
        RGSM=${samplename}  \
        RGPL=ILLUMINA  \
        RGLB=${rglb}  \
        RGPU=${rgpu}  \
        RGCN=IPCT

[human_hg19_470: shared={'sorted_unique':'output[0]'}]
#Alignment: Remove fusion reads in preparation for variant calling.
# step 475, samtools index is defined above
output:
    '${alignment_out}/${samplename}-sorted_uniq_nonF.bam'

run:
    samtools view -h ${input} \
        | awk -F '\t' '{ if($0 ~ "^@") {print} else { for(i=12;i&lt;=NF;i++){ if ($i ~ "NH:i:1$"){print}} } }' \
        | awk '{ if($0 ~ "^@") {print} else { if ($0 !~ "XF:Z") {print} }  }' \
        | samtools view -bS - &gt; ${alignment_out}/${samplename}-sorted_uniq_nonF.bam


[human_hg19_480]
#Alignment: Count the number of junction reads. 
input:
    accepted_hits

output:
    '${alignment_out}/junction.count'

run:
    samtools view -h ${input} \
   | awk '$6 ~/N/' | awk '{ if ($9 ~ /^-/) {print $1"\t-"} else print $1"\t+" }' \
   | sort -T ${temp_dir} -u | wc -l &gt; ${alignment_out}/junction.count


[human_hg19_500 (gene body coverage): shared={'rseqc_pdf':'output'}]
# this version uses BAM files to count geneBody coverage, using .bw might be a more resource
# saving idea but needs more testing, especially because the output of bigWig is currently normalized
# and does not show the correct depth of coverage.
input:
    sorted_unique

depends:
    refgene_bed

output:
    '${alignmentqc_out}/${samplename}.geneBodyCoverage.curves.pdf'

# this job can take a long time to execute so we tend to execute as a separate job
task:
run:
    geneBody_coverage.py -r ${refgene_bed} -i ${input} -o ${alignmentqc_out}/${samplename}

#[human_hg19_560]
#RSeQC: Coverage of gene body
#RunCommand('geneBody_coverage2.py -r ${refgene_bed} -i ${input551: input551[0][:-3] + "bw"} -t pdf -o ${alignmentqc_out}/${samplename}',
#    output='${alignmentqc_out}/${samplename}.geneBodyCoverage.curves.pdf')

[human_hg19_510]
#RSeQC: Check if current sequencing depth is deep enough to perform alternative splicing analyses
#RSeQC: Annotating splicing event and splicing junction.
#RSeQC: Plot the distribution of inner distance between paired reads, which can be used to 
#   determine parameters --mate-inner-dist and --mate-std-dev of the pipeline.
#RSeQC: read duplication plot
#RSeQC: Read distribution (statistics)
#RSeQC: Generate scaled wig track for depth
input:
    accepted_hits

output:
    '${alignmentqc_out}/${samplename}.junctionSaturation_plot.pdf',
    '${alignmentqc_out}/${samplename}.splice_events.pdf',
    '${alignmentqc_out}/${samplename}.splice_junction.pdf',
    '${alignmentqc_out}/${samplename}.inner_distance_plot.pdf',
    '${alignmentqc_out}//${samplename}.DupRate_plot.pdf',
    '${alignmentqc_out}//${samplename}.read_distribution'

task:
run:
    junction_saturation.py  -r ${refgene_bed} -i ${input} -o ${alignmentqc_out}/${samplename}
    junction_annotation.py -r ${refgene_bed} -i ${input} -o ${alignmentqc_out}/${samplename}
    inner_distance.py -r ${refgene_bed} -i ${input} -o ${alignmentqc_out}/${samplename}
    read_duplication.py -i ${input} -o ${alignmentqc_out}/${samplename}
    read_distribution.py -r ${refgene_bed} -i ${input} &gt; ${alignmentqc_out}/${samplename}.read_distribution

[human_hg19_600 (SortSam): shared={'sorted_bam':'output'}]
#Count: Order reads by name so that the BAM file can be processed by htseq-count
input:
    accepted_hits

output:
    '${alignment_out}/${samplename}-sorted_by_id.bam'

task:
run:
    java -Xmx16g -Xms512m   \
        -Djava.io.tmpdir=${temp_dir}  \
        -jar ${picard_path}/SortSam.jar  \
        input=${input}  \
        OUTPUT=${alignment_out}/${samplename}-sorted_by_id.bam  \
        SO=queryname   \
        MAX_RECORDS_IN_RAM=1000000   \
        TMP_DIR=${temp_dir}  \
        VALIDATION_STRINGENCY=SILENT


[human_hg19_610 (htseq-count)]
# Running htseq-count to count reads for genes. This step uses the union
#    intersection mode, and a gene definition file with chromosome name added to gene_id.
input:
    sorted_bam

output:
    '${count_out}/${samplename}-gene_counts.tsv'

# We cannot use --order=pos because htseq-count would fail with an error message 
run:
    htseq-count --stranded=no --format=bam --type=exon \
     --mode=union ${input} ${genes_with_chr_gtf} \
     &gt; ${count_out}/${samplename}-gene_counts.tsv


[get_genesize: provides=genesize_txt]
#    Genes.gtf file in the format of
#
#    chr1    unknown    exon    17233    17368    .    -    .    gene_id "WASH7P"; gene_name "WASH7P"; transcript_id "NR_024540"; tss_id "TSS8151";
#    chr1    unknown    exon    17369    17436    .    -    .    gene_id "MIR6859-2"; gene_name "MIR6859-2"; transcript_id "NR_107062"; tss_id "TSS24460";
#    chr1    unknown    exon    17369    17436    .    -    .    gene_id "MIR6859-1"; gene_name "MIR6859-1"; transcript_id "NR_106918_2"; tss_id "TSS24460";
#
# output (string):
#    A text file in the format of
#
#    chr1  MIR6859-2  112324
input:
    genes_gtf


python:

    import sys
    def _size(features):
        # this is definitely not efficient but we do not really care because it will be run
        # only once. 
        bases = set()
        for feature in features:
            # Add 1 because GTF input is one based
            bases.update(range(int(feature[0]), int(feature[1]) + 1))
        return len(bases)

    # only work on selected chromosomes
    allowed_chromosomes = set([str(x) for x in range(25)] + ['chr{}'.format(x) for x in range(25)] + \
                ['X', 'Y', 'M', 'chrX', 'chrY', 'chrM'])
 
    gene_size = {}
    with open(input[0], 'rU') as ifile:
        for line in ifile:
            if line.startswith('#'):
                continue
            ls = line.strip().split('\t')
            #
            # ls[0]: seqname (chromosome name)
            # ls[1]: source (--ignored--)
            # ls[2]: feature (exon only)
            # ls[3]: start position (1-based)
            # ls[4]: end position (1-based)
            # ls[5]: score (--ignored---)
            # ls[6]: strand (--ignored--)
            # ls[7]: frame (--ignored--)
            # ls[8]: attributes, look for gene_id
            #
            if ls[0] not in allowwd_chromosomes:
                continue
            if len(ls) &lt; 5:
                continue
            if ls[2] != 'exon':
                continue
            try:
                gene_id = [x.strip().rsplit(' ', 1)[-1].strip('"').strip("'") for x in ls[8].split(';') if x.strip().startswith('gene_id')]
            except Exception as e:
                sys.stderr.write('Failed to parse {}: {}'.format(line, e))
                continue
            #
            if not gene_id:
                sys.stderr.write('No gene_id for record {}'.format(line))
            #
            # use gene_id.chr as key
            try:
                key = gene_id[0] + '.' + ls[0]
                if key in gene_size:
                    if (int(ls[3]), int(ls[4])) != gene_size[key][3][-1]:
                        gene_size[key][1] = min(gene_size[key][1], int(ls[3]))
                        gene_size[key][2] = max(gene_size[key][2], int(ls[4]))
                        gene_size[key][3].append((int(ls[3]), int(ls[4])))
                else:
                    gene_size[key] = [ls[0], int(ls[3]), int(ls[4]), [(int(ls[3]), int(ls[4]))]]
            except Exception as e:
                sys.stderr.write('Failed to store key {}: {}'.format(key, e))
                continue
    #
    with open(${output!r}, 'w') as ofile:
        for key in sorted(gene_size.keys()):
            val = gene_size[key]
            #
            # Remove chr from key for output
            ofile.write('{0}\t{1}\t{2}\t{3}\t{4}\n'.format(val[0], key[:-len(val[0])-1], val[1], val[2], 
                _size(val[3])))


[flagstat: provides='{filename}.bam.flagstat']
# Alignment (samtools): Generating statistics of aligned reads.
input:
    '${filename}.bam'

run:
    samtools flagstat ${input} &gt; ${input}.flagstat
 
## # format of flagstat file. 
## # 
## #2181882 + 0 in total (QC-passed reads + QC-failed reads)
## #0 + 0 duplicates
## #2181882 + 0 mapped (100.00%:nan%)
## #2181882 + 0 paired in sequencing
## #mapped_reads = ${eval([line.split('m')[0] for line in open(input[0] + '.flagstat').read().split('\n') if 'mapped' in line][0])}
## 

[mapped_reads: shared='mapped_reads']
input:
    '${accepted_hits}.flagstat'

depends:
    sos_variable('accepted_hits')

with open(input[0]) as flgs:
    for line in flgs:
        if 'mapped' in line:
            mapped_reads = eval(line.split('m')[0])
            break


[human_hg19_620 (Gene Count RPKM)]

input:
     '${count_out}/${samplename}-gene_counts.tsv'

depends: 
    genesize_txt,
    sos_variable('mapped_reads')

output:
    "${count_out}/${samplename}-gene_counts_rpkm.xls"

python:
    import sys
    gene_size = {}
    ambiguous_size = set()
    with open(${depends[0]!r}) as genesize:
        for line in genesize:
            chr, gene, start, stop, size = line.split('\t')
            if gene in gene_size and gene_size[gene][3] != int(size):
                ambiguous_size.add(gene)
            gene_size[gene] = (chr, start, stop, int(size))
            gene_size[gene + '.' + chr] = (chr, start, stop, int(size))
    #
    print('Size of {} genes imported.'.format(len(gene_size)/2))
    with open(${input!r}) as counts, open(${output!r}, 'w') as output:
        output.write('Chr\tGeneID\tStart\tStop\tCodingLength\tReadCount\tRPKM\n')
        for line in counts:
            # '__no_feature', '__ambiguous', '__too_low_aQual', 'not_aligned', '__alignment_not_unique'
            if line.startswith('__'):
                continue
            try:
                gene, value = line.split('\t', 1)
            except:
                sys.stderr.write('Unprocess line: {}'.format(line))
            if not gene.strip():
                continue
            elif gene not in gene_size:
                sys.stderr.write('Unknown size of gene {}'.format(gene))
                output.write('NA\t{}\tNA\tNA\tNA\t{}\tNA\n'.format(gene, value.strip()))
            else:
                if gene in ambiguous_size:
                    sys.stderr.write('{} appear in different chromosomes. Its RPKM value is based on the size of one of them'.format(gene))
                output.write('{}\t{}\t{}\t{}\t{}\t{}\t{:.2f}\n'
                    .format(gene_size[gene][0],
                        # output gene name if the genename was appended with chromosome name
                        gene[:-len(gene_size[gene][0])-1] if gene.endswith('.' + gene_size[gene][0]) else gene,
                        gene_size[gene][1],
                        gene_size[gene][2], gene_size[gene][3], value.strip(),
                        1.e9*int(value)/(${mapped_reads} * gene_size[gene][3])))

[human_hg19_630]
#Count: Running bedtools interectBed to get exon count.
#Count: Calculate RPKM for exon counts.
input:
    sorted_bam

output:
    '${count_out}/${samplename}-exon_counts.tsv'

run:
    coverageBed -abam ${input} -b ${genes_exon_bed}  \
        &gt; ${count_out}/${samplename}-exon_counts.tsv

[human_hg19_631]

output:
    "${count_out}/${samplename}-exon_counts_rpkm.xls"
    
depends:
    sos_variable('mapped_reads')

python:
    import sys
    from collections import defaultdict

    exon_counts = defaultdict(int)
    with open(${input!r}) as counts:
        for line in counts:
            fields = line.split()
            if len(fields) == 16:
                exon_counts[tuple(fields[12:16])] += 1
            elif len(fields) == 8:
                if int(fields[4]) != 0:
                    exon_counts[tuple(fields[0:4])] = int(fields[4])
            else:
                sys.stderr.write('Incorrect input line: {}'.format(line))
                continue
    sys.stderr.write('Writing counts for {} exons'.format(len(exon_counts)))
    with open(${output!r}, 'w') as output:
        output.write('Chr\tStart\tStop\tExon\tReadCount\tRPKM\n')
        for exon in sorted(exon_counts.keys(), key=lambda x: x[3]):
            count = exon_counts[exon]
            # the input files are in BED format with 0-based start position
            # we need to add 1 to the start position for consistency
            output.write('{}\t{}\t{}\t{}\t{}\t{:.2f}\n'
                .format(exon[0], int(exon[1]) + 1, exon[2], exon[3], count,
                        1.e9*count/(${mapped_reads} * (int(exon[2]) - int(exon[1])))))



[update_blast: provides='${ncbi_resource_dir}/blast/db/human_genomic.nal']
#Resource: Download NCBI Blast databases and link the directory
# to output dir for blast to work
depends:
    executable('update_blastdb.pl')

output:
    '${ncbi_resource_dir}/blast/db/human_genomic.nal',
    '${ncbi_resource_dir}/blast/db/nt.nal'

task:

sh: workdir='${ncbi_resource_dir}/blast/db'
    update_blastdb.pl human_genomic --decompress || true
    update_blastdb.pl nt --decompress || true

sh: workdir=output_dir
    [ -d blast ] || ln -s ${ncbi_resource_dir}/blast/db blast


[before_tophat_fusion: provides=['${output_dir}/refGene.txt', '${output_dir}/ensGene.txt']]
#Fusion (prepare): Create an alias for tophat output directory because
#    tophat-fusion-post require the directory to have name tophat-XXX
#Fusion (prepare): Link refGene.txt and ensGene.txt to output directory to be used by tophat-fusion-post.

depends:
    refgene_txt, ensgene_txt

run:    workdir=output_dir
    [ -f refGene.txt ] || ln -s ${refgene_txt} refGene.txt
    [ -f ensGene.txt ] || ln -s ${ensgene_txt} ensGene.txt

stop_if(alignment_out.rstrip('/').rsplit('/', 1)[-1].startswith('tophat_') \
    or os.path.isdir(output_dir + '/tophat_' + alignment_out.rsplit("/",1)[-1]))

run:    workdir=output_dir
    ln -s ${alignment_out.rstrip("/").rsplit("/",1)[-1]} ${'tophat_' + alignment_out.rsplit("/",1)[-1]}


[human_hg19_710]
# tophat-fusion for fusion detection
output:
    "${fusion_out}/potential_fusion.txt"

depends:
    '${output_dir}/refGene.txt',
    '${output_dir}/ensGene.txt',
    '${ncbi_resource_dir}/blast/db/human_genomic.nal'

run: workdir=output_dir

    tophat-fusion-post -p 3  -o ${fusion_out.rstrip("/").rsplit("/")[-1]} \
        ${reference_dir}/BowtieIndex/genome


[human_hg19_730]
# (Oncofuse): filter fusion candidates using Oncofuse

output:
    '${fusion_out}/fusions_oncofuse.xls'

run:
    java -Xmx1G -jar ${oncofuse_path}/Oncofuse.jar \
        ${input} tophat-post AVG ${fusion_out}/fusions_oncofuse.xls

[download_gatk: provides=gatk_resource_dir + '/dbsnp_138.hg19.vcf']
download: dest_dir=gatk_resource_dir
    ${gatk_url}/1000G_omni2.5.hg19.sites.vcf.gz
    ${gatk_url}/1000G_omni2.5.hg19.sites.vcf.gz.md5
    ${gatk_url}/1000G_omni2.5.hg19.sites.vcf.idx.gz
    ${gatk_url}/1000G_omni2.5.hg19.sites.vcf.idx.gz.md5
    ${gatk_url}/dbsnp_138.hg19.vcf.gz
    ${gatk_url}/dbsnp_138.hg19.vcf.gz.md5
    ${gatk_url}/dbsnp_138.hg19.vcf.idx.gz
    ${gatk_url}/dbsnp_138.hg19.vcf.idx.gz.md5
    ${gatk_url}/hapmap_3.3.hg19.sites.vcf.gz
    ${gatk_url}/hapmap_3.3.hg19.sites.vcf.gz.md5
    ${gatk_url}/hapmap_3.3.hg19.sites.vcf.idx.gz
    ${gatk_url}/hapmap_3.3.hg19.sites.vcf.idx.gz.md5

[human_hg19_800 (GATK UnifiedGenotyper)]
# Use GATK UnifiedGenotyper caller to call variants
input:
    sorted_unique

depends:
    '${input}.bai',
    sos_variable('mapped_reads')

# An option -L can be used to limit the variant call to targeted regions
# -L $ref_dir/hg19.master.gene.bed 

output:
    '${variant_out}/${samplename}-gatk.vcf'

run:
    java -Xmx16g -Xms512m -Djava.io.tmpdir=${temp_dir}  \
        -jar ${gatk_path}/GenomeAnalysisTK.jar   \
        -R ${reference_dir}/WholeGenomeFasta/genome.fa   \
        --filter_reads_with_N_cigar   \
        -T UnifiedGenotyper   \
        -I ${input}  \
        --out ${variant_out}/${samplename}-gatk.vcf  

[human_hg19_810 (Recalibrate Variants)]
# recalibrate and filter variants
output:
    '${variant_out}/${samplename}-filter.vcf'

depends:
    '${reference_dir}/WholeGenomeFasta/genome.fa',
    '${gatk_resource_dir}/hapmap_3.3.hg19.sites.vcf',
    '${gatk_resource_dir}/1000G_omni2.5.hg19.sites.vcf',
    '${gatk_resource_dir}/dbsnp_138.hg19.vcf'

run:
    java -Xmx16g -Xms512m  \
        -Djava.io.tmpdir=${temp_dir}  \
        -jar ${gatk_path}/GenomeAnalysisTK.jar   \
        -R ${reference_dir}/WholeGenomeFasta/genome.fa   \
        -T VariantRecalibrator  \
        -mode SNP   \
        -input ${input}  \
        -resource:hapmap,known=false,training=true,truth=true,prior=15.0 ${gatk_resource_dir}/hapmap_3.3.hg19.sites.vcf  \
        -resource:omni,known=false,training=true,truth=false,prior=12.0  ${gatk_resource_dir}/1000G_omni2.5.hg19.sites.vcf  \
        -resource:dbsnp,known=true,training=false,truth=false,prior=2.0  ${gatk_resource_dir}/dbsnp_138.hg19.vcf  \
        -an ReadPosRankSum -an FS   \
        -recalFile ${variant_out}/${samplename}.recal   \
        -tranchesFile ${variant_out}/${samplename}.tranches   \
        --maxGaussians ${4 if mapped_reads &gt; 4000000 else 2} || true  \

if os.path.getsize('${variant_out}/${samplename}.recal') &gt; 0:
    run(r'''
        java -Xmx16g -Xms512m   \
            -Djava.io.tmpdir=${temp_dir}   \
            -jar ${gatk_path}/GenomeAnalysisTK.jar   \
            -R ${reference_dir}/WholeGenomeFasta/genome.fa  \
            -T ApplyRecalibration   \
            -mode SNP   \
            -input ${input}   \
            -recalFile ${variant_out}/${samplename}.recal  \
            -tranchesFile ${variant_out}/${samplename}.tranches  \
            -o ${variant_out}/${samplename}-filter.vcf   \
            --ts_filter_level 99.0
    ''')
else:
    run(r'''
        java -Xmx16g -Xms512m  \
            -Djava.io.tmpdir=${temp_dir}  \
            -jar ${gatk_path}/GenomeAnalysisTK.jar  \
            -R ${reference_dir}/WholeGenomeFasta/genome.fa  \
            -V ${input}  \
            -l INFO -T VariantFiltration  \
            --filterExpression "FS &gt; 20.0" --filterName FSFilter  \
            --filterExpression "ED &gt; 5" --filterName EDFilter  \
            --filterExpression "ReadPosRankSum &lt; -8.0" --filterName RPRSFilter  \
            --filterExpression "ReadPosRankSum &gt; -8.0" --filterName RPRSFilter  \
            -o ${variant_out}/${samplename}-filter.vcf
    ''')

[human_hg19_900]
#RSeQC: Generate JPEG versions of the RSeQC plots to be embedded in final report 
input:
    rseqc_pdf

output:
    '${alignmentqc_out}/${samplename}.geneBodyCoverage.curves.jpg',
    '${alignmentqc_out}/${samplename}.junctionSaturation_plot.jpg',
    '${alignmentqc_out}/${samplename}.splice_events.jpg',
    '${alignmentqc_out}/${samplename}.splice_junction.jpg',
    '${alignmentqc_out}/${samplename}.inner_distance_plot.jpg',
    '${alignmentqc_out}//${samplename}.DupRate_plot.jpg'

# we need to wait the completion of geneBody_coverage
run:
    cat ${' '.join([os.path.join(alignmentqc_out, x) for x in os.listdir(alignmentqc_out) if x.endswith('.r')])}  \
    | sed 's/pdf(\(.*\).pdf\(.\)/jpeg(\1.jpg\2, width=600, height=600/' | R --no-save


[human_hg19_120: shared='read_length']

input:
   fastq_files

import gzip
with gzip.open(input[0], 'rb') if input[0].endswith('.gz') else open(input[0], 'rb') as fastq:
    # skip the first line
    count = 0
    total_length = 0
    for index,line in enumerate(fastq):
        if index % 4 == 1:
            # get the second line
            total_length += len(line.decode().strip())
            count += 1
        if count == 100:
            break
    read_length = total_length / count



[alignment_summary: shared=('total_reads', 'used_reads', 'mapped_reads_perc', 'uniq_mapped_reads', 'uniq_mapped_reads_perc', 'mapped_pairs', 'mapped_pairs_perc', 'concordant_mapped_pairs', 'concordant_mapped_pairs_perc', 'genome_mapped_count', 'genome_mapped_perc', 'junction_mapped_count', 'junction_mapped_perc')]

depends:
    sos_variable('accepted_hits')

with open('${alignment_out}/prep_reads.info') as prep_reads:
    total_reads = 0
    used_reads = 0
    for line in prep_reads:
        if 'reads_in' in line:
            total_reads += int(line.split('=')[-1])
        if 'reads_out' in line:
            used_reads += int(line.split('=')[-1])
#
# get mapped reads from align_summary.txt
with open('${alignment_out}/align_summary.txt') as align_summary:
    uniq_mapped_reads = 0
    uniq_mapped_pairs = 0
    mapped_reads = 0
    mapped_pairs = 0
    concordant_mapped_pairs = 0
    for idx, line in enumerate(align_summary):
        if idx in (2, 6):
            #  Mapped   :    770754 (77.1% of input)
            mapped_reads += int(line.split(':')[1].split('(')[0].strip())
        if idx == 10:
            # Aligned pairs:
            mapped_pairs += int(line.split(':')[1].strip())
        if idx in (3, 7):
            uniq_mapped_reads += int(line.split(':')[1].split('(')[0].strip())
        if idx == 12:
            concordant_mapped_pairs = int(line.split()[0])
    uniq_mapped_reads = mapped_reads - uniq_mapped_reads
    uniq_mapped_reads_perc = uniq_mapped_reads * 100. / total_reads
    mapped_reads_perc = mapped_reads * 100. / total_reads
    concordant_mapped_pairs = mapped_pairs - concordant_mapped_pairs
    concordant_mapped_pairs_perc = concordant_mapped_pairs * 200.0 / total_reads
    mapped_pairs_perc = mapped_pairs * 200.0 / total_reads

# get junction reads
with open('${alignment_out}/junction.count') as junction_count:
    junction_mapped_count = int(junction_count.readline().strip())
    genome_mapped_count = mapped_reads - junction_mapped_count
    junction_mapped_perc = junction_mapped_count * 100. / total_reads
    genome_mapped_perc = genome_mapped_count * 100. / total_reads

uniq_mapped_reads = mapped_reads - uniq_mapped_reads
uniq_mapped_reads_perc = uniq_mapped_reads * 100. / total_reads
mapped_reads_perc = mapped_reads * 100. / total_reads
concordant_mapped_pairs = mapped_pairs - concordant_mapped_pairs
concordant_mapped_pairs_perc = concordant_mapped_pairs * 200.0 / total_reads
mapped_pairs_perc = mapped_pairs * 200.0 / total_reads
genome_mapped_count = mapped_reads - junction_mapped_count
junction_mapped_perc = junction_mapped_count * 100. / total_reads
genome_mapped_perc = genome_mapped_count * 100. / total_reads


[mean_depth: shared='mean_depth']

input:
    '${count_out}/${samplename}-exon_counts_rpkm.xls'

depends:
    sos_variable('read_length')

mean_depth = 0
with open(input[0]) as exons:
    exons.readline()
    total_reads = 0
    total_length = 0
    for line in exons:
        chr, start, end, name, count, rpkm = line.split()
        if int(count) &lt; 2:
            continue
        total_reads += int(count)
        total_length += int(end) - int(start) + 1
    mean_depth = float(total_reads) * int(read_length) / total_length


[get_counts: shared=('num_genes', 'num_covered_genes', 'num_exons', 'num_variants', 'num_valid_variants', 'num_oncofuse_fusions')]

depends:
    '${count_out}/${samplename}-gene_counts_rpkm.xls',
    '${count_out}/${samplename}-exon_counts_rpkm.xls',
    '${fusion_out}/fusions_oncofuse.xls',
    '${variant_out}/${samplename}-filter.vcf'

def _lineCount(filename, matching=None):
    count = 0
    count_no_comment = 0
    count_matching = 0
    with open(filename, 'r') as ifile:
        for line in ifile:
            count += 1
            if not line.startswith('#'):
                count_no_comment += 1
            if matching is not None:
                try:
                    if line.split()[matching[0]].strip() == matching[1]:
                        count_matching += 1
                except:
                    pass
    return count, count_no_comment, count_matching

gene_count = _lineCount('${count_out}/${samplename}-gene_counts_rpkm.xls', matching=(5, '0'))
num_genes = gene_count[0] - 1
# all lines - 0 count
num_covered_genes = gene_count[0] - gene_count[2] - 1
num_exons = _lineCount('${count_out}/${samplename}-exon_counts_rpkm.xls')[0] - 1

num_oncofuse_fusions = _lineCount('${fusion_out}/fusions_oncofuse.xls')[0] - 1
variant_count = _lineCount('${variant_out}/${samplename}-filter.vcf', matching=(6, 'PASS'))

num_variants = variant_count[1]
num_valid_variants = variant_count[2]

[human_hg19_910]
#Summary: Write summary.html to report all results.

output:
    '${output_dir}/summary.html'

depends:
    [sos_variable(x) for x in (
        'read_length',
        'mean_depth',
        'mate_inner_dist',
        'mate_std_dev',
        'total_reads',
        'used_reads',
        'mapped_reads',
        'mapped_reads_perc',
        'uniq_mapped_reads',
        'uniq_mapped_reads_perc',
        'mapped_pairs',
        'mapped_pairs_perc',
        'concordant_mapped_pairs',
        'concordant_mapped_pairs_perc',
        'genome_mapped_count',
        'genome_mapped_perc',
        'junction_mapped_count',
        'junction_mapped_perc',
        'reference_genes',
        'reference_genome',
        'num_genes',
        'num_covered_genes',
        'num_exons',
        'num_variants',
        'num_valid_variants',
        'num_oncofuse_fusions')]

parameter: css = 'default.css'

if not os.path.isfile(css):
    logger.warning('{} does not exist. No CSS is used in HTML output'.format(css))
    css_style = ''
else:
    with open(css) as style:
        css_style = style.read()

with open(output[0], 'w') as summary:
    summary.write(r'''
&lt;!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"&gt;
&lt;html&gt;
&lt;head&gt;
  &lt;meta content="text/html; charset=ISO-8859-1"
    http-equiv="content-type"&gt;
  &lt;title&gt;${samplename}&lt;/title&gt;
 &lt;style type="text/css"&gt;
 ${css_style}
 &lt;/style&gt;
&lt;/head&gt;
&lt;body&gt;
&lt;h1&gt;RNA Seq analysis for sample ${samplename}&lt;/h1&gt;
&lt;p&gt;Generated by pipeline ${pipeline_name} ${pipeline_version}&lt;/p&gt;

&lt;h2&gt;Project description&lt;/h2&gt;

The goal of the analysis is to evaluate how well RNA Seq sequencing worked on the sample,
obtain gene and exon expression, fusion candidates, and variant calls from the sample.

&lt;h3&gt;Basic information&lt;/h3&gt;

&lt;table class="gridtable"&gt;
&lt;tr&gt;
&lt;th&gt;Sample name&lt;/th&gt;&lt;td&gt;${samplename}&lt;/td&gt;
&lt;/tr&gt;
&lt;tr&gt;
&lt;th&gt;Type&lt;/th&gt;&lt;td&gt;Paired ends&lt;/td&gt;
&lt;/tr&gt;
&lt;tr&gt;
&lt;th&gt;Read length&lt;/th&gt;&lt;td&gt;${read_length}&lt;/td&gt;
&lt;/tr&gt;
&lt;tr&gt;
&lt;th&gt;Mean Depth of Coverage &lt;sup&gt;*&lt;/sup&gt;&lt;/th&gt;&lt;td&gt;${mean_depth:.1f} &lt;/td&gt;
&lt;/tr&gt;
&lt;tr&gt;
&lt;th&gt;Inner distance between paired reads&lt;sup&gt;**&lt;/sup&gt;&lt;/th&gt;&lt;td&gt;${mate_inner_dist} (stdev=${mate_std_dev})&lt;/td&gt;
&lt;/tr&gt;
&lt;/table&gt;

&lt;p&gt;* The mean depth of coverage is roughly estimated by number of mapped reads times read length and divided by the total length of exons. 
A low depth of coverage might appear if the mapping rate is low.&lt;/p&gt;
&lt;p&gt;** Mean and standard deviation of inner distance is estimated by distances between 1M pairs of aligned reads.&lt;/p&gt;

&lt;h2&gt;Result summary&lt;/h2&gt;

&lt;h3&gt;Read QC -- FastQC report&lt;/h3&gt;
&lt;p&gt;&lt;a href="http://www.bioinformatics.babraham.ac.uk/projects/fastqc/"&gt;FastQC&lt;/a&gt; is a quality control tool for high throughput sequence data. 
Please check &lt;a href="http://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/"&gt;here&lt;/a&gt; for details of each plot.
Note that it is normal for RNA Seq samples to have high duplication level and uneven distribution of Kmer content. Warnings 
on items such GC content distribution are more alarming.&lt;/p&gt;

&lt;table class="gridtable"&gt;
&lt;tr&gt;
&lt;tr&gt;&lt;th&gt;Per base sequence quality&lt;/th&gt;&lt;th&gt;Per base sequence content&lt;/th&gt;&lt;th&gt;Per sequence GC content&lt;/th&gt;&lt;/tr&gt;
&lt;tr&gt;
    &lt;td&gt;
    &lt;a href="../${readqc_out}/${samplename}_fastqc.html#M1"&gt;
    &lt;IMG SRC="../${readqc_out}/per_base_quality.png" WIDTH=300 ALT="Per base sequence quality"&gt;
    &lt;/a&gt;&lt;/td&gt;
    &lt;td&gt;
     &lt;a href="../${readqc_out}/${samplename}_fastqc.html#M4"&gt;
    &lt;IMG SRC="../${readqc_out}/per_base_sequence_content.png" WIDTH=300 ALT="Per base sequence content"&gt;
    &lt;/a&gt;&lt;/td&gt;
    &lt;td&gt;
    &lt;a href="../${readqc_out}/${samplename}_fastqc.html#M5"&gt;
    &lt;IMG SRC="../${readqc_out}/per_sequence_gc_content.png" WIDTH=300 ALT="Per sequence GC content"&gt;
    &lt;/a&gt;&lt;/td&gt;
&lt;/tr&gt;
&lt;/table&gt;

&lt;p&gt;&lt;a href="../${readqc_out}/${samplename}_fastqc.html"&gt;Complete FastQC reports&lt;/a&gt; &lt;/p&gt;

&lt;h3&gt;Alignment summary&lt;/h3&gt;
&lt;p&gt;Input reads are aligned to the transcriptome of ${reference_genes} genes using reference genome ${reference_genome}.
The mapping rate might appear higher if the reads are mapped to the whole genome.&lt;/p&gt;

&lt;table class="gridtable"&gt;
&lt;tr&gt;
&lt;th&gt;Total reads&lt;/th&gt;
&lt;th&gt;Used reads&lt;/th&gt;
&lt;th&gt;Mapped reads&lt;/th&gt;
&lt;th&gt;Uniquely mapped reads&lt;/th&gt;
&lt;th&gt;Mapped pairs&lt;/th&gt;
&lt;th&gt;Concordant mapped pairs&lt;/th&gt;
&lt;th&gt;Mapped reads (genome)&lt;/th&gt;
&lt;th&gt;Mapped reads (junction)&lt;/th&gt;
&lt;/tr&gt;
&lt;tr&gt;
&lt;td&gt;${total_reads:,}&lt;/td&gt;
&lt;td&gt;${used_reads:,}&lt;/td&gt;
&lt;td&gt;${mapped_reads:,} (${mapped_reads_perc:.1f}%)&lt;/td&gt;
&lt;td&gt;${uniq_mapped_reads:,} (${uniq_mapped_reads_perc:.1f}%)&lt;/td&gt;
&lt;td&gt;${mapped_pairs:,} (${mapped_pairs_perc:.1f}%)&lt;/td&gt;
&lt;td&gt;${concordant_mapped_pairs:,} (${concordant_mapped_pairs_perc:.1f}%)&lt;/td&gt;
&lt;td&gt;${genome_mapped_count:,} (${genome_mapped_perc:.1f}%)&lt;/td&gt;
&lt;td&gt;${junction_mapped_count:,} (${junction_mapped_perc:.1f}%)&lt;/td&gt;
&lt;/tr&gt;
&lt;/table&gt;

&lt;h3&gt;Alignment QC -- RSeQC plots&lt;/h3&gt;
&lt;table class="gridtable"&gt;
&lt;tr&gt;
&lt;tr&gt;&lt;th&gt;Junction saturation&lt;/th&gt;&lt;th&gt;Inner distance&lt;/th&gt;&lt;th&gt;Splicing junction&lt;/th&gt;&lt;/tr&gt;
&lt;tr&gt;
    &lt;td&gt;
    &lt;a href="../${alignmentqc_out}/${samplename}.junctionSaturation_plot.pdf"&gt;
    &lt;IMG SRC="../${alignmentqc_out}/${samplename}.junctionSaturation_plot.jpg" WIDTH=300 ALT="Junction saturation plot"&gt;
    &lt;/a&gt;&lt;/td&gt;
    &lt;td&gt;
     &lt;a href="../${alignmentqc_out}/${samplename}.inner_distance_plot.pdf"&gt;
    &lt;IMG SRC="../${alignmentqc_out}/${samplename}.inner_distance_plot.jpg" WIDTH=300 ALT="Inner distance plot"&gt;
    &lt;/a&gt;&lt;/td&gt;
    &lt;td&gt;
    &lt;a href="../${alignmentqc_out}/${samplename}.splice_junction.pdf"&gt;
    &lt;IMG SRC="../${alignmentqc_out}/${samplename}.splice_junction.jpg" WIDTH=300 ALT="Splicing junction plot"&gt;
    &lt;/a&gt;&lt;/td&gt;
&lt;/tr&gt;
&lt;tr&gt;&lt;th&gt;Read duplication&lt;/th&gt;&lt;th&gt;Gene body coverage&lt;/th&gt;&lt;th&gt;Splicing events&lt;/th&gt;&lt;/tr&gt;
&lt;tr&gt;
    &lt;td&gt;
    &lt;a href="../${alignmentqc_out}/${samplename}.DupRate_plot.pdf"&gt;
    &lt;IMG SRC="../${alignmentqc_out}/${samplename}.DupRate_plot.jpg" WIDTH=300 ALT="Read duplication plot"&gt;
    &lt;/a&gt;&lt;/td&gt;
    &lt;td&gt;
    &lt;a href="../${alignmentqc_out}/${samplename}.geneBodyCoverage.curves.pdf"&gt;
    &lt;IMG SRC="../${alignmentqc_out}/${samplename}.geneBodyCoverage.curves.jpg" WIDTH=300 ALT="Gene body coverage plot"&gt;
    &lt;/a&gt;&lt;/td&gt;
    &lt;td&gt;
    &lt;a href="../${alignmentqc_out}/${samplename}.splice_events.pdf"&gt;
    &lt;IMG SRC="../${alignmentqc_out}/${samplename}.splice_events.jpg" WIDTH=300 ALT="Splicing events plot"&gt;
    &lt;/a&gt;&lt;/td&gt;
    &lt;/tr&gt;
&lt;/table&gt;

&lt;h3&gt;Gene and exon counts&lt;/h3&gt;

&lt;ul&gt;
&lt;li&gt;&lt;a href="../${count_out}/${samplename}-gene_counts_rpkm.xls"&gt;Gene counts with RPKM &lt;/a&gt; (${num_genes:,} genes, ${num_covered_genes:,} overlap with at least 1 read)&lt;/li&gt;
&lt;li&gt;&lt;a href="../${count_out}/${samplename}-exon_counts_rpkm.xls"&gt;Exon counts with RPKM &lt;/a&gt; (${num_exons:,} exons overlapped with at least 1 read)&lt;/li&gt;
&lt;/ul&gt;

&lt;p&gt;The gene count was performed by &lt;a href="http://www-huber.embl.de/HTSeq/doc/count.html"&gt;htseq-count&lt;/a&gt; under
the &lt;b&gt;union&lt;/b&gt; model. The exon count was performed by &lt;a href="https://code.google.com/p/bedtools/"&gt;bedtools&lt;/a&gt;
which counts the number of reads that overlap with each exon. These files are tab-delimited files with .xls extension.&lt;/p&gt;

&lt;h3&gt;Fusion candidates&lt;/h3&gt;
The following are fusion candidates produced by tophat-fusion:
&lt;ul&gt;
&lt;li&gt;&lt;a href="../${fusion_out}/fusions_oncofuse.xls"&gt;Oncofuse filtered fusion candidates&lt;/a&gt; (${num_oncofuse_fusions:,} fusions in tab-delimited file with .xls extension)&lt;/li&gt;
&lt;/ul&gt;

&lt;h3&gt;Variant calls&lt;/h3&gt;
&lt;ul&gt;
&lt;li&gt; &lt;a href="../${variant_out}/${samplename}-filter.vcf"&gt;GATK detected SNV in VCF format&lt;/a&gt;(${num_variants:,} variants, ${num_valid_variants:,} passed basic filtering)&lt;/li&gt;
&lt;/ul&gt;

The variants are called using the GATK Unified Genotyper and include only Single Nucleotide Variants, no indel detection was performed.&lt;/p&gt;
&lt;p&gt;
&lt;p&gt;
&lt;/body&gt;
&lt;/html&gt;
''')

[human_hg19_1000]
#Summary: Compress all result files into a deliverable package.
input:
    '${output_dir}/summary.html',
    '${readqc_out}/${samplename}_fastqc.html',
    '${readqc_out}/per_base_quality.png',
    '${readqc_out}/per_base_sequence_content.png',
    '${readqc_out}/per_sequence_gc_content.png',
    '${alignmentqc_out}/${samplename}.geneBodyCoverage.curves.pdf',
    '${alignmentqc_out}/${samplename}.junctionSaturation_plot.pdf',
    '${alignmentqc_out}/${samplename}.splice_events.pdf',
    '${alignmentqc_out}/${samplename}.splice_junction.pdf',
    '${alignmentqc_out}/${samplename}.inner_distance_plot.pdf',
    '${alignmentqc_out}//${samplename}.DupRate_plot.pdf',
    '${alignmentqc_out}/${samplename}.geneBodyCoverage.curves.jpg',
    '${alignmentqc_out}/${samplename}.junctionSaturation_plot.jpg',
    '${alignmentqc_out}/${samplename}.splice_events.jpg',
    '${alignmentqc_out}/${samplename}.splice_junction.jpg',
    '${alignmentqc_out}/${samplename}.inner_distance_plot.jpg',
    '${alignmentqc_out}/${samplename}.DupRate_plot.jpg',
    '${count_out}/${samplename}-gene_counts_rpkm.xls',
    '${count_out}/${samplename}-exon_counts_rpkm.xls',
    '${fusion_out}/fusions_oncofuse.xls',
    '${variant_out}/${samplename}-filter.vcf'

output:
    '${output_dir}/${samplename}.zip'

run:
    zip ${output_dir}/${samplename}.zip ${input}


</code></pre>

<h2 id="workflow-to-install-required-tools">Workflow to install required tools</h2>

<pre><code class="language-python">#!/usr/bin/env sos-runner
#fileformat=SOS1.0

#
# Auxiliary steps to install required NGS tools if it is not avaible
#

# default installation directory. This directory is recognized by
# default by SoS so that commands installed would be immediately usable
# by sos workflows. You can use commands such as
#
#    sudo sos run install_ngs install_fastqc --install-dir /usr/local/bin/
#
# to install programs in a different directory.

parameter: install_dir  = '~/.sos/bin'
parameter: download_dir = '~/.sos/bin/download'

[install_fastqc: provides=executable('fastqc')]
ver = 'v0.11.5'

download: dest_dir=download_dir
    http://www.bioinformatics.babraham.ac.uk/projects/fastqc/fastqc_${ver}.zip

warn_if(os.path.isdir('${install_dir!e}/FastQC_${ver}'),
	"Replacing FastQC version ${ver} under ${install_dir!e}/FastQC_${ver}.")

warn_if(os.path.isfile('${install_dir!e}/fastqc}'),
	"Replacing fastqc command ${install_dir!e}/fastqc.")

run:    
    cd ${install_dir!e} &amp;&amp; \
    rm -rf ${install_dir!e}/FastQC_${ver} &amp;&amp; \
	rm -f ${install_dir!e}/fastqc &amp;&amp; \
    unzip -o ${download_dir!e}/fastqc_*.zip  &amp;&amp; \
    mv ${install_dir!e}/FastQC ${install_dir!e}/FastQC_${ver} &amp;&amp; \
    chmod 755 ${install_dir}/FastQC_${ver}/fastqc &amp;&amp; \
    ln -s ${install_dir}/FastQC_${ver}/fastqc ${install_dir}
	
[install_bowtie: provides=executable('bowtie')]
download: dest_dir=download_dir
    https://sourceforge.net/projects/bowtie-bio/files/bowtie/1.1.2/bowtie-1.1.2-macos-x86_64.zip

run:    workdir=download_dir
    unzip bowtie-1.1.2-macos-x86_64.zip &amp;&amp; \
	rm -rf ${install_dir!e}/bowtie-1.1.2 &amp;&amp; \
    mv bowtie-1.1.2 ${install_dir!e} &amp;&amp; \
	ln -s -f ${install_dir!e}/bowtie-1.1.2/* ${install_dir!e}

[install_bowtie2: provides=executable('bowtie2')]
download: dest_dir=download_dir
    https://sourceforge.net/projects/bowtie-bio/files/bowtie2/2.2.9/bowtie2-2.2.9-macos-x86_64.zip

run:    workdir=download_dir
    unzip bowtie2-2.2.9-macos-x86_64.zip &amp;&amp; \
	rm -rf ${install_dir!e}/bowtie2-2.2.9 &amp;&amp; \
    mv bowtie2-2.2.9 ${install_dir!e} &amp;&amp; \
	ln -s -f ${install_dir!e}/bowtie2-2.2.9/* ${install_dir!e}


[install_tophat2: provides=executable('tophat2')]
# install tophat 2, which requires bowtie2
#
# NOTE: The pre-compiled binary does not work well under mac so it is
# highly recommened that you compile from source. However, compiling from
# source requires xcode and boost, and the later is not among the easiest
# software package to install

depends:  executable('bowtie2')

#download: dest_dir=download_dir
#	https://ccb.jhu.edu/software/tophat/downloads/tophat-2.0.13.tar.gz

#run:	workdir=download_dir
#	tar zxf tophat-2.0.13.tar.gz &amp;&amp; \
#	cd tophat-2.0.13 &amp;&amp; \
#	./configure --prefix=${os.path.dirname(install_dir)!e} &amp;&amp; \
#	make -B &amp;&amp; make install
	
download: dest_dir=download_dir
    https://ccb.jhu.edu/software/tophat/downloads/tophat-2.0.13.OSX_x86_64.tar.gz

run:    workdir=download_dir
    tar zxf tophat-2.0.13.OSX_x86_64.tar.gz &amp;&amp; \
	rm -rf ${install_dir!e}/tophat-2.0.13.OSX_x86_64 &amp;&amp; \
    mv tophat-2.0.13.OSX_x86_64 ${install_dir!e} &amp;&amp; \
	ln -s -f ${install_dir!e}/tophat-2.0.13.OSX_x86_64/* ${install_dir!e}


[install_rseqc: provides=executable('inner_distance.py')]

run:
	pip install RSeQC

</code></pre>

                            </div>
                            <button type="button" class="btn btn-default" data-dismiss="modal"><i class="fa fa-times"></i> Close</button>
                        </div>
                    </div> 
                    <div class="footer-wrapper">
                      <footer role="contentinfo">
                        <span>&copy; 2016 Bo Peng.</span>
                      </footer>
                    </div>
                </div>
            </div>
        </div>
    </div> 


   <div class="portfolio-modal modal fade" id="SoS-Docker-Guide" tabindex="-1" role="dialog" aria-hidden="true">
        <div class="modal-content">
            <div class="close-modal" data-dismiss="modal">
                <div class="lr">
                    <div class="rl">
                    </div>
                </div>
            </div>
            <div class="container">
                <div class="row">
                 <div class="col-lg-8 col-lg-offset-2">
                        <h1>6.SoS-Docker-Guide</h1>
                        <div class="modal-body">
                         
                            <hr class="star-primary">
                            <p>This documentation is under development.</p>

<h2 id="what-is-docker-and-why-it-is-helpful">What is docker and why it is helpful</h2>

<p>This is a big question to answer but in essence you can think docker containers as virtual machines with applications but without the bulky OS part, or applications with stripped down OSes. Docker contains are much more lightweight than virtual machines because all docker containers share the same core OS and related containers (e.g. different applications derived from the same CentOS or Ubuntu OS) share the same base container. Please refer to the <a href="https://www.docker.com/">docker website</a> for details about docker. I have found it helpful to watch a few youtube videos on docker.</p>

<p>The reason why docker is very helpful in building (bioinformatics) pipelines are that</p>

<ol>
  <li>
    <p>Applications are encapsulated in docker containers so that they do not interfere with the underlying OS, and with other applications. For example, we can run a workflow with applications that based on different versions of Python2 and Python 3 without having to install them locally and calling the correct version of Python, because all applications use the specific version of Python and required libraries and tools inside their own container.</p>
  </li>
  <li>
    <p>Workflows will be more stable and reproducible because unlike a local installation of Python that can be affected by other software and upgrades of python, Docker containers are stable and will not change.</p>
  </li>
  <li>
    <p>The same docker containers can be executed on different OS (e.g. various version of Linux, MacOSX etc) so your workflow built on a Mac OS workstation can be executed on a cluster environment.</p>
  </li>
</ol>

<p>There are of course some complexity in the use of docker but SoS has made it extremely easy to use docker in your workflows.</p>

<h2 id="installing-and-configuring-docker">Installing and configuring docker</h2>

<p>Docker is relatively new and is evolving very fast. It is crucial for you to install the latest version from <a href="https://www.docker.com/">docker website</a>. This website provides very detailed step by step instruction and you should have no problem installing docker on your machine.</p>

<p>After installation, you should be able to start a docker terminal (On Mac, it is called <em>Docker QuickStart Terminal</em>) and run command</p>

<pre><code class="language-python">$ docker run hello-world
</code></pre>

<p>as suggested by the documentation.</p>

<h3 id="configuration-for-mac">Configuration for Mac</h3>

<p>If you are using a Mac machine, docker containers are executed under a VirtualBox virtual machine. The docker daemon cannot see anything outside of <code>/Users</code> so if your data resides in a large volume outside of <code>/Users</code>, you will not be able to access it from inside a docker container. This can be very frustrating because Mac OS Xs boot disk is usually very small so all important data likely reside in separate volumes under <code>/Volumes</code>.</p>

<p>In addition, the default docker machine has only 1 CPU, 2GB of RAM and 20GB diskspace. If you are using docker for any serious bioinformatics work and download a few docker images, you will need to add more RAM and more storage to the virtual box. To solve these problems, you would need to remove the default virtual machine and create a new one.</p>

<p>From your command line, stop and remove the default virtual machine
<code>python
$ docker-machine stop
$ docker-machine rm default
</code>
then create a new one with 
<code>python
$ docker-machine create -d virtualbox --virtualbox-cpu-count 4 \
    --virtualbox-memory 30000 --virtualbox-disk-size 100000 default
</code>
Here I am using 4CPU/core, 30G of RAM and 100G of disk. You can adjust these parameters according to the configuration of your machine.</p>

<p>After this, you would need to share <code>/volumes</code> to the virtual machine,</p>

<pre><code class="language-python"># stop docker virtual machine
$ docker-machine stop
# add all /Volumes to /Volumes
$ VBoxManage sharedfolder add default --name Volumes --hostpath /Volumes --automount
</code></pre>

<p>You would also need to mount <code>/Volumes</code> to the docker-machine using the following command after the docker machine starts,
<code>python
$ docker-machine ssh default 'sudo mkdir -p /Volumes ; sudo  mount -t vboxsf Volumes /Volumes'
</code>
but SoS would do this automatically for you if you are using SoS to interact with docker.</p>

<h2 id="running-a-workflow-with-docker">Running a workflow with docker</h2>

<p>Running a docker-based workflow is easy because SoS will automatically download docker images and execute scripts inside docker container. Just make sure you start the script from a docker terminal (e.g. test with command <code>docker ps</code>)</p>

<h2 id="building-a-docker-image">Building a docker image</h2>

<p>Building a docker image is usually done outside of SoS if you are maintaining a collection of docker containers to be shared by your workflows, your groups, or everyone. However, if you need to create a docker image on-the-fly or would like to embed the Dockerfile inside a SoS script, you can use the <code>docker_build</code> action to build a docker container.</p>

<p>For example, you can build a container for MISO as follows:</p>

<pre><code class="language-python">[miso_build]
# building miso from a Dockerfile
docker_build: tag='mdabioinfo/miso:latest'

    ############################################################
    # Dockerfile to build MISO container images
    # Based on Anaconda python
    ############################################################

    # Set the base image to anaconda Python 2.7 (miso does not support python 3)
    FROM continuumio/anaconda

    # File Author / Maintainer
    MAINTAINER Bo Peng &lt;bpeng@mdanderson.org&gt;

    # Update the repository sources list
    RUN apt-get update

    # Install compiler and python stuff, samtools and git
    RUN apt-get install --yes \
     build-essential \
     gcc-multilib \
     gfortran \ 
     apt-utils \
     libblas3 \ 
     liblapack3 \
     libc6 \
     cython \ 
     samtools \
     libbam-dev \
     bedtools \
     wget \
     zlib1g-dev \ 
     tar \
     gzip

    WORKDIR /usr/local
    RUN pip install misopy
</code></pre>

<p>Command</p>

<pre><code class="language-python">sos run script miso_build
</code></pre>

<p>would build a docker image <code>mdabioinfo/miso:latest</code> that can be used by other SoS steps.</p>

<h2 id="writing-a-workflow-with-docker-support">Writing a workflow with docker support</h2>

<p>Writing a workflow with docker support is a bit more complicated because you will need to understand a few concepts of docker, so reading through the <a href="https://docs.docker.com/engine/reference/run/">docker run manual</a> should be helpful. The most important concept is <strong>Volumes</strong>, whch is how the host directories are mounted to a docker container so that the command executed inside the container can access (and change) files on the host machine. SoS helps the use of docker by</p>

<ul>
  <li>Automatically mounts <code>/tmp</code> to <code>/tmp</code></li>
  <li>Automatically mounts <code>/Users</code> to <code>/Users</code> under MacOS X</li>
  <li>Automatically mounts user script inside docker and execute it as <code>/var/lib/sos/xxxxx</code></li>
</ul>

<p>so that step <code>input</code> and <code>output</code> are almost always identical inside and outside of docker.</p>

<p>To use existing public docker container, you will need to specify its tag using option <code>docker_image</code>. For example, to use <code>compbio/ngseasy-fastqc</code> container to run <code>fastqc</code>, instead of installing <code>fastqc</code> locally, you can do</p>

<pre><code class="language-python">[MISO_1]
run:     docker_image='compbio/ngseasy-fastqc:1.0-r001'
    fastqc ${input} -o /tmp
</code></pre>

<p>(More to follow)</p>

<h2 id="limitations">Limitations</h2>

<ul>
  <li>Virtual Box virtual machine does not support symbolic link so running <code>ln -s</code> inside a docker machine under Mac will cause a strange error message <code>Read-only file system</code>.</li>
</ul>

                            </div>
                            <button type="button" class="btn btn-default" data-dismiss="modal"><i class="fa fa-times"></i> Close</button>
                        </div>
                    </div> 
                    <div class="footer-wrapper">
                      <footer role="contentinfo">
                        <span>&copy; 2016 Bo Peng.</span>
                      </footer>
                    </div>
                </div>
            </div>
        </div>
    </div> 


   <div class="portfolio-modal modal fade" id="Generating-Report-with-SoS" tabindex="-1" role="dialog" aria-hidden="true">
        <div class="modal-content">
            <div class="close-modal" data-dismiss="modal">
                <div class="lr">
                    <div class="rl">
                    </div>
                </div>
            </div>
            <div class="container">
                <div class="row">
                 <div class="col-lg-8 col-lg-offset-2">
                        <h1>7.Generating-Report-with-SoS</h1>
                        <div class="modal-body">
                         
                            <hr class="star-primary">
                            <p>This documentation is under development.</p>

<h2 id="reporting-features-of-script-of-scripts">Reporting features of Script of Scripts</h2>

<p>SoS provides a simple yet powerful reporting mechanism. That is to say,</p>

<ul>
  <li>Any lines that start with <code>'! '</code> are report lines that will be written to a default report file. A space after <code>!</code> is required for better formatting, but lines with a single <code>!</code> is also acceptable.</li>
  <li>Action <code>report</code> can be used to append text to this report, or optionally to another report file.</li>
</ul>

<p>Similar to other texts in a SoS script, reports will be interpolated before they are written to the output. For example</p>

<pre><code class="language-python">#!/usr/bin/env sos-runner
#fileformat=SOS1.0

! == My report
! Date: Apr. 2016

# human
# align raw reads to human genome
[parameters]
# path to STAR
star_path = '~/bin/STAR'

[*_1]
output:   'figure.png'
R:
   R script

! == some analysis
! ![result](${output})
</code></pre>

<p>will generate a text file</p>

<pre><code class="language-python">== My report
Date: Apr. 2016
== some analysis
![result](figure.png)
</code></pre>

<p>after the execution of the pipeline. Action report can also write to this report and provides more flexible and powerful ways of reporting different types of results. Action pandoc can be used to convert the report to various formats. Note that</p>

<ul>
  <li>
    <p>Although markdown is generally used, SoS does not restrict the type of report a script generates. If so desired, you can use this mechanism to generate a log file in plain text format, a HTML file with fancy style, any flavor of markdown that you can post-process, or a .Rmd file that can be loaded into RMarkdown.</p>
  </li>
  <li>
    <p>The report is generated from the steps that are executed so the content can be controlled with parameters and configuration files.</p>
  </li>
</ul>

<h2 id="action-report">Action <code>report</code></h2>

<p>Action <code>report</code> can be used write a block of text</p>

<pre><code class="language-python">report:
    == My report
    Date: Apr. 2016
</code></pre>

<p>or content of a (dynamically generated) file to the report.</p>

<pre><code class="language-python">report(filename='report_source.txt')
</code></pre>

<p>Alternatively, it can write to another file so that multiple reports could be generated from a SoS run.</p>

<pre><code class="language-python">report(output='alternative.txt')
</code></pre>

<h2 id="pandoc-with-html-and-pdf-output">Pandoc with HTML and pdf output</h2>

<h2 id="rmarkdown">Rmarkdown</h2>

<h2 id="multiple-reports">Multiple reports</h2>


                            </div>
                            <button type="button" class="btn btn-default" data-dismiss="modal"><i class="fa fa-times"></i> Close</button>
                        </div>
                    </div> 
                    <div class="footer-wrapper">
                      <footer role="contentinfo">
                        <span>&copy; 2016 Bo Peng.</span>
                      </footer>
                    </div>
                </div>
            </div>
        </div>
    </div> 



<script src="//ajax.googleapis.com/ajax/libs/jquery/1.9.1/jquery.min.js"></script>
<script>window.jQuery || document.write('<script src="https://bopeng.github.io/assets/js/vendor/jquery-1.9.1.min.js"><\/script>')</script>


<script src="/SOS/assets/js/scripts.min.js"></script>
<script src="/SOS/assets/js/vendor/bootstrap.min.js"></script>


          

</body>
</html>