<!doctype html>
<!--[if lt IE 7]><html class="no-js lt-ie9 lt-ie8 lt-ie7" lang="en"> <![endif]-->
<!--[if (IE 7)&!(IEMobile)]><html class="no-js lt-ie9 lt-ie8" lang="en"><![endif]-->
<!--[if (IE 8)&!(IEMobile)]><html class="no-js lt-ie9" lang="en"><![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en"><!--<![endif]-->
<head>
<meta charset="utf-8">
<title>Tag Archive &#8211; SOS</title>
<meta name="description" content="An archive of posts sorted by tag.">




<!-- Open Graph -->
<meta property="og:locale" content="en_US">
<meta property="og:type" content="article">
<meta property="og:title" content="Tag Archive">
<meta property="og:description" content="An archive of posts sorted by tag.">
<meta property="og:url" content="https://bopeng.github.io/tags/">
<meta property="og:site_name" content="SOS">





<link rel="canonical" href="https://bopeng.github.io/tags/">
<link href="https://bopeng.github.io/feed.xml" type="application/atom+xml" rel="alternate" title="SOS Feed">

<!-- http://t.co/dKP3o1e -->
<meta name="HandheldFriendly" content="True">
<meta name="MobileOptimized" content="320">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<!-- For all browsers -->
<link rel="stylesheet" href="/SOS/assets/css/bootstrap.min.css">


<!-- <link rel="stylesheet" href="https://bopeng.github.io/style.css"> -->
<!-- Webfonts -->
    <link rel="stylesheet" href="/SOS/assets/css/font-awesome/css/font-awesome.min.css">
    <link href="https://fonts.googleapis.com/css?family=Montserrat:400,700" rel="stylesheet" type="text/css">
    <link href='https://fonts.googleapis.com/css?family=Kaushan+Script' rel='stylesheet' type='text/css'>
    <link href="https://fonts.googleapis.com/css?family=Lato:400,700,400italic,700italic" rel="stylesheet" type="text/css">
    <link href='https://fonts.googleapis.com/css?family=Droid+Serif:400,700,400italic,700italic' rel='stylesheet' type='text/css'>
    <link href='https://fonts.googleapis.com/css?family=Roboto+Slab:400,100,300,700' rel='stylesheet' type='text/css'>


<link rel="stylesheet" href="/SOS/assets/css/main.css">
<link rel="stylesheet" href="/SOS/assets/css/agency.css">
<link rel="stylesheet" href="/SOS/assets/css/pygmentsSoS.css">

<link href="//fonts.googleapis.com/css?family=Lato:300,400,700,300italic,400italic" rel="stylesheet" type="text/css">

<meta http-equiv="cleartype" content="on">

<!-- Load Modernizr -->
<script src="/SOS/assets/js/vendor/modernizr-2.6.2.custom.min.js"></script>

<!-- Icons -->
<!-- 16x16 -->
<link rel="shortcut icon" href="https://bopeng.github.io/favicon.ico">
<!-- 32x32 -->
<link rel="shortcut icon" href="https://bopeng.github.io/favicon.png">
<!-- 57x57 (precomposed) for iPhone 3GS, pre-2011 iPod Touch and older Android devices -->
<link rel="apple-touch-icon-precomposed" href="https://bopeng.github.io/images/apple-touch-icon-precomposed.png">
<!-- 72x72 (precomposed) for 1st generation iPad, iPad 2 and iPad mini -->
<link rel="apple-touch-icon-precomposed" sizes="72x72" href="https://bopeng.github.io/images/apple-touch-icon-72x72-precomposed.png">
<!-- 114x114 (precomposed) for iPhone 4, 4S, 5 and post-2011 iPod Touch -->
<link rel="apple-touch-icon-precomposed" sizes="114x114" href="https://bopeng.github.io/images/apple-touch-icon-114x114-precomposed.png">
<!-- 144x144 (precomposed) for iPad 3rd and 4th generation -->
<link rel="apple-touch-icon-precomposed" sizes="144x144" href="https://bopeng.github.io/images/apple-touch-icon-144x144-precomposed.png">



</head>

<body id="post-index" >

<!-- Navigation -->
    <nav class="navbar navbar-default navbar-fixed-top">
        <div class="container">
            <!-- Brand and toggle get grouped for better mobile display -->
            <div class="navbar-header page-scroll">
                <button type="button" class="navbar-toggle" data-toggle="collapse" data-target="#bs-example-navbar-collapse-1">
                    <span class="sr-only">Toggle navigation</span>
                    <span class="icon-bar"></span>
                    <span class="icon-bar"></span>
                    <span class="icon-bar"></span>
                </button>
                <a class="navbar-brand page-scroll" href="#page-top">SOS</a>
            </div>

            <!-- Collect the nav links, forms, and other content for toggling -->
            <div class="collapse navbar-collapse" id="bs-example-navbar-collapse-1">
                <ul class="nav navbar-nav navbar-right">
                    <li class="hidden">
                        <a href="#page-top"></a>
                    </li>
                    <li>
                        <a class="page-scroll" href="#intro">Introduction</a>
                    </li>
                    <li>
                        <a class="page-scroll" href="#documentation">Documentation</a>
                    </li>
                      <li>
                        <a class="page-scroll" href="#tutorial">Tutorial</a>
                    </li>
                    <li>
                        <a class="page-scroll" href="#about">About</a>
                    </li>
                    <li>
                        <a class="page-scroll" href="#team">Team</a>
                    </li>
                    <li>
                        <a class="page-scroll" href="#contact">Contact</a>
                    </li>
                </ul>
            </div>
            <!-- /.navbar-collapse -->
        </div>
        <!-- /.container-fluid -->
    </nav>

    <!-- Header -->
   <header style="background-image: url('https://bopeng.github.io//SOS/images/header-bg.jpg');"">
      
        <div class="container">
            <div class="intro-text">
                 <div class="intro-heading">SOS</div>
                <div class="intro-lead-in">Script of Scripts, a lightweight workflow system for the creation of readable workflows</div>
               
                <a href="#services" class="page-scroll btn btn-xl">Tell Me More</a>
            </div>
        </div>
    </header>
<!--
        
          <a class="page-link" href="/SOS/404.html">Page Not Found</a>
        
          <a class="page-link" href="/SOS/posts/">All Posts</a>
        
          <a class="page-link" href="/SOS/"></a>
        
          <a class="page-link" href="/SOS/tags/">Tag Archive</a>
        
          
        
          
        
 -->


<!--[if lt IE 9]><div class="upgrade"><strong><a href="http://whatbrowser.org/">Your browser is quite old!</strong> Why not upgrade to a different browser to better enjoy this site?</a></div><![endif]-->



    <section id="intro" class="bg-medium-gray">
        <div class="container">
            <div class="row">
              <div class="col-lg-12 text-center">
                    <h2 class="section-heading">Introduction</h2>
                </div>
              </div>
            <div class="row" id="main" role="main">
                


<ul class="entry-meta inline-list">
  
</ul>


              </div><!-- /#main -->
          </div>
        </section>  

<!-- Portfolio Grid Section -->
    <section id="documentation" class="bg-light-gray">
        <div class="container">
            <div class="row">
                <div class="col-lg-12 text-center">
                    <h2 class="section-heading">Documentation</h2>
                </div>
            </div>
            <div class="row">
        
            
                <div class="col-md-4 col-sm-6 portfolio-item">
                    <a href="#Introduction" class="portfolio-link" data-toggle="modal">
              <!--           <div class="portfolio-hover">
                            <div class="portfolio-hover-content">
                                <i class="fa fa-plus fa-3x"></i>
                            </div>
                        </div> -->
                        1.Introduction
                     <!--    <img src="img/portfolio/" class="img-responsive img-centered" alt=""> -->
                    </a>
                    <!-- <div class="portfolio-caption">
                        <h4></h4>
                        <p class="text-muted"></p>
                    </div> -->
                </div>
            
                <div class="col-md-4 col-sm-6 portfolio-item">
                    <a href="#SoS-Syntax" class="portfolio-link" data-toggle="modal">
              <!--           <div class="portfolio-hover">
                            <div class="portfolio-hover-content">
                                <i class="fa fa-plus fa-3x"></i>
                            </div>
                        </div> -->
                        2.SoS-Syntax
                     <!--    <img src="img/portfolio/" class="img-responsive img-centered" alt=""> -->
                    </a>
                    <!-- <div class="portfolio-caption">
                        <h4></h4>
                        <p class="text-muted"></p>
                    </div> -->
                </div>
            
                <div class="col-md-4 col-sm-6 portfolio-item">
                    <a href="#File-Structure" class="portfolio-link" data-toggle="modal">
              <!--           <div class="portfolio-hover">
                            <div class="portfolio-hover-content">
                                <i class="fa fa-plus fa-3x"></i>
                            </div>
                        </div> -->
                        3.File-Structure
                     <!--    <img src="img/portfolio/" class="img-responsive img-centered" alt=""> -->
                    </a>
                    <!-- <div class="portfolio-caption">
                        <h4></h4>
                        <p class="text-muted"></p>
                    </div> -->
                </div>
            
                <div class="col-md-4 col-sm-6 portfolio-item">
                    <a href="#Command-line-options-and-configuration-file" class="portfolio-link" data-toggle="modal">
              <!--           <div class="portfolio-hover">
                            <div class="portfolio-hover-content">
                                <i class="fa fa-plus fa-3x"></i>
                            </div>
                        </div> -->
                        4.Command-line-options-and-configuration-file
                     <!--    <img src="img/portfolio/" class="img-responsive img-centered" alt=""> -->
                    </a>
                    <!-- <div class="portfolio-caption">
                        <h4></h4>
                        <p class="text-muted"></p>
                    </div> -->
                </div>
            
                <div class="col-md-4 col-sm-6 portfolio-item">
                    <a href="#Workflow-Definition" class="portfolio-link" data-toggle="modal">
              <!--           <div class="portfolio-hover">
                            <div class="portfolio-hover-content">
                                <i class="fa fa-plus fa-3x"></i>
                            </div>
                        </div> -->
                        5.Workflow-Definition
                     <!--    <img src="img/portfolio/" class="img-responsive img-centered" alt=""> -->
                    </a>
                    <!-- <div class="portfolio-caption">
                        <h4></h4>
                        <p class="text-muted"></p>
                    </div> -->
                </div>
            
                <div class="col-md-4 col-sm-6 portfolio-item">
                    <a href="#SoS-Steps" class="portfolio-link" data-toggle="modal">
              <!--           <div class="portfolio-hover">
                            <div class="portfolio-hover-content">
                                <i class="fa fa-plus fa-3x"></i>
                            </div>
                        </div> -->
                        6.SoS-Steps
                     <!--    <img src="img/portfolio/" class="img-responsive img-centered" alt=""> -->
                    </a>
                    <!-- <div class="portfolio-caption">
                        <h4></h4>
                        <p class="text-muted"></p>
                    </div> -->
                </div>
            
                <div class="col-md-4 col-sm-6 portfolio-item">
                    <a href="#Make-Style-Rules" class="portfolio-link" data-toggle="modal">
              <!--           <div class="portfolio-hover">
                            <div class="portfolio-hover-content">
                                <i class="fa fa-plus fa-3x"></i>
                            </div>
                        </div> -->
                        7.Make-Style-Rules
                     <!--    <img src="img/portfolio/" class="img-responsive img-centered" alt=""> -->
                    </a>
                    <!-- <div class="portfolio-caption">
                        <h4></h4>
                        <p class="text-muted"></p>
                    </div> -->
                </div>
            
                <div class="col-md-4 col-sm-6 portfolio-item">
                    <a href="#SoS-Actions" class="portfolio-link" data-toggle="modal">
              <!--           <div class="portfolio-hover">
                            <div class="portfolio-hover-content">
                                <i class="fa fa-plus fa-3x"></i>
                            </div>
                        </div> -->
                        8.SoS-Actions
                     <!--    <img src="img/portfolio/" class="img-responsive img-centered" alt=""> -->
                    </a>
                    <!-- <div class="portfolio-caption">
                        <h4></h4>
                        <p class="text-muted"></p>
                    </div> -->
                </div>
            
                <div class="col-md-4 col-sm-6 portfolio-item">
                    <a href="#Workflow-Execution" class="portfolio-link" data-toggle="modal">
              <!--           <div class="portfolio-hover">
                            <div class="portfolio-hover-content">
                                <i class="fa fa-plus fa-3x"></i>
                            </div>
                        </div> -->
                        9.Workflow-Execution
                     <!--    <img src="img/portfolio/" class="img-responsive img-centered" alt=""> -->
                    </a>
                    <!-- <div class="portfolio-caption">
                        <h4></h4>
                        <p class="text-muted"></p>
                    </div> -->
                </div>
            
                <div class="col-md-4 col-sm-6 portfolio-item">
                    <a href="#User-Interface" class="portfolio-link" data-toggle="modal">
              <!--           <div class="portfolio-hover">
                            <div class="portfolio-hover-content">
                                <i class="fa fa-plus fa-3x"></i>
                            </div>
                        </div> -->
                        10.User-Interface
                     <!--    <img src="img/portfolio/" class="img-responsive img-centered" alt=""> -->
                    </a>
                    <!-- <div class="portfolio-caption">
                        <h4></h4>
                        <p class="text-muted"></p>
                    </div> -->
                </div>
            
            </div>
        </div>
    </section>

<!-- Portfolio Grid Section -->
    <section id="tutorial" class="bg-medium-gray">
        <div class="container">
            <div class="row">
                <div class="col-lg-12 text-center">
                    <h2 class="section-heading">Tutorial</h2>
                </div>
            </div>
            <div class="row">
            
                <div class="col-md-4 col-sm-6 portfolio-item">
                    <a href="#Quick-start" class="portfolio-link" data-toggle="modal">
              <!--           <div class="portfolio-hover">
                            <div class="portfolio-hover-content">
                                <i class="fa fa-plus fa-3x"></i>
                            </div>
                        </div> -->
                        1.Quick-Start
                     <!--    <img src="img/portfolio/" class="img-responsive img-centered" alt=""> -->
                    </a>
                    <!-- <div class="portfolio-caption">
                        <h4></h4>
                        <p class="text-muted"></p>
                    </div> -->
                </div>
            
                <div class="col-md-4 col-sm-6 portfolio-item">
                    <a href="#Using-SoS-with-iPython" class="portfolio-link" data-toggle="modal">
              <!--           <div class="portfolio-hover">
                            <div class="portfolio-hover-content">
                                <i class="fa fa-plus fa-3x"></i>
                            </div>
                        </div> -->
                        2.Using-SoS-with-iPython
                     <!--    <img src="img/portfolio/" class="img-responsive img-centered" alt=""> -->
                    </a>
                    <!-- <div class="portfolio-caption">
                        <h4></h4>
                        <p class="text-muted"></p>
                    </div> -->
                </div>
            
                <div class="col-md-4 col-sm-6 portfolio-item">
                    <a href="#SoS-Notebook-Using-Jupyter" class="portfolio-link" data-toggle="modal">
              <!--           <div class="portfolio-hover">
                            <div class="portfolio-hover-content">
                                <i class="fa fa-plus fa-3x"></i>
                            </div>
                        </div> -->
                        3.SoS-Notebook-Using-Jupyter
                     <!--    <img src="img/portfolio/" class="img-responsive img-centered" alt=""> -->
                    </a>
                    <!-- <div class="portfolio-caption">
                        <h4></h4>
                        <p class="text-muted"></p>
                    </div> -->
                </div>
            
                <div class="col-md-4 col-sm-6 portfolio-item">
                    <a href="#Using-Spyder-as-SoS-IDE" class="portfolio-link" data-toggle="modal">
              <!--           <div class="portfolio-hover">
                            <div class="portfolio-hover-content">
                                <i class="fa fa-plus fa-3x"></i>
                            </div>
                        </div> -->
                        4.Using-Spyder-as-SoS-IDE
                     <!--    <img src="img/portfolio/" class="img-responsive img-centered" alt=""> -->
                    </a>
                    <!-- <div class="portfolio-caption">
                        <h4></h4>
                        <p class="text-muted"></p>
                    </div> -->
                </div>
            
                <div class="col-md-4 col-sm-6 portfolio-item">
                    <a href="#A-Complete-Example" class="portfolio-link" data-toggle="modal">
              <!--           <div class="portfolio-hover">
                            <div class="portfolio-hover-content">
                                <i class="fa fa-plus fa-3x"></i>
                            </div>
                        </div> -->
                        5.A-Complete-Example
                     <!--    <img src="img/portfolio/" class="img-responsive img-centered" alt=""> -->
                    </a>
                    <!-- <div class="portfolio-caption">
                        <h4></h4>
                        <p class="text-muted"></p>
                    </div> -->
                </div>
            
                <div class="col-md-4 col-sm-6 portfolio-item">
                    <a href="#SoS-Docker-Guide" class="portfolio-link" data-toggle="modal">
              <!--           <div class="portfolio-hover">
                            <div class="portfolio-hover-content">
                                <i class="fa fa-plus fa-3x"></i>
                            </div>
                        </div> -->
                        6.SoS-Docker-Guide
                     <!--    <img src="img/portfolio/" class="img-responsive img-centered" alt=""> -->
                    </a>
                    <!-- <div class="portfolio-caption">
                        <h4></h4>
                        <p class="text-muted"></p>
                    </div> -->
                </div>
            
                <div class="col-md-4 col-sm-6 portfolio-item">
                    <a href="#Generating-Report-with-SoS" class="portfolio-link" data-toggle="modal">
              <!--           <div class="portfolio-hover">
                            <div class="portfolio-hover-content">
                                <i class="fa fa-plus fa-3x"></i>
                            </div>
                        </div> -->
                        7.Generating-Report-with-SoS
                     <!--    <img src="img/portfolio/" class="img-responsive img-centered" alt=""> -->
                    </a>
                    <!-- <div class="portfolio-caption">
                        <h4></h4>
                        <p class="text-muted"></p>
                    </div> -->
                </div>
            
            </div>
        </div>
    </section>
<!-- Team Section -->
    <section id="team" class="bg-light-gray">
        <div class="container">
            <div class="row">
                <div class="text-center">
                    <h2 class="section-heading">Our Amazing Team</h2>
                </div>
            </div>
            <div class="row">
        
                    <div class="team-member">
                        <h4>Bo Peng</h4>
                        <p class="text-muted">Assistant Professor, Department of Bioinformatics and Computational Biology, MD Anderson Cancer Center.</p>
                    </div>
                </div>
            <!-- <div class="row">
                <div class="col-lg-8 col-lg-offset-2 text-center">
                    <p class="large text-muted">Lorem ipsum dolor sit amet, consectetur adipisicing elit. Aut eaque, laboriosam veritatis, quos non quis ad perspiciatis, totam corporis ea, alias ut unde.</p>
                </div>
            </div> -->
        </div>
    </section>
<section id="contact">
        <div class="container">
            <div class="row">
                <div class="col-lg-12 text-center">
                    <h2 class="section-heading">Contact Us</h2>
                </div>
            </div>
            <div class="row">
                <div class="col-lg-12">
                    <form name="sentMessage" id="contactForm" novalidate>
                        <div class="row">
                            <div class="col-md-6">
                                <div class="control-group form-group">
                                    <input type="text" class="form-control" placeholder="Your Name *" id="name" required data-validation-required-message="Please enter your name.">
                                    <p class="help-block text-danger"></p>
                                </div>
                                <div class="control-group form-group">
                                    <input type="email" class="form-control" placeholder="Your Email *" id="email" required data-validation-required-message="Please enter your email address.">
                                    <p class="help-block text-danger"></p>
                                </div>
                                <div class="control-group form-group">
                                    <input type="tel" class="form-control" placeholder="Your Phone *" id="phone" required data-validation-required-message="Please enter your phone number.">
                                    <p class="help-block text-danger"></p>
                                </div>
                            </div>
                            <div class="col-md-6">
                                <div class="control-group form-group">
                                    <textarea class="form-control" placeholder="Your Message *" id="message" required data-validation-required-message="Please enter a message."></textarea>
                                    <p class="help-block text-danger"></p>
                                </div>
                            </div>
                            <div class="clearfix"></div>
                            <div class="col-lg-12 text-center">
                                <div id="success"></div>
                                <button type="submit" class="btn btn-xl">Send Message</button>
                            </div>
                        </div>
                    </form>
                </div>
            </div>
        </div>
    </section>


<div class="footer-wrapper">
  <footer role="contentinfo">
    <span>&copy; 2016 Bo Peng.</span>
  </footer>
</div><!-- /.footer-wrapper -->

 <!-- Portfolio Modals -->
 
   <div class="portfolio-modal modal fade" id="Introduction" tabindex="-1" role="dialog" aria-hidden="true">
        <div class="modal-content">
            <div class="close-modal" data-dismiss="modal">
                <div class="lr">
                    <div class="rl">
                    </div>
                </div>
            </div>
            <div class="container">
                <div class="row">
                 <div class="col-lg-8 col-lg-offset-2">
                        <h1>1.Introduction</h1>
                        <div class="modal-body">
                         
                            <hr class="star-primary">
                            <h2 id="overview-of-sos-concepts">Overview of SoS concepts</h2>

<p>A SoS <strong>script</strong> defines one or more <strong>workflows</strong>, and each workflow consists of one or more <strong>steps</strong>.</p>

<p><img src="/SOS/media/workflow.png" alt="workflow" class="img-responsive" /></p>

<p>Although the input and output can be more general, each step typically has its <strong>input</strong>, <strong>output</strong>, and <strong>dependents</strong> files, it executes a <strong>step process</strong> that consists of one or more Python statements and SoS actions (special python functions). Part or all the step process, called <strong>tasks</strong>, can be executed and monitored externally.</p>

<p><img src="/SOS/media/sos_step.png" alt="sos_step" class="img-responsive" /></p>

<p>A SoS script contains <strong>comments</strong>, <strong>statements</strong>, and one or more SoS <strong>steps</strong>. A SoS <strong>step</strong> consists of a <strong>header</strong>
with one or more step names and optional options. The body of a SoS step consists of optional <strong>comments</strong>, 
<strong>statements</strong>, <strong>input</strong>, <strong>output</strong>, <strong>depends</strong> files, <strong>parameter</strong> definitions, followed by step <strong>process</strong>. The following figure 
shows a sample script that defines a workflow with two steps:</p>

<p><img src="/SOS/media/sample_script.jpg" alt="sample_script" class="img-responsive" /></p>

<h2 id="formal-definitions-of-terminology--grammar">Formal definitions of terminology &amp; grammar</h2>

<ul>
  <li><strong>Script</strong>: A SoS script that defines one or more workflows.</li>
  <li><strong>Workflow</strong>: A sequence of processes that can be executed to complete certain task.</li>
  <li><strong>Step</strong>: A step of a workflow that perform one piece of the workflow.</li>
  <li><strong>Target</strong>: Objects that are input and result of a SoS step, which are usually files, but can also be objects such as an executable command (with variable locations), and a SoS variable.</li>
  <li><strong>Step options</strong>: Options of the step that assist the definition of the workflow.</li>
  <li><strong>Step input</strong>: Specifies the input files of the step.</li>
  <li><strong>Step output</strong>: Specifies the output files and targets of the step.</li>
  <li><strong>Step dependencies</strong>: Specifies the files and targets that are required by the step.</li>
  <li><strong>Step process</strong>: The process that a step executes to complete specified work, specified as one or more Python statements.</li>
  <li><strong>Task</strong>: Part or all step process that will be executed and monitored outside of SoS. These are usually resource intensive jobs that will take long time to complete.</li>
  <li><strong>Action</strong>: SoS or user-defined Python functions. They differ from regular Python functions in that they may behave differently in different running mode of SoS (e.g. ignore when executed in dryrun mode).</li>
</ul>

<p>More formally defined, the SoS syntax obeys the following grammar, given in extended Backus-Naur form (EBNF):</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="n">Script</span>         <span class="o">=</span> <span class="p">{</span><span class="n">comment</span><span class="p">},</span> <span class="p">{</span><span class="n">statement</span><span class="p">},</span> <span class="p">{</span><span class="n">step</span><span class="p">};</span>
<span class="n">comment</span>        <span class="o">=</span> <span class="s">&quot;#&quot;</span><span class="p">,</span> <span class="n">text</span><span class="p">,</span> <span class="n">NEWLINE</span>
<span class="n">assignment</span>     <span class="o">=</span> <span class="n">name</span><span class="p">,</span> <span class="s">&quot;=&quot;</span><span class="p">,</span> <span class="n">expression</span><span class="p">,</span> <span class="n">NEWLINE</span></code></pre></figure>

<p>with SoS steps defined as</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="n">step</span>           <span class="o">=</span> <span class="n">step_header</span><span class="p">,</span>
                 <span class="p">{</span><span class="n">comment</span><span class="p">},</span> <span class="p">{</span> <span class="p">{</span><span class="n">statement</span><span class="p">},</span> <span class="p">[</span><span class="nb">input</span> <span class="o">|</span> <span class="n">output</span> <span class="o">|</span> <span class="n">depends</span> <span class="p">]},</span>
                 <span class="p">[</span><span class="n">process</span><span class="p">,</span> <span class="n">NEWLINE</span><span class="p">,</span> <span class="p">{</span><span class="n">script</span><span class="p">}</span> <span class="p">]</span>
<span class="n">step_header</span>    <span class="o">=</span> <span class="s">&quot;[&quot;</span><span class="p">,</span> <span class="n">section_names</span><span class="p">,</span> <span class="p">[</span><span class="s">&quot;:&quot;</span><span class="p">,</span> <span class="n">names</span> <span class="o">|</span> <span class="n">options</span><span class="p">],</span> <span class="s">&quot;]&quot;</span><span class="p">,</span> <span class="n">NEWLINE</span>
<span class="n">parameter</span>      <span class="o">=</span> <span class="s">&quot;parameter&quot;</span><span class="p">,</span> <span class="s">&quot;:&quot;</span><span class="p">,</span> <span class="n">assignment</span>
<span class="nb">input</span>          <span class="o">=</span> <span class="s">&quot;input&quot;</span><span class="p">,</span> <span class="s">&quot;:&quot;</span><span class="p">,</span> <span class="p">[</span><span class="n">expressions</span><span class="p">],</span> <span class="p">[</span><span class="s">&quot;,&quot;</span><span class="p">,</span> <span class="n">options</span><span class="p">],</span> <span class="n">NEWLINE</span>
<span class="n">output</span>         <span class="o">=</span> <span class="s">&quot;output&quot;</span><span class="p">,</span> <span class="s">&quot;:&quot;</span><span class="p">,</span> <span class="p">[</span><span class="n">expressions</span><span class="p">],</span> <span class="p">[</span><span class="s">&quot;,&quot;</span><span class="p">,</span> <span class="n">options</span><span class="p">],</span> <span class="n">NEWLINE</span>
<span class="n">depends</span>        <span class="o">=</span> <span class="s">&quot;depends&quot;</span><span class="p">,</span> <span class="s">&quot;:&quot;</span><span class="p">,</span> <span class="p">[</span><span class="n">expressions</span><span class="p">],</span> <span class="p">[</span><span class="s">&quot;,&quot;</span><span class="p">,</span> <span class="n">options</span><span class="p">],</span> <span class="n">NEWLINE</span>
<span class="n">task</span>           <span class="o">=</span> <span class="s">&quot;task&quot;</span><span class="p">,</span> <span class="s">&quot;:&quot;</span><span class="p">,</span>  <span class="p">[</span><span class="n">options</span><span class="p">]</span>
<span class="n">action</span>         <span class="o">=</span> <span class="n">func_format</span> <span class="o">|</span> <span class="n">script_format</span>
<span class="n">func_format</span>    <span class="o">=</span> <span class="n">name</span><span class="p">,</span> <span class="s">&quot;(&quot;</span><span class="p">,</span> <span class="p">[</span><span class="n">options</span><span class="p">],</span> <span class="s">&quot;)&quot;</span>
<span class="n">script_format</span>  <span class="o">=</span> <span class="n">name</span><span class="p">,</span> <span class="s">&quot;:&quot;</span><span class="p">,</span> <span class="p">[</span><span class="n">options</span><span class="p">],</span> <span class="n">NEWLINE</span><span class="p">,</span> <span class="n">script</span> 
<span class="n">section_names</span>  <span class="o">=</span> <span class="n">section_name</span><span class="p">,</span> <span class="s">&quot;,&quot;</span><span class="p">,</span> <span class="n">section_name</span>
<span class="n">section_name</span>   <span class="o">=</span> <span class="n">name</span><span class="p">,</span> <span class="s">&quot;(&quot;</span><span class="p">,</span> <span class="n">text</span><span class="p">,</span> <span class="s">&quot;)&quot;</span>
<span class="n">names</span>          <span class="o">=</span> <span class="n">name</span><span class="p">,</span> <span class="p">{</span><span class="s">&quot;,&quot;</span><span class="p">,</span> <span class="n">name</span><span class="p">}</span>
<span class="n">workflow</span>       <span class="o">=</span> <span class="n">name</span><span class="p">,</span> <span class="p">[</span><span class="s">&#39;_&#39;</span><span class="p">,</span> <span class="n">steps</span><span class="p">],</span> <span class="p">{</span><span class="s">&quot;+&quot;</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="p">[</span><span class="s">&#39;_&#39;</span><span class="p">,</span> <span class="n">steps</span><span class="p">}</span>
<span class="n">assignment</span>     <span class="o">=</span> <span class="n">name</span><span class="p">,</span> <span class="s">&quot;=&quot;</span><span class="p">,</span> <span class="n">expression</span><span class="p">,</span> <span class="n">NEWLINW</span>
<span class="n">expressions</span>    <span class="o">=</span> <span class="n">expression</span><span class="p">,</span> <span class="p">{</span><span class="s">&quot;,&quot;</span><span class="p">,</span> <span class="n">expression</span><span class="p">}</span>
<span class="n">options</span>        <span class="o">=</span> <span class="n">option</span><span class="p">,</span> <span class="p">{</span><span class="s">&quot;,&quot;</span> <span class="n">option</span><span class="p">}</span>
<span class="n">option</span>         <span class="o">=</span> <span class="n">name</span><span class="p">,</span> <span class="s">&quot;=&quot;</span><span class="p">,</span> <span class="n">expression</span></code></pre></figure>

<p>Here <code>name</code>, <code>expression</code> and <code>statement</code> are arbitrary <a href="http://www.python.org">Python</a> names, expression and statements with added SoS features. <strong>SoS requires Python 3 and does not support Python 2.x specific syntax</strong></p>

<p>This documentation is a comprehensive reference to all SoS features. Please refer to this <a href="https://github.com/BoPeng/SOS/wiki/1.-Quick-Start">SoS quick start tutorial</a> to learn some basics of SoS before diving into the details.</p>


                            </div>
                            <button type="button" class="btn btn-default" data-dismiss="modal"><i class="fa fa-times"></i> Close</button>
                        </div>
                    </div> 
                    <div class="footer-wrapper">
                      <footer role="contentinfo">
                        <span>&copy; 2016 Bo Peng.</span>
                      </footer>
                    </div>
                </div>
            </div>
        </div>
    </div> 


   <div class="portfolio-modal modal fade" id="User-Interface" tabindex="-1" role="dialog" aria-hidden="true">
        <div class="modal-content">
            <div class="close-modal" data-dismiss="modal">
                <div class="lr">
                    <div class="rl">
                    </div>
                </div>
            </div>
            <div class="container">
                <div class="row">
                 <div class="col-lg-8 col-lg-offset-2">
                        <h1>10.User-Interface</h1>
                        <div class="modal-body">
                         
                            <hr class="star-primary">
                            <h2 id="command-sos">Command <code>sos</code></h2>

<p>Command <code>sos</code> accepts a number of subcommands (similar to <code>svn</code>, <code>git</code> etc). Its syntax follows</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="n">sos</span> <span class="n">subcommand</span> <span class="p">[</span><span class="n">subcommand</span><span class="o">-</span><span class="n">options</span><span class="p">]</span></code></pre></figure>

<p>You can use command</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="err">$</span> <span class="n">sos</span> <span class="o">-</span><span class="n">h</span>
<span class="kn">usage:</span> <span class="n">sos</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">version</span><span class="p">]</span>
           <span class="p">{</span><span class="n">run</span><span class="p">,</span><span class="n">dryrun</span><span class="p">,</span><span class="n">prepare</span><span class="p">,</span><span class="n">convert</span><span class="p">,</span><span class="n">remove</span><span class="p">,</span><span class="n">start</span><span class="p">,</span><span class="n">config</span><span class="p">}</span> <span class="o">...</span>

<span class="n">A</span> <span class="n">workflow</span> <span class="n">system</span> <span class="k">for</span> <span class="n">the</span> <span class="n">execution</span> <span class="n">of</span> <span class="n">commands</span> <span class="ow">and</span> <span class="n">scripts</span> <span class="ow">in</span> <span class="n">different</span>
<span class="n">languages</span><span class="o">.</span>

<span class="n">optional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="o">-</span><span class="n">h</span><span class="p">,</span> <span class="o">--</span><span class="n">help</span>            <span class="n">show</span> <span class="n">this</span> <span class="n">help</span> <span class="n">message</span> <span class="ow">and</span> <span class="nb">exit</span>
  <span class="o">--</span><span class="n">version</span>             <span class="n">show</span> <span class="n">program</span><span class="s">&#39;s version number and exit</span>

<span class="kn">subcommands:</span>
  <span class="p">{</span><span class="n">run</span><span class="p">,</span><span class="n">dryrun</span><span class="p">,</span><span class="n">prepare</span><span class="p">,</span><span class="n">convert</span><span class="p">,</span><span class="n">remove</span><span class="p">,</span><span class="n">start</span><span class="p">,</span><span class="n">config</span><span class="p">}</span>
    <span class="n">run</span>                 <span class="n">Execute</span> <span class="n">a</span> <span class="n">SoS</span> <span class="n">script</span>
    <span class="n">dryrun</span>              <span class="n">Execute</span> <span class="n">a</span> <span class="n">SoS</span> <span class="n">script</span> <span class="ow">in</span> <span class="n">dryrun</span> <span class="n">mode</span>
    <span class="n">prepare</span>             <span class="n">Execute</span> <span class="n">a</span> <span class="n">SoS</span> <span class="n">script</span> <span class="ow">in</span> <span class="n">prepare</span> <span class="n">mode</span>
    <span class="n">convert</span>             <span class="n">Convert</span> <span class="n">between</span> <span class="n">sos</span> <span class="ow">and</span> <span class="n">other</span> <span class="nb">file</span> <span class="n">formats</span> <span class="n">such</span> <span class="k">as</span>
                        <span class="n">html</span> <span class="ow">and</span> <span class="n">Jupyter</span> <span class="n">notebooks</span>
    <span class="n">remove</span>              <span class="n">Remove</span> <span class="n">tracked</span> <span class="ow">and</span><span class="o">/</span><span class="ow">or</span> <span class="n">untracked</span> <span class="n">files</span> <span class="k">with</span> <span class="n">their</span>
                        <span class="n">signatures</span>
    <span class="n">config</span>              <span class="n">Set</span><span class="p">,</span> <span class="n">unset</span> <span class="ow">or</span> <span class="n">get</span> <span class="n">the</span> <span class="n">value</span> <span class="n">of</span> <span class="n">system</span> <span class="ow">or</span> <span class="n">local</span>
                        <span class="n">configuration</span> <span class="n">files</span>

<span class="n">Use</span> <span class="s">&#39;sos cmd -h&#39;</span> <span class="k">for</span> <span class="n">details</span> <span class="n">about</span> <span class="n">each</span> <span class="n">subcommand</span><span class="o">.</span> <span class="n">Please</span> <span class="n">contact</span> <span class="n">Bo</span> <span class="n">Peng</span>
<span class="p">(</span><span class="n">bpeng</span> <span class="n">at</span> <span class="n">mdanderson</span><span class="o">.</span><span class="n">org</span><span class="p">)</span> <span class="k">if</span> <span class="n">you</span> <span class="n">have</span> <span class="nb">any</span> <span class="n">question</span><span class="o">.</span></code></pre></figure>

<p>to get a list of subcommands with brief descriptions and</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="n">sos</span> <span class="n">subcommand</span> <span class="o">-</span><span class="n">h</span></code></pre></figure>

<p>to get detailed description of a particular subcommand.</p>

<h2 id="subcommand-run">subcommand <code>run</code></h2>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="err">$</span> <span class="n">sos</span> <span class="n">run</span> <span class="o">-</span><span class="n">h</span>
<span class="kn">usage:</span> <span class="n">sos</span> <span class="n">run</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">j</span> <span class="n">JOBS</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">c</span> <span class="n">CONFIG_FILE</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">t</span> <span class="n">FILE</span> <span class="p">[</span><span class="n">FILE</span> <span class="o">...</span><span class="p">]]</span>
               <span class="p">[</span><span class="o">-</span><span class="n">b</span> <span class="p">[</span><span class="n">BIN_DIR</span> <span class="p">[</span><span class="n">BIN_DIR</span> <span class="o">...</span><span class="p">]]]</span> <span class="p">[</span><span class="o">-</span><span class="n">q</span> <span class="n">QUEUE</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">n</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">p</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">f</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">F</span><span class="p">]</span>
               <span class="p">[</span><span class="o">-</span><span class="n">v</span> <span class="p">{</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">}]</span>
               <span class="n">SCRIPT</span> <span class="p">[</span><span class="n">WORKFLOW</span><span class="p">]</span>

<span class="n">Execute</span> <span class="n">a</span> <span class="n">workflow</span> <span class="n">defined</span> <span class="ow">in</span> <span class="n">script</span>

<span class="n">positional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="n">SCRIPT</span>                <span class="n">A</span> <span class="n">SoS</span> <span class="n">script</span> <span class="n">that</span> <span class="n">defines</span> <span class="n">one</span> <span class="ow">or</span> <span class="n">more</span> <span class="n">workflows</span><span class="o">.</span> <span class="n">The</span>
                        <span class="n">script</span> <span class="n">can</span> <span class="n">be</span> <span class="n">a</span> <span class="n">filename</span> <span class="ow">or</span> <span class="n">a</span> <span class="n">URL</span> <span class="kn">from</span> <span class="nn">which</span> <span class="nn">the</span>
                        <span class="n">content</span> <span class="n">of</span> <span class="n">a</span> <span class="n">SoS</span> <span class="n">will</span> <span class="n">be</span> <span class="n">read</span><span class="o">.</span> <span class="n">If</span> <span class="n">a</span> <span class="n">valid</span> <span class="nb">file</span> <span class="n">cannot</span>
                        <span class="n">be</span> <span class="n">located</span> <span class="ow">or</span> <span class="n">downloaded</span><span class="p">,</span> <span class="n">SoS</span> <span class="n">will</span> <span class="n">search</span> <span class="k">for</span> <span class="n">the</span>
                        <span class="n">script</span> <span class="ow">in</span> <span class="n">a</span> <span class="n">search</span> <span class="n">path</span> <span class="n">specified</span> <span class="n">by</span> <span class="n">variable</span>
                        <span class="sb">`sos_path`</span> <span class="n">defined</span> <span class="ow">in</span> <span class="n">the</span> <span class="k">global</span> <span class="n">SoS</span> <span class="n">configuration</span>
                        <span class="nb">file</span> <span class="p">(</span><span class="o">~/.</span><span class="n">sos</span><span class="o">/</span><span class="n">config</span><span class="o">.</span><span class="n">yaml</span><span class="p">)</span><span class="o">.</span>
  <span class="n">WORKFLOW</span>              <span class="n">Name</span> <span class="n">of</span> <span class="n">the</span> <span class="n">workflow</span> <span class="n">to</span> <span class="n">execute</span><span class="o">.</span> <span class="n">This</span> <span class="n">option</span> <span class="n">can</span> <span class="n">be</span>
                        <span class="n">ignored</span> <span class="k">if</span> <span class="n">the</span> <span class="n">script</span> <span class="n">defines</span> <span class="n">a</span> <span class="n">default</span> <span class="n">workflow</span> <span class="p">(</span><span class="k">with</span>
                        <span class="n">no</span> <span class="n">name</span> <span class="ow">or</span> <span class="k">with</span> <span class="n">name</span> <span class="sb">`default`</span><span class="p">)</span> <span class="ow">or</span> <span class="n">defines</span> <span class="n">only</span> <span class="n">a</span>
                        <span class="n">single</span> <span class="n">workflow</span><span class="o">.</span> <span class="n">A</span> <span class="n">subworkflow</span> <span class="ow">or</span> <span class="n">a</span> <span class="n">combined</span> <span class="n">workflow</span>
                        <span class="n">can</span> <span class="n">also</span> <span class="n">be</span> <span class="n">specified</span><span class="p">,</span> <span class="n">where</span> <span class="n">a</span> <span class="n">subworkflow</span> <span class="n">executes</span> <span class="n">a</span>
                        <span class="n">subset</span> <span class="n">of</span> <span class="n">workflow</span> <span class="p">(</span><span class="sb">`name_steps`</span> <span class="n">where</span> <span class="sb">`steps`</span> <span class="n">can</span> <span class="n">be</span>
                        <span class="sb">`n`</span> <span class="p">(</span><span class="n">a</span> <span class="n">step</span> <span class="sb">`n`</span><span class="p">),</span> <span class="sb">`:n`</span> <span class="p">(</span><span class="n">up</span> <span class="n">to</span> <span class="n">step</span> <span class="sb">`n`</span><span class="p">),</span> <span class="sb">`n:m`</span> <span class="p">(</span><span class="kn">from</span>
                        <span class="nn">step</span> <span class="sb">`n`</span> <span class="n">to</span> <span class="sb">`m`</span><span class="p">),</span> <span class="ow">and</span> <span class="sb">`n:`</span> <span class="p">(</span><span class="kn">from</span> <span class="nn">step</span> <span class="sb">`n`</span><span class="p">)),</span> <span class="ow">and</span> <span class="n">a</span>
                        <span class="n">combined</span> <span class="n">workflow</span> <span class="n">executes</span> <span class="n">to</span> <span class="n">multiple</span> <span class="p">(</span><span class="n">sub</span><span class="p">)</span><span class="n">workflows</span>
                        <span class="n">combined</span> <span class="n">by</span> <span class="sb">`+`</span> <span class="p">(</span><span class="n">e</span><span class="o">.</span><span class="n">g</span><span class="o">.</span> <span class="sb">`A_0+B+C`</span><span class="p">)</span><span class="o">.</span>

<span class="n">optional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="o">-</span><span class="n">h</span><span class="p">,</span> <span class="o">--</span><span class="n">help</span>            <span class="n">show</span> <span class="n">this</span> <span class="n">help</span> <span class="n">message</span> <span class="ow">and</span> <span class="nb">exit</span>
  <span class="o">-</span><span class="n">j</span> <span class="n">JOBS</span>               <span class="n">Number</span> <span class="n">of</span> <span class="kp">concurrent</span> <span class="n">process</span> <span class="n">allowed</span><span class="o">.</span> <span class="n">A</span> <span class="n">workflow</span> <span class="ow">is</span> <span class="n">by</span>
                        <span class="n">default</span> <span class="n">executed</span> <span class="n">sequentially</span> <span class="p">(</span><span class="o">-</span><span class="n">j</span> <span class="mi">1</span><span class="p">)</span><span class="o">.</span> <span class="n">If</span> <span class="n">a</span> <span class="n">greater</span>
                        <span class="n">than</span> <span class="mi">1</span> <span class="n">number</span> <span class="ow">is</span> <span class="n">specified</span> <span class="n">SoS</span> <span class="n">will</span> <span class="n">execute</span> <span class="n">the</span>
                        <span class="n">workflow</span> <span class="ow">in</span> <span class="n">parallel</span> <span class="n">mode</span> <span class="ow">and</span> <span class="n">execute</span> <span class="n">up</span> <span class="n">to</span> <span class="n">specified</span>
                        <span class="n">processes</span> <span class="n">concurrently</span><span class="o">.</span> <span class="n">These</span> <span class="n">include</span> <span class="n">looped</span> <span class="n">processes</span>
                        <span class="n">within</span> <span class="n">a</span> <span class="n">step</span> <span class="p">(</span><span class="k">with</span> <span class="n">runtime</span> <span class="n">option</span> <span class="sb">`concurrent=True`</span><span class="p">)</span>
                        <span class="ow">and</span> <span class="n">steps</span> <span class="k">with</span> <span class="n">non</span><span class="o">-</span><span class="n">missing</span> <span class="n">required</span> <span class="n">files</span><span class="o">.</span>
  <span class="o">-</span><span class="n">c</span> <span class="n">CONFIG_FILE</span>        <span class="n">A</span> <span class="n">configuration</span> <span class="nb">file</span> <span class="ow">in</span> <span class="n">the</span> <span class="n">format</span> <span class="n">of</span> <span class="n">YAML</span><span class="o">/</span><span class="n">JSON</span><span class="o">.</span> <span class="n">The</span>
                        <span class="n">content</span> <span class="n">of</span> <span class="n">the</span> <span class="n">configuration</span> <span class="nb">file</span> <span class="n">will</span> <span class="n">be</span> <span class="n">available</span> <span class="k">as</span>
                        <span class="n">a</span> <span class="n">dictionary</span> <span class="n">CONF</span> <span class="ow">in</span> <span class="n">the</span> <span class="n">SoS</span> <span class="n">script</span> <span class="n">being</span> <span class="n">executed</span><span class="o">.</span>
  <span class="o">-</span><span class="n">t</span> <span class="n">FILE</span> <span class="p">[</span><span class="n">FILE</span> <span class="o">...</span><span class="p">]</span>    <span class="n">One</span> <span class="n">of</span> <span class="n">more</span> <span class="n">files</span> <span class="ow">or</span> <span class="n">alias</span> <span class="n">of</span> <span class="n">other</span> <span class="n">targets</span> <span class="n">that</span> <span class="n">will</span>
                        <span class="n">be</span> <span class="n">the</span> <span class="n">target</span> <span class="n">of</span> <span class="n">execution</span><span class="o">.</span> <span class="n">If</span> <span class="n">specified</span><span class="p">,</span> <span class="n">SoS</span> <span class="n">will</span>
                        <span class="n">execute</span> <span class="n">only</span> <span class="n">part</span> <span class="n">of</span> <span class="n">a</span> <span class="n">workflow</span> <span class="ow">or</span> <span class="n">multiple</span> <span class="n">workflows</span>
                        <span class="ow">or</span> <span class="n">auxiliary</span> <span class="n">steps</span> <span class="n">to</span> <span class="n">generate</span> <span class="n">specified</span> <span class="n">targets</span><span class="o">.</span>
  <span class="o">-</span><span class="n">b</span> <span class="p">[</span><span class="n">BIN_DIR</span> <span class="p">[</span><span class="n">BIN_DIR</span> <span class="o">...</span><span class="p">]]</span>
                        <span class="n">Extra</span> <span class="n">directories</span> <span class="ow">in</span> <span class="n">which</span> <span class="n">SoS</span> <span class="n">will</span> <span class="n">look</span> <span class="k">for</span>
                        <span class="n">executables</span> <span class="n">before</span> <span class="n">standard</span> <span class="err">$</span><span class="n">PATH</span><span class="o">.</span> <span class="n">This</span> <span class="n">option</span>
                        <span class="n">essentially</span> <span class="n">prefix</span> <span class="err">$</span><span class="n">PATH</span> <span class="k">with</span> <span class="n">these</span> <span class="n">directories</span><span class="o">.</span> <span class="n">Note</span>
                        <span class="n">that</span> <span class="n">the</span> <span class="n">default</span> <span class="n">value</span> <span class="s">&#39;~/.sos/bin&#39;</span> <span class="ow">is</span> <span class="n">by</span> <span class="n">convention</span> <span class="n">a</span>
                        <span class="n">default</span> <span class="n">directory</span> <span class="k">for</span> <span class="n">commands</span> <span class="n">that</span> <span class="n">are</span> <span class="n">installed</span> <span class="n">by</span>
                        <span class="n">SoS</span><span class="o">.</span> <span class="n">You</span> <span class="n">can</span> <span class="n">use</span> <span class="n">option</span> <span class="s">&#39;-b&#39;</span> <span class="n">without</span> <span class="n">value</span> <span class="n">to</span> <span class="n">disallow</span>
                        <span class="n">commands</span> <span class="n">under</span> <span class="o">~/.</span><span class="n">sos</span><span class="o">/</span><span class="nb">bin</span><span class="o">.</span>
  <span class="o">-</span><span class="n">q</span> <span class="n">QUEUE</span>              <span class="n">Task</span><span class="o">-</span><span class="n">processing</span> <span class="n">queue</span><span class="o">.</span> <span class="n">SoS</span> <span class="n">by</span> <span class="n">default</span> <span class="n">uses</span> <span class="n">a</span> <span class="n">local</span>
                        <span class="n">multiprocessing</span> <span class="n">queue</span> <span class="n">where</span> <span class="n">tasks</span> <span class="n">are</span> <span class="n">executed</span> <span class="n">by</span>
                        <span class="n">different</span> <span class="n">processes</span><span class="o">.</span> <span class="n">Supported</span> <span class="n">task</span> <span class="n">queues</span> <span class="n">include</span> <span class="n">a</span>
                        <span class="s">&#39;rq&#39;</span> <span class="n">engine</span> <span class="n">where</span> <span class="n">tasks</span> <span class="n">will</span> <span class="n">be</span> <span class="n">distributed</span> <span class="n">to</span> <span class="n">one</span> <span class="ow">or</span>
                        <span class="n">more</span> <span class="n">rq</span><span class="o">-</span><span class="n">workers</span> <span class="k">with</span> <span class="n">assistance</span> <span class="kn">from</span> <span class="nn">a</span> <span class="nn">redis</span> <span class="nn">server</span><span class="p">,</span>
                        <span class="ow">and</span> <span class="n">a</span> <span class="s">&#39;celery&#39;</span> <span class="n">quque</span> <span class="n">where</span> <span class="n">tasks</span> <span class="n">will</span> <span class="n">be</span> <span class="n">distributed</span>
                        <span class="n">to</span> <span class="n">celery</span> <span class="n">workers</span><span class="o">.</span>
  <span class="o">-</span><span class="n">v</span> <span class="p">{</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">},</span> <span class="o">--</span><span class="n">verbosity</span> <span class="p">{</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">}</span>
                        <span class="n">Output</span> <span class="n">error</span> <span class="p">(</span><span class="mi">0</span><span class="p">),</span> <span class="n">warning</span> <span class="p">(</span><span class="mi">1</span><span class="p">),</span> <span class="n">info</span> <span class="p">(</span><span class="mi">2</span><span class="p">),</span> <span class="n">debug</span> <span class="p">(</span><span class="mi">3</span><span class="p">)</span> <span class="ow">and</span>
                        <span class="n">trace</span> <span class="p">(</span><span class="mi">4</span><span class="p">)</span> <span class="n">information</span> <span class="n">to</span> <span class="n">standard</span> <span class="n">output</span> <span class="p">(</span><span class="n">default</span> <span class="n">to</span>
                        <span class="mi">2</span><span class="p">)</span><span class="o">.</span>

<span class="n">Run</span> <span class="n">mode</span> <span class="n">options</span><span class="p">:</span>
  <span class="n">SoS</span> <span class="n">scripts</span> <span class="n">are</span> <span class="n">by</span> <span class="n">default</span> <span class="n">executed</span> <span class="ow">in</span> <span class="n">run</span> <span class="n">mode</span> <span class="n">where</span> <span class="nb">all</span> <span class="n">the</span> <span class="n">script</span> <span class="ow">is</span>
  <span class="n">run</span> <span class="ow">in</span> <span class="n">dryrun</span> <span class="n">mode</span> <span class="n">to</span> <span class="n">check</span> <span class="n">syntax</span> <span class="n">error</span><span class="p">,</span> <span class="n">prepare</span> <span class="n">mode</span> <span class="n">to</span> <span class="n">prepare</span>
  <span class="n">resources</span><span class="p">,</span> <span class="ow">and</span> <span class="n">run</span> <span class="n">mode</span> <span class="n">to</span> <span class="n">execute</span> <span class="n">the</span> <span class="n">pipelines</span><span class="o">.</span> <span class="n">Run</span> <span class="n">mode</span> <span class="n">options</span> <span class="n">allow</span>
  <span class="n">you</span> <span class="n">to</span> <span class="n">execute</span> <span class="n">these</span> <span class="n">steps</span> <span class="n">selectively</span><span class="o">.</span>

  <span class="o">-</span><span class="n">n</span>                    <span class="n">Execute</span> <span class="n">a</span> <span class="n">workflow</span> <span class="n">without</span> <span class="n">executing</span> <span class="nb">any</span> <span class="n">actions</span><span class="o">.</span> <span class="n">This</span>
                        <span class="n">can</span> <span class="n">be</span> <span class="n">used</span> <span class="n">to</span> <span class="n">check</span> <span class="n">the</span> <span class="n">syntax</span> <span class="n">of</span> <span class="n">a</span> <span class="n">SoS</span> <span class="nb">file</span><span class="o">.</span>
  <span class="o">-</span><span class="n">p</span>                    <span class="n">Execute</span> <span class="n">the</span> <span class="n">workflow</span> <span class="ow">in</span> <span class="n">preparation</span> <span class="n">mode</span> <span class="ow">in</span> <span class="n">which</span> <span class="n">SoS</span>
                        <span class="n">prepare</span> <span class="n">the</span> <span class="n">execution</span> <span class="n">of</span> <span class="n">workflow</span> <span class="n">by</span><span class="p">,</span> <span class="k">for</span> <span class="n">example</span><span class="p">,</span>
                        <span class="n">download</span> <span class="n">required</span> <span class="n">resources</span> <span class="ow">and</span> <span class="n">docker</span> <span class="n">images</span><span class="o">.</span>
  <span class="o">-</span><span class="n">f</span>                    <span class="n">Execute</span> <span class="n">the</span> <span class="n">workflow</span> <span class="ow">in</span> <span class="n">a</span> <span class="n">special</span> <span class="n">run</span> <span class="n">mode</span> <span class="n">that</span>
                        <span class="n">ignores</span> <span class="n">saved</span> <span class="n">runtime</span> <span class="n">signatures</span> <span class="ow">and</span> <span class="n">re</span><span class="o">-</span><span class="n">execute</span> <span class="nb">all</span>
                        <span class="n">the</span> <span class="n">steps</span><span class="o">.</span>
  <span class="o">-</span><span class="n">F</span>                    <span class="n">Execute</span> <span class="n">the</span> <span class="n">workflow</span> <span class="ow">in</span> <span class="n">a</span> <span class="n">special</span> <span class="n">run</span> <span class="n">mode</span> <span class="n">that</span> <span class="n">re</span><span class="o">-</span><span class="n">use</span>
                        <span class="n">existing</span> <span class="n">output</span> <span class="n">files</span> <span class="ow">and</span> <span class="n">recontruct</span> <span class="n">runtime</span>
                        <span class="n">signatures</span> <span class="k">if</span> <span class="n">output</span> <span class="n">files</span> <span class="n">exist</span><span class="o">.</span>

<span class="n">Arbitrary</span> <span class="n">parameters</span> <span class="n">defined</span> <span class="n">by</span> <span class="n">the</span> <span class="p">[</span><span class="n">parameters</span><span class="p">]</span> <span class="n">step</span> <span class="n">of</span> <span class="n">the</span> <span class="n">script</span><span class="p">,</span> <span class="ow">and</span>
<span class="p">[</span><span class="n">parameters</span><span class="p">]</span> <span class="n">steps</span> <span class="n">of</span> <span class="n">other</span> <span class="n">scripts</span> <span class="k">if</span> <span class="n">nested</span> <span class="n">workflows</span> <span class="n">are</span> <span class="n">defined</span> <span class="ow">in</span> <span class="n">other</span>
<span class="n">SoS</span> <span class="n">files</span> <span class="p">(</span><span class="n">option</span> <span class="sb">`source`</span><span class="p">)</span><span class="o">.</span> <span class="n">The</span> <span class="n">name</span><span class="p">,</span> <span class="n">default</span> <span class="ow">and</span> <span class="nb">type</span> <span class="n">of</span> <span class="n">the</span> <span class="n">parameters</span> <span class="n">are</span>
<span class="n">specified</span> <span class="ow">in</span> <span class="n">the</span> <span class="n">script</span><span class="o">.</span> <span class="n">Single</span> <span class="n">value</span> <span class="n">parameters</span> <span class="n">should</span> <span class="n">be</span> <span class="n">passed</span> <span class="n">using</span> <span class="n">option</span>
<span class="sb">`--name value`</span> <span class="ow">and</span> <span class="n">multi</span><span class="o">-</span><span class="n">value</span> <span class="n">parameters</span> <span class="n">should</span> <span class="n">be</span> <span class="n">passed</span> <span class="n">using</span> <span class="n">option</span>
<span class="sb">`--name value1 value2`</span><span class="o">.</span></code></pre></figure>

<p>Examples of the <code>sos run</code> include</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="n">sos</span> <span class="n">run</span> <span class="o">-</span><span class="n">h</span>                       <span class="c"># get help message</span>
<span class="n">sos</span> <span class="n">run</span> <span class="n">myscript</span>                 <span class="c"># run default workflow defined in myscript</span>
<span class="n">sos</span> <span class="n">run</span> <span class="n">myscript</span> <span class="n">align</span>           <span class="c"># run workflow align defined in myscript</span>
<span class="n">sos</span> <span class="n">run</span> <span class="n">myscript</span> <span class="n">align</span><span class="o">+</span><span class="n">call</span>      <span class="c"># run a combined workflow align+call</span>
<span class="n">sos</span> <span class="n">run</span> <span class="n">myscript</span> <span class="n">align</span> <span class="o">-</span><span class="n">h</span>        <span class="c"># help message for the align workflow defined in myscript</span>
<span class="n">sos</span> <span class="n">run</span> <span class="o">-</span><span class="n">n</span> <span class="n">myscript</span> <span class="n">align</span>        <span class="c"># run align workflow in inspect mode </span></code></pre></figure>

<h2 id="subcommand-dryrun">subcommand <code>dryrun</code></h2>

<p>This command execute the script in dryrun mode. It is alias to command <code>sos run -n</code>.</p>

<h2 id="subcommand-prepare">subcommand <code>prepare</code></h2>

<p>This command execute the script in prepare mode. It is alias to command <code>sos run -p</code>.</p>

<h2 id="subcommand-convert">subcommand <code>convert</code></h2>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="err">$</span> <span class="n">sos</span> <span class="n">convert</span> <span class="o">-</span><span class="n">h</span>
<span class="kn">usage:</span> <span class="n">sos</span> <span class="n">convert</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">html</span> <span class="p">[</span><span class="n">FILENAME</span><span class="p">]]</span> <span class="p">[</span><span class="o">--</span><span class="n">markdown</span> <span class="p">[</span><span class="n">FILENAME</span><span class="p">]]</span> <span class="p">[</span><span class="o">--</span><span class="n">term</span><span class="p">]</span>
                   <span class="p">[</span><span class="o">--</span><span class="n">notebook</span> <span class="p">[</span><span class="n">FILENAME</span><span class="p">]]</span> <span class="p">[</span><span class="o">--</span><span class="n">sos</span> <span class="p">[</span><span class="n">SCRIPT</span><span class="p">]]</span> <span class="p">[</span><span class="o">-</span><span class="n">v</span> <span class="p">{</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">}]</span>
                   <span class="n">FILENAME</span> <span class="p">[</span><span class="n">WORKFLOW</span><span class="p">]</span>

<span class="n">The</span> <span class="n">show</span> <span class="n">command</span> <span class="n">displays</span> <span class="n">details</span> <span class="n">of</span> <span class="nb">all</span> <span class="n">workflows</span> <span class="n">defined</span> <span class="ow">in</span> <span class="n">a</span> <span class="n">script</span><span class="p">,</span>
<span class="n">including</span> <span class="n">description</span> <span class="n">of</span> <span class="n">script</span><span class="p">,</span> <span class="n">workflow</span><span class="p">,</span> <span class="n">steps</span><span class="p">,</span> <span class="ow">and</span> <span class="n">command</span> <span class="n">line</span> <span class="n">parameters</span><span class="o">.</span>
<span class="n">The</span> <span class="n">output</span> <span class="n">can</span> <span class="n">be</span> <span class="n">limited</span> <span class="n">to</span> <span class="n">a</span> <span class="n">specified</span> <span class="n">workflow</span> <span class="p">(</span><span class="n">which</span> <span class="n">can</span> <span class="n">be</span> <span class="n">a</span> <span class="n">subworkflow</span>
<span class="ow">or</span> <span class="n">a</span> <span class="n">combined</span> <span class="n">workflow</span><span class="p">)</span> <span class="k">if</span> <span class="n">a</span> <span class="n">workflow</span> <span class="ow">is</span> <span class="n">specified</span><span class="o">.</span>

<span class="n">positional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="n">FILENAME</span>              <span class="n">File</span> <span class="n">to</span> <span class="n">be</span> <span class="n">converted</span><span class="p">,</span> <span class="n">can</span> <span class="n">be</span> <span class="n">a</span> <span class="n">SoS</span> <span class="n">script</span> <span class="ow">or</span> <span class="n">a</span> <span class="n">Jupyter</span>
                        <span class="n">notebook</span><span class="o">.</span>
  <span class="n">WORKFLOW</span>              <span class="n">Workflow</span> <span class="n">to</span> <span class="n">be</span> <span class="n">converted</span> <span class="k">if</span> <span class="n">the</span> <span class="nb">file</span> <span class="n">being</span> <span class="n">converted</span>
                        <span class="ow">is</span> <span class="n">a</span> <span class="n">SoS</span> <span class="n">script</span><span class="o">.</span>

<span class="n">optional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="o">-</span><span class="n">h</span><span class="p">,</span> <span class="o">--</span><span class="n">help</span>            <span class="n">show</span> <span class="n">this</span> <span class="n">help</span> <span class="n">message</span> <span class="ow">and</span> <span class="nb">exit</span>
  <span class="o">--</span><span class="n">html</span> <span class="p">[</span><span class="n">FILENAME</span><span class="p">]</span>     <span class="n">Generate</span> <span class="n">a</span> <span class="n">syntax</span><span class="o">-</span><span class="n">highlighted</span> <span class="n">HTML</span> <span class="nb">file</span><span class="p">,</span> <span class="n">write</span> <span class="n">it</span> <span class="n">to</span> <span class="n">a</span>
                        <span class="n">specified</span> <span class="nb">file</span><span class="p">,</span> <span class="ow">or</span> <span class="n">view</span> <span class="ow">in</span> <span class="n">a</span> <span class="n">browser</span> <span class="k">if</span> <span class="n">no</span> <span class="n">filename</span> <span class="ow">is</span>
                        <span class="n">specified</span><span class="o">.</span> <span class="n">Additional</span> <span class="n">argument</span> <span class="o">--</span><span class="n">raw</span> <span class="n">can</span> <span class="n">be</span> <span class="n">used</span> <span class="n">to</span>
                        <span class="n">specify</span> <span class="n">a</span> <span class="n">URL</span> <span class="n">to</span> <span class="n">raw</span> <span class="nb">file</span><span class="p">,</span> <span class="n">arguments</span> <span class="o">--</span><span class="n">linenos</span> <span class="ow">and</span>
                        <span class="o">--</span><span class="n">style</span> <span class="n">can</span> <span class="n">be</span> <span class="n">used</span> <span class="n">to</span> <span class="n">customize</span> <span class="n">style</span> <span class="n">of</span> <span class="n">html</span> <span class="n">output</span><span class="o">.</span>
                        <span class="n">You</span> <span class="n">can</span> <span class="k">pass</span> <span class="n">an</span> <span class="n">arbitrary</span> <span class="n">name</span> <span class="n">to</span> <span class="n">option</span> <span class="o">--</span><span class="n">style</span> <span class="n">get</span> <span class="n">a</span>
                        <span class="nb">list</span> <span class="n">of</span> <span class="n">available</span> <span class="n">styles</span><span class="o">.</span>
  <span class="o">--</span><span class="n">markdown</span> <span class="p">[</span><span class="n">FILENAME</span><span class="p">]</span>
                        <span class="n">Convert</span> <span class="n">script</span> <span class="ow">or</span> <span class="n">workflow</span> <span class="n">to</span> <span class="n">markdown</span> <span class="n">format</span> <span class="ow">and</span>
                        <span class="n">write</span> <span class="n">it</span> <span class="n">to</span> <span class="n">specified</span> <span class="nb">file</span><span class="p">,</span> <span class="ow">or</span> <span class="n">standard</span> <span class="n">output</span> <span class="k">if</span> <span class="ow">not</span>
                        <span class="n">filename</span> <span class="ow">is</span> <span class="n">specified</span><span class="o">.</span>
  <span class="o">--</span><span class="n">term</span>                <span class="n">Output</span> <span class="n">syntax</span><span class="o">-</span><span class="n">highlighted</span> <span class="n">script</span> <span class="ow">or</span> <span class="n">workflow</span> <span class="n">to</span> <span class="n">the</span>
                        <span class="n">terminal</span><span class="o">.</span> <span class="n">Additional</span> <span class="n">arguments</span> <span class="o">--</span><span class="n">bg</span><span class="o">=</span><span class="n">light</span><span class="o">|</span><span class="n">dark</span>
                        <span class="o">--</span><span class="n">lineno</span> <span class="n">can</span> <span class="n">be</span> <span class="n">used</span> <span class="n">to</span> <span class="n">customized</span> <span class="n">output</span><span class="o">.</span>
  <span class="o">--</span><span class="n">notebook</span> <span class="p">[</span><span class="n">FILENAME</span><span class="p">]</span>
                        <span class="n">Convert</span> <span class="n">script</span> <span class="ow">or</span> <span class="n">workflow</span> <span class="n">to</span> <span class="n">jupyter</span> <span class="n">notebook</span> <span class="n">format</span>
                        <span class="ow">and</span> <span class="n">write</span> <span class="n">it</span> <span class="n">to</span> <span class="n">specified</span> <span class="nb">file</span><span class="p">,</span> <span class="ow">or</span> <span class="n">standard</span> <span class="n">output</span> <span class="k">if</span>
                        <span class="n">no</span> <span class="n">filename</span> <span class="ow">is</span> <span class="n">specified</span><span class="o">.</span> <span class="n">If</span> <span class="n">the</span> <span class="nb">input</span> <span class="nb">file</span> <span class="ow">is</span> <span class="n">a</span>
                        <span class="n">notebook</span><span class="p">,</span> <span class="n">it</span> <span class="n">will</span> <span class="n">be</span> <span class="n">converted</span> <span class="n">to</span> <span class="o">.</span><span class="n">sos</span> <span class="p">(</span><span class="n">see</span> <span class="n">option</span>
                        <span class="o">--</span><span class="n">sos</span><span class="p">)</span> <span class="n">then</span> <span class="n">to</span> <span class="n">notebook</span><span class="p">,</span> <span class="n">resetting</span> <span class="n">indexes</span> <span class="ow">and</span>
                        <span class="n">removing</span> <span class="nb">all</span> <span class="n">output</span> <span class="n">cells</span><span class="o">.</span>
  <span class="o">--</span><span class="n">sos</span> <span class="p">[</span><span class="n">SCRIPT</span><span class="p">]</span>        <span class="n">Convert</span> <span class="n">specified</span> <span class="n">Jupyter</span> <span class="n">notebook</span> <span class="n">to</span> <span class="n">SoS</span> <span class="n">format</span><span class="o">.</span> <span class="n">The</span>
                        <span class="n">output</span> <span class="ow">is</span> <span class="n">the</span> <span class="n">same</span> <span class="k">as</span> <span class="n">you</span> <span class="n">use</span> <span class="n">File</span> <span class="o">-&gt;</span> <span class="n">Download</span> <span class="k">as</span> <span class="o">-&gt;</span>
                        <span class="n">SoS</span> <span class="p">(</span><span class="o">.</span><span class="n">sos</span><span class="p">)</span> <span class="kn">from</span> <span class="nn">Jupyter</span> <span class="nn">with</span> <span class="nn">nbconvert</span> <span class="nn">version</span> <span class="mf">4.2</span><span class="o">.</span><span class="mi">0</span>
                        <span class="ow">or</span> <span class="n">higher</span> <span class="n">although</span> <span class="n">you</span> <span class="n">can</span> <span class="n">customize</span> <span class="n">output</span> <span class="n">using</span>
                        <span class="n">options</span> <span class="o">--</span><span class="n">reorder</span> <span class="p">(</span><span class="n">rearrange</span> <span class="n">notebook</span> <span class="n">cells</span> <span class="k">with</span>
                        <span class="n">execution</span> <span class="n">order</span><span class="p">),</span> <span class="o">--</span><span class="n">reset</span><span class="o">-</span><span class="n">index</span> <span class="p">(</span><span class="n">reset</span> <span class="n">indexes</span> <span class="n">to</span> <span class="mi">1</span><span class="p">,</span>
                        <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="o">..</span><span class="p">),</span> <span class="o">--</span><span class="n">add</span><span class="o">-</span><span class="n">header</span> <span class="p">(</span><span class="n">add</span> <span class="n">section</span> <span class="n">header</span> <span class="p">[</span><span class="n">index</span><span class="p">]</span> <span class="k">if</span>
                        <span class="n">the</span> <span class="n">cell</span> <span class="n">does</span> <span class="ow">not</span> <span class="n">start</span> <span class="k">with</span> <span class="n">a</span> <span class="n">header</span><span class="p">),</span> <span class="o">--</span><span class="n">no</span><span class="o">-</span><span class="n">index</span>
                        <span class="p">(</span><span class="n">does</span> <span class="ow">not</span> <span class="n">save</span> <span class="n">cell</span> <span class="n">index</span><span class="p">),</span> <span class="o">--</span><span class="n">remove</span><span class="o">-</span><span class="n">magic</span> <span class="p">(</span><span class="n">remove</span>
                        <span class="n">cell</span> <span class="n">magic</span><span class="p">),</span> <span class="ow">and</span> <span class="o">--</span><span class="n">md</span><span class="o">-</span><span class="n">to</span><span class="o">-</span><span class="n">report</span> <span class="p">(</span><span class="n">convert</span> <span class="n">markdown</span> <span class="n">cell</span>
                        <span class="n">to</span> <span class="n">code</span> <span class="n">cell</span> <span class="k">with</span> <span class="n">report</span><span class="o">.</span><span class="p">)</span>
  <span class="o">-</span><span class="n">v</span> <span class="p">{</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">},</span> <span class="o">--</span><span class="n">verbosity</span> <span class="p">{</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">}</span>
                        <span class="n">Output</span> <span class="n">error</span> <span class="p">(</span><span class="mi">0</span><span class="p">),</span> <span class="n">warning</span> <span class="p">(</span><span class="mi">1</span><span class="p">),</span> <span class="n">info</span> <span class="p">(</span><span class="mi">2</span><span class="p">),</span> <span class="n">debug</span> <span class="p">(</span><span class="mi">3</span><span class="p">)</span> <span class="ow">and</span>
                        <span class="n">trace</span> <span class="p">(</span><span class="mi">4</span><span class="p">)</span> <span class="n">information</span> <span class="n">to</span> <span class="n">standard</span> <span class="n">output</span> <span class="p">(</span><span class="n">default</span> <span class="n">to</span>
                        <span class="mi">2</span><span class="p">)</span><span class="o">.</span>

<span class="n">Extra</span> <span class="n">command</span> <span class="n">line</span> <span class="n">argument</span> <span class="n">could</span> <span class="n">be</span> <span class="n">specified</span> <span class="n">to</span> <span class="n">customize</span> <span class="n">the</span> <span class="n">style</span> <span class="n">of</span> <span class="n">html</span><span class="p">,</span>
<span class="n">markdown</span><span class="p">,</span> <span class="ow">and</span> <span class="n">terminal</span> <span class="n">output</span><span class="o">.</span></code></pre></figure>

<p>The <code>--html</code> option is very useful in displaying SoS scripts with proper syntax highlighting for different embedded languages. It can be used to generate a nice-looking .html file if you run</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="n">sos</span> <span class="n">convert</span> <span class="n">myscript</span><span class="o">.</span><span class="n">sos</span> <span class="o">--</span><span class="n">html</span> <span class="n">myscript</span><span class="o">.</span><span class="n">sos</span><span class="o">.</span><span class="n">html</span></code></pre></figure>

<p>or generate a temporary file and display it in a browser if you do not specify an output filename.</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="n">sos</span> <span class="n">convert</span> <span class="n">myscript</span><span class="o">.</span><span class="n">sos</span> <span class="o">--</span><span class="n">html</span> <span class="n">myscript</span><span class="o">.</span><span class="n">sos</span><span class="o">.</span><span class="n">html</span></code></pre></figure>

<p>For example, the example in our <a href="https://github.com/bpeng2000/SOS/wiki/Quick-Start">QuickStart Guide</a> can be displayed as</p>

<p><img src="/SOS/media/QuickStart.sos.html.png" alt="QuickStart.sos.html" class="img-responsive" /></p>

<p>It is also worth noting that you are a VIM user, it is highly recommended that you install the SoS syntax file so that your SoS scripts can be properly highlighted during editing. This is not only visually appealing but is very helpful in identifying syntax errors as you write. The same script in vim with a dark background would look look like</p>

<p><img src="/SOS/media/QuickStart.sos.vim.png" alt="QuickStart.sos.vim" class="img-responsive" /></p>

<h2 id="subcommand-config">subcommand <code>config</code></h2>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="kn">usage:</span> <span class="n">sos</span> <span class="n">config</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">g</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">c</span> <span class="n">CONFIG_FILE</span><span class="p">]</span>
                  <span class="p">(</span><span class="o">--</span><span class="n">get</span> <span class="p">[</span><span class="n">OPTION</span> <span class="p">[</span><span class="n">OPTION</span> <span class="o">...</span><span class="p">]]</span> <span class="o">|</span> <span class="o">--</span><span class="n">unset</span> <span class="n">OPTION</span> <span class="p">[</span><span class="n">OPTION</span> <span class="o">...</span><span class="p">]</span> <span class="o">|</span> <span class="o">--</span><span class="nb">set</span> <span class="n">KEY</span> <span class="n">VALUE</span> <span class="p">[</span><span class="n">KEY</span> <span class="n">VALUE</span> <span class="o">...</span><span class="p">])</span>
                  <span class="p">[</span><span class="o">-</span><span class="n">v</span> <span class="p">{</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">}]</span>

<span class="n">The</span> <span class="n">config</span> <span class="n">command</span> <span class="n">displays</span><span class="p">,</span> <span class="nb">set</span><span class="p">,</span> <span class="ow">and</span> <span class="n">unset</span> <span class="n">configuration</span> <span class="n">variables</span> <span class="n">defined</span> <span class="ow">in</span>
<span class="k">global</span> <span class="ow">or</span> <span class="n">local</span> <span class="n">configuration</span> <span class="n">files</span><span class="o">.</span>

<span class="n">optional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="o">-</span><span class="n">h</span><span class="p">,</span> <span class="o">--</span><span class="n">help</span>            <span class="n">show</span> <span class="n">this</span> <span class="n">help</span> <span class="n">message</span> <span class="ow">and</span> <span class="nb">exit</span>
  <span class="o">-</span><span class="n">g</span><span class="p">,</span> <span class="o">--</span><span class="k">global</span>          <span class="n">If</span> <span class="nb">set</span><span class="p">,</span> <span class="n">change</span> <span class="k">global</span> <span class="p">(</span><span class="o">~/.</span><span class="n">sos</span><span class="o">/</span><span class="n">config</span><span class="o">.</span><span class="n">yaml</span><span class="p">)</span> <span class="n">instead</span> <span class="n">of</span>
                        <span class="n">local</span> <span class="p">(</span><span class="o">.</span><span class="n">sos</span><span class="o">/</span><span class="n">config</span><span class="o">.</span><span class="n">yaml</span><span class="p">)</span> <span class="n">configuration</span>
  <span class="o">-</span><span class="n">c</span> <span class="n">CONFIG_FILE</span><span class="p">,</span> <span class="o">--</span><span class="n">config</span> <span class="n">CONFIG_FILE</span>
                        <span class="n">User</span> <span class="n">specified</span> <span class="n">configuration</span> <span class="nb">file</span> <span class="ow">in</span> <span class="n">YAML</span> <span class="n">format</span><span class="o">.</span> <span class="n">This</span>
                        <span class="nb">file</span> <span class="n">will</span> <span class="ow">not</span> <span class="n">be</span> <span class="n">automatically</span> <span class="n">loaded</span> <span class="n">by</span> <span class="n">SoS</span> <span class="n">but</span> <span class="n">can</span>
                        <span class="n">be</span> <span class="n">specified</span> <span class="n">using</span> <span class="n">option</span> <span class="sb">`-c`</span>
  <span class="o">--</span><span class="n">get</span> <span class="p">[</span><span class="n">OPTION</span> <span class="p">[</span><span class="n">OPTION</span> <span class="o">...</span><span class="p">]]</span>
                        <span class="n">Display</span> <span class="n">values</span> <span class="n">of</span> <span class="n">specified</span> <span class="n">configuration</span><span class="o">.</span> <span class="n">The</span>
                        <span class="n">arguments</span> <span class="n">of</span> <span class="n">this</span> <span class="n">option</span> <span class="n">can</span> <span class="n">be</span> <span class="n">a</span> <span class="n">single</span> <span class="n">configuration</span>
                        <span class="n">option</span> <span class="ow">or</span> <span class="n">a</span> <span class="nb">list</span> <span class="n">of</span> <span class="n">option</span><span class="o">.</span> <span class="n">Wildcard</span> <span class="n">characters</span> <span class="n">are</span>
                        <span class="n">allowed</span> <span class="n">to</span> <span class="n">match</span> <span class="n">more</span> <span class="n">options</span> <span class="p">(</span><span class="n">e</span><span class="o">.</span><span class="n">g</span><span class="o">.</span> <span class="s">&#39;*timeout&#39;</span><span class="p">,</span>
                        <span class="n">quotation</span> <span class="ow">is</span> <span class="n">needed</span> <span class="n">to</span> <span class="n">avoid</span> <span class="n">shell</span> <span class="n">expansion</span><span class="p">)</span><span class="o">.</span> <span class="n">If</span> <span class="n">no</span>
                        <span class="n">option</span> <span class="ow">is</span> <span class="n">given</span><span class="p">,</span> <span class="nb">all</span> <span class="n">options</span> <span class="n">will</span> <span class="n">be</span> <span class="n">outputted</span><span class="o">.</span>
  <span class="o">--</span><span class="n">unset</span> <span class="n">OPTION</span> <span class="p">[</span><span class="n">OPTION</span> <span class="o">...</span><span class="p">]</span>
                        <span class="n">Unset</span> <span class="p">(</span><span class="n">remove</span><span class="p">)</span> <span class="n">settings</span> <span class="k">for</span> <span class="n">specified</span> <span class="n">options</span><span class="o">.</span> <span class="n">The</span>
                        <span class="n">arguments</span> <span class="n">of</span> <span class="n">this</span> <span class="n">option</span> <span class="n">can</span> <span class="n">be</span> <span class="n">a</span> <span class="n">single</span> <span class="n">configuration</span>
                        <span class="n">option</span> <span class="ow">or</span> <span class="n">a</span> <span class="nb">list</span> <span class="n">of</span> <span class="n">option</span><span class="o">.</span> <span class="n">Wildcard</span> <span class="n">characters</span> <span class="n">are</span>
                        <span class="n">allowed</span> <span class="n">to</span> <span class="n">match</span> <span class="n">more</span> <span class="n">options</span> <span class="p">(</span><span class="n">e</span><span class="o">.</span><span class="n">g</span><span class="o">.</span> <span class="s">&#39;*timeout&#39;</span><span class="p">,</span> <span class="ow">or</span> <span class="s">&#39;*&#39;</span>
                        <span class="k">for</span> <span class="nb">all</span> <span class="n">options</span><span class="p">,</span> <span class="n">quotation</span> <span class="ow">is</span> <span class="n">needed</span> <span class="n">to</span> <span class="n">avoid</span> <span class="n">shell</span>
                        <span class="n">expansion</span><span class="p">)</span><span class="o">.</span>
  <span class="o">--</span><span class="nb">set</span> <span class="n">KEY</span> <span class="n">VALUE</span> <span class="p">[</span><span class="n">KEY</span> <span class="n">VALUE</span> <span class="o">...</span><span class="p">]</span>
                        <span class="o">--</span><span class="nb">set</span> <span class="n">KEY</span> <span class="n">VALUE</span> <span class="n">sets</span> <span class="n">VALUE</span> <span class="n">to</span> <span class="n">variable</span> <span class="n">KEY</span><span class="o">.</span> <span class="n">The</span> <span class="n">value</span>
                        <span class="n">can</span> <span class="n">be</span> <span class="nb">any</span> <span class="n">valid</span> <span class="n">python</span> <span class="n">expression</span> <span class="p">(</span><span class="n">e</span><span class="o">.</span><span class="n">g</span><span class="o">.</span> <span class="mi">5</span> <span class="k">for</span> <span class="n">integer</span>
                        <span class="mi">5</span> <span class="ow">and</span> <span class="s">&#39;{&quot;c&quot;: 2, &quot;d&quot;: 1}&#39;</span> <span class="k">for</span> <span class="n">a</span> <span class="n">dictionary</span><span class="p">)</span> <span class="k">with</span>
                        <span class="n">invalid</span> <span class="n">expression</span> <span class="p">(</span><span class="n">e</span><span class="o">.</span><span class="n">g</span><span class="o">.</span> <span class="n">val</span> <span class="n">without</span> <span class="n">quote</span><span class="p">)</span> <span class="n">considered</span>
                        <span class="k">as</span> <span class="n">string</span><span class="o">.</span> <span class="n">Syntax</span> <span class="s">&#39;A.B=v&#39;</span> <span class="n">can</span> <span class="n">be</span> <span class="n">used</span> <span class="n">to</span> <span class="n">add</span> <span class="p">{</span><span class="s">&#39;B&#39;</span><span class="p">:</span> <span class="n">v</span><span class="p">}</span>
                        <span class="n">to</span> <span class="n">dictionary</span> <span class="s">&#39;A&#39;</span><span class="p">,</span> <span class="ow">and</span> <span class="o">--</span><span class="nb">set</span> <span class="n">KEY</span> <span class="n">VALUE1</span> <span class="n">VALUE2</span> <span class="o">...</span>
                        <span class="n">will</span> <span class="n">create</span> <span class="n">a</span> <span class="nb">list</span> <span class="k">with</span> <span class="n">multiple</span> <span class="n">values</span><span class="o">.</span>
  <span class="o">-</span><span class="n">v</span> <span class="p">{</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">},</span> <span class="o">--</span><span class="n">verbosity</span> <span class="p">{</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">}</span>
                        <span class="n">Output</span> <span class="n">error</span> <span class="p">(</span><span class="mi">0</span><span class="p">),</span> <span class="n">warning</span> <span class="p">(</span><span class="mi">1</span><span class="p">),</span> <span class="n">info</span> <span class="p">(</span><span class="mi">2</span><span class="p">),</span> <span class="n">debug</span> <span class="p">(</span><span class="mi">3</span><span class="p">)</span> <span class="ow">and</span>
                        <span class="n">trace</span> <span class="p">(</span><span class="mi">4</span><span class="p">)</span> <span class="n">information</span> <span class="n">to</span> <span class="n">standard</span> <span class="n">output</span> <span class="p">(</span><span class="n">default</span> <span class="n">to</span>
                        <span class="mi">2</span><span class="p">)</span><span class="o">.</span></code></pre></figure>

<p>This subcommand can read/write three configuration files</p>

<ol>
  <li><strong><code>~/.sos/config.yaml</code></strong> A global SoS configuration file. (option <code>--global</code>)</li>
  <li><strong><code>.sos/config.yaml</code></strong> Project specific SoS configuration file. (default)</li>
  <li><strong>User specified</strong> Any configuration file specified with option <code>-c</code> (<code>--config</code>).</li>
</ol>

<p>The global and local configuration files will be loaded by default. Other configuration files can be specified using option <code>-c</code> of the <code>sos run</code> command. Options defined in these configuration file will be available to the script as dictionary <code>CONFIG</code>.</p>

<p>Note that you can set complex datatypes from command line using Python expressions,</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="n">sos</span> <span class="n">config</span> <span class="o">--</span><span class="nb">set</span> <span class="n">a</span> <span class="s">&#39;{&quot;b&quot;:1}&#39;</span></code></pre></figure>

<p>although</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="n">sos</span> <span class="n">config</span> <span class="o">--</span><span class="nb">set</span> <span class="n">a</span><span class="o">.</span><span class="n">b</span> <span class="mi">1</span></code></pre></figure>

<p>is easier to write if you only need to set one key <code>b</code> in dictionary <code>a</code>.</p>

<h2 id="subcommand-remove">subcommand <code>remove</code></h2>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="err">$</span> <span class="n">sos</span> <span class="n">remove</span> <span class="o">-</span><span class="n">h</span>
<span class="kn">usage:</span> <span class="n">sos</span> <span class="n">remove</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">t</span> <span class="o">|</span> <span class="o">-</span><span class="n">u</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">n</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">y</span><span class="p">]</span> <span class="p">[</span><span class="n">FILE_OR_DIR</span> <span class="p">[</span><span class="n">FILE_OR_DIR</span> <span class="o">...</span><span class="p">]]</span>

<span class="n">Remove</span> <span class="n">specified</span> <span class="n">files</span> <span class="ow">and</span> <span class="n">directories</span> <span class="ow">and</span> <span class="n">their</span> <span class="n">signatures</span> <span class="p">(</span><span class="k">if</span> <span class="n">available</span><span class="p">)</span><span class="o">.</span>
<span class="n">Optionally</span><span class="p">,</span> <span class="n">you</span> <span class="n">can</span> <span class="n">remove</span> <span class="n">only</span> <span class="n">tracked</span> <span class="n">files</span> <span class="p">(</span><span class="nb">input</span><span class="p">,</span> <span class="n">output</span> <span class="ow">and</span> <span class="n">intermediate</span>
<span class="n">files</span> <span class="n">of</span> <span class="n">executed</span> <span class="n">workflows</span><span class="p">)</span> <span class="ow">or</span> <span class="n">untracked</span> <span class="nb">file</span> <span class="kn">from</span> <span class="nn">specified</span> <span class="nn">files</span> <span class="nn">and</span><span class="o">/</span><span class="ow">or</span>
<span class="n">directories</span><span class="o">.</span>

<span class="n">positional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="n">FILE_OR_DIR</span>  <span class="n">Files</span> <span class="ow">and</span> <span class="n">directories</span> <span class="n">to</span> <span class="n">be</span> <span class="n">removed</span><span class="p">,</span> <span class="n">which</span> <span class="n">should</span> <span class="n">be</span> <span class="n">under</span> <span class="n">the</span>
               <span class="n">current</span> <span class="n">directory</span> <span class="p">(</span><span class="n">default</span><span class="p">)</span><span class="o">.</span> <span class="n">All</span><span class="p">,</span> <span class="n">tracked</span><span class="p">,</span> <span class="ow">or</span> <span class="n">untracked</span> <span class="n">files</span>
               <span class="n">will</span> <span class="n">be</span> <span class="n">removed</span> <span class="n">depending</span> <span class="n">on</span> <span class="n">other</span> <span class="n">options</span> <span class="p">(</span><span class="s">&#39;-t&#39;</span> <span class="ow">or</span> <span class="s">&#39;-u&#39;</span><span class="p">)</span><span class="o">.</span> <span class="n">For</span>
               <span class="n">safety</span> <span class="n">reasons</span><span class="p">,</span> <span class="n">files</span> <span class="n">under</span> <span class="n">the</span> <span class="n">current</span> <span class="n">directory</span> <span class="n">have</span> <span class="n">to</span> <span class="n">be</span>
               <span class="n">listed</span> <span class="p">(</span><span class="ow">not</span> <span class="k">as</span> <span class="n">files</span> <span class="n">under</span> <span class="o">.</span><span class="p">)</span> <span class="n">to</span> <span class="n">be</span> <span class="n">removed</span><span class="o">.</span>

<span class="n">optional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="o">-</span><span class="n">h</span><span class="p">,</span> <span class="o">--</span><span class="n">help</span>   <span class="n">show</span> <span class="n">this</span> <span class="n">help</span> <span class="n">message</span> <span class="ow">and</span> <span class="nb">exit</span>
  <span class="o">-</span><span class="n">t</span>           <span class="n">Remove</span> <span class="n">tracked</span> <span class="n">files</span> <span class="ow">and</span> <span class="n">their</span> <span class="n">signatures</span> <span class="kn">from</span> <span class="nn">specified</span> <span class="nn">files</span>
               <span class="ow">and</span> <span class="n">directories</span><span class="o">.</span>
  <span class="o">-</span><span class="n">u</span>           <span class="n">Remove</span> <span class="n">untracked</span> <span class="n">files</span> <span class="kn">from</span> <span class="nn">specified</span> <span class="nn">files</span> <span class="nn">and</span> <span class="nn">directories.</span>
  <span class="o">-</span><span class="n">n</span>           <span class="n">List</span> <span class="n">files</span> <span class="ow">or</span> <span class="n">directories</span> <span class="n">to</span> <span class="n">be</span> <span class="n">removed</span><span class="p">,</span> <span class="n">without</span> <span class="n">actually</span>
               <span class="n">removing</span> <span class="n">them</span><span class="o">.</span>
  <span class="o">-</span><span class="n">y</span>           <span class="n">Remove</span> <span class="n">files</span> <span class="n">without</span> <span class="n">confirmation</span><span class="p">,</span> <span class="n">suitable</span> <span class="k">for</span> <span class="n">batch</span> <span class="n">removal</span>
               <span class="n">of</span> <span class="n">files</span><span class="o">.</span></code></pre></figure>

<h2 id="subcommand-pack">subcommand <code>pack</code></h2>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="err">$</span> <span class="n">sos</span> <span class="n">pack</span> <span class="o">-</span><span class="n">h</span>
<span class="kn">usage:</span> <span class="n">sos</span> <span class="n">pack</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">o</span> <span class="n">OUTPUT</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">i</span> <span class="p">[</span><span class="n">INCLUDE</span> <span class="p">[</span><span class="n">INCLUDE</span> <span class="o">...</span><span class="p">]]]</span>
                <span class="p">[</span><span class="o">-</span><span class="n">e</span> <span class="p">[</span><span class="n">EXCLUDE</span> <span class="p">[</span><span class="n">EXCLUDE</span> <span class="o">...</span><span class="p">]]]</span> <span class="p">[</span><span class="o">-</span><span class="n">a</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">m</span> <span class="n">MESSAGE</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">y</span><span class="p">]</span>
                <span class="p">[</span><span class="o">-</span><span class="n">v</span> <span class="p">{</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">}]</span>
                <span class="p">[</span><span class="n">session</span><span class="p">]</span>

<span class="n">positional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="n">session</span>               <span class="n">ID</span> <span class="n">of</span> <span class="n">the</span> <span class="n">session</span> <span class="n">to</span> <span class="n">be</span> <span class="n">saved</span><span class="p">,</span> <span class="n">which</span> <span class="n">can</span> <span class="n">be</span> <span class="nb">any</span> <span class="n">number</span>
                        <span class="n">of</span> <span class="n">digits</span> <span class="k">as</span> <span class="nb">long</span> <span class="k">as</span> <span class="n">it</span> <span class="n">can</span> <span class="n">uniquely</span> <span class="n">determine</span> <span class="n">a</span>
                        <span class="n">workflow</span> <span class="n">session</span><span class="o">.</span> <span class="n">This</span> <span class="n">parameter</span> <span class="n">can</span> <span class="n">be</span> <span class="n">ignored</span> <span class="k">if</span>
                        <span class="n">only</span> <span class="n">one</span> <span class="n">session</span> <span class="ow">is</span> <span class="n">available</span><span class="o">.</span>

<span class="n">optional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="o">-</span><span class="n">h</span><span class="p">,</span> <span class="o">--</span><span class="n">help</span>            <span class="n">show</span> <span class="n">this</span> <span class="n">help</span> <span class="n">message</span> <span class="ow">and</span> <span class="nb">exit</span>
  <span class="o">-</span><span class="n">o</span> <span class="n">OUTPUT</span><span class="p">,</span> <span class="o">--</span><span class="n">output</span> <span class="n">OUTPUT</span>
                        <span class="n">Output</span> <span class="nb">file</span><span class="p">,</span> <span class="n">which</span> <span class="n">can</span> <span class="n">be</span> <span class="n">a</span> <span class="nb">file</span> <span class="k">with</span> <span class="n">extension</span> <span class="s">&quot;.sar&quot;</span>
                        <span class="p">(</span><span class="n">the</span> <span class="n">extension</span> <span class="n">will</span> <span class="n">be</span> <span class="n">be</span> <span class="n">automatically</span> <span class="n">appended</span> <span class="k">if</span>
                        <span class="n">needed</span><span class="p">),</span> <span class="ow">or</span> <span class="s">&quot;-&quot;</span> <span class="k">for</span> <span class="n">standard</span> <span class="n">output</span> <span class="p">(</span><span class="n">default</span><span class="p">)</span><span class="o">.</span>
  <span class="o">-</span><span class="n">i</span> <span class="p">[</span><span class="n">INCLUDE</span> <span class="p">[</span><span class="n">INCLUDE</span> <span class="o">...</span><span class="p">]],</span> <span class="o">--</span><span class="n">include</span> <span class="p">[</span><span class="n">INCLUDE</span> <span class="p">[</span><span class="n">INCLUDE</span> <span class="o">...</span><span class="p">]]</span>
                        <span class="n">Additional</span> <span class="n">files</span> <span class="ow">or</span> <span class="n">directories</span> <span class="n">to</span> <span class="n">be</span> <span class="n">incldued</span> <span class="ow">in</span> <span class="n">the</span>
                        <span class="n">archive</span><span class="o">.</span> <span class="n">SoS</span> <span class="n">will</span> <span class="n">archive</span> <span class="nb">all</span> <span class="n">files</span> <span class="n">under</span> <span class="n">specified</span>
                        <span class="n">directories</span><span class="p">,</span> <span class="n">including</span> <span class="n">hidden</span> <span class="n">directories</span> <span class="n">such</span> <span class="k">as</span>
                        <span class="s">&quot;.git&quot;</span><span class="o">.</span> <span class="n">Option</span> <span class="o">--</span><span class="n">exclude</span> <span class="n">could</span> <span class="n">be</span> <span class="n">used</span> <span class="n">to</span> <span class="n">exclude</span>
                        <span class="n">these</span> <span class="n">files</span><span class="o">.</span>
  <span class="o">-</span><span class="n">e</span> <span class="p">[</span><span class="n">EXCLUDE</span> <span class="p">[</span><span class="n">EXCLUDE</span> <span class="o">...</span><span class="p">]],</span> <span class="o">--</span><span class="n">exclude</span> <span class="p">[</span><span class="n">EXCLUDE</span> <span class="p">[</span><span class="n">EXCLUDE</span> <span class="o">...</span><span class="p">]]</span>
                        <span class="n">Files</span> <span class="n">that</span> <span class="n">should</span> <span class="n">be</span> <span class="n">excluded</span> <span class="kn">from</span> <span class="nn">archive.</span> <span class="nn">The</span>
                        <span class="n">parameter</span> <span class="n">should</span> <span class="n">be</span> <span class="n">one</span> <span class="ow">or</span> <span class="n">more</span> <span class="n">patterns</span> <span class="n">that</span> <span class="n">match</span>
                        <span class="n">the</span> <span class="n">whole</span> <span class="n">path</span> <span class="p">(</span><span class="n">e</span><span class="o">.</span><span class="n">g</span><span class="o">.</span> <span class="s">&quot;output/*.log&quot;</span> <span class="ow">or</span> <span class="nb">file</span> <span class="ow">or</span>
                        <span class="n">directory</span> <span class="n">names</span> <span class="n">such</span> <span class="k">as</span> <span class="s">&quot;tmp&quot;</span> <span class="ow">or</span> <span class="s">&quot;*.bam&quot;</span><span class="o">.</span>
  <span class="o">-</span><span class="n">a</span><span class="p">,</span> <span class="o">--</span><span class="nb">all</span>             <span class="n">Include</span> <span class="nb">all</span> <span class="n">tracked</span> <span class="n">files</span> <span class="n">even</span> <span class="k">if</span> <span class="n">they</span> <span class="n">reside</span> <span class="n">outside</span>
                        <span class="n">of</span> <span class="n">the</span> <span class="n">current</span> <span class="n">working</span> <span class="n">directory</span><span class="o">.</span>
  <span class="o">-</span><span class="n">m</span> <span class="n">MESSAGE</span><span class="p">,</span> <span class="o">--</span><span class="n">message</span> <span class="n">MESSAGE</span>
                        <span class="n">A</span> <span class="n">short</span> <span class="n">message</span> <span class="n">to</span> <span class="n">be</span> <span class="n">included</span> <span class="n">into</span> <span class="n">the</span> <span class="n">archive</span><span class="o">.</span>
                        <span class="n">Because</span> <span class="n">the</span> <span class="n">message</span> <span class="n">would</span> <span class="n">be</span> <span class="n">lost</span> <span class="n">during</span> <span class="n">unpacking</span><span class="p">,</span> <span class="n">it</span>
                        <span class="ow">is</span> <span class="n">highly</span> <span class="n">recommended</span> <span class="n">that</span> <span class="n">you</span> <span class="n">create</span> <span class="n">a</span> <span class="n">README</span> <span class="nb">file</span>
                        <span class="ow">and</span> <span class="n">include</span> <span class="n">it</span> <span class="k">with</span> <span class="n">option</span> <span class="o">--</span><span class="n">include</span><span class="o">.</span>
  <span class="o">-</span><span class="n">y</span><span class="p">,</span> <span class="o">--</span><span class="n">yes</span>             <span class="n">Overwrite</span> <span class="n">output</span> <span class="nb">file</span> <span class="k">if</span> <span class="n">it</span> <span class="n">already</span> <span class="n">exists</span>
  <span class="o">-</span><span class="n">v</span> <span class="p">{</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">},</span> <span class="o">--</span><span class="n">verbosity</span> <span class="p">{</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">}</span>
                        <span class="n">Output</span> <span class="n">error</span> <span class="p">(</span><span class="mi">0</span><span class="p">),</span> <span class="n">warning</span> <span class="p">(</span><span class="mi">1</span><span class="p">),</span> <span class="n">info</span> <span class="p">(</span><span class="mi">2</span><span class="p">),</span> <span class="n">debug</span> <span class="p">(</span><span class="mi">3</span><span class="p">)</span> <span class="ow">and</span>
                        <span class="n">trace</span> <span class="p">(</span><span class="mi">4</span><span class="p">)</span> <span class="n">information</span> <span class="n">to</span> <span class="n">standard</span> <span class="n">output</span> <span class="p">(</span><span class="n">default</span> <span class="n">to</span>
                        <span class="mi">2</span><span class="p">)</span><span class="o">.</span></code></pre></figure>

<h2 id="subcommand-unpack">subcommand <code>unpack</code></h2>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="err">$</span> <span class="n">sos</span> <span class="n">unpack</span> <span class="o">-</span><span class="n">h</span>
<span class="kn">usage:</span> <span class="n">sos</span> <span class="n">unpack</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">d</span> <span class="n">DEST</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">l</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">e</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">n</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">y</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">v</span> <span class="p">{</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">}]</span>
                  <span class="n">archive</span> <span class="p">[</span><span class="n">files</span> <span class="p">[</span><span class="n">files</span> <span class="o">...</span><span class="p">]]</span>

<span class="n">positional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="n">archive</span>               <span class="n">SoS</span> <span class="n">archive</span> <span class="n">saved</span> <span class="n">by</span> <span class="n">command</span> <span class="n">sos</span> <span class="n">pack</span>
  <span class="n">files</span>                 <span class="n">An</span> <span class="n">optional</span> <span class="nb">list</span> <span class="n">of</span> <span class="n">files</span> <span class="n">to</span> <span class="n">be</span> <span class="n">processed</span><span class="p">,</span> <span class="n">which</span> <span class="n">can</span>
                        <span class="n">be</span> <span class="n">exact</span> <span class="n">filenames</span> <span class="ow">or</span> <span class="n">patterns</span> <span class="p">(</span><span class="n">e</span><span class="o">.</span><span class="n">g</span><span class="o">.</span> <span class="s">&quot;*.bam&quot;</span><span class="p">)</span>

<span class="n">optional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="o">-</span><span class="n">h</span><span class="p">,</span> <span class="o">--</span><span class="n">help</span>            <span class="n">show</span> <span class="n">this</span> <span class="n">help</span> <span class="n">message</span> <span class="ow">and</span> <span class="nb">exit</span>
  <span class="o">-</span><span class="n">d</span> <span class="n">DEST</span><span class="p">,</span> <span class="o">--</span><span class="n">dest</span> <span class="n">DEST</span>  <span class="n">Directory</span> <span class="n">where</span> <span class="n">a</span> <span class="n">sos</span> <span class="n">archive</span> <span class="n">would</span> <span class="n">be</span> <span class="n">unpacked</span><span class="o">.</span>
                        <span class="n">Default</span> <span class="n">to</span> <span class="n">current</span> <span class="n">directory</span><span class="o">.</span>
  <span class="o">-</span><span class="n">l</span><span class="p">,</span> <span class="o">--</span><span class="nb">list</span>            <span class="n">List</span> <span class="n">content</span> <span class="n">of</span> <span class="n">the</span> <span class="n">archive</span> <span class="n">instead</span> <span class="n">of</span> <span class="n">extracting</span> <span class="n">it</span><span class="o">.</span>
                        <span class="n">The</span> <span class="n">names</span><span class="p">,</span> <span class="n">uncompressed</span> <span class="nb">file</span> <span class="n">sizes</span> <span class="ow">and</span> <span class="n">modification</span>
                        <span class="n">dates</span> <span class="ow">and</span> <span class="n">times</span> <span class="n">of</span> <span class="n">the</span> <span class="n">specified</span> <span class="n">files</span> <span class="n">are</span> <span class="n">printed</span><span class="p">,</span>
                        <span class="n">along</span> <span class="k">with</span> <span class="n">totals</span> <span class="k">for</span> <span class="nb">all</span> <span class="n">files</span> <span class="n">specified</span><span class="o">.</span>
  <span class="o">-</span><span class="n">e</span><span class="p">,</span> <span class="o">--</span><span class="n">external</span>        <span class="n">Extract</span> <span class="n">files</span> <span class="n">outside</span> <span class="n">of</span> <span class="n">the</span> <span class="n">project</span> <span class="n">to</span> <span class="n">their</span> <span class="n">external</span>
                        <span class="n">destinations</span><span class="o">.</span> <span class="n">This</span> <span class="n">option</span> <span class="n">can</span> <span class="n">be</span> <span class="n">dangerous</span> <span class="n">because</span> <span class="n">it</span>
                        <span class="n">can</span> <span class="n">overwrite</span> <span class="n">system</span> <span class="n">files</span> <span class="n">silently</span> <span class="k">if</span> <span class="n">accompanied</span>
                        <span class="k">with</span> <span class="n">option</span> <span class="o">-</span><span class="n">y</span><span class="o">.</span>
  <span class="o">-</span><span class="n">n</span><span class="p">,</span> <span class="o">--</span><span class="n">no</span>              <span class="n">Do</span> <span class="ow">not</span> <span class="n">overwrite</span> <span class="n">existing</span> <span class="n">files</span> <span class="n">without</span> <span class="n">promoting</span>
                        <span class="n">users</span><span class="o">.</span>
  <span class="o">-</span><span class="n">y</span><span class="p">,</span> <span class="o">--</span><span class="n">yes</span>             <span class="n">Overwrite</span> <span class="n">existing</span> <span class="n">files</span> <span class="n">without</span> <span class="n">promoting</span> <span class="n">users</span><span class="o">.</span> <span class="n">This</span>
                        <span class="n">option</span> <span class="n">can</span> <span class="n">be</span> <span class="n">dangerous</span> <span class="n">to</span> <span class="n">use</span><span class="o">.</span> <span class="n">Note</span> <span class="n">that</span> <span class="n">SoS</span> <span class="n">checks</span>
                        <span class="nb">file</span> <span class="n">signature</span> <span class="ow">and</span> <span class="n">ignores</span> <span class="n">existing</span> <span class="n">files</span> <span class="n">that</span> <span class="n">are</span>
                        <span class="n">identical</span> <span class="n">to</span> <span class="n">those</span> <span class="ow">in</span> <span class="n">the</span> <span class="n">archive</span><span class="o">.</span>
  <span class="o">-</span><span class="n">v</span> <span class="p">{</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">},</span> <span class="o">--</span><span class="n">verbosity</span> <span class="p">{</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">}</span>
                        <span class="n">Output</span> <span class="n">error</span> <span class="p">(</span><span class="mi">0</span><span class="p">),</span> <span class="n">warning</span> <span class="p">(</span><span class="mi">1</span><span class="p">),</span> <span class="n">info</span> <span class="p">(</span><span class="mi">2</span><span class="p">),</span> <span class="n">debug</span> <span class="p">(</span><span class="mi">3</span><span class="p">)</span> <span class="ow">and</span>
                        <span class="n">trace</span> <span class="p">(</span><span class="mi">4</span><span class="p">)</span> <span class="n">information</span> <span class="n">to</span> <span class="n">standard</span> <span class="n">output</span> <span class="p">(</span><span class="n">default</span> <span class="n">to</span>
                        <span class="mi">2</span><span class="p">)</span><span class="o">.</span></code></pre></figure>

<h2 id="command-sos-runner">Command <code>sos-runner</code></h2>

<p>Command <code>sos-runner</code> is a shortcut for <code>sos run</code> so</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="n">sos</span><span class="o">-</span><span class="n">runner</span> <span class="n">script</span></code></pre></figure>

<p>is equivalent to</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="n">sos</span> <span class="n">run</span> <span class="n">script</span></code></pre></figure>

<p>This allows a SoS script to be executed directly if it is executable with shebang line</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="c">#!/usr/bin/env sos-runner</span></code></pre></figure>


                            </div>
                            <button type="button" class="btn btn-default" data-dismiss="modal"><i class="fa fa-times"></i> Close</button>
                        </div>
                    </div> 
                    <div class="footer-wrapper">
                      <footer role="contentinfo">
                        <span>&copy; 2016 Bo Peng.</span>
                      </footer>
                    </div>
                </div>
            </div>
        </div>
    </div> 


   <div class="portfolio-modal modal fade" id="SoS-Syntax" tabindex="-1" role="dialog" aria-hidden="true">
        <div class="modal-content">
            <div class="close-modal" data-dismiss="modal">
                <div class="lr">
                    <div class="rl">
                    </div>
                </div>
            </div>
            <div class="container">
                <div class="row">
                 <div class="col-lg-8 col-lg-offset-2">
                        <h1>2.SoS-Syntax</h1>
                        <div class="modal-body">
                         
                            <hr class="star-primary">
                            <p>SoS uses <a href="http://www.python.org">Python</a> expressions and statements. If you are unfamiliar with Python, you can learn some basics of Python, usually in less than half a day, by reading some Python tutorials (e.g. <a href="https://docs.python.org/3/tutorial/">the official python tutorial</a>). This <a href="https://docs.python.org/3/tutorial/introduction.html">short introduction</a> is good enough for you to use SoS.</p>

<p>The only difference between SoS’ python syntax and standard Python syntax is that <strong>string literals are format string that will be <a href="#string-interpolation">interpolated</a> using SoS defined sigil</strong>.</p>

<p>and introduces the following SoS-specific syntax:</p>

<ul>
  <li><code>[names: options]</code> line to define steps of workflows</li>
  <li>special directives such as <code>input:</code></li>
  <li>A special script style of function calls for SoS actions</li>
  <li>structural preprocessors such as <code>%include</code>, <code>%if</code>, <code>%cell</code></li>
</ul>

<h2 id="string-interpolation">String interpolation</h2>

<p>On top of python string manipulation functions and similar to recently introduced (Python 3.6) format string, SoS uses string interpolation to replace variables and expressions within a string with their values. For example, expressions <code>resource_path</code>, <code>sample_names[0]</code> and <code>sample_names</code> would be replaced by their values in the following string definitions.</p>

<figure class="highlight"><pre><code class="language-python" data-lang="python"><span class="n">resource_path</span> <span class="o">=</span> <span class="s">&#39;~/.sos/resources&#39;</span>
<span class="n">ref_genome</span>    <span class="o">=</span> <span class="s">&#39;${resource_path}/hg19/refGenome.fasta&#39;</span>     <span class="c"># &#39;~/.sos/resources/hg19/refGenome.fasta&#39;</span>

<span class="n">sample_names</span>  <span class="o">=</span> <span class="p">[</span><span class="s">&#39;A&#39;</span><span class="p">,</span> <span class="s">&#39;B&#39;</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">]</span>
<span class="n">title</span>         <span class="o">=</span> <span class="s">&#39;Sample ${sample_names[0]} results&#39;</span>         <span class="c"># &#39;Sample A results&#39;</span>
<span class="n">all_names</span>     <span class="o">=</span> <span class="s">&#39;Samples ${sample_names}&#39;</span>                   <span class="c"># &#39;Samples A B C&#39;</span></code></pre></figure>

<p>Depending on the return type of the expression</p>

<ul>
  <li>String representations (<code>repr(obj)</code>) are returned for objects in simple Python types (e.g. string, True, False, None, numeric numbers)</li>
  <li>
    <p>For objects with an iterator interface (e.g. Python <code>list</code>, <code>tuple</code>, <code>dict</code>, and <code>set</code>), SoS join the string representation of each item by a space or comma as specified by conversion flag <code>!,</code> (see <a href="#conversion-and-format">conversion and format</a> for details). More specifically,</p>

    <ul>
      <li>List of strings will be converted to a string by joining strings with a space or comma.</li>
      <li>Dictionary of strings will be converted to a string by joining dictionary keys with no guarantee on the order of values.</li>
    </ul>
  </li>
</ul>

<p>If you are unhappy with the string conversion, you can always format your object using python expressions such as <code>${repr(obj)}</code> and <code>${', '.join(x for x in obj)}</code>.</p>

<p>SoS supports nested interpolation, for example</p>

<figure class="highlight"><pre><code class="language-python" data-lang="python"><span class="s">&#39;${_input[${index}]}&#39;</span></code></pre></figure>

<p>would evaluate <code>${index}</code> and then <code>${_input[?]}</code> where <code>?</code> is the result of <code>${index}</code>.</p>

<p>Note that you can continue to use Python string functions such as</p>

<figure class="highlight"><pre><code class="language-python" data-lang="python"><span class="n">ref_genome</span> <span class="o">=</span> <span class="n">resource_path</span> <span class="o">+</span> <span class="s">&#39;/hg19/refGenome.fasta&#39;</span>
<span class="n">title</span>      <span class="o">=</span> <span class="s">&#39;Result for the first sample {} is invalid&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">sample_names</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span></code></pre></figure>

<p>but string interpolation is recommended for multi-line scripts because it is easier to read.</p>

<p>SoS interpolation also support all string format and conversion specification as in the <a href="https://docs.python.org/2/library/string.html#formatspec">Python string format specifier</a>, that is to say, you can use <code>: specifier</code> at the end of the expression to control the format of the output. For example</p>

<ul>
  <li><code>${1/3. :.2f}</code> in a string literal yields <code>0.33</code> instead of its long representation <code>0.3333333333333333</code>, and</li>
  <li><code>${filename!r}</code> in a string literal yields <code>'Article by "Jane Eyre"'</code> (note the quotation mark) if <code>filename='Article by "Jane Eyre"'</code>.</li>
</ul>

<p>The latter is interesting because SoS uses correct quotation marks for filenames with quotation marks, making</p>

<figure class="highlight"><pre><code class="language-python" data-lang="python"><span class="n">R</span><span class="p">(</span><span class="sb">``</span><span class="err">`</span>
<span class="n">read</span><span class="o">.</span><span class="n">csv</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="err">!</span><span class="n">r</span><span class="p">})</span>
<span class="sb">``</span><span class="err">`</span><span class="p">)</span></code></pre></figure>

<p>a safer choice than</p>

<figure class="highlight"><pre><code class="language-python" data-lang="python"><span class="n">R</span><span class="p">(</span><span class="s">&#39;&#39;&#39;</span>
<span class="s">read.csv(&quot;${_input}&quot;)</span>
<span class="s">&#39;&#39;&#39;</span><span class="p">)</span></code></pre></figure>

<p>because <code>${_input}</code> might contain quotation marks.</p>

<p>SoS also supports a few other conversion operators, for</p>

<figure class="highlight"><pre><code class="language-python" data-lang="python"><span class="n">file1</span><span class="o">=</span><span class="s">&#39;file 1.txt&#39;</span>
<span class="n">file2</span><span class="o">=</span><span class="s">&#39;~/SoS/test.sos&#39;</span>
<span class="n">files</span><span class="o">=</span><span class="p">[</span><span class="s">&#39;a.txt&#39;</span><span class="p">,</span> <span class="s">&#39;b.txt&#39;</span><span class="p">]</span></code></pre></figure>

<table>
  <thead>
    <tr>
      <th style="text-align: left">convertor</th>
      <th style="text-align: left">usage</th>
      <th style="text-align: left">input</th>
      <th style="text-align: left">output</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td style="text-align: left"><code>s</code></td>
      <td style="text-align: left"><code>str()</code></td>
      <td style="text-align: left"><code>${file1!s}</code></td>
      <td style="text-align: left"><code>file 1.txt</code></td>
    </tr>
    <tr>
      <td style="text-align: left"><code>r</code></td>
      <td style="text-align: left"><code>repr()</code></td>
      <td style="text-align: left"><code>${file1!r}</code></td>
      <td style="text-align: left"><code>'file 1.txt'</code></td>
    </tr>
    <tr>
      <td style="text-align: left"><code>q</code></td>
      <td style="text-align: left"><code>quoted()</code></td>
      <td style="text-align: left"><code>${file1!q}</code></td>
      <td style="text-align: left"><code>file\ 1.txt</code></td>
    </tr>
    <tr>
      <td style="text-align: left"><code>e</code></td>
      <td style="text-align: left"><code>expanduser()</code></td>
      <td style="text-align: left"><code>${file2!e}</code></td>
      <td style="text-align: left"><code>/path/to/user/SoS/test.sos</code></td>
    </tr>
    <tr>
      <td style="text-align: left"><code>a</code></td>
      <td style="text-align: left"><code>abspath(expanduser())</code></td>
      <td style="text-align: left"><code>${file2!a}</code></td>
      <td style="text-align: left"><code>/path/to/user/SoS/test.sos</code></td>
    </tr>
    <tr>
      <td style="text-align: left"><code>b</code></td>
      <td style="text-align: left"><code>basename())</code></td>
      <td style="text-align: left"><code>${file2!b}</code></td>
      <td style="text-align: left"><code>test.sos</code></td>
    </tr>
    <tr>
      <td style="text-align: left"><code>d</code></td>
      <td style="text-align: left"><code>dirname())</code></td>
      <td style="text-align: left"><code>${file2!d}</code></td>
      <td style="text-align: left"><code>/path/to/user/SoS/</code></td>
    </tr>
    <tr>
      <td style="text-align: left"><code>,</code></td>
      <td style="text-align: left"><code>', '.join()</code></td>
      <td style="text-align: left"><code>${files!,}</code></td>
      <td style="text-align: left"><code>a.txt, b.txt</code></td>
    </tr>
  </tbody>
</table>

<p>For example, the following command</p>

<figure class="highlight"><pre><code class="language-python" data-lang="python"><span class="n">run</span><span class="p">(</span><span class="s">&#39;cat ${_input!q}`)</span></code></pre></figure>

<p>would produce the correct command and execute</p>

<figure class="highlight"><pre><code class="language-python" data-lang="python"><span class="n">cat</span> <span class="s">&#39;Bon Jovi.txt&#39;</span></code></pre></figure>

<p>instead of</p>

<figure class="highlight"><pre><code class="language-python" data-lang="python"><span class="n">cat</span> <span class="n">Bon</span> <span class="n">Jovi</span><span class="o">.</span><span class="n">txt</span></code></pre></figure>

<p>for <code>_input=['Bon Jovi.txt']</code>. The conversion flags can also be combined, as shown in</p>

<figure class="highlight"><pre><code class="language-python" data-lang="python"><span class="n">files</span>   <span class="o">=</span> <span class="p">[</span><span class="s">&#39;AB.txt&#39;</span><span class="p">,</span> <span class="s">&#39;C D.txt&#39;</span><span class="p">]</span>
<span class="n">fname</span>   <span class="o">=</span> <span class="s">&#39;${files!q,}&#39;</span>            <span class="c"># AB.txt,C\D.txt or AB.txt,&#39;C D.txt&#39;</span></code></pre></figure>

<p>If the default sigil conflicts with the script used, an alternative sigil can be used. Please see section <a href="#option-sigil">option <code>sigil</code></a> for details.</p>

<h2 id="section-header-for-workflow-steps">Section header for workflow steps</h2>

<p>SoS defines workflows that consists of multiple steps. A step is marked by a section head in the format of</p>

<figure class="highlight"><pre><code class="language-python" data-lang="python"><span class="p">[</span><span class="n">names</span><span class="p">:</span> <span class="n">options</span><span class="p">]</span></code></pre></figure>

<p>The header should start with a <code>[</code> from the beginning of a line and end with a <code>]</code>. It can contain one or more names with optional section options.</p>

<h2 id="sos-directives">SoS directives</h2>

<p>SoS introduces a number of directives to specify elements of workflows, including</p>

<ul>
  <li><code>parameter:</code> to define command line options</li>
  <li><code>input:</code>, <code>output:</code>, <code>depends:</code> to define input, output, and dependent files of workflow steps, and</li>
  <li><code>task:</code> to define (long) tasks that will be executed as separate tasks</li>
</ul>

<h2 id="script-style-function">Script style function</h2>

<p>SoS allows you to write SoS <code>action</code> (basically a Python function) in a special script format. For example,</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="n">R</span><span class="p">(</span><span class="s">&#39;&#39;&#39;</span>
<span class="s">pdf(&#39;${input}&#39;)</span>
<span class="s">plot(0, 0)</span>
<span class="s">dev.off()</span>
<span class="s">&#39;&#39;&#39;</span><span class="p">,</span> <span class="kp">workdir</span><span class="o">=</span><span class="s">&#39;result&#39;</span><span class="p">)</span></code></pre></figure>

<p>can be written as</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="kn">R:</span>     <span class="kp">workdir</span><span class="o">=</span><span class="s">&#39;result&#39;</span>
<span class="n">pdf</span><span class="p">(</span><span class="s">&#39;${_input}&#39;</span><span class="p">)</span>
<span class="n">plot</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
<span class="n">dev</span><span class="o">.</span><span class="n">off</span><span class="p">()</span></code></pre></figure>

<p><strong>The script is a string without quotation marks</strong> and the normal string interpolation will take place. You can also indent the script (add leading white spaces to all lines) and write the action as</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="kn">R:</span>  <span class="kp">workdir</span><span class="o">=</span><span class="s">&#39;result&#39;</span>
   <span class="n">pdf</span><span class="p">(</span><span class="s">&#39;${_input}&#39;</span><span class="p">)</span>
   <span class="n">plot</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
   <span class="n">dev</span><span class="o">.</span><span class="n">off</span><span class="p">()</span></code></pre></figure>

<p>The latter is much preferred because it avoids trouble if your script contains strings such as <code>[1]</code> and <code>option:</code> (and be treated as SoS directives), and more importantly, allows starting a new statement from a non-indented line. For example, <code>check_command('dot')</code> would be considered part of a R script in</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="kn">R:</span>  <span class="kp">workdir</span><span class="o">=</span><span class="s">&#39;result&#39;</span>
<span class="n">pdf</span><span class="p">(</span><span class="s">&#39;${_input}&#39;</span><span class="p">)</span>
<span class="n">plot</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
<span class="n">dev</span><span class="o">.</span><span class="n">off</span><span class="p">()</span>

<span class="n">check_command</span><span class="p">(</span><span class="s">&#39;dot&#39;</span><span class="p">)</span></code></pre></figure>

<p>but a separate action in</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="kn">R:</span>  <span class="kp">workdir</span><span class="o">=</span><span class="s">&#39;result&#39;</span>
   <span class="n">pdf</span><span class="p">(</span><span class="s">&#39;${_input}&#39;</span><span class="p">)</span>
   <span class="n">plot</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
   <span class="n">dev</span><span class="o">.</span><span class="n">off</span><span class="p">()</span>

<span class="n">check_command</span><span class="p">(</span><span class="s">&#39;dot&#39;</span><span class="p">)</span></code></pre></figure>

<p>Although the script format is more concise and easier to read, it is limited to actions that accept a string as its first parameter and cannot return value or be used within <code>try/except</code> of <code>if/else</code> statements.</p>

<p>One final difference between SoS and regular Python 3 syntax is that SoS is more lenient on the use of mixed tab and spaces for indentation. Although it is highly recommended that you use all spaces for indentation, SoS will give an warning and treat tabs as 4 spaces during execution.</p>

                            </div>
                            <button type="button" class="btn btn-default" data-dismiss="modal"><i class="fa fa-times"></i> Close</button>
                        </div>
                    </div> 
                    <div class="footer-wrapper">
                      <footer role="contentinfo">
                        <span>&copy; 2016 Bo Peng.</span>
                      </footer>
                    </div>
                </div>
            </div>
        </div>
    </div> 


   <div class="portfolio-modal modal fade" id="File-Structure" tabindex="-1" role="dialog" aria-hidden="true">
        <div class="modal-content">
            <div class="close-modal" data-dismiss="modal">
                <div class="lr">
                    <div class="rl">
                    </div>
                </div>
            </div>
            <div class="container">
                <div class="row">
                 <div class="col-lg-8 col-lg-offset-2">
                        <h1>3.File-Structure</h1>
                        <div class="modal-body">
                         
                            <hr class="star-primary">
                            <p>A SoS script usually starts with lines</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="c">#!/usr/bin/env sos-runner</span>
<span class="c">#fileformat=SOS1.0</span></code></pre></figure>

<p>The first line allows the script to be executed by command <code>sos-runner</code> if it is executed as an executable script. The second line tells SoS the format of the script. The <code>#fileformat</code> line does not have to be the first or second line but should be in the first comment block. SOS format 1.0 is assumed if no format line is present.</p>

<h2 id="global-definitions">Global definitions</h2>

<p>Python functions, classes, variables can be defined or imported (using Python <code>import</code> statement) before any SoS step is defined. These definitions usually contains variables such as version and date of the script, paths to various resources, and utility functions that will be used by later steps. <strong>These definitions are visible to all steps of workflows and are assumed to be readonly</strong>.</p>

<p>In addition to user-defined global variables, SoS defines the following variables before any variables are defined</p>

<ul>
  <li><strong><code>SOS_VERSION</code></strong>: version of SoS command.</li>
  <li><strong><code>CONFIG</code></strong>: A dictionary of configurations specified by the global sos configuration file (<code>~/.sos/config.yaml</code>), local configuration file (<code>~/.sos/config.yaml</code>) and command line option <code>-c config_file</code>.</li>
  <li><strong><code>run_mode</code></strong>: A string indicating the run mode in which the script is executed, which can be <code>inspect</code>, <code>prepare</code> or <code>run</code>. Because the script and therefore all Python statements in the script will be executed in all three modes, this variable allows the execution of some statement only in selected modes.</li>
</ul>

<h2 id="include-preprocessor"><code>%include</code> preprocessor</h2>

<p>SoS allows you to include variables and steps of another script into the current script using preprocessor <code>%include</code>. The <code>%include</code> statement should be used before any other SoS statements and follows the same syntax as the python <code>import</code> keyword. For example, the following statements are allowed</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="o">%</span><span class="n">include</span> <span class="n">alignment</span>
<span class="o">%</span><span class="n">include</span> <span class="n">alignment</span> <span class="k">as</span> <span class="n">ag</span>
<span class="o">%</span><span class="n">include</span> <span class="n">alignment</span><span class="p">,</span> <span class="n">calling</span>
<span class="o">%</span><span class="n">include</span> <span class="n">alignment</span> <span class="k">as</span> <span class="n">ag</span><span class="p">,</span> <span class="n">calling</span> <span class="k">as</span> <span class="n">ca</span>

<span class="o">%</span><span class="kn">from</span> <span class="nn">alignment</span> <span class="nn">include</span> <span class="o">*</span>
<span class="o">%</span><span class="kn">from</span> <span class="nn">alignment</span> <span class="nn">include</span> <span class="nn">var1</span><span class="p">,</span> <span class="n">workflow1</span>
<span class="o">%</span><span class="kn">from</span> <span class="nn">alignment</span> <span class="nn">include</span> <span class="nn">workflow1</span> <span class="nn">as</span> <span class="nn">wf</span></code></pre></figure>

<p>Similar to python <code>import</code> statement, variables and workflows included using the <code>%include ... as ...</code> syntax can be accessed using <code>module.name</code> (e.g. <code>alignment.var1</code>, <code>ag.var1</code> (for <code>include alignment as ag</code>)), whereas variables and workflows included using the <code>from ... include ...</code> syntax are available directly (e.g. <code>var1</code>, <code>workflow1</code>, and <code>wf</code> (for <code>workflow1 as wf</code>).</p>

<p>For example</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="c">#!/usr/bin/env sos-runner</span>
<span class="c">#fileformat=SOS1.0</span>

<span class="o">%</span><span class="n">include</span> <span class="n">alignment</span>
<span class="o">%</span><span class="n">include</span> <span class="n">call</span>
<span class="p">[</span><span class="mi">10</span><span class="p">]</span>
<span class="n">sos_run</span><span class="p">(</span><span class="s">&#39;alignment.default + call.default&#39;</span><span class="p">)</span></code></pre></figure>

<p>would execute two workflows defined in <code>alignment.sos</code> and <code>call.sos</code>. The files should be in SoS search path to be included.</p>

<h2 id="if--elif-else--endif"><code>%if .. %elif ..%else .. %endif</code></h2>

<p>SoS allows you to conditionally include part of the script using macos <code>%if</code>, <code>%elif</code>, <code>%else</code>, and <code>%endif</code>. The condition in <code>%if cond</code> and <code>elif cond</code> should be valid python expression. The condition should not expand multiple lines and should not have trailing <code>:</code>. The conditions are evaluated at parsing time so no SoS variables are allowed. For example, you can write</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="o">%</span><span class="k">if</span> <span class="n">sys</span><span class="o">.</span><span class="n">platform</span> <span class="o">==</span> <span class="s">&#39;darwin&#39;</span>

<span class="p">[</span><span class="n">step</span><span class="p">]</span>
<span class="n">Mac</span> <span class="n">OSX</span> <span class="n">step</span>

<span class="o">%</span><span class="k">else</span>

<span class="p">[</span><span class="n">step</span><span class="p">]</span>
<span class="n">Linux</span> <span class="n">step</span>

<span class="o">%</span><span class="n">endif</span></code></pre></figure>


                            </div>
                            <button type="button" class="btn btn-default" data-dismiss="modal"><i class="fa fa-times"></i> Close</button>
                        </div>
                    </div> 
                    <div class="footer-wrapper">
                      <footer role="contentinfo">
                        <span>&copy; 2016 Bo Peng.</span>
                      </footer>
                    </div>
                </div>
            </div>
        </div>
    </div> 


   <div class="portfolio-modal modal fade" id="Command-line-options-and-configuration-file" tabindex="-1" role="dialog" aria-hidden="true">
        <div class="modal-content">
            <div class="close-modal" data-dismiss="modal">
                <div class="lr">
                    <div class="rl">
                    </div>
                </div>
            </div>
            <div class="container">
                <div class="row">
                 <div class="col-lg-8 col-lg-offset-2">
                        <h1>4.Command-line-options-and-configuration-file</h1>
                        <div class="modal-body">
                         
                            <hr class="star-primary">
                            <h2 id="command-line-options">Command line options</h2>

<h3 id="optional-arguments">Optional arguments</h3>

<p>Any variable assignment prefixed with <code>parameter:</code> can accept values from command  line. The format of such lines are</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="c"># comment</span>
<span class="kn">parameter:</span> <span class="n">var_name</span> <span class="o">=</span> <span class="n">default_value</span></code></pre></figure>

<p>The default value can be number, string, list of string, or expressions that return values of these types. Other types can be used as long as they can be converted to these types from user-provided values. For example</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="c"># path to tool gatk</span>
<span class="kn">parameter:</span> <span class="n">gatk_path</span> <span class="o">=</span> <span class="s">&#39;~/bin/GATK&#39;</span></code></pre></figure>

<p>defines a variable <code>gatk_path</code> with default value <code>'~/bin/GATK'</code>.</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="c"># A list of sample names</span>
<span class="kn">parameter:</span> <span class="n">sample_names</span><span class="o">=</span><span class="p">[]</span></code></pre></figure>

<p>defines a variable <code>sample_name</code> with default value <code>[]</code>, and</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="c"># path to gatk</span>
<span class="kn">parameter:</span> <span class="n">gatk_path</span> <span class="o">=</span> <span class="n">CONFIG</span><span class="p">[</span><span class="s">&#39;gatk_path&#39;</span><span class="p">]</span></code></pre></figure>

<p>uses <code>gatk_path</code> from a YAML-based configuration file (specified from command line using option <code>-c</code>) as default value. You can set default value for <code>CONFIG</code> inside SoS script via:</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="c"># path to gatk</span>
<span class="kn">parameter:</span> <span class="n">gatk_path</span> <span class="o">=</span> <span class="n">CONFIG</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s">&#39;gatk_path&#39;</span><span class="p">,</span> <span class="s">&#39;/default/path/to/gatk&#39;</span><span class="p">)</span></code></pre></figure>

<p>so that if configuration files are not supplied the default value <code>/default/path/to/gatk</code> will be used.</p>

<p>The default values not only determines the values of variable when they are not specified from command line or configuration files, but also determines the type of input these parameters accept. For example, with the above definitions for command arguments <code>--gatk_path</code> and <code>--sample_names</code>, you can pass values to these variables from command line,</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="n">sos</span> <span class="n">run</span> <span class="n">myscript</span> <span class="o">--</span><span class="n">gatk_path</span> <span class="o">/</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">gatk</span> <span class="o">--</span><span class="n">sample_names</span> <span class="n">A1</span> <span class="n">A2</span> <span class="n">A3</span></code></pre></figure>

<p>A list will be passed to <code>sample_names</code> even if only a single value is provided (e.g. <code>sample_names=['A1']</code> for <code>--sample_name A1</code>).
Attempts to pass more than one values (a list) to <code>gatk_path</code> (e.g. <code>--gatk_path /path1 /path2</code>) will trigger an error.</p>

<p>Note that boolean values can be specified from command line as <code>--param</code> for <code>True</code> and <code>--no-param</code> for <code>False</code>. It is also allowed to use <code>--gatk-path</code> in addition to <code>--gatk_path</code> for parameter <code>--gatk_path</code>.</p>

<h3 id="required-arguments">Required arguments</h3>

<p>In cases where there is no suitable default values and/or command line arguments are mandatary, you can list the type of arguments (e.g. <code>int</code>, <code>bool</code>, <code>str</code>, <code>list</code> of strings) in place of default values. For example, if an integer parameter <code>cutoff</code> is required, you can define it as</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="c"># cutoff value</span>
<span class="kn">parameter:</span> <span class="n">cutoff</span> <span class="o">=</span> <span class="nb">int</span></code></pre></figure>

<p>This will force the users to provide an integer to this parameter. You can do the same for lists but SoS assumes that you need a <strong>list of strings</strong>. For example, the following definition</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="c"># input bam files</span>
<span class="kn">parameter:</span> <span class="n">bam_files</span> <span class="o">=</span> <span class="nb">list</span></code></pre></figure>

<p>request a list of strings from command line. SoS will return a list even if only one value is provided.</p>

<h3 id="redefinition-of-parameter">Redefinition of <code>parameter</code></h3>

<p>A very important property of the <code>parameter</code> directive is that <strong>parameter definition is only effective for undefined variable</strong>. That is to say, the parameter statement will be ignored if a parameter has been specified in anyway, as shown in the following example:</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="n">a</span> <span class="o">=</span> <span class="mi">10</span>
<span class="kn">parameter:</span> <span class="n">a</span> <span class="o">=</span> <span class="nb">int</span></code></pre></figure>

<p>An useful implication of this rule is that you can define a workflow and allow some of its parameters to be specified from command line and from other workflows. For example, the following workflow</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="p">[</span><span class="n">default_1</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">cutoff</span><span class="o">=</span><span class="mi">10</span>
<span class="kn">sh:</span>
  <span class="n">do</span> <span class="n">something</span> <span class="k">with</span> <span class="err">$</span><span class="p">{</span><span class="n">cutoff</span><span class="p">}</span>

<span class="p">[</span><span class="n">default_2</span><span class="p">]</span>
<span class="kn">sh:</span>
  <span class="n">do</span> <span class="n">something</span> <span class="k">else</span>

<span class="p">[</span><span class="n">batch</span><span class="p">]</span>
<span class="k">for</span> <span class="n">cutoff</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">10</span><span class="p">):</span>
    <span class="n">sos_run</span><span class="p">(</span><span class="s">&#39;default&#39;</span><span class="p">)</span></code></pre></figure>

<p>would be called in a number of ways</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="n">sos</span> <span class="n">run</span> <span class="n">myscript</span> 
<span class="n">sos</span> <span class="n">run</span> <span class="n">myscript</span> <span class="o">--</span><span class="n">cutoff</span> <span class="mi">5</span>
<span class="n">sos</span> <span class="n">run</span> <span class="n">myscript</span> <span class="n">batch</span></code></pre></figure>

<p>The first command uses default value <code>10</code> for parameter <code>cutoff</code>, the second command assigns value <code>5</code> from command line, and the third command uses a series of values <code>0</code>, <code>1</code>, …. Because <code>cutoff</code> is already defined for workflow <code>default</code>, the parameter statement will be ignored.</p>

<h2 id="configuration-files">Configuration files</h2>

<p>The configuration files should be in the format of <a href="http://yaml.org/"><code>YAML</code></a> or its subset format <a href="http://json-schema.org/implementations.html"><code>JSON</code></a>. Inside SoS script you can access these variables via, for example, either <code>CONFIG['gatk_path']</code> or <code>CONFIG.gatk_path</code>.</p>

<p>The <code>-c</code> option allows the specification of a configuration file in YAML format. YAML is a superset of JSON so any configuration file in JSON format should also be acceptable. Variables defined in the configuration file are available in SoS script as a dictionary <code>CONFIG</code>. For example</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="n">gatk_path</span> <span class="o">=</span> <span class="n">CONFIG</span><span class="p">[</span><span class="s">&#39;gatk_path&#39;</span><span class="p">]</span></code></pre></figure>

<p>or equivalently</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="n">gatk_path</span> <span class="o">=</span> <span class="n">CONFIG</span><span class="o">.</span><span class="n">gatk_path</span></code></pre></figure>

<p>requires a configuration file with <code>gatk_path</code> defined. If you do not want to require a configuration file, you can define <code>gatk_path</code> as</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="n">gatk_path</span> <span class="o">=</span> <span class="n">CONFIG</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s">&#39;gatk_path&#39;</span><span class="p">,</span> <span class="s">&#39;/path/to/default/gatk&#39;</span><span class="p">)</span></code></pre></figure>

<p>In this way, a default path would be used if no configuration file is specified (so <code>CONFIG</code> is an empty dictionary) or if <code>gatk_path</code> is not defined in the specified configuration file.</p>

<p>If you would further want to allow modification of this value from command line, you can place this definition in the <code>[parameters]</code> section</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="p">[</span><span class="n">parameters</span><span class="p">]</span>
<span class="c"># path to gatk executable</span>
<span class="n">gatk_path</span> <span class="o">=</span> <span class="n">CONFIG</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s">&#39;gatk_path&#39;</span><span class="p">,</span> <span class="s">&#39;/path/to/default/gatk&#39;</span><span class="p">)</span></code></pre></figure>

<p>In this way, users have the freedom to use the default value, define a value in a configuration file, and provide another value from command line.</p>

                            </div>
                            <button type="button" class="btn btn-default" data-dismiss="modal"><i class="fa fa-times"></i> Close</button>
                        </div>
                    </div> 
                    <div class="footer-wrapper">
                      <footer role="contentinfo">
                        <span>&copy; 2016 Bo Peng.</span>
                      </footer>
                    </div>
                </div>
            </div>
        </div>
    </div> 


   <div class="portfolio-modal modal fade" id="Workflow-Definition" tabindex="-1" role="dialog" aria-hidden="true">
        <div class="modal-content">
            <div class="close-modal" data-dismiss="modal">
                <div class="lr">
                    <div class="rl">
                    </div>
                </div>
            </div>
            <div class="container">
                <div class="row">
                 <div class="col-lg-8 col-lg-offset-2">
                        <h1>5.Workflow-Definition</h1>
                        <div class="modal-body">
                         
                            <hr class="star-primary">
                            <p>A SoS script can specify one or more workflows. Each workflow consists of one or more numbered steps. The numbers (should be non-negative) specify the <strong>logical order</strong> by which the steps are executed, but a later step might be executed before the completion of previous steps if it does not depend on the output of these steps.</p>

<h2 id="single-workflow">Single workflow</h2>

<p>A single workflow can be specified without a name in a SoS script. For example, the following sections specify a workflow with four steps <code>5</code>, <code>10</code>, <code>20</code>, and <code>100</code>.
As you can see, the workflow steps can be specified in any order and do not have to be consecutive (which is actually preferred because it allows easy insertion of extra steps).</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="p">[</span><span class="mi">5</span><span class="p">]</span>
<span class="p">[</span><span class="mi">20</span><span class="p">]</span>
<span class="p">[</span><span class="mi">10</span><span class="p">]</span>
<span class="p">[</span><span class="mi">100</span><span class="p">]</span></code></pre></figure>

<p>Workflows specified in this way is the <code>default</code> workflow and are actually called <code>default</code> in SoS output. If you want to give it a meaningful name, you can specify the steps as</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="p">[</span><span class="n">mapping_5</span><span class="p">]</span>
<span class="p">[</span><span class="n">mapping_20</span><span class="p">]</span>
<span class="p">[</span><span class="n">mapping_10</span><span class="p">]</span>
<span class="p">[</span><span class="n">mapping_100</span><span class="p">]</span></code></pre></figure>

<p>Because this SoS script defines only one workflow (<code>mapping</code>), you do not have to specify the name of workflow from SoS command</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="n">sos</span> <span class="n">run</span> <span class="n">myscript</span><span class="o">.</span><span class="n">sos</span> <span class="o">--</span><span class="nb">input</span> <span class="n">input1</span><span class="o">.</span><span class="n">fasta</span></code></pre></figure>

<p>Unnumbered workflow steps are assumed to be the first steps (with index <code>0</code>) of a workflow unless they have other meanings (e.g. <a href="auxiliary-workflow-steps-and-makefile-style-dependency-rules">auxiliary step</a>).</p>

<h2 id="multiple-workflows">Multiple workflows</h2>

<p>A SoS script can define multiple workflows. For example, the following sections of SoS script defines two workflows named <code>mouse</code> and <code>human</code>.</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="p">[</span><span class="n">mouse_10</span><span class="p">]</span>
<span class="p">[</span><span class="n">mouse_20</span><span class="p">]</span>
<span class="p">[</span><span class="n">mouse_30</span><span class="p">]</span>
<span class="p">[</span><span class="n">human_10</span><span class="p">]</span>
<span class="p">[</span><span class="n">human_20</span><span class="p">]</span>
<span class="p">[</span><span class="n">human_30</span><span class="p">]</span></code></pre></figure>

<p>You will have to specify which workflow to execute from the command line, e.g.</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="n">sos</span> <span class="n">run</span> <span class="n">myscript</span> <span class="n">mouse</span> <span class="o">--</span><span class="nb">input</span> <span class="n">input1</span><span class="o">.</span><span class="n">fasta</span></code></pre></figure>

<p>If you would like to define a <code>default</code> and a named workflow, you can define them as</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="p">[</span><span class="mi">10</span><span class="p">]</span>
<span class="p">[</span><span class="mi">20</span><span class="p">]</span>
<span class="p">[</span><span class="mi">30</span><span class="p">]</span>
<span class="p">[</span><span class="n">test_10</span><span class="p">]</span>
<span class="p">[</span><span class="n">test_20</span><span class="p">]</span>
<span class="p">[</span><span class="n">test_30</span><span class="p">]</span></code></pre></figure>

<p>The <code>default</code> workflow will be executed by default using command</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="n">sos</span> <span class="n">run</span> <span class="n">myscript</span><span class="o">.</span><span class="n">sos</span></code></pre></figure>

<p>The <code>test</code> workflow will be executed if its name is specified from the command line</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="n">sos</span> <span class="n">run</span> <span class="n">myscript</span><span class="o">.</span><span class="n">sos</span> <span class="n">test</span></code></pre></figure>

<h2 id="shared-workflow-steps">Shared workflow steps</h2>

<p>The most common motivation of defining multiple workflows in a single SoS script is that they share certain processing steps. If this is the case, you can define sections such as</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="p">[</span><span class="n">mouse_10</span><span class="p">,</span><span class="n">human_10</span><span class="p">]</span>
<span class="p">[</span><span class="n">mouse_20</span><span class="p">]</span>
<span class="p">[</span><span class="n">human_20</span><span class="p">]</span>
<span class="p">[</span><span class="n">mouse_30</span><span class="p">,</span><span class="n">human_30</span><span class="p">]</span></code></pre></figure>

<p>or</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="p">[</span><span class="o">*</span><span class="n">_10</span><span class="p">]</span>
<span class="p">[</span><span class="n">mouse_20</span><span class="p">]</span>
<span class="p">[</span><span class="n">human_20</span><span class="p">]</span>
<span class="p">[</span><span class="o">*</span><span class="n">_30</span><span class="p">]</span></code></pre></figure>

<p>or</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="p">[</span><span class="o">*</span><span class="n">_10</span><span class="p">]</span>
<span class="p">[</span><span class="n">mouse_20</span><span class="p">,</span><span class="n">human_20</span><span class="p">]</span>
<span class="p">[</span><span class="n">fly_20</span><span class="p">]</span>
<span class="p">[</span><span class="o">*</span><span class="n">_30</span><span class="p">,</span><span class="n">fly_50</span><span class="p">]</span>
<span class="p">[</span><span class="n">fly_40</span><span class="p">]</span></code></pre></figure>

<p>In the last case, step defined by <code>[*_30,fly_40]</code> will be expanded to <code>mouse_30</code>, <code>human_30</code>, <code>fly_30</code>, and <code>fly_50</code> and will be executed twice for the <code>fly</code> workflow. Note that workflow steps can use variable <code>step_name</code> to act (slightly) differently for different workflows. For example,</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="p">[</span><span class="n">mouse_20</span><span class="p">,</span><span class="n">human_20</span><span class="p">]</span>
<span class="n">reference</span> <span class="o">=</span> <span class="n">mouse_reference</span> <span class="k">if</span> <span class="n">step_name</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s">&#39;mouse&#39;</span><span class="p">)</span> <span class="k">else</span> <span class="n">human_reference</span>

<span class="kn">input:</span> <span class="n">fasta_files</span>
<span class="kn">depends:</span> <span class="n">reference</span></code></pre></figure>

<p>Here the variable <code>step_name</code> is <code>mouse_20</code> or <code>human_20</code> depending on the workflow being executed, and expression <code>mouse_reference if step_name.startswith('mouse') else human_reference</code> returns <code>mouse_reference</code> if the workflow <code>mouse</code> is executed, and <code>human_reference</code> otherwise.</p>

<h2 id="subworkflow">Subworkflow</h2>

<p>Although workflows are defined separately with all their steps, they do not have to be executed in their entirety. A <code>subworkflow</code> refers to a workflow that is defined from one or more steps of an existing workflows. It is specified using syntax <code>workflow[_steps]</code> where step can be <code>n</code> (step <code>n</code>), <code>:n</code> (up to <code>n</code>), <code>n:m</code> (step <code>n</code> to <code>m</code>) and <code>m:</code> (from <code>m</code>). For example</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="n">A</span>              <span class="c"># complete workflow A</span>
<span class="kn">A:</span><span class="mi">5</span><span class="o">-</span><span class="mi">10</span>         <span class="c"># step 5 to 10 of A</span>
<span class="kn">A:</span><span class="mi">50</span><span class="o">-</span>          <span class="c"># step 50 up</span>
<span class="kn">A:</span><span class="o">-</span><span class="mi">10</span>          <span class="c"># up to step 10 of A</span>
<span class="kn">A:</span><span class="mi">10</span>           <span class="c"># step 10 of workflow A can be considered a subworkflow</span></code></pre></figure>

<h2 id="combined-workflow">Combined workflow</h2>

<p>You can also combine subworkflows to execute multiple workflows one after another. For example,</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="n">A</span> <span class="o">+</span> <span class="n">B</span>          <span class="c"># workflow A, followed by B</span>
<span class="n">A_0</span> <span class="o">+</span> <span class="n">B</span>        <span class="c"># step 0 of A, followed by B</span>
<span class="kn">A:</span><span class="o">-</span><span class="mi">50</span> <span class="o">+</span> <span class="n">B</span> <span class="o">+</span> <span class="n">C</span>  <span class="c"># up to step 50 of workflow A, followed by B, and C</span></code></pre></figure>

<p>This syntax can be used from the command line (option <code>workflow</code>, e.g. <code>sos-runner myscript.sos align+call</code>) or used to execute <a href="#nested-workflow">nested workflows</a> inside the SoS script.</p>

<p>It is worth noting that combined workflow might work differently from when they are executed individually (e.g. default input of <code>B</code> is changed from empty to output of <code>A_0</code>), and it is up to the user to resolve conflicts between them.</p>

<h2 id="nested-workflow">Nested workflow</h2>

<p>SoS also supports nested workflow in which a complete workflow is treated as part of a step process.
The workflow is execute by SoS action <code>sos_run</code>, e.g.</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="n">sos_run</span><span class="p">(</span><span class="s">&#39;A&#39;</span><span class="p">)</span>                       <span class="c"># execute workflow A</span>
<span class="n">sos_run</span><span class="p">(</span><span class="s">&#39;A + B&#39;</span><span class="p">)</span>                   <span class="c"># execute workflow B after A</span>
<span class="n">sos_run</span><span class="p">(</span><span class="s">&#39;D:-10 + C&#39;</span><span class="p">)</span>               <span class="c"># execute up to step 10 of D and workflow C</span>
<span class="n">sos_run</span><span class="p">(</span><span class="s">&#39;${aligner} + ${caller}&#39;</span><span class="p">)</span>  <span class="c"># execute user-specified aligner and caller workflows</span></code></pre></figure>

<p>In its simplest form, nested workflow allows you to define another workflow from existing ones. For example,</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="p">[</span><span class="n">default</span><span class="p">]</span>
<span class="n">sos_run</span><span class="p">(</span><span class="s">&#39;align+call&#39;</span><span class="p">)</span></code></pre></figure>

<p>defines a nested workflow that combines workflows <code>align</code> and <code>call</code> so that the workflow will by default execute two workflows, but can also execute one of them as separate workflows <code>align</code> and <code>call</code>.</p>

<p>Nested workflow also allows you to define multiple mini-workflows and connect them freely. For example</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="p">[</span><span class="n">a_1</span><span class="p">]</span>
<span class="p">[</span><span class="n">a_2</span><span class="p">]</span>
<span class="p">[</span><span class="n">b</span><span class="p">]</span>
<span class="p">[</span><span class="n">c</span><span class="p">]</span>
<span class="p">[</span><span class="n">d_1</span><span class="p">]</span>
<span class="n">sos_run</span><span class="p">(</span><span class="s">&#39;a+b&#39;</span><span class="p">)</span>
<span class="p">[</span><span class="n">d_2</span><span class="p">]</span>
<span class="n">sos_run</span><span class="p">(</span><span class="s">&#39;a+c&#39;</span><span class="p">)</span></code></pre></figure>

<p>defines workflows <code>d</code> that will execute steps <code>d_1</code>, <code>a_1</code>, <code>a_2</code>, <code>b_0</code>, <code>d_2</code>,  <code>a_1</code>, <code>a_2</code>, and <code>c_0</code>.</p>

<p>Nested workflows, like other SoS actions, can be executed repeatedly, for example,</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="p">[</span><span class="n">b_1</span><span class="p">]</span>
<span class="p">[</span><span class="n">b_2</span><span class="p">]</span>
<span class="p">[</span><span class="n">b_3</span><span class="p">]</span>

<span class="p">[</span><span class="n">a</span><span class="p">]</span>
<span class="n">parameters</span> <span class="o">=</span> <span class="nb">range</span><span class="p">(</span><span class="mi">20</span><span class="p">)</span>
<span class="kn">input:</span> <span class="s">&#39;some.txt&#39;</span><span class="p">,</span> <span class="kp">for_each</span><span class="o">=</span><span class="s">&#39;parameters&#39;</span>
<span class="kn">output:</span> <span class="s">&#39;${input}_${_parameters}.res&#39;</span>
<span class="n">sos_run</span><span class="p">(</span><span class="s">&#39;b&#39;</span><span class="p">)</span></code></pre></figure>

<p>would execute the complete workflow <code>b</code> 20 times each with a different parameter. Similarly you can let the nested workflow process groups of input files.</p>

<p>Nested workflows can also be used to compose workflows from user-provided options through command line arguments, configuration files, and even results from previous steps. For example, the following example</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="c"># aligner steps to use to align the reads </span>
<span class="kn">parameter:</span> <span class="n">aligner</span> <span class="o">=</span> <span class="n">CONFIG</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s">&#39;aligner&#39;</span><span class="p">,</span> <span class="s">&#39;bwa&#39;</span><span class="p">)</span>

<span class="p">[</span><span class="n">bwa_1</span><span class="p">]</span>
<span class="p">[</span><span class="n">bwa_2</span><span class="p">]</span>
<span class="p">[</span><span class="n">novaalign_1</span><span class="p">]</span>
<span class="p">[</span><span class="n">novaalign_2</span><span class="p">]</span>

<span class="p">[</span><span class="n">align</span><span class="p">]</span>
<span class="n">sos_run</span><span class="p">(</span><span class="n">aligner</span><span class="p">)</span></code></pre></figure>

<p>defines workflows <code>bwa</code> and <code>novaalign</code> to align raw reads. The <code>align</code> workflow is a master workflow that executes <code>bwa</code> or <code>novaalign</code> determined by option <code>aligner</code> defined in a configuration file (command line option <code>-c</code>) and command line option <code>--aligner</code>.</p>


                            </div>
                            <button type="button" class="btn btn-default" data-dismiss="modal"><i class="fa fa-times"></i> Close</button>
                        </div>
                    </div> 
                    <div class="footer-wrapper">
                      <footer role="contentinfo">
                        <span>&copy; 2016 Bo Peng.</span>
                      </footer>
                    </div>
                </div>
            </div>
        </div>
    </div> 


   <div class="portfolio-modal modal fade" id="SoS-Steps" tabindex="-1" role="dialog" aria-hidden="true">
        <div class="modal-content">
            <div class="close-modal" data-dismiss="modal">
                <div class="lr">
                    <div class="rl">
                    </div>
                </div>
            </div>
            <div class="container">
                <div class="row">
                 <div class="col-lg-8 col-lg-offset-2">
                        <h1>6.SoS-Steps</h1>
                        <div class="modal-body">
                         
                            <hr class="star-primary">
                            <p>Although no item is required for a SoS step, a complete SoS step can have the following form</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="p">[</span><span class="n">step_name</span><span class="p">:</span> <span class="n">option1</span><span class="p">,</span> <span class="n">option2</span><span class="p">,</span> <span class="o">...</span><span class="p">]</span>
<span class="c">#</span>
<span class="c"># description of the step</span>
<span class="c">#</span>

<span class="n">statements</span>
<span class="kn">input:</span>    <span class="nb">input</span> <span class="n">files</span><span class="p">,</span> <span class="n">opt1</span><span class="o">=</span><span class="n">value1</span><span class="p">,</span> <span class="n">opt2</span><span class="o">=</span><span class="n">value2</span>

<span class="n">statements</span> <span class="p">(</span><span class="n">executed</span> <span class="ow">in</span> <span class="n">SoS</span><span class="p">)</span>
<span class="kn">output:</span>	  <span class="n">output</span> <span class="n">files</span><span class="p">,</span> <span class="n">opt1</span><span class="o">=</span><span class="n">value1</span><span class="p">,</span> <span class="n">opt2</span><span class="o">=</span><span class="n">value2</span>

<span class="n">statements</span> <span class="p">(</span><span class="n">executed</span> <span class="ow">in</span> <span class="n">SoS</span><span class="p">)</span>
<span class="kn">depends:</span>  <span class="n">dependent</span> <span class="n">files</span><span class="p">,</span> <span class="n">opt1</span><span class="o">=</span><span class="n">value1</span><span class="p">,</span> <span class="n">opt2</span><span class="o">=</span><span class="n">value2</span>

<span class="n">statements</span> <span class="p">(</span><span class="n">executed</span> <span class="ow">in</span> <span class="n">SoS</span><span class="p">)</span>
<span class="kn">task:</span>  <span class="n">opt1</span><span class="o">=</span><span class="n">value1</span><span class="p">,</span> <span class="n">opt2</span><span class="o">=</span><span class="n">value2</span> <span class="p">(</span><span class="n">executed</span> <span class="n">outside</span> <span class="n">of</span> <span class="n">SoS</span><span class="p">)</span>
<span class="n">statements</span> <span class="ow">or</span> <span class="n">script</span> </code></pre></figure>

<p>Basically, a step can have <code>input</code>, <code>output</code>, <code>depends</code>, <code>task</code> keywords with their options, with arbitrary Python statements around them.</p>

<h2 id="targets">Targets</h2>

<p>Targets are objects that a SoS step can input, output, or dependent on. They are usually files that are presented by filenames, but can also be other targets.</p>

<h3 id="executable-target"><code>executable</code> target</h3>

<p><code>executable</code> targets are commands that should be accessible and executable by SoS. These targets are usually listed in the <code>depends</code> section of a SoS step. For example, SoS would stop if a command <code>fastqc</code> is not found.</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="p">[</span><span class="mi">10</span><span class="p">]</span>
<span class="kn">input:</span>     <span class="n">fastq_files</span>
<span class="kn">depends:</span>   <span class="n">executable</span><span class="p">(</span><span class="s">&#39;fastqc&#39;</span><span class="p">)</span>
<span class="kn">sh:</span>
    <span class="n">fastqc</span> <span class="err">$</span><span class="p">{</span><span class="n">fastq_files</span><span class="p">}</span></code></pre></figure>

<p><code>executable</code> target can also be output of a step but installing executables can be tricky because the commands should be installed to existing <code>$PATH</code> so that they can be immediately accessible by SoS. Because SoS automatically adds <code>~/.sos/bin</code> to <code>$PATH</code> (option <code>-b</code>), an environment-neutral way for on-the-fly installation is to install commands to this directory. For example</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="p">[</span><span class="mi">10</span><span class="p">:</span> <span class="kp">provides</span><span class="o">=</span><span class="n">executable</span><span class="p">(</span><span class="s">&#39;fastqc&#39;</span><span class="p">)]</span>
<span class="kn">sh:</span>
    <span class="c"># download fastqc</span>
    <span class="c"># install fastqc</span>
    <span class="c"># link fastqc to ~/.sos/bin</span></code></pre></figure>

<p>would be called if executable <code>fastqc</code> is not found.</p>

<h3 id="sosvariable-target"><code>sos_variable</code> target</h3>

<p><code>sos_variable(name)</code> targets represent SoS variables that are created by a SoS step and shared to other steps. These targets can be used to provide information to other steps. For example,</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="p">[</span><span class="n">get_total_reads</span><span class="p">:</span> <span class="kp">shared</span><span class="o">=</span><span class="s">&#39;total_reads&#39;</span><span class="p">]</span>
<span class="kn">input:</span>  <span class="s">&#39;alignment/summary.txt&#39;</span>
<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="nb">input</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="k">as</span> <span class="n">ifile</span><span class="p">:</span>
    <span class="o">...</span>
    <span class="n">total_reads</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="o">...</span><span class="p">)</span>


<span class="p">[</span><span class="mi">100</span><span class="p">]</span>
<span class="kn">depends:</span> <span class="n">sos_variable</span><span class="p">(</span><span class="s">&#39;total_reads&#39;</span><span class="p">)</span>
<span class="kn">output:</span>  <span class="s">&#39;report.txt&#39;</span>

<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">output</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="k">as</span> <span class="n">report</span><span class="p">:</span>
    <span class="n">report</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;total reads: ${total_reads}&#39;</span><span class="p">)</span></code></pre></figure>

<p>Step <code>100</code> needed some information extracted from output of another step. You can either parse the information in step <code>100</code> or use another step to provide the information. The latter is recommended because the information could be requested by multiple steps.</p>

<h3 id="envvariable-target"><code>env_variable</code> target</h3>

<p>SoS keeps tract of runtime environment and creates signatures of executed steps so that they do not have to be executed again. Some commands, especially shell scripts, could however behave differently with different environmental variables. To make sure a step would be re-executed with changing environments, you should list these variables as dependencies of the step. For example</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="p">[</span><span class="mi">10</span><span class="p">]</span>
<span class="kn">depends:</span>   <span class="n">env_variable</span><span class="p">(</span><span class="s">&#39;DEBUG&#39;</span><span class="p">)</span>
<span class="kn">sh:</span>
    <span class="n">echo</span> <span class="n">DEBUG</span> <span class="ow">is</span> <span class="nb">set</span> <span class="n">to</span> <span class="err">$</span><span class="n">DEBUG</span></code></pre></figure>

<h3 id="dynamic-target"><code>dynamic</code> target</h3>

<p>A <code>dynamic</code> target is a target that can only be determined when the step is actually executed. For example, the input of the following step would not be evaluated until step <code>20</code> is executed.</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="p">[</span><span class="mi">10</span><span class="p">]</span>
<span class="c">#</span>
<span class="c"># a step that generate some text with dynamically determined names</span>
<span class="c">#</span>

<span class="p">[</span><span class="mi">20</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">dynamic</span><span class="p">(</span><span class="s">&#39;*.txt&#39;</span><span class="p">)</span></code></pre></figure>

<p>In this particular case, if you specify</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="p">[</span><span class="mi">20</span><span class="p">]</span>
<span class="kn">input:</span> <span class="s">&#39;*.txt&#39;</span></code></pre></figure>

<p>SoS would find all <code>.txt</code> files at the beginning of the execution of the workflow and feed them to step <code>20</code>.</p>

<h2 id="step-name-and-alias">Step name and alias</h2>

<p>A section can define one or more steps, with step names separated by comma. For example</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="p">[</span><span class="n">human_10</span><span class="p">,</span> <span class="n">mouse_10</span><span class="p">]</span></code></pre></figure>

<p>defines a section with two steps for two different workflows. The step name can be followed by a numeric index, which indicates the relative position of the step in a workflow.</p>

<p>An alias, or description, is allowed for step name. Although the description does not affect the workflow and index of the step, it helps identify the step in the SoS output. For example</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="p">[</span><span class="n">human_10</span> <span class="p">(</span><span class="n">quality</span> <span class="n">check</span><span class="p">)]</span></code></pre></figure>

<p>define a step with description <code>quality check</code>.</p>

<h2 id="step-options">Step options</h2>

<p><strong>Step options</strong> are specified after step name that assists the specification of workflows. SoS provides the following options</p>

<h3 id="option-skip">Option <code>skip</code></h3>

<p>Option <code>skip</code> takes two formats, the first format has no value</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="p">[</span><span class="mi">10</span><span class="p">:</span> <span class="kp">skip</span><span class="p">]</span></code></pre></figure>

<p>and is equivalent to</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="p">[</span><span class="mi">10</span><span class="p">:</span> <span class="kp">skip</span><span class="o">=</span><span class="bp">True</span><span class="p">]</span></code></pre></figure>

<p>The whole step will be skipped as if it is not defined at all in the script. This option provides a quick method to disable a step.</p>

<p>The second format takes a value, which is usually an expression that will be evaluated when the step is executed. For example, in the following workflow,</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="c"># tool suite to align the reads</span>
<span class="kn">parameter:</span> <span class="n">quality_check</span> <span class="o">=</span> <span class="bp">True</span>

<span class="p">[</span><span class="mi">10</span><span class="p">:</span> <span class="kp">skip</span><span class="o">=</span> <span class="ow">not</span> <span class="n">quality_check</span><span class="p">]</span></code></pre></figure>

<p>Step 10 will be skipped if option <code>--quality_check no</code> is specified from command line.</p>

<h3 id="option-sigil">Option <code>sigil</code></h3>

<p>Because a SoS script can include scripts in different languages with different sigils (special characters used to enclose expression), <strong>SoS allows the use of different sigils</strong> in the script. Whereas the default sigil (<code>${ }</code>) has to be used for global variables, other sigils can be used for different sections of the script. For example</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="p">[</span><span class="n">step_10</span><span class="p">:</span> <span class="kp">sigil</span><span class="o">=</span><span class="s">&#39;%( )&#39;</span><span class="p">]</span>
<span class="n">title</span> <span class="o">=</span> <span class="s">&#39;Sample %(sample_names[0]) results&#39;</span>
<span class="kn">run:</span>
	<span class="k">for</span> <span class="nb">file</span> <span class="ow">in</span> <span class="o">*.</span><span class="n">fastq</span>
	<span class="n">do</span>
		<span class="n">echo</span> <span class="n">Processing</span> <span class="err">$</span><span class="p">{</span><span class="nb">file</span><span class="p">}</span> <span class="o">...</span>
		<span class="n">echo</span> <span class="o">%</span><span class="p">(</span><span class="n">title</span><span class="p">)</span>
	<span class="n">done</span></code></pre></figure>

<p>uses a different sigil style because the embedded shell script uses <code>${ }</code>. In this example <code>${file}</code> keeps its meaning in the shell script while <code>%(sample_names[0])</code> and <code>%(title)</code> are replaced by SoS to their values.</p>

<p>Option <code>sigil</code> accept only a constant string (not a variable or expression).</p>

<h3 id="option-shared">Option <code>shared</code></h3>

<p>SoS executes each step in a separate process and by default does not return any result to the master SoS process. Option <code>shared</code> is used to share variables between steps. This option accepts:</p>

<ul>
  <li>A string (variable name), or</li>
  <li>A map between variable names and expressions (strings) that will be evaluated upon the completion of the step.</li>
  <li>A sequence of strings (variables) or maps.</li>
</ul>

<p>After the completion of a step, local variables defined in the <code>shared</code> option will be sent to the global environment so that later steps can use the updated value. For example,</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="kn">parameter:</span> <span class="n">trim</span> <span class="o">=</span> <span class="bp">False</span>
<span class="kn">parameter:</span> <span class="n">input_files</span> <span class="o">=</span> <span class="nb">list</span>

<span class="p">[</span><span class="mi">10</span><span class="p">:</span> <span class="kp">shared</span><span class="o">=</span><span class="p">{</span><span class="s">&#39;input_files&#39;</span><span class="p">:</span><span class="s">&#39;output&#39;</span><span class="p">},</span> <span class="kp">skip</span><span class="o">=</span><span class="n">trim</span> <span class="ow">is</span> <span class="bp">False</span><span class="p">]</span>
<span class="kn">input:</span>  <span class="n">input_files</span><span class="p">,</span> <span class="kp">group_by</span><span class="o">=</span><span class="mi">1</span>
<span class="kn">output:</span> <span class="s">&#39;${_input[0]}.trimmed&#39;</span>

<span class="kn">sh:</span>
    <span class="n">trim</span> <span class="err">$</span><span class="p">{</span><span class="nb">input</span><span class="p">}</span> <span class="o">&gt;</span> <span class="err">$</span><span class="p">{</span><span class="n">output</span><span class="p">}</span>

<span class="p">[</span><span class="mi">20</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">input_files</span>
<span class="c"># do something</span></code></pre></figure>

<p>In this example, <code>trim</code> is an option to trim input files defined in another parameter <code>input_files</code>. If <code>trim</code> is false, step 10 is ignored and step 20 handles <code>input_files</code>. If <code>trim</code> is True, the files are trimmed and produces <code>${_input[0]}.trimmed</code>. Variable <code>input_files</code> are sent back to the global environment so that step 20 would be handling the trimmed version of input files.</p>

<p>A map syntax is recommended to share <code>output</code> of one step with others, because the variable assignment will be evaluated only after the step is complete:</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="p">[</span><span class="mi">1</span><span class="p">:</span> <span class="kp">shared</span> <span class="o">=</span> <span class="p">{</span><span class="s">&#39;test_output&#39;</span><span class="p">:</span> <span class="s">&#39;output&#39;</span><span class="p">}]</span>
<span class="o">...</span>
<span class="p">[</span><span class="mi">2</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">test_output</span></code></pre></figure>

<p>The map syntax is evaluated as expressions; therefore it is possible to finer control what specific output, or variations of output, to share with others. For example:</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="p">[</span><span class="mi">1</span><span class="p">:</span> <span class="kp">shared</span><span class="o">=</span><span class="p">{</span><span class="s">&#39;test_output_1&#39;</span><span class="p">:</span><span class="s">&#39;output[0]&#39;</span><span class="p">,</span> <span class="s">&#39;test_output_2&#39;</span><span class="p">:</span> <span class="s">&#39;output[1]&#39;</span><span class="p">}]</span>
<span class="o">...</span>
<span class="p">[</span><span class="mi">2</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">test_output_1</span></code></pre></figure>

<p>to shared the first file in <code>output</code> (filename <code>output[0]</code>) instead of the entire output file list.</p>

<p>NOTE: Because global variables are readonly, they cannot be changed by individual steps through the <code>shared</code> option. Shared variable therefore has to be either new variables or parameters.</p>

<h3 id="option-provides">Option <code>provides</code></h3>

<p>This option lists files or targets a step generates so that it can be called if the target is required but does not exist. Please see section <a href="Format-Specification#auxiliary-workflow-steps-and-makefile-style-dependency-rules">auxiliary step</a> for details. A step can be treated as both a forward step and an auxiliary step despite of the differences on how they are invoked.</p>

<h2 id="step-input">Step input</h2>

<h3 id="input-files">Input files</h3>

<p>The input of SoS step follows the following rules:</p>

<ul>
  <li><strong>the input of a SoS step is by default the output of the previous step</strong>, which is <code>None</code> for the first step.</li>
  <li><strong><code>input</code> option</strong>, which could be <strong>a list</strong> of filenames (string literal, variables, or expressions that return filenames).  Wildcard characters (<code>*</code>, which matches everything and <code>?</code>, which matches any single character) are acceptable. Nested lists are flattened.</li>
</ul>

<p>Examples of input specification are as follows:</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="kn">input:</span> <span class="p">[]</span>

<span class="kn">input:</span> <span class="s">&#39;file1.fasta&#39;</span><span class="p">,</span> <span class="s">&#39;file2.fasta&#39;</span>

<span class="kn">input:</span> <span class="s">&#39;file*.fasta&#39;</span><span class="p">,</span> <span class="s">&#39;filename with space.fasta&#39;</span>

<span class="kn">input:</span>
    <span class="s">&#39;file*.txt&#39;</span><span class="p">,</span>
    <span class="s">&#39;directory/file2.txt&#39;</span>

<span class="kn">input:</span> <span class="n">aligned_reads</span>

<span class="kn">input:</span> <span class="n">aligned_reads</span><span class="p">,</span> <span class="n">reference_genome</span>

<span class="kn">input:</span> <span class="n">aligned_reads</span><span class="p">[</span><span class="mi">2</span><span class="p">:]</span>

<span class="kn">input:</span> <span class="s">&#39;data/*.fastq&#39;</span>

<span class="kn">input:</span> <span class="s">&#39;*/GXT*.fastq&#39;</span>

<span class="kn">input:</span> <span class="n">func</span><span class="p">(</span><span class="n">parameter</span><span class="p">)</span></code></pre></figure>

<p>It is worth noting that</p>

<ul>
  <li>The first examples shows that the step does not need any input file (so it does not depend on any other step).</li>
  <li>It does not matter if <code>aligned_reads</code> and <code>reference_genome</code> are strings or lists of strings because SoS will flatten nested lists to a single list of filenames.</li>
  <li>The <code>input</code> option tries to expand filenames with wildcard characters (<code>*</code> and <code>?</code>). This can be very useful for workflows that, for example, regularly scan a directory and process unprocessed files. However, because the value of this step depends on availability of files, the output of <code>sos show script</code> and the execution path will be unpredictable, and even wrong if there is no available file during the execution of <code>sos show script</code>.</li>
</ul>

<p>The input files will be evaluated and form a list of input files. They are by default sent to the step process all at once as varible <code>_input</code>, but can also be sent in groups, each time
with different <code>_input</code>. Here <code>_input</code> is a temporary variable that is available only within the step.</p>

<h3 id="option-filetype">Option <code>filetype</code></h3>

<p>SoS allows the specification of input options, which are appended to input file list as comma separated <code>name=value</code> pairs.</p>

<p>Option <code>filetype</code> accepts one or more filetypes or a lambda function. For example,</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="p">[</span><span class="n">step</span><span class="p">]</span>
<span class="kn">input:</span>
	<span class="n">input_files</span><span class="p">,</span> <span class="kp">filetype</span><span class="o">=</span><span class="s">&#39;*.fastq&#39;</span></code></pre></figure>

<p>passes only files with extension <code>*.fastq</code>.</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="p">[</span><span class="n">step</span><span class="p">]</span>
<span class="kn">input:</span>
	<span class="n">input_files</span><span class="p">,</span>
	<span class="kp">filetype</span><span class="o">=</span><span class="p">[</span><span class="s">&#39;*.fastq&#39;</span><span class="p">,</span> <span class="s">&#39;*.fastq.gz&#39;</span><span class="p">]</span></code></pre></figure>

<p>passes only files with extension <code>.fastq</code> or <code>.fastq.gz</code>. Under the hood SoS treats the pattern as Unix shell-stype wildcard pattern (with <code>*</code>, <code>?</code>, <code>[seq]</code> and <code>[!seq]</code>, see <a href="https://docs.python.org/2/library/fnmatch.html#module-fnmatch">doc</a> for details) so</p>

<ul>
  <li><strong><code>filetype='.txt'</code> does not match <code>file.txt</code></strong></li>
  <li><code>filetype='*.fastq*'</code> matches <code>a.fastq</code>, <code>a.fastq.gz</code> and <code>a.fastq.zip</code></li>
  <li><code>filetype='[!_]*.txt'</code> matches <code>file1.txt</code> but not <code>_file1.txt</code></li>
</ul>

<p>If you need more refined control over the selection of files, you can use lambda functions (a bit python knowledge is required). For example,</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="p">[</span><span class="n">step</span><span class="p">]</span>
<span class="kn">input:</span>
	<span class="n">input_files</span><span class="p">,</span>
	<span class="kp">filetype</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">open</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="o">.</span><span class="n">readline</span><span class="p">()</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s">&#39;##fileformat=VCF4.1&#39;</span><span class="p">)</span></code></pre></figure>

<p>passes only files with the first line starting with string <code>##fileformat=VCF4.1</code>`.</p>

<h3 id="option-groupby">Option <code>group_by</code></h3>

<p>SoS by default passes all input files to step process as a single list. Option <code>group_by</code> pass input files in groups, each time with a subset of input files named <code>_input</code>. SoS allows you to group input by <code>single</code> (individual file), <code>pairs</code> (match first half of files with the second half),  <code>combinations</code> (all unordered combinations of 2-sets), <code>pairwise</code> (all adjacent 2-sets), or chunks of size <code>N</code> (integer <code>group_by=3</code> or string <code>group_by='4'</code>). For example, with the following sos script</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="c"># how to group input files</span>
<span class="kn">parameter:</span> <span class="n">group</span> <span class="o">=</span> <span class="s">&#39;all&#39;</span>

<span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="k">print</span><span class="p">(</span><span class="s">&#39;group_by=${group}&#39;</span><span class="p">)</span>
<span class="kn">input:</span> <span class="s">&#39;file1&#39;</span><span class="p">,</span> <span class="s">&#39;file2&#39;</span><span class="p">,</span> <span class="s">&#39;file3&#39;</span><span class="p">,</span> <span class="s">&#39;file4&#39;</span><span class="p">,</span> <span class="kp">group_by</span><span class="o">=</span><span class="n">group</span>
<span class="k">print</span><span class="p">(</span><span class="s">&#39;${_index}: ${_input}&#39;</span><span class="p">)</span></code></pre></figure>

<p>The output of commands are</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="err">$</span> <span class="n">sos</span> <span class="n">run</span> <span class="n">test</span><span class="o">.</span><span class="n">sos</span>  <span class="o">-</span><span class="n">i</span> <span class="o">-</span><span class="n">v0</span>
<span class="kp">group_by</span><span class="o">=</span><span class="nb">all</span>
<span class="kn">0:</span> <span class="n">file1</span> <span class="n">file2</span> <span class="n">file3</span> <span class="n">file4</span>

<span class="err">$</span> <span class="n">sos</span> <span class="n">run</span> <span class="n">test</span><span class="o">.</span><span class="n">sos</span> <span class="o">--</span><span class="n">group</span> <span class="s">&#39;single&#39;</span> <span class="o">-</span><span class="n">i</span> <span class="o">-</span><span class="n">v0</span>
<span class="kp">group_by</span><span class="o">=</span><span class="n">single</span>
<span class="kn">0:</span> <span class="n">file1</span>
<span class="kn">1:</span> <span class="n">file2</span>
<span class="kn">2:</span> <span class="n">file3</span>
<span class="kn">3:</span> <span class="n">file4</span>

<span class="err">$</span> <span class="n">sos</span> <span class="n">run</span> <span class="n">test</span><span class="o">.</span><span class="n">sos</span> <span class="o">--</span><span class="n">group</span> <span class="s">&#39;pairwise&#39;</span> <span class="o">-</span><span class="n">i</span> <span class="o">-</span><span class="n">v0</span>
<span class="kp">group_by</span><span class="o">=</span><span class="n">pairwise</span>
<span class="kn">0:</span> <span class="n">file1</span> <span class="n">file2</span>
<span class="kn">1:</span> <span class="n">file2</span> <span class="n">file3</span>
<span class="kn">2:</span> <span class="n">file3</span> <span class="n">file4</span>

<span class="err">$</span> <span class="n">sos</span> <span class="n">run</span> <span class="n">test</span><span class="o">.</span><span class="n">sos</span> <span class="o">--</span><span class="n">group</span> <span class="s">&#39;pairs&#39;</span> <span class="o">-</span><span class="n">i</span> <span class="o">-</span><span class="n">v0</span>
<span class="kp">group_by</span><span class="o">=</span><span class="n">pairs</span>
<span class="kn">0:</span> <span class="n">file1</span> <span class="n">file3</span>
<span class="kn">1:</span> <span class="n">file2</span> <span class="n">file4</span>

<span class="err">$</span> <span class="n">sos</span> <span class="n">run</span> <span class="n">test</span><span class="o">.</span><span class="n">sos</span> <span class="o">--</span><span class="n">group</span> <span class="s">&#39;combinations&#39;</span> <span class="o">-</span><span class="n">i</span> <span class="o">-</span><span class="n">v0</span>
<span class="kp">group_by</span><span class="o">=</span><span class="n">combinations</span>
<span class="kn">0:</span> <span class="n">file1</span> <span class="n">file2</span>
<span class="kn">1:</span> <span class="n">file1</span> <span class="n">file3</span>
<span class="kn">2:</span> <span class="n">file1</span> <span class="n">file4</span>
<span class="kn">3:</span> <span class="n">file2</span> <span class="n">file3</span>
<span class="kn">4:</span> <span class="n">file2</span> <span class="n">file4</span>
<span class="kn">5:</span> <span class="n">file3</span> <span class="n">file4</span>

<span class="err">$</span> <span class="n">sos</span> <span class="n">run</span> <span class="n">test</span><span class="o">.</span><span class="n">sos</span> <span class="o">--</span><span class="n">group</span> <span class="mi">1</span> <span class="o">-</span><span class="n">i</span> <span class="o">-</span><span class="n">v0</span>
<span class="kp">group_by</span><span class="o">=</span><span class="mi">1</span>
<span class="kn">0:</span> <span class="n">file1</span>
<span class="kn">1:</span> <span class="n">file2</span>
<span class="kn">2:</span> <span class="n">file3</span>
<span class="kn">3:</span> <span class="n">file4</span>

<span class="err">$</span> <span class="n">sos</span> <span class="n">run</span> <span class="n">test</span><span class="o">.</span><span class="n">sos</span> <span class="o">--</span><span class="n">group</span> <span class="mi">2</span> <span class="o">-</span><span class="n">i</span> <span class="o">-</span><span class="n">v0</span>
<span class="kp">group_by</span><span class="o">=</span><span class="mi">2</span>
<span class="kn">0:</span> <span class="n">file1</span> <span class="n">file2</span>
<span class="kn">1:</span> <span class="n">file3</span> <span class="n">file4</span></code></pre></figure>

<p>Here we are running the script in inspect mode to avoid “file not found” errors for input files.</p>

<p>Obviously, the output of the <code>pairs</code> cases depends on the order of files. If you need to pair files in any particular order, you can control it in input. For example</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="p">[</span><span class="n">step</span><span class="p">]</span>
<span class="kn">input:</span>
	<span class="nb">sorted</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">fastq_files</span> <span class="k">if</span> <span class="s">&#39;_R1_&#39;</span> <span class="ow">in</span> <span class="n">x</span><span class="p">]),</span>
	<span class="nb">sorted</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">fastq_files</span> <span class="k">if</span> <span class="s">&#39;_R2_&#39;</span> <span class="ow">in</span> <span class="n">x</span><span class="p">]),</span>
	<span class="kp">group_by</span><span class="o">=</span><span class="s">&#39;pairs&#39;</span>

<span class="n">run</span><span class="p">(</span><span class="s">&#39;echo ${input}&#39;</span><span class="p">)</span></code></pre></figure>

<p>will take all input files and sort them by <code>_R1_</code> and <code>_R2_</code> and by filename, and pair them.</p>

<h3 id="option-foreach">Option <code>for_each</code></h3>

<p>Option <code>for_each</code> allows you to repeat step process for each value of a variable. For example,</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="n">method</span> <span class="o">=</span> <span class="p">[</span><span class="s">&#39;m1&#39;</span><span class="p">,</span> <span class="s">&#39;m2&#39;</span><span class="p">]</span>
<span class="kn">input:</span> <span class="s">&#39;file1&#39;</span><span class="p">,</span> <span class="s">&#39;file2&#39;</span><span class="p">,</span> <span class="kp">for_each</span><span class="o">=</span><span class="s">&#39;method&#39;</span>
<span class="k">print</span><span class="p">(</span><span class="s">&#39;${_index}: ${_input} ${_method}&#39;</span><span class="p">)</span></code></pre></figure>

<p>will repeat the step with each item of variable <code>method</code></p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="err">$</span> <span class="n">sos</span> <span class="n">run</span> <span class="n">test</span><span class="o">.</span><span class="n">sos</span> <span class="o">-</span><span class="n">v0</span>
<span class="kn">0:</span> <span class="n">file1</span> <span class="n">file2</span> <span class="n">m1</span>
<span class="kn">1:</span> <span class="n">file1</span> <span class="n">file2</span> <span class="n">m2</span></code></pre></figure>

<p>Note that the SoS automatically creates a loop variable <code>_method</code> for variable <code>method</code>. If the variable is an attribute of an object (e.g. <code>aligned.output</code>), the iterator variable will be named without the attribute part.</p>

<p>Nested loops are also allowed. For example,</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="n">method</span> <span class="o">=</span> <span class="p">[</span><span class="s">&#39;m1&#39;</span><span class="p">,</span> <span class="s">&#39;m2&#39;</span><span class="p">]</span>
<span class="n">pars</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">]</span>
<span class="kn">input:</span> <span class="s">&#39;file1&#39;</span><span class="p">,</span> <span class="s">&#39;file2&#39;</span><span class="p">,</span> <span class="kp">for_each</span><span class="o">=</span><span class="p">[</span><span class="s">&#39;method&#39;</span><span class="p">,</span> <span class="s">&#39;pars&#39;</span><span class="p">]</span>
<span class="k">print</span><span class="p">(</span><span class="s">&#39;${_index}: _input=${_input} _method=${_method}, _pars=${_pars}&#39;</span><span class="p">)</span></code></pre></figure>

<p>would produce</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="err">$</span> <span class="n">sos</span> <span class="n">run</span> <span class="n">test</span><span class="o">.</span><span class="n">sos</span> <span class="o">-</span><span class="n">v0</span>
<span class="kn">0:</span> <span class="n">_input</span><span class="o">=</span><span class="n">file1</span> <span class="n">file2</span> <span class="n">_method</span><span class="o">=</span><span class="n">m1</span><span class="p">,</span> <span class="n">_pars</span><span class="o">=</span><span class="mi">1</span>
<span class="kn">1:</span> <span class="n">_input</span><span class="o">=</span><span class="n">file1</span> <span class="n">file2</span> <span class="n">_method</span><span class="o">=</span><span class="n">m2</span><span class="p">,</span> <span class="n">_pars</span><span class="o">=</span><span class="mi">1</span>
<span class="kn">2:</span> <span class="n">_input</span><span class="o">=</span><span class="n">file1</span> <span class="n">file2</span> <span class="n">_method</span><span class="o">=</span><span class="n">m1</span><span class="p">,</span> <span class="n">_pars</span><span class="o">=</span><span class="mi">2</span>
<span class="kn">3:</span> <span class="n">_input</span><span class="o">=</span><span class="n">file1</span> <span class="n">file2</span> <span class="n">_method</span><span class="o">=</span><span class="n">m2</span><span class="p">,</span> <span class="n">_pars</span><span class="o">=</span><span class="mi">2</span></code></pre></figure>

<p>If you would like to loop the process with several parameters, you can put them into the same level by ‘var1,var2’. For example,</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="n">method</span> <span class="o">=</span> <span class="p">[</span><span class="s">&#39;m1&#39;</span><span class="p">,</span> <span class="s">&#39;m2&#39;</span><span class="p">]</span>
<span class="n">pars</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">]</span>
<span class="kn">input:</span> <span class="s">&#39;file1&#39;</span><span class="p">,</span> <span class="s">&#39;file2&#39;</span><span class="p">,</span> <span class="kp">for_each</span><span class="o">=</span><span class="s">&#39;method,pars&#39;</span>
<span class="k">print</span><span class="p">(</span><span class="s">&#39;${_index}: _input=${_input} _method=${_method}, _pars=${_pars}&#39;</span><span class="p">)</span></code></pre></figure>

<p>would produce</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="err">$</span> <span class="n">sos</span> <span class="n">run</span> <span class="n">test</span><span class="o">.</span><span class="n">sos</span> <span class="o">-</span><span class="n">v0</span>
<span class="kn">0:</span> <span class="n">_input</span><span class="o">=</span><span class="n">file1</span> <span class="n">file2</span> <span class="n">_method</span><span class="o">=</span><span class="n">m1</span><span class="p">,</span> <span class="n">_pars</span><span class="o">=</span><span class="mi">1</span>
<span class="kn">1:</span> <span class="n">_input</span><span class="o">=</span><span class="n">file1</span> <span class="n">file2</span> <span class="n">_method</span><span class="o">=</span><span class="n">m2</span><span class="p">,</span> <span class="n">_pars</span><span class="o">=</span><span class="mi">2</span></code></pre></figure>

<p>The variable passed to option <code>for_each</code> can a sequence, or a Pandas dataframe. In this case, each <code>_loop</code> variable presents a line in the dataframe and you can access single values using format <code>_loop["header"]</code>. For example</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="kn">as</span> <span class="nn">pd</span>
<span class="n">data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">([(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="s">&#39;Hello&#39;</span><span class="p">),</span> <span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="s">&#39;World&#39;</span><span class="p">)],</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s">&#39;A&#39;</span><span class="p">,</span> <span class="s">&#39;B&#39;</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">])</span>
<span class="kn">input:</span> <span class="kp">for_each</span><span class="o">=</span><span class="s">&#39;data&#39;</span>
<span class="kn">output:</span> <span class="s">&#39;${_data[&quot;A&quot;]}_${_data[&quot;B&quot;]}_${_data[&quot;C&quot;]}.txt&#39;</span>
<span class="kn">sh:</span>
    <span class="n">touch</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">}</span></code></pre></figure>

<p>would produce results with names <code>1_2_Hello.txt</code> and <code>2_4_World.txt</code>.</p>

<h3 id="option-pairedwith">Option <code>paired_with</code></h3>

<p>Input files might might come with additional information such as sample type, and sample name, and you can pair these information to input files using the <code>paired_with</code> option. For example, <code>bam_files</code> in the following example have matched <code>mutated</code> and <code>sample_name</code>, you can atteched these information to groups of input files <code>_input</code> with looped paired with variables.</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="n">bam_files</span> <span class="o">=</span> <span class="p">[</span><span class="s">&#39;case/A1.bam&#39;</span><span class="p">,</span> <span class="s">&#39;case/A2.bam&#39;</span><span class="p">,</span> <span class="s">&#39;ctrl/A1.bam&#39;</span><span class="p">,</span> <span class="s">&#39;ctrl/A2.bam&#39;</span><span class="p">]</span>
<span class="n">mutated</span> <span class="o">=</span> <span class="p">[</span><span class="s">&#39;case&#39;</span><span class="p">,</span> <span class="s">&#39;case&#39;</span><span class="p">,</span> <span class="s">&#39;ctrl&#39;</span><span class="p">,</span> <span class="s">&#39;ctrl&#39;</span><span class="p">]</span>
<span class="n">sample_name</span> <span class="o">=</span> <span class="p">[</span><span class="s">&#39;A1&#39;</span><span class="p">,</span> <span class="s">&#39;A2&#39;</span><span class="p">,</span> <span class="s">&#39;A1&#39;</span><span class="p">,</span> <span class="s">&#39;A2&#39;</span><span class="p">]</span>

<span class="kn">input:</span> <span class="n">bam_files</span><span class="p">,</span> <span class="kp">paired_with</span><span class="o">=</span><span class="p">[</span><span class="s">&#39;mutated&#39;</span><span class="p">,</span> <span class="s">&#39;sample_name&#39;</span><span class="p">],</span> <span class="kp">group_by</span><span class="o">=</span><span class="s">&#39;pairs&#39;</span>

<span class="k">print</span><span class="p">(</span><span class="s">&#39;${_index}: _input=${_input} _mutated=${_mutated}, _sample_name=${_sample_name}&#39;</span><span class="p">)</span></code></pre></figure>

<p>Output:</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="err">$</span> <span class="n">sos</span> <span class="n">run</span> <span class="n">test</span><span class="o">.</span><span class="n">sos</span> <span class="o">-</span><span class="n">v0</span>
<span class="kn">0:</span> <span class="n">_input</span><span class="o">=</span><span class="n">case</span><span class="o">/</span><span class="n">A1</span><span class="o">.</span><span class="n">bam</span> <span class="n">ctrl</span><span class="o">/</span><span class="n">A1</span><span class="o">.</span><span class="n">bam</span> <span class="n">_mutated</span><span class="o">=</span><span class="n">case</span> <span class="n">ctrl</span><span class="p">,</span> <span class="n">_sample_name</span><span class="o">=</span><span class="n">A1</span> <span class="n">A1</span>
<span class="kn">1:</span> <span class="n">_input</span><span class="o">=</span><span class="n">case</span><span class="o">/</span><span class="n">A2</span><span class="o">.</span><span class="n">bam</span> <span class="n">ctrl</span><span class="o">/</span><span class="n">A2</span><span class="o">.</span><span class="n">bam</span> <span class="n">_mutated</span><span class="o">=</span><span class="n">case</span> <span class="n">ctrl</span><span class="p">,</span> <span class="n">_sample_name</span><span class="o">=</span><span class="n">A2</span> <span class="n">A2</span></code></pre></figure>

<p>Similar to option <code>for_each</code>, if the variable is an attribute of an object (e.g. <code>aligned.output</code>), the iterator variable will be named without the attribute part (e.g. <code>_aligned</code> for <code>paired_with='aligned.output'</code>).</p>

<h3 id="option-pattern">Option <code>pattern</code></h3>

<p>This option does the reverse of function <code>expand_pattern</code>. It uses named wildcards to match pattern to all input files, and creates step variables for these wildcard objects. For example,</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="p">[</span><span class="n">step</span><span class="p">]</span>
<span class="kn">input:</span>  <span class="s">&#39;a-20.txt&#39;</span><span class="p">,</span> <span class="s">&#39;b-10.txt&#39;</span><span class="p">,</span> <span class="kp">pattern</span> <span class="o">=</span> <span class="s">&#39;{name}-{par}.txt&#39;</span>
<span class="kn">output:</span> <span class="n">expand_pattern</span><span class="p">(</span><span class="s">&#39;{name}-processed-{par}.txt&#39;</span><span class="p">)</span>
<span class="n">run</span><span class="p">(</span><span class="s">&#39;echo ${output}; touch ${output}&#39;</span><span class="p">)</span></code></pre></figure>

<p>will take all input files and extract <code>name</code> and <code>par</code> from each file name as variables <code>name</code> and <code>par</code>. It is then used to create output file names adding the word <code>processed</code> in between these wildcard objects. The outcome of the SoS script above is creation of files <code>a-processed-10.txt</code> and <code>b-processed-20.txt</code>.</p>

<p>When wildcard objects are accessed as step variables, both variable names with and without <code>_</code> prefix is available, e.g. in this example, both <code>_name</code> and <code>name</code>, <code>_par</code> and <code>par</code> are avaiable and are the same. The two conventions will only differ when <a href="Documentation#option-group_by"><code>group_by</code></a> or <a href="Documentation#option-for_each">‘for_each’</a> is also used.</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="p">[</span><span class="n">step</span><span class="p">]</span>
<span class="kn">input:</span>  <span class="s">&#39;a-20.txt&#39;</span><span class="p">,</span> <span class="s">&#39;b-10.txt&#39;</span><span class="p">,</span> <span class="n">extract</span> <span class="o">=</span> <span class="s">&#39;{name}-{par}.txt&#39;</span><span class="p">,</span> <span class="kp">group_by</span><span class="o">=</span><span class="s">&#39;single&#39;</span>
<span class="kn">output:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s">&#39;{_name}-processed-{_par}.txt&#39;</span>
<span class="n">run</span><span class="p">(</span><span class="s">&#39;echo ${_output}; touch ${_output}&#39;</span><span class="p">)</span></code></pre></figure>

<p>Please note that the sigil <code>{}</code> is exclusively used for named wildcards and you should not use <code>${}</code> because it will be interpreted as strings.</p>

<h3 id="option-skip-1">Option <code>skip</code></h3>

<p>Option <code>skip</code> takes either a constant (<code>True</code> or <code>False</code>) or a function. Option <code>skip=True</code>  will make SoS skip the execution of the current step. Using <code>skip=True</code> is not very useful so this option is often used with a SoS variable. For example</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="p">[</span><span class="mi">10</span><span class="p">]</span>
<span class="kn">input:</span>
	<span class="n">fasta_files</span><span class="p">,</span>
	<span class="kp">skip</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">fasta_failes</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span>

<span class="kn">output:</span> <span class="s">&#39;merged.fasta&#39;</span>

<span class="n">run</span><span class="p">(</span><span class="s">&#39;command to merge multiple fasta files.&#39;</span><span class="p">)</span></code></pre></figure>

<p>Here the <code>skip</code> option gets the value of <code>True</code> if there is only one input file. The command to merge multiple input files would then be skipped.</p>

<p>Another use of option <code>skip</code> is to assign it a function. In this case, this function will be applied
to each input group with <code>_input</code> as the first variable and all <code>paired_with</code>, <code>pattern</code>, <code>loop</code> variables (e.g. <code>_loopvar</code>) as keyword arguments. The input group will be skipped if this function returns <code>False</code>. For example,</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="p">[</span><span class="mi">10</span><span class="p">]</span>
<span class="k">def</span> <span class="nf">filter_group</span><span class="p">(</span><span class="n">ifiles</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
    <span class="k">return</span> <span class="nb">all</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">getfile</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">ifiles</span><span class="p">)</span>

<span class="kn">input:</span> <span class="kp">group_by</span><span class="o">=</span><span class="s">&#39;combinations&#39;</span><span class="p">,</span> <span class="kp">skip</span><span class="o">=</span><span class="n">filter_group</span></code></pre></figure>

<p>will check all input groups and skip groups with one or two empty files (file size = 0).</p>

<h3 id="dynamic-input-files"><code>dynamic</code> input files</h3>

<p>In order to determine the best execution strategy, SoS evaluates all expressions for all steps before the execution of a workflow to figure
out input and output of steps. This works most of the time but sometimes the input of a step can only be determined at runtime. For example,
if you would like your workflow to automatically scan an input directory and process all fasta files under it, or if a previous step produces
files that cannot be determined beforehand, you can specify input files as follows,</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="kn">input:</span> <span class="s">&#39;input/*.fasta&#39;</span></code></pre></figure>

<p>The problem is that no file or a wrong set files might exist during the planing stage so SoS might skip this step or start the step
with a wrong set of files. To address this problem, you can declare the input files as <strong>dynamic</strong> by passing a <code>dynamic</code> object</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="kn">input:</span> <span class="n">dynamic</span><span class="p">(</span><span class="s">&#39;input/*.fasta&#39;</span><span class="p">)</span></code></pre></figure>

<p>This tells SoS that the input of this step can only be determined at runtime and will execute the step only after all its previous
steps have been completed.</p>

<h3 id="summary">Summary</h3>

<p>Options of step <code>input</code> are evaluated in the following orders:</p>

<ol>
  <li>A list of input files, if specified, would replace <code>input</code>, which is by default output from the previous step.</li>
  <li>Option <code>filetype</code> filters input files. <strong>The output becomes <code>input</code></strong>.</li>
  <li>Option <code>group_by</code> groups the files into several groups, named <code>_input</code></li>
  <li>Option <code>for_each</code> repeat <code>_input</code> for each loop var, named <code>_loopvar</code> if <code>for_each='loopvar'</code>.</li>
  <li>Option <code>paired_with</code> pairs one or more variables with <code>input</code>, variable <code>paired</code> is paired with <code>input</code>
 and variable <code>_paired</code> is paired with <code>_input</code> in each loop if <code>paired_with='paired'</code></li>
  <li>Option <code>pattern</code> extract variables from filenames in <code>input</code>. Variable <code>extracted</code> is paired with <code>input</code>
 and variable <code>_extracted</code> is paired with <code>_input</code> in each loop if <code>extract='{extracted}_other_part'</code>.</li>
  <li>Option <code>skip</code> optionally skip all or part of the input groups.</li>
</ol>

<p>The differences between looped and non-loop steps are sumarized in the following figure</p>

<p><img src="/SOS/media/step_loop.jpg" alt="step_loop" class="img-responsive" /></p>

<h2 id="step-output">Step output</h2>

<h3 id="output-files">Output files</h3>

<p>Output files of a step can be specified by step <code>output</code>. Whereas step <code>input</code> override or
change SoS provided variable <code>input</code> to produce one or more variable <code>_input</code>,
the step <code>output</code> specify <code>output</code> which should be <code>[]</code> if no output is generated
(or of interest). Similar to <code>input</code>, step output accepts strings, variables, expressions, and allows wildcard characters. For example, the following are acceptable output files</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="kn">output:</span>  <span class="p">[]</span>

<span class="kn">output:</span>  <span class="s">&#39;accepted_hits.bam&#39;</span>

<span class="kn">output:</span>  <span class="n">aligned_reads</span><span class="p">,</span> <span class="n">bam_stats</span>

<span class="kn">output:</span>  <span class="s">&#39;aligned/*.bam&#39;</span>

<span class="kn">output:</span>  <span class="n">expand_pattern</span><span class="p">(</span><span class="s">&#39;aligned_{samples}.bam&#39;</span><span class="p">)</span></code></pre></figure>

<p>In the last example, function <code>expand_pattern</code> is used to contruct list of files from items of a sequence <code>samples</code>. In case of <code>input</code> loop, step <code>output</code> actually determines variable <code>_output</code> for each input loop. For example, the following step accepts one or more bam files and index them using command <code>samtools index</code>. The input files are passed one
by one so generate multiple <code>output</code> files (<code>'${_input}.bai'</code>).</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="p">[</span><span class="mi">10</span><span class="p">]</span>
<span class="kn">input:</span>
	<span class="n">bamfiles</span><span class="p">,</span> <span class="kp">group_by</span><span class="o">=</span><span class="s">&#39;single&#39;</span>

<span class="kn">output:</span>
	<span class="s">&#39;${_input}.bai&#39;</span>

<span class="n">run</span><span class="p">(</span><span class="s">&#39;&#39;&#39;samtools index ${_input} &#39;&#39;&#39;</span><span class="p">)</span></code></pre></figure>

<p>The use of variable <code>output</code> in this scenario is discouraged because <code>output</code>, as the collection of all <code>_output</code> increases with each input group. The accumulation only happens when <code>_output</code> and <code>output</code> are different so <code>output</code> will stay the same if all output files are specified as in the following example</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="kn">output:</span> <span class="p">[</span><span class="n">x</span> <span class="o">+</span> <span class="s">&#39;.bai&#39;</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">bamfiles</span><span class="p">]</span></code></pre></figure>

<h3 id="dynamic-output-files"><code>dynamic</code> output files</h3>

<p>Similar to the cases with <a href="#dynamically-determined-input-files-function-dynamic">dynamic input files</a>, the
output of some steps could also not be determined beforehand. For example, with the following script that generates <code>html</code> files that cannot be determined during dry run,</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="kn">run:</span>
    <span class="n">rm</span> <span class="o">-</span><span class="n">f</span> <span class="o">*.</span><span class="n">html</span>

<span class="p">[</span><span class="mi">10</span><span class="p">]</span>
<span class="kn">input:</span> <span class="p">[]</span>
<span class="kn">output:</span> <span class="s">&#39;*.html&#39;</span>

<span class="kn">import</span> <span class="nn">random</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">5</span><span class="p">):</span>
    <span class="n">run</span><span class="p">(</span><span class="s">&#39;touch result_${random.randint(1, 20)}.html&#39;</span><span class="p">)</span></code></pre></figure>

<p>SoS will determine that you do not have any output file (no <code>*.html</code> file) and produce the following output</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="err">$</span> <span class="n">sos</span> <span class="n">run</span> <span class="n">test</span><span class="o">.</span><span class="n">sos</span>
<span class="kn">INFO:</span> <span class="n">Execute</span> <span class="n">default_0</span><span class="p">:</span>
<span class="kn">INFO:</span> <span class="nb">input</span><span class="p">:</span>   <span class="p">[]</span>
<span class="kn">INFO:</span> <span class="n">output</span><span class="p">:</span>  <span class="n">unspecified</span>
<span class="kn">INFO:</span> <span class="n">Execute</span> <span class="n">default_10</span><span class="p">:</span>
<span class="kn">INFO:</span> <span class="nb">input</span><span class="p">:</span>   <span class="p">[]</span>
<span class="kn">WARNING:</span> <span class="o">*.</span><span class="n">html</span> <span class="n">does</span> <span class="ow">not</span> <span class="n">expand</span> <span class="n">to</span> <span class="nb">any</span> <span class="n">valid</span> <span class="nb">file</span><span class="o">.</span>
<span class="kn">INFO:</span> <span class="n">output</span><span class="p">:</span>  <span class="p">[]</span></code></pre></figure>

<p>In this case, you will need to define the output as <code>dynamic</code> using</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="kn">run:</span>
    <span class="n">rm</span> <span class="o">-</span><span class="n">f</span> <span class="o">*.</span><span class="n">html</span>

<span class="p">[</span><span class="mi">10</span><span class="p">]</span>
<span class="kn">input:</span> <span class="p">[]</span>
<span class="kn">output:</span> <span class="n">dynamic</span><span class="p">(</span><span class="s">&#39;*.html&#39;</span><span class="p">)</span>

<span class="kn">import</span> <span class="nn">random</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">5</span><span class="p">):</span>
    <span class="n">run</span><span class="p">(</span><span class="s">&#39;touch result_${random.randint(1, 20)}.html&#39;</span><span class="p">)</span></code></pre></figure>

<p>so that SoS knows that the output can only be determined after the completion of the step. Because of this, variable <code>output</code> is unavailable to the step process.</p>

<h2 id="step-dependencies">Step dependencies</h2>

<p>This item specifies files that are required for the step. Although not required, it is a good practice to list resource files and other dependency files for a particular step. For example</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="p">[</span><span class="mi">10</span><span class="p">]</span>
<span class="kn">input:</span>
	<span class="n">fasta_files</span>

<span class="kn">depends:</span>
	<span class="n">reference_seq</span></code></pre></figure>

<p>Similar to <code>output</code> options, dependent files can also be defined after <code>input</code> options and consist of
dependent files determined from loop variables.</p>

<p>The following figure summarizes the effect of <code>input</code>
and <code>output</code> options and input options <code>group_by</code> and <code>for_each</code> on the flow
of input and output files and related variables.</p>

<p><img src="/SOS/media/step_options.jpg" alt="step_options" class="img-responsive" /></p>

<h2 id="step-process">Step process</h2>

<p>A step process is the Python statements that perform certain tasks and produce step output from step input. A step process can contain arbitrary Python statements. For example,</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="p">[</span><span class="mi">10</span><span class="p">]</span>
<span class="kn">output:</span> <span class="s">&#39;a.txt&#39;</span>
<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="err">!</span><span class="n">r</span><span class="p">},</span> <span class="s">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">dest</span><span class="p">:</span>
   <span class="n">dest</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;some text&#39;</span><span class="p">)</span></code></pre></figure>

<p>and</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="p">[</span><span class="mi">10</span><span class="p">]</span>
<span class="kn">output:</span> <span class="s">&#39;a.txt&#39;</span>
<span class="n">run</span><span class="p">(</span><span class="s">&#39;echo &quot;some text&quot; &gt; ${_output!q}&#39;</span><span class="p">)</span></code></pre></figure>

<p>use inline (interpreted and executed by SoS) python code or shell script to generate <code>a.txt</code>.</p>

<p>Step processes are executed within SoS and are executed sequentially. However, part or all of the step process can be executed externally and potentially in parallel as step <code>task</code>.</p>

<h2 id="external-task">External <code>task</code></h2>

<p>If a job is long and time consuming, it is much preferred to submit them as separate tasks to be executed, for example, on a cluster system. These jobs should be specified using the <code>task</code> keyword, which marks the beginning of a task, with optional runtime options to control its execution. For example,</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="p">[</span><span class="mi">10</span><span class="p">]</span>
<span class="kn">input:</span> <span class="kp">group_by</span><span class="o">=</span><span class="s">&#39;single&#39;</span>

<span class="kn">task:</span> <span class="kp">concurrent</span><span class="o">=</span><span class="bp">True</span>

<span class="n">run</span><span class="p">(</span><span class="s">&#39;&#39;&#39;</span>
<span class="s">samtools index {_input}</span>
<span class="s">&#39;&#39;&#39;</span><span class="p">)</span></code></pre></figure>

<p>execute a shell script in parallel (with <code>concurrent=True</code>). The step process can consists of arbitrary python statements and execute multiple step actions. For example,</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="kn">task:</span>
<span class="kn">try:</span>
   <span class="n">action1</span><span class="p">()</span>
<span class="k">except</span> <span class="ne">RuntimeError</span><span class="p">:</span>
   <span class="n">action2</span><span class="p">()</span></code></pre></figure>

<p>execute <code>action1</code> and <code>action2</code> if <code>action1</code> raises an error.</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="kn">task:</span>
<span class="k">for</span> <span class="n">par</span> <span class="ow">in</span> <span class="p">[</span><span class="s">&#39;-4&#39;</span><span class="p">,</span> <span class="s">&#39;-6&#39;</span><span class="p">]:</span>
   <span class="n">run</span><span class="p">(</span><span class="s">&#39;command with ${par}&#39;</span><span class="p">)</span></code></pre></figure>

<p>executes commands in a loop. This is similar to</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="n">pars</span> <span class="o">=</span> <span class="p">[</span><span class="s">&#39;-4&#39;</span><span class="p">,</span> <span class="s">&#39;-6&#39;</span><span class="p">]</span>
<span class="kn">input:</span> <span class="kp">for_each</span><span class="o">=</span><span class="n">pars</span>
<span class="kn">task:</span>
<span class="n">run</span><span class="p">(</span><span class="s">&#39;command with ${_pars}&#39;</span><span class="p">)</span></code></pre></figure>

<p>but the <code>for</code> loop version would not be able to be executed in parallel. Note that SoS actions can be used outside of <code>step process</code> but only statements specified after the <code>process</code> keyword can have runtime options and be executed in separate processes. That is to say,</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="n">pars</span> <span class="o">=</span> <span class="p">[</span><span class="s">&#39;-4&#39;</span><span class="p">,</span> <span class="s">&#39;-6&#39;</span><span class="p">]</span>
<span class="kn">input:</span> <span class="kp">for_each</span><span class="o">=</span><span class="n">pars</span>
<span class="n">run</span><span class="p">(</span><span class="s">&#39;command with ${_pars}&#39;</span><span class="p">)</span></code></pre></figure>

<p>is equivalent to</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="n">pars</span> <span class="o">=</span> <span class="p">[</span><span class="s">&#39;-4&#39;</span><span class="p">,</span> <span class="s">&#39;-6&#39;</span><span class="p">]</span>
<span class="kn">input:</span> <span class="kp">for_each</span><span class="o">=</span><span class="n">pars</span>
<span class="kn">task:</span>
<span class="n">run</span><span class="p">(</span><span class="s">&#39;command with ${_pars}&#39;</span><span class="p">)</span></code></pre></figure>

<p>but the latter can have additional runtime options to run commands in parallel</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="n">pars</span> <span class="o">=</span> <span class="p">[</span><span class="s">&#39;-4&#39;</span><span class="p">,</span> <span class="s">&#39;-6&#39;</span><span class="p">]</span>
<span class="kn">input:</span> <span class="kp">for_each</span><span class="o">=</span><span class="n">pars</span>
<span class="kn">task:</span> <span class="kp">concurrent</span><span class="o">=</span><span class="bp">True</span>
<span class="n">run</span><span class="p">(</span><span class="s">&#39;command with ${_pars}&#39;</span><span class="p">)</span></code></pre></figure>

<p>Because step tasks are executed outside of SoS, variables assigned in step tasks are not accessible to SoS. For example,</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="p">[</span><span class="mi">10</span><span class="p">:</span> <span class="kp">shared</span><span class="o">=</span><span class="s">&#39;res&#39;</span><span class="p">]</span>
<span class="n">res</span> <span class="o">=</span> <span class="n">some_action</span><span class="p">()</span></code></pre></figure>

<p>executes <code>some_action()</code> in step process and return its result as a shared variable <code>res</code>. The following script,</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="p">[</span><span class="mi">10</span><span class="p">:</span> <span class="kp">shared</span><span class="o">=</span><span class="s">&#39;res&#39;</span><span class="p">]</span>
<span class="kn">task:</span>
<span class="n">res</span> <span class="o">=</span> <span class="n">some_action</span><span class="p">()</span></code></pre></figure>

<p>however, does not work because <code>res</code> is assigned in step task and is not accessible from the step.</p>

<h3 id="option-workdir">Option <code>workdir</code></h3>

<p>Default to current working directory.</p>

<p>Option <code>workdir</code> controls the working directory of the process. For example, the following step downloads a file to the <code>resource_dir</code> using command <code>wget</code>.</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="p">[</span><span class="mi">10</span><span class="p">]</span>

<span class="kn">run:</span> <span class="kp">workdir</span><span class="o">=</span><span class="n">resource_dir</span>

  <span class="n">wget</span> <span class="n">a_url</span> <span class="o">-</span><span class="n">O</span> <span class="n">filename</span></code></pre></figure>

<h3 id="option-concurrent">Option <code>concurrent</code></h3>

<p>Default to <code>False</code>.</p>

<p>If the step process is repeated for different input files or parameters (using input options <code>group_by</code> or <code>for_each</code>), the loop process can be execute in parallel, up to the maximum number of concurrent jobs specified by command line option <code>-j</code>.</p>

<h3 id="option-env">Option <code>env</code></h3>

<p>The <code>env</code> option allow you to modify runtime environment, similar to the <code>env</code> parameter of the <code>subprocess.Popen</code> function. For example, you can execute your command with in a specific directory using</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="kn">task:</span>  <span class="kp">env</span><span class="o">=</span><span class="p">{</span><span class="s">&#39;PATH&#39;</span><span class="p">:</span> <span class="s">&#39;/path/to/mycommand&#39;</span> <span class="o">+</span> <span class="n">os</span><span class="o">.</span><span class="n">sep</span> <span class="o">+</span> <span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="p">[</span><span class="s">&#39;PATH&#39;</span><span class="p">]}</span>
<span class="kn">run:</span>
   <span class="n">mycommand</span> </code></pre></figure>

<h3 id="option-prependpath">Option <code>prepend_path</code></h3>

<p>Option <code>prepend_path</code> is a shortcut to option <code>env</code> to prepend one (a string) or more (a list of strings) paths to system path. For example, the above example can be shortened to</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="kn">task:</span>  <span class="kp">prepend_path</span><span class="o">=</span><span class="s">&#39;/path/to/mycommand&#39;</span>
<span class="kn">run:</span>
   <span class="n">mycommand</span> </code></pre></figure>

<h3 id="option-walltime">Option <code>walltime</code></h3>


                            </div>
                            <button type="button" class="btn btn-default" data-dismiss="modal"><i class="fa fa-times"></i> Close</button>
                        </div>
                    </div> 
                    <div class="footer-wrapper">
                      <footer role="contentinfo">
                        <span>&copy; 2016 Bo Peng.</span>
                      </footer>
                    </div>
                </div>
            </div>
        </div>
    </div> 


   <div class="portfolio-modal modal fade" id="Make-Style-Rules" tabindex="-1" role="dialog" aria-hidden="true">
        <div class="modal-content">
            <div class="close-modal" data-dismiss="modal">
                <div class="lr">
                    <div class="rl">
                    </div>
                </div>
            </div>
            <div class="container">
                <div class="row">
                 <div class="col-lg-8 col-lg-offset-2">
                        <h1>7.Make-Style-Rules</h1>
                        <div class="modal-body">
                         
                            <hr class="star-primary">
                            <h2 id="makefile-style-steps">Makefile-style steps</h2>

<p>Auxiliary steps are special steps that are used only when a target is needed but not available. Such steps are defined in the format of</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="p">[</span><span class="n">step_name</span> <span class="p">:</span> <span class="kp">provides</span><span class="o">=</span><span class="kp">pattern</span><span class="p">]</span></code></pre></figure>

<p>where <code>pattern</code> can be</p>

<ul>
  <li>A file pattern such as <code>{sample}.bam.idx</code></li>
  <li>A <code>target</code> such as <code>executable("ms")</code></li>
  <li>A list (sequence) of one or more file patterns and targets.</li>
</ul>

<p>When a target is required, SoS will look in the script for steps that provides such a target. It will then</p>

<ol>
  <li>Create one or more variables defined by the pattern. For example, if the step matches to <code>AS123.bam.idx</code>, a variable named <code>sample</code> would be defined with value <code>AS123</code>.</li>
  <li>The default <code>input</code> of the step would be <code>None</code> instead of output of previous step because there is no previous step for auxiliary steps.</li>
  <li>The default <code>output</code> of the step would be the matched target, although you are free to generate more output files.</li>
</ol>

<p>For example, step</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="p">[</span><span class="n">index_bam</span> <span class="p">:</span> <span class="kp">provides</span><span class="o">=</span><span class="s">&#39;{sample}.bam.bai&#39;</span><span class="p">]</span>
<span class="kn">input:</span> <span class="s">&#39;${sample}.bam&#39;</span>
<span class="kn">run:</span>
     <span class="n">samtools</span> <span class="n">index</span> <span class="err">$</span><span class="p">{</span><span class="nb">input</span><span class="p">}</span></code></pre></figure>

<p>would be called when a target <code>AS123.bam.bai</code> is required. This step would have variable <code>sample</code> defined as <code>AS123</code>, default input <code>None</code>, and default output <code>["AS123.bam.bai"]</code>. The input would be redefined by <code>input: ${sample}.bam</code> as <code>AS123.bam</code> and be passed to the <code>run</code> action.</p>

<p>You might have already realized that an auxiliary step is a makefile style step and you can use this technique to build complete
workflows in a way similar to <a href="https://bitbucket.org/johanneskoester/snakemake">snakemake</a>. That is to say, you can define multiple
auxiliary steps (rules in snakemake’s term) and let SoS determine what steps to execute depending on what workflow target to produce.
You can even mix the forward, input-oriented style with backward, output-oriented style in SoS.</p>


                            </div>
                            <button type="button" class="btn btn-default" data-dismiss="modal"><i class="fa fa-times"></i> Close</button>
                        </div>
                    </div> 
                    <div class="footer-wrapper">
                      <footer role="contentinfo">
                        <span>&copy; 2016 Bo Peng.</span>
                      </footer>
                    </div>
                </div>
            </div>
        </div>
    </div> 


   <div class="portfolio-modal modal fade" id="SoS-Actions" tabindex="-1" role="dialog" aria-hidden="true">
        <div class="modal-content">
            <div class="close-modal" data-dismiss="modal">
                <div class="lr">
                    <div class="rl">
                    </div>
                </div>
            </div>
            <div class="container">
                <div class="row">
                 <div class="col-lg-8 col-lg-offset-2">
                        <h1>8.SoS-Actions</h1>
                        <div class="modal-body">
                         
                            <hr class="star-primary">
                            <p>Although arbitrary python functions can be used in SoS step process, SoS defines some <strong><code>actions</code></strong> (e.g. the <code>run</code> function in the aforementioned examples)
that can be used in a SoS script. The only difference between a SoS action and a regular Python function is that they can behave differently
in different <code>run_mode</code>. For example,</p>

<ul>
  <li>Most SoS actions return 0 directly in <code>inspect</code> mode.</li>
  <li>Some SoS actions such as <code>check_command</code> work in both <code>run</code> and <code>inspect</code> mode so that you can check if you script can be executed by running it in inspect mode.</li>
</ul>

<p>Because SoS executes all step processes in <code>inspect</code> and <code>prepare</code> mode for the planning of large workflows, <strong>it is important to enclose data processing steps in SoS actions or execute conditionally (with <code>if run_mode == 'run'</code>)</strong>. For example, running script</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="kn">import</span> <span class="nn">time</span>
<span class="n">time</span><span class="o">.</span><span class="n">sleep</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span></code></pre></figure>

<p>with command <code>sos run</code> will sleep for 15 seconds instead of 5 seconds because the <code>sleep</code> function will be called in inspect, prepare, and run mode. To fix this, you can either put the function in an action</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="kn">python:</span>
   <span class="kn">import</span> <span class="nn">time</span>
   <span class="n">time</span><span class="o">.</span><span class="n">sleep</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span></code></pre></figure>

<p>or use <code>run_mode</code> condition</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="k">if</span> <span class="n">run_mode</span> <span class="o">==</span> <span class="s">&#39;run&#39;</span><span class="p">:</span>
    <span class="kn">import</span> <span class="nn">time</span>
    <span class="n">time</span><span class="o">.</span><span class="n">sleep</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span></code></pre></figure>

<p>It is easy to define your own actions. All you need to do is to define a function and decorate it with a <code>SoS_Action</code> decorator. For example</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="kn">from</span> <span class="nn">pysos</span> <span class="kn">import</span> <span class="n">SoS_Action</span>

<span class="nd">@SoS_Action</span><span class="p">(</span><span class="n">run_mode</span><span class="o">=</span><span class="s">&#39;run&#39;</span><span class="p">)</span>
<span class="k">def</span> <span class="nf">my_action</span><span class="p">(</span><span class="n">parameters</span><span class="p">):</span>
    <span class="n">do_something_with_parameters</span>
	<span class="k">return</span> <span class="mi">1</span></code></pre></figure>

<p><code>run_mode='run'</code> can also be <code>run_mode=('inspect', 'run')</code> or <code>run_mode='inspect'</code> indicating in which mode the function will be executed (instead of returning 0 directly).</p>

<h2 id="option-runmode">Option <code>run_mode</code></h2>

<p>All actions have pre-defined run_mode. For example, <code>sh</code> action runs at <code>run</code> mode and <code>download</code> runs at <code>prepare</code> mode. If for some applications you would like to run an action in different mode, you can change it by setting option <code>run_mode</code> to a mode at which an action would be executed. For example, the following step</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="p">[</span><span class="n">install_fastq</span><span class="p">:</span> <span class="kp">provides</span><span class="o">=</span><span class="n">executable</span><span class="p">(</span><span class="s">&#39;fastq&#39;</span><span class="p">)]</span>
<span class="kn">run:</span> <span class="n">run_mode</span><span class="o">=</span><span class="s">&#39;prepare&#39;</span>
  <span class="n">script_to_install_fastq</span></code></pre></figure>

<p>would be called when an executable <code>fastq</code> is required. The commands would be running in <code>prepare</code> mode to install the program for the execution of the workflow.</p>

<h2 id="option-dockerimage">Option <code>docker_image</code></h2>

<p>If a docker image is specified (either a name, an Id, or a file), the action is assumed to be executed in the specified docker. The image will be automatically downloaded (pulled) or loaded (if a <code>.tar</code> or <code>.tar.gz</code> file is specified<code>) if it is not available locally. Note that this option only affect script executing actions such as </code>run<code>, </code>python<code> and </code>perl<code> so other actions such as </code>check_command` will be executed on the host machine even if it is included in the step process.</p>

<p>For example, executing the following script</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="p">[</span><span class="mi">10</span><span class="p">]</span>
<span class="kn">python3:</span> <span class="n">docker_image</span><span class="o">=</span><span class="s">&#39;python&#39;</span>
  <span class="nb">set</span> <span class="o">=</span> <span class="p">{</span><span class="s">&#39;a&#39;</span><span class="p">,</span> <span class="s">&#39;b&#39;</span><span class="p">}</span>
  <span class="k">print</span><span class="p">(</span><span class="nb">set</span><span class="p">)</span></code></pre></figure>

<p>under a docker terminal (that is connected to the docker daemon) will</p>

<ol>
  <li>Pull docker image <code>python</code>,  which is the official docker image for Python 2 and 3.</li>
  <li>Create a python script with the specified content</li>
  <li>Run the docker container <code>python</code> and make the script available inside the container</li>
  <li>Use the <code>python3</code> command inside the container to execute the script.</li>
</ol>

<p>Additional <code>docker_run</code> parameters can be passed to actions when the action
is executed in a docker image. These options include</p>

<ul>
  <li><code>name</code>: name of the container (option <code>--name</code>)</li>
  <li><code>tty</code>: if a tty is attached (default to <code>True</code>, option <code>-t</code>)</li>
  <li><code>stdin_open</code>: if stdin should be open (default to <code>False</code>, option <code>-i</code>)</li>
  <li><code>user</code>: username (default o <code>root</code>, option <code>-u</code>)</li>
  <li><code>environment</code>: Can be a string, a list of string or dictinary of environment variables for docker (option <code>-e</code>)</li>
  <li><code>volumes</code>: string or list of string, extra volumes that need to be link, in addition to SoS mounted (<code>/tmp</code>, <code>/Users</code> (if mac), <code>/Volumes</code> (if <a href="https://github.com/bpeng2000/SOS/wiki/SoS-Docker-guide">properly configured</a> under mac) and script file)</li>
  <li><code>volumes_from</code>: container names or Ids to get volumes from</li>
  <li><code>working_dir</code>: working directory (option <code>-w</code>), default working directory, or working directory set by runtime option <code>workdir</code>.</li>
  <li><code>port</code>: port opened (option <code>-p</code>)</li>
  <li><code>extra_args</code>: If there is any extra arguments you would like to pass to the <code>docker run</code> process (after you check the actual command of <code>docker run</code> of SoS</li>
</ul>

<h2 id="option-dockerfile">Option <code>docker_file</code></h2>

<p>This option allows you to import a docker from specified <code>docker_file</code>, which can be an archive file (<code>.tar</code>, <code>.tar.gz</code>, <code>.tgz</code>, <code>.bzip</code>, <code>.tar.xz</code>, <code>.txz</code>) or a URL to an archive file (e.g. <code>http://example.com/exampleimage.tgz</code>). SoS will use command <code>docker import</code> to import the <code>docker_file</code>. However, because SoS does not know the repository and tag names of the imported docker file, you will still need to use option <code>docker_image</code> to specify the image to use.</p>

<h2 id="action-sosrun">Action <code>sos_run</code></h2>

<p>Action <code>sos_run(workflow, source)</code> executes a specified workflow. The workflow can be a single workflow, a subworkflow (e.g. <code>A_-10</code>), or a combined workflow (e.g. <code>A + B</code>). Because the workflow is executed from a step, it takes step <code>_input</code> as the input of the nested workflow and it can access local step variables. For example,</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="kn">import</span> <span class="nn">random</span>

<span class="p">[</span><span class="n">simulate</span><span class="p">]</span>
<span class="kn">output:</span>   <span class="s">&#39;result_${_reps}.txt&#39;</span>
<span class="kn">run:</span>
<span class="n">simulate_experiment</span> <span class="o">--</span><span class="n">seed</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="n">seed</span><span class="p">}</span> <span class="o">--</span><span class="n">output</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">}</span>

<span class="p">[</span><span class="mi">10</span><span class="p">]</span>
<span class="n">reps</span> <span class="o">=</span> <span class="nb">range</span><span class="p">(</span><span class="mi">100</span><span class="p">)</span>
<span class="kn">input:</span> <span class="kp">for_each</span><span class="o">=</span><span class="s">&#39;reps&#39;</span>
<span class="kn">outout:</span> <span class="s">&#39;result_${_reps}.txt&#39;</span>

<span class="n">seed</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="o">**</span><span class="mi">32</span><span class="p">)</span>
<span class="n">sos_run</span><span class="p">(</span><span class="s">&#39;simulate&#39;</span><span class="p">)</span></code></pre></figure>

<p>would run the nested pipeline <code>simulate</code> (which is a single step in this example) 100 times with their own <code>seed</code>, and <code>_reps</code>.</p>

<p>The workflow can be defined in the current script, or in other SoS scripts, in which case the name or full path of the SoS script should be provided to parameter <code>source</code>. For example,</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="p">[</span><span class="n">myworkflow</span><span class="p">]</span>
<span class="n">sos_run</span><span class="p">(</span><span class="s">&#39;A+B&#39;</span><span class="p">,</span> <span class="n">source</span><span class="o">=</span><span class="s">&quot;AB.sos&quot;</span><span class="p">)</span></code></pre></figure>

<p>defines a nested workflow with workflow <code>A</code> and/or <code>B</code> defined in <code>AB.sos</code>. The nested workflow is a combination of two workflows <code>A</code> and <code>B</code> with their own parameters sections. SoS searches the specified files in the current working directory, the directory of the master script, and a search path specified by variable <code>sos_path</code> defined in the SoS global (<code>~/.sos/config.yaml</code>) or local (<code>.sos/config.yaml</code>) configuration files, and will produce an error if no file can be found.</p>

<h2 id="action-run-bash-sh-csh-tcsh-zsh">Action <code>run</code>, <code>bash</code>, <code>sh</code>, <code>csh</code>, <code>tcsh</code>, <code>zsh</code></h2>

<p>Actions <code>run(script)</code> and <code>bash(script)</code> accepts a shell script and execute it using <code>bash</code>. <code>sh</code>, <code>csh</code>, <code>tcsh</code>, <code>zsh</code> uses respective shell to execute the provided script.</p>

<p>These actions, as well as all script-executing actions such as <code>python</code>, also accept an option <code>args</code> and allows you to pass additional arguments to the interpreter. For example</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="kn">run:</span> <span class="n">args</span><span class="o">=</span><span class="s">&#39;-n&#39;</span>
  <span class="n">echo</span> <span class="s">&quot;a&quot;</span></code></pre></figure>

<p>will execute the script with command <code>bash -n</code> (check syntax).</p>

<h2 id="action-python-and-python3">Action <code>python</code> and <code>python3</code></h2>

<p>Action <code>python(script)</code> and <code>python3(script)</code> accepts a Python script and execute it with python or python3, respectively.</p>

<p>Because SoS can include Python statements directly in a SoS script, it is important to note that embedded Python
statements are interpreted by SoS and the <code>python</code> and <code>python3</code> actions are execute in separate processes without
access to the SoS environment.</p>

<p>For example, the following SoS step execute some python statements <strong>within</strong> SoS with direct access to SoS variables
such as <code>input</code>, and with <code>result</code> writing directly to the SoS environment,</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="p">[</span><span class="mi">10</span><span class="p">]</span>
<span class="k">for</span> <span class="n">filename</span> <span class="ow">in</span> <span class="nb">input</span><span class="p">:</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span> <span class="k">as</span> <span class="n">data</span><span class="p">:</span>
        <span class="n">result</span> <span class="o">=</span> <span class="n">filename</span> <span class="o">+</span> <span class="s">&#39;.res&#39;</span>
        <span class="o">....</span></code></pre></figure>

<p>while</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="p">[</span><span class="mi">10</span><span class="p">]</span>
<span class="kn">input:</span> <span class="kp">group_by</span><span class="o">=</span><span class="s">&#39;single&#39;</span>

<span class="kn">python:</span>

<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="nb">input</span><span class="err">!</span><span class="n">r</span><span class="p">})</span> <span class="k">as</span> <span class="n">data</span><span class="p">:</span>
   <span class="n">result</span> <span class="o">=</span> <span class="err">$</span><span class="p">{</span><span class="nb">input</span><span class="err">!</span><span class="n">r</span><span class="p">}</span> <span class="o">+</span> <span class="s">&#39;.res&#39;</span>
   <span class="o">...</span></code></pre></figure>

<p>composes a Python script for each input file and calls separate Python interpreters to execute them. Whereas
the Python statement in the first example will always be executed, the statements in <code>python</code> will not be executed
in <code>inspect</code> mode.</p>

<h2 id="action-r-and-checkrlibrary">Action <code>R</code> and <code>check_R_library</code></h2>

<p>Action <code>R(script)</code> exexute the passed script using <code>Rscript</code> command. Action <code>check_R_library</code> accept an R library and optionally can check for required versions. If the libraries are not available, it will try to install it from <a href="https://cran.r-project.org/">CRAN</a>, <a href="https://www.bioconductor.org/">bioconductor</a>, or <a href="https://github.com/">github</a>. Github package name should be formatted as <code>repo/pkg</code>. Action <code>check_R_library</code> is available in both <code>run</code> and <code>inspect</code> mode.</p>

<p>For example, <code>check_R_library('edgeR')</code> will check and install (if necessary) <code>edgeR</code> from bioconductor. <code>check_R_library('stephens999/ashr')</code> will check and install <code>ashr</code> package from a github repository https://github.com/stephens999/ashr.</p>

<p><code>check_R_library</code> can also be used to check for required version of packages. For example:</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="n">check_R_library</span><span class="p">(</span><span class="s">&#39;edgeR&#39;</span><span class="p">,</span> <span class="s">&#39;3.12.0&#39;</span><span class="p">)</span></code></pre></figure>

<p>will result in a warning if edgeR version is not 3.12.0, and</p>

<p>Multiple versions is allowed, for example:</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="n">check_R_library</span><span class="p">(</span><span class="s">&#39;edgeR&#39;</span><span class="p">,</span> <span class="p">[</span><span class="s">&#39;3.12.0&#39;</span><span class="p">,</span> <span class="s">&#39;3.12.1&#39;</span><span class="p">])</span></code></pre></figure>

<p>Or, simply:</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="n">check_R_library</span><span class="p">(</span><span class="s">&#39;edgeR&#39;</span><span class="p">,</span> <span class="s">&#39;3.12.0+&#39;</span><span class="p">)</span></code></pre></figure>

<p>It is also possible to restrict package to old version, for example:</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="n">check_R_library</span><span class="p">(</span><span class="s">&#39;ggplot2&#39;</span><span class="p">,</span> <span class="s">&#39;1.0.0-&#39;</span><span class="p">)</span></code></pre></figure>

<p>Note that version mismatch triggers a warning message, not an error. If you would like to enforce certain version, use</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="n">fail_if</span><span class="p">(</span><span class="n">check_R_library</span><span class="p">(</span><span class="s">&#39;ggplot2&#39;</span><span class="p">,</span> <span class="s">&#39;1.0.0+&#39;</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">,</span> <span class="s">&#39;Version 1.0.0 or newer version of ggplot2 is required&#39;</span><span class="p">)</span></code></pre></figure>

<p>The default R library repo is <code>http://cran.us.r-project.org</code>. It is possible to customize the repo, for example:</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="n">check_R_library</span><span class="p">(</span><span class="s">&#39;Rmosek&#39;</span><span class="p">,</span> <span class="n">repos</span> <span class="o">=</span> <span class="s">&quot;http://download.mosek.com/R/7&quot;</span><span class="p">)</span></code></pre></figure>

<h2 id="action-perl-ruby">Action <code>perl</code>, <code>ruby</code></h2>

<p>Action <code>perl(script)</code> execute the passed script using <code>perl</code> interpreter. Action <code>ruby(script)</code> execute the passed script using <code>ruby</code> interpreter.</p>

<h2 id="action-node-javascript">Action <code>node</code>, <code>JavaScript</code></h2>

<p>Action <code>node(script)</code> and <code>JavaScript(script)</code> execute the passed script using <code>node</code> interpreter.</p>

<h2 id="action-dockerbuild">Action <code>docker_build</code></h2>

<p>Build a docker image from an inline Docker file. The inline version of the action currently does not support adding any file from local machine because the docker file will be saved to a random directory. You can walk around this problem by creating a <code>Dockerfile</code> and pass it to the action through option <code>path</code>. This action accepts all parameters as specified in https://docker-py.readthedocs.org/en/stable/api/#build because SoS simply pass additional parameters to the <code>build</code> function.</p>

<p>For example, the following step builds a docker container for <a href="http://miso.readthedocs.org/en/fastmiso/">MISO</a> based on anaconda python 2.7.</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="p">[</span><span class="n">build_1</span><span class="p">]</span>
<span class="c"># building miso from a Dockerfile</span>
<span class="kn">docker_build:</span> <span class="n">tag</span><span class="o">=</span><span class="s">&#39;mdabioinfo/miso:latest&#39;</span>

	<span class="c">############################################################</span>
	<span class="c"># Dockerfile to build MISO container images</span>
	<span class="c"># Based on Anaconda python</span>
	<span class="c">############################################################</span>

	<span class="c"># Set the base image to anaconda Python 2.7 (miso does not support python 3)</span>
	<span class="n">FROM</span> <span class="n">continuumio</span><span class="o">/</span><span class="n">anaconda</span>

	<span class="c"># File Author / Maintainer</span>
	<span class="n">MAINTAINER</span> <span class="n">Bo</span> <span class="n">Peng</span> <span class="o">&lt;</span><span class="n">bpeng</span><span class="nd">@mdanderson.org</span><span class="o">&gt;</span>

	<span class="c"># Update the repository sources list</span>
	<span class="n">RUN</span> <span class="n">apt</span><span class="o">-</span><span class="n">get</span> <span class="n">update</span>

	<span class="c"># Install compiler and python stuff, samtools and git</span>
	<span class="n">RUN</span> <span class="n">apt</span><span class="o">-</span><span class="n">get</span> <span class="n">install</span> <span class="o">--</span><span class="n">yes</span> \
	 <span class="n">build</span><span class="o">-</span><span class="n">essential</span> \
	 <span class="n">gcc</span><span class="o">-</span><span class="n">multilib</span> \
	 <span class="n">gfortran</span> \ 
	 <span class="n">apt</span><span class="o">-</span><span class="n">utils</span> \
	 <span class="n">libblas3</span> \ 
	 <span class="n">liblapack3</span> \
	 <span class="n">libc6</span> \
	 <span class="n">cython</span> \ 
	 <span class="n">samtools</span> \
	 <span class="n">libbam</span><span class="o">-</span><span class="n">dev</span> \
	 <span class="n">bedtools</span> \
	 <span class="n">wget</span> \
	 <span class="n">zlib1g</span><span class="o">-</span><span class="n">dev</span> \ 
	 <span class="n">tar</span> \
	 <span class="n">gzip</span>

	<span class="n">WORKDIR</span> <span class="o">/</span><span class="n">usr</span><span class="o">/</span><span class="n">local</span>
	<span class="n">RUN</span> <span class="n">pip</span> <span class="n">install</span> <span class="n">misopy</span></code></pre></figure>

<h2 id="action-download">Action <code>download</code></h2>

<p>Action <code>download(URLs, dest_dir='.', dest_file=None, decompress=False)</code> download files from specified URLs, which can be a list of URLs, or a string with tab, space or newline separated URLs.</p>

<ul>
  <li>If <code>dest_file</code> is specified, only one URL is allowed and the URL can have any form.</li>
  <li>Otherwise all files will be downloaded to <code>dest_dir</code>. Filenames are determined from URLs so the URLs must have the last portion as the filename to save.</li>
  <li>If <code>decompress</code> is True, <code>.zip</code> file, compressed or plan <code>tar</code> (e.g. <code>.tar.gz</code>) files, and <code>.gz</code> files will be decompressed to the same directory as the downloaded file.</li>
</ul>

<p>For example,</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="p">[</span><span class="mi">10</span><span class="p">]</span>
<span class="n">GATK_RESOURCE_DIR</span> <span class="o">=</span> <span class="s">&#39;/path/to/resource&#39;</span>
<span class="n">GATK_URL</span> <span class="o">=</span> <span class="s">&#39;ftp://gsapubftp-anonymous@ftp.broadinstitute.org/bundle/2.8/hg19/&#39;</span>

<span class="kn">download:</span>   <span class="n">dest</span><span class="o">=</span><span class="n">GATK_RESOURCE_DIR</span>
    <span class="err">$</span><span class="p">{</span><span class="n">GATK_URL</span><span class="p">}</span><span class="o">/</span><span class="mi">1000</span><span class="n">G_omni2</span><span class="o">.</span><span class="mf">5.</span><span class="n">hg19</span><span class="o">.</span><span class="n">sites</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span>
    <span class="err">$</span><span class="p">{</span><span class="n">GATK_URL</span><span class="p">}</span><span class="o">/</span><span class="mi">1000</span><span class="n">G_omni2</span><span class="o">.</span><span class="mf">5.</span><span class="n">hg19</span><span class="o">.</span><span class="n">sites</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span><span class="o">.</span><span class="n">md5</span>
    <span class="err">$</span><span class="p">{</span><span class="n">GATK_URL</span><span class="p">}</span><span class="o">/</span><span class="mi">1000</span><span class="n">G_omni2</span><span class="o">.</span><span class="mf">5.</span><span class="n">hg19</span><span class="o">.</span><span class="n">sites</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">idx</span><span class="o">.</span><span class="n">gz</span>
    <span class="err">$</span><span class="p">{</span><span class="n">GATK_URL</span><span class="p">}</span><span class="o">/</span><span class="mi">1000</span><span class="n">G_omni2</span><span class="o">.</span><span class="mf">5.</span><span class="n">hg19</span><span class="o">.</span><span class="n">sites</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">idx</span><span class="o">.</span><span class="n">gz</span><span class="o">.</span><span class="n">md5</span>
    <span class="err">$</span><span class="p">{</span><span class="n">GATK_URL</span><span class="p">}</span><span class="o">/</span><span class="n">dbsnp_138</span><span class="o">.</span><span class="n">hg19</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span>
    <span class="err">$</span><span class="p">{</span><span class="n">GATK_URL</span><span class="p">}</span><span class="o">/</span><span class="n">dbsnp_138</span><span class="o">.</span><span class="n">hg19</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span><span class="o">.</span><span class="n">md5</span>
    <span class="err">$</span><span class="p">{</span><span class="n">GATK_URL</span><span class="p">}</span><span class="o">/</span><span class="n">dbsnp_138</span><span class="o">.</span><span class="n">hg19</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">idx</span><span class="o">.</span><span class="n">gz</span>
    <span class="err">$</span><span class="p">{</span><span class="n">GATK_URL</span><span class="p">}</span><span class="o">/</span><span class="n">dbsnp_138</span><span class="o">.</span><span class="n">hg19</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">idx</span><span class="o">.</span><span class="n">gz</span><span class="o">.</span><span class="n">md5</span>
    <span class="err">$</span><span class="p">{</span><span class="n">GATK_URL</span><span class="p">}</span><span class="o">/</span><span class="n">hapmap_3</span><span class="o">.</span><span class="mf">3.</span><span class="n">hg19</span><span class="o">.</span><span class="n">sites</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span>
    <span class="err">$</span><span class="p">{</span><span class="n">GATK_URL</span><span class="p">}</span><span class="o">/</span><span class="n">hapmap_3</span><span class="o">.</span><span class="mf">3.</span><span class="n">hg19</span><span class="o">.</span><span class="n">sites</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span><span class="o">.</span><span class="n">md5</span>
    <span class="err">$</span><span class="p">{</span><span class="n">GATK_URL</span><span class="p">}</span><span class="o">/</span><span class="n">hapmap_3</span><span class="o">.</span><span class="mf">3.</span><span class="n">hg19</span><span class="o">.</span><span class="n">sites</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">idx</span><span class="o">.</span><span class="n">gz</span>
    <span class="err">$</span><span class="p">{</span><span class="n">GATK_URL</span><span class="p">}</span><span class="o">/</span><span class="n">hapmap_3</span><span class="o">.</span><span class="mf">3.</span><span class="n">hg19</span><span class="o">.</span><span class="n">sites</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">idx</span><span class="o">.</span><span class="n">gz</span><span class="o">.</span><span class="n">md5</span></code></pre></figure>

<p>download the specified files to <code>GATK_RESOURCE_DIR</code>. The <code>.md5</code> files will be automatically used to validate the content of the associated files. Note that</p>

<p>SoS automatically save signature of downloaded and decompressed files so the files will not be re-downloaded if the action is called multiple times. You can however still still specifies input and output of the step to use step signature</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="p">[</span><span class="mi">10</span><span class="p">]</span>
<span class="n">GATK_RESOURCE_DIR</span> <span class="o">=</span> <span class="s">&#39;/path/to/resource&#39;</span>
<span class="n">GATK_URL</span> <span class="o">=</span> <span class="s">&#39;ftp://gsapubftp-anonymous@ftp.broadinstitute.org/bundle/2.8/hg19/&#39;</span>
<span class="n">RESOUCE_FILES</span> <span class="o">=</span>  <span class="s">&#39;&#39;&#39;1000G_omni2.5.hg19.sites.vcf.gz</span>
<span class="s">    1000G_omni2.5.hg19.sites.vcf.gz.md5</span>
<span class="s">    1000G_omni2.5.hg19.sites.vcf.idx.gz</span>
<span class="s">    1000G_omni2.5.hg19.sites.vcf.idx.gz.md5</span>
<span class="s">    dbsnp_138.hg19.vcf.gz</span>
<span class="s">    dbsnp_138.hg19.vcf.gz.md5</span>
<span class="s">    dbsnp_138.hg19.vcf.idx.gz</span>
<span class="s">    dbsnp_138.hg19.vcf.idx.gz.md5</span>
<span class="s">    hapmap_3.3.hg19.sites.vcf.gz</span>
<span class="s">    hapmap_3.3.hg19.sites.vcf.gz.md5</span>
<span class="s">    hapmap_3.3.hg19.sites.vcf.idx.gz</span>
<span class="s">    hapmap_3.3.hg19.sites.vcf.idx.gz.md5&#39;&#39;&#39;</span><span class="o">.</span><span class="n">split</span><span class="p">()</span> 
<span class="kn">input:</span> <span class="p">[]</span>
<span class="kn">output:</span>  <span class="p">[</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">GATK_RESOURCE_DIR</span><span class="p">,</span> <span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">GATK_RESOURCE_FILES</span><span class="p">]</span>
<span class="n">download</span><span class="p">([</span><span class="s">&#39;${GATK_URL}/${x}&#39;</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">GATK_RESOURCE_FILES</span><span class="p">],</span> <span class="n">dest</span><span class="o">=</span><span class="n">GATK_RESOURCE_DIR</span><span class="p">)</span></code></pre></figure>

<p>Note that the <code>download</code> action uses up to 5 processes to download files. You can change this number by adjusting system configuration <code>sos_download_processes</code>.</p>

<h2 id="action-report">Action <code>report</code></h2>

<p>Action <code>report(script, filename=None, mode='a')</code> writes the specified script to the report channel, or to a file with specified <code>filename</code>. It will by default append to the report (<code>mode='a'</code>) but you can start a new report with <code>mode='w'</code>.</p>

<p>Note that:</p>

<ul>
  <li>
    <p>Although markdown is generally used, SoS does not restrict the type of report a script generates. If so desired, you can use this mechanism to generate a log file in plain text format, a HTML file with fancy style, any flavor of markdown that you can post-process, or a <code>.Rmd</code> file that can be loaded into RMarkdown.</p>
  </li>
  <li>
    <p>The report is generated from the steps that are executed so the content can be controlled with parameters and configuration files.</p>
  </li>
</ul>

<h2 id="action-pandoc">Action <code>pandoc</code></h2>

<p>Action <code>pandoc</code> converts the report generated so far to another format. Parameters <code>outputfile</code>, <code>format</code>, <code>to</code>, <code>filter</code> and <code>extra_args</code> can be used to customize the output. Please refer to <a href="https://pypi.python.org/pypi/pypandoc/">pandoc and pypandoc documentation</a> for details about these parameters.</p>

<h2 id="action-checkcommand">Action <code>check_command</code></h2>

<p>Action <code>check_command(cmd, pattern)</code> check existence of command or the output of the command. This action works in both dyrun and run mode so it can be used to check the existence of command in inspect mode.</p>

<p>For example, if a script contains step</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="p">[</span><span class="mi">100</span><span class="p">]</span>
<span class="n">check_command</span><span class="p">(</span><span class="s">&#39;tophat&#39;</span><span class="p">)</span>
<span class="n">run</span><span class="p">(</span><span class="s">&#39;tophat ...&#39;</span><span class="p">)</span></code></pre></figure>

<p>Command ‘sos run script -i’ would check the existence of command <code>tophat</code> without actually running
the <code>tophat ...</code> command.</p>

<p>Action <code>check_command(cmd, pattern)</code> execute specified <code>cmd</code> and searches specified pattern in its output. <code>pattern</code>  should be one or a list of python regular expressions (see <a href="https://docs.python.org/2/library/re.html">Python re module</a>, especially the <code>search</code> function for details). It raises an exception if the output does not match any of the patterns. This function is usually used to check the version of commands.</p>

<p>For example, action</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="n">check_command</span><span class="p">(</span><span class="s">&#39;STAR --version&#39;</span><span class="p">,</span> <span class="p">[</span><span class="s">&#39;2.4.0&#39;</span><span class="p">,</span> <span class="s">&#39;2.5.0&#39;</span><span class="p">])</span></code></pre></figure>

<p>checks the output of command <code>STAR --version</code> and raises an error if it does not contain string <code>2.4.0</code> or <code>2.5.0</code>.</p>

<p>Note that</p>

<ol>
  <li>
    <p><code>check_command</code> will return a non-zero value if the command exist but returns non-zero value (e.g. because of incorrect command line argument). A warning message will also be printed.</p>
  </li>
  <li>
    <p><code>check_command</code> will timeout after 2 seconds because it is intended to check basic information of specified command. The action will produce a warning message and return 1 in this case.</p>
  </li>
</ol>

<h2 id="action-failif">Action <code>fail_if</code></h2>

<p>Action <code>fail_if(expr, msg='')</code> raises an exception with <code>msg</code> (and terminate the execution of the workflow if the exception is not caught) if <code>expr</code> returns True.</p>

<h2 id="action-warnif">Action <code>warn_if</code></h2>

<p>Action <code>warn_if(expr, msg)</code> yields a warning message <code>msg</code> if <code>expr</code> is evaluate to be true.</p>

<h2 id="action-abortif">Action <code>abort_if</code></h2>

<p>Action <code>abort_if(expr, msg='')</code> stops the execution of the current step (or current processes if within <code>for_each</code> loop) and gives a warning message if <code>msg</code> is specified. For example,</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="p">[</span><span class="mi">10</span><span class="p">]</span>
<span class="kn">input:</span> <span class="s">&#39;*.txt&#39;</span><span class="p">,</span> <span class="kp">for_each</span><span class="o">=</span><span class="mi">1</span>

<span class="n">abort_if</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">getsize</span><span class="p">(</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="o">&gt;</span> <span class="mi">10000</span><span class="p">)</span>
<span class="n">do_something_else</span></code></pre></figure>

<p>skips files that are larger than <code>10k</code>.</p>

<h2 id="utility-functions-and-logger">Utility functions and <code>logger</code></h2>

<p>SoS exposes a few utility functions that can be helpful from time to time.</p>

<h2 id="function-getoutput">Function <code>get_output</code></h2>

<p>Function <code>get_output(cmd)</code> returns the output of command (decoded in <code>UTF-8</code>), which is a shortcut for <code>subprocess.check_output(cmd, shell=True).decode()</code>. It is worth noting that SoS kills any function call after 5 seconds in inspect mode so you will need to put this function call inside a step process if it will take more than 5 seconds to execute.</p>

<h2 id="function-expandpattern">Function <code>expand_pattern</code></h2>

<p>Function <code>expand_pattern</code> expands a string to multiple ones using items of variables quoted between <code>{ }</code>. For example,</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="kn">output:</span> <span class="n">expand_pattern</span><span class="p">(</span><span class="s">&#39;{a}_{b}.txt&#39;</span><span class="p">)</span></code></pre></figure>

<p>is equivalent to</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="kn">output:</span> <span class="p">[</span><span class="s">&#39;{x}_{y}.txt&#39;</span> <span class="k">for</span> <span class="n">x</span><span class="p">,</span><span class="n">y</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">a</span><span class="p">,</span> <span class="n">b</span><span class="p">)]</span></code></pre></figure>

<p>if <code>a</code> and <code>b</code> are sequences of the same length.</p>

<h2 id="sos-logger-object">SoS <code>logger</code> object</h2>

<p>The SoS logger object is a <code>logging</code> object used by SoS to produce various outputs. You can use this object
to output error, warning, info, debug, and trace messages to terminal. For example,</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&#39;I am at ${step_name}&#39;</span><span class="p">)</span></code></pre></figure>

<p>would print a logging message <code>I am at default_0</code> when the first step is execute.</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="err">$</span> <span class="n">sos</span> <span class="n">run</span> <span class="n">test</span><span class="o">.</span><span class="n">sos</span>
<span class="kn">INFO:</span> <span class="n">Execute</span> <span class="n">default_0</span><span class="p">:</span> 
<span class="kn">INFO:</span> <span class="nb">input</span><span class="p">:</span>   <span class="p">[]</span>
<span class="kn">INFO:</span> <span class="n">I</span> <span class="n">am</span> <span class="n">at</span> <span class="n">default_0</span>
<span class="kn">INFO:</span> <span class="n">output</span><span class="p">:</span>  <span class="n">unspecified</span></code></pre></figure>


                            </div>
                            <button type="button" class="btn btn-default" data-dismiss="modal"><i class="fa fa-times"></i> Close</button>
                        </div>
                    </div> 
                    <div class="footer-wrapper">
                      <footer role="contentinfo">
                        <span>&copy; 2016 Bo Peng.</span>
                      </footer>
                    </div>
                </div>
            </div>
        </div>
    </div> 


   <div class="portfolio-modal modal fade" id="Workflow-Execution" tabindex="-1" role="dialog" aria-hidden="true">
        <div class="modal-content">
            <div class="close-modal" data-dismiss="modal">
                <div class="lr">
                    <div class="rl">
                    </div>
                </div>
            </div>
            <div class="container">
                <div class="row">
                 <div class="col-lg-8 col-lg-offset-2">
                        <h1>9.Workflow-Execution</h1>
                        <div class="modal-body">
                         
                            <hr class="star-primary">
                            <h2 id="specification-of-workflow-or-targets">Specification of workflow or targets</h2>

<p>There are two ways to specify a workflow</p>

<ol>
  <li>
    <p>Specify name of a workflow (<code>sos run script workflow</code>), this includes the execution of default workflow, single workflow (e.g. <code>hg18</code>), subworflow (e.g. <code>hg18_:100</code> (up to step 100)), and combined workflows (e.g. <code>hg18+report</code>). SoS will execute steps of the specified workflow, in addition to any auxiliary steps that are needed. The <strong>target</strong> of the execution would be the output of the last step of the workflow.</p>
  </li>
  <li>
    <p>Specify targets that needs to be generated (<code>sos run script -t target</code>). SoS will construct a DAG (Directed Acyclic Graph) with auxiliary steps that are need to generate specified targets and execute the graph.</p>
  </li>
</ol>

<h2 id="dag-and-parallel-execution">DAG and Parallel execution</h2>

<p>Logically speaking, with command</p>

<ul>
  <li>Evaluate parameters with either command line input or their default values</li>
  <li>Evaluate the rest of steps in numeric order</li>
</ul>

<p>Actual execution order depends on step actions, input output options of SoS steps, and option <code>-j</code>.
In the sequential execution mode (by default),
all steps are executed one by one with auxiliary steps called when necessary. If a script is written without any step option and
input and output files, it will be executed sequentially even in parallel execution mode (with <code>-j</code> option).</p>

<p>If the steps are described with necessary input and output information, steps in SoS workflows can be executed in parallel. The
following figure illustrates the impact input/output options on the execution order of workflows. Note that a step with
unknown <code>input</code> (no <code>input</code> at present step and no <code>output</code> at previous step) can only be executed after all its previous steps
are completed can become bottlenecks of the workflow.</p>

<p>[[/SOS/media/workflow.jpg]]</p>

<h2 id="runtime-signature">Runtime signature</h2>

<h2 id="removal-of-intermediate-files">Removal of intermediate files</h2>

<h2 id="targets-of-workflow">Targets of workflow</h2>

<h2 id="configuration-file">Configuration file</h2>

<p>SoS reads a global (<code>~/.sos/config.yaml</code>) and a local (<code>.sos/config.yaml</code>) configuration files before executing any SoS script. The content of these files are available in the <code>CONFIG</code> variable.</p>

<p>SoS currently support the following configuration variables</p>

<ul>
  <li><code>sos_path</code> (default to <code>[]</code>): A list of directories from which sos will try to locate a script if the script is not in the current directory. For example, if you set <code>sos_path</code> to <code>['~/scripts']</code>, <code>sos run myscript.sos</code> will execute <code>~/scripts/myscript.sos</code> if <code>myscript.sos</code> does not exist in the current directory.</li>
  <li><code>sos_download_processes</code> (default to <code>5</code>): Number of download processes to download files specified in action <code>download</code>.</li>
</ul>

<p>The default value of 5 for option <code>sos_download_processes</code> is usually good enough but you can set it to a larger number if you have enough bandwidthonger time if you are working on a machine with, for example, slow disk access. To change this option, you will need to run</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="err">$</span> <span class="n">sos</span> <span class="n">config</span> <span class="o">--</span><span class="k">global</span> <span class="o">--</span><span class="nb">set</span> <span class="s">&#39;sos_download_processes=10&#39;</span></code></pre></figure>

<h2 id="inspect-and-prepare-workflows"><code>inspect</code> and <code>prepare</code> workflows</h2>

<h2 id="alternative-search-path-path">Alternative search path <code>$PATH</code></h2>

<p>SoS executes command in the environment the script is executed.</p>

                            </div>
                            <button type="button" class="btn btn-default" data-dismiss="modal"><i class="fa fa-times"></i> Close</button>
                        </div>
                    </div> 
                    <div class="footer-wrapper">
                      <footer role="contentinfo">
                        <span>&copy; 2016 Bo Peng.</span>
                      </footer>
                    </div>
                </div>
            </div>
        </div>
    </div> 



 
   <div class="portfolio-modal modal fade" id="Quick-start" tabindex="-1" role="dialog" aria-hidden="true">
        <div class="modal-content">
            <div class="close-modal" data-dismiss="modal">
                <div class="lr">
                    <div class="rl">
                    </div>
                </div>
            </div>
            <div class="container">
                <div class="row">
                 <div class="col-lg-8 col-lg-offset-2">
                        <h1>1.Quick-Start</h1>
                        <div class="modal-body">
                         
                            <hr class="star-primary">
                            <p>A SoS script consists of one or more scripts, comments and optional SoS directives. In its simplest form, a sos script is simply a series of scripts that can be executed sequentially by different intepreters.</p>

<p>Let us assume that you are a bioinformaticist needed to compare the expression levels between two samples. After reading some online tutorials, you ended up with some working commands</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="c"># index reference genome</span>
<span class="n">STAR</span> <span class="o">--</span><span class="n">runMode</span> <span class="n">genomeGenerate</span> <span class="o">--</span><span class="n">genomeFastaFile</span> <span class="n">human38</span><span class="o">.</span><span class="n">fasta</span> <span class="o">--</span><span class="n">genomeDir</span> <span class="n">STAR_index</span>
<span class="c"># align reads to the reference genome</span>
<span class="n">STAR</span> <span class="o">--</span><span class="n">genomeDir</span> <span class="n">STAR_index</span> <span class="o">--</span><span class="n">outSAMtype</span> <span class="n">BAM</span> <span class="n">SortedByCoordinate</span> <span class="o">--</span><span class="n">readFilesIn</span> <span class="n">control</span><span class="o">.</span><span class="n">fasta</span> \
    <span class="o">--</span><span class="n">quantMode</span> <span class="n">GeneCounts</span> <span class="o">--</span><span class="n">outFileNamePrefix</span> <span class="n">aligned</span><span class="o">/</span><span class="n">control</span>
<span class="n">STAR</span> <span class="o">--</span><span class="n">genomeDir</span> <span class="n">STAR_index</span> <span class="o">--</span><span class="n">outSAMtype</span> <span class="n">BAM</span> <span class="n">SortedByCoordinate</span> <span class="o">--</span><span class="n">readFilesIn</span> <span class="n">mutated</span><span class="o">.</span><span class="n">fasta</span> \
    <span class="o">--</span><span class="n">quantMode</span> <span class="n">GeneCounts</span> <span class="o">--</span><span class="n">outFileNamePrefix</span> <span class="n">aligned</span><span class="o">/</span><span class="n">mutated</span></code></pre></figure>

<p>The first command builds an index of the reference genome in preparation for the latter steps, the second command aligns reads from the first sample to the reference genome, and the third command aligns reads from the second sample to the reference genome. Do not panic if you do not know what these commands are doing, this is just an example.</p>

<p>These commands generate, among other files, two files named <code>aligned/control.out.tab</code> and <code>aligned/mutated.out.tab</code> with expression counts of all genes. You then wrote a <a href="https://www.r-project.org/">R</a> script to analyze the results, something like</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="n">control</span><span class="o">.</span><span class="n">count</span> <span class="o">&lt;-</span> <span class="n">read</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="s">&#39;aligned/control.out.tab&#39;</span><span class="p">)</span>
<span class="n">mutated</span><span class="o">.</span><span class="n">count</span> <span class="o">&lt;-</span> <span class="n">read</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="s">&#39;aligned/mutated.out.tab&#39;</span><span class="p">)</span>
<span class="c"># normalize, compare, output etc, ignored.</span>
<span class="n">pdf</span><span class="p">(</span><span class="s">&#39;myfigure.pdf&#39;</span><span class="p">)</span>
<span class="c"># plot results</span>
<span class="n">dev</span><span class="o">.</span><span class="n">off</span><span class="p">()</span></code></pre></figure>

<h1 id="your-first-sos-script">Your first SoS script</h1>

<p>The project completed successfully and you needed to archive the scripts for later reference. Instead of having two files lying around with perhaps another <code>README</code> file to describe what you have done, you can write a single SoS script named <code>myanalysis.sos</code> with content</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="c">#!/usr/bin/env sos-runner</span>
<span class="c">#fileformat=SOS1.0</span>

<span class="c"># This script aligns raw reads of a control and a mutated sample</span>
<span class="c"># to the reference genome and compare the expression values</span>
<span class="c"># of the samples at genes A, B and C.</span>

<span class="kn">run:</span>
<span class="c"># index reference genome</span>
<span class="n">STAR</span> <span class="o">--</span><span class="n">runMode</span> <span class="n">genomeGenerate</span> <span class="o">--</span><span class="n">genomeFastaFile</span> <span class="n">human38</span><span class="o">.</span><span class="n">fasta</span> <span class="o">--</span><span class="n">genomeDir</span> <span class="n">STAR_index</span>

<span class="c"># align reads to the reference genome</span>
<span class="n">STAR</span> <span class="o">--</span><span class="n">genomeDir</span> <span class="n">STAR_index</span> <span class="o">--</span><span class="n">outSAMtype</span> <span class="n">BAM</span> <span class="n">SortedByCoordinate</span>  <span class="o">--</span><span class="n">readFilesIn</span> <span class="n">control</span><span class="o">.</span><span class="n">fasta</span> \
    <span class="o">--</span><span class="n">quantMode</span> <span class="n">GeneCounts</span> <span class="o">--</span><span class="n">outFileNamePrefix</span> <span class="n">aligned</span><span class="o">/</span><span class="n">control</span>
<span class="n">STAR</span> <span class="o">--</span><span class="n">genomeDir</span> <span class="n">STAR_index</span> <span class="o">--</span><span class="n">outSAMtype</span> <span class="n">BAM</span> <span class="n">SortedByCoordinate</span> <span class="o">--</span><span class="n">readFilesIn</span> <span class="n">mutated</span><span class="o">.</span><span class="n">fasta</span> \
    <span class="o">--</span><span class="n">quantMode</span> <span class="n">GeneCounts</span> <span class="o">--</span><span class="n">outFileNamePrefix</span> <span class="n">aligned</span><span class="o">/</span><span class="n">mutated</span>

<span class="kn">R:</span>
<span class="n">control</span><span class="o">.</span><span class="n">count</span> <span class="o">&lt;-</span> <span class="n">read</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="s">&#39;aligned/control.out.tab&#39;</span><span class="p">)</span>
<span class="n">mutated</span><span class="o">.</span><span class="n">count</span> <span class="o">&lt;-</span> <span class="n">read</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="s">&#39;aligned/mutated.out.tab&#39;</span><span class="p">)</span>
<span class="c"># normalize, compare, output etc, ignored.</span>
<span class="n">pdf</span><span class="p">(</span><span class="s">&#39;myfigure.pdf&#39;</span><span class="p">)</span>
<span class="c"># plot results</span>
<span class="n">dev</span><span class="o">.</span><span class="n">off</span><span class="p">()</span></code></pre></figure>

<p>Here <strong><code>run</code></strong> and <strong><code>R</code></strong> are SoS actions that executes the following scripts in <code>bash</code> and <code>R</code> respectively. The scripts are included verbatim and end after reaching another SoS action or directive. It is however clearer to indent the script and write your SoS script as:</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="c">#!/usr/bin/env sos-runner</span>
<span class="c">#fileformat=SOS1.0</span>

<span class="c"># This script aligns raw reads of a control and a mutated sample</span>
<span class="c"># to the reference genome and compare the expression values</span>
<span class="c"># of the samples at genes A, B and C.</span>

<span class="kn">run:</span>
    <span class="c"># index reference genome</span>
    <span class="n">STAR</span> <span class="o">--</span><span class="n">runMode</span> <span class="n">genomeGenerate</span> <span class="o">--</span><span class="n">genomeFastaFile</span> <span class="n">human38</span><span class="o">.</span><span class="n">fasta</span> <span class="o">--</span><span class="n">genomeDir</span> <span class="n">STAR_index</span>

    <span class="c"># align reads to the reference genome</span>
    <span class="n">STAR</span> <span class="o">--</span><span class="n">genomeDir</span> <span class="n">STAR_index</span> <span class="o">--</span><span class="n">outSAMtype</span> <span class="n">BAM</span> <span class="n">SortedByCoordinate</span>  <span class="o">--</span><span class="n">readFilesIn</span> <span class="n">control</span><span class="o">.</span><span class="n">fasta</span> \
        <span class="o">--</span><span class="n">quantMode</span> <span class="n">GeneCounts</span> <span class="o">--</span><span class="n">outFileNamePrefix</span> <span class="n">aligned</span><span class="o">/</span><span class="n">control</span>
    <span class="n">STAR</span> <span class="o">--</span><span class="n">genomeDir</span> <span class="n">STAR_index</span> <span class="o">--</span><span class="n">outSAMtype</span> <span class="n">BAM</span> <span class="n">SortedByCoordinate</span> <span class="o">--</span><span class="n">readFilesIn</span> <span class="n">mutated</span><span class="o">.</span><span class="n">fasta</span> \
        <span class="o">--</span><span class="n">quantMode</span> <span class="n">GeneCounts</span> <span class="o">--</span><span class="n">outFileNamePrefix</span> <span class="n">aligned</span><span class="o">/</span><span class="n">mutated</span>

<span class="c"># compare expression values</span>
<span class="kn">R:</span>
    <span class="n">control</span><span class="o">.</span><span class="n">count</span> <span class="o">&lt;-</span> <span class="n">read</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="s">&#39;aligned/control.out.tab&#39;</span><span class="p">)</span>
    <span class="n">mutated</span><span class="o">.</span><span class="n">count</span> <span class="o">&lt;-</span> <span class="n">read</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="s">&#39;aligned/mutated.out.tab&#39;</span><span class="p">)</span>
    <span class="c"># normalize, compare, output etc, ignored.</span>
    <span class="n">pdf</span><span class="p">(</span><span class="s">&#39;myfigure.pdf&#39;</span><span class="p">)</span>
    <span class="c"># plot results</span>
    <span class="n">dev</span><span class="o">.</span><span class="n">off</span><span class="p">()</span></code></pre></figure>

<p>The scripts in this case end with the end of indentation so that you can add some comments for the <code>R</code> script without being considered as part of the previous script.</p>

<p>You can execute the shell and R scripts defined in this SoS script sequentially by running command</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="n">sos</span> <span class="n">run</span> <span class="n">myanalysis</span></code></pre></figure>

<p>or simply</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="n">myanalysis</span><span class="o">.</span><span class="n">sos</span></code></pre></figure>

<p>if you give <code>myanalyis.sos</code> executable permission (<code>chmod +x myanalysis.sos</code>).</p>

<h1 id="separate-scripts-into-steps">Separate scripts into steps</h1>

<p>Because the scripts perform different tasks, it is logically clearer to separate them into different <strong>steps</strong>. Actually, because the first reference-generating command is not a data processing step, it makes sense to separate it into two scripts. This can be done by inserting step headers to the script as follows:</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="c">#!/usr/bin/env sos-runner</span>
<span class="c">#fileformat=SOS1.0</span>

<span class="c"># This script aligns raw reads of a control and a mutated sample</span>
<span class="c"># to the reference genome and compare the expression values</span>
<span class="c"># of the samples at genes A, B and C.</span>

<span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<span class="c"># index reference genome</span>
<span class="kn">run:</span>
    <span class="n">STAR</span> <span class="o">--</span><span class="n">runMode</span> <span class="n">genomeGenerate</span> <span class="o">--</span><span class="n">genomeFastaFile</span> <span class="n">human38</span><span class="o">.</span><span class="n">fasta</span> <span class="o">--</span><span class="n">genomeDir</span> <span class="n">STAR_index</span>

<span class="p">[</span><span class="mi">2</span><span class="p">]</span>
<span class="c"># align reads to the reference genome</span>
<span class="kn">run:</span>
    <span class="n">STAR</span> <span class="o">--</span><span class="n">genomeDir</span> <span class="n">STAR_index</span> <span class="o">--</span><span class="n">outSAMtype</span> <span class="n">BAM</span> <span class="n">SortedByCoordinate</span>  <span class="o">--</span><span class="n">readFilesIn</span> <span class="n">control</span><span class="o">.</span><span class="n">fasta</span> \
        <span class="o">--</span><span class="n">quantMode</span> <span class="n">GeneCounts</span> <span class="o">--</span><span class="n">outFileNamePrefix</span> <span class="n">aligned</span><span class="o">/</span><span class="n">control</span>
    <span class="n">STAR</span> <span class="o">--</span><span class="n">genomeDir</span> <span class="n">STAR_index</span> <span class="o">--</span><span class="n">outSAMtype</span> <span class="n">BAM</span> <span class="n">SortedByCoordinate</span> <span class="o">--</span><span class="n">readFilesIn</span> <span class="n">mutated</span><span class="o">.</span><span class="n">fasta</span> \
        <span class="o">--</span><span class="n">quantMode</span> <span class="n">GeneCounts</span> <span class="o">--</span><span class="n">outFileNamePrefix</span> <span class="n">aligned</span><span class="o">/</span><span class="n">mutated</span>

<span class="p">[</span><span class="mi">3</span><span class="p">]</span>
<span class="c"># compare expression values</span>
<span class="kn">R:</span>
    <span class="n">control</span><span class="o">.</span><span class="n">count</span> <span class="o">&lt;-</span> <span class="n">read</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="s">&#39;aligned/control.out.tab&#39;</span><span class="p">)</span>
    <span class="n">mutated</span><span class="o">.</span><span class="n">count</span> <span class="o">&lt;-</span> <span class="n">read</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="s">&#39;aligned/mutated.out.tab&#39;</span><span class="p">)</span>
    <span class="c"># normalize, compare, output etc, ignored.</span>
    <span class="n">pdf</span><span class="p">(</span><span class="s">&#39;myfigure.pdf&#39;</span><span class="p">)</span>
    <span class="c"># plot results</span>
    <span class="n">dev</span><span class="o">.</span><span class="n">off</span><span class="p">()</span></code></pre></figure>

<p>Now, when you execute the script with command</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="n">sos</span> <span class="n">run</span> <span class="n">myscript</span></code></pre></figure>

<p>SoS will display <code>default_1</code>, <code>default_2</code> and <code>default_3</code> to report progress. The comments after section heads are considered step comments and will also be displayed during execution.</p>

<h1 id="make-the-script-work-for-other-input-files">Make the script work for other input files</h1>

<p>After a while, before you almost forgot about this analysis, you needed to analyze another pair of samples. You could copy <code>myanalysis.sos</code> to <code>myanalysis2.sos</code>, change filenames and run it, but an easier way is to change your SoS file to accommodate other input files. This can be done by defining a command line argument and passing files name to a <strong>SoS variable</strong>:</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="c">#!/usr/bin/env sos-runner</span>
<span class="c">#fileformat=SOS1.0</span>

<span class="c"># This script aligns raw reads of a control and a mutated sample</span>
<span class="c"># to the reference genome and compare the expression values</span>
<span class="c"># of the samples at genes A, B and C.</span>

<span class="c"># Two input files in .fasta formats. The first one for control sample</span>
<span class="c"># and the second one for mutated sample.</span>
<span class="kn">parameter:</span> <span class="n">fasta_files</span><span class="o">=</span><span class="p">[</span><span class="s">&#39;control.fasta&#39;</span><span class="p">,</span> <span class="s">&#39;mutated.fasta&#39;</span><span class="p">]</span>

<span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<span class="c"># index reference genome</span>
<span class="kn">run:</span>
    <span class="n">STAR</span> <span class="o">--</span><span class="n">runMode</span> <span class="n">genomeGenerate</span> <span class="o">--</span><span class="n">genomeFastaFile</span> <span class="n">human38</span><span class="o">.</span><span class="n">fasta</span> <span class="o">--</span><span class="n">genomeDir</span> <span class="n">STAR_index</span>

<span class="p">[</span><span class="mi">2</span><span class="p">]</span>
<span class="c"># align reads to the reference genome</span>
<span class="kn">run:</span>
    <span class="n">STAR</span> <span class="o">--</span><span class="n">genomeDir</span> <span class="n">STAR_index</span> <span class="o">--</span><span class="n">outSAMtype</span> <span class="n">BAM</span> <span class="n">SortedByCoordinate</span> \
        <span class="o">--</span><span class="n">readFilesIn</span> <span class="err">$</span><span class="p">{</span><span class="n">fasta_files</span><span class="p">[</span><span class="mi">0</span><span class="p">]}</span> <span class="o">--</span><span class="n">quantMode</span> <span class="n">GeneCounts</span> \
        <span class="o">--</span><span class="n">outFileNamePrefix</span> <span class="n">aligned</span><span class="o">/</span><span class="n">control</span>
    <span class="n">STAR</span> <span class="o">--</span><span class="n">genomeDir</span> <span class="n">STAR_index</span> <span class="o">--</span><span class="n">outSAMtype</span> <span class="n">BAM</span> <span class="n">SortedByCoordinate</span> \
        <span class="o">--</span><span class="n">readFilesIn</span> <span class="err">$</span><span class="p">{</span><span class="n">fasta_files</span><span class="p">[</span><span class="mi">1</span><span class="p">]}</span> <span class="o">--</span><span class="n">quantMode</span> <span class="n">GeneCounts</span> \
        <span class="o">--</span><span class="n">outFileNamePrefix</span> <span class="n">aligned</span><span class="o">/</span><span class="n">mutated</span>

<span class="p">[</span><span class="mi">3</span><span class="p">]</span>
<span class="c"># compare expression values</span>
<span class="kn">R:</span>
    <span class="n">control</span><span class="o">.</span><span class="n">count</span> <span class="o">&lt;-</span> <span class="n">read</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="s">&#39;aligned/control.out.tab&#39;</span><span class="p">)</span>
    <span class="n">mutated</span><span class="o">.</span><span class="n">count</span> <span class="o">&lt;-</span> <span class="n">read</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="s">&#39;aligned/mutated.out.tab&#39;</span><span class="p">)</span>
    <span class="c"># normalize, compare, output etc, ignored.</span>
    <span class="n">pdf</span><span class="p">(</span><span class="s">&#39;myfigure.pdf&#39;</span><span class="p">)</span>
    <span class="c"># plot results</span>
    <span class="n">dev</span><span class="o">.</span><span class="n">off</span><span class="p">()</span></code></pre></figure>

<p>A command line argument <code>fasta_files</code> is defined in with <code>parameter</code> keyword. With this definition, you can pass two filenames to variable <code>fasta_files</code> from command line</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="n">sos</span> <span class="n">run</span> <span class="n">myanalysis</span> <span class="o">--</span><span class="n">fasta_files</span> <span class="n">control1</span><span class="o">.</span><span class="n">fasta</span> <span class="n">control2</span><span class="o">.</span><span class="n">fasta</span></code></pre></figure>

<p><code>${fasta_files[0]}</code> and <code>${fasta_files[1]}</code> in command <code>STAR --genomeDir
...</code> will be replaced with their values before the commands are executed. Here <code>fasta_files[0]</code> and <code>fasta_files[1]</code> are Python expressions that will be evaluated during execution.</p>

<h1 id="ignore-steps-that-do-not-need-to-be-rerun">Ignore steps that do not need to be rerun</h1>

<p>Although the SoS script now accepts command line arguments, it is still no more than a compilation of scripts and you immediately realized that it is a waste of time to execute the first command each time. To solve this problem, you can convert the SoS script to a real workflow by telling SoS the input and output of each step:</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="c">#!/usr/bin/env sos-runner</span>
<span class="c">#fileformat=SOS1.0</span>

<span class="c"># This script aligns raw reads of a control and a mutated sample</span>
<span class="c"># to the reference genome and compare the expression values</span>
<span class="c"># of the samples at genes A, B and C.</span>

<span class="c"># Two input files in .fasta formats. The first one for control sample</span>
<span class="c"># and the second one for mutated sample.</span>
<span class="kn">parameter:</span> <span class="n">fasta_files</span><span class="o">=</span><span class="p">[</span><span class="s">&#39;control.fasta&#39;</span><span class="p">,</span> <span class="s">&#39;mutated.fasta&#39;</span><span class="p">]</span>

<span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<span class="c"># create a index for reference genome</span>
<span class="kn">output:</span> <span class="s">&#39;STAR_index/chrName.txt&#39;</span>
<span class="kn">run:</span>
    <span class="n">STAR</span> <span class="o">--</span><span class="n">runMode</span> <span class="n">genomeGenerate</span> <span class="o">--</span><span class="n">genomeFastaFile</span> <span class="n">human38</span><span class="o">.</span><span class="n">fasta</span> <span class="o">--</span><span class="n">genomeDir</span> <span class="n">STAR_index</span>

<span class="p">[</span><span class="mi">2</span><span class="p">]</span>
<span class="c"># align the reads to the reference genome</span>
<span class="kn">input:</span>    <span class="n">fasta_files</span>
<span class="kn">depends:</span>  <span class="s">&#39;STAR_index/chrName.txt&#39;</span>
<span class="kn">output:</span>   <span class="p">[</span><span class="s">&#39;aligned/control.out.tab&#39;</span><span class="p">,</span> <span class="s">&#39;aligned/mutated.out.tab&#39;</span><span class="p">]</span>
<span class="kn">run:</span>
    <span class="n">STAR</span> <span class="o">--</span><span class="n">genomeDir</span> <span class="n">STAR_index</span> <span class="o">--</span><span class="n">outSAMtype</span> <span class="n">BAM</span> <span class="n">SortedByCoordinate</span>  <span class="o">--</span><span class="n">readFilesIn</span> <span class="err">$</span><span class="p">{</span><span class="nb">input</span><span class="p">[</span><span class="mi">0</span><span class="p">]}</span>  \
        <span class="o">--</span><span class="n">quantMode</span> <span class="n">GeneCounts</span> <span class="o">--</span><span class="n">outFileNamePrefix</span> <span class="n">aligned</span><span class="o">/</span><span class="n">control</span>
    <span class="n">STAR</span> <span class="o">--</span><span class="n">genomeDir</span> <span class="n">STAR_index</span> <span class="o">--</span><span class="n">outSAMtype</span> <span class="n">BAM</span> <span class="n">SortedByCoordinate</span> <span class="o">--</span><span class="n">readFilesIn</span> <span class="err">$</span><span class="p">{</span><span class="nb">input</span><span class="p">[</span><span class="mi">1</span><span class="p">]}</span>  \
        <span class="o">--</span><span class="n">quantMode</span> <span class="n">GeneCounts</span> <span class="o">--</span><span class="n">outFileNamePrefix</span> <span class="n">aligned</span><span class="o">/</span><span class="n">mutated</span>

<span class="p">[</span><span class="mi">3</span><span class="p">]</span>
<span class="c"># compare expression values</span>
<span class="kn">output:</span> <span class="s">&#39;myfigure.pdf&#39;</span>
<span class="kn">R:</span>
    <span class="n">control</span><span class="o">.</span><span class="n">count</span> <span class="o">&lt;-</span> <span class="n">read</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="s">&#39;${input[0]}&#39;</span><span class="p">)</span>
    <span class="n">mutated</span><span class="o">.</span><span class="n">count</span> <span class="o">&lt;-</span> <span class="n">read</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="s">&#39;${input[1]}&#39;</span><span class="p">)</span>
    <span class="c"># normalize, compare, output etc, ignored.</span>
    <span class="n">pdf</span><span class="p">(</span><span class="s">&#39;${output}&#39;</span><span class="p">)</span>
    <span class="c"># plot results</span>
    <span class="n">dev</span><span class="o">.</span><span class="n">off</span><span class="p">()</span></code></pre></figure>

<p>Here we</p>

<ul>
  <li>Use <strong>output directive</strong> to specify the expected output of all steps.</li>
  <li>Use <strong>input directive</strong> to specify the input of step 2. Step 1 by default has no input and input for step 3 by default is the output of step 2, its previous step.</li>
  <li>Use <strong>depends directive</strong> to let step 2 depend on the output of step 1.</li>
  <li>Use <code>${input[0]}</code> and <code>${input[1]}</code> in step 2 and 3 because these steps now have properly-defined <code>input</code>. This variable is defined by step input as the input file of the step.</li>
</ul>

<p>With such information, when you run the same command</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="n">sos</span> <span class="n">run</span> <span class="n">myanalysis</span> <span class="o">--</span><span class="nb">input</span> <span class="n">control1</span><span class="o">.</span><span class="n">fasta</span> <span class="n">control2</span><span class="o">.</span><span class="n">fasta</span></code></pre></figure>

<p>SoS will ignore step 1 if this step has been run with output <code>STAR_index/chrName.txt</code>. The same happens to step 2 and 3 so all steps will be ignored if you run the script repeatedly with the same input and processing scripts. SoS uses <strong>runtime signature</strong> for each step and will re-run the step if and only if the content or filename of input, output files or the processing scripts are changed.</p>

<h1 id="use-make-rule-to-define-resource-providing-steps">Use make-rule to define resource-providing steps</h1>

<p>Instead of using runtime signature to avoid re-running the first step, we can also make step 1 an optional step that will be executed only necessary. That is to say, we can define this step as an <code>auxiliary step</code> that will only be called when the file it <strong>provides</strong> does not exist.</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="c">#!/usr/bin/env sos-runner</span>
<span class="c">#fileformat=SOS1.0</span>

<span class="c"># This script aligns raw reads of a control and a mutated sample</span>
<span class="c"># to the reference genome and compare the expression values</span>
<span class="c"># of the samples at genes A, B and C.</span>

<span class="c"># Two input files in .fasta formats. The first one for control sample</span>
<span class="c"># and the second one for mutated sample.</span>
<span class="kn">parameter:</span> <span class="n">fasta_files</span><span class="o">=</span><span class="p">[</span><span class="s">&#39;control.fasta&#39;</span><span class="p">,</span> <span class="s">&#39;mutated.fasta&#39;</span><span class="p">]</span>

<span class="p">[</span><span class="n">build_index</span><span class="p">:</span> <span class="kp">provides</span><span class="o">=</span><span class="s">&#39;STAR_index/chrName.txt&#39;</span><span class="p">]</span>
<span class="c"># create a index for reference genome</span>
<span class="kn">run:</span>
    <span class="n">STAR</span> <span class="o">--</span><span class="n">runMode</span> <span class="n">genomeGenerate</span> <span class="o">--</span><span class="n">genomeFastaFile</span> <span class="n">human38</span><span class="o">.</span><span class="n">fasta</span> <span class="o">--</span><span class="n">genomeDir</span> <span class="n">STAR_index</span>

<span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<span class="c"># align the reads to the reference genome</span>
<span class="kn">input:</span>    <span class="n">fasta_files</span>
<span class="kn">depends:</span>  <span class="s">&#39;STAR_index/chrName.txt&#39;</span>
<span class="kn">output:</span>   <span class="p">[</span><span class="s">&#39;aligned/control.out.tab&#39;</span><span class="p">,</span> <span class="s">&#39;aligned/mutated.out.tab&#39;</span><span class="p">]</span>
<span class="kn">run:</span>
    <span class="n">STAR</span> <span class="o">--</span><span class="n">genomeDir</span> <span class="n">STAR_index</span> <span class="o">--</span><span class="n">outSAMtype</span> <span class="n">BAM</span> <span class="n">SortedByCoordinate</span>  <span class="o">--</span><span class="n">readFilesIn</span> <span class="err">$</span><span class="p">{</span><span class="nb">input</span><span class="p">[</span><span class="mi">0</span><span class="p">]}</span>  \
        <span class="o">--</span><span class="n">quantMode</span> <span class="n">GeneCounts</span> <span class="o">--</span><span class="n">outFileNamePrefix</span> <span class="n">aligned</span><span class="o">/</span><span class="n">control</span>
    <span class="n">STAR</span> <span class="o">--</span><span class="n">genomeDir</span> <span class="n">STAR_index</span> <span class="o">--</span><span class="n">outSAMtype</span> <span class="n">BAM</span> <span class="n">SortedByCoordinate</span> <span class="o">--</span><span class="n">readFilesIn</span> <span class="err">$</span><span class="p">{</span><span class="nb">input</span><span class="p">[</span><span class="mi">1</span><span class="p">]}</span>  \
        <span class="o">--</span><span class="n">quantMode</span> <span class="n">GeneCounts</span> <span class="o">--</span><span class="n">outFileNamePrefix</span> <span class="n">aligned</span><span class="o">/</span><span class="n">mutated</span>

<span class="p">[</span><span class="mi">2</span><span class="p">]</span>
<span class="c"># compare expression values</span>
<span class="kn">output:</span> <span class="s">&#39;myfigure.pdf&#39;</span>
<span class="kn">R:</span>
    <span class="n">control</span><span class="o">.</span><span class="n">count</span> <span class="o">&lt;-</span> <span class="n">read</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="s">&#39;${input[0]}&#39;</span><span class="p">)</span>
    <span class="n">mutated</span><span class="o">.</span><span class="n">count</span> <span class="o">&lt;-</span> <span class="n">read</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="s">&#39;${input[1]}&#39;</span><span class="p">)</span>
    <span class="c"># normalize, compare, output etc, ignored.</span>
    <span class="n">pdf</span><span class="p">(</span><span class="s">&#39;${output}&#39;</span><span class="p">)</span>
    <span class="c"># plot results</span>
    <span class="n">dev</span><span class="o">.</span><span class="n">off</span><span class="p">()</span></code></pre></figure>

<p>The new script consists of two steps, and an auxiliary step <code>build_index</code> that will only be executed when <code>STAR_index/chrName.txt</code> is not available. A slight difference between this version and the previous one is that while the previous version will always execute the first step, this version will not execute it if <code>STAR_index/chrName.txt</code> has been generated before in any way, perhaps not by SoS.</p>

<h1 id="execute-long-running-jobs-externally">Execute long-running jobs externally</h1>

<p>The first step will takes a long time to execute. Instead of executing them by SoS, you might want to submit the commands to a job-queue and be executed and monitored externally. This is especially useful when you need to execute multiple SoS workflows on a cluster-based system. Without going through the details on how to set up your job-queue (see another tutorial for details), it is almost trivial to modify your script to define part of a <strong>step process</strong> to a <strong>task</strong>:</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="c">#!/usr/bin/env sos-runner</span>
<span class="c">#fileformat=SOS1.0</span>

<span class="c"># This script aligns raw reads of a control and a mutated sample</span>
<span class="c"># to the reference genome and compare the expression values</span>
<span class="c"># of the samples at genes A, B and C.</span>

<span class="c"># Two input files in .fasta formats. The first one for control sample</span>
<span class="c"># and the second one for mutated sample.</span>
<span class="kn">parameter:</span> <span class="n">fasta_files</span><span class="o">=</span><span class="p">[</span><span class="s">&#39;control.fasta&#39;</span><span class="p">,</span> <span class="s">&#39;mutated.fasta&#39;</span><span class="p">]</span>

<span class="p">[</span><span class="n">build_index</span><span class="p">:</span> <span class="kp">provides</span><span class="o">=</span><span class="s">&#39;STAR_index/chrName.txt&#39;</span><span class="p">]</span>
<span class="c"># create a index for reference genome</span>
<span class="kn">run:</span>
    <span class="n">STAR</span> <span class="o">--</span><span class="n">runMode</span> <span class="n">genomeGenerate</span> <span class="o">--</span><span class="n">genomeFastaFile</span> <span class="n">human38</span><span class="o">.</span><span class="n">fasta</span> <span class="o">--</span><span class="n">genomeDir</span> <span class="n">STAR_index</span>

<span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<span class="c"># align the reads to the reference genome</span>
<span class="kn">input:</span>    <span class="n">fasta_files</span>
<span class="kn">depends:</span>  <span class="s">&#39;STAR_index/chrName.txt&#39;</span>
<span class="kn">output:</span>   <span class="p">[</span><span class="s">&#39;aligned/control.out.tab&#39;</span><span class="p">,</span> <span class="s">&#39;aligned/mutated.out.tab&#39;</span><span class="p">]</span>
<span class="kn">task:</span>
<span class="kn">run:</span>
    <span class="n">STAR</span> <span class="o">--</span><span class="n">genomeDir</span> <span class="n">STAR_index</span> <span class="o">--</span><span class="n">outSAMtype</span> <span class="n">BAM</span> <span class="n">SortedByCoordinate</span>  <span class="o">--</span><span class="n">readFilesIn</span> <span class="err">$</span><span class="p">{</span><span class="nb">input</span><span class="p">[</span><span class="mi">0</span><span class="p">]}</span>  \
        <span class="o">--</span><span class="n">quantMode</span> <span class="n">GeneCounts</span> <span class="o">--</span><span class="n">outFileNamePrefix</span> <span class="n">aligned</span><span class="o">/</span><span class="n">control</span>
    <span class="n">STAR</span> <span class="o">--</span><span class="n">genomeDir</span> <span class="n">STAR_index</span> <span class="o">--</span><span class="n">outSAMtype</span> <span class="n">BAM</span> <span class="n">SortedByCoordinate</span> <span class="o">--</span><span class="n">readFilesIn</span> <span class="err">$</span><span class="p">{</span><span class="nb">input</span><span class="p">[</span><span class="mi">1</span><span class="p">]}</span>  \
        <span class="o">--</span><span class="n">quantMode</span> <span class="n">GeneCounts</span> <span class="o">--</span><span class="n">outFileNamePrefix</span> <span class="n">aligned</span><span class="o">/</span><span class="n">mutated</span>

<span class="p">[</span><span class="mi">2</span><span class="p">]</span>
<span class="c"># compare expression values</span>
<span class="kn">output:</span> <span class="s">&#39;myfigure.pdf&#39;</span>
<span class="kn">R:</span>
    <span class="n">control</span><span class="o">.</span><span class="n">count</span> <span class="o">&lt;-</span> <span class="n">read</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="s">&#39;${input[0]}&#39;</span><span class="p">)</span>
    <span class="n">mutated</span><span class="o">.</span><span class="n">count</span> <span class="o">&lt;-</span> <span class="n">read</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="s">&#39;${input[1]}&#39;</span><span class="p">)</span>
    <span class="c"># normalize, compare, output etc, ignored.</span>
    <span class="n">pdf</span><span class="p">(</span><span class="s">&#39;${output}&#39;</span><span class="p">)</span>
    <span class="c"># plot results</span>
    <span class="n">dev</span><span class="o">.</span><span class="n">off</span><span class="p">()</span></code></pre></figure>

<p>As you can see, the only difference is the insertion of <code>task:</code> directive before <code>run</code>. Now, when you execute the command, the <code>STAR</code> commands will be executed externally and you can monitor the status of the jobs using your web browser. If no job-queue is set up, the command will be executed in a separate process.</p>

<h1 id="execute-steps-in-parallel">Execute steps in parallel</h1>

<p>Although the two steps of this example have to be executed sequentially, the first step runs the <code>STAR</code> command twice on two input files, and can be executed in parallel. You can tell this to SoS by modifying the script as follows</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="c">#!/usr/bin/env sos-runner</span>
<span class="c">#fileformat=SOS1.0</span>

<span class="c"># This script aligns raw reads of a control and a mutated sample</span>
<span class="c"># to the reference genome and compare the expression values</span>
<span class="c"># of the samples at genes A, B and C.</span>

<span class="c"># Two input files in .fasta formats. The first one for control sample</span>
<span class="c"># and the second one for mutated sample.</span>
<span class="kn">parameter:</span> <span class="n">fasta_files</span><span class="o">=</span><span class="p">[</span><span class="s">&#39;control.fasta&#39;</span><span class="p">,</span> <span class="s">&#39;mutated.fasta&#39;</span><span class="p">]</span>

<span class="p">[</span><span class="n">build_index</span><span class="p">:</span> <span class="kp">provides</span><span class="o">=</span><span class="s">&#39;STAR_index/chrName.txt&#39;</span><span class="p">]</span>
<span class="c"># create a index for reference genome</span>
<span class="kn">run:</span>
    <span class="n">STAR</span> <span class="o">--</span><span class="n">runMode</span> <span class="n">genomeGenerate</span> <span class="o">--</span><span class="n">genomeFastaFile</span> <span class="n">human38</span><span class="o">.</span><span class="n">fasta</span> <span class="o">--</span><span class="n">genomeDir</span> <span class="n">STAR_index</span>

<span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<span class="c"># align the reads to the reference genome</span>
<span class="n">sample_type</span> <span class="o">=</span> <span class="p">[</span><span class="s">&#39;control&#39;</span><span class="p">,</span> <span class="s">&#39;mutated&#39;</span><span class="p">]</span>
<span class="kn">input:</span>    <span class="n">fasta_files</span><span class="p">,</span> <span class="kp">group_by</span><span class="o">=</span><span class="s">&#39;single&#39;</span><span class="p">,</span> <span class="kp">paired_with</span><span class="o">=</span><span class="s">&#39;sample_type&#39;</span>
<span class="kn">depends:</span>  <span class="s">&#39;STAR_index/chrName.txt&#39;</span>
<span class="kn">output:</span>   <span class="s">&#39;aligned/${_sample_type}.out.tab&#39;</span>

<span class="kn">task:</span>     <span class="kp">concurrent</span><span class="o">=</span><span class="bp">True</span>
<span class="kn">run:</span>
    <span class="n">STAR</span> <span class="o">--</span><span class="n">genomeDir</span> <span class="n">STAR_index</span> <span class="o">--</span><span class="n">outSAMtype</span> <span class="n">BAM</span> <span class="n">SortedByCoordinate</span>  <span class="o">--</span><span class="n">readFilesIn</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="err">!</span><span class="n">q</span><span class="p">}</span>  \
        <span class="o">--</span><span class="n">quantMode</span> <span class="n">GeneCounts</span> <span class="o">--</span><span class="n">outFileNamePrefix</span> <span class="n">aligned</span><span class="o">/</span><span class="err">$</span><span class="p">{</span><span class="n">_sample_type</span><span class="p">}</span>

<span class="p">[</span><span class="mi">2</span><span class="p">]</span>
<span class="c"># compare expression values</span>
<span class="kn">output:</span> <span class="s">&#39;myfigure.pdf&#39;</span>
<span class="kn">R:</span>
    <span class="n">control</span><span class="o">.</span><span class="n">count</span> <span class="o">&lt;-</span> <span class="n">read</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="s">&#39;${input[0]}&#39;</span><span class="p">)</span>
    <span class="n">mutated</span><span class="o">.</span><span class="n">count</span> <span class="o">&lt;-</span> <span class="n">read</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="s">&#39;${input[1]}&#39;</span><span class="p">)</span>
    <span class="c"># normalize, compare, output etc, ignored.</span>
    <span class="n">pdf</span><span class="p">(</span><span class="s">&#39;${output}&#39;</span><span class="p">)</span>
    <span class="c"># plot results</span>
    <span class="n">dev</span><span class="o">.</span><span class="n">off</span><span class="p">()</span></code></pre></figure>

<p>Here we</p>

<ol>
  <li>Use option <code>group_by='single'</code> to pass input one by one to action. The action will be executed twice with <code>_input</code> set to the first and second input file respectively.</li>
  <li>Define a variable <code>sample_type</code> and pair it with input files (option <code>paired_with</code>). This will generate a variable <code>_sample_type</code> for each input file so <code>_sample_type</code> will be <code>control</code> for the first input file, and <code>mutated</code> for the second.</li>
  <li>Use <code>${_input}</code> and <code>${_sample_type}</code> to define partial <code>output</code>.</li>
  <li>Use <code>${_input!q}</code> instead of <code>${_input}</code> in the script. This is a small trick to shell-quote filenames so that filenames with spaces and other special characters can be properly quoted in shell commands.</li>
</ol>

<p>Now, if you execute the script with option <code>-j 2</code> (2 concurrent processes),</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="n">sos</span> <span class="n">run</span> <span class="n">myanalysis</span><span class="o">.</span><span class="n">sos</span> <span class="o">--</span><span class="nb">input</span> <span class="n">control1</span><span class="o">.</span><span class="n">fasta</span> <span class="n">control2</span><span class="o">.</span><span class="n">fasta</span> <span class="o">-</span><span class="n">j</span> <span class="mi">2</span></code></pre></figure>

<p>the second step submit two jobs to job-queue, or execute them in two separate processes if a job-queue is not setup.</p>

<p>We have showed you multiple versions of the same SoS script, each using more features of SoS. This actually demonstrates one of the advantages of the SoS system, namely you can start using SoS in minutes without knowing any of its advanced features, and gradually improve your script if needed.</p>

                            </div>
                            <button type="button" class="btn btn-default" data-dismiss="modal"><i class="fa fa-times"></i> Close</button>
                        </div>
                    </div> 
                    <div class="footer-wrapper">
                      <footer role="contentinfo">
                        <span>&copy; 2016 Bo Peng.</span>
                      </footer>
                    </div>
                </div>
            </div>
        </div>
    </div> 


   <div class="portfolio-modal modal fade" id="Using-SoS-with-iPython" tabindex="-1" role="dialog" aria-hidden="true">
        <div class="modal-content">
            <div class="close-modal" data-dismiss="modal">
                <div class="lr">
                    <div class="rl">
                    </div>
                </div>
            </div>
            <div class="container">
                <div class="row">
                 <div class="col-lg-8 col-lg-offset-2">
                        <h1>2.Using-SoS-with-iPython</h1>
                        <div class="modal-body">
                         
                            <hr class="star-primary">
                            <p>Using sos magic within iPython allows you to run sos within an iPython session. It does not provide a full-blown SoS system but on the other hand you can use all the iPython features that you are familiar with.</p>

<h2 id="setting-up-ipython">Setting up ipython</h2>

<p>sos magic is installed by default when you install sos. A profile has been created for you so that you can use it by command</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="n">ipython</span> <span class="o">--</span><span class="n">profile</span> <span class="n">sos</span> </code></pre></figure>

<p>You can also load the extension using command <code>%load_ext sos_magic</code> after ipython starts in its default profile, or edit <code>~/.ipython/profile_default/ipython_config.py</code> and add <code>sos_magic</code> to <code>c.InteractiveShellApp.extensions</code> so that the extension is loaded automatically to your default profile. The latter is useful if you would like to use the ipython mode in Jupyter notebook (<code>jupyter notebook</code> and select python as your kernel).</p>

<h2 id="sos-ipython-magics">SoS ipython magics</h2>

<h3 id="magic-sosdict">Magic <code>sosdict</code></h3>

<p>Let us starts with the basic. When you start an <code>ipython</code> interactive shell through <code>ipython</code> command or <a href="https://ipython.org/notebook.html">ipython/Jupyter notebook</a>, you are given a cell that you can enter python expression or statements, or iPython magic. There are two kinds of magics</p>

<ul>
  <li><strong>Line magic</strong> that starts with <code>%</code> or starts at the beginning of line without <code>%</code> if <code>automagic</code> is set to <code>True</code> (default). Line magic takes the words after it as parameter.</li>
  <li><strong>Cell magic</strong> that starts with <code>%%</code> that takes both the words and lines after the magic word.</li>
</ul>

<p>If the <code>sos_magic</code> extension is loaded, it already has a SoS dictionary. You can get the dictionary using line magic <code>%sosdict</code>, for example</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="n">In</span> <span class="p">[</span><span class="mi">1</span><span class="p">]:</span> <span class="n">sosdict</span>
<span class="n">Out</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span> <span class="p">{}</span></code></pre></figure>

<p>The dictionary is empty because we have not assigned anything to it. Let us run a sos statement</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="n">In</span> <span class="p">[</span><span class="mi">2</span><span class="p">]:</span> <span class="n">sos</span> <span class="n">a</span> <span class="o">=</span> <span class="mi">1</span>

<span class="n">In</span> <span class="p">[</span><span class="mi">3</span><span class="p">]:</span> <span class="n">sosdict</span>
<span class="n">Out</span><span class="p">[</span><span class="mi">3</span><span class="p">]:</span> <span class="p">{</span><span class="s">&#39;a&#39;</span><span class="p">:</span> <span class="mi">1</span><span class="p">}</span></code></pre></figure>

<p>and you can see the sos dictionary contains one item. There are other usages of the <code>%sosdict</code> magic, you can get the dictionary by assigning the dictionary to a variable</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="n">In</span> <span class="p">[</span><span class="mi">4</span><span class="p">]:</span> <span class="n">d</span> <span class="o">=</span> <span class="o">%</span><span class="n">sosdict</span>

<span class="n">In</span> <span class="p">[</span><span class="mi">5</span><span class="p">]:</span> <span class="n">d</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
<span class="n">Out</span><span class="p">[</span><span class="mi">5</span><span class="p">]:</span> <span class="n">dict_keys</span><span class="p">([</span><span class="s">&#39;a&#39;</span><span class="p">])</span></code></pre></figure>

<p>If you are interested in only a subset of variables, you can list them after <code>%sosdict</code></p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="n">In</span> <span class="p">[</span><span class="mi">6</span><span class="p">]:</span> <span class="n">d</span> <span class="o">=</span> <span class="o">%</span><span class="n">sosdict</span> <span class="n">a</span>

<span class="n">In</span> <span class="p">[</span><span class="mi">7</span><span class="p">]:</span> <span class="n">d</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
<span class="n">Out</span><span class="p">[</span><span class="mi">7</span><span class="p">]:</span> <span class="n">dict_keys</span><span class="p">([</span><span class="s">&#39;a&#39;</span><span class="p">])</span></code></pre></figure>

<p>You can get the keys of the dictionary easier using</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="n">In</span> <span class="p">[</span><span class="mi">8</span><span class="p">]:</span> <span class="o">%</span><span class="n">sosdict</span> <span class="o">--</span><span class="n">keys</span>
<span class="n">Out</span><span class="p">[</span><span class="mi">8</span><span class="p">]:</span> <span class="p">{</span><span class="s">&#39;a&#39;</span><span class="p">}</span></code></pre></figure>

<p>If after a while you would like to reset the dictionary and run a SoS script from fresh, you can reset the dictionary using option <code>--reset</code></p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="n">In</span> <span class="p">[</span><span class="mi">9</span><span class="p">]:</span> <span class="n">sosdict</span> <span class="o">--</span><span class="n">reset</span>

<span class="n">In</span> <span class="p">[</span><span class="mi">10</span><span class="p">]:</span> <span class="n">sosdict</span>
<span class="n">Out</span><span class="p">[</span><span class="mi">10</span><span class="p">]:</span> <span class="p">{}</span></code></pre></figure>

<p>The SoS dictionary actually contains other items such as all the SoS actions and functions. If you would like to see all of them, use option <code>--all</code>. For example,</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="n">In</span> <span class="p">[</span><span class="mi">11</span><span class="p">]:</span> <span class="n">sosdict</span> <span class="o">--</span><span class="n">keys</span> <span class="o">--</span><span class="nb">all</span>
<span class="n">Out</span><span class="p">[</span><span class="mi">11</span><span class="p">]:</span> <span class="n">dict_keys</span><span class="p">([</span><span class="s">&#39;fail_if&#39;</span><span class="p">,</span> <span class="s">&#39;expand_pattern&#39;</span><span class="p">,</span> <span class="s">&#39;SoS_Script&#39;</span><span class="p">,</span> <span class="s">&#39;zsh&#39;</span><span class="p">,</span> <span class="s">&#39;download&#39;</span><span class="p">,</span> <span class="s">&#39;perl&#39;</span><span class="p">,</span> <span class="s">&#39;__builtins__&#39;</span><span class="p">,</span> <span class="s">&#39;R&#39;</span><span class="p">,</span> <span class="s">&#39;logger&#39;</span><span class="p">,</span> <span class="s">&#39;ruby&#39;</span><span class="p">,</span> <span class="s">&#39;docker_commit&#39;</span><span class="p">,</span> <span class="s">&#39;Rmarkdown&#39;</span><span class="p">,</span> <span class="s">&#39;python3&#39;</span><span class="p">,</span> <span class="s">&#39;report&#39;</span><span class="p">,</span> <span class="s">&#39;env_variable&#39;</span><span class="p">,</span> <span class="s">&#39;stop_if&#39;</span><span class="p">,</span> <span class="s">&#39;SoS_Action&#39;</span><span class="p">,</span> <span class="s">&#39;sos_variable&#39;</span><span class="p">,</span> <span class="s">&#39;tcsh&#39;</span><span class="p">,</span> <span class="s">&#39;warn_if&#39;</span><span class="p">,</span> <span class="s">&#39;dynamic&#39;</span><span class="p">,</span> <span class="s">&#39;executable&#39;</span><span class="p">,</span> <span class="s">&#39;__interactive__&#39;</span><span class="p">,</span> <span class="s">&#39;check_R_library&#39;</span><span class="p">,</span> <span class="s">&#39;JavaScript&#39;</span><span class="p">,</span> <span class="s">&#39;run&#39;</span><span class="p">,</span> <span class="s">&#39;sos_handle_parameter_&#39;</span><span class="p">,</span> <span class="s">&#39;get_output&#39;</span><span class="p">,</span> <span class="s">&#39;node&#39;</span><span class="p">,</span> <span class="s">&#39;docker_build&#39;</span><span class="p">,</span> <span class="s">&#39;interpolate&#39;</span><span class="p">,</span> <span class="s">&#39;python&#39;</span><span class="p">,</span> <span class="s">&#39;sos_run&#39;</span><span class="p">,</span> <span class="s">&#39;csh&#39;</span><span class="p">,</span> <span class="s">&#39;bash&#39;</span><span class="p">,</span> <span class="s">&#39;pandoc&#39;</span><span class="p">,</span> <span class="s">&#39;sos_namespace_&#39;</span><span class="p">,</span> <span class="s">&#39;execute_script&#39;</span><span class="p">,</span> <span class="s">&#39;check_command&#39;</span><span class="p">,</span> <span class="s">&#39;sh&#39;</span><span class="p">])</span></code></pre></figure>

<p>In summary, the <code>%sosdict</code> magic accepts</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="o">%</span><span class="n">sosdict</span> <span class="p">[</span><span class="o">-</span><span class="n">a</span><span class="o">|-</span><span class="nb">all</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">k</span><span class="o">|--</span><span class="n">keys</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">r</span><span class="o">|--</span><span class="n">reset</span><span class="p">]</span> <span class="p">[</span><span class="n">var1</span><span class="p">]</span> <span class="p">[</span><span class="n">var2</span><span class="p">]</span> <span class="o">...</span></code></pre></figure>

<p>where</p>

<ul>
  <li><code>var1</code>, <code>var2</code> etc are name of variables. All variables will be displayed if no variable is specified.</li>
  <li><code>-a|-all</code>: list all dictionary keys, including SoS functions and variables.</li>
  <li><code>-k|--keys</code>: list only keys, not their values</li>
  <li><code>-r|--reset</code>: reset the dictionary to its original content (with only SoS internal values)</li>
</ul>

<h3 id="magic-sos">Magic <code>sos</code></h3>

<p>Magic <code>sos</code> can be either a line magic or cell magic. Using <code>sos</code> as a line magic, it simply executes the SoS (python) expression or statement in SoS. For example,</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="n">In</span> <span class="p">[</span><span class="mi">5</span><span class="p">]:</span> <span class="n">sos</span> <span class="n">a</span><span class="o">=</span><span class="mi">10</span>

<span class="n">In</span> <span class="p">[</span><span class="mi">6</span><span class="p">]:</span> <span class="n">sos</span> <span class="s">&#39;a + 100 = ${a+100}&#39;</span>
<span class="n">Out</span><span class="p">[</span><span class="mi">6</span><span class="p">]:</span> <span class="s">&#39;a + 100 = 110&#39;</span>

<span class="n">In</span> <span class="p">[</span><span class="mi">7</span><span class="p">]:</span> <span class="n">sos</span> <span class="n">b</span><span class="o">=</span><span class="p">[</span><span class="s">&#39;file1.txt&#39;</span><span class="p">,</span> <span class="s">&#39;file2.txt&#39;</span><span class="p">]</span>

<span class="n">In</span> <span class="p">[</span><span class="mi">8</span><span class="p">]:</span> <span class="n">sos</span> <span class="s">&#39;${b!r,}&#39;</span>
<span class="n">Out</span><span class="p">[</span><span class="mi">8</span><span class="p">]:</span> <span class="s">&quot;&#39;file1.txt&#39;,&#39;file2.txt&#39;&quot;</span></code></pre></figure>

<p>If you would like to execute multi-line SoS statements or scripts, you will need to use the magic in cell mode (with <code>%%</code> prefix), for example,</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="n">In</span> <span class="p">[</span><span class="mi">9</span><span class="p">]:</span> <span class="o">%%</span><span class="n">sos</span>
<span class="kn">run:</span>
    <span class="n">echo</span> <span class="s">&quot;something&quot;</span>
   <span class="o">....</span><span class="p">:</span> 
<span class="n">something</span></code></pre></figure>

<p>runs a shell script within iPython. Similarly, you can run arbitrary shell, R, perl, python code in ipython/SoS, with string interpolation. Note that ipython already has a magic called <a href="https://ipython.org/ipython-doc/3/interactive/magics.html"><code>%%script</code></a> that allows you to execute scripts in a cell. The difference here is that SoS process passed script with string interpolation before executing it.</p>

<p>You can also input one or more SoS steps and run the workflow</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="n">In</span> <span class="p">[</span><span class="mi">10</span><span class="p">]:</span> <span class="o">%%</span><span class="n">sos</span>
   <span class="o">...</span><span class="p">:</span> <span class="n">resource</span>   <span class="o">=</span> <span class="s">&#39;~/resources&#39;</span>
   <span class="o">...</span><span class="p">:</span> <span class="n">ref_genome</span> <span class="o">=</span> <span class="s">&#39;${resource}/hg19&#39;</span>
   <span class="o">...</span><span class="p">:</span> <span class="n">parameter</span><span class="p">:</span> <span class="n">rep</span><span class="o">=</span><span class="mi">5</span>
   <span class="o">...</span><span class="p">:</span> </code></pre></figure>

<p>Here a SoS script with a parameter section is executed, but without parameter. The <code>%%sos</code> cell magic allows you to pass command line arguments after <code>%%sos</code>, for example,</p>

<pre><code class="language-python">In [11]: %%sos --rep 3
resource   = '~/resources'
ref_genome = '${resource}/hg19'
parameter: rep=5
   ...: 

In [12]: sos rep
Out[12]: 3
</code></pre>

<h3 id="magic-sospaste">Magic <code>sospaste</code></h3>

<p>You will soon notice a problem with the cell magic <code>%%sos</code> in that it does not accept blank new lines, which is problematic for large piece of code. In this case, you can use line magic <code>%sospaste</code> to read the content directly from clipboard. Using the same example, you can select the text (with newline), and run</p>

<pre><code class="language-python">In [5]: sospaste --rep 4
resource   = '~/resources'
ref_genome = '${resource}/hg19'

parameter: rep = 5
## -- End pasted text --
</code></pre>

<p>This is the most convenient way to execute pieces of SoS script in iPython and is used most frequently.</p>

<h3 id="magics-sosget-and-sosput">Magics <code>sosget</code> and <code>sosput</code></h3>

<p>Magics <code>sosget</code> and <code>sosput</code> are used to exchange variables between ipython and sos. For example, if you defined a variable in SoS and would like to use it in ipython, you can use magic <code>%sosget var1 var2</code> to copy the value of <code>var1</code> and <code>var2</code> from SoS dictionary to ipython namespace. Similarly, <code>%sosput</code> copies variables from the ipython namespace to SoS dictionary. For example,</p>

<pre><code>In [1]: %sos a = 20

In [2]: %sos b = 'a**2 = ${a**2}'

In [3]: %sosget a b

In [4]: a
Out[4]: 20

In [5]: b
Out[5]: 'a**2 = 400'

In [6]: b = 'something else'

In [7]: %sosput b

In [8]: %sosdict
Out[8]: {'a': 20, 'b': 'something else'}
</code></pre>

<h3 id="magic-sosset">Magic <code>sosset</code></h3>

<p>If you are tired of entering certain SoS options after magic <code>%%sos</code> or <code>%sospaste</code>, you can set these options persistently using magic <code>sosset</code>. For example, if you set</p>

<pre><code>sosset -v 3
</code></pre>

<p>SoS will be running in a more verbose mode and print more messages. Other acceptable parameters include</p>

<ul>
  <li><code>-c CONFIG</code>: read from configuration file</li>
  <li><code>-f</code>: force execution, regardless of signature</li>
  <li><code>-F</code>: construct signature from existing files</li>
</ul>

<h2 id="a-complete-example">A complete example</h2>

<p>Suppose you are working on a script</p>

<pre><code class="language-python">resource   = '~/resources'
ref_genome = '${resource}/hg19'

parameter: rep = 5

[1]
print(rep)
seq = range(rep)
input:  for_each='seq'

python:
  import time
  print('sleep {} seconds.'.format(_seq))
  time.sleep(_seq)
</code></pre>

<p>To check the global definition, you can copy and paste the definitions to ipython as</p>

<pre><code class="language-python">In [1]: %%sos
   ...: resource   = '~/resources'
   ...: ref_genome = '${resource}/hg19'
   ...: 

</code></pre>

<p>The statements are executed and you can check the result using</p>

<pre><code class="language-python">In [2]: sos ref_genome
Out[2]: '~/resources/hg19'
</code></pre>

<p>Now we select the global definitions and execute them</p>

<pre><code class="language-python">In [3]: sospaste
resource   = '~/resources'
ref_genome = '${resource}/hg19'

parameter: rep = 5
## -- End pasted text --

In [4]: sos rep
Out[4]: 5
</code></pre>

<p>We can see the default parameter is passed. If you want, you can set the parameter to 10,</p>

<pre><code class="language-python">In [5]: sospaste --rep 10
resource   = '~/resources'
ref_genome = '${resource}/hg19'

parameter: rep = 5
## -- End pasted text --

In [6]: sos rep
Out[6]: 10
</code></pre>

<p>Now, let us continue, select the next step, run <code>sospaste</code> and ….  a big block of errors!</p>

<pre><code class="language-python">In [7]: sospaste 

[1]
print(rep)
seq = range(rep)
input:  for_each='seq'

python:
  import time
  print('sleep {} seconds.'.format(_seq))
  time.sleep(_seq)
## -- End pasted text --
INFO: Execute default_1: 
10
INFO: input:   []
Traceback (most recent call last):
  File "/var/folders/ys/gnzk0qbx5wbdgm531v82xxljv5yqy8/T/tmpm5g6f7k7.py", line 2, in &lt;module&gt;
    print('sleep {} seconds.'.format(_seq))
NameError: name '_seq' is not defined

# IGNORED

RuntimeError: Failed to execute process
"python(r'''import time
print('sleep {} seconds.'.format(_seq))
time.sleep(_seq)
''')
"
Failed to execute script. The script is saved to .sos/default_1.py. Please use command "python .sos/default_1.py" under /Users/bpeng1/SOS to test it.
</code></pre>

<p>Here 10 is printed so <code>rep</code> is valid. This means the <code>rep</code> we set last time is available and correct, then what might be the problem? Let us see what is saved in <code>.sos/default_1.py</code> using a bit magic of ipython</p>

<pre><code class="language-python">In [8]: cat .sos/default_1.py
import time
print('sleep {} seconds.'.format(_seq))
time.sleep(_seq)
</code></pre>

<p>The error seems to be obvious, we need to use <code>${_rep}</code> for the value to be passed through string interpolation. Let us make some changes to the script, selct and run <code>sospaste</code> again:</p>

<pre><code class="language-python">In [10]: sospaste

[1]
print(rep)
seq = range(rep)
input:  for_each='seq'

python:
  import time
  print('sleep ${_seq} seconds.')
  time.sleep(${_seq})
## -- End pasted text --
INFO: Execute default_1: 
10
INFO: input:   []
sleep 0 seconds.
sleep 1 seconds.
sleep 2 seconds.
sleep 3 seconds.
sleep 4 seconds.
sleep 5 seconds.
sleep 6 seconds.
sleep 7 seconds.
sleep 8 seconds.
sleep 9 seconds.
</code></pre>

<p>Using this method, you can execute your SoS script step by step and make sure everything works, before you execute it from a command line.</p>


                            </div>
                            <button type="button" class="btn btn-default" data-dismiss="modal"><i class="fa fa-times"></i> Close</button>
                        </div>
                    </div> 
                    <div class="footer-wrapper">
                      <footer role="contentinfo">
                        <span>&copy; 2016 Bo Peng.</span>
                      </footer>
                    </div>
                </div>
            </div>
        </div>
    </div> 


   <div class="portfolio-modal modal fade" id="SoS-Notebook-Using-Jupyter" tabindex="-1" role="dialog" aria-hidden="true">
        <div class="modal-content">
            <div class="close-modal" data-dismiss="modal">
                <div class="lr">
                    <div class="rl">
                    </div>
                </div>
            </div>
            <div class="container">
                <div class="row">
                 <div class="col-lg-8 col-lg-offset-2">
                        <h1>3.SoS-Notebook-Using-Jupyter</h1>
                        <div class="modal-body">
                         
                            <hr class="star-primary">
                            <p>Jupyter notebook is a web application that allows you to create documents with live code and results. The underlying interpreters of the cells are called <strong>kernel</strong>. For example, you can use a <code>python</code> kernel to interpret python expressions, or a <code>ir</code> kernel to interpret <code>R</code> commands. You can use sos within a <code>python</code> kernel by loading the <code>sos_magic</code> ipython extension, or use a <code>sos</code> kernel to create a complete SoS environment. In addition to the ability to execute SoS workflows within Jupyter, the SoS kernel allows you to start subkernels (e.g. <code>ir</code>) and switch between the subkernels in the same notebook. That allows you to freely mix code in different languages in the same notebook.</p>

<h2 id="installing-the-sos-kernel">Installing the SoS kernel</h2>

<p>The SoS installation process will install a SoS kernel by default. To verify if you have the sos kernel installed, type</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="err">$</span> <span class="n">jupyter</span> <span class="n">kernelspec</span> <span class="nb">list</span>
<span class="n">Available</span> <span class="n">kernels</span><span class="p">:</span>
  <span class="n">python3</span>    <span class="o">/</span><span class="n">Users</span><span class="o">/</span><span class="n">bpeng1</span><span class="o">/</span><span class="nb">bin</span><span class="o">/</span><span class="n">anaconda</span><span class="o">/</span><span class="n">lib</span><span class="o">/</span><span class="n">python3</span><span class="o">.</span><span class="mi">5</span><span class="o">/</span><span class="n">site</span><span class="o">-</span><span class="n">packages</span><span class="o">/</span><span class="n">ipykernel</span><span class="o">/</span><span class="n">resources</span>
  <span class="n">sos</span>        <span class="o">/</span><span class="n">usr</span><span class="o">/</span><span class="n">local</span><span class="o">/</span><span class="n">share</span><span class="o">/</span><span class="n">jupyter</span><span class="o">/</span><span class="n">kernels</span><span class="o">/</span><span class="n">sos</span></code></pre></figure>

<p>and check if you have <code>sos</code> listed as one of the kernels. If not, try to re-install SoS using <code>pip3 install sos --upgrade</code> or clone SoS locally and run <code>python setup.py install</code>. At the end of the installation process, you should see</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="n">IPython</span> <span class="n">kernel</span> <span class="n">named</span> <span class="s">&quot;sos&quot;</span> <span class="ow">is</span> <span class="n">installed</span><span class="o">.</span></code></pre></figure>

<p>If the kernel is installed, fire Jupyter using command</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="err">$</span> <span class="n">jupyter</span> <span class="n">notebook</span></code></pre></figure>

<p>from the <code>New</code> button to the top right corner, select <code>SoS</code>, and create a SoS notebook.</p>

<p>SoS uses <a href="http://www.imagemagick.org/script/index.php">imagemagick</a> and python package <a href="http://docs.wand-py.org/en/0.4.2/"><code>wand</code></a> to preview output files so it is recommended that you install imagemagick.</p>

<h2 id="convert-a-sos-workflow-to-jupyter-notebook">Convert a SoS workflow to Jupyter notebook</h2>

<p>You can convert an existing SoS script to the <code>.ipynb</code> format using command</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="err">$</span> <span class="n">sos</span> <span class="n">convert</span> <span class="n">myscript</span><span class="o">.</span><span class="n">sos</span> <span class="o">--</span><span class="n">notebook</span> <span class="n">myscript</span><span class="o">.</span><span class="n">ipynb</span></code></pre></figure>

<p>and open the resulting notebook from the web interface. The converter uses lines that starts with <code>%cell</code> to split SoS a script into cells in Jupyter notebook.</p>

<h2 id="using-the-sos-kernel">Using the SoS kernel</h2>

<p>You can enter any SoS expression and statements in a Jupyter notebook but each cell needs to be a valid SoS script. That is to say,</p>

<ul>
  <li>You can enter any Python expression and statements. For example, you can write pieces of SoS code</li>
</ul>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos">    <span class="n">res_path</span> <span class="o">=</span> <span class="s">&#39;/path/to/resource&#39;</span>
    <span class="n">ref_genome</span> <span class="o">=</span> <span class="s">&#39;${res_path}/hg19&#39;</span>
    </code></pre></figure>

<pre><code>and evaluate them.
</code></pre>

<ul>
  <li>You can enter any SoS actions in function or script format. For example, you can execute a shell script using</li>
</ul>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos">    <span class="n">run</span><span class="p">:</span>
        <span class="n">echo</span> <span class="n">Reference</span> <span class="n">genome</span> <span class="ow">is</span> <span class="n">located</span> <span class="n">at</span> <span class="err">$</span><span class="p">{</span><span class="n">ref_genome</span><span class="p">}</span>
  </code></pre></figure>

<ul>
  <li>
    <p>You can create markdown cells to document your workflow. These cells are not evaluated by SoS so string interpolations are not available to these cells.</p>
  </li>
  <li>
    <p>Enter complete SoS steps with or without headers. Section headers are usually ignored because the cells are executed interactively regardless of the specified step indexes.</p>
  </li>
</ul>

<h2 id="sos-magics">SOS Magics</h2>

<p>In addition to SoS statements, you can use a few SoS magics in Jupyter notebook. They are similar to ipython sos magics but the syntax are a bit different. More specifically,</p>

<ul>
  <li>
    <p><code>%dict [reset|keys|all]</code></p>

    <p>The <code>%dict</code> magic lists or rests the content of SoS dict, using syntax</p>
  </li>
</ul>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos">    <span class="o">%</span><span class="nb">dict</span> <span class="p">[</span><span class="o">-</span><span class="n">a</span><span class="o">|-</span><span class="nb">all</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">k</span><span class="o">|--</span><span class="n">keys</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">r</span><span class="o">|--</span><span class="n">reset</span><span class="p">]</span> <span class="p">[</span><span class="n">var1</span><span class="p">]</span> <span class="p">[</span><span class="n">var2</span><span class="p">]</span> <span class="o">...</span>
    </code></pre></figure>

<pre><code>where

* `var1`, `var2` etc are name of variables. All variables will be displayed if no variable is specified.
* `-a|-all`: list all dictionary keys, including SoS functions and variables.
* `-k|--keys`: list only keys, not their values
* `-r|--reset`: reset the dictionary to its original content (with only SoS internal values)
</code></pre>

<ul>
  <li>
    <p><code>%run [command-options] [workflow-options]</code></p>

    <p>The <code>%sos</code> magic allows you to specify SoS options such as <code>-v</code> (verbosity), <code>-j</code> (max number of jobs), and workflow options as defined by <code>parameter:</code> keyword.</p>
  </li>
  <li>
    <p><code>%paste [command-options] [workflow-options]</code></p>

    <p>This magic is used only for SoS kernel on a console (e.g. <code>ipython qtconsole --kernel sos</code>) where you can grab and execute content from the clipboard.</p>
  </li>
  <li>
    <p><code>%set [command-options] [workflow-options]</code></p>

    <p>The <code>%set</code> magic sets a persistent sos options so you do not have to enter them each time after <code>%run</code> or <code>%paste</code>. For example, if you set <code>%set -v3</code>, you can execute all cells in the notebook at verbosity level 3 (<code>DEBUG</code>).</p>
  </li>
  <li>
    <p><code>%matplotlib [GUI]</code></p>

    <p>Similar to ipython’s <code>matplotlib</code> magic, the <code>%matplotlib inline</code> magic allows the display of matplotlib figures inline in Jupyter notebook or qtconsole.</p>
  </li>
  <li>
    <p><code>%use kernel</code> 
 This magic starts (or switches to) an alternative jupyter kernel to which expanded code (with string interpolation) will be sent. Kernel set by <code>%use kernel</code> will be persistent, affecting all subsequent input, including text pasted using magic <code>%paste</code>.</p>

    <p><code>kernel</code> can be any <a href="https://github.com/ipython/ipython/wiki/IPython-kernels-for-other-languages">Jupyter supported kernels</a>. In particular, you can use <code>ir</code> for R kernel (alias <code>R</code> is also acceptable), <code>python3</code> for python3, <code>iperl</code> for perl, <code>bash</code> for bash. Please refer to <a href="https://github.com/IRkernel/IRkernel">IRKernel</a> for instructions on how to install the <code>ir</code> kernel for <code>R</code>.</p>

    <p><code>%use</code> magic accepts parameters <code>--in (-i)</code> and <code>--out (-o)</code>, which are variables that will be transferred to the subkernel after starting (or switching to) the subkernel, and variables that will be transferred back to SoS before magic <code>%use sos</code>. More specifically,</p>
  </li>
</ul>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos">  <span class="o">%</span><span class="n">use</span> <span class="n">R</span> <span class="o">-</span><span class="n">i</span> <span class="n">var1</span> <span class="n">var2</span> <span class="o">-</span><span class="n">o</span> <span class="n">var3</span>
  <span class="n">do_something</span>
  <span class="o">---</span> <span class="p">(</span><span class="n">another</span> <span class="n">cell</span><span class="p">)</span>
  <span class="o">%</span><span class="n">use</span> <span class="n">sos</span></code></pre></figure>

<p>is equivalent to</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos">  <span class="o">%</span><span class="n">use</span> <span class="n">R</span>
  <span class="o">%</span><span class="n">get</span> <span class="n">var1</span> <span class="n">var2</span>
  <span class="n">do_something</span>
  <span class="o">---</span> <span class="p">(</span><span class="n">another</span> <span class="n">cell</span><span class="p">)</span>
  <span class="o">%</span><span class="n">put</span> <span class="n">var3</span>
  <span class="o">%</span><span class="n">use</span> <span class="n">sos</span></code></pre></figure>

<ul>
  <li>
    <p><code>%with kernel</code></p>

    <p><code>%with kernel</code> also starts (or switches to) a subkernel, but kernel specified by <code>%with kernel</code> will be reset as soon as the cell is executed. This magic is therefore suitable for a quick errand to another kernel. This magic also accepts options <code>--in</code> (or <code>-i</code>) and <code>--out</code> (or <code>-o</code>), so you could use</p>
  </li>
</ul>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos">  <span class="o">%</span><span class="k">with</span> <span class="n">R</span> <span class="o">-</span><span class="n">i</span> <span class="n">n</span> <span class="o">-</span><span class="n">o</span> <span class="n">rn</span>
  <span class="n">rn</span> <span class="o">&lt;-</span> <span class="n">rnorm</span><span class="p">(</span><span class="n">n</span><span class="p">)</span></code></pre></figure>

<p>to get a list of normally distributed numbers using R’s <code>rnorm</code> function.</p>

<ul>
  <li>
    <p><code>%restart kernel</code></p>

    <p><code>%restart</code> specified kernel without affecting the current working kernel.</p>
  </li>
  <li>
    <p><code>%put</code> (<code>ir</code> and <code>python</code> kernels only)</p>

    <p>Magics <code>%put</code> are used to return results from subkernel to SoS kernel when the current kernel is a subkernel started by magic <code>%use</code>. Only the <code>ir</code> and <code>python</code> kernels are supported now. Whereas there is virtually no limit on exchangeable datatypes between <code>SoS</code> and <code>python</code> kernels, there are no perfect map between Python (SoS) and <code>R</code> datatypes so SoS tries to find the best match between data types. More specifically,</p>

    <table>
      <thead>
        <tr>
          <th>R</th>
          <th>length (n)</th>
          <th>Python</th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>NULL</td>
          <td> </td>
          <td>None</td>
        </tr>
        <tr>
          <td>logical</td>
          <td>1</td>
          <td>boolean</td>
        </tr>
        <tr>
          <td>integer</td>
          <td>1</td>
          <td>integer</td>
        </tr>
        <tr>
          <td>numeric</td>
          <td>1</td>
          <td>double</td>
        </tr>
        <tr>
          <td>character</td>
          <td>1</td>
          <td>unicode</td>
        </tr>
        <tr>
          <td>logical</td>
          <td>n &gt; 1</td>
          <td>list</td>
        </tr>
        <tr>
          <td>integer</td>
          <td>n &gt; 1</td>
          <td>list</td>
        </tr>
        <tr>
          <td>numeric</td>
          <td>n &gt; 1</td>
          <td>list</td>
        </tr>
        <tr>
          <td>character</td>
          <td>n &gt; 1</td>
          <td>list</td>
        </tr>
        <tr>
          <td>list without names</td>
          <td>n &gt; 0</td>
          <td>list</td>
        </tr>
        <tr>
          <td>list with names</td>
          <td>n &gt; 0</td>
          <td>dict</td>
        </tr>
        <tr>
          <td>matrix</td>
          <td>n &gt; 0</td>
          <td>numpy.array</td>
        </tr>
        <tr>
          <td>data.frame</td>
          <td>n &gt; 0</td>
          <td>DataFrame</td>
        </tr>
      </tbody>
    </table>
  </li>
  <li>
    <p><code>%get</code> (<code>ir</code> and <code>python</code> kernels only)</p>

    <p>Magics <code>%get</code> retrieve variables from the SoS to the current subkernel started by magic <code>%use</code>. Only the <code>ir</code> and <code>python</code> kernels are supported now. SoS tries to use the best matching data type for the conversion. More specifically,</p>

    <table>
      <thead>
        <tr>
          <th>Python</th>
          <th>cond</th>
          <th>R</th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>None</td>
          <td> </td>
          <td>NULL</td>
        </tr>
        <tr>
          <td>boolean</td>
          <td> </td>
          <td>logical</td>
        </tr>
        <tr>
          <td>integer</td>
          <td> </td>
          <td>integer</td>
        </tr>
        <tr>
          <td>float</td>
          <td> </td>
          <td>float</td>
        </tr>
        <tr>
          <td>str</td>
          <td> </td>
          <td>character</td>
        </tr>
        <tr>
          <td>Sequence (list, tuple, …)</td>
          <td>homogenous type</td>
          <td>c()</td>
        </tr>
        <tr>
          <td>Sequence (list, tuple, …)</td>
          <td>multiple types</td>
          <td>list</td>
        </tr>
        <tr>
          <td>set</td>
          <td> </td>
          <td>list</td>
        </tr>
        <tr>
          <td>dictionary</td>
          <td> </td>
          <td>list with names</td>
        </tr>
        <tr>
          <td>numpy.ndarray</td>
          <td> </td>
          <td>c() array</td>
        </tr>
        <tr>
          <td>numpy.matrix</td>
          <td> </td>
          <td>matrix</td>
        </tr>
        <tr>
          <td>pandas.DataFrame</td>
          <td> </td>
          <td>R data.frame</td>
        </tr>
      </tbody>
    </table>

    <p>Python objects in other datatypes are transferred as string <code>"Unsupported datatype"</code>.</p>
  </li>
  <li>
    <p><code>!any-shell-command</code> such as <code>!ls</code> and <code>!pwd</code></p>

    <p>If any other command is entered after <code>!</code>, sos will treat the rest of the line as a shell command and execute it. Only single-line commands are supported. String interpolation is supported. Most commands are executed in a separate process and have no impact on SoS, except for command <code>!cd</code> that changes the current working directory of SoS.</p>
  </li>
</ul>

<h2 id="convert-ipynb-files-to-sos">Convert <code>.ipynb</code> files to <code>.sos</code></h2>

<p>You can save the notebook to a SoS script using <code>File</code> -&gt; <code>Download As</code> -&gt; <code>SoS</code>, or using command</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="err">$</span> <span class="n">sos</span> <span class="n">convert</span> <span class="n">myscript</span><span class="o">.</span><span class="n">ipynb</span> <span class="o">--</span><span class="n">sos</span> <span class="n">myscript</span><span class="o">.</span><span class="n">sos</span></code></pre></figure>

<p>The conversion process will strip all results (which can be useful if you would like to version control your notebook without results) but keep cell information and SoS magics. However, <strong>SoS files converted from Jupyter notebook cannot be executed by the <code>sos</code> command if they contain SoS magics such as <code>%use R</code></strong> because of non-interactive execution of SoS steps and more rigorous structure required for SoS scripts. You will generally need to</p>

<ol>
  <li>Remove <code>%</code> magic from the notebook and replace cells with</li>
</ol>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos">    <span class="o">%</span><span class="k">with</span> <span class="n">R</span>
      <span class="n">R</span> <span class="n">code</span>
 </code></pre></figure>

<pre><code>to
</code></pre>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos">    <span class="n">R</span><span class="p">:</span>
      <span class="n">R</span> <span class="n">code</span></code></pre></figure>

<pre><code>The latter quires the `R code` to be self-contained and does not require definitions from other cells.
</code></pre>

<ol>
  <li>Add section headers such as <code>[10]</code>, and <code>[human_10]</code> to the SoS steps if they are executed separately.</li>
</ol>

<p>These can be done either in Jupyter notebook or a text editor. Note that Jupyter notebooks converted from a working <code>.sos</code> file already have these features so they will work just fine if you convert back from Jupyter notebook, and you can keep your Jupyter notebook executable in batch mode by removing <code>%</code> magics after using them for debug, and use section headers for different steps.</p>

                            </div>
                            <button type="button" class="btn btn-default" data-dismiss="modal"><i class="fa fa-times"></i> Close</button>
                        </div>
                    </div> 
                    <div class="footer-wrapper">
                      <footer role="contentinfo">
                        <span>&copy; 2016 Bo Peng.</span>
                      </footer>
                    </div>
                </div>
            </div>
        </div>
    </div> 


   <div class="portfolio-modal modal fade" id="Using-Spyder-as-SoS-IDE" tabindex="-1" role="dialog" aria-hidden="true">
        <div class="modal-content">
            <div class="close-modal" data-dismiss="modal">
                <div class="lr">
                    <div class="rl">
                    </div>
                </div>
            </div>
            <div class="container">
                <div class="row">
                 <div class="col-lg-8 col-lg-offset-2">
                        <h1>4.Using-Spyder-as-SoS-IDE</h1>
                        <div class="modal-body">
                         
                            <hr class="star-primary">
                            <h2 id="using-sos-with-spyder">Using SoS with spyder</h2>

<table>
  <tbody>
    <tr>
      <td>[spyder</td>
      <td>https://pythonhosted.org/spyder/) is a Python GUI that works well with ipython, and therefore SoS if you configure it properly. Spyder should be readily available if you use Anaconda python, or you can install spyder according to instructions on the <a href="https://pythonhosted.org/spyder/">spyder website</a>.</td>
    </tr>
  </tbody>
</table>

<p>Disclaimer: Spyder does not support third-party kernels and does not yet support <code>.sos</code> files. The SoS installer tries to patch Spyder so that it can recognize <code>.sos</code> files and the <code>%cell</code> syntax. These are supposed to change after Spyder provides native support for SoS in its next major release.</p>

<p>Therefore two ways to use spyder with SoS:</p>

<ol>
  <li>
    <p>Connect spyder to an existing qtconsole with sos kernel</p>

    <ul>
      <li>Start a qtconsole with sos kernel using command <code>jupyter qtconsole --kernel sos</code></li>
      <li>Record connection file id from command line, or by running <code>%connect_info</code> from the console if you cannot find it.</li>
      <li>Start <code>spyder</code>, select <code>consoles</code> -&gt; <code>Connect an existing kernel</code>. Put connection id and connect.</li>
    </ul>

    <p>It is certainly possible to set up a remote Jupyter server and connect to a remote SoS kernel but this usage is beyond the scope of this tutorial.</p>
  </li>
  <li>
    <p>Make sos your default kernel by adding</p>
  </li>
</ol>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos">    <span class="n">c</span><span class="o">.</span><span class="n">IPKernelApp</span><span class="o">.</span><span class="n">kernel_class</span> <span class="o">=</span>  <span class="s">&#39;pysos.kernel.SoS_Kernel&#39;</span>
    </code></pre></figure>

<pre><code>to `~/.ipython/profile_default/ipython_config.py` . You can then start `spyder` and use spyder as usual.
</code></pre>


                            </div>
                            <button type="button" class="btn btn-default" data-dismiss="modal"><i class="fa fa-times"></i> Close</button>
                        </div>
                    </div> 
                    <div class="footer-wrapper">
                      <footer role="contentinfo">
                        <span>&copy; 2016 Bo Peng.</span>
                      </footer>
                    </div>
                </div>
            </div>
        </div>
    </div> 


   <div class="portfolio-modal modal fade" id="A-Complete-Example" tabindex="-1" role="dialog" aria-hidden="true">
        <div class="modal-content">
            <div class="close-modal" data-dismiss="modal">
                <div class="lr">
                    <div class="rl">
                    </div>
                </div>
            </div>
            <div class="container">
                <div class="row">
                 <div class="col-lg-8 col-lg-offset-2">
                        <h1>5.A-Complete-Example</h1>
                        <div class="modal-body">
                         
                            <hr class="star-primary">
                            <h2 id="a-rna-seq-data-processing-workflow">A RNA Seq data processing workflow</h2>

<p>If you are curious about how to write large workflows using SoS or just how a long SoS script looks like, here is a RNA Seq data processing workflow written in SoS. The workflow consists of many steps and was written in a mixed style, namely a spine of forward-moving steps and several auxiliary steps that supply required information. The DAG of the workflow looks like</p>

<p><img src="/SOS/media/example_dag.png" alt="example_dag.png" class="img-responsive" /></p>

<p>although this DAG does not include many auxiliary steps (e.g. those that download required resources) because these steps were called once and would not be called once the resources are available. The DAG is generated using command</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="n">sos</span> <span class="n">prepare</span> <span class="n">RNASeq_hg10</span><span class="o">.</span><span class="n">sos</span>
<span class="n">cat</span> <span class="o">.</span><span class="n">sos</span><span class="o">/</span><span class="n">human_hg19</span><span class="o">.</span><span class="n">dot</span> <span class="o">|</span> <span class="n">dot</span> <span class="o">-</span><span class="n">T</span> <span class="n">png</span> <span class="o">&gt;</span> <span class="n">example_dag</span><span class="o">.</span><span class="n">png</span></code></pre></figure>

<p>This tutorial will divide the workflow into several parts and briefly describe each of them.</p>

<h3 id="header">Header</h3>

<p>The header of script contains description of the workflow.</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="c">#!/usr/bin/env sos-runner</span>
<span class="c">#fileformat=SOS1.0</span>

<span class="c"># Workflow for a comprehensive analysis of paired-end RNA seq data</span>
<span class="c"># that performs read-level quality assessment, alignment, alignment quality</span>
<span class="c"># assessment, summary statistics, fusion detection and filtering, gene and exon</span>
<span class="c"># level expression counts, and variant calling. The input of this pipeline</span>
<span class="c"># should be a list of fastq or fastq.gz files with paired reads differentiated</span>
<span class="c"># by _R1_ and _R2_ in filenames. &lt;p&gt;The output of the pipeline should be a</span>
<span class="c"># directory to which all results will be written. The directory will be </span>
<span class="c"># created if it does not exist. Note that it is highly recommended that</span>
<span class="c"># you execute the pipeline with a subset of samples (using option --sampling</span>
<span class="c"># 1000000) to test the pipeline before you apply it to the full dataset. </span>
<span class="c"># Network connection will be needed for the first run (to download resources)</span>
<span class="c"># but not for subsequent runs. For details of the pipeline including the</span>
<span class="c"># accepted parameters, please run command &quot;vtools show pipeline RNASeq_hg19_v3&quot;.</span>

<span class="c"># human_hg19</span>
<span class="c"># </span>
<span class="c"># This pipeline uses the hg19 version of the reference</span>
<span class="c"># genome, genes downloaded from Illumina iGenome, and existing variants from</span>
<span class="c"># the GATK resource bundle. This pipeline uses &lt;ul&gt;</span>
<span class="c">#   &lt;li&gt;tophat 2.0.13 </span>
<span class="c">#   &lt;li&gt;bowtie 1.1.1</span>
<span class="c">#   &lt;li&gt;samtools 0.1.19</span>
<span class="c">#   &lt;li&gt;picard 1.82</span>
<span class="c">#   &lt;li&gt;GATK 3.3</span>
<span class="c">#   &lt;li&gt;fastqc 0.11.2</span>
<span class="c">#   &lt;li&gt;BEDTools 2.17.0</span>
<span class="c">#   &lt;li&gt;RSeQC 2.4, and</span>
<span class="c">#   &lt;li&gt;Oncofuse 1.0.7</span>
<span class="c">#   &lt;/ul&gt;</span>
<span class="c"># Other versions of the tools can be used if you execute the pipeline with option</span>
<span class="c"># --strict_version False, but the pipeline might or might not work as expected. </span>
<span class="c"># Please download all these tools and make them available to the pipeline (set $PATH),</span>
<span class="c"># and start the pipeline using command &quot;vtools execute /path/to/this_file options&quot;.</span>
<span class="c"># Noted that the precompiled version of tophat 2.0.13 can fail under mac so it is</span>
<span class="c"># recommended that you compile tophat2 from source.</span></code></pre></figure>

<h3 id="inclusion-of-auxiliary-steps">inclusion of auxiliary steps</h3>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="o">%</span><span class="kn">from</span> <span class="nn">install_ngs</span> <span class="nn">include</span> <span class="o">*</span></code></pre></figure>

<h3 id="global-definitions">Global definitions</h3>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isdir</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">expanduser</span><span class="p">(</span><span class="s">&#39;~/rsrch1/variant_tools_resource/pipeline_resource&#39;</span><span class="p">)):</span>
    <span class="n">resource_dir</span>  <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">expanduser</span><span class="p">(</span><span class="s">&#39;~/rsrch1/variant_tools_resource/pipeline_resource&#39;</span><span class="p">)</span>
<span class="k">elif</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isdir</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">expanduser</span><span class="p">(</span><span class="s">&#39;~/.variant_tools/pipeline_resource&#39;</span><span class="p">)):</span>
    <span class="n">resource_dir</span>  <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">expanduser</span><span class="p">(</span><span class="s">&#39;~/.variant_tools/pipeline_resource&#39;</span><span class="p">)</span>
<span class="kn">else:</span>
    <span class="n">fail_if</span><span class="p">(</span><span class="bp">True</span><span class="p">,</span> <span class="s">&#39;Unable to locate a resource directory&#39;</span><span class="p">)</span>

<span class="c"># Path to picard jar files</span>
<span class="kn">parameter:</span> <span class="n">picard_path</span> <span class="o">=</span> <span class="s">&#39;~/bin/Picard&#39;</span>

<span class="c"># Path to GenomeAnalysisTK.jar</span>
<span class="kn">parameter:</span> <span class="n">gatk_path</span> <span class="o">=</span> <span class="s">&#39;~/bin/GATK&#39;</span>

<span class="c"># Path to Oncofuse.jar</span>
<span class="kn">parameter:</span> <span class="n">oncofuse_path</span> <span class="o">=</span> <span class="s">&#39;~/bin/Oncofuse&#39;</span>

<span class="c"># input fastq files</span>
<span class="kn">parameter:</span> <span class="n">fastq_files</span> <span class="o">=</span> <span class="nb">list</span>

<span class="n">fail_if</span><span class="p">(</span><span class="nb">len</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">fastq_files</span> <span class="k">if</span> <span class="s">&#39;_R1_&#39;</span> <span class="ow">in</span> <span class="n">x</span><span class="p">])</span> <span class="o">==</span> <span class="mi">0</span>  \
    <span class="ow">or</span> <span class="nb">len</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">fastq_files</span> <span class="k">if</span> <span class="s">&#39;_R1_&#39;</span> <span class="ow">in</span> <span class="n">x</span><span class="p">])</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">fastq_files</span> <span class="k">if</span> <span class="s">&#39;_R2_&#39;</span> <span class="ow">in</span> <span class="n">x</span><span class="p">]),</span>
    <span class="s">&#39;Input file names are not paired fastq files differentiated by _R1_ and _R2_ in filename.&#39;</span><span class="p">)</span>

<span class="c"># Name of the sample being processed, which will be used for filename</span>
<span class="c"># and read group of aligned reads.</span>
<span class="kn">parameter:</span> <span class="n">samplename</span> <span class="o">=</span> <span class="nb">str</span>

<span class="c"># If set to a positive number, this pipeline will sample specified</span>
<span class="c"># number of reads (pairs) from the input. This is useful to test if the pipeline</span>
<span class="c"># works for your environment, download all needed resources, created indexes,</span>
<span class="c"># and check the quality of data.</span>
<span class="kn">parameter:</span> <span class="n">sampling</span> <span class="o">=</span> <span class="mi">0</span>


<span class="c"># An optional filename that contains detailed information</span>
<span class="c"># about the sample. The sample sheet should be a tab or comma-delimited </span>
<span class="c"># file with a header, and one or more lines with one of which contain the </span>
<span class="c"># information of the sample. The information from the sample sheet will be</span>
<span class="c"># written to the final report (summary.html).</span>
<span class="kn">parameter:</span> <span class="n">sample_sheet</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>

<span class="c">#</span>
<span class="c"># CONSTANT VALUES: not configurable in this pipeline</span>
<span class="c"># </span>
<span class="n">pipeline_name</span>     <span class="o">=</span> <span class="s">&#39;RNASeq_hg19&#39;</span>
<span class="n">pipeline_version</span>  <span class="o">=</span> <span class="s">&#39;ver1.0 (2016/8/29&#39;</span>
<span class="n">reference_genome</span>  <span class="o">=</span> <span class="s">&#39;hg19&#39;</span>
<span class="n">reference_genes</span>   <span class="o">=</span> <span class="s">&#39;UCSC/Illumina iGenome&#39;</span>
<span class="c"># HG19 version of the resource files</span>
<span class="n">gatk_resource_dir</span> <span class="o">=</span> <span class="s">&#39;${resource_dir}/GATK/2.8/hg19/&#39;</span>
<span class="n">igenome_resource_dir</span> <span class="o">=</span> <span class="s">&#39;${resource_dir}/iGenome&#39;</span>
<span class="n">ncbi_resource_dir</span> <span class="o">=</span> <span class="s">&#39;${resource_dir}/NCBI&#39;</span>

<span class="n">igenome_url</span>       <span class="o">=</span> <span class="s">&#39;ftp://igenome:G3nom3s4u@ussd-ftp.illumina.com/Homo_sapiens/UCSC/hg19/Homo_sapiens_UCSC_hg19.tar.gz&#39;</span>
<span class="n">gene_info_url</span>     <span class="o">=</span> <span class="s">&#39;ftp://ftp.ncbi.nih.gov/gene/DATA/GENE_INFO/Mammalia/Homo_sapiens.gene_info.gz&#39;</span>
<span class="n">reference_dir</span>     <span class="o">=</span> <span class="s">&#39;${igenome_resource_dir}/Homo_sapiens/UCSC/hg19/Sequence&#39;</span>
<span class="n">transcriptome_dir</span> <span class="o">=</span> <span class="s">&#39;${igenome_resource_dir}/Homo_sapiens/UCSC/hg19/Annotation/transcriptome_index&#39;</span>
<span class="n">annotation_dir</span>    <span class="o">=</span> <span class="s">&#39;${igenome_resource_dir}/Homo_sapiens/UCSC/hg19/Annotation&#39;</span>
<span class="n">refgene_txt</span>       <span class="o">=</span> <span class="s">&#39;${annotation_dir}/Genes/refGene.txt&#39;</span>
<span class="n">genes_gtf</span>         <span class="o">=</span> <span class="s">&#39;${annotation_dir}/Genes/genes.gtf&#39;</span>
<span class="c"># generated files&#39;</span>
<span class="n">refgene_bed</span>       <span class="o">=</span> <span class="s">&#39;${annotation_dir}/Genes/refGene.bed&#39;</span>
<span class="n">genes_exon_bed</span>    <span class="o">=</span> <span class="s">&#39;${annotation_dir}/Genes/genes_exon.bed&#39;</span>
<span class="n">genes_with_chr_gtf</span><span class="o">=</span> <span class="s">&#39;${annotation_dir}/Genes/genes_chr.gtf&#39;</span>
<span class="n">genesize_txt</span>      <span class="o">=</span> <span class="s">&#39;${annotation_dir}/Genes/geneSize.txt&#39;</span>
<span class="c"># Download from UCSC&#39;</span>
<span class="n">ensgene_url</span>       <span class="o">=</span> <span class="s">&#39;http://hgdownload.cse.ucsc.edu/goldenpath/hg19/database/ensGene.txt.gz&#39;</span>
<span class="n">ensgene_txt</span>       <span class="o">=</span> <span class="s">&#39;${annotation_dir}/Genes/ensGene.txt&#39;</span>
<span class="n">chromsize_url</span>     <span class="o">=</span> <span class="s">&#39;ftp://hgdownload.cse.ucsc.edu/goldenPath/hg19/database/chromInfo.txt.gz&#39;</span>
<span class="n">chromsize_txt</span>     <span class="o">=</span> <span class="s">&#39;${annotation_dir}/chromsize.txt&#39;</span>
<span class="c"># GATK resource bundle&#39;</span>
<span class="n">gatk_url</span>          <span class="o">=</span> <span class="s">&#39;ftp://gsapubftp-anonymous@ftp.broadinstitute.org/bundle/2.8/hg19/&#39;</span>
<span class="c"># &#39;</span>
<span class="c"># Output directories&#39;</span>
<span class="n">output_dir</span>        <span class="o">=</span> <span class="n">samplename</span>
<span class="n">temp_dir</span>          <span class="o">=</span> <span class="s">&#39;${output_dir}/tmp&#39;</span>
<span class="n">alignment_out</span>     <span class="o">=</span> <span class="s">&#39;${output_dir}/alignment&#39;</span>
<span class="n">alignmentqc_out</span>   <span class="o">=</span> <span class="s">&#39;${output_dir}/alignment_qc&#39;</span>
<span class="n">fusion_out</span>        <span class="o">=</span> <span class="s">&#39;${output_dir}/fusion&#39;</span>
<span class="n">readqc_out</span>        <span class="o">=</span> <span class="s">&#39;${output_dir}/read_qc&#39;</span>
<span class="n">count_out</span>         <span class="o">=</span> <span class="s">&#39;${output_dir}/counts&#39;</span>
<span class="n">variant_out</span>       <span class="o">=</span> <span class="s">&#39;${output_dir}/variants&#39;</span>



<span class="p">[</span><span class="n">human_hg19_100</span><span class="p">:</span> <span class="kp">skip</span><span class="o">=</span><span class="n">sampling</span> <span class="o">==</span> <span class="mi">0</span><span class="p">,</span> <span class="kp">shared</span><span class="o">=</span><span class="s">&#39;fastq_files&#39;</span><span class="p">]</span>
<span class="c"># Select a subset of reads if parameter --sampling is specified. Output will be </span>
<span class="c"># written to the project cache directory.</span>
<span class="kn">input:</span>
    <span class="n">fastq_files</span>

<span class="kn">depends:</span>
    <span class="n">executable</span><span class="p">(</span><span class="s">&#39;gunzip&#39;</span><span class="p">)</span>

<span class="kn">output:</span>
    <span class="s">&#39;${temp_dir}/${samplename}_R1_${sampling}.fastq&#39;</span><span class="p">,</span>
    <span class="s">&#39;${temp_dir}/${samplename}_R2_${sampling}.fastq&#39;</span>

<span class="kn">run:</span>
    <span class="n">gunzip</span> <span class="o">-</span><span class="n">c</span> <span class="err">$</span><span class="p">{</span><span class="s">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">sorted</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">input</span> <span class="k">if</span> <span class="s">&#39;_R1_&#39;</span> <span class="ow">in</span> <span class="n">x</span><span class="p">]))}</span> <span class="o">|</span> <span class="n">head</span> <span class="o">-</span><span class="n">n</span> <span class="err">$</span><span class="p">{</span><span class="mi">4</span><span class="o">*</span><span class="n">sampling</span><span class="p">}</span> <span class="o">&gt;</span> <span class="err">$</span><span class="p">{</span><span class="n">output</span><span class="p">[</span><span class="mi">0</span><span class="p">]}</span>
    <span class="n">gunzip</span> <span class="o">-</span><span class="n">c</span> <span class="err">$</span><span class="p">{</span><span class="s">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">sorted</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">input</span> <span class="k">if</span> <span class="s">&#39;_R2_&#39;</span> <span class="ow">in</span> <span class="n">x</span><span class="p">]))}</span> <span class="o">|</span> <span class="n">head</span> <span class="o">-</span><span class="n">n</span> <span class="err">$</span><span class="p">{</span><span class="mi">4</span><span class="o">*</span><span class="n">sampling</span><span class="p">}</span> <span class="o">&gt;</span> <span class="err">$</span><span class="p">{</span><span class="n">output</span><span class="p">[</span><span class="mi">1</span><span class="p">]}</span>

<span class="n">fastq_files</span> <span class="o">=</span> <span class="n">output</span>


<span class="p">[</span><span class="n">human_hg19_110</span><span class="p">]</span>
<span class="c"># Use fastqc to check the quality of input reads</span>
<span class="c">#</span>
<span class="kn">input:</span>
    <span class="n">fastq_files</span>

<span class="kn">depends:</span>
    <span class="n">executable</span><span class="p">(</span><span class="s">&#39;fastqc&#39;</span><span class="p">),</span>
    <span class="n">executable</span><span class="p">(</span><span class="s">&#39;gunzip&#39;</span><span class="p">)</span>

<span class="kn">output:</span>
    <span class="s">&#39;${readqc_out}/${samplename}_fastqc.html&#39;</span><span class="p">,</span>
    <span class="s">&#39;${readqc_out}/${samplename}_fastqc.zip&#39;</span>

<span class="kn">task:</span>

<span class="kn">run:</span>
    <span class="c"># the input can be original .fast.gz file or extracted </span>
    <span class="err">$</span><span class="p">{</span><span class="s">&quot;gunzip -c&quot;</span> <span class="k">if</span> <span class="nb">input</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s">&quot;.gz&quot;</span><span class="p">)</span> <span class="k">else</span> <span class="s">&quot;cat&quot;</span><span class="p">}</span> <span class="err">$</span><span class="p">{</span><span class="nb">input</span><span class="p">}</span> <span class="o">|</span> <span class="n">fastqc</span> <span class="n">stdin</span> <span class="o">--</span><span class="n">outdir</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="n">readqc_out</span><span class="p">}</span>
    <span class="n">mv</span> <span class="err">$</span><span class="p">{</span><span class="n">readqc_out</span><span class="p">}</span><span class="o">/</span><span class="n">stdin_fastqc</span><span class="o">.</span><span class="n">html</span>  <span class="err">$</span><span class="p">{</span><span class="n">output</span><span class="p">[</span><span class="mi">0</span><span class="p">]}</span>
    <span class="n">mv</span> <span class="err">$</span><span class="p">{</span><span class="n">readqc_out</span><span class="p">}</span><span class="o">/</span><span class="n">stdin_fastqc</span><span class="o">.</span><span class="n">zip</span> <span class="err">$</span><span class="p">{</span><span class="n">output</span><span class="p">[</span><span class="mi">1</span><span class="p">]}</span>

<span class="p">[</span><span class="n">extract_fastqc_figures</span><span class="p">:</span> <span class="kp">provides</span><span class="o">=</span><span class="p">[</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">readqc_out</span><span class="p">,</span> <span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="p">(</span><span class="s">&#39;per_base_quality.png&#39;</span><span class="p">,</span> <span class="s">&#39;per_sequence_gc_content.png&#39;</span><span class="p">,</span> <span class="s">&#39;per_base_sequence_content.png&#39;</span><span class="p">)]]</span>
<span class="kn">input:</span>
    <span class="s">&#39;${readqc_out}/${samplename}_fastqc.zip&#39;</span>

<span class="kn">import</span> <span class="nn">zipfile</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="k">with</span> <span class="n">zipfile</span><span class="o">.</span><span class="n">ZipFile</span><span class="p">(</span><span class="nb">input</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="k">as</span> <span class="n">qczip</span><span class="p">:</span>
    <span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="n">qczip</span><span class="o">.</span><span class="n">namelist</span><span class="p">():</span>
        <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">name</span><span class="p">)</span> <span class="ow">in</span> <span class="p">[</span><span class="s">&#39;per_base_quality.png&#39;</span><span class="p">,</span> <span class="s">&#39;per_sequence_gc_content.png&#39;</span><span class="p">,</span> <span class="s">&#39;per_base_sequence_content.png&#39;</span><span class="p">]:</span>
            <span class="n">qczip</span><span class="o">.</span><span class="n">extract</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="n">readqc_out</span><span class="p">)</span>
            <span class="c"># move the file to top directory</span>
            <span class="n">os</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">readqc_out</span><span class="p">,</span> <span class="n">name</span><span class="p">),</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">readqc_out</span><span class="p">,</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">name</span><span class="p">)))</span>


<span class="p">[</span><span class="n">download_ensgene</span><span class="p">:</span> <span class="kp">provides</span><span class="o">=</span><span class="n">genes_gtf</span><span class="p">]</span>
<span class="kn">download:</span>  <span class="s">&#39;${annotation_dir}/Genes/&#39;</span>
    <span class="err">$</span><span class="p">{</span><span class="n">ensgene_url</span><span class="p">}</span>

<span class="p">[</span><span class="n">chr_gtf</span><span class="p">:</span> <span class="kp">provides</span> <span class="o">=</span> <span class="n">genes_with_chr_gtf</span><span class="p">]</span>
<span class="c"># takes gtf and and chromosome names to gene name to avoid problems</span>
<span class="c"># with genes with the same names on different chromosomes</span>
<span class="kn">input:</span>
    <span class="n">genes_gtf</span>

<span class="kn">output:</span>
    <span class="n">genes_with_chr_gtf</span>

<span class="kn">python:</span>
    <span class="c">#</span>
    <span class="c">#    Input format:</span>
    <span class="c">#</span>
    <span class="c">#        chr1 unknown exon 17233 17368 . - . gene_id &quot;WASH7P&quot;; gene_name &quot;WASH7P&quot;; transcript_id &quot;NR_024540&quot;; tss_id &quot;TSS8151&quot;;</span>
    <span class="c">#</span>
    <span class="c">#    Output format:</span>
    <span class="c">#           chr1 unknown exon 17233 17368 . - . gene_id &quot;WASH7P&quot;; gene_name &quot;WASH7P.chr1&quot;; transcript_id &quot;NR_024540&quot;; tss_id &quot;TSS8151&quot;;</span>
    <span class="c"># </span>
    <span class="kn">import</span> <span class="nn">re</span>
    <span class="kp">pattern</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="s">r&#39;&#39;&#39;(gene_id\s+[&#39;&quot;]*)([\w.-]+)&#39;&#39;&#39;</span><span class="p">)</span>
    <span class="c"># only work on selected chromosomes</span>
    <span class="n">allowed_chromosomes</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([</span><span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">25</span><span class="p">)]</span> <span class="o">+</span> <span class="p">[</span><span class="s">&#39;chr{}&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">25</span><span class="p">)]</span> <span class="o">+</span> \
                <span class="p">[</span><span class="s">&#39;X&#39;</span><span class="p">,</span> <span class="s">&#39;Y&#39;</span><span class="p">,</span> <span class="s">&#39;M&#39;</span><span class="p">,</span> <span class="s">&#39;chrX&#39;</span><span class="p">,</span> <span class="s">&#39;chrY&#39;</span><span class="p">,</span> <span class="s">&#39;chrM&#39;</span><span class="p">])</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="nb">input</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="s">&#39;rU&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">ifile</span><span class="p">,</span> <span class="nb">open</span><span class="p">(</span><span class="n">output</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="s">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">ofile</span><span class="p">:</span>
       <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">ifile</span><span class="p">:</span>
           <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s">&#39;#&#39;</span><span class="p">):</span>
               <span class="k">continue</span>
           <span class="n">ls</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&#39;</span><span class="se">\t</span><span class="s">&#39;</span><span class="p">)</span>
           <span class="k">if</span> <span class="n">ls</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">allowed_chromosomes</span><span class="p">:</span>
               <span class="k">continue</span>
           <span class="n">ofile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="kp">pattern</span><span class="p">,</span> <span class="s">r&#39;\1\2.{}&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">ls</span><span class="p">[</span><span class="mi">0</span><span class="p">]),</span> <span class="n">line</span><span class="p">))</span> 

<span class="p">[</span><span class="n">download_igenome</span><span class="p">:</span> <span class="kp">provides</span><span class="o">=</span><span class="n">refgene_bed</span><span class="p">]</span>
<span class="kn">download:</span> <span class="n">dest_dir</span><span class="o">=</span><span class="n">igenome_resource_dir</span>
    <span class="err">$</span><span class="p">{</span><span class="n">igenome_url</span><span class="p">}</span>

<span class="p">[</span><span class="n">human_hg19_400</span><span class="p">:</span> <span class="kp">shared</span><span class="o">=</span><span class="p">{</span><span class="s">&#39;param&#39;</span><span class="p">:</span> <span class="s">&#39;output[0]&#39;</span><span class="p">}]</span>
<span class="c">#</span>
<span class="c"># Align a subset of reads to obtain two key parameters parameters</span>
<span class="c"># --mate-inner-dist and --mate-std-dev for full scale alignment</span>
<span class="c">#</span>
<span class="kn">input:</span>
    <span class="n">fastq_files</span>

<span class="kn">depends:</span>
    <span class="n">genes_with_chr_gtf</span><span class="p">,</span>
    <span class="n">refgene_bed</span><span class="p">,</span>
    <span class="n">executable</span><span class="p">(</span><span class="s">&#39;tophat2&#39;</span><span class="p">),</span>
    <span class="n">executable</span><span class="p">(</span><span class="s">&#39;bowtie&#39;</span><span class="p">),</span>
    <span class="n">executable</span><span class="p">(</span><span class="s">&#39;inner_distance.py&#39;</span><span class="p">)</span>
    
<span class="kn">output:</span>
    <span class="s">&#39;${temp_dir}/${samplename}.inner_distance_freq.txt&#39;</span>

<span class="n">test_files</span> <span class="o">=</span> <span class="s">&#39;${temp_dir}/${samplename}_R1_test.fastq&#39;</span><span class="p">,</span> <span class="s">&#39;${temp_dir}/${samplename}_R2_test.fastq&#39;</span>

<span class="kn">task:</span>

<span class="kn">run:</span>
    <span class="err">$</span><span class="p">{</span><span class="s">&#39;gunzip -c&#39;</span> <span class="k">if</span> <span class="nb">input</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s">&#39;.gz&#39;</span><span class="p">)</span> <span class="k">else</span> <span class="s">&#39;cat&#39;</span><span class="p">}</span>  <span class="err">$</span><span class="p">{</span><span class="s">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">sorted</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">input</span> <span class="k">if</span> <span class="s">&#39;_R1_&#39;</span> <span class="ow">in</span> <span class="n">x</span><span class="p">]))}</span> \
        <span class="o">|</span> <span class="n">head</span> <span class="o">-</span><span class="n">n</span> <span class="mi">40000</span> <span class="o">&gt;</span> <span class="err">$</span><span class="p">{</span><span class="n">test_files</span><span class="p">[</span><span class="mi">0</span><span class="p">]}</span>

    <span class="err">$</span><span class="p">{</span><span class="s">&#39;gunzip -c&#39;</span> <span class="k">if</span> <span class="nb">input</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s">&#39;.gz&#39;</span><span class="p">)</span> <span class="k">else</span> <span class="s">&#39;cat&#39;</span><span class="p">}</span>  <span class="err">$</span><span class="p">{</span><span class="s">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">sorted</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">input</span> <span class="k">if</span> <span class="s">&#39;_R2_&#39;</span> <span class="ow">in</span> <span class="n">x</span><span class="p">]))}</span> \
        <span class="o">|</span> <span class="n">head</span> <span class="o">-</span><span class="n">n</span> <span class="mi">40000</span> <span class="o">&gt;</span> <span class="err">$</span><span class="p">{</span><span class="n">test_files</span><span class="p">[</span><span class="mi">1</span><span class="p">]}</span>

<span class="kn">run:</span>
    <span class="n">tophat2</span> <span class="o">--</span><span class="n">zpacker</span> <span class="mi">0</span> <span class="o">-</span><span class="n">no</span><span class="o">-</span><span class="n">coverage</span><span class="o">-</span><span class="n">search</span> <span class="o">--</span><span class="n">library</span><span class="o">-</span><span class="nb">type</span> <span class="n">fr</span><span class="o">-</span><span class="n">unstranded</span> \
        <span class="o">--</span><span class="nb">max</span><span class="o">-</span><span class="n">multihits</span> <span class="mi">20</span> <span class="o">--</span><span class="n">prefilter</span><span class="o">-</span><span class="n">multihits</span> <span class="o">--</span><span class="n">num</span><span class="o">-</span><span class="n">threads</span> <span class="mi">8</span> \
        <span class="o">--</span><span class="n">GTF</span> <span class="err">$</span><span class="p">{</span><span class="n">genes_with_chr_gtf</span><span class="p">}</span> <span class="o">--</span><span class="n">segment</span><span class="o">-</span><span class="n">length</span> <span class="mi">25</span> \
        <span class="o">--</span><span class="n">transcriptome</span><span class="o">-</span><span class="n">index</span> <span class="err">$</span><span class="p">{</span><span class="n">transcriptome_dir</span><span class="p">}</span><span class="o">/</span><span class="n">genes</span> <span class="o">--</span><span class="n">bowtie1</span> \
        <span class="o">--</span><span class="n">output</span><span class="o">-</span><span class="nb">dir</span> <span class="err">$</span><span class="p">{</span><span class="n">temp_dir</span><span class="p">}</span> <span class="err">$</span><span class="p">{</span><span class="n">reference_dir</span><span class="p">}</span><span class="o">/</span><span class="n">BowtieIndex</span><span class="o">/</span><span class="n">genome</span> \
        <span class="err">$</span><span class="p">{</span><span class="n">test_files</span><span class="p">[</span><span class="mi">0</span><span class="p">]}</span> <span class="err">$</span><span class="p">{</span><span class="n">test_files</span><span class="p">[</span><span class="mi">1</span><span class="p">]}</span>

<span class="kn">run:</span>
    <span class="n">inner_distance</span><span class="o">.</span><span class="n">py</span> <span class="o">-</span><span class="n">r</span> <span class="err">$</span><span class="p">{</span><span class="n">refgene_bed</span><span class="p">}</span> <span class="o">-</span><span class="n">i</span> <span class="err">$</span><span class="p">{</span><span class="n">temp_dir</span><span class="p">}</span><span class="o">/</span><span class="n">accepted_hits</span><span class="o">.</span><span class="n">bam</span> <span class="o">-</span><span class="n">o</span> <span class="err">$</span><span class="p">{</span><span class="n">temp_dir</span><span class="p">}</span><span class="o">/</span><span class="err">$</span><span class="p">{</span><span class="n">samplename</span><span class="p">}</span>


<span class="p">[</span><span class="n">human_hg19_440</span><span class="p">:</span> <span class="kp">shared</span><span class="o">=</span><span class="p">{</span><span class="s">&#39;accepted_hits&#39;</span><span class="p">:</span><span class="s">&#39;output[0]&#39;</span><span class="p">,</span> <span class="s">&#39;mate_inner_dist&#39;</span><span class="p">:</span><span class="s">&#39;mate_inner_dist&#39;</span><span class="p">,</span> <span class="s">&#39;mate_std_dev&#39;</span><span class="p">:</span><span class="s">&#39;mate_std_dev&#39;</span><span class="p">}]</span>
<span class="c">#Alignment (Tophat2): Running tophat2 with fusion search.</span>
<span class="c">#Alignment Preparation: Calculate mean and standard devision of inner distance and set parameters</span>
<span class="c"># --mate-inner-dist and --mate-std-dev</span>
<span class="kn">import</span> <span class="nn">math</span>
<span class="n">values</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">param</span><span class="p">)</span> <span class="k">as</span> <span class="n">freq</span><span class="p">:</span>
    <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">freq</span><span class="p">:</span>
        <span class="n">s</span><span class="p">,</span> <span class="n">e</span><span class="p">,</span> <span class="n">c</span> <span class="o">=</span> <span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">()]</span>
        <span class="n">values</span><span class="o">.</span><span class="n">extend</span><span class="p">([</span><span class="nb">int</span><span class="p">(</span><span class="n">s</span><span class="o">+</span><span class="n">e</span><span class="p">)</span><span class="o">/</span><span class="mf">2.</span><span class="p">]</span> <span class="o">*</span> <span class="n">c</span><span class="p">)</span>
<span class="k">if</span> <span class="ow">not</span> <span class="n">values</span><span class="p">:</span>
    <span class="c"># no mapped reads...</span>
    <span class="n">mate_inner_dist</span> <span class="o">=</span> <span class="mi">50</span>
    <span class="n">mate_std_dev</span> <span class="o">=</span> <span class="mi">20</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&#39;Mean and standard deviation of inner distance is set to default values 50 and 20&#39;</span><span class="p">)</span>
<span class="kn">else:</span>
    <span class="n">mean</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">values</span><span class="p">)</span><span class="o">/</span><span class="nb">len</span><span class="p">(</span><span class="n">values</span><span class="p">)</span>
    <span class="n">mean_sq</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">([</span><span class="n">x</span><span class="o">*</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">values</span><span class="p">])</span><span class="o">/</span><span class="nb">len</span><span class="p">(</span><span class="n">values</span><span class="p">)</span>
    <span class="n">sd</span> <span class="o">=</span> <span class="n">math</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">mean_sq</span> <span class="o">-</span> <span class="n">mean</span><span class="o">*</span><span class="n">mean</span><span class="p">)</span>
    <span class="n">mate_inner_dist</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">mean</span><span class="p">)</span>
    <span class="n">mate_std_dev</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">sd</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s">&#39;Mean and standard deviation of inner distance is set to {} and {}&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">mean</span><span class="p">),</span> <span class="nb">int</span><span class="p">(</span><span class="n">sd</span><span class="p">)))</span>

<span class="kn">input:</span>
    <span class="n">fastq_files</span>

<span class="kn">depends:</span>
    <span class="n">genes_with_chr_gtf</span>

<span class="kn">output:</span>
    <span class="s">&#39;${alignment_out}/accepted_hits.bam&#39;</span>

<span class="c"># use --zpacker 0 to disable compression of temporary file</span>
<span class="c"># which makes the program run faster</span>
<span class="kn">run:</span>
    <span class="k">if</span> <span class="p">[[</span> <span class="o">-</span><span class="n">f</span> <span class="err">$</span><span class="p">{</span><span class="n">alignment_out</span><span class="p">}</span><span class="o">/</span><span class="n">logs</span><span class="o">/</span><span class="n">run</span><span class="o">.</span><span class="n">log</span> <span class="p">]]</span> <span class="o">&amp;&amp;</span> <span class="p">[[</span> <span class="err">!</span> <span class="o">-</span><span class="n">f</span> <span class="err">$</span><span class="p">{</span><span class="n">alignment_out</span><span class="p">}</span><span class="o">/</span><span class="n">accepted_hits</span><span class="o">.</span><span class="n">bam</span> <span class="p">]];</span>  \
        <span class="n">then</span>  \
        <span class="n">tophat2</span> <span class="o">--</span><span class="n">resume</span> <span class="err">$</span><span class="p">{</span><span class="n">alignment_out</span><span class="p">}</span> <span class="o">||</span>  \
        <span class="n">tophat2</span>  \
        <span class="o">-</span><span class="n">no</span><span class="o">-</span><span class="n">coverage</span><span class="o">-</span><span class="n">search</span>  \
        <span class="o">--</span><span class="n">library</span><span class="o">-</span><span class="nb">type</span> <span class="n">fr</span><span class="o">-</span><span class="n">unstranded</span>  \
        <span class="o">--</span><span class="nb">max</span><span class="o">-</span><span class="n">multihits</span> <span class="mi">20</span>  \
        <span class="o">--</span><span class="n">prefilter</span><span class="o">-</span><span class="n">multihits</span>  \
        <span class="o">--</span><span class="n">num</span><span class="o">-</span><span class="n">threads</span> <span class="mi">8</span>  \
        <span class="o">--</span><span class="n">GTF</span> <span class="err">$</span><span class="p">{</span><span class="n">genes_with_chr_gtf</span><span class="p">}</span>  \
        <span class="o">--</span><span class="n">segment</span><span class="o">-</span><span class="n">length</span> <span class="mi">25</span>  \
        <span class="o">--</span><span class="n">keep</span><span class="o">-</span><span class="n">fasta</span><span class="o">-</span><span class="n">order</span>  \
        <span class="o">--</span><span class="n">transcriptome</span><span class="o">-</span><span class="n">index</span> <span class="err">$</span><span class="p">{</span><span class="n">transcriptome_dir</span><span class="p">}</span><span class="o">/</span><span class="n">genes</span>  \
        <span class="o">--</span><span class="n">fusion</span><span class="o">-</span><span class="n">search</span> <span class="o">--</span><span class="n">fusion</span><span class="o">-</span><span class="n">ignore</span><span class="o">-</span><span class="n">chromosomes</span> <span class="n">chrM</span><span class="p">,</span><span class="n">M</span>  \
        <span class="o">--</span><span class="n">fusion</span><span class="o">-</span><span class="nb">min</span><span class="o">-</span><span class="n">dist</span> <span class="mi">50000</span>  \
        <span class="o">--</span><span class="n">bowtie1</span>  \
        <span class="o">--</span><span class="n">mate</span><span class="o">-</span><span class="n">inner</span><span class="o">-</span><span class="n">dist</span> <span class="err">$</span><span class="p">{</span><span class="n">mate_inner_dist</span><span class="p">}</span>  \
        <span class="o">--</span><span class="n">mate</span><span class="o">-</span><span class="n">std</span><span class="o">-</span><span class="n">dev</span> <span class="err">$</span><span class="p">{</span><span class="n">mate_std_dev</span><span class="p">}</span>  \
        <span class="o">--</span><span class="n">output</span><span class="o">-</span><span class="nb">dir</span> <span class="err">$</span><span class="p">{</span><span class="n">alignment_out</span><span class="p">}</span>  \
        <span class="err">$</span><span class="p">{</span><span class="n">reference_dir</span><span class="p">}</span><span class="o">/</span><span class="n">BowtieIndex</span><span class="o">/</span><span class="n">genome</span>  \
        <span class="err">$</span><span class="p">{</span><span class="s">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">sorted</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">input</span> <span class="k">if</span> <span class="s">&#39;_R1_&#39;</span> <span class="ow">in</span> <span class="n">x</span><span class="p">]))}</span>  \
        <span class="err">$</span><span class="p">{</span><span class="s">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">sorted</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">input</span> <span class="k">if</span> <span class="s">&#39;_R2_&#39;</span> <span class="ow">in</span> <span class="n">x</span><span class="p">]))}</span>  \
        <span class="p">;</span>  \
    <span class="k">else</span>   \
        <span class="n">tophat2</span>  \
        <span class="o">-</span><span class="n">no</span><span class="o">-</span><span class="n">coverage</span><span class="o">-</span><span class="n">search</span>  \
        <span class="o">--</span><span class="n">library</span><span class="o">-</span><span class="nb">type</span> <span class="n">fr</span><span class="o">-</span><span class="n">unstranded</span>   \
        <span class="o">--</span><span class="nb">max</span><span class="o">-</span><span class="n">multihits</span> <span class="mi">20</span>   \
        <span class="o">--</span><span class="n">prefilter</span><span class="o">-</span><span class="n">multihits</span>  \
        <span class="o">--</span><span class="n">num</span><span class="o">-</span><span class="n">threads</span> <span class="mi">8</span>   \
        <span class="o">--</span><span class="n">GTF</span> <span class="err">$</span><span class="p">{</span><span class="n">genes_with_chr_gtf</span><span class="p">}</span>  \
        <span class="o">--</span><span class="n">segment</span><span class="o">-</span><span class="n">length</span> <span class="mi">25</span>   \
        <span class="o">--</span><span class="n">keep</span><span class="o">-</span><span class="n">fasta</span><span class="o">-</span><span class="n">order</span>   \
        <span class="o">--</span><span class="n">transcriptome</span><span class="o">-</span><span class="n">index</span> <span class="err">$</span><span class="p">{</span><span class="n">transcriptome_dir</span><span class="p">}</span><span class="o">/</span><span class="n">genes</span>  \
        <span class="o">--</span><span class="n">fusion</span><span class="o">-</span><span class="n">search</span> <span class="o">--</span><span class="n">fusion</span><span class="o">-</span><span class="n">ignore</span><span class="o">-</span><span class="n">chromosomes</span> <span class="n">chrM</span><span class="p">,</span><span class="n">M</span>   \
        <span class="o">--</span><span class="n">fusion</span><span class="o">-</span><span class="nb">min</span><span class="o">-</span><span class="n">dist</span> <span class="mi">50000</span>  \
        <span class="o">--</span><span class="n">bowtie1</span>   \
        <span class="o">--</span><span class="n">mate</span><span class="o">-</span><span class="n">inner</span><span class="o">-</span><span class="n">dist</span> <span class="err">$</span><span class="p">{</span><span class="n">mate_inner_dist</span><span class="p">}</span>  \
        <span class="o">--</span><span class="n">mate</span><span class="o">-</span><span class="n">std</span><span class="o">-</span><span class="n">dev</span> <span class="err">$</span><span class="p">{</span><span class="n">mate_std_dev</span><span class="p">}</span>  \
        <span class="o">--</span><span class="n">output</span><span class="o">-</span><span class="nb">dir</span> <span class="err">$</span><span class="p">{</span><span class="n">alignment_out</span><span class="p">}</span>  \
        <span class="err">$</span><span class="p">{</span><span class="n">reference_dir</span><span class="p">}</span><span class="o">/</span><span class="n">BowtieIndex</span><span class="o">/</span><span class="n">genome</span>  \
        <span class="err">$</span><span class="p">{</span><span class="s">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">sorted</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">input</span> <span class="k">if</span> <span class="s">&#39;_R1_&#39;</span> <span class="ow">in</span> <span class="n">x</span><span class="p">]))}</span>  \
        <span class="err">$</span><span class="p">{</span><span class="s">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">sorted</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">input</span> <span class="k">if</span> <span class="s">&#39;_R2_&#39;</span> <span class="ow">in</span> <span class="n">x</span><span class="p">]))};</span>  \
        <span class="n">fi</span> 

 
<span class="p">[</span><span class="n">index_bam</span><span class="p">:</span> <span class="kp">provides</span><span class="o">=</span><span class="s">&#39;{filename}.bam.bai&#39;</span><span class="p">]</span>
<span class="c"># Index generated bam file</span>
<span class="kn">input:</span>
    <span class="s">&#39;${filename}.bam&#39;</span>

<span class="kn">run:</span>
    <span class="n">samtools</span> <span class="n">index</span> <span class="err">$</span><span class="p">{</span><span class="nb">input</span><span class="p">}</span>


<span class="p">[</span><span class="n">human_hg19_460</span><span class="p">]</span>
<span class="c"># Add ReadGroup information to sorted bam file.</span>

<span class="c"># Read group identifier, which should be an unique identifier for each</span>
<span class="c"># instrument run (e.g. flowcell number + lane name + number). The filename before</span>
<span class="c"># the &#39;_R1_&#39; part will be used if this option is left unspecified.</span>
<span class="kn">parameter:</span> <span class="n">rgid</span> <span class="o">=</span> <span class="s">&#39;NA&#39;</span>
<span class="n">warn_if</span><span class="p">(</span><span class="n">rgid</span> <span class="o">==</span> <span class="s">&#39;NA&#39;</span><span class="p">,</span>
    <span class="s">&#39;A default (NA) value is used for parameter --rgid (intrument run ID)&#39;</span><span class="p">)</span>

<span class="c"># Platform/technology used to produce the read. Valid values include</span>
<span class="c"># ILLUMINA, SOLID, LS454, HELICOS and PACBIO</span>
<span class="kn">parameter:</span> <span class="n">rgpl</span> <span class="o">=</span> <span class="s">&#39;ILLUMINA&#39;</span>
<span class="c"># check input parameters</span>
<span class="n">warn_if</span><span class="p">(</span><span class="ow">not</span> <span class="n">rgpl</span> <span class="ow">in</span> <span class="p">[</span><span class="s">&#39;ILLUMINA&#39;</span><span class="p">,</span> <span class="s">&#39;SOLID&#39;</span><span class="p">,</span> <span class="s">&#39;LS454&#39;</span><span class="p">,</span> <span class="s">&#39;HELICOS&#39;</span><span class="p">,</span> <span class="s">&#39;PACBIO&#39;</span><span class="p">],</span>
    <span class="s">&#39;PL (platform) value if not one of ILLUMINA, SOLID, LS454, HELICOS and PACBIO&#39;</span><span class="p">)</span>

<span class="c"># DNA preparation library identity.</span>
<span class="kn">parameter:</span> <span class="n">rglb</span> <span class="o">=</span> <span class="s">&#39;NA&#39;</span>
<span class="n">warn_if</span><span class="p">(</span><span class="n">rglb</span> <span class="o">==</span> <span class="s">&#39;NA&#39;</span><span class="p">,</span>
    <span class="s">&#39;A default (NA) value is used for parameter --rglb (library identification string)&#39;</span><span class="p">)</span>
<span class="c">#</span>
<span class="c"># Platform unit (e.g. barcode). This field is not used by GATK.</span>
<span class="kn">parameter:</span> <span class="n">rgpu</span> <span class="o">=</span> <span class="s">&#39;NA&#39;</span>
<span class="n">warn_if</span><span class="p">(</span><span class="n">rgpu</span> <span class="o">==</span> <span class="s">&#39;NA&#39;</span><span class="p">,</span>
    <span class="s">&#39;A default (NA) value is used for parameter --rgpu (platform unit, barcode)&#39;</span><span class="p">)</span>


<span class="kn">input:</span>
     <span class="n">accepted_hits</span>

<span class="kn">output:</span>
    <span class="s">&#39;${temp_dir}/${samplename}-with_rg.bam&#39;</span>

<span class="kn">run:</span>
    <span class="n">java</span> <span class="o">-</span><span class="n">Xmx16g</span> <span class="o">-</span><span class="n">Xms512m</span>   \
        <span class="o">-</span><span class="n">Djava</span><span class="o">.</span><span class="n">io</span><span class="o">.</span><span class="n">tmpdir</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="n">temp_dir</span><span class="p">}</span>   \
        <span class="o">-</span><span class="n">jar</span> <span class="err">$</span><span class="p">{</span><span class="n">picard_path</span><span class="p">}</span><span class="o">/</span><span class="n">AddOrReplaceReadGroups</span><span class="o">.</span><span class="n">jar</span>  \
        <span class="nb">input</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="nb">input</span><span class="p">}</span>  \
        <span class="n">OUTPUT</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="n">temp_dir</span><span class="p">}</span><span class="o">/</span><span class="err">$</span><span class="p">{</span><span class="n">samplename</span><span class="p">}</span><span class="o">-</span><span class="n">with_rg</span><span class="o">.</span><span class="n">bam</span>  \
        <span class="n">VALIDATION_STRINGENCY</span><span class="o">=</span><span class="n">LENIENT</span>  \
        <span class="n">RGID</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="n">rgid</span><span class="p">}</span>  \
        <span class="n">RGSM</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="n">samplename</span><span class="p">}</span>  \
        <span class="n">RGPL</span><span class="o">=</span><span class="n">ILLUMINA</span>  \
        <span class="n">RGLB</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="n">rglb</span><span class="p">}</span>  \
        <span class="n">RGPU</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="n">rgpu</span><span class="p">}</span>  \
        <span class="n">RGCN</span><span class="o">=</span><span class="n">IPCT</span>

<span class="p">[</span><span class="n">human_hg19_470</span><span class="p">:</span> <span class="kp">shared</span><span class="o">=</span><span class="p">{</span><span class="s">&#39;sorted_unique&#39;</span><span class="p">:</span><span class="s">&#39;output[0]&#39;</span><span class="p">}]</span>
<span class="c">#Alignment: Remove fusion reads in preparation for variant calling.</span>
<span class="c"># step 475, samtools index is defined above</span>
<span class="kn">output:</span>
    <span class="s">&#39;${alignment_out}/${samplename}-sorted_uniq_nonF.bam&#39;</span>

<span class="kn">run:</span>
    <span class="n">samtools</span> <span class="n">view</span> <span class="o">-</span><span class="n">h</span> <span class="err">$</span><span class="p">{</span><span class="nb">input</span><span class="p">}</span> \
        <span class="o">|</span> <span class="n">awk</span> <span class="o">-</span><span class="n">F</span> <span class="s">&#39;</span><span class="se">\t</span><span class="s">&#39;</span> <span class="s">&#39;{ if($0 ~ &quot;^@&quot;) {print} else { for(i=12;i&lt;=NF;i++){ if ($i ~ &quot;NH:i:1$&quot;){print}} } }&#39;</span> \
        <span class="o">|</span> <span class="n">awk</span> <span class="s">&#39;{ if($0 ~ &quot;^@&quot;) {print} else { if ($0 !~ &quot;XF:Z&quot;) {print} }  }&#39;</span> \
        <span class="o">|</span> <span class="n">samtools</span> <span class="n">view</span> <span class="o">-</span><span class="n">bS</span> <span class="o">-</span> <span class="o">&gt;</span> <span class="err">$</span><span class="p">{</span><span class="n">alignment_out</span><span class="p">}</span><span class="o">/</span><span class="err">$</span><span class="p">{</span><span class="n">samplename</span><span class="p">}</span><span class="o">-</span><span class="n">sorted_uniq_nonF</span><span class="o">.</span><span class="n">bam</span>


<span class="p">[</span><span class="n">human_hg19_480</span><span class="p">]</span>
<span class="c">#Alignment: Count the number of junction reads. </span>
<span class="kn">input:</span>
    <span class="n">accepted_hits</span>

<span class="kn">output:</span>
    <span class="s">&#39;${alignment_out}/junction.count&#39;</span>

<span class="kn">run:</span>
    <span class="n">samtools</span> <span class="n">view</span> <span class="o">-</span><span class="n">h</span> <span class="err">$</span><span class="p">{</span><span class="nb">input</span><span class="p">}</span> \
   <span class="o">|</span> <span class="n">awk</span> <span class="s">&#39;$6 ~/N/&#39;</span> <span class="o">|</span> <span class="n">awk</span> <span class="s">&#39;{ if ($9 ~ /^-/) {print $1&quot;</span><span class="se">\t</span><span class="s">-&quot;} else print $1&quot;</span><span class="se">\t</span><span class="s">+&quot; }&#39;</span> \
   <span class="o">|</span> <span class="n">sort</span> <span class="o">-</span><span class="n">T</span> <span class="err">$</span><span class="p">{</span><span class="n">temp_dir</span><span class="p">}</span> <span class="o">-</span><span class="n">u</span> <span class="o">|</span> <span class="n">wc</span> <span class="o">-</span><span class="n">l</span> <span class="o">&gt;</span> <span class="err">$</span><span class="p">{</span><span class="n">alignment_out</span><span class="p">}</span><span class="o">/</span><span class="n">junction</span><span class="o">.</span><span class="n">count</span>


<span class="p">[</span><span class="n">human_hg19_500</span> <span class="p">(</span><span class="n">gene</span> <span class="n">body</span> <span class="n">coverage</span><span class="p">):</span> <span class="kp">shared</span><span class="o">=</span><span class="p">{</span><span class="s">&#39;rseqc_pdf&#39;</span><span class="p">:</span><span class="s">&#39;output&#39;</span><span class="p">}]</span>
<span class="c"># this version uses BAM files to count geneBody coverage, using .bw might be a more resource</span>
<span class="c"># saving idea but needs more testing, especially because the output of bigWig is currently normalized</span>
<span class="c"># and does not show the correct depth of coverage.</span>
<span class="kn">input:</span>
    <span class="n">sorted_unique</span>

<span class="kn">depends:</span>
    <span class="n">refgene_bed</span>

<span class="kn">output:</span>
    <span class="s">&#39;${alignmentqc_out}/${samplename}.geneBodyCoverage.curves.pdf&#39;</span>

<span class="c"># this job can take a long time to execute so we tend to execute as a separate job</span>
<span class="kn">task:</span>
<span class="kn">run:</span>
    <span class="n">geneBody_coverage</span><span class="o">.</span><span class="n">py</span> <span class="o">-</span><span class="n">r</span> <span class="err">$</span><span class="p">{</span><span class="n">refgene_bed</span><span class="p">}</span> <span class="o">-</span><span class="n">i</span> <span class="err">$</span><span class="p">{</span><span class="nb">input</span><span class="p">}</span> <span class="o">-</span><span class="n">o</span> <span class="err">$</span><span class="p">{</span><span class="n">alignmentqc_out</span><span class="p">}</span><span class="o">/</span><span class="err">$</span><span class="p">{</span><span class="n">samplename</span><span class="p">}</span>

<span class="c">#[human_hg19_560]</span>
<span class="c">#RSeQC: Coverage of gene body</span>
<span class="c">#RunCommand(&#39;geneBody_coverage2.py -r ${refgene_bed} -i ${input551: input551[0][:-3] + &quot;bw&quot;} -t pdf -o ${alignmentqc_out}/${samplename}&#39;,</span>
<span class="c">#    output=&#39;${alignmentqc_out}/${samplename}.geneBodyCoverage.curves.pdf&#39;)</span>

<span class="p">[</span><span class="n">human_hg19_510</span><span class="p">]</span>
<span class="c">#RSeQC: Check if current sequencing depth is deep enough to perform alternative splicing analyses</span>
<span class="c">#RSeQC: Annotating splicing event and splicing junction.</span>
<span class="c">#RSeQC: Plot the distribution of inner distance between paired reads, which can be used to </span>
<span class="c">#   determine parameters --mate-inner-dist and --mate-std-dev of the pipeline.</span>
<span class="c">#RSeQC: read duplication plot</span>
<span class="c">#RSeQC: Read distribution (statistics)</span>
<span class="c">#RSeQC: Generate scaled wig track for depth</span>
<span class="kn">input:</span>
    <span class="n">accepted_hits</span>

<span class="kn">output:</span>
    <span class="s">&#39;${alignmentqc_out}/${samplename}.junctionSaturation_plot.pdf&#39;</span><span class="p">,</span>
    <span class="s">&#39;${alignmentqc_out}/${samplename}.splice_events.pdf&#39;</span><span class="p">,</span>
    <span class="s">&#39;${alignmentqc_out}/${samplename}.splice_junction.pdf&#39;</span><span class="p">,</span>
    <span class="s">&#39;${alignmentqc_out}/${samplename}.inner_distance_plot.pdf&#39;</span><span class="p">,</span>
    <span class="s">&#39;${alignmentqc_out}//${samplename}.DupRate_plot.pdf&#39;</span><span class="p">,</span>
    <span class="s">&#39;${alignmentqc_out}//${samplename}.read_distribution&#39;</span>

<span class="kn">task:</span>
<span class="kn">run:</span>
    <span class="n">junction_saturation</span><span class="o">.</span><span class="n">py</span>  <span class="o">-</span><span class="n">r</span> <span class="err">$</span><span class="p">{</span><span class="n">refgene_bed</span><span class="p">}</span> <span class="o">-</span><span class="n">i</span> <span class="err">$</span><span class="p">{</span><span class="nb">input</span><span class="p">}</span> <span class="o">-</span><span class="n">o</span> <span class="err">$</span><span class="p">{</span><span class="n">alignmentqc_out</span><span class="p">}</span><span class="o">/</span><span class="err">$</span><span class="p">{</span><span class="n">samplename</span><span class="p">}</span>
    <span class="n">junction_annotation</span><span class="o">.</span><span class="n">py</span> <span class="o">-</span><span class="n">r</span> <span class="err">$</span><span class="p">{</span><span class="n">refgene_bed</span><span class="p">}</span> <span class="o">-</span><span class="n">i</span> <span class="err">$</span><span class="p">{</span><span class="nb">input</span><span class="p">}</span> <span class="o">-</span><span class="n">o</span> <span class="err">$</span><span class="p">{</span><span class="n">alignmentqc_out</span><span class="p">}</span><span class="o">/</span><span class="err">$</span><span class="p">{</span><span class="n">samplename</span><span class="p">}</span>
    <span class="n">inner_distance</span><span class="o">.</span><span class="n">py</span> <span class="o">-</span><span class="n">r</span> <span class="err">$</span><span class="p">{</span><span class="n">refgene_bed</span><span class="p">}</span> <span class="o">-</span><span class="n">i</span> <span class="err">$</span><span class="p">{</span><span class="nb">input</span><span class="p">}</span> <span class="o">-</span><span class="n">o</span> <span class="err">$</span><span class="p">{</span><span class="n">alignmentqc_out</span><span class="p">}</span><span class="o">/</span><span class="err">$</span><span class="p">{</span><span class="n">samplename</span><span class="p">}</span>
    <span class="n">read_duplication</span><span class="o">.</span><span class="n">py</span> <span class="o">-</span><span class="n">i</span> <span class="err">$</span><span class="p">{</span><span class="nb">input</span><span class="p">}</span> <span class="o">-</span><span class="n">o</span> <span class="err">$</span><span class="p">{</span><span class="n">alignmentqc_out</span><span class="p">}</span><span class="o">/</span><span class="err">$</span><span class="p">{</span><span class="n">samplename</span><span class="p">}</span>
    <span class="n">read_distribution</span><span class="o">.</span><span class="n">py</span> <span class="o">-</span><span class="n">r</span> <span class="err">$</span><span class="p">{</span><span class="n">refgene_bed</span><span class="p">}</span> <span class="o">-</span><span class="n">i</span> <span class="err">$</span><span class="p">{</span><span class="nb">input</span><span class="p">}</span> <span class="o">&gt;</span> <span class="err">$</span><span class="p">{</span><span class="n">alignmentqc_out</span><span class="p">}</span><span class="o">/</span><span class="err">$</span><span class="p">{</span><span class="n">samplename</span><span class="p">}</span><span class="o">.</span><span class="n">read_distribution</span>

<span class="p">[</span><span class="n">human_hg19_600</span> <span class="p">(</span><span class="n">SortSam</span><span class="p">):</span> <span class="kp">shared</span><span class="o">=</span><span class="p">{</span><span class="s">&#39;sorted_bam&#39;</span><span class="p">:</span><span class="s">&#39;output&#39;</span><span class="p">}]</span>
<span class="c">#Count: Order reads by name so that the BAM file can be processed by htseq-count</span>
<span class="kn">input:</span>
    <span class="n">accepted_hits</span>

<span class="kn">output:</span>
    <span class="s">&#39;${alignment_out}/${samplename}-sorted_by_id.bam&#39;</span>

<span class="kn">task:</span>
<span class="kn">run:</span>
    <span class="n">java</span> <span class="o">-</span><span class="n">Xmx16g</span> <span class="o">-</span><span class="n">Xms512m</span>   \
        <span class="o">-</span><span class="n">Djava</span><span class="o">.</span><span class="n">io</span><span class="o">.</span><span class="n">tmpdir</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="n">temp_dir</span><span class="p">}</span>  \
        <span class="o">-</span><span class="n">jar</span> <span class="err">$</span><span class="p">{</span><span class="n">picard_path</span><span class="p">}</span><span class="o">/</span><span class="n">SortSam</span><span class="o">.</span><span class="n">jar</span>  \
        <span class="nb">input</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="nb">input</span><span class="p">}</span>  \
        <span class="n">OUTPUT</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="n">alignment_out</span><span class="p">}</span><span class="o">/</span><span class="err">$</span><span class="p">{</span><span class="n">samplename</span><span class="p">}</span><span class="o">-</span><span class="n">sorted_by_id</span><span class="o">.</span><span class="n">bam</span>  \
        <span class="n">SO</span><span class="o">=</span><span class="n">queryname</span>   \
        <span class="n">MAX_RECORDS_IN_RAM</span><span class="o">=</span><span class="mi">1000000</span>   \
        <span class="n">TMP_DIR</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="n">temp_dir</span><span class="p">}</span>  \
        <span class="n">VALIDATION_STRINGENCY</span><span class="o">=</span><span class="n">SILENT</span>


<span class="p">[</span><span class="n">human_hg19_610</span> <span class="p">(</span><span class="n">htseq</span><span class="o">-</span><span class="n">count</span><span class="p">)]</span>
<span class="c"># Running htseq-count to count reads for genes. This step uses the union</span>
<span class="c">#    intersection mode, and a gene definition file with chromosome name added to gene_id.</span>
<span class="kn">input:</span>
    <span class="n">sorted_bam</span>

<span class="kn">output:</span>
    <span class="s">&#39;${count_out}/${samplename}-gene_counts.tsv&#39;</span>

<span class="c"># We cannot use --order=pos because htseq-count would fail with an error message </span>
<span class="kn">run:</span>
    <span class="n">htseq</span><span class="o">-</span><span class="n">count</span> <span class="o">--</span><span class="n">stranded</span><span class="o">=</span><span class="n">no</span> <span class="o">--</span><span class="n">format</span><span class="o">=</span><span class="n">bam</span> <span class="o">--</span><span class="nb">type</span><span class="o">=</span><span class="n">exon</span> \
     <span class="o">--</span><span class="n">mode</span><span class="o">=</span><span class="n">union</span> <span class="err">$</span><span class="p">{</span><span class="nb">input</span><span class="p">}</span> <span class="err">$</span><span class="p">{</span><span class="n">genes_with_chr_gtf</span><span class="p">}</span> \
     <span class="o">&gt;</span> <span class="err">$</span><span class="p">{</span><span class="n">count_out</span><span class="p">}</span><span class="o">/</span><span class="err">$</span><span class="p">{</span><span class="n">samplename</span><span class="p">}</span><span class="o">-</span><span class="n">gene_counts</span><span class="o">.</span><span class="n">tsv</span>


<span class="p">[</span><span class="n">get_genesize</span><span class="p">:</span> <span class="kp">provides</span><span class="o">=</span><span class="n">genesize_txt</span><span class="p">]</span>
<span class="c">#    Genes.gtf file in the format of</span>
<span class="c">#</span>
<span class="c">#    chr1    unknown    exon    17233    17368    .    -    .    gene_id &quot;WASH7P&quot;; gene_name &quot;WASH7P&quot;; transcript_id &quot;NR_024540&quot;; tss_id &quot;TSS8151&quot;;</span>
<span class="c">#    chr1    unknown    exon    17369    17436    .    -    .    gene_id &quot;MIR6859-2&quot;; gene_name &quot;MIR6859-2&quot;; transcript_id &quot;NR_107062&quot;; tss_id &quot;TSS24460&quot;;</span>
<span class="c">#    chr1    unknown    exon    17369    17436    .    -    .    gene_id &quot;MIR6859-1&quot;; gene_name &quot;MIR6859-1&quot;; transcript_id &quot;NR_106918_2&quot;; tss_id &quot;TSS24460&quot;;</span>
<span class="c">#</span>
<span class="c"># output (string):</span>
<span class="c">#    A text file in the format of</span>
<span class="c">#</span>
<span class="c">#    chr1  MIR6859-2  112324</span>
<span class="kn">input:</span>
    <span class="n">genes_gtf</span>


<span class="kn">python:</span>

    <span class="kn">import</span> <span class="nn">sys</span>
    <span class="k">def</span> <span class="nf">_size</span><span class="p">(</span><span class="n">features</span><span class="p">):</span>
        <span class="c"># this is definitely not efficient but we do not really care because it will be run</span>
        <span class="c"># only once. </span>
        <span class="n">bases</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">feature</span> <span class="ow">in</span> <span class="n">features</span><span class="p">:</span>
            <span class="c"># Add 1 because GTF input is one based</span>
            <span class="n">bases</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">feature</span><span class="p">[</span><span class="mi">0</span><span class="p">]),</span> <span class="nb">int</span><span class="p">(</span><span class="n">feature</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="o">+</span> <span class="mi">1</span><span class="p">))</span>
        <span class="k">return</span> <span class="nb">len</span><span class="p">(</span><span class="n">bases</span><span class="p">)</span>

    <span class="c"># only work on selected chromosomes</span>
    <span class="n">allowed_chromosomes</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([</span><span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">25</span><span class="p">)]</span> <span class="o">+</span> <span class="p">[</span><span class="s">&#39;chr{}&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">25</span><span class="p">)]</span> <span class="o">+</span> \
                <span class="p">[</span><span class="s">&#39;X&#39;</span><span class="p">,</span> <span class="s">&#39;Y&#39;</span><span class="p">,</span> <span class="s">&#39;M&#39;</span><span class="p">,</span> <span class="s">&#39;chrX&#39;</span><span class="p">,</span> <span class="s">&#39;chrY&#39;</span><span class="p">,</span> <span class="s">&#39;chrM&#39;</span><span class="p">])</span>
 
    <span class="n">gene_size</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="nb">input</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="s">&#39;rU&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">ifile</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">ifile</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s">&#39;#&#39;</span><span class="p">):</span>
                <span class="k">continue</span>
            <span class="n">ls</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&#39;</span><span class="se">\t</span><span class="s">&#39;</span><span class="p">)</span>
            <span class="c">#</span>
            <span class="c"># ls[0]: seqname (chromosome name)</span>
            <span class="c"># ls[1]: source (--ignored--)</span>
            <span class="c"># ls[2]: feature (exon only)</span>
            <span class="c"># ls[3]: start position (1-based)</span>
            <span class="c"># ls[4]: end position (1-based)</span>
            <span class="c"># ls[5]: score (--ignored---)</span>
            <span class="c"># ls[6]: strand (--ignored--)</span>
            <span class="c"># ls[7]: frame (--ignored--)</span>
            <span class="c"># ls[8]: attributes, look for gene_id</span>
            <span class="c">#</span>
            <span class="k">if</span> <span class="n">ls</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">allowwd_chromosomes</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">ls</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">5</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="k">if</span> <span class="n">ls</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="o">!=</span> <span class="s">&#39;exon&#39;</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">gene_id</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">rsplit</span><span class="p">(</span><span class="s">&#39; &#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">strip</span><span class="p">(</span><span class="s">&#39;&quot;&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">strip</span><span class="p">(</span><span class="s">&quot;&#39;&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">ls</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&#39;;&#39;</span><span class="p">)</span> <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s">&#39;gene_id&#39;</span><span class="p">)]</span>
            <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                <span class="n">sys</span><span class="o">.</span><span class="n">stderr</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;Failed to parse {}: {}&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">line</span><span class="p">,</span> <span class="n">e</span><span class="p">))</span>
                <span class="k">continue</span>
            <span class="c">#</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">gene_id</span><span class="p">:</span>
                <span class="n">sys</span><span class="o">.</span><span class="n">stderr</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;No gene_id for record {}&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">line</span><span class="p">))</span>
            <span class="c">#</span>
            <span class="c"># use gene_id.chr as key</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">key</span> <span class="o">=</span> <span class="n">gene_id</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s">&#39;.&#39;</span> <span class="o">+</span> <span class="n">ls</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                <span class="k">if</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">gene_size</span><span class="p">:</span>
                    <span class="k">if</span> <span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">ls</span><span class="p">[</span><span class="mi">3</span><span class="p">]),</span> <span class="nb">int</span><span class="p">(</span><span class="n">ls</span><span class="p">[</span><span class="mi">4</span><span class="p">]))</span> <span class="o">!=</span> <span class="n">gene_size</span><span class="p">[</span><span class="n">key</span><span class="p">][</span><span class="mi">3</span><span class="p">][</span><span class="o">-</span><span class="mi">1</span><span class="p">]:</span>
                        <span class="n">gene_size</span><span class="p">[</span><span class="n">key</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">gene_size</span><span class="p">[</span><span class="n">key</span><span class="p">][</span><span class="mi">1</span><span class="p">],</span> <span class="nb">int</span><span class="p">(</span><span class="n">ls</span><span class="p">[</span><span class="mi">3</span><span class="p">]))</span>
                        <span class="n">gene_size</span><span class="p">[</span><span class="n">key</span><span class="p">][</span><span class="mi">2</span><span class="p">]</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">gene_size</span><span class="p">[</span><span class="n">key</span><span class="p">][</span><span class="mi">2</span><span class="p">],</span> <span class="nb">int</span><span class="p">(</span><span class="n">ls</span><span class="p">[</span><span class="mi">4</span><span class="p">]))</span>
                        <span class="n">gene_size</span><span class="p">[</span><span class="n">key</span><span class="p">][</span><span class="mi">3</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="nb">int</span><span class="p">(</span><span class="n">ls</span><span class="p">[</span><span class="mi">3</span><span class="p">]),</span> <span class="nb">int</span><span class="p">(</span><span class="n">ls</span><span class="p">[</span><span class="mi">4</span><span class="p">])))</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">gene_size</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">ls</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="nb">int</span><span class="p">(</span><span class="n">ls</span><span class="p">[</span><span class="mi">3</span><span class="p">]),</span> <span class="nb">int</span><span class="p">(</span><span class="n">ls</span><span class="p">[</span><span class="mi">4</span><span class="p">]),</span> <span class="p">[(</span><span class="nb">int</span><span class="p">(</span><span class="n">ls</span><span class="p">[</span><span class="mi">3</span><span class="p">]),</span> <span class="nb">int</span><span class="p">(</span><span class="n">ls</span><span class="p">[</span><span class="mi">4</span><span class="p">]))]]</span>
            <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                <span class="n">sys</span><span class="o">.</span><span class="n">stderr</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;Failed to store key {}: {}&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">key</span><span class="p">,</span> <span class="n">e</span><span class="p">))</span>
                <span class="k">continue</span>
    <span class="c">#</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">output</span><span class="err">!</span><span class="n">r</span><span class="p">},</span> <span class="s">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">ofile</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">gene_size</span><span class="o">.</span><span class="n">keys</span><span class="p">()):</span>
            <span class="n">val</span> <span class="o">=</span> <span class="n">gene_size</span><span class="p">[</span><span class="n">key</span><span class="p">]</span>
            <span class="c">#</span>
            <span class="c"># Remove chr from key for output</span>
            <span class="n">ofile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;{0}</span><span class="se">\t</span><span class="s">{1}</span><span class="se">\t</span><span class="s">{2}</span><span class="se">\t</span><span class="s">{3}</span><span class="se">\t</span><span class="s">{4}</span><span class="se">\n</span><span class="s">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">val</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">key</span><span class="p">[:</span><span class="o">-</span><span class="nb">len</span><span class="p">(</span><span class="n">val</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">val</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">val</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span> 
                <span class="n">_size</span><span class="p">(</span><span class="n">val</span><span class="p">[</span><span class="mi">3</span><span class="p">])))</span>


<span class="p">[</span><span class="n">flagstat</span><span class="p">:</span> <span class="kp">provides</span><span class="o">=</span><span class="s">&#39;{filename}.bam.flagstat&#39;</span><span class="p">]</span>
<span class="c"># Alignment (samtools): Generating statistics of aligned reads.</span>
<span class="kn">input:</span>
    <span class="s">&#39;${filename}.bam&#39;</span>

<span class="kn">run:</span>
    <span class="n">samtools</span> <span class="n">flagstat</span> <span class="err">$</span><span class="p">{</span><span class="nb">input</span><span class="p">}</span> <span class="o">&gt;</span> <span class="err">$</span><span class="p">{</span><span class="nb">input</span><span class="p">}</span><span class="o">.</span><span class="n">flagstat</span>
 
<span class="c">## # format of flagstat file. </span>
<span class="c">## # </span>
<span class="c">## #2181882 + 0 in total (QC-passed reads + QC-failed reads)</span>
<span class="c">## #0 + 0 duplicates</span>
<span class="c">## #2181882 + 0 mapped (100.00%:nan%)</span>
<span class="c">## #2181882 + 0 paired in sequencing</span>
<span class="c">## #mapped_reads = ${eval([line.split(&#39;m&#39;)[0] for line in open(input[0] + &#39;.flagstat&#39;).read().split(&#39;\n&#39;) if &#39;mapped&#39; in line][0])}</span>
<span class="c">## </span>

<span class="p">[</span><span class="n">mapped_reads</span><span class="p">:</span> <span class="kp">shared</span><span class="o">=</span><span class="s">&#39;mapped_reads&#39;</span><span class="p">]</span>
<span class="kn">input:</span>
    <span class="s">&#39;${accepted_hits}.flagstat&#39;</span>

<span class="kn">depends:</span>
    <span class="n">sos_variable</span><span class="p">(</span><span class="s">&#39;accepted_hits&#39;</span><span class="p">)</span>

<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="nb">input</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="k">as</span> <span class="n">flgs</span><span class="p">:</span>
    <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">flgs</span><span class="p">:</span>
        <span class="k">if</span> <span class="s">&#39;mapped&#39;</span> <span class="ow">in</span> <span class="n">line</span><span class="p">:</span>
            <span class="n">mapped_reads</span> <span class="o">=</span> <span class="nb">eval</span><span class="p">(</span><span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&#39;m&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">])</span>
            <span class="k">break</span>


<span class="p">[</span><span class="n">human_hg19_620</span> <span class="p">(</span><span class="n">Gene</span> <span class="n">Count</span> <span class="n">RPKM</span><span class="p">)]</span>

<span class="kn">input:</span>
     <span class="s">&#39;${count_out}/${samplename}-gene_counts.tsv&#39;</span>

<span class="kn">depends:</span> 
    <span class="n">genesize_txt</span><span class="p">,</span>
    <span class="n">sos_variable</span><span class="p">(</span><span class="s">&#39;mapped_reads&#39;</span><span class="p">)</span>

<span class="kn">output:</span>
    <span class="s">&quot;${count_out}/${samplename}-gene_counts_rpkm.xls&quot;</span>

<span class="kn">python:</span>
    <span class="kn">import</span> <span class="nn">sys</span>
    <span class="n">gene_size</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">ambiguous_size</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">depends</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="err">!</span><span class="n">r</span><span class="p">})</span> <span class="k">as</span> <span class="n">genesize</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">genesize</span><span class="p">:</span>
            <span class="nb">chr</span><span class="p">,</span> <span class="n">gene</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">stop</span><span class="p">,</span> <span class="n">size</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&#39;</span><span class="se">\t</span><span class="s">&#39;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">gene_size</span> <span class="ow">and</span> <span class="n">gene_size</span><span class="p">[</span><span class="n">gene</span><span class="p">][</span><span class="mi">3</span><span class="p">]</span> <span class="o">!=</span> <span class="nb">int</span><span class="p">(</span><span class="n">size</span><span class="p">):</span>
                <span class="n">ambiguous_size</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
            <span class="n">gene_size</span><span class="p">[</span><span class="n">gene</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="nb">chr</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">stop</span><span class="p">,</span> <span class="nb">int</span><span class="p">(</span><span class="n">size</span><span class="p">))</span>
            <span class="n">gene_size</span><span class="p">[</span><span class="n">gene</span> <span class="o">+</span> <span class="s">&#39;.&#39;</span> <span class="o">+</span> <span class="nb">chr</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="nb">chr</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">stop</span><span class="p">,</span> <span class="nb">int</span><span class="p">(</span><span class="n">size</span><span class="p">))</span>
    <span class="c">#</span>
    <span class="k">print</span><span class="p">(</span><span class="s">&#39;Size of {} genes imported.&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">gene_size</span><span class="p">)</span><span class="o">/</span><span class="mi">2</span><span class="p">))</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="nb">input</span><span class="err">!</span><span class="n">r</span><span class="p">})</span> <span class="k">as</span> <span class="n">counts</span><span class="p">,</span> <span class="nb">open</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">output</span><span class="err">!</span><span class="n">r</span><span class="p">},</span> <span class="s">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">output</span><span class="p">:</span>
        <span class="n">output</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;Chr</span><span class="se">\t</span><span class="s">GeneID</span><span class="se">\t</span><span class="s">Start</span><span class="se">\t</span><span class="s">Stop</span><span class="se">\t</span><span class="s">CodingLength</span><span class="se">\t</span><span class="s">ReadCount</span><span class="se">\t</span><span class="s">RPKM</span><span class="se">\n</span><span class="s">&#39;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">counts</span><span class="p">:</span>
            <span class="c"># &#39;__no_feature&#39;, &#39;__ambiguous&#39;, &#39;__too_low_aQual&#39;, &#39;not_aligned&#39;, &#39;__alignment_not_unique&#39;</span>
            <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s">&#39;__&#39;</span><span class="p">):</span>
                <span class="k">continue</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">gene</span><span class="p">,</span> <span class="n">value</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&#39;</span><span class="se">\t</span><span class="s">&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
            <span class="k">except</span><span class="p">:</span>
                <span class="n">sys</span><span class="o">.</span><span class="n">stderr</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;Unprocess line: {}&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">line</span><span class="p">))</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">gene</span><span class="o">.</span><span class="n">strip</span><span class="p">():</span>
                <span class="k">continue</span>
            <span class="k">elif</span> <span class="n">gene</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">gene_size</span><span class="p">:</span>
                <span class="n">sys</span><span class="o">.</span><span class="n">stderr</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;Unknown size of gene {}&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">gene</span><span class="p">))</span>
                <span class="n">output</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;NA</span><span class="se">\t</span><span class="s">{}</span><span class="se">\t</span><span class="s">NA</span><span class="se">\t</span><span class="s">NA</span><span class="se">\t</span><span class="s">NA</span><span class="se">\t</span><span class="s">{}</span><span class="se">\t</span><span class="s">NA</span><span class="se">\n</span><span class="s">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">value</span><span class="o">.</span><span class="n">strip</span><span class="p">()))</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">ambiguous_size</span><span class="p">:</span>
                    <span class="n">sys</span><span class="o">.</span><span class="n">stderr</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;{} appear in different chromosomes. Its RPKM value is based on the size of one of them&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">gene</span><span class="p">))</span>
                <span class="n">output</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;{}</span><span class="se">\t</span><span class="s">{}</span><span class="se">\t</span><span class="s">{}</span><span class="se">\t</span><span class="s">{}</span><span class="se">\t</span><span class="s">{}</span><span class="se">\t</span><span class="s">{}</span><span class="se">\t</span><span class="s">{:.2f}</span><span class="se">\n</span><span class="s">&#39;</span>
                    <span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">gene_size</span><span class="p">[</span><span class="n">gene</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span>
                        <span class="c"># output gene name if the genename was appended with chromosome name</span>
                        <span class="n">gene</span><span class="p">[:</span><span class="o">-</span><span class="nb">len</span><span class="p">(</span><span class="n">gene_size</span><span class="p">[</span><span class="n">gene</span><span class="p">][</span><span class="mi">0</span><span class="p">])</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="k">if</span> <span class="n">gene</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s">&#39;.&#39;</span> <span class="o">+</span> <span class="n">gene_size</span><span class="p">[</span><span class="n">gene</span><span class="p">][</span><span class="mi">0</span><span class="p">])</span> <span class="k">else</span> <span class="n">gene</span><span class="p">,</span>
                        <span class="n">gene_size</span><span class="p">[</span><span class="n">gene</span><span class="p">][</span><span class="mi">1</span><span class="p">],</span>
                        <span class="n">gene_size</span><span class="p">[</span><span class="n">gene</span><span class="p">][</span><span class="mi">2</span><span class="p">],</span> <span class="n">gene_size</span><span class="p">[</span><span class="n">gene</span><span class="p">][</span><span class="mi">3</span><span class="p">],</span> <span class="n">value</span><span class="o">.</span><span class="n">strip</span><span class="p">(),</span>
                        <span class="mf">1.e9</span><span class="o">*</span><span class="nb">int</span><span class="p">(</span><span class="n">value</span><span class="p">)</span><span class="o">/</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">mapped_reads</span><span class="p">}</span> <span class="o">*</span> <span class="n">gene_size</span><span class="p">[</span><span class="n">gene</span><span class="p">][</span><span class="mi">3</span><span class="p">])))</span>

<span class="p">[</span><span class="n">human_hg19_630</span><span class="p">]</span>
<span class="c">#Count: Running bedtools interectBed to get exon count.</span>
<span class="c">#Count: Calculate RPKM for exon counts.</span>
<span class="kn">input:</span>
    <span class="n">sorted_bam</span>

<span class="kn">output:</span>
    <span class="s">&#39;${count_out}/${samplename}-exon_counts.tsv&#39;</span>

<span class="kn">run:</span>
    <span class="n">coverageBed</span> <span class="o">-</span><span class="n">abam</span> <span class="err">$</span><span class="p">{</span><span class="nb">input</span><span class="p">}</span> <span class="o">-</span><span class="n">b</span> <span class="err">$</span><span class="p">{</span><span class="n">genes_exon_bed</span><span class="p">}</span>  \
        <span class="o">&gt;</span> <span class="err">$</span><span class="p">{</span><span class="n">count_out</span><span class="p">}</span><span class="o">/</span><span class="err">$</span><span class="p">{</span><span class="n">samplename</span><span class="p">}</span><span class="o">-</span><span class="n">exon_counts</span><span class="o">.</span><span class="n">tsv</span>

<span class="p">[</span><span class="n">human_hg19_631</span><span class="p">]</span>

<span class="kn">output:</span>
    <span class="s">&quot;${count_out}/${samplename}-exon_counts_rpkm.xls&quot;</span>
    
<span class="kn">depends:</span>
    <span class="n">sos_variable</span><span class="p">(</span><span class="s">&#39;mapped_reads&#39;</span><span class="p">)</span>

<span class="kn">python:</span>
    <span class="kn">import</span> <span class="nn">sys</span>
    <span class="kn">from</span> <span class="nn">collections</span> <span class="kn">import</span> <span class="n">defaultdict</span>

    <span class="n">exon_counts</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="nb">input</span><span class="err">!</span><span class="n">r</span><span class="p">})</span> <span class="k">as</span> <span class="n">counts</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">counts</span><span class="p">:</span>
            <span class="n">fields</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">fields</span><span class="p">)</span> <span class="o">==</span> <span class="mi">16</span><span class="p">:</span>
                <span class="n">exon_counts</span><span class="p">[</span><span class="nb">tuple</span><span class="p">(</span><span class="n">fields</span><span class="p">[</span><span class="mi">12</span><span class="p">:</span><span class="mi">16</span><span class="p">])]</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="k">elif</span> <span class="nb">len</span><span class="p">(</span><span class="n">fields</span><span class="p">)</span> <span class="o">==</span> <span class="mi">8</span><span class="p">:</span>
                <span class="k">if</span> <span class="nb">int</span><span class="p">(</span><span class="n">fields</span><span class="p">[</span><span class="mi">4</span><span class="p">])</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="n">exon_counts</span><span class="p">[</span><span class="nb">tuple</span><span class="p">(</span><span class="n">fields</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">4</span><span class="p">])]</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">fields</span><span class="p">[</span><span class="mi">4</span><span class="p">])</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">sys</span><span class="o">.</span><span class="n">stderr</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;Incorrect input line: {}&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">line</span><span class="p">))</span>
                <span class="k">continue</span>
    <span class="n">sys</span><span class="o">.</span><span class="n">stderr</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;Writing counts for {} exons&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">exon_counts</span><span class="p">)))</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">output</span><span class="err">!</span><span class="n">r</span><span class="p">},</span> <span class="s">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">output</span><span class="p">:</span>
        <span class="n">output</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;Chr</span><span class="se">\t</span><span class="s">Start</span><span class="se">\t</span><span class="s">Stop</span><span class="se">\t</span><span class="s">Exon</span><span class="se">\t</span><span class="s">ReadCount</span><span class="se">\t</span><span class="s">RPKM</span><span class="se">\n</span><span class="s">&#39;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">exon</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">exon_counts</span><span class="o">.</span><span class="n">keys</span><span class="p">(),</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="mi">3</span><span class="p">]):</span>
            <span class="n">count</span> <span class="o">=</span> <span class="n">exon_counts</span><span class="p">[</span><span class="n">exon</span><span class="p">]</span>
            <span class="c"># the input files are in BED format with 0-based start position</span>
            <span class="c"># we need to add 1 to the start position for consistency</span>
            <span class="n">output</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">&#39;{}</span><span class="se">\t</span><span class="s">{}</span><span class="se">\t</span><span class="s">{}</span><span class="se">\t</span><span class="s">{}</span><span class="se">\t</span><span class="s">{}</span><span class="se">\t</span><span class="s">{:.2f}</span><span class="se">\n</span><span class="s">&#39;</span>
                <span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">exon</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="nb">int</span><span class="p">(</span><span class="n">exon</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="n">exon</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span> <span class="n">exon</span><span class="p">[</span><span class="mi">3</span><span class="p">],</span> <span class="n">count</span><span class="p">,</span>
                        <span class="mf">1.e9</span><span class="o">*</span><span class="n">count</span><span class="o">/</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">mapped_reads</span><span class="p">}</span> <span class="o">*</span> <span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">exon</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span> <span class="o">-</span> <span class="nb">int</span><span class="p">(</span><span class="n">exon</span><span class="p">[</span><span class="mi">1</span><span class="p">])))))</span>



<span class="p">[</span><span class="n">update_blast</span><span class="p">:</span> <span class="kp">provides</span><span class="o">=</span><span class="s">&#39;${ncbi_resource_dir}/blast/db/human_genomic.nal&#39;</span><span class="p">]</span>
<span class="c">#Resource: Download NCBI Blast databases and link the directory</span>
<span class="c"># to output dir for blast to work</span>
<span class="kn">depends:</span>
    <span class="n">executable</span><span class="p">(</span><span class="s">&#39;update_blastdb.pl&#39;</span><span class="p">)</span>

<span class="kn">output:</span>
    <span class="s">&#39;${ncbi_resource_dir}/blast/db/human_genomic.nal&#39;</span><span class="p">,</span>
    <span class="s">&#39;${ncbi_resource_dir}/blast/db/nt.nal&#39;</span>

<span class="kn">task:</span>

<span class="kn">sh:</span> <span class="kp">workdir</span><span class="o">=</span><span class="s">&#39;${ncbi_resource_dir}/blast/db&#39;</span>
    <span class="n">update_blastdb</span><span class="o">.</span><span class="n">pl</span> <span class="n">human_genomic</span> <span class="o">--</span><span class="n">decompress</span> <span class="o">||</span> <span class="n">true</span>
    <span class="n">update_blastdb</span><span class="o">.</span><span class="n">pl</span> <span class="n">nt</span> <span class="o">--</span><span class="n">decompress</span> <span class="o">||</span> <span class="n">true</span>

<span class="kn">sh:</span> <span class="kp">workdir</span><span class="o">=</span><span class="n">output_dir</span>
    <span class="p">[</span> <span class="o">-</span><span class="n">d</span> <span class="n">blast</span> <span class="p">]</span> <span class="o">||</span> <span class="n">ln</span> <span class="o">-</span><span class="n">s</span> <span class="err">$</span><span class="p">{</span><span class="n">ncbi_resource_dir</span><span class="p">}</span><span class="o">/</span><span class="n">blast</span><span class="o">/</span><span class="n">db</span> <span class="n">blast</span>


<span class="p">[</span><span class="n">before_tophat_fusion</span><span class="p">:</span> <span class="kp">provides</span><span class="o">=</span><span class="p">[</span><span class="s">&#39;${output_dir}/refGene.txt&#39;</span><span class="p">,</span> <span class="s">&#39;${output_dir}/ensGene.txt&#39;</span><span class="p">]]</span>
<span class="c">#Fusion (prepare): Create an alias for tophat output directory because</span>
<span class="c">#    tophat-fusion-post require the directory to have name tophat-XXX</span>
<span class="c">#Fusion (prepare): Link refGene.txt and ensGene.txt to output directory to be used by tophat-fusion-post.</span>

<span class="kn">depends:</span>
    <span class="n">refgene_txt</span><span class="p">,</span> <span class="n">ensgene_txt</span>

<span class="kn">run:</span>    <span class="kp">workdir</span><span class="o">=</span><span class="n">output_dir</span>
    <span class="p">[</span> <span class="o">-</span><span class="n">f</span> <span class="n">refGene</span><span class="o">.</span><span class="n">txt</span> <span class="p">]</span> <span class="o">||</span> <span class="n">ln</span> <span class="o">-</span><span class="n">s</span> <span class="err">$</span><span class="p">{</span><span class="n">refgene_txt</span><span class="p">}</span> <span class="n">refGene</span><span class="o">.</span><span class="n">txt</span>
    <span class="p">[</span> <span class="o">-</span><span class="n">f</span> <span class="n">ensGene</span><span class="o">.</span><span class="n">txt</span> <span class="p">]</span> <span class="o">||</span> <span class="n">ln</span> <span class="o">-</span><span class="n">s</span> <span class="err">$</span><span class="p">{</span><span class="n">ensgene_txt</span><span class="p">}</span> <span class="n">ensGene</span><span class="o">.</span><span class="n">txt</span>

<span class="n">stop_if</span><span class="p">(</span><span class="n">alignment_out</span><span class="o">.</span><span class="n">rstrip</span><span class="p">(</span><span class="s">&#39;/&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">rsplit</span><span class="p">(</span><span class="s">&#39;/&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s">&#39;tophat_&#39;</span><span class="p">)</span> \
    <span class="ow">or</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isdir</span><span class="p">(</span><span class="n">output_dir</span> <span class="o">+</span> <span class="s">&#39;/tophat_&#39;</span> <span class="o">+</span> <span class="n">alignment_out</span><span class="o">.</span><span class="n">rsplit</span><span class="p">(</span><span class="s">&quot;/&quot;</span><span class="p">,</span><span class="mi">1</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]))</span>

<span class="kn">run:</span>    <span class="kp">workdir</span><span class="o">=</span><span class="n">output_dir</span>
    <span class="n">ln</span> <span class="o">-</span><span class="n">s</span> <span class="err">$</span><span class="p">{</span><span class="n">alignment_out</span><span class="o">.</span><span class="n">rstrip</span><span class="p">(</span><span class="s">&quot;/&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">rsplit</span><span class="p">(</span><span class="s">&quot;/&quot;</span><span class="p">,</span><span class="mi">1</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]}</span> <span class="err">$</span><span class="p">{</span><span class="s">&#39;tophat_&#39;</span> <span class="o">+</span> <span class="n">alignment_out</span><span class="o">.</span><span class="n">rsplit</span><span class="p">(</span><span class="s">&quot;/&quot;</span><span class="p">,</span><span class="mi">1</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]}</span>


<span class="p">[</span><span class="n">human_hg19_710</span><span class="p">]</span>
<span class="c"># tophat-fusion for fusion detection</span>
<span class="kn">output:</span>
    <span class="s">&quot;${fusion_out}/potential_fusion.txt&quot;</span>

<span class="kn">depends:</span>
    <span class="s">&#39;${output_dir}/refGene.txt&#39;</span><span class="p">,</span>
    <span class="s">&#39;${output_dir}/ensGene.txt&#39;</span><span class="p">,</span>
    <span class="s">&#39;${ncbi_resource_dir}/blast/db/human_genomic.nal&#39;</span>

<span class="kn">run:</span> <span class="kp">workdir</span><span class="o">=</span><span class="n">output_dir</span>

    <span class="n">tophat</span><span class="o">-</span><span class="n">fusion</span><span class="o">-</span><span class="n">post</span> <span class="o">-</span><span class="n">p</span> <span class="mi">3</span>  <span class="o">-</span><span class="n">o</span> <span class="err">$</span><span class="p">{</span><span class="n">fusion_out</span><span class="o">.</span><span class="n">rstrip</span><span class="p">(</span><span class="s">&quot;/&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">rsplit</span><span class="p">(</span><span class="s">&quot;/&quot;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]}</span> \
        <span class="err">$</span><span class="p">{</span><span class="n">reference_dir</span><span class="p">}</span><span class="o">/</span><span class="n">BowtieIndex</span><span class="o">/</span><span class="n">genome</span>


<span class="p">[</span><span class="n">human_hg19_730</span><span class="p">]</span>
<span class="c"># (Oncofuse): filter fusion candidates using Oncofuse</span>

<span class="kn">output:</span>
    <span class="s">&#39;${fusion_out}/fusions_oncofuse.xls&#39;</span>

<span class="kn">run:</span>
    <span class="n">java</span> <span class="o">-</span><span class="n">Xmx1G</span> <span class="o">-</span><span class="n">jar</span> <span class="err">$</span><span class="p">{</span><span class="n">oncofuse_path</span><span class="p">}</span><span class="o">/</span><span class="n">Oncofuse</span><span class="o">.</span><span class="n">jar</span> \
        <span class="err">$</span><span class="p">{</span><span class="nb">input</span><span class="p">}</span> <span class="n">tophat</span><span class="o">-</span><span class="n">post</span> <span class="n">AVG</span> <span class="err">$</span><span class="p">{</span><span class="n">fusion_out</span><span class="p">}</span><span class="o">/</span><span class="n">fusions_oncofuse</span><span class="o">.</span><span class="n">xls</span>

<span class="p">[</span><span class="n">download_gatk</span><span class="p">:</span> <span class="kp">provides</span><span class="o">=</span><span class="n">gatk_resource_dir</span> <span class="o">+</span> <span class="s">&#39;/dbsnp_138.hg19.vcf&#39;</span><span class="p">]</span>
<span class="kn">download:</span> <span class="n">dest_dir</span><span class="o">=</span><span class="n">gatk_resource_dir</span>
    <span class="err">$</span><span class="p">{</span><span class="n">gatk_url</span><span class="p">}</span><span class="o">/</span><span class="mi">1000</span><span class="n">G_omni2</span><span class="o">.</span><span class="mf">5.</span><span class="n">hg19</span><span class="o">.</span><span class="n">sites</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span>
    <span class="err">$</span><span class="p">{</span><span class="n">gatk_url</span><span class="p">}</span><span class="o">/</span><span class="mi">1000</span><span class="n">G_omni2</span><span class="o">.</span><span class="mf">5.</span><span class="n">hg19</span><span class="o">.</span><span class="n">sites</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span><span class="o">.</span><span class="n">md5</span>
    <span class="err">$</span><span class="p">{</span><span class="n">gatk_url</span><span class="p">}</span><span class="o">/</span><span class="mi">1000</span><span class="n">G_omni2</span><span class="o">.</span><span class="mf">5.</span><span class="n">hg19</span><span class="o">.</span><span class="n">sites</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">idx</span><span class="o">.</span><span class="n">gz</span>
    <span class="err">$</span><span class="p">{</span><span class="n">gatk_url</span><span class="p">}</span><span class="o">/</span><span class="mi">1000</span><span class="n">G_omni2</span><span class="o">.</span><span class="mf">5.</span><span class="n">hg19</span><span class="o">.</span><span class="n">sites</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">idx</span><span class="o">.</span><span class="n">gz</span><span class="o">.</span><span class="n">md5</span>
    <span class="err">$</span><span class="p">{</span><span class="n">gatk_url</span><span class="p">}</span><span class="o">/</span><span class="n">dbsnp_138</span><span class="o">.</span><span class="n">hg19</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span>
    <span class="err">$</span><span class="p">{</span><span class="n">gatk_url</span><span class="p">}</span><span class="o">/</span><span class="n">dbsnp_138</span><span class="o">.</span><span class="n">hg19</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span><span class="o">.</span><span class="n">md5</span>
    <span class="err">$</span><span class="p">{</span><span class="n">gatk_url</span><span class="p">}</span><span class="o">/</span><span class="n">dbsnp_138</span><span class="o">.</span><span class="n">hg19</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">idx</span><span class="o">.</span><span class="n">gz</span>
    <span class="err">$</span><span class="p">{</span><span class="n">gatk_url</span><span class="p">}</span><span class="o">/</span><span class="n">dbsnp_138</span><span class="o">.</span><span class="n">hg19</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">idx</span><span class="o">.</span><span class="n">gz</span><span class="o">.</span><span class="n">md5</span>
    <span class="err">$</span><span class="p">{</span><span class="n">gatk_url</span><span class="p">}</span><span class="o">/</span><span class="n">hapmap_3</span><span class="o">.</span><span class="mf">3.</span><span class="n">hg19</span><span class="o">.</span><span class="n">sites</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span>
    <span class="err">$</span><span class="p">{</span><span class="n">gatk_url</span><span class="p">}</span><span class="o">/</span><span class="n">hapmap_3</span><span class="o">.</span><span class="mf">3.</span><span class="n">hg19</span><span class="o">.</span><span class="n">sites</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span><span class="o">.</span><span class="n">md5</span>
    <span class="err">$</span><span class="p">{</span><span class="n">gatk_url</span><span class="p">}</span><span class="o">/</span><span class="n">hapmap_3</span><span class="o">.</span><span class="mf">3.</span><span class="n">hg19</span><span class="o">.</span><span class="n">sites</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">idx</span><span class="o">.</span><span class="n">gz</span>
    <span class="err">$</span><span class="p">{</span><span class="n">gatk_url</span><span class="p">}</span><span class="o">/</span><span class="n">hapmap_3</span><span class="o">.</span><span class="mf">3.</span><span class="n">hg19</span><span class="o">.</span><span class="n">sites</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">idx</span><span class="o">.</span><span class="n">gz</span><span class="o">.</span><span class="n">md5</span>

<span class="p">[</span><span class="n">human_hg19_800</span> <span class="p">(</span><span class="n">GATK</span> <span class="n">UnifiedGenotyper</span><span class="p">)]</span>
<span class="c"># Use GATK UnifiedGenotyper caller to call variants</span>
<span class="kn">input:</span>
    <span class="n">sorted_unique</span>

<span class="kn">depends:</span>
    <span class="s">&#39;${input}.bai&#39;</span><span class="p">,</span>
    <span class="n">sos_variable</span><span class="p">(</span><span class="s">&#39;mapped_reads&#39;</span><span class="p">)</span>

<span class="c"># An option -L can be used to limit the variant call to targeted regions</span>
<span class="c"># -L $ref_dir/hg19.master.gene.bed </span>

<span class="kn">output:</span>
    <span class="s">&#39;${variant_out}/${samplename}-gatk.vcf&#39;</span>

<span class="kn">run:</span>
    <span class="n">java</span> <span class="o">-</span><span class="n">Xmx16g</span> <span class="o">-</span><span class="n">Xms512m</span> <span class="o">-</span><span class="n">Djava</span><span class="o">.</span><span class="n">io</span><span class="o">.</span><span class="n">tmpdir</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="n">temp_dir</span><span class="p">}</span>  \
        <span class="o">-</span><span class="n">jar</span> <span class="err">$</span><span class="p">{</span><span class="n">gatk_path</span><span class="p">}</span><span class="o">/</span><span class="n">GenomeAnalysisTK</span><span class="o">.</span><span class="n">jar</span>   \
        <span class="o">-</span><span class="n">R</span> <span class="err">$</span><span class="p">{</span><span class="n">reference_dir</span><span class="p">}</span><span class="o">/</span><span class="n">WholeGenomeFasta</span><span class="o">/</span><span class="n">genome</span><span class="o">.</span><span class="n">fa</span>   \
        <span class="o">--</span><span class="n">filter_reads_with_N_cigar</span>   \
        <span class="o">-</span><span class="n">T</span> <span class="n">UnifiedGenotyper</span>   \
        <span class="o">-</span><span class="n">I</span> <span class="err">$</span><span class="p">{</span><span class="nb">input</span><span class="p">}</span>  \
        <span class="o">--</span><span class="n">out</span> <span class="err">$</span><span class="p">{</span><span class="n">variant_out</span><span class="p">}</span><span class="o">/</span><span class="err">$</span><span class="p">{</span><span class="n">samplename</span><span class="p">}</span><span class="o">-</span><span class="n">gatk</span><span class="o">.</span><span class="n">vcf</span>  

<span class="p">[</span><span class="n">human_hg19_810</span> <span class="p">(</span><span class="n">Recalibrate</span> <span class="n">Variants</span><span class="p">)]</span>
<span class="c"># recalibrate and filter variants</span>
<span class="kn">output:</span>
    <span class="s">&#39;${variant_out}/${samplename}-filter.vcf&#39;</span>

<span class="kn">depends:</span>
    <span class="s">&#39;${reference_dir}/WholeGenomeFasta/genome.fa&#39;</span><span class="p">,</span>
    <span class="s">&#39;${gatk_resource_dir}/hapmap_3.3.hg19.sites.vcf&#39;</span><span class="p">,</span>
    <span class="s">&#39;${gatk_resource_dir}/1000G_omni2.5.hg19.sites.vcf&#39;</span><span class="p">,</span>
    <span class="s">&#39;${gatk_resource_dir}/dbsnp_138.hg19.vcf&#39;</span>

<span class="kn">run:</span>
    <span class="n">java</span> <span class="o">-</span><span class="n">Xmx16g</span> <span class="o">-</span><span class="n">Xms512m</span>  \
        <span class="o">-</span><span class="n">Djava</span><span class="o">.</span><span class="n">io</span><span class="o">.</span><span class="n">tmpdir</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="n">temp_dir</span><span class="p">}</span>  \
        <span class="o">-</span><span class="n">jar</span> <span class="err">$</span><span class="p">{</span><span class="n">gatk_path</span><span class="p">}</span><span class="o">/</span><span class="n">GenomeAnalysisTK</span><span class="o">.</span><span class="n">jar</span>   \
        <span class="o">-</span><span class="n">R</span> <span class="err">$</span><span class="p">{</span><span class="n">reference_dir</span><span class="p">}</span><span class="o">/</span><span class="n">WholeGenomeFasta</span><span class="o">/</span><span class="n">genome</span><span class="o">.</span><span class="n">fa</span>   \
        <span class="o">-</span><span class="n">T</span> <span class="n">VariantRecalibrator</span>  \
        <span class="o">-</span><span class="n">mode</span> <span class="n">SNP</span>   \
        <span class="o">-</span><span class="nb">input</span> <span class="err">$</span><span class="p">{</span><span class="nb">input</span><span class="p">}</span>  \
        <span class="o">-</span><span class="n">resource</span><span class="p">:</span><span class="n">hapmap</span><span class="p">,</span><span class="n">known</span><span class="o">=</span><span class="n">false</span><span class="p">,</span><span class="n">training</span><span class="o">=</span><span class="n">true</span><span class="p">,</span><span class="n">truth</span><span class="o">=</span><span class="n">true</span><span class="p">,</span><span class="n">prior</span><span class="o">=</span><span class="mf">15.0</span> <span class="err">$</span><span class="p">{</span><span class="n">gatk_resource_dir</span><span class="p">}</span><span class="o">/</span><span class="n">hapmap_3</span><span class="o">.</span><span class="mf">3.</span><span class="n">hg19</span><span class="o">.</span><span class="n">sites</span><span class="o">.</span><span class="n">vcf</span>  \
        <span class="o">-</span><span class="n">resource</span><span class="p">:</span><span class="n">omni</span><span class="p">,</span><span class="n">known</span><span class="o">=</span><span class="n">false</span><span class="p">,</span><span class="n">training</span><span class="o">=</span><span class="n">true</span><span class="p">,</span><span class="n">truth</span><span class="o">=</span><span class="n">false</span><span class="p">,</span><span class="n">prior</span><span class="o">=</span><span class="mf">12.0</span>  <span class="err">$</span><span class="p">{</span><span class="n">gatk_resource_dir</span><span class="p">}</span><span class="o">/</span><span class="mi">1000</span><span class="n">G_omni2</span><span class="o">.</span><span class="mf">5.</span><span class="n">hg19</span><span class="o">.</span><span class="n">sites</span><span class="o">.</span><span class="n">vcf</span>  \
        <span class="o">-</span><span class="n">resource</span><span class="p">:</span><span class="n">dbsnp</span><span class="p">,</span><span class="n">known</span><span class="o">=</span><span class="n">true</span><span class="p">,</span><span class="n">training</span><span class="o">=</span><span class="n">false</span><span class="p">,</span><span class="n">truth</span><span class="o">=</span><span class="n">false</span><span class="p">,</span><span class="n">prior</span><span class="o">=</span><span class="mf">2.0</span>  <span class="err">$</span><span class="p">{</span><span class="n">gatk_resource_dir</span><span class="p">}</span><span class="o">/</span><span class="n">dbsnp_138</span><span class="o">.</span><span class="n">hg19</span><span class="o">.</span><span class="n">vcf</span>  \
        <span class="o">-</span><span class="n">an</span> <span class="n">ReadPosRankSum</span> <span class="o">-</span><span class="n">an</span> <span class="n">FS</span>   \
        <span class="o">-</span><span class="n">recalFile</span> <span class="err">$</span><span class="p">{</span><span class="n">variant_out</span><span class="p">}</span><span class="o">/</span><span class="err">$</span><span class="p">{</span><span class="n">samplename</span><span class="p">}</span><span class="o">.</span><span class="n">recal</span>   \
        <span class="o">-</span><span class="n">tranchesFile</span> <span class="err">$</span><span class="p">{</span><span class="n">variant_out</span><span class="p">}</span><span class="o">/</span><span class="err">$</span><span class="p">{</span><span class="n">samplename</span><span class="p">}</span><span class="o">.</span><span class="n">tranches</span>   \
        <span class="o">--</span><span class="n">maxGaussians</span> <span class="err">$</span><span class="p">{</span><span class="mi">4</span> <span class="k">if</span> <span class="n">mapped_reads</span> <span class="o">&gt;</span> <span class="mi">4000000</span> <span class="k">else</span> <span class="mi">2</span><span class="p">}</span> <span class="o">||</span> <span class="n">true</span>  \

<span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">getsize</span><span class="p">(</span><span class="s">&#39;${variant_out}/${samplename}.recal&#39;</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
    <span class="n">run</span><span class="p">(</span><span class="s">r&#39;&#39;&#39;</span>
<span class="s">        java -Xmx16g -Xms512m   \</span>
<span class="s">            -Djava.io.tmpdir=${temp_dir}   \</span>
<span class="s">            -jar ${gatk_path}/GenomeAnalysisTK.jar   \</span>
<span class="s">            -R ${reference_dir}/WholeGenomeFasta/genome.fa  \</span>
<span class="s">            -T ApplyRecalibration   \</span>
<span class="s">            -mode SNP   \</span>
<span class="s">            -input ${input}   \</span>
<span class="s">            -recalFile ${variant_out}/${samplename}.recal  \</span>
<span class="s">            -tranchesFile ${variant_out}/${samplename}.tranches  \</span>
<span class="s">            -o ${variant_out}/${samplename}-filter.vcf   \</span>
<span class="s">            --ts_filter_level 99.0</span>
<span class="s">    &#39;&#39;&#39;</span><span class="p">)</span>
<span class="kn">else:</span>
    <span class="n">run</span><span class="p">(</span><span class="s">r&#39;&#39;&#39;</span>
<span class="s">        java -Xmx16g -Xms512m  \</span>
<span class="s">            -Djava.io.tmpdir=${temp_dir}  \</span>
<span class="s">            -jar ${gatk_path}/GenomeAnalysisTK.jar  \</span>
<span class="s">            -R ${reference_dir}/WholeGenomeFasta/genome.fa  \</span>
<span class="s">            -V ${input}  \</span>
<span class="s">            -l INFO -T VariantFiltration  \</span>
<span class="s">            --filterExpression &quot;FS &gt; 20.0&quot; --filterName FSFilter  \</span>
<span class="s">            --filterExpression &quot;ED &gt; 5&quot; --filterName EDFilter  \</span>
<span class="s">            --filterExpression &quot;ReadPosRankSum &lt; -8.0&quot; --filterName RPRSFilter  \</span>
<span class="s">            --filterExpression &quot;ReadPosRankSum &gt; -8.0&quot; --filterName RPRSFilter  \</span>
<span class="s">            -o ${variant_out}/${samplename}-filter.vcf</span>
<span class="s">    &#39;&#39;&#39;</span><span class="p">)</span>

<span class="p">[</span><span class="n">human_hg19_900</span><span class="p">]</span>
<span class="c">#RSeQC: Generate JPEG versions of the RSeQC plots to be embedded in final report </span>
<span class="kn">input:</span>
    <span class="n">rseqc_pdf</span>

<span class="kn">output:</span>
    <span class="s">&#39;${alignmentqc_out}/${samplename}.geneBodyCoverage.curves.jpg&#39;</span><span class="p">,</span>
    <span class="s">&#39;${alignmentqc_out}/${samplename}.junctionSaturation_plot.jpg&#39;</span><span class="p">,</span>
    <span class="s">&#39;${alignmentqc_out}/${samplename}.splice_events.jpg&#39;</span><span class="p">,</span>
    <span class="s">&#39;${alignmentqc_out}/${samplename}.splice_junction.jpg&#39;</span><span class="p">,</span>
    <span class="s">&#39;${alignmentqc_out}/${samplename}.inner_distance_plot.jpg&#39;</span><span class="p">,</span>
    <span class="s">&#39;${alignmentqc_out}//${samplename}.DupRate_plot.jpg&#39;</span>

<span class="c"># we need to wait the completion of geneBody_coverage</span>
<span class="kn">run:</span>
    <span class="n">cat</span> <span class="err">$</span><span class="p">{</span><span class="s">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">alignmentqc_out</span><span class="p">,</span> <span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="n">alignmentqc_out</span><span class="p">)</span> <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s">&#39;.r&#39;</span><span class="p">)])}</span>  \
    <span class="o">|</span> <span class="n">sed</span> <span class="s">&#39;s/pdf(\(.*\).pdf\(.\)/jpeg(</span><span class="se">\1</span><span class="s">.jpg</span><span class="se">\2</span><span class="s">, width=600, height=600/&#39;</span> <span class="o">|</span> <span class="n">R</span> <span class="o">--</span><span class="n">no</span><span class="o">-</span><span class="n">save</span>


<span class="p">[</span><span class="n">human_hg19_120</span><span class="p">:</span> <span class="kp">shared</span><span class="o">=</span><span class="s">&#39;read_length&#39;</span><span class="p">]</span>

<span class="kn">input:</span>
   <span class="n">fastq_files</span>

<span class="kn">import</span> <span class="nn">gzip</span>
<span class="k">with</span> <span class="n">gzip</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="nb">input</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="s">&#39;rb&#39;</span><span class="p">)</span> <span class="k">if</span> <span class="nb">input</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s">&#39;.gz&#39;</span><span class="p">)</span> <span class="k">else</span> <span class="nb">open</span><span class="p">(</span><span class="nb">input</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="s">&#39;rb&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fastq</span><span class="p">:</span>
    <span class="c"># skip the first line</span>
    <span class="n">count</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">total_length</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">for</span> <span class="n">index</span><span class="p">,</span><span class="n">line</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">fastq</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">index</span> <span class="o">%</span> <span class="mi">4</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
            <span class="c"># get the second line</span>
            <span class="n">total_length</span> <span class="o">+=</span> <span class="nb">len</span><span class="p">(</span><span class="n">line</span><span class="o">.</span><span class="n">decode</span><span class="p">()</span><span class="o">.</span><span class="n">strip</span><span class="p">())</span>
            <span class="n">count</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="k">if</span> <span class="n">count</span> <span class="o">==</span> <span class="mi">100</span><span class="p">:</span>
            <span class="k">break</span>
    <span class="n">read_length</span> <span class="o">=</span> <span class="n">total_length</span> <span class="o">/</span> <span class="n">count</span>



<span class="p">[</span><span class="n">alignment_summary</span><span class="p">:</span> <span class="kp">shared</span><span class="o">=</span><span class="p">(</span><span class="s">&#39;total_reads&#39;</span><span class="p">,</span> <span class="s">&#39;used_reads&#39;</span><span class="p">,</span> <span class="s">&#39;mapped_reads_perc&#39;</span><span class="p">,</span> <span class="s">&#39;uniq_mapped_reads&#39;</span><span class="p">,</span> <span class="s">&#39;uniq_mapped_reads_perc&#39;</span><span class="p">,</span> <span class="s">&#39;mapped_pairs&#39;</span><span class="p">,</span> <span class="s">&#39;mapped_pairs_perc&#39;</span><span class="p">,</span> <span class="s">&#39;concordant_mapped_pairs&#39;</span><span class="p">,</span> <span class="s">&#39;concordant_mapped_pairs_perc&#39;</span><span class="p">,</span> <span class="s">&#39;genome_mapped_count&#39;</span><span class="p">,</span> <span class="s">&#39;genome_mapped_perc&#39;</span><span class="p">,</span> <span class="s">&#39;junction_mapped_count&#39;</span><span class="p">,</span> <span class="s">&#39;junction_mapped_perc&#39;</span><span class="p">)]</span>

<span class="kn">depends:</span>
    <span class="n">sos_variable</span><span class="p">(</span><span class="s">&#39;accepted_hits&#39;</span><span class="p">)</span>

<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s">&#39;${alignment_out}/prep_reads.info&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">prep_reads</span><span class="p">:</span>
    <span class="n">total_reads</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">used_reads</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">prep_reads</span><span class="p">:</span>
        <span class="k">if</span> <span class="s">&#39;reads_in&#39;</span> <span class="ow">in</span> <span class="n">line</span><span class="p">:</span>
            <span class="n">total_reads</span> <span class="o">+=</span> <span class="nb">int</span><span class="p">(</span><span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&#39;=&#39;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
        <span class="k">if</span> <span class="s">&#39;reads_out&#39;</span> <span class="ow">in</span> <span class="n">line</span><span class="p">:</span>
            <span class="n">used_reads</span> <span class="o">+=</span> <span class="nb">int</span><span class="p">(</span><span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&#39;=&#39;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
<span class="c">#</span>
<span class="c"># get mapped reads from align_summary.txt</span>
<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s">&#39;${alignment_out}/align_summary.txt&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">align_summary</span><span class="p">:</span>
    <span class="n">uniq_mapped_reads</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">uniq_mapped_pairs</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">mapped_reads</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">mapped_pairs</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">concordant_mapped_pairs</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">line</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">align_summary</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">idx</span> <span class="ow">in</span> <span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">6</span><span class="p">):</span>
            <span class="c">#  Mapped   :    770754 (77.1% of input)</span>
            <span class="n">mapped_reads</span> <span class="o">+=</span> <span class="nb">int</span><span class="p">(</span><span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&#39;:&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&#39;(&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">strip</span><span class="p">())</span>
        <span class="k">if</span> <span class="n">idx</span> <span class="o">==</span> <span class="mi">10</span><span class="p">:</span>
            <span class="c"># Aligned pairs:</span>
            <span class="n">mapped_pairs</span> <span class="o">+=</span> <span class="nb">int</span><span class="p">(</span><span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&#39;:&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">strip</span><span class="p">())</span>
        <span class="k">if</span> <span class="n">idx</span> <span class="ow">in</span> <span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">7</span><span class="p">):</span>
            <span class="n">uniq_mapped_reads</span> <span class="o">+=</span> <span class="nb">int</span><span class="p">(</span><span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&#39;:&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s">&#39;(&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">strip</span><span class="p">())</span>
        <span class="k">if</span> <span class="n">idx</span> <span class="o">==</span> <span class="mi">12</span><span class="p">:</span>
            <span class="n">concordant_mapped_pairs</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">()[</span><span class="mi">0</span><span class="p">])</span>
    <span class="n">uniq_mapped_reads</span> <span class="o">=</span> <span class="n">mapped_reads</span> <span class="o">-</span> <span class="n">uniq_mapped_reads</span>
    <span class="n">uniq_mapped_reads_perc</span> <span class="o">=</span> <span class="n">uniq_mapped_reads</span> <span class="o">*</span> <span class="mf">100.</span> <span class="o">/</span> <span class="n">total_reads</span>
    <span class="n">mapped_reads_perc</span> <span class="o">=</span> <span class="n">mapped_reads</span> <span class="o">*</span> <span class="mf">100.</span> <span class="o">/</span> <span class="n">total_reads</span>
    <span class="n">concordant_mapped_pairs</span> <span class="o">=</span> <span class="n">mapped_pairs</span> <span class="o">-</span> <span class="n">concordant_mapped_pairs</span>
    <span class="n">concordant_mapped_pairs_perc</span> <span class="o">=</span> <span class="n">concordant_mapped_pairs</span> <span class="o">*</span> <span class="mf">200.0</span> <span class="o">/</span> <span class="n">total_reads</span>
    <span class="n">mapped_pairs_perc</span> <span class="o">=</span> <span class="n">mapped_pairs</span> <span class="o">*</span> <span class="mf">200.0</span> <span class="o">/</span> <span class="n">total_reads</span>

<span class="c"># get junction reads</span>
<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s">&#39;${alignment_out}/junction.count&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">junction_count</span><span class="p">:</span>
    <span class="n">junction_mapped_count</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">junction_count</span><span class="o">.</span><span class="n">readline</span><span class="p">()</span><span class="o">.</span><span class="n">strip</span><span class="p">())</span>
    <span class="n">genome_mapped_count</span> <span class="o">=</span> <span class="n">mapped_reads</span> <span class="o">-</span> <span class="n">junction_mapped_count</span>
    <span class="n">junction_mapped_perc</span> <span class="o">=</span> <span class="n">junction_mapped_count</span> <span class="o">*</span> <span class="mf">100.</span> <span class="o">/</span> <span class="n">total_reads</span>
    <span class="n">genome_mapped_perc</span> <span class="o">=</span> <span class="n">genome_mapped_count</span> <span class="o">*</span> <span class="mf">100.</span> <span class="o">/</span> <span class="n">total_reads</span>

<span class="n">uniq_mapped_reads</span> <span class="o">=</span> <span class="n">mapped_reads</span> <span class="o">-</span> <span class="n">uniq_mapped_reads</span>
<span class="n">uniq_mapped_reads_perc</span> <span class="o">=</span> <span class="n">uniq_mapped_reads</span> <span class="o">*</span> <span class="mf">100.</span> <span class="o">/</span> <span class="n">total_reads</span>
<span class="n">mapped_reads_perc</span> <span class="o">=</span> <span class="n">mapped_reads</span> <span class="o">*</span> <span class="mf">100.</span> <span class="o">/</span> <span class="n">total_reads</span>
<span class="n">concordant_mapped_pairs</span> <span class="o">=</span> <span class="n">mapped_pairs</span> <span class="o">-</span> <span class="n">concordant_mapped_pairs</span>
<span class="n">concordant_mapped_pairs_perc</span> <span class="o">=</span> <span class="n">concordant_mapped_pairs</span> <span class="o">*</span> <span class="mf">200.0</span> <span class="o">/</span> <span class="n">total_reads</span>
<span class="n">mapped_pairs_perc</span> <span class="o">=</span> <span class="n">mapped_pairs</span> <span class="o">*</span> <span class="mf">200.0</span> <span class="o">/</span> <span class="n">total_reads</span>
<span class="n">genome_mapped_count</span> <span class="o">=</span> <span class="n">mapped_reads</span> <span class="o">-</span> <span class="n">junction_mapped_count</span>
<span class="n">junction_mapped_perc</span> <span class="o">=</span> <span class="n">junction_mapped_count</span> <span class="o">*</span> <span class="mf">100.</span> <span class="o">/</span> <span class="n">total_reads</span>
<span class="n">genome_mapped_perc</span> <span class="o">=</span> <span class="n">genome_mapped_count</span> <span class="o">*</span> <span class="mf">100.</span> <span class="o">/</span> <span class="n">total_reads</span>


<span class="p">[</span><span class="n">mean_depth</span><span class="p">:</span> <span class="kp">shared</span><span class="o">=</span><span class="s">&#39;mean_depth&#39;</span><span class="p">]</span>

<span class="kn">input:</span>
    <span class="s">&#39;${count_out}/${samplename}-exon_counts_rpkm.xls&#39;</span>

<span class="kn">depends:</span>
    <span class="n">sos_variable</span><span class="p">(</span><span class="s">&#39;read_length&#39;</span><span class="p">)</span>

<span class="n">mean_depth</span> <span class="o">=</span> <span class="mi">0</span>
<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="nb">input</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="k">as</span> <span class="n">exons</span><span class="p">:</span>
    <span class="n">exons</span><span class="o">.</span><span class="n">readline</span><span class="p">()</span>
    <span class="n">total_reads</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">total_length</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">exons</span><span class="p">:</span>
        <span class="nb">chr</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="n">count</span><span class="p">,</span> <span class="n">rpkm</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
        <span class="k">if</span> <span class="nb">int</span><span class="p">(</span><span class="n">count</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">2</span><span class="p">:</span>
            <span class="k">continue</span>
        <span class="n">total_reads</span> <span class="o">+=</span> <span class="nb">int</span><span class="p">(</span><span class="n">count</span><span class="p">)</span>
        <span class="n">total_length</span> <span class="o">+=</span> <span class="nb">int</span><span class="p">(</span><span class="n">end</span><span class="p">)</span> <span class="o">-</span> <span class="nb">int</span><span class="p">(</span><span class="n">start</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span>
    <span class="n">mean_depth</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">total_reads</span><span class="p">)</span> <span class="o">*</span> <span class="nb">int</span><span class="p">(</span><span class="n">read_length</span><span class="p">)</span> <span class="o">/</span> <span class="n">total_length</span>


<span class="p">[</span><span class="n">get_counts</span><span class="p">:</span> <span class="kp">shared</span><span class="o">=</span><span class="p">(</span><span class="s">&#39;num_genes&#39;</span><span class="p">,</span> <span class="s">&#39;num_covered_genes&#39;</span><span class="p">,</span> <span class="s">&#39;num_exons&#39;</span><span class="p">,</span> <span class="s">&#39;num_variants&#39;</span><span class="p">,</span> <span class="s">&#39;num_valid_variants&#39;</span><span class="p">,</span> <span class="s">&#39;num_oncofuse_fusions&#39;</span><span class="p">)]</span>

<span class="kn">depends:</span>
    <span class="s">&#39;${count_out}/${samplename}-gene_counts_rpkm.xls&#39;</span><span class="p">,</span>
    <span class="s">&#39;${count_out}/${samplename}-exon_counts_rpkm.xls&#39;</span><span class="p">,</span>
    <span class="s">&#39;${fusion_out}/fusions_oncofuse.xls&#39;</span><span class="p">,</span>
    <span class="s">&#39;${variant_out}/${samplename}-filter.vcf&#39;</span>

<span class="k">def</span> <span class="nf">_lineCount</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">matching</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>
    <span class="n">count</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">count_no_comment</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">count_matching</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="s">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">ifile</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">ifile</span><span class="p">:</span>
            <span class="n">count</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s">&#39;#&#39;</span><span class="p">):</span>
                <span class="n">count_no_comment</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="k">if</span> <span class="n">matching</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">()[</span><span class="n">matching</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="o">==</span> <span class="n">matching</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span>
                        <span class="n">count_matching</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="k">except</span><span class="p">:</span>
                    <span class="k">pass</span>
    <span class="k">return</span> <span class="n">count</span><span class="p">,</span> <span class="n">count_no_comment</span><span class="p">,</span> <span class="n">count_matching</span>

<span class="n">gene_count</span> <span class="o">=</span> <span class="n">_lineCount</span><span class="p">(</span><span class="s">&#39;${count_out}/${samplename}-gene_counts_rpkm.xls&#39;</span><span class="p">,</span> <span class="n">matching</span><span class="o">=</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="s">&#39;0&#39;</span><span class="p">))</span>
<span class="n">num_genes</span> <span class="o">=</span> <span class="n">gene_count</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">-</span> <span class="mi">1</span>
<span class="c"># all lines - 0 count</span>
<span class="n">num_covered_genes</span> <span class="o">=</span> <span class="n">gene_count</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">-</span> <span class="n">gene_count</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="o">-</span> <span class="mi">1</span>
<span class="n">num_exons</span> <span class="o">=</span> <span class="n">_lineCount</span><span class="p">(</span><span class="s">&#39;${count_out}/${samplename}-exon_counts_rpkm.xls&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">-</span> <span class="mi">1</span>

<span class="n">num_oncofuse_fusions</span> <span class="o">=</span> <span class="n">_lineCount</span><span class="p">(</span><span class="s">&#39;${fusion_out}/fusions_oncofuse.xls&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">-</span> <span class="mi">1</span>
<span class="n">variant_count</span> <span class="o">=</span> <span class="n">_lineCount</span><span class="p">(</span><span class="s">&#39;${variant_out}/${samplename}-filter.vcf&#39;</span><span class="p">,</span> <span class="n">matching</span><span class="o">=</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="s">&#39;PASS&#39;</span><span class="p">))</span>

<span class="n">num_variants</span> <span class="o">=</span> <span class="n">variant_count</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<span class="n">num_valid_variants</span> <span class="o">=</span> <span class="n">variant_count</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>

<span class="p">[</span><span class="n">human_hg19_910</span><span class="p">]</span>
<span class="c">#Summary: Write summary.html to report all results.</span>

<span class="kn">output:</span>
    <span class="s">&#39;${output_dir}/summary.html&#39;</span>

<span class="kn">depends:</span>
    <span class="p">[</span><span class="n">sos_variable</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="p">(</span>
        <span class="s">&#39;read_length&#39;</span><span class="p">,</span>
        <span class="s">&#39;mean_depth&#39;</span><span class="p">,</span>
        <span class="s">&#39;mate_inner_dist&#39;</span><span class="p">,</span>
        <span class="s">&#39;mate_std_dev&#39;</span><span class="p">,</span>
        <span class="s">&#39;total_reads&#39;</span><span class="p">,</span>
        <span class="s">&#39;used_reads&#39;</span><span class="p">,</span>
        <span class="s">&#39;mapped_reads&#39;</span><span class="p">,</span>
        <span class="s">&#39;mapped_reads_perc&#39;</span><span class="p">,</span>
        <span class="s">&#39;uniq_mapped_reads&#39;</span><span class="p">,</span>
        <span class="s">&#39;uniq_mapped_reads_perc&#39;</span><span class="p">,</span>
        <span class="s">&#39;mapped_pairs&#39;</span><span class="p">,</span>
        <span class="s">&#39;mapped_pairs_perc&#39;</span><span class="p">,</span>
        <span class="s">&#39;concordant_mapped_pairs&#39;</span><span class="p">,</span>
        <span class="s">&#39;concordant_mapped_pairs_perc&#39;</span><span class="p">,</span>
        <span class="s">&#39;genome_mapped_count&#39;</span><span class="p">,</span>
        <span class="s">&#39;genome_mapped_perc&#39;</span><span class="p">,</span>
        <span class="s">&#39;junction_mapped_count&#39;</span><span class="p">,</span>
        <span class="s">&#39;junction_mapped_perc&#39;</span><span class="p">,</span>
        <span class="s">&#39;reference_genes&#39;</span><span class="p">,</span>
        <span class="s">&#39;reference_genome&#39;</span><span class="p">,</span>
        <span class="s">&#39;num_genes&#39;</span><span class="p">,</span>
        <span class="s">&#39;num_covered_genes&#39;</span><span class="p">,</span>
        <span class="s">&#39;num_exons&#39;</span><span class="p">,</span>
        <span class="s">&#39;num_variants&#39;</span><span class="p">,</span>
        <span class="s">&#39;num_valid_variants&#39;</span><span class="p">,</span>
        <span class="s">&#39;num_oncofuse_fusions&#39;</span><span class="p">)]</span>

<span class="kn">parameter:</span> <span class="n">css</span> <span class="o">=</span> <span class="s">&#39;default.css&#39;</span>

<span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">css</span><span class="p">):</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s">&#39;{} does not exist. No CSS is used in HTML output&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">css</span><span class="p">))</span>
    <span class="n">css_style</span> <span class="o">=</span> <span class="s">&#39;&#39;</span>
<span class="kn">else:</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">css</span><span class="p">)</span> <span class="k">as</span> <span class="n">style</span><span class="p">:</span>
        <span class="n">css_style</span> <span class="o">=</span> <span class="n">style</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>

<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">output</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="s">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">summary</span><span class="p">:</span>
    <span class="n">summary</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s">r&#39;&#39;&#39;</span>
<span class="s">&lt;!DOCTYPE html PUBLIC &quot;-//W3C//DTD HTML 4.01 Transitional//EN&quot;&gt;</span>
<span class="s">&lt;html&gt;</span>
<span class="s">&lt;head&gt;</span>
<span class="s">  &lt;meta content=&quot;text/html; charset=ISO-8859-1&quot;</span>
<span class="s">    http-equiv=&quot;content-type&quot;&gt;</span>
<span class="s">  &lt;title&gt;${samplename}&lt;/title&gt;</span>
<span class="s"> &lt;style type=&quot;text/css&quot;&gt;</span>
<span class="s"> ${css_style}</span>
<span class="s"> &lt;/style&gt;</span>
<span class="s">&lt;/head&gt;</span>
<span class="s">&lt;body&gt;</span>
<span class="s">&lt;h1&gt;RNA Seq analysis for sample ${samplename}&lt;/h1&gt;</span>
<span class="s">&lt;p&gt;Generated by pipeline ${pipeline_name} ${pipeline_version}&lt;/p&gt;</span>

<span class="s">&lt;h2&gt;Project description&lt;/h2&gt;</span>

<span class="s">The goal of the analysis is to evaluate how well RNA Seq sequencing worked on the sample,</span>
<span class="s">obtain gene and exon expression, fusion candidates, and variant calls from the sample.</span>

<span class="s">&lt;h3&gt;Basic information&lt;/h3&gt;</span>

<span class="s">&lt;table class=&quot;gridtable&quot;&gt;</span>
<span class="s">&lt;tr&gt;</span>
<span class="s">&lt;th&gt;Sample name&lt;/th&gt;&lt;td&gt;${samplename}&lt;/td&gt;</span>
<span class="s">&lt;/tr&gt;</span>
<span class="s">&lt;tr&gt;</span>
<span class="s">&lt;th&gt;Type&lt;/th&gt;&lt;td&gt;Paired ends&lt;/td&gt;</span>
<span class="s">&lt;/tr&gt;</span>
<span class="s">&lt;tr&gt;</span>
<span class="s">&lt;th&gt;Read length&lt;/th&gt;&lt;td&gt;${read_length}&lt;/td&gt;</span>
<span class="s">&lt;/tr&gt;</span>
<span class="s">&lt;tr&gt;</span>
<span class="s">&lt;th&gt;Mean Depth of Coverage &lt;sup&gt;*&lt;/sup&gt;&lt;/th&gt;&lt;td&gt;${mean_depth:.1f} &lt;/td&gt;</span>
<span class="s">&lt;/tr&gt;</span>
<span class="s">&lt;tr&gt;</span>
<span class="s">&lt;th&gt;Inner distance between paired reads&lt;sup&gt;**&lt;/sup&gt;&lt;/th&gt;&lt;td&gt;${mate_inner_dist} (stdev=${mate_std_dev})&lt;/td&gt;</span>
<span class="s">&lt;/tr&gt;</span>
<span class="s">&lt;/table&gt;</span>

<span class="s">&lt;p&gt;* The mean depth of coverage is roughly estimated by number of mapped reads times read length and divided by the total length of exons. </span>
<span class="s">A low depth of coverage might appear if the mapping rate is low.&lt;/p&gt;</span>
<span class="s">&lt;p&gt;** Mean and standard deviation of inner distance is estimated by distances between 1M pairs of aligned reads.&lt;/p&gt;</span>

<span class="s">&lt;h2&gt;Result summary&lt;/h2&gt;</span>

<span class="s">&lt;h3&gt;Read QC -- FastQC report&lt;/h3&gt;</span>
<span class="s">&lt;p&gt;&lt;a href=&quot;http://www.bioinformatics.babraham.ac.uk/projects/fastqc/&quot;&gt;FastQC&lt;/a&gt; is a quality control tool for high throughput sequence data. </span>
<span class="s">Please check &lt;a href=&quot;http://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/&quot;&gt;here&lt;/a&gt; for details of each plot.</span>
<span class="s">Note that it is normal for RNA Seq samples to have high duplication level and uneven distribution of Kmer content. Warnings </span>
<span class="s">on items such GC content distribution are more alarming.&lt;/p&gt;</span>

<span class="s">&lt;table class=&quot;gridtable&quot;&gt;</span>
<span class="s">&lt;tr&gt;</span>
<span class="s">&lt;tr&gt;&lt;th&gt;Per base sequence quality&lt;/th&gt;&lt;th&gt;Per base sequence content&lt;/th&gt;&lt;th&gt;Per sequence GC content&lt;/th&gt;&lt;/tr&gt;</span>
<span class="s">&lt;tr&gt;</span>
<span class="s">    &lt;td&gt;</span>
<span class="s">    &lt;a href=&quot;../${readqc_out}/${samplename}_fastqc.html#M1&quot;&gt;</span>
<span class="s">    &lt;IMG SRC=&quot;../${readqc_out}/per_base_quality.png&quot; WIDTH=300 ALT=&quot;Per base sequence quality&quot;&gt;</span>
<span class="s">    &lt;/a&gt;&lt;/td&gt;</span>
<span class="s">    &lt;td&gt;</span>
<span class="s">     &lt;a href=&quot;../${readqc_out}/${samplename}_fastqc.html#M4&quot;&gt;</span>
<span class="s">    &lt;IMG SRC=&quot;../${readqc_out}/per_base_sequence_content.png&quot; WIDTH=300 ALT=&quot;Per base sequence content&quot;&gt;</span>
<span class="s">    &lt;/a&gt;&lt;/td&gt;</span>
<span class="s">    &lt;td&gt;</span>
<span class="s">    &lt;a href=&quot;../${readqc_out}/${samplename}_fastqc.html#M5&quot;&gt;</span>
<span class="s">    &lt;IMG SRC=&quot;../${readqc_out}/per_sequence_gc_content.png&quot; WIDTH=300 ALT=&quot;Per sequence GC content&quot;&gt;</span>
<span class="s">    &lt;/a&gt;&lt;/td&gt;</span>
<span class="s">&lt;/tr&gt;</span>
<span class="s">&lt;/table&gt;</span>

<span class="s">&lt;p&gt;&lt;a href=&quot;../${readqc_out}/${samplename}_fastqc.html&quot;&gt;Complete FastQC reports&lt;/a&gt; &lt;/p&gt;</span>

<span class="s">&lt;h3&gt;Alignment summary&lt;/h3&gt;</span>
<span class="s">&lt;p&gt;Input reads are aligned to the transcriptome of ${reference_genes} genes using reference genome ${reference_genome}.</span>
<span class="s">The mapping rate might appear higher if the reads are mapped to the whole genome.&lt;/p&gt;</span>

<span class="s">&lt;table class=&quot;gridtable&quot;&gt;</span>
<span class="s">&lt;tr&gt;</span>
<span class="s">&lt;th&gt;Total reads&lt;/th&gt;</span>
<span class="s">&lt;th&gt;Used reads&lt;/th&gt;</span>
<span class="s">&lt;th&gt;Mapped reads&lt;/th&gt;</span>
<span class="s">&lt;th&gt;Uniquely mapped reads&lt;/th&gt;</span>
<span class="s">&lt;th&gt;Mapped pairs&lt;/th&gt;</span>
<span class="s">&lt;th&gt;Concordant mapped pairs&lt;/th&gt;</span>
<span class="s">&lt;th&gt;Mapped reads (genome)&lt;/th&gt;</span>
<span class="s">&lt;th&gt;Mapped reads (junction)&lt;/th&gt;</span>
<span class="s">&lt;/tr&gt;</span>
<span class="s">&lt;tr&gt;</span>
<span class="s">&lt;td&gt;${total_reads:,}&lt;/td&gt;</span>
<span class="s">&lt;td&gt;${used_reads:,}&lt;/td&gt;</span>
<span class="s">&lt;td&gt;${mapped_reads:,} (${mapped_reads_perc:.1f}%)&lt;/td&gt;</span>
<span class="s">&lt;td&gt;${uniq_mapped_reads:,} (${uniq_mapped_reads_perc:.1f}%)&lt;/td&gt;</span>
<span class="s">&lt;td&gt;${mapped_pairs:,} (${mapped_pairs_perc:.1f}%)&lt;/td&gt;</span>
<span class="s">&lt;td&gt;${concordant_mapped_pairs:,} (${concordant_mapped_pairs_perc:.1f}%)&lt;/td&gt;</span>
<span class="s">&lt;td&gt;${genome_mapped_count:,} (${genome_mapped_perc:.1f}%)&lt;/td&gt;</span>
<span class="s">&lt;td&gt;${junction_mapped_count:,} (${junction_mapped_perc:.1f}%)&lt;/td&gt;</span>
<span class="s">&lt;/tr&gt;</span>
<span class="s">&lt;/table&gt;</span>

<span class="s">&lt;h3&gt;Alignment QC -- RSeQC plots&lt;/h3&gt;</span>
<span class="s">&lt;table class=&quot;gridtable&quot;&gt;</span>
<span class="s">&lt;tr&gt;</span>
<span class="s">&lt;tr&gt;&lt;th&gt;Junction saturation&lt;/th&gt;&lt;th&gt;Inner distance&lt;/th&gt;&lt;th&gt;Splicing junction&lt;/th&gt;&lt;/tr&gt;</span>
<span class="s">&lt;tr&gt;</span>
<span class="s">    &lt;td&gt;</span>
<span class="s">    &lt;a href=&quot;../${alignmentqc_out}/${samplename}.junctionSaturation_plot.pdf&quot;&gt;</span>
<span class="s">    &lt;IMG SRC=&quot;../${alignmentqc_out}/${samplename}.junctionSaturation_plot.jpg&quot; WIDTH=300 ALT=&quot;Junction saturation plot&quot;&gt;</span>
<span class="s">    &lt;/a&gt;&lt;/td&gt;</span>
<span class="s">    &lt;td&gt;</span>
<span class="s">     &lt;a href=&quot;../${alignmentqc_out}/${samplename}.inner_distance_plot.pdf&quot;&gt;</span>
<span class="s">    &lt;IMG SRC=&quot;../${alignmentqc_out}/${samplename}.inner_distance_plot.jpg&quot; WIDTH=300 ALT=&quot;Inner distance plot&quot;&gt;</span>
<span class="s">    &lt;/a&gt;&lt;/td&gt;</span>
<span class="s">    &lt;td&gt;</span>
<span class="s">    &lt;a href=&quot;../${alignmentqc_out}/${samplename}.splice_junction.pdf&quot;&gt;</span>
<span class="s">    &lt;IMG SRC=&quot;../${alignmentqc_out}/${samplename}.splice_junction.jpg&quot; WIDTH=300 ALT=&quot;Splicing junction plot&quot;&gt;</span>
<span class="s">    &lt;/a&gt;&lt;/td&gt;</span>
<span class="s">&lt;/tr&gt;</span>
<span class="s">&lt;tr&gt;&lt;th&gt;Read duplication&lt;/th&gt;&lt;th&gt;Gene body coverage&lt;/th&gt;&lt;th&gt;Splicing events&lt;/th&gt;&lt;/tr&gt;</span>
<span class="s">&lt;tr&gt;</span>
<span class="s">    &lt;td&gt;</span>
<span class="s">    &lt;a href=&quot;../${alignmentqc_out}/${samplename}.DupRate_plot.pdf&quot;&gt;</span>
<span class="s">    &lt;IMG SRC=&quot;../${alignmentqc_out}/${samplename}.DupRate_plot.jpg&quot; WIDTH=300 ALT=&quot;Read duplication plot&quot;&gt;</span>
<span class="s">    &lt;/a&gt;&lt;/td&gt;</span>
<span class="s">    &lt;td&gt;</span>
<span class="s">    &lt;a href=&quot;../${alignmentqc_out}/${samplename}.geneBodyCoverage.curves.pdf&quot;&gt;</span>
<span class="s">    &lt;IMG SRC=&quot;../${alignmentqc_out}/${samplename}.geneBodyCoverage.curves.jpg&quot; WIDTH=300 ALT=&quot;Gene body coverage plot&quot;&gt;</span>
<span class="s">    &lt;/a&gt;&lt;/td&gt;</span>
<span class="s">    &lt;td&gt;</span>
<span class="s">    &lt;a href=&quot;../${alignmentqc_out}/${samplename}.splice_events.pdf&quot;&gt;</span>
<span class="s">    &lt;IMG SRC=&quot;../${alignmentqc_out}/${samplename}.splice_events.jpg&quot; WIDTH=300 ALT=&quot;Splicing events plot&quot;&gt;</span>
<span class="s">    &lt;/a&gt;&lt;/td&gt;</span>
<span class="s">    &lt;/tr&gt;</span>
<span class="s">&lt;/table&gt;</span>

<span class="s">&lt;h3&gt;Gene and exon counts&lt;/h3&gt;</span>

<span class="s">&lt;ul&gt;</span>
<span class="s">&lt;li&gt;&lt;a href=&quot;../${count_out}/${samplename}-gene_counts_rpkm.xls&quot;&gt;Gene counts with RPKM &lt;/a&gt; (${num_genes:,} genes, ${num_covered_genes:,} overlap with at least 1 read)&lt;/li&gt;</span>
<span class="s">&lt;li&gt;&lt;a href=&quot;../${count_out}/${samplename}-exon_counts_rpkm.xls&quot;&gt;Exon counts with RPKM &lt;/a&gt; (${num_exons:,} exons overlapped with at least 1 read)&lt;/li&gt;</span>
<span class="s">&lt;/ul&gt;</span>

<span class="s">&lt;p&gt;The gene count was performed by &lt;a href=&quot;http://www-huber.embl.de/HTSeq/doc/count.html&quot;&gt;htseq-count&lt;/a&gt; under</span>
<span class="s">the &lt;b&gt;union&lt;/b&gt; model. The exon count was performed by &lt;a href=&quot;https://code.google.com/p/bedtools/&quot;&gt;bedtools&lt;/a&gt;</span>
<span class="s">which counts the number of reads that overlap with each exon. These files are tab-delimited files with .xls extension.&lt;/p&gt;</span>

<span class="s">&lt;h3&gt;Fusion candidates&lt;/h3&gt;</span>
<span class="s">The following are fusion candidates produced by tophat-fusion:</span>
<span class="s">&lt;ul&gt;</span>
<span class="s">&lt;li&gt;&lt;a href=&quot;../${fusion_out}/fusions_oncofuse.xls&quot;&gt;Oncofuse filtered fusion candidates&lt;/a&gt; (${num_oncofuse_fusions:,} fusions in tab-delimited file with .xls extension)&lt;/li&gt;</span>
<span class="s">&lt;/ul&gt;</span>

<span class="s">&lt;h3&gt;Variant calls&lt;/h3&gt;</span>
<span class="s">&lt;ul&gt;</span>
<span class="s">&lt;li&gt; &lt;a href=&quot;../${variant_out}/${samplename}-filter.vcf&quot;&gt;GATK detected SNV in VCF format&lt;/a&gt;(${num_variants:,} variants, ${num_valid_variants:,} passed basic filtering)&lt;/li&gt;</span>
<span class="s">&lt;/ul&gt;</span>

<span class="s">The variants are called using the GATK Unified Genotyper and include only Single Nucleotide Variants, no indel detection was performed.&lt;/p&gt;</span>
<span class="s">&lt;p&gt;</span>
<span class="s">&lt;p&gt;</span>
<span class="s">&lt;/body&gt;</span>
<span class="s">&lt;/html&gt;</span>
<span class="s">&#39;&#39;&#39;</span><span class="p">)</span>

<span class="p">[</span><span class="n">human_hg19_1000</span><span class="p">]</span>
<span class="c">#Summary: Compress all result files into a deliverable package.</span>
<span class="kn">input:</span>
    <span class="s">&#39;${output_dir}/summary.html&#39;</span><span class="p">,</span>
    <span class="s">&#39;${readqc_out}/${samplename}_fastqc.html&#39;</span><span class="p">,</span>
    <span class="s">&#39;${readqc_out}/per_base_quality.png&#39;</span><span class="p">,</span>
    <span class="s">&#39;${readqc_out}/per_base_sequence_content.png&#39;</span><span class="p">,</span>
    <span class="s">&#39;${readqc_out}/per_sequence_gc_content.png&#39;</span><span class="p">,</span>
    <span class="s">&#39;${alignmentqc_out}/${samplename}.geneBodyCoverage.curves.pdf&#39;</span><span class="p">,</span>
    <span class="s">&#39;${alignmentqc_out}/${samplename}.junctionSaturation_plot.pdf&#39;</span><span class="p">,</span>
    <span class="s">&#39;${alignmentqc_out}/${samplename}.splice_events.pdf&#39;</span><span class="p">,</span>
    <span class="s">&#39;${alignmentqc_out}/${samplename}.splice_junction.pdf&#39;</span><span class="p">,</span>
    <span class="s">&#39;${alignmentqc_out}/${samplename}.inner_distance_plot.pdf&#39;</span><span class="p">,</span>
    <span class="s">&#39;${alignmentqc_out}//${samplename}.DupRate_plot.pdf&#39;</span><span class="p">,</span>
    <span class="s">&#39;${alignmentqc_out}/${samplename}.geneBodyCoverage.curves.jpg&#39;</span><span class="p">,</span>
    <span class="s">&#39;${alignmentqc_out}/${samplename}.junctionSaturation_plot.jpg&#39;</span><span class="p">,</span>
    <span class="s">&#39;${alignmentqc_out}/${samplename}.splice_events.jpg&#39;</span><span class="p">,</span>
    <span class="s">&#39;${alignmentqc_out}/${samplename}.splice_junction.jpg&#39;</span><span class="p">,</span>
    <span class="s">&#39;${alignmentqc_out}/${samplename}.inner_distance_plot.jpg&#39;</span><span class="p">,</span>
    <span class="s">&#39;${alignmentqc_out}/${samplename}.DupRate_plot.jpg&#39;</span><span class="p">,</span>
    <span class="s">&#39;${count_out}/${samplename}-gene_counts_rpkm.xls&#39;</span><span class="p">,</span>
    <span class="s">&#39;${count_out}/${samplename}-exon_counts_rpkm.xls&#39;</span><span class="p">,</span>
    <span class="s">&#39;${fusion_out}/fusions_oncofuse.xls&#39;</span><span class="p">,</span>
    <span class="s">&#39;${variant_out}/${samplename}-filter.vcf&#39;</span>

<span class="kn">output:</span>
    <span class="s">&#39;${output_dir}/${samplename}.zip&#39;</span>

<span class="kn">run:</span>
    <span class="nb">zip</span> <span class="err">$</span><span class="p">{</span><span class="n">output_dir</span><span class="p">}</span><span class="o">/</span><span class="err">$</span><span class="p">{</span><span class="n">samplename</span><span class="p">}</span><span class="o">.</span><span class="n">zip</span> <span class="err">$</span><span class="p">{</span><span class="nb">input</span><span class="p">}</span></code></pre></figure>

<h2 id="workflow-to-install-required-tools">Workflow to install required tools</h2>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="c">#!/usr/bin/env sos-runner</span>
<span class="c">#fileformat=SOS1.0</span>

<span class="c">#</span>
<span class="c"># Auxiliary steps to install required NGS tools if it is not avaible</span>
<span class="c">#</span>

<span class="c"># default installation directory. This directory is recognized by</span>
<span class="c"># default by SoS so that commands installed would be immediately usable</span>
<span class="c"># by sos workflows. You can use commands such as</span>
<span class="c">#</span>
<span class="c">#    sudo sos run install_ngs install_fastqc --install-dir /usr/local/bin/</span>
<span class="c">#</span>
<span class="c"># to install programs in a different directory.</span>

<span class="kn">parameter:</span> <span class="n">install_dir</span>  <span class="o">=</span> <span class="s">&#39;~/.sos/bin&#39;</span>
<span class="kn">parameter:</span> <span class="n">download_dir</span> <span class="o">=</span> <span class="s">&#39;~/.sos/bin/download&#39;</span>

<span class="p">[</span><span class="n">install_fastqc</span><span class="p">:</span> <span class="kp">provides</span><span class="o">=</span><span class="n">executable</span><span class="p">(</span><span class="s">&#39;fastqc&#39;</span><span class="p">)]</span>
<span class="n">ver</span> <span class="o">=</span> <span class="s">&#39;v0.11.5&#39;</span>

<span class="kn">download:</span> <span class="n">dest_dir</span><span class="o">=</span><span class="n">download_dir</span>
    <span class="n">http</span><span class="p">:</span><span class="o">//</span><span class="n">www</span><span class="o">.</span><span class="n">bioinformatics</span><span class="o">.</span><span class="n">babraham</span><span class="o">.</span><span class="n">ac</span><span class="o">.</span><span class="n">uk</span><span class="o">/</span><span class="n">projects</span><span class="o">/</span><span class="n">fastqc</span><span class="o">/</span><span class="n">fastqc_</span><span class="err">$</span><span class="p">{</span><span class="n">ver</span><span class="p">}</span><span class="o">.</span><span class="n">zip</span>

<span class="n">warn_if</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isdir</span><span class="p">(</span><span class="s">&#39;${install_dir!e}/FastQC_${ver}&#39;</span><span class="p">),</span>
	<span class="s">&quot;Replacing FastQC version ${ver} under ${install_dir!e}/FastQC_${ver}.&quot;</span><span class="p">)</span>

<span class="n">warn_if</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="s">&#39;${install_dir!e}/fastqc}&#39;</span><span class="p">),</span>
	<span class="s">&quot;Replacing fastqc command ${install_dir!e}/fastqc.&quot;</span><span class="p">)</span>

<span class="kn">run:</span>    
    <span class="n">cd</span> <span class="err">$</span><span class="p">{</span><span class="n">install_dir</span><span class="err">!</span><span class="n">e</span><span class="p">}</span> <span class="o">&amp;&amp;</span> \
    <span class="n">rm</span> <span class="o">-</span><span class="n">rf</span> <span class="err">$</span><span class="p">{</span><span class="n">install_dir</span><span class="err">!</span><span class="n">e</span><span class="p">}</span><span class="o">/</span><span class="n">FastQC_</span><span class="err">$</span><span class="p">{</span><span class="n">ver</span><span class="p">}</span> <span class="o">&amp;&amp;</span> \
	<span class="n">rm</span> <span class="o">-</span><span class="n">f</span> <span class="err">$</span><span class="p">{</span><span class="n">install_dir</span><span class="err">!</span><span class="n">e</span><span class="p">}</span><span class="o">/</span><span class="n">fastqc</span> <span class="o">&amp;&amp;</span> \
    <span class="n">unzip</span> <span class="o">-</span><span class="n">o</span> <span class="err">$</span><span class="p">{</span><span class="n">download_dir</span><span class="err">!</span><span class="n">e</span><span class="p">}</span><span class="o">/</span><span class="n">fastqc_</span><span class="o">*.</span><span class="n">zip</span>  <span class="o">&amp;&amp;</span> \
    <span class="n">mv</span> <span class="err">$</span><span class="p">{</span><span class="n">install_dir</span><span class="err">!</span><span class="n">e</span><span class="p">}</span><span class="o">/</span><span class="n">FastQC</span> <span class="err">$</span><span class="p">{</span><span class="n">install_dir</span><span class="err">!</span><span class="n">e</span><span class="p">}</span><span class="o">/</span><span class="n">FastQC_</span><span class="err">$</span><span class="p">{</span><span class="n">ver</span><span class="p">}</span> <span class="o">&amp;&amp;</span> \
    <span class="n">chmod</span> <span class="mi">755</span> <span class="err">$</span><span class="p">{</span><span class="n">install_dir</span><span class="p">}</span><span class="o">/</span><span class="n">FastQC_</span><span class="err">$</span><span class="p">{</span><span class="n">ver</span><span class="p">}</span><span class="o">/</span><span class="n">fastqc</span> <span class="o">&amp;&amp;</span> \
    <span class="n">ln</span> <span class="o">-</span><span class="n">s</span> <span class="err">$</span><span class="p">{</span><span class="n">install_dir</span><span class="p">}</span><span class="o">/</span><span class="n">FastQC_</span><span class="err">$</span><span class="p">{</span><span class="n">ver</span><span class="p">}</span><span class="o">/</span><span class="n">fastqc</span> <span class="err">$</span><span class="p">{</span><span class="n">install_dir</span><span class="p">}</span>
	
<span class="p">[</span><span class="n">install_bowtie</span><span class="p">:</span> <span class="kp">provides</span><span class="o">=</span><span class="n">executable</span><span class="p">(</span><span class="s">&#39;bowtie&#39;</span><span class="p">)]</span>
<span class="kn">download:</span> <span class="n">dest_dir</span><span class="o">=</span><span class="n">download_dir</span>
    <span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">sourceforge</span><span class="o">.</span><span class="n">net</span><span class="o">/</span><span class="n">projects</span><span class="o">/</span><span class="n">bowtie</span><span class="o">-</span><span class="n">bio</span><span class="o">/</span><span class="n">files</span><span class="o">/</span><span class="n">bowtie</span><span class="o">/</span><span class="mf">1.1</span><span class="o">.</span><span class="mi">2</span><span class="o">/</span><span class="n">bowtie</span><span class="o">-</span><span class="mf">1.1</span><span class="o">.</span><span class="mi">2</span><span class="o">-</span><span class="n">macos</span><span class="o">-</span><span class="n">x86_64</span><span class="o">.</span><span class="n">zip</span>

<span class="kn">run:</span>    <span class="kp">workdir</span><span class="o">=</span><span class="n">download_dir</span>
    <span class="n">unzip</span> <span class="n">bowtie</span><span class="o">-</span><span class="mf">1.1</span><span class="o">.</span><span class="mi">2</span><span class="o">-</span><span class="n">macos</span><span class="o">-</span><span class="n">x86_64</span><span class="o">.</span><span class="n">zip</span> <span class="o">&amp;&amp;</span> \
	<span class="n">rm</span> <span class="o">-</span><span class="n">rf</span> <span class="err">$</span><span class="p">{</span><span class="n">install_dir</span><span class="err">!</span><span class="n">e</span><span class="p">}</span><span class="o">/</span><span class="n">bowtie</span><span class="o">-</span><span class="mf">1.1</span><span class="o">.</span><span class="mi">2</span> <span class="o">&amp;&amp;</span> \
    <span class="n">mv</span> <span class="n">bowtie</span><span class="o">-</span><span class="mf">1.1</span><span class="o">.</span><span class="mi">2</span> <span class="err">$</span><span class="p">{</span><span class="n">install_dir</span><span class="err">!</span><span class="n">e</span><span class="p">}</span> <span class="o">&amp;&amp;</span> \
	<span class="n">ln</span> <span class="o">-</span><span class="n">s</span> <span class="o">-</span><span class="n">f</span> <span class="err">$</span><span class="p">{</span><span class="n">install_dir</span><span class="err">!</span><span class="n">e</span><span class="p">}</span><span class="o">/</span><span class="n">bowtie</span><span class="o">-</span><span class="mf">1.1</span><span class="o">.</span><span class="mi">2</span><span class="o">/*</span> <span class="err">$</span><span class="p">{</span><span class="n">install_dir</span><span class="err">!</span><span class="n">e</span><span class="p">}</span>

<span class="p">[</span><span class="n">install_bowtie2</span><span class="p">:</span> <span class="kp">provides</span><span class="o">=</span><span class="n">executable</span><span class="p">(</span><span class="s">&#39;bowtie2&#39;</span><span class="p">)]</span>
<span class="kn">download:</span> <span class="n">dest_dir</span><span class="o">=</span><span class="n">download_dir</span>
    <span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">sourceforge</span><span class="o">.</span><span class="n">net</span><span class="o">/</span><span class="n">projects</span><span class="o">/</span><span class="n">bowtie</span><span class="o">-</span><span class="n">bio</span><span class="o">/</span><span class="n">files</span><span class="o">/</span><span class="n">bowtie2</span><span class="o">/</span><span class="mf">2.2</span><span class="o">.</span><span class="mi">9</span><span class="o">/</span><span class="n">bowtie2</span><span class="o">-</span><span class="mf">2.2</span><span class="o">.</span><span class="mi">9</span><span class="o">-</span><span class="n">macos</span><span class="o">-</span><span class="n">x86_64</span><span class="o">.</span><span class="n">zip</span>

<span class="kn">run:</span>    <span class="kp">workdir</span><span class="o">=</span><span class="n">download_dir</span>
    <span class="n">unzip</span> <span class="n">bowtie2</span><span class="o">-</span><span class="mf">2.2</span><span class="o">.</span><span class="mi">9</span><span class="o">-</span><span class="n">macos</span><span class="o">-</span><span class="n">x86_64</span><span class="o">.</span><span class="n">zip</span> <span class="o">&amp;&amp;</span> \
	<span class="n">rm</span> <span class="o">-</span><span class="n">rf</span> <span class="err">$</span><span class="p">{</span><span class="n">install_dir</span><span class="err">!</span><span class="n">e</span><span class="p">}</span><span class="o">/</span><span class="n">bowtie2</span><span class="o">-</span><span class="mf">2.2</span><span class="o">.</span><span class="mi">9</span> <span class="o">&amp;&amp;</span> \
    <span class="n">mv</span> <span class="n">bowtie2</span><span class="o">-</span><span class="mf">2.2</span><span class="o">.</span><span class="mi">9</span> <span class="err">$</span><span class="p">{</span><span class="n">install_dir</span><span class="err">!</span><span class="n">e</span><span class="p">}</span> <span class="o">&amp;&amp;</span> \
	<span class="n">ln</span> <span class="o">-</span><span class="n">s</span> <span class="o">-</span><span class="n">f</span> <span class="err">$</span><span class="p">{</span><span class="n">install_dir</span><span class="err">!</span><span class="n">e</span><span class="p">}</span><span class="o">/</span><span class="n">bowtie2</span><span class="o">-</span><span class="mf">2.2</span><span class="o">.</span><span class="mi">9</span><span class="o">/*</span> <span class="err">$</span><span class="p">{</span><span class="n">install_dir</span><span class="err">!</span><span class="n">e</span><span class="p">}</span>


<span class="p">[</span><span class="n">install_tophat2</span><span class="p">:</span> <span class="kp">provides</span><span class="o">=</span><span class="n">executable</span><span class="p">(</span><span class="s">&#39;tophat2&#39;</span><span class="p">)]</span>
<span class="c"># install tophat 2, which requires bowtie2</span>
<span class="c">#</span>
<span class="c"># NOTE: The pre-compiled binary does not work well under mac so it is</span>
<span class="c"># highly recommened that you compile from source. However, compiling from</span>
<span class="c"># source requires xcode and boost, and the later is not among the easiest</span>
<span class="c"># software package to install</span>

<span class="kn">depends:</span>  <span class="n">executable</span><span class="p">(</span><span class="s">&#39;bowtie2&#39;</span><span class="p">)</span>

<span class="c">#download: dest_dir=download_dir</span>
<span class="c">#	https://ccb.jhu.edu/software/tophat/downloads/tophat-2.0.13.tar.gz</span>

<span class="c">#run:	workdir=download_dir</span>
<span class="c">#	tar zxf tophat-2.0.13.tar.gz &amp;&amp; \</span>
<span class="c">#	cd tophat-2.0.13 &amp;&amp; \</span>
<span class="c">#	./configure --prefix=${os.path.dirname(install_dir)!e} &amp;&amp; \</span>
<span class="c">#	make -B &amp;&amp; make install</span>
	
<span class="kn">download:</span> <span class="n">dest_dir</span><span class="o">=</span><span class="n">download_dir</span>
    <span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">ccb</span><span class="o">.</span><span class="n">jhu</span><span class="o">.</span><span class="n">edu</span><span class="o">/</span><span class="n">software</span><span class="o">/</span><span class="n">tophat</span><span class="o">/</span><span class="n">downloads</span><span class="o">/</span><span class="n">tophat</span><span class="o">-</span><span class="mf">2.0</span><span class="o">.</span><span class="mf">13.</span><span class="n">OSX_x86_64</span><span class="o">.</span><span class="n">tar</span><span class="o">.</span><span class="n">gz</span>

<span class="kn">run:</span>    <span class="kp">workdir</span><span class="o">=</span><span class="n">download_dir</span>
    <span class="n">tar</span> <span class="n">zxf</span> <span class="n">tophat</span><span class="o">-</span><span class="mf">2.0</span><span class="o">.</span><span class="mf">13.</span><span class="n">OSX_x86_64</span><span class="o">.</span><span class="n">tar</span><span class="o">.</span><span class="n">gz</span> <span class="o">&amp;&amp;</span> \
	<span class="n">rm</span> <span class="o">-</span><span class="n">rf</span> <span class="err">$</span><span class="p">{</span><span class="n">install_dir</span><span class="err">!</span><span class="n">e</span><span class="p">}</span><span class="o">/</span><span class="n">tophat</span><span class="o">-</span><span class="mf">2.0</span><span class="o">.</span><span class="mf">13.</span><span class="n">OSX_x86_64</span> <span class="o">&amp;&amp;</span> \
    <span class="n">mv</span> <span class="n">tophat</span><span class="o">-</span><span class="mf">2.0</span><span class="o">.</span><span class="mf">13.</span><span class="n">OSX_x86_64</span> <span class="err">$</span><span class="p">{</span><span class="n">install_dir</span><span class="err">!</span><span class="n">e</span><span class="p">}</span> <span class="o">&amp;&amp;</span> \
	<span class="n">ln</span> <span class="o">-</span><span class="n">s</span> <span class="o">-</span><span class="n">f</span> <span class="err">$</span><span class="p">{</span><span class="n">install_dir</span><span class="err">!</span><span class="n">e</span><span class="p">}</span><span class="o">/</span><span class="n">tophat</span><span class="o">-</span><span class="mf">2.0</span><span class="o">.</span><span class="mf">13.</span><span class="n">OSX_x86_64</span><span class="o">/*</span> <span class="err">$</span><span class="p">{</span><span class="n">install_dir</span><span class="err">!</span><span class="n">e</span><span class="p">}</span>


<span class="p">[</span><span class="n">install_rseqc</span><span class="p">:</span> <span class="kp">provides</span><span class="o">=</span><span class="n">executable</span><span class="p">(</span><span class="s">&#39;inner_distance.py&#39;</span><span class="p">)]</span>

<span class="kn">run:</span>
	<span class="n">pip</span> <span class="n">install</span> <span class="n">RSeQC</span></code></pre></figure>


                            </div>
                            <button type="button" class="btn btn-default" data-dismiss="modal"><i class="fa fa-times"></i> Close</button>
                        </div>
                    </div> 
                    <div class="footer-wrapper">
                      <footer role="contentinfo">
                        <span>&copy; 2016 Bo Peng.</span>
                      </footer>
                    </div>
                </div>
            </div>
        </div>
    </div> 


   <div class="portfolio-modal modal fade" id="SoS-Docker-Guide" tabindex="-1" role="dialog" aria-hidden="true">
        <div class="modal-content">
            <div class="close-modal" data-dismiss="modal">
                <div class="lr">
                    <div class="rl">
                    </div>
                </div>
            </div>
            <div class="container">
                <div class="row">
                 <div class="col-lg-8 col-lg-offset-2">
                        <h1>6.SoS-Docker-Guide</h1>
                        <div class="modal-body">
                         
                            <hr class="star-primary">
                            <p>This documentation is under development.</p>

<h2 id="what-is-docker-and-why-it-is-helpful">What is docker and why it is helpful</h2>

<p>This is a big question to answer but in essence you can think docker containers as virtual machines with applications but without the bulky OS part, or applications with stripped down OSes. Docker contains are much more lightweight than virtual machines because all docker containers share the same core OS and related containers (e.g. different applications derived from the same CentOS or Ubuntu OS) share the same base container. Please refer to the <a href="https://www.docker.com/">docker website</a> for details about docker. I have found it helpful to watch a few youtube videos on docker.</p>

<p>The reason why docker is very helpful in building (bioinformatics) pipelines are that</p>

<ol>
  <li>
    <p>Applications are encapsulated in docker containers so that they do not interfere with the underlying OS, and with other applications. For example, we can run a workflow with applications that based on different versions of Python2 and Python 3 without having to install them locally and calling the correct version of Python, because all applications use the specific version of Python and required libraries and tools inside their own container.</p>
  </li>
  <li>
    <p>Workflows will be more stable and reproducible because unlike a local installation of Python that can be affected by other software and upgrades of python, Docker containers are stable and will not change.</p>
  </li>
  <li>
    <p>The same docker containers can be executed on different OS (e.g. various version of Linux, MacOSX etc) so your workflow built on a Mac OS workstation can be executed on a cluster environment.</p>
  </li>
</ol>

<p>There are of course some complexity in the use of docker but SoS has made it extremely easy to use docker in your workflows.</p>

<h2 id="installing-and-configuring-docker">Installing and configuring docker</h2>

<p>Docker is relatively new and is evolving very fast. It is crucial for you to install the latest version from <a href="https://www.docker.com/">docker website</a>. This website provides very detailed step by step instruction and you should have no problem installing docker on your machine.</p>

<p>After installation, you should be able to start a docker terminal (On Mac, it is called <em>Docker QuickStart Terminal</em>) and run command</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="err">$</span> <span class="n">docker</span> <span class="n">run</span> <span class="n">hello</span><span class="o">-</span><span class="n">world</span></code></pre></figure>

<p>as suggested by the documentation.</p>

<h3 id="configuration-for-mac">Configuration for Mac</h3>

<p>If you are using a Mac machine, docker containers are executed under a VirtualBox virtual machine. The docker daemon cannot see anything outside of <code>/Users</code> so if your data resides in a large volume outside of <code>/Users</code>, you will not be able to access it from inside a docker container. This can be very frustrating because Mac OS X’s boot disk is usually very small so all important data likely reside in separate volumes under <code>/Volumes</code>.</p>

<p>In addition, the default docker machine has only 1 CPU, 2GB of RAM and 20GB diskspace. If you are using docker for any serious bioinformatics work and download a few docker images, you will need to add more RAM and more storage to the virtual box. To solve these problems, you would need to remove the default virtual machine and create a new one.</p>

<p>From your command line, stop and remove the default virtual machine</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="err">$</span> <span class="n">docker</span><span class="o">-</span><span class="n">machine</span> <span class="n">stop</span>
<span class="err">$</span> <span class="n">docker</span><span class="o">-</span><span class="n">machine</span> <span class="n">rm</span> <span class="n">default</span></code></pre></figure>

<p>then create a new one with</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="err">$</span> <span class="n">docker</span><span class="o">-</span><span class="n">machine</span> <span class="n">create</span> <span class="o">-</span><span class="n">d</span> <span class="n">virtualbox</span> <span class="o">--</span><span class="n">virtualbox</span><span class="o">-</span><span class="n">cpu</span><span class="o">-</span><span class="n">count</span> <span class="mi">4</span> \
    <span class="o">--</span><span class="n">virtualbox</span><span class="o">-</span><span class="n">memory</span> <span class="mi">30000</span> <span class="o">--</span><span class="n">virtualbox</span><span class="o">-</span><span class="n">disk</span><span class="o">-</span><span class="n">size</span> <span class="mi">100000</span> <span class="n">default</span></code></pre></figure>

<p>Here I am using 4CPU/core, 30G of RAM and 100G of disk. You can adjust these parameters according to the configuration of your machine.</p>

<p>After this, you would need to share <code>/volumes</code> to the virtual machine,</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="c"># stop docker virtual machine</span>
<span class="err">$</span> <span class="n">docker</span><span class="o">-</span><span class="n">machine</span> <span class="n">stop</span>
<span class="c"># add all /Volumes to /Volumes</span>
<span class="err">$</span> <span class="n">VBoxManage</span> <span class="n">sharedfolder</span> <span class="n">add</span> <span class="n">default</span> <span class="o">--</span><span class="n">name</span> <span class="n">Volumes</span> <span class="o">--</span><span class="n">hostpath</span> <span class="o">/</span><span class="n">Volumes</span> <span class="o">--</span><span class="n">automount</span></code></pre></figure>

<p>You would also need to mount <code>/Volumes</code> to the docker-machine using the following command after the docker machine starts,</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="err">$</span> <span class="n">docker</span><span class="o">-</span><span class="n">machine</span> <span class="n">ssh</span> <span class="n">default</span> <span class="s">&#39;sudo mkdir -p /Volumes ; sudo  mount -t vboxsf Volumes /Volumes&#39;</span></code></pre></figure>

<p>but SoS would do this automatically for you if you are using SoS to interact with docker.</p>

<h2 id="running-a-workflow-with-docker">Running a workflow with docker</h2>

<p>Running a docker-based workflow is easy because SoS will automatically download docker images and execute scripts inside docker container. Just make sure you start the script from a docker terminal (e.g. test with command <code>docker ps</code>)</p>

<h2 id="building-a-docker-image">Building a docker image</h2>

<p>Building a docker image is usually done outside of SoS if you are maintaining a collection of docker containers to be shared by your workflows, your groups, or everyone. However, if you need to create a docker image on-the-fly or would like to embed the Dockerfile inside a SoS script, you can use the <code>docker_build</code> action to build a docker container.</p>

<p>For example, you can build a container for MISO as follows:</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="p">[</span><span class="n">miso_build</span><span class="p">]</span>
<span class="c"># building miso from a Dockerfile</span>
<span class="kn">docker_build:</span> <span class="n">tag</span><span class="o">=</span><span class="s">&#39;mdabioinfo/miso:latest&#39;</span>

    <span class="c">############################################################</span>
    <span class="c"># Dockerfile to build MISO container images</span>
    <span class="c"># Based on Anaconda python</span>
    <span class="c">############################################################</span>

    <span class="c"># Set the base image to anaconda Python 2.7 (miso does not support python 3)</span>
    <span class="n">FROM</span> <span class="n">continuumio</span><span class="o">/</span><span class="n">anaconda</span>

    <span class="c"># File Author / Maintainer</span>
    <span class="n">MAINTAINER</span> <span class="n">Bo</span> <span class="n">Peng</span> <span class="o">&lt;</span><span class="n">bpeng</span><span class="nd">@mdanderson.org</span><span class="o">&gt;</span>

    <span class="c"># Update the repository sources list</span>
    <span class="n">RUN</span> <span class="n">apt</span><span class="o">-</span><span class="n">get</span> <span class="n">update</span>

    <span class="c"># Install compiler and python stuff, samtools and git</span>
    <span class="n">RUN</span> <span class="n">apt</span><span class="o">-</span><span class="n">get</span> <span class="n">install</span> <span class="o">--</span><span class="n">yes</span> \
     <span class="n">build</span><span class="o">-</span><span class="n">essential</span> \
     <span class="n">gcc</span><span class="o">-</span><span class="n">multilib</span> \
     <span class="n">gfortran</span> \ 
     <span class="n">apt</span><span class="o">-</span><span class="n">utils</span> \
     <span class="n">libblas3</span> \ 
     <span class="n">liblapack3</span> \
     <span class="n">libc6</span> \
     <span class="n">cython</span> \ 
     <span class="n">samtools</span> \
     <span class="n">libbam</span><span class="o">-</span><span class="n">dev</span> \
     <span class="n">bedtools</span> \
     <span class="n">wget</span> \
     <span class="n">zlib1g</span><span class="o">-</span><span class="n">dev</span> \ 
     <span class="n">tar</span> \
     <span class="n">gzip</span>

    <span class="n">WORKDIR</span> <span class="o">/</span><span class="n">usr</span><span class="o">/</span><span class="n">local</span>
    <span class="n">RUN</span> <span class="n">pip</span> <span class="n">install</span> <span class="n">misopy</span></code></pre></figure>

<p>Command</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="n">sos</span> <span class="n">run</span> <span class="n">script</span> <span class="n">miso_build</span></code></pre></figure>

<p>would build a docker image <code>mdabioinfo/miso:latest</code> that can be used by other SoS steps.</p>

<h2 id="writing-a-workflow-with-docker-support">Writing a workflow with docker support</h2>

<p>Writing a workflow with docker support is a bit more complicated because you will need to understand a few concepts of docker, so reading through the <a href="https://docs.docker.com/engine/reference/run/">docker run manual</a> should be helpful. The most important concept is <strong>Volumes</strong>, whch is how the host directories are mounted to a docker container so that the command executed inside the container can access (and change) files on the host machine. SoS helps the use of docker by</p>

<ul>
  <li>Automatically mounts <code>/tmp</code> to <code>/tmp</code></li>
  <li>Automatically mounts <code>/Users</code> to <code>/Users</code> under MacOS X</li>
  <li>Automatically mounts user script inside docker and execute it as <code>/var/lib/sos/xxxxx</code></li>
</ul>

<p>so that step <code>input</code> and <code>output</code> are almost always identical inside and outside of docker.</p>

<p>To use existing public docker container, you will need to specify its tag using option <code>docker_image</code>. For example, to use <code>compbio/ngseasy-fastqc</code> container to run <code>fastqc</code>, instead of installing <code>fastqc</code> locally, you can do</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="p">[</span><span class="n">MISO_1</span><span class="p">]</span>
<span class="kn">run:</span>     <span class="n">docker_image</span><span class="o">=</span><span class="s">&#39;compbio/ngseasy-fastqc:1.0-r001&#39;</span>
    <span class="n">fastqc</span> <span class="err">$</span><span class="p">{</span><span class="nb">input</span><span class="p">}</span> <span class="o">-</span><span class="n">o</span> <span class="o">/</span><span class="n">tmp</span></code></pre></figure>

<p>(More to follow)</p>

<h2 id="limitations">Limitations</h2>

<ul>
  <li>Virtual Box virtual machine does not support symbolic link so running <code>ln -s</code> inside a docker machine under Mac will cause a strange error message <code>Read-only file system</code>.</li>
</ul>

                            </div>
                            <button type="button" class="btn btn-default" data-dismiss="modal"><i class="fa fa-times"></i> Close</button>
                        </div>
                    </div> 
                    <div class="footer-wrapper">
                      <footer role="contentinfo">
                        <span>&copy; 2016 Bo Peng.</span>
                      </footer>
                    </div>
                </div>
            </div>
        </div>
    </div> 


   <div class="portfolio-modal modal fade" id="Generating-Report-with-SoS" tabindex="-1" role="dialog" aria-hidden="true">
        <div class="modal-content">
            <div class="close-modal" data-dismiss="modal">
                <div class="lr">
                    <div class="rl">
                    </div>
                </div>
            </div>
            <div class="container">
                <div class="row">
                 <div class="col-lg-8 col-lg-offset-2">
                        <h1>7.Generating-Report-with-SoS</h1>
                        <div class="modal-body">
                         
                            <hr class="star-primary">
                            <p>This documentation is under development.</p>

<h2 id="reporting-features-of-script-of-scripts">Reporting features of Script of Scripts</h2>

<p>SoS provides a simple yet powerful reporting mechanism. That is to say,</p>

<ul>
  <li>Any lines that start with <code>'! '</code> are report lines that will be written to a default report file. A space after <code>!</code> is required for better formatting, but lines with a single <code>!</code> is also acceptable.</li>
  <li>Action <code>report</code> can be used to append text to this report, or optionally to another report file.</li>
</ul>

<p>Similar to other texts in a SoS script, reports will be interpolated before they are written to the output. For example</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="c">#!/usr/bin/env sos-runner</span>
<span class="c">#fileformat=SOS1.0</span>

<span class="err">!</span> <span class="o">==</span> <span class="n">My</span> <span class="n">report</span>
<span class="err">!</span> <span class="n">Date</span><span class="p">:</span> <span class="n">Apr</span><span class="o">.</span> <span class="mi">2016</span>

<span class="c"># human</span>
<span class="c"># align raw reads to human genome</span>
<span class="p">[</span><span class="n">parameters</span><span class="p">]</span>
<span class="c"># path to STAR</span>
<span class="n">star_path</span> <span class="o">=</span> <span class="s">&#39;~/bin/STAR&#39;</span>

<span class="p">[</span><span class="o">*</span><span class="n">_1</span><span class="p">]</span>
<span class="kn">output:</span>   <span class="s">&#39;figure.png&#39;</span>
<span class="kn">R:</span>
   <span class="n">R</span> <span class="n">script</span>

<span class="err">!</span> <span class="o">==</span> <span class="n">some</span> <span class="n">analysis</span>
<span class="err">!</span> <span class="err">!</span><span class="p">[</span><span class="n">result</span><span class="p">](</span><span class="err">$</span><span class="p">{</span><span class="n">output</span><span class="p">})</span></code></pre></figure>

<p>will generate a text file</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="o">==</span> <span class="n">My</span> <span class="n">report</span>
<span class="kn">Date:</span> <span class="n">Apr</span><span class="o">.</span> <span class="mi">2016</span>
<span class="o">==</span> <span class="n">some</span> <span class="n">analysis</span>
<span class="err">!</span><span class="p">[</span><span class="n">result</span><span class="p">](</span><span class="n">figure</span><span class="o">.</span><span class="n">png</span><span class="p">)</span></code></pre></figure>

<p>after the execution of the pipeline. Action report can also write to this report and provides more flexible and powerful ways of reporting different types of results. Action pandoc can be used to convert the report to various formats. Note that</p>

<ul>
  <li>
    <p>Although markdown is generally used, SoS does not restrict the type of report a script generates. If so desired, you can use this mechanism to generate a log file in plain text format, a HTML file with fancy style, any flavor of markdown that you can post-process, or a .Rmd file that can be loaded into RMarkdown.</p>
  </li>
  <li>
    <p>The report is generated from the steps that are executed so the content can be controlled with parameters and configuration files.</p>
  </li>
</ul>

<h2 id="action-report">Action <code>report</code></h2>

<p>Action <code>report</code> can be used write a block of text</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="kn">report:</span>
    <span class="o">==</span> <span class="n">My</span> <span class="n">report</span>
    <span class="n">Date</span><span class="p">:</span> <span class="n">Apr</span><span class="o">.</span> <span class="mi">2016</span></code></pre></figure>

<p>or content of a (dynamically generated) file to the report.</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="n">report</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span><span class="s">&#39;report_source.txt&#39;</span><span class="p">)</span></code></pre></figure>

<p>Alternatively, it can write to another file so that multiple reports could be generated from a SoS run.</p>

<figure class="highlight"><pre><code class="language-sos" data-lang="sos"><span class="n">report</span><span class="p">(</span><span class="n">output</span><span class="o">=</span><span class="s">&#39;alternative.txt&#39;</span><span class="p">)</span></code></pre></figure>

<h2 id="pandoc-with-html-and-pdf-output">Pandoc with HTML and pdf output</h2>

<h2 id="rmarkdown">Rmarkdown</h2>

<h2 id="multiple-reports">Multiple reports</h2>


                            </div>
                            <button type="button" class="btn btn-default" data-dismiss="modal"><i class="fa fa-times"></i> Close</button>
                        </div>
                    </div> 
                    <div class="footer-wrapper">
                      <footer role="contentinfo">
                        <span>&copy; 2016 Bo Peng.</span>
                      </footer>
                    </div>
                </div>
            </div>
        </div>
    </div> 



<script src="//ajax.googleapis.com/ajax/libs/jquery/1.9.1/jquery.min.js"></script>
<script>window.jQuery || document.write('<script src="https://bopeng.github.io/assets/js/vendor/jquery-1.9.1.min.js"><\/script>')</script>


<script src="/SOS/assets/js/scripts.min.js"></script>
<script src="/SOS/assets/js/vendor/bootstrap.min.js"></script>


          

</body>
</html>