<!DOCTYPE html>
<html>
<head><meta charset="utf-8" />
<title>A_Complete_Example</title>

<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.1.10/require.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/2.0.3/jquery.min.js"></script>

<link rel="stylesheet" href="../../assets/css/bootstrap.min.css">
<link rel="stylesheet" href="../../assets/css/pygmentsSoS.css">




<!-- Loading mathjax macro -->
<!-- Load mathjax -->
    <script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS_HTML"></script>
    <!-- MathJax configuration -->
    <script type="text/x-mathjax-config">
    MathJax.Hub.Config({
        tex2jax: {
            inlineMath: [ ['$','$'], ["\\(","\\)"] ],
            displayMath: [ ['$$','$$'], ["\\[","\\]"] ],
            processEscapes: true,
            processEnvironments: true
        },
        // Center justify equations in code and markdown cells. Elsewhere
        // we use CSS to left justify single line equations in code cells.
        displayAlign: 'center',
        "HTML-CSS": {
            styles: {'.MathJax_Display': {"margin": 0}},
            linebreaks: { automatic: true }
        }
    });
    </script>
    <!-- End of mathjax configuration --></head>
<body>
  <div tabindex="-1" id="notebook" class="border-box-sizing">
    <div class="container" id="notebook-container">

<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="A-RNA-Seq-data-processing-workflow">A RNA Seq data processing workflow<a class="anchor-link" href="#A-RNA-Seq-data-processing-workflow">&#182;</a></h2>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>If you are curious about how to write large workflows using SoS or just how a long SoS script looks like, here is a RNA Seq data processing workflow written in SoS. The workflow consists of many steps and was written in a mixed style, namely a spine of forward-moving steps and several auxiliary steps that supply required information. The DAG of the workflow looks like</p>
<p><img src="media/example_dag.png" alt="DAG"></p>
<p>although this DAG does not include many auxiliary steps (e.g. those that download required resources) because these steps were called once and would not be called once the resources are available. The DAG is generated using command</p>
<div class="highlight"><pre><span></span>sos prepare RNASeq_hg10.sos
cat .sos/human_hg19.dot <span class="p">|</span> dot -T png &gt; example_dag.png
</pre></div>
<p>This tutorial will divide the workflow into several parts and briefly describe each of them.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Header">Header<a class="anchor-link" href="#Header">&#182;</a></h3><p>The header of script contains description of the workflow.</p>
<div class="highlight"><pre><span></span><span class="ch">#!/usr/bin/env sos-runner</span>
<span class="c1">#fileformat=SOS1.0</span>

<span class="c1"># Workflow for a comprehensive analysis of paired-end RNA seq data</span>
<span class="c1"># that performs read-level quality assessment, alignment, alignment quality</span>
<span class="c1"># assessment, summary statistics, fusion detection and filtering, gene and exon</span>
<span class="c1"># level expression counts, and variant calling. The input of this pipeline</span>
<span class="c1"># should be a list of fastq or fastq.gz files with paired reads differentiated</span>
<span class="c1"># by _R1_ and _R2_ in filenames. &lt;p&gt;The output of the pipeline should be a</span>
<span class="c1"># directory to which all results will be written. The directory will be </span>
<span class="c1"># created if it does not exist. Note that it is highly recommended that</span>
<span class="c1"># you execute the pipeline with a subset of samples (using option --sampling</span>
<span class="c1"># 1000000) to test the pipeline before you apply it to the full dataset. </span>
<span class="c1"># Network connection will be needed for the first run (to download resources)</span>
<span class="c1"># but not for subsequent runs. For details of the pipeline including the</span>
<span class="c1"># accepted parameters, please run command &quot;vtools show pipeline RNASeq_hg19_v3&quot;.</span>

<span class="c1"># human_hg19</span>
<span class="c1"># </span>
<span class="c1"># This pipeline uses the hg19 version of the reference</span>
<span class="c1"># genome, genes downloaded from Illumina iGenome, and existing variants from</span>
<span class="c1"># the GATK resource bundle. This pipeline uses &lt;ul&gt;</span>
<span class="c1">#   &lt;li&gt;tophat 2.0.13 </span>
<span class="c1">#   &lt;li&gt;bowtie 1.1.1</span>
<span class="c1">#   &lt;li&gt;samtools 0.1.19</span>
<span class="c1">#   &lt;li&gt;picard 1.82</span>
<span class="c1">#   &lt;li&gt;GATK 3.3</span>
<span class="c1">#   &lt;li&gt;fastqc 0.11.2</span>
<span class="c1">#   &lt;li&gt;BEDTools 2.17.0</span>
<span class="c1">#   &lt;li&gt;RSeQC 2.4, and</span>
<span class="c1">#   &lt;li&gt;Oncofuse 1.0.7</span>
<span class="c1">#   &lt;/ul&gt;</span>
<span class="c1"># Other versions of the tools can be used if you execute the pipeline with option</span>
<span class="c1"># --strict_version False, but the pipeline might or might not work as expected. </span>
<span class="c1"># Please download all these tools and make them available to the pipeline (set $PATH),</span>
<span class="c1"># and start the pipeline using command &quot;vtools execute /path/to/this_file options&quot;.</span>
<span class="c1"># Noted that the precompiled version of tophat 2.0.13 can fail under mac so it is</span>
<span class="c1"># recommended that you compile tophat2 from source.</span>
</pre></div>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="inclusion-of-auxiliary-steps">inclusion of auxiliary steps<a class="anchor-link" href="#inclusion-of-auxiliary-steps">&#182;</a></h3><div class="highlight"><pre><span></span><span class="o">%</span><span class="kn">from</span> <span class="nn">install_ngs</span> <span class="nn">include</span> <span class="o">*</span>
</pre></div>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered">
<div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Global-definitions">Global definitions<a class="anchor-link" href="#Global-definitions">&#182;</a></h2><div class="highlight"><pre><span></span><span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isdir</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">expanduser</span><span class="p">(</span><span class="s1">&#39;~/rsrch1/variant_tools_resource/pipeline_resource&#39;</span><span class="p">)):</span>
    <span class="n">resource_dir</span>  <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">expanduser</span><span class="p">(</span><span class="s1">&#39;~/rsrch1/variant_tools_resource/pipeline_resource&#39;</span><span class="p">)</span>
<span class="k">elif</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isdir</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">expanduser</span><span class="p">(</span><span class="s1">&#39;~/.variant_tools/pipeline_resource&#39;</span><span class="p">)):</span>
    <span class="n">resource_dir</span>  <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">expanduser</span><span class="p">(</span><span class="s1">&#39;~/.variant_tools/pipeline_resource&#39;</span><span class="p">)</span>
<span class="kn">else:</span>
    <span class="kp">fail_if</span><span class="p">(</span><span class="bp">True</span><span class="p">,</span> <span class="s1">&#39;Unable to locate a resource directory&#39;</span><span class="p">)</span>

<span class="c1"># Path to picard jar files</span>
<span class="kn">parameter:</span> <span class="n">picard_path</span> <span class="o">=</span> <span class="s1">&#39;~/bin/Picard&#39;</span>

<span class="c1"># Path to GenomeAnalysisTK.jar</span>
<span class="kn">parameter:</span> <span class="n">gatk_path</span> <span class="o">=</span> <span class="s1">&#39;~/bin/GATK&#39;</span>

<span class="c1"># Path to Oncofuse.jar</span>
<span class="kn">parameter:</span> <span class="n">oncofuse_path</span> <span class="o">=</span> <span class="s1">&#39;~/bin/Oncofuse&#39;</span>

<span class="c1"># input fastq files</span>
<span class="kn">parameter:</span> <span class="n">fastq_files</span> <span class="o">=</span> <span class="nb">list</span>

<span class="kp">fail_if</span><span class="p">(</span><span class="nb">len</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">fastq_files</span> <span class="k">if</span> <span class="s1">&#39;_R1_&#39;</span> <span class="ow">in</span> <span class="n">x</span><span class="p">])</span> <span class="o">==</span> <span class="mi">0</span>  \
    <span class="ow">or</span> <span class="nb">len</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">fastq_files</span> <span class="k">if</span> <span class="s1">&#39;_R1_&#39;</span> <span class="ow">in</span> <span class="n">x</span><span class="p">])</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">fastq_files</span> <span class="k">if</span> <span class="s1">&#39;_R2_&#39;</span> <span class="ow">in</span> <span class="n">x</span><span class="p">]),</span>
    <span class="s1">&#39;Input file names are not paired fastq files differentiated by _R1_ and _R2_ in filename.&#39;</span><span class="p">)</span>

<span class="c1"># Name of the sample being processed, which will be used for filename</span>
<span class="c1"># and read group of aligned reads.</span>
<span class="kn">parameter:</span> <span class="n">samplename</span> <span class="o">=</span> <span class="nb">str</span>

<span class="c1"># If set to a positive number, this pipeline will sample specified</span>
<span class="c1"># number of reads (pairs) from the input. This is useful to test if the pipeline</span>
<span class="c1"># works for your environment, download all needed resources, created indexes,</span>
<span class="c1"># and check the quality of data.</span>
<span class="kn">parameter:</span> <span class="n">sampling</span> <span class="o">=</span> <span class="mi">0</span>


<span class="c1"># An optional filename that contains detailed information</span>
<span class="c1"># about the sample. The sample sheet should be a tab or comma-delimited </span>
<span class="c1"># file with a header, and one or more lines with one of which contain the </span>
<span class="c1"># information of the sample. The information from the sample sheet will be</span>
<span class="c1"># written to the final report (summary.html).</span>
<span class="kn">parameter:</span> <span class="n">sample_sheet</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>

<span class="c1">#</span>
<span class="c1"># CONSTANT VALUES: not configurable in this pipeline</span>
<span class="c1"># </span>
<span class="n">pipeline_name</span>     <span class="o">=</span> <span class="s1">&#39;RNASeq_hg19&#39;</span>
<span class="n">pipeline_version</span>  <span class="o">=</span> <span class="s1">&#39;ver1.0 (2016/8/29&#39;</span>
<span class="n">reference_genome</span>  <span class="o">=</span> <span class="s1">&#39;hg19&#39;</span>
<span class="n">reference_genes</span>   <span class="o">=</span> <span class="s1">&#39;UCSC/Illumina iGenome&#39;</span>
<span class="c1"># HG19 version of the resource files</span>
<span class="n">gatk_resource_dir</span> <span class="o">=</span> <span class="s1">&#39;${resource_dir}/GATK/2.8/hg19/&#39;</span>
<span class="n">igenome_resource_dir</span> <span class="o">=</span> <span class="s1">&#39;${resource_dir}/iGenome&#39;</span>
<span class="n">ncbi_resource_dir</span> <span class="o">=</span> <span class="s1">&#39;${resource_dir}/NCBI&#39;</span>

<span class="n">igenome_url</span>       <span class="o">=</span> <span class="s1">&#39;ftp://igenome:G3nom3s4u@ussd-ftp.illumina.com/Homo_sapiens/UCSC/hg19/Homo_sapiens_UCSC_hg19.tar.gz&#39;</span>
<span class="n">gene_info_url</span>     <span class="o">=</span> <span class="s1">&#39;ftp://ftp.ncbi.nih.gov/gene/DATA/GENE_INFO/Mammalia/Homo_sapiens.gene_info.gz&#39;</span>
<span class="n">reference_dir</span>     <span class="o">=</span> <span class="s1">&#39;${igenome_resource_dir}/Homo_sapiens/UCSC/hg19/Sequence&#39;</span>
<span class="n">transcriptome_dir</span> <span class="o">=</span> <span class="s1">&#39;${igenome_resource_dir}/Homo_sapiens/UCSC/hg19/Annotation/transcriptome_index&#39;</span>
<span class="n">annotation_dir</span>    <span class="o">=</span> <span class="s1">&#39;${igenome_resource_dir}/Homo_sapiens/UCSC/hg19/Annotation&#39;</span>
<span class="n">refgene_txt</span>       <span class="o">=</span> <span class="s1">&#39;${annotation_dir}/Genes/refGene.txt&#39;</span>
<span class="n">genes_gtf</span>         <span class="o">=</span> <span class="s1">&#39;${annotation_dir}/Genes/genes.gtf&#39;</span>
<span class="c1"># generated files&#39;</span>
<span class="n">refgene_bed</span>       <span class="o">=</span> <span class="s1">&#39;${annotation_dir}/Genes/refGene.bed&#39;</span>
<span class="n">genes_exon_bed</span>    <span class="o">=</span> <span class="s1">&#39;${annotation_dir}/Genes/genes_exon.bed&#39;</span>
<span class="n">genes_with_chr_gtf</span><span class="o">=</span> <span class="s1">&#39;${annotation_dir}/Genes/genes_chr.gtf&#39;</span>
<span class="n">genesize_txt</span>      <span class="o">=</span> <span class="s1">&#39;${annotation_dir}/Genes/geneSize.txt&#39;</span>
<span class="c1"># Download from UCSC&#39;</span>
<span class="n">ensgene_url</span>       <span class="o">=</span> <span class="s1">&#39;http://hgdownload.cse.ucsc.edu/goldenpath/hg19/database/ensGene.txt.gz&#39;</span>
<span class="n">ensgene_txt</span>       <span class="o">=</span> <span class="s1">&#39;${annotation_dir}/Genes/ensGene.txt&#39;</span>
<span class="n">chromsize_url</span>     <span class="o">=</span> <span class="s1">&#39;ftp://hgdownload.cse.ucsc.edu/goldenPath/hg19/database/chromInfo.txt.gz&#39;</span>
<span class="n">chromsize_txt</span>     <span class="o">=</span> <span class="s1">&#39;${annotation_dir}/chromsize.txt&#39;</span>
<span class="c1"># GATK resource bundle&#39;</span>
<span class="n">gatk_url</span>          <span class="o">=</span> <span class="s1">&#39;ftp://gsapubftp-anonymous@ftp.broadinstitute.org/bundle/2.8/hg19/&#39;</span>
<span class="c1"># &#39;</span>
<span class="c1"># Output directories&#39;</span>
<span class="n">output_dir</span>        <span class="o">=</span> <span class="n">samplename</span>
<span class="n">temp_dir</span>          <span class="o">=</span> <span class="s1">&#39;${output_dir}/tmp&#39;</span>
<span class="n">alignment_out</span>     <span class="o">=</span> <span class="s1">&#39;${output_dir}/alignment&#39;</span>
<span class="n">alignmentqc_out</span>   <span class="o">=</span> <span class="s1">&#39;${output_dir}/alignment_qc&#39;</span>
<span class="n">fusion_out</span>        <span class="o">=</span> <span class="s1">&#39;${output_dir}/fusion&#39;</span>
<span class="n">readqc_out</span>        <span class="o">=</span> <span class="s1">&#39;${output_dir}/read_qc&#39;</span>
<span class="n">count_out</span>         <span class="o">=</span> <span class="s1">&#39;${output_dir}/counts&#39;</span>
<span class="n">variant_out</span>       <span class="o">=</span> <span class="s1">&#39;${output_dir}/variants&#39;</span>



<span class="p">[</span><span class="n">human_hg19_100</span><span class="p">:</span> <span class="kp">skip</span><span class="o">=</span><span class="n">sampling</span> <span class="o">==</span> <span class="mi">0</span><span class="p">,</span> <span class="kp">shared</span><span class="o">=</span><span class="s1">&#39;fastq_files&#39;</span><span class="p">]</span>
<span class="c1"># Select a subset of reads if parameter --sampling is specified. Output will be </span>
<span class="c1"># written to the project cache directory.</span>
<span class="kn">input:</span>
    <span class="n">fastq_files</span>

<span class="kn">depends:</span>
    <span class="n">executable</span><span class="p">(</span><span class="s1">&#39;gunzip&#39;</span><span class="p">)</span>

<span class="kn">output:</span>
    <span class="s1">&#39;${temp_dir}/${samplename}_R1_${sampling}.fastq&#39;</span><span class="p">,</span>
    <span class="s1">&#39;${temp_dir}/${samplename}_R2_${sampling}.fastq&#39;</span>

<span class="kn">run:</span>
    <span class="n">gunzip</span> <span class="o">-</span><span class="n">c</span> <span class="err">$</span><span class="p">{</span><span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">sorted</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">input</span> <span class="k">if</span> <span class="s1">&#39;_R1_&#39;</span> <span class="ow">in</span> <span class="n">x</span><span class="p">]))}</span> <span class="o">|</span> <span class="n">head</span> <span class="o">-</span><span class="n">n</span> <span class="err">$</span><span class="p">{</span><span class="mi">4</span><span class="o">*</span><span class="n">sampling</span><span class="p">}</span> <span class="o">&gt;</span> <span class="err">$</span><span class="p">{</span><span class="n">output</span><span class="p">[</span><span class="mi">0</span><span class="p">]}</span>
    <span class="n">gunzip</span> <span class="o">-</span><span class="n">c</span> <span class="err">$</span><span class="p">{</span><span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">sorted</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">input</span> <span class="k">if</span> <span class="s1">&#39;_R2_&#39;</span> <span class="ow">in</span> <span class="n">x</span><span class="p">]))}</span> <span class="o">|</span> <span class="n">head</span> <span class="o">-</span><span class="n">n</span> <span class="err">$</span><span class="p">{</span><span class="mi">4</span><span class="o">*</span><span class="n">sampling</span><span class="p">}</span> <span class="o">&gt;</span> <span class="err">$</span><span class="p">{</span><span class="n">output</span><span class="p">[</span><span class="mi">1</span><span class="p">]}</span>

<span class="n">fastq_files</span> <span class="o">=</span> <span class="n">output</span>


<span class="p">[</span><span class="n">human_hg19_110</span><span class="p">]</span>
<span class="c1"># Use fastqc to check the quality of input reads</span>
<span class="c1">#</span>
<span class="kn">input:</span>
    <span class="n">fastq_files</span>

<span class="kn">depends:</span>
    <span class="n">executable</span><span class="p">(</span><span class="s1">&#39;fastqc&#39;</span><span class="p">),</span>
    <span class="n">executable</span><span class="p">(</span><span class="s1">&#39;gunzip&#39;</span><span class="p">)</span>

<span class="kn">output:</span>
    <span class="s1">&#39;${readqc_out}/${samplename}_fastqc.html&#39;</span><span class="p">,</span>
    <span class="s1">&#39;${readqc_out}/${samplename}_fastqc.zip&#39;</span>

<span class="kn">task:</span>

<span class="kn">run:</span>
    <span class="c1"># the input can be original .fast.gz file or extracted </span>
    <span class="err">$</span><span class="p">{</span><span class="s2">&quot;gunzip -c&quot;</span> <span class="k">if</span> <span class="nb">input</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.gz&quot;</span><span class="p">)</span> <span class="k">else</span> <span class="s2">&quot;cat&quot;</span><span class="p">}</span> <span class="err">$</span><span class="p">{</span><span class="nb">input</span><span class="p">}</span> <span class="o">|</span> <span class="n">fastqc</span> <span class="n">stdin</span> <span class="o">--</span><span class="n">outdir</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="n">readqc_out</span><span class="p">}</span>
    <span class="n">mv</span> <span class="err">$</span><span class="p">{</span><span class="n">readqc_out</span><span class="p">}</span><span class="o">/</span><span class="n">stdin_fastqc</span><span class="o">.</span><span class="n">html</span>  <span class="err">$</span><span class="p">{</span><span class="n">output</span><span class="p">[</span><span class="mi">0</span><span class="p">]}</span>
    <span class="n">mv</span> <span class="err">$</span><span class="p">{</span><span class="n">readqc_out</span><span class="p">}</span><span class="o">/</span><span class="n">stdin_fastqc</span><span class="o">.</span><span class="n">zip</span> <span class="err">$</span><span class="p">{</span><span class="n">output</span><span class="p">[</span><span class="mi">1</span><span class="p">]}</span>

<span class="p">[</span><span class="n">extract_fastqc_figures</span><span class="p">:</span> <span class="kp">provides</span><span class="o">=</span><span class="p">[</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">readqc_out</span><span class="p">,</span> <span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="p">(</span><span class="s1">&#39;per_base_quality.png&#39;</span><span class="p">,</span> <span class="s1">&#39;per_sequence_gc_content.png&#39;</span><span class="p">,</span> <span class="s1">&#39;per_base_sequence_content.png&#39;</span><span class="p">)]]</span>
<span class="kn">input:</span>
    <span class="s1">&#39;${readqc_out}/${samplename}_fastqc.zip&#39;</span>

<span class="kn">import</span> <span class="nn">zipfile</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="k">with</span> <span class="n">zipfile</span><span class="o">.</span><span class="n">ZipFile</span><span class="p">(</span><span class="nb">input</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="k">as</span> <span class="n">qczip</span><span class="p">:</span>
    <span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="n">qczip</span><span class="o">.</span><span class="n">namelist</span><span class="p">():</span>
        <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">name</span><span class="p">)</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;per_base_quality.png&#39;</span><span class="p">,</span> <span class="s1">&#39;per_sequence_gc_content.png&#39;</span><span class="p">,</span> <span class="s1">&#39;per_base_sequence_content.png&#39;</span><span class="p">]:</span>
            <span class="n">qczip</span><span class="o">.</span><span class="n">extract</span><span class="p">(</span><span class="n">name</span><span class="p">,</span> <span class="n">readqc_out</span><span class="p">)</span>
            <span class="c1"># move the file to top directory</span>
            <span class="n">os</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">readqc_out</span><span class="p">,</span> <span class="n">name</span><span class="p">),</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">readqc_out</span><span class="p">,</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">name</span><span class="p">)))</span>


<span class="p">[</span><span class="n">download_ensgene</span><span class="p">:</span> <span class="kp">provides</span><span class="o">=</span><span class="n">genes_gtf</span><span class="p">]</span>
<span class="kn">download:</span>  <span class="s1">&#39;${annotation_dir}/Genes/&#39;</span>
    <span class="err">$</span><span class="p">{</span><span class="n">ensgene_url</span><span class="p">}</span>

<span class="p">[</span><span class="n">chr_gtf</span><span class="p">:</span> <span class="kp">provides</span> <span class="o">=</span> <span class="n">genes_with_chr_gtf</span><span class="p">]</span>
<span class="c1"># takes gtf and and chromosome names to gene name to avoid problems</span>
<span class="c1"># with genes with the same names on different chromosomes</span>
<span class="kn">input:</span>
    <span class="n">genes_gtf</span>

<span class="kn">output:</span>
    <span class="n">genes_with_chr_gtf</span>

<span class="kn">python:</span>
    <span class="c1">#</span>
    <span class="c1">#    Input format:</span>
    <span class="c1">#</span>
    <span class="c1">#        chr1 unknown exon 17233 17368 . - . gene_id &quot;WASH7P&quot;; gene_name &quot;WASH7P&quot;; transcript_id &quot;NR_024540&quot;; tss_id &quot;TSS8151&quot;;</span>
    <span class="c1">#</span>
    <span class="c1">#    Output format:</span>
    <span class="c1">#           chr1 unknown exon 17233 17368 . - . gene_id &quot;WASH7P&quot;; gene_name &quot;WASH7P.chr1&quot;; transcript_id &quot;NR_024540&quot;; tss_id &quot;TSS8151&quot;;</span>
    <span class="c1"># </span>
    <span class="kn">import</span> <span class="nn">re</span>
    <span class="kp">pattern</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="s1">r&#39;&#39;&#39;(gene_id\s+[&#39;&quot;]*)([\w.-]+)&#39;&#39;&#39;</span><span class="p">)</span>
    <span class="c1"># only work on selected chromosomes</span>
    <span class="n">allowed_chromosomes</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([</span><span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">25</span><span class="p">)]</span> <span class="o">+</span> <span class="p">[</span><span class="s1">&#39;chr{}&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">25</span><span class="p">)]</span> <span class="o">+</span> \
                <span class="p">[</span><span class="s1">&#39;X&#39;</span><span class="p">,</span> <span class="s1">&#39;Y&#39;</span><span class="p">,</span> <span class="s1">&#39;M&#39;</span><span class="p">,</span> <span class="s1">&#39;chrX&#39;</span><span class="p">,</span> <span class="s1">&#39;chrY&#39;</span><span class="p">,</span> <span class="s1">&#39;chrM&#39;</span><span class="p">])</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="nb">input</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="s1">&#39;rU&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">ifile</span><span class="p">,</span> <span class="nb">open</span><span class="p">(</span><span class="n">output</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">ofile</span><span class="p">:</span>
       <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">ifile</span><span class="p">:</span>
           <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;#&#39;</span><span class="p">):</span>
               <span class="k">continue</span>
           <span class="n">ls</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
           <span class="k">if</span> <span class="n">ls</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">allowed_chromosomes</span><span class="p">:</span>
               <span class="k">continue</span>
           <span class="n">ofile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="kp">pattern</span><span class="p">,</span> <span class="s1">r&#39;\1\2.{}&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">ls</span><span class="p">[</span><span class="mi">0</span><span class="p">]),</span> <span class="n">line</span><span class="p">))</span> 

<span class="p">[</span><span class="n">download_igenome</span><span class="p">:</span> <span class="kp">provides</span><span class="o">=</span><span class="n">refgene_bed</span><span class="p">]</span>
<span class="kn">download:</span> <span class="n">dest_dir</span><span class="o">=</span><span class="n">igenome_resource_dir</span>
    <span class="err">$</span><span class="p">{</span><span class="n">igenome_url</span><span class="p">}</span>

<span class="p">[</span><span class="n">human_hg19_400</span><span class="p">:</span> <span class="kp">shared</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;param&#39;</span><span class="p">:</span> <span class="s1">&#39;output[0]&#39;</span><span class="p">}]</span>
<span class="c1">#</span>
<span class="c1"># Align a subset of reads to obtain two key parameters parameters</span>
<span class="c1"># --mate-inner-dist and --mate-std-dev for full scale alignment</span>
<span class="c1">#</span>
<span class="kn">input:</span>
    <span class="n">fastq_files</span>

<span class="kn">depends:</span>
    <span class="n">genes_with_chr_gtf</span><span class="p">,</span>
    <span class="n">refgene_bed</span><span class="p">,</span>
    <span class="n">executable</span><span class="p">(</span><span class="s1">&#39;tophat2&#39;</span><span class="p">),</span>
    <span class="n">executable</span><span class="p">(</span><span class="s1">&#39;bowtie&#39;</span><span class="p">),</span>
    <span class="n">executable</span><span class="p">(</span><span class="s1">&#39;inner_distance.py&#39;</span><span class="p">)</span>

<span class="kn">output:</span>
    <span class="s1">&#39;${temp_dir}/${samplename}.inner_distance_freq.txt&#39;</span>

<span class="n">test_files</span> <span class="o">=</span> <span class="s1">&#39;${temp_dir}/${samplename}_R1_test.fastq&#39;</span><span class="p">,</span> <span class="s1">&#39;${temp_dir}/${samplename}_R2_test.fastq&#39;</span>

<span class="kn">task:</span>

<span class="kn">run:</span>
    <span class="err">$</span><span class="p">{</span><span class="s1">&#39;gunzip -c&#39;</span> <span class="k">if</span> <span class="nb">input</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s1">&#39;.gz&#39;</span><span class="p">)</span> <span class="k">else</span> <span class="s1">&#39;cat&#39;</span><span class="p">}</span>  <span class="err">$</span><span class="p">{</span><span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">sorted</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">input</span> <span class="k">if</span> <span class="s1">&#39;_R1_&#39;</span> <span class="ow">in</span> <span class="n">x</span><span class="p">]))}</span> \
        <span class="o">|</span> <span class="n">head</span> <span class="o">-</span><span class="n">n</span> <span class="mi">40000</span> <span class="o">&gt;</span> <span class="err">$</span><span class="p">{</span><span class="n">test_files</span><span class="p">[</span><span class="mi">0</span><span class="p">]}</span>

    <span class="err">$</span><span class="p">{</span><span class="s1">&#39;gunzip -c&#39;</span> <span class="k">if</span> <span class="nb">input</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s1">&#39;.gz&#39;</span><span class="p">)</span> <span class="k">else</span> <span class="s1">&#39;cat&#39;</span><span class="p">}</span>  <span class="err">$</span><span class="p">{</span><span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">sorted</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">input</span> <span class="k">if</span> <span class="s1">&#39;_R2_&#39;</span> <span class="ow">in</span> <span class="n">x</span><span class="p">]))}</span> \
        <span class="o">|</span> <span class="n">head</span> <span class="o">-</span><span class="n">n</span> <span class="mi">40000</span> <span class="o">&gt;</span> <span class="err">$</span><span class="p">{</span><span class="n">test_files</span><span class="p">[</span><span class="mi">1</span><span class="p">]}</span>

<span class="kn">run:</span>
    <span class="n">tophat2</span> <span class="o">--</span><span class="n">zpacker</span> <span class="mi">0</span> <span class="o">-</span><span class="n">no</span><span class="o">-</span><span class="n">coverage</span><span class="o">-</span><span class="n">search</span> <span class="o">--</span><span class="n">library</span><span class="o">-</span><span class="nb">type</span> <span class="n">fr</span><span class="o">-</span><span class="n">unstranded</span> \
        <span class="o">--</span><span class="nb">max</span><span class="o">-</span><span class="n">multihits</span> <span class="mi">20</span> <span class="o">--</span><span class="n">prefilter</span><span class="o">-</span><span class="n">multihits</span> <span class="o">--</span><span class="n">num</span><span class="o">-</span><span class="n">threads</span> <span class="mi">8</span> \
        <span class="o">--</span><span class="n">GTF</span> <span class="err">$</span><span class="p">{</span><span class="n">genes_with_chr_gtf</span><span class="p">}</span> <span class="o">--</span><span class="n">segment</span><span class="o">-</span><span class="n">length</span> <span class="mi">25</span> \
        <span class="o">--</span><span class="n">transcriptome</span><span class="o">-</span><span class="n">index</span> <span class="err">$</span><span class="p">{</span><span class="n">transcriptome_dir</span><span class="p">}</span><span class="o">/</span><span class="n">genes</span> <span class="o">--</span><span class="n">bowtie1</span> \
        <span class="o">--</span><span class="n">output</span><span class="o">-</span><span class="nb">dir</span> <span class="err">$</span><span class="p">{</span><span class="n">temp_dir</span><span class="p">}</span> <span class="err">$</span><span class="p">{</span><span class="n">reference_dir</span><span class="p">}</span><span class="o">/</span><span class="n">BowtieIndex</span><span class="o">/</span><span class="n">genome</span> \
        <span class="err">$</span><span class="p">{</span><span class="n">test_files</span><span class="p">[</span><span class="mi">0</span><span class="p">]}</span> <span class="err">$</span><span class="p">{</span><span class="n">test_files</span><span class="p">[</span><span class="mi">1</span><span class="p">]}</span>

<span class="kn">run:</span>
    <span class="n">inner_distance</span><span class="o">.</span><span class="n">py</span> <span class="o">-</span><span class="n">r</span> <span class="err">$</span><span class="p">{</span><span class="n">refgene_bed</span><span class="p">}</span> <span class="o">-</span><span class="n">i</span> <span class="err">$</span><span class="p">{</span><span class="n">temp_dir</span><span class="p">}</span><span class="o">/</span><span class="n">accepted_hits</span><span class="o">.</span><span class="n">bam</span> <span class="o">-</span><span class="n">o</span> <span class="err">$</span><span class="p">{</span><span class="n">temp_dir</span><span class="p">}</span><span class="o">/</span><span class="err">$</span><span class="p">{</span><span class="n">samplename</span><span class="p">}</span>


<span class="p">[</span><span class="n">human_hg19_440</span><span class="p">:</span> <span class="kp">shared</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;accepted_hits&#39;</span><span class="p">:</span><span class="s1">&#39;output[0]&#39;</span><span class="p">,</span> <span class="s1">&#39;mate_inner_dist&#39;</span><span class="p">:</span><span class="s1">&#39;mate_inner_dist&#39;</span><span class="p">,</span> <span class="s1">&#39;mate_std_dev&#39;</span><span class="p">:</span><span class="s1">&#39;mate_std_dev&#39;</span><span class="p">}]</span>
<span class="c1">#Alignment (Tophat2): Running tophat2 with fusion search.</span>
<span class="c1">#Alignment Preparation: Calculate mean and standard devision of inner distance and set parameters</span>
<span class="c1"># --mate-inner-dist and --mate-std-dev</span>
<span class="kn">import</span> <span class="nn">math</span>
<span class="n">values</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">param</span><span class="p">)</span> <span class="k">as</span> <span class="n">freq</span><span class="p">:</span>
    <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">freq</span><span class="p">:</span>
        <span class="n">s</span><span class="p">,</span> <span class="n">e</span><span class="p">,</span> <span class="n">c</span> <span class="o">=</span> <span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">()]</span>
        <span class="n">values</span><span class="o">.</span><span class="n">extend</span><span class="p">([</span><span class="nb">int</span><span class="p">(</span><span class="n">s</span><span class="o">+</span><span class="n">e</span><span class="p">)</span><span class="o">/</span><span class="mf">2.</span><span class="p">]</span> <span class="o">*</span> <span class="n">c</span><span class="p">)</span>
<span class="k">if</span> <span class="ow">not</span> <span class="n">values</span><span class="p">:</span>
    <span class="c1"># no mapped reads...</span>
    <span class="n">mate_inner_dist</span> <span class="o">=</span> <span class="mi">50</span>
    <span class="n">mate_std_dev</span> <span class="o">=</span> <span class="mi">20</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;Mean and standard deviation of inner distance is set to default values 50 and 20&#39;</span><span class="p">)</span>
<span class="kn">else:</span>
    <span class="n">mean</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">values</span><span class="p">)</span><span class="o">/</span><span class="nb">len</span><span class="p">(</span><span class="n">values</span><span class="p">)</span>
    <span class="n">mean_sq</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">([</span><span class="n">x</span><span class="o">*</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">values</span><span class="p">])</span><span class="o">/</span><span class="nb">len</span><span class="p">(</span><span class="n">values</span><span class="p">)</span>
    <span class="n">sd</span> <span class="o">=</span> <span class="n">math</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">mean_sq</span> <span class="o">-</span> <span class="n">mean</span><span class="o">*</span><span class="n">mean</span><span class="p">)</span>
    <span class="n">mate_inner_dist</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">mean</span><span class="p">)</span>
    <span class="n">mate_std_dev</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">sd</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;Mean and standard deviation of inner distance is set to {} and {}&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">mean</span><span class="p">),</span> <span class="nb">int</span><span class="p">(</span><span class="n">sd</span><span class="p">)))</span>

<span class="kn">input:</span>
    <span class="n">fastq_files</span>

<span class="kn">depends:</span>
    <span class="n">genes_with_chr_gtf</span>

<span class="kn">output:</span>
    <span class="s1">&#39;${alignment_out}/accepted_hits.bam&#39;</span>

<span class="c1"># use --zpacker 0 to disable compression of temporary file</span>
<span class="c1"># which makes the program run faster</span>
<span class="kn">run:</span>
    <span class="k">if</span> <span class="p">[[</span> <span class="o">-</span><span class="n">f</span> <span class="err">$</span><span class="p">{</span><span class="n">alignment_out</span><span class="p">}</span><span class="o">/</span><span class="n">logs</span><span class="o">/</span><span class="kp">run</span><span class="o">.</span><span class="n">log</span> <span class="p">]]</span> <span class="o">&amp;&amp;</span> <span class="p">[[</span> <span class="err">!</span> <span class="o">-</span><span class="n">f</span> <span class="err">$</span><span class="p">{</span><span class="n">alignment_out</span><span class="p">}</span><span class="o">/</span><span class="n">accepted_hits</span><span class="o">.</span><span class="n">bam</span> <span class="p">]];</span>  \
        <span class="n">then</span>  \
        <span class="n">tophat2</span> <span class="o">--</span><span class="n">resume</span> <span class="err">$</span><span class="p">{</span><span class="n">alignment_out</span><span class="p">}</span> <span class="o">||</span>  \
        <span class="n">tophat2</span>  \
        <span class="o">-</span><span class="n">no</span><span class="o">-</span><span class="n">coverage</span><span class="o">-</span><span class="n">search</span>  \
        <span class="o">--</span><span class="n">library</span><span class="o">-</span><span class="nb">type</span> <span class="n">fr</span><span class="o">-</span><span class="n">unstranded</span>  \
        <span class="o">--</span><span class="nb">max</span><span class="o">-</span><span class="n">multihits</span> <span class="mi">20</span>  \
        <span class="o">--</span><span class="n">prefilter</span><span class="o">-</span><span class="n">multihits</span>  \
        <span class="o">--</span><span class="n">num</span><span class="o">-</span><span class="n">threads</span> <span class="mi">8</span>  \
        <span class="o">--</span><span class="n">GTF</span> <span class="err">$</span><span class="p">{</span><span class="n">genes_with_chr_gtf</span><span class="p">}</span>  \
        <span class="o">--</span><span class="n">segment</span><span class="o">-</span><span class="n">length</span> <span class="mi">25</span>  \
        <span class="o">--</span><span class="n">keep</span><span class="o">-</span><span class="n">fasta</span><span class="o">-</span><span class="n">order</span>  \
        <span class="o">--</span><span class="n">transcriptome</span><span class="o">-</span><span class="n">index</span> <span class="err">$</span><span class="p">{</span><span class="n">transcriptome_dir</span><span class="p">}</span><span class="o">/</span><span class="n">genes</span>  \
        <span class="o">--</span><span class="n">fusion</span><span class="o">-</span><span class="n">search</span> <span class="o">--</span><span class="n">fusion</span><span class="o">-</span><span class="n">ignore</span><span class="o">-</span><span class="n">chromosomes</span> <span class="n">chrM</span><span class="p">,</span><span class="n">M</span>  \
        <span class="o">--</span><span class="n">fusion</span><span class="o">-</span><span class="nb">min</span><span class="o">-</span><span class="n">dist</span> <span class="mi">50000</span>  \
        <span class="o">--</span><span class="n">bowtie1</span>  \
        <span class="o">--</span><span class="n">mate</span><span class="o">-</span><span class="n">inner</span><span class="o">-</span><span class="n">dist</span> <span class="err">$</span><span class="p">{</span><span class="n">mate_inner_dist</span><span class="p">}</span>  \
        <span class="o">--</span><span class="n">mate</span><span class="o">-</span><span class="n">std</span><span class="o">-</span><span class="n">dev</span> <span class="err">$</span><span class="p">{</span><span class="n">mate_std_dev</span><span class="p">}</span>  \
        <span class="o">--</span><span class="n">output</span><span class="o">-</span><span class="nb">dir</span> <span class="err">$</span><span class="p">{</span><span class="n">alignment_out</span><span class="p">}</span>  \
        <span class="err">$</span><span class="p">{</span><span class="n">reference_dir</span><span class="p">}</span><span class="o">/</span><span class="n">BowtieIndex</span><span class="o">/</span><span class="n">genome</span>  \
        <span class="err">$</span><span class="p">{</span><span class="s1">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">sorted</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">input</span> <span class="k">if</span> <span class="s1">&#39;_R1_&#39;</span> <span class="ow">in</span> <span class="n">x</span><span class="p">]))}</span>  \
        <span class="err">$</span><span class="p">{</span><span class="s1">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">sorted</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">input</span> <span class="k">if</span> <span class="s1">&#39;_R2_&#39;</span> <span class="ow">in</span> <span class="n">x</span><span class="p">]))}</span>  \
        <span class="p">;</span>  \
    <span class="k">else</span>   \
        <span class="n">tophat2</span>  \
        <span class="o">-</span><span class="n">no</span><span class="o">-</span><span class="n">coverage</span><span class="o">-</span><span class="n">search</span>  \
        <span class="o">--</span><span class="n">library</span><span class="o">-</span><span class="nb">type</span> <span class="n">fr</span><span class="o">-</span><span class="n">unstranded</span>   \
        <span class="o">--</span><span class="nb">max</span><span class="o">-</span><span class="n">multihits</span> <span class="mi">20</span>   \
        <span class="o">--</span><span class="n">prefilter</span><span class="o">-</span><span class="n">multihits</span>  \
        <span class="o">--</span><span class="n">num</span><span class="o">-</span><span class="n">threads</span> <span class="mi">8</span>   \
        <span class="o">--</span><span class="n">GTF</span> <span class="err">$</span><span class="p">{</span><span class="n">genes_with_chr_gtf</span><span class="p">}</span>  \
        <span class="o">--</span><span class="n">segment</span><span class="o">-</span><span class="n">length</span> <span class="mi">25</span>   \
        <span class="o">--</span><span class="n">keep</span><span class="o">-</span><span class="n">fasta</span><span class="o">-</span><span class="n">order</span>   \
        <span class="o">--</span><span class="n">transcriptome</span><span class="o">-</span><span class="n">index</span> <span class="err">$</span><span class="p">{</span><span class="n">transcriptome_dir</span><span class="p">}</span><span class="o">/</span><span class="n">genes</span>  \
        <span class="o">--</span><span class="n">fusion</span><span class="o">-</span><span class="n">search</span> <span class="o">--</span><span class="n">fusion</span><span class="o">-</span><span class="n">ignore</span><span class="o">-</span><span class="n">chromosomes</span> <span class="n">chrM</span><span class="p">,</span><span class="n">M</span>   \
        <span class="o">--</span><span class="n">fusion</span><span class="o">-</span><span class="nb">min</span><span class="o">-</span><span class="n">dist</span> <span class="mi">50000</span>  \
        <span class="o">--</span><span class="n">bowtie1</span>   \
        <span class="o">--</span><span class="n">mate</span><span class="o">-</span><span class="n">inner</span><span class="o">-</span><span class="n">dist</span> <span class="err">$</span><span class="p">{</span><span class="n">mate_inner_dist</span><span class="p">}</span>  \
        <span class="o">--</span><span class="n">mate</span><span class="o">-</span><span class="n">std</span><span class="o">-</span><span class="n">dev</span> <span class="err">$</span><span class="p">{</span><span class="n">mate_std_dev</span><span class="p">}</span>  \
        <span class="o">--</span><span class="n">output</span><span class="o">-</span><span class="nb">dir</span> <span class="err">$</span><span class="p">{</span><span class="n">alignment_out</span><span class="p">}</span>  \
        <span class="err">$</span><span class="p">{</span><span class="n">reference_dir</span><span class="p">}</span><span class="o">/</span><span class="n">BowtieIndex</span><span class="o">/</span><span class="n">genome</span>  \
        <span class="err">$</span><span class="p">{</span><span class="s1">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">sorted</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">input</span> <span class="k">if</span> <span class="s1">&#39;_R1_&#39;</span> <span class="ow">in</span> <span class="n">x</span><span class="p">]))}</span>  \
        <span class="err">$</span><span class="p">{</span><span class="s1">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">sorted</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">input</span> <span class="k">if</span> <span class="s1">&#39;_R2_&#39;</span> <span class="ow">in</span> <span class="n">x</span><span class="p">]))};</span>  \
        <span class="n">fi</span> 


<span class="p">[</span><span class="n">index_bam</span><span class="p">:</span> <span class="kp">provides</span><span class="o">=</span><span class="s1">&#39;{filename}.bam.bai&#39;</span><span class="p">]</span>
<span class="c1"># Index generated bam file</span>
<span class="kn">input:</span>
    <span class="s1">&#39;${filename}.bam&#39;</span>

<span class="kn">run:</span>
    <span class="n">samtools</span> <span class="n">index</span> <span class="err">$</span><span class="p">{</span><span class="nb">input</span><span class="p">}</span>


<span class="p">[</span><span class="n">human_hg19_460</span><span class="p">]</span>
<span class="c1"># Add ReadGroup information to sorted bam file.</span>

<span class="c1"># Read group identifier, which should be an unique identifier for each</span>
<span class="c1"># instrument run (e.g. flowcell number + lane name + number). The filename before</span>
<span class="c1"># the &#39;_R1_&#39; part will be used if this option is left unspecified.</span>
<span class="kn">parameter:</span> <span class="n">rgid</span> <span class="o">=</span> <span class="s1">&#39;NA&#39;</span>
<span class="kp">warn_if</span><span class="p">(</span><span class="n">rgid</span> <span class="o">==</span> <span class="s1">&#39;NA&#39;</span><span class="p">,</span>
    <span class="s1">&#39;A default (NA) value is used for parameter --rgid (intrument run ID)&#39;</span><span class="p">)</span>

<span class="c1"># Platform/technology used to produce the read. Valid values include</span>
<span class="c1"># ILLUMINA, SOLID, LS454, HELICOS and PACBIO</span>
<span class="kn">parameter:</span> <span class="n">rgpl</span> <span class="o">=</span> <span class="s1">&#39;ILLUMINA&#39;</span>
<span class="c1"># check input parameters</span>
<span class="kp">warn_if</span><span class="p">(</span><span class="ow">not</span> <span class="n">rgpl</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;ILLUMINA&#39;</span><span class="p">,</span> <span class="s1">&#39;SOLID&#39;</span><span class="p">,</span> <span class="s1">&#39;LS454&#39;</span><span class="p">,</span> <span class="s1">&#39;HELICOS&#39;</span><span class="p">,</span> <span class="s1">&#39;PACBIO&#39;</span><span class="p">],</span>
    <span class="s1">&#39;PL (platform) value if not one of ILLUMINA, SOLID, LS454, HELICOS and PACBIO&#39;</span><span class="p">)</span>

<span class="c1"># DNA preparation library identity.</span>
<span class="kn">parameter:</span> <span class="n">rglb</span> <span class="o">=</span> <span class="s1">&#39;NA&#39;</span>
<span class="kp">warn_if</span><span class="p">(</span><span class="n">rglb</span> <span class="o">==</span> <span class="s1">&#39;NA&#39;</span><span class="p">,</span>
    <span class="s1">&#39;A default (NA) value is used for parameter --rglb (library identification string)&#39;</span><span class="p">)</span>
<span class="c1">#</span>
<span class="c1"># Platform unit (e.g. barcode). This field is not used by GATK.</span>
<span class="kn">parameter:</span> <span class="n">rgpu</span> <span class="o">=</span> <span class="s1">&#39;NA&#39;</span>
<span class="kp">warn_if</span><span class="p">(</span><span class="n">rgpu</span> <span class="o">==</span> <span class="s1">&#39;NA&#39;</span><span class="p">,</span>
    <span class="s1">&#39;A default (NA) value is used for parameter --rgpu (platform unit, barcode)&#39;</span><span class="p">)</span>


<span class="kn">input:</span>
     <span class="n">accepted_hits</span>

<span class="kn">output:</span>
    <span class="s1">&#39;${temp_dir}/${samplename}-with_rg.bam&#39;</span>

<span class="kn">run:</span>
    <span class="n">java</span> <span class="o">-</span><span class="n">Xmx16g</span> <span class="o">-</span><span class="n">Xms512m</span>   \
        <span class="o">-</span><span class="n">Djava</span><span class="o">.</span><span class="n">io</span><span class="o">.</span><span class="n">tmpdir</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="n">temp_dir</span><span class="p">}</span>   \
        <span class="o">-</span><span class="n">jar</span> <span class="err">$</span><span class="p">{</span><span class="n">picard_path</span><span class="p">}</span><span class="o">/</span><span class="n">AddOrReplaceReadGroups</span><span class="o">.</span><span class="n">jar</span>  \
        <span class="nb">input</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="nb">input</span><span class="p">}</span>  \
        <span class="n">OUTPUT</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="n">temp_dir</span><span class="p">}</span><span class="o">/</span><span class="err">$</span><span class="p">{</span><span class="n">samplename</span><span class="p">}</span><span class="o">-</span><span class="n">with_rg</span><span class="o">.</span><span class="n">bam</span>  \
        <span class="n">VALIDATION_STRINGENCY</span><span class="o">=</span><span class="n">LENIENT</span>  \
        <span class="n">RGID</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="n">rgid</span><span class="p">}</span>  \
        <span class="n">RGSM</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="n">samplename</span><span class="p">}</span>  \
        <span class="n">RGPL</span><span class="o">=</span><span class="n">ILLUMINA</span>  \
        <span class="n">RGLB</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="n">rglb</span><span class="p">}</span>  \
        <span class="n">RGPU</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="n">rgpu</span><span class="p">}</span>  \
        <span class="n">RGCN</span><span class="o">=</span><span class="n">IPCT</span>

<span class="p">[</span><span class="n">human_hg19_470</span><span class="p">:</span> <span class="kp">shared</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;sorted_unique&#39;</span><span class="p">:</span><span class="s1">&#39;output[0]&#39;</span><span class="p">}]</span>
<span class="c1">#Alignment: Remove fusion reads in preparation for variant calling.</span>
<span class="c1"># step 475, samtools index is defined above</span>
<span class="kn">output:</span>
    <span class="s1">&#39;${alignment_out}/${samplename}-sorted_uniq_nonF.bam&#39;</span>

<span class="kn">run:</span>
    <span class="n">samtools</span> <span class="n">view</span> <span class="o">-</span><span class="n">h</span> <span class="err">$</span><span class="p">{</span><span class="nb">input</span><span class="p">}</span> \
        <span class="o">|</span> <span class="n">awk</span> <span class="o">-</span><span class="n">F</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span> <span class="s1">&#39;{ if($0 ~ &quot;^@&quot;) {print} else { for(i=12;i&lt;=NF;i++){ if ($i ~ &quot;NH:i:1$&quot;){print}} } }&#39;</span> \
        <span class="o">|</span> <span class="n">awk</span> <span class="s1">&#39;{ if($0 ~ &quot;^@&quot;) {print} else { if ($0 !~ &quot;XF:Z&quot;) {print} }  }&#39;</span> \
        <span class="o">|</span> <span class="n">samtools</span> <span class="n">view</span> <span class="o">-</span><span class="n">bS</span> <span class="o">-</span> <span class="o">&gt;</span> <span class="err">$</span><span class="p">{</span><span class="n">alignment_out</span><span class="p">}</span><span class="o">/</span><span class="err">$</span><span class="p">{</span><span class="n">samplename</span><span class="p">}</span><span class="o">-</span><span class="n">sorted_uniq_nonF</span><span class="o">.</span><span class="n">bam</span>


<span class="p">[</span><span class="n">human_hg19_480</span><span class="p">]</span>
<span class="c1">#Alignment: Count the number of junction reads. </span>
<span class="kn">input:</span>
    <span class="n">accepted_hits</span>

<span class="kn">output:</span>
    <span class="s1">&#39;${alignment_out}/junction.count&#39;</span>

<span class="kn">run:</span>
    <span class="n">samtools</span> <span class="n">view</span> <span class="o">-</span><span class="n">h</span> <span class="err">$</span><span class="p">{</span><span class="nb">input</span><span class="p">}</span> \
   <span class="o">|</span> <span class="n">awk</span> <span class="s1">&#39;$6 ~/N/&#39;</span> <span class="o">|</span> <span class="n">awk</span> <span class="s1">&#39;{ if ($9 ~ /^-/) {print $1&quot;</span><span class="se">\t</span><span class="s1">-&quot;} else print $1&quot;</span><span class="se">\t</span><span class="s1">+&quot; }&#39;</span> \
   <span class="o">|</span> <span class="n">sort</span> <span class="o">-</span><span class="n">T</span> <span class="err">$</span><span class="p">{</span><span class="n">temp_dir</span><span class="p">}</span> <span class="o">-</span><span class="n">u</span> <span class="o">|</span> <span class="n">wc</span> <span class="o">-</span><span class="n">l</span> <span class="o">&gt;</span> <span class="err">$</span><span class="p">{</span><span class="n">alignment_out</span><span class="p">}</span><span class="o">/</span><span class="n">junction</span><span class="o">.</span><span class="n">count</span>


<span class="p">[</span><span class="n">human_hg19_500</span> <span class="p">(</span><span class="n">gene</span> <span class="n">body</span> <span class="n">coverage</span><span class="p">):</span> <span class="kp">shared</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;rseqc_pdf&#39;</span><span class="p">:</span><span class="s1">&#39;output&#39;</span><span class="p">}]</span>
<span class="c1"># this version uses BAM files to count geneBody coverage, using .bw might be a more resource</span>
<span class="c1"># saving idea but needs more testing, especially because the output of bigWig is currently normalized</span>
<span class="c1"># and does not show the correct depth of coverage.</span>
<span class="kn">input:</span>
    <span class="n">sorted_unique</span>

<span class="kn">depends:</span>
    <span class="n">refgene_bed</span>

<span class="kn">output:</span>
    <span class="s1">&#39;${alignmentqc_out}/${samplename}.geneBodyCoverage.curves.pdf&#39;</span>

<span class="c1"># this job can take a long time to execute so we tend to execute as a separate job</span>
<span class="kn">task:</span>
<span class="kn">run:</span>
    <span class="n">geneBody_coverage</span><span class="o">.</span><span class="n">py</span> <span class="o">-</span><span class="n">r</span> <span class="err">$</span><span class="p">{</span><span class="n">refgene_bed</span><span class="p">}</span> <span class="o">-</span><span class="n">i</span> <span class="err">$</span><span class="p">{</span><span class="nb">input</span><span class="p">}</span> <span class="o">-</span><span class="n">o</span> <span class="err">$</span><span class="p">{</span><span class="n">alignmentqc_out</span><span class="p">}</span><span class="o">/</span><span class="err">$</span><span class="p">{</span><span class="n">samplename</span><span class="p">}</span>

<span class="c1">#[human_hg19_560]</span>
<span class="c1">#RSeQC: Coverage of gene body</span>
<span class="c1">#RunCommand(&#39;geneBody_coverage2.py -r ${refgene_bed} -i ${input551: input551[0][:-3] + &quot;bw&quot;} -t pdf -o ${alignmentqc_out}/${samplename}&#39;,</span>
<span class="c1">#    output=&#39;${alignmentqc_out}/${samplename}.geneBodyCoverage.curves.pdf&#39;)</span>

<span class="p">[</span><span class="n">human_hg19_510</span><span class="p">]</span>
<span class="c1">#RSeQC: Check if current sequencing depth is deep enough to perform alternative splicing analyses</span>
<span class="c1">#RSeQC: Annotating splicing event and splicing junction.</span>
<span class="c1">#RSeQC: Plot the distribution of inner distance between paired reads, which can be used to </span>
<span class="c1">#   determine parameters --mate-inner-dist and --mate-std-dev of the pipeline.</span>
<span class="c1">#RSeQC: read duplication plot</span>
<span class="c1">#RSeQC: Read distribution (statistics)</span>
<span class="c1">#RSeQC: Generate scaled wig track for depth</span>
<span class="kn">input:</span>
    <span class="n">accepted_hits</span>

<span class="kn">output:</span>
    <span class="s1">&#39;${alignmentqc_out}/${samplename}.junctionSaturation_plot.pdf&#39;</span><span class="p">,</span>
    <span class="s1">&#39;${alignmentqc_out}/${samplename}.splice_events.pdf&#39;</span><span class="p">,</span>
    <span class="s1">&#39;${alignmentqc_out}/${samplename}.splice_junction.pdf&#39;</span><span class="p">,</span>
    <span class="s1">&#39;${alignmentqc_out}/${samplename}.inner_distance_plot.pdf&#39;</span><span class="p">,</span>
    <span class="s1">&#39;${alignmentqc_out}//${samplename}.DupRate_plot.pdf&#39;</span><span class="p">,</span>
    <span class="s1">&#39;${alignmentqc_out}//${samplename}.read_distribution&#39;</span>

<span class="kn">task:</span>
<span class="kn">run:</span>
    <span class="n">junction_saturation</span><span class="o">.</span><span class="n">py</span>  <span class="o">-</span><span class="n">r</span> <span class="err">$</span><span class="p">{</span><span class="n">refgene_bed</span><span class="p">}</span> <span class="o">-</span><span class="n">i</span> <span class="err">$</span><span class="p">{</span><span class="nb">input</span><span class="p">}</span> <span class="o">-</span><span class="n">o</span> <span class="err">$</span><span class="p">{</span><span class="n">alignmentqc_out</span><span class="p">}</span><span class="o">/</span><span class="err">$</span><span class="p">{</span><span class="n">samplename</span><span class="p">}</span>
    <span class="n">junction_annotation</span><span class="o">.</span><span class="n">py</span> <span class="o">-</span><span class="n">r</span> <span class="err">$</span><span class="p">{</span><span class="n">refgene_bed</span><span class="p">}</span> <span class="o">-</span><span class="n">i</span> <span class="err">$</span><span class="p">{</span><span class="nb">input</span><span class="p">}</span> <span class="o">-</span><span class="n">o</span> <span class="err">$</span><span class="p">{</span><span class="n">alignmentqc_out</span><span class="p">}</span><span class="o">/</span><span class="err">$</span><span class="p">{</span><span class="n">samplename</span><span class="p">}</span>
    <span class="n">inner_distance</span><span class="o">.</span><span class="n">py</span> <span class="o">-</span><span class="n">r</span> <span class="err">$</span><span class="p">{</span><span class="n">refgene_bed</span><span class="p">}</span> <span class="o">-</span><span class="n">i</span> <span class="err">$</span><span class="p">{</span><span class="nb">input</span><span class="p">}</span> <span class="o">-</span><span class="n">o</span> <span class="err">$</span><span class="p">{</span><span class="n">alignmentqc_out</span><span class="p">}</span><span class="o">/</span><span class="err">$</span><span class="p">{</span><span class="n">samplename</span><span class="p">}</span>
    <span class="n">read_duplication</span><span class="o">.</span><span class="n">py</span> <span class="o">-</span><span class="n">i</span> <span class="err">$</span><span class="p">{</span><span class="nb">input</span><span class="p">}</span> <span class="o">-</span><span class="n">o</span> <span class="err">$</span><span class="p">{</span><span class="n">alignmentqc_out</span><span class="p">}</span><span class="o">/</span><span class="err">$</span><span class="p">{</span><span class="n">samplename</span><span class="p">}</span>
    <span class="n">read_distribution</span><span class="o">.</span><span class="n">py</span> <span class="o">-</span><span class="n">r</span> <span class="err">$</span><span class="p">{</span><span class="n">refgene_bed</span><span class="p">}</span> <span class="o">-</span><span class="n">i</span> <span class="err">$</span><span class="p">{</span><span class="nb">input</span><span class="p">}</span> <span class="o">&gt;</span> <span class="err">$</span><span class="p">{</span><span class="n">alignmentqc_out</span><span class="p">}</span><span class="o">/</span><span class="err">$</span><span class="p">{</span><span class="n">samplename</span><span class="p">}</span><span class="o">.</span><span class="n">read_distribution</span>

<span class="p">[</span><span class="n">human_hg19_600</span> <span class="p">(</span><span class="n">SortSam</span><span class="p">):</span> <span class="kp">shared</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;sorted_bam&#39;</span><span class="p">:</span><span class="s1">&#39;output&#39;</span><span class="p">}]</span>
<span class="c1">#Count: Order reads by name so that the BAM file can be processed by htseq-count</span>
<span class="kn">input:</span>
    <span class="n">accepted_hits</span>

<span class="kn">output:</span>
    <span class="s1">&#39;${alignment_out}/${samplename}-sorted_by_id.bam&#39;</span>

<span class="kn">task:</span>
<span class="kn">run:</span>
    <span class="n">java</span> <span class="o">-</span><span class="n">Xmx16g</span> <span class="o">-</span><span class="n">Xms512m</span>   \
        <span class="o">-</span><span class="n">Djava</span><span class="o">.</span><span class="n">io</span><span class="o">.</span><span class="n">tmpdir</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="n">temp_dir</span><span class="p">}</span>  \
        <span class="o">-</span><span class="n">jar</span> <span class="err">$</span><span class="p">{</span><span class="n">picard_path</span><span class="p">}</span><span class="o">/</span><span class="n">SortSam</span><span class="o">.</span><span class="n">jar</span>  \
        <span class="nb">input</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="nb">input</span><span class="p">}</span>  \
        <span class="n">OUTPUT</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="n">alignment_out</span><span class="p">}</span><span class="o">/</span><span class="err">$</span><span class="p">{</span><span class="n">samplename</span><span class="p">}</span><span class="o">-</span><span class="n">sorted_by_id</span><span class="o">.</span><span class="n">bam</span>  \
        <span class="n">SO</span><span class="o">=</span><span class="n">queryname</span>   \
        <span class="n">MAX_RECORDS_IN_RAM</span><span class="o">=</span><span class="mi">1000000</span>   \
        <span class="n">TMP_DIR</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="n">temp_dir</span><span class="p">}</span>  \
        <span class="n">VALIDATION_STRINGENCY</span><span class="o">=</span><span class="n">SILENT</span>


<span class="p">[</span><span class="n">human_hg19_610</span> <span class="p">(</span><span class="n">htseq</span><span class="o">-</span><span class="n">count</span><span class="p">)]</span>
<span class="c1"># Running htseq-count to count reads for genes. This step uses the union</span>
<span class="c1">#    intersection mode, and a gene definition file with chromosome name added to gene_id.</span>
<span class="kn">input:</span>
    <span class="n">sorted_bam</span>

<span class="kn">output:</span>
    <span class="s1">&#39;${count_out}/${samplename}-gene_counts.tsv&#39;</span>

<span class="c1"># We cannot use --order=pos because htseq-count would fail with an error message </span>
<span class="kn">run:</span>
    <span class="n">htseq</span><span class="o">-</span><span class="n">count</span> <span class="o">--</span><span class="n">stranded</span><span class="o">=</span><span class="n">no</span> <span class="o">--</span><span class="n">format</span><span class="o">=</span><span class="n">bam</span> <span class="o">--</span><span class="nb">type</span><span class="o">=</span><span class="n">exon</span> \
     <span class="o">--</span><span class="n">mode</span><span class="o">=</span><span class="n">union</span> <span class="err">$</span><span class="p">{</span><span class="nb">input</span><span class="p">}</span> <span class="err">$</span><span class="p">{</span><span class="n">genes_with_chr_gtf</span><span class="p">}</span> \
     <span class="o">&gt;</span> <span class="err">$</span><span class="p">{</span><span class="n">count_out</span><span class="p">}</span><span class="o">/</span><span class="err">$</span><span class="p">{</span><span class="n">samplename</span><span class="p">}</span><span class="o">-</span><span class="n">gene_counts</span><span class="o">.</span><span class="n">tsv</span>


<span class="p">[</span><span class="n">get_genesize</span><span class="p">:</span> <span class="kp">provides</span><span class="o">=</span><span class="n">genesize_txt</span><span class="p">]</span>
<span class="c1">#    Genes.gtf file in the format of</span>
<span class="c1">#</span>
<span class="c1">#    chr1    unknown    exon    17233    17368    .    -    .    gene_id &quot;WASH7P&quot;; gene_name &quot;WASH7P&quot;; transcript_id &quot;NR_024540&quot;; tss_id &quot;TSS8151&quot;;</span>
<span class="c1">#    chr1    unknown    exon    17369    17436    .    -    .    gene_id &quot;MIR6859-2&quot;; gene_name &quot;MIR6859-2&quot;; transcript_id &quot;NR_107062&quot;; tss_id &quot;TSS24460&quot;;</span>
<span class="c1">#    chr1    unknown    exon    17369    17436    .    -    .    gene_id &quot;MIR6859-1&quot;; gene_name &quot;MIR6859-1&quot;; transcript_id &quot;NR_106918_2&quot;; tss_id &quot;TSS24460&quot;;</span>
<span class="c1">#</span>
<span class="c1"># output (string):</span>
<span class="c1">#    A text file in the format of</span>
<span class="c1">#</span>
<span class="c1">#    chr1  MIR6859-2  112324</span>
<span class="kn">input:</span>
    <span class="n">genes_gtf</span>


<span class="kn">python:</span>

    <span class="kn">import</span> <span class="nn">sys</span>
    <span class="k">def</span> <span class="nf">_size</span><span class="p">(</span><span class="n">features</span><span class="p">):</span>
        <span class="c1"># this is definitely not efficient but we do not really care because it will be run</span>
        <span class="c1"># only once. </span>
        <span class="n">bases</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">feature</span> <span class="ow">in</span> <span class="n">features</span><span class="p">:</span>
            <span class="c1"># Add 1 because GTF input is one based</span>
            <span class="n">bases</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">feature</span><span class="p">[</span><span class="mi">0</span><span class="p">]),</span> <span class="nb">int</span><span class="p">(</span><span class="n">feature</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="o">+</span> <span class="mi">1</span><span class="p">))</span>
        <span class="k">return</span> <span class="nb">len</span><span class="p">(</span><span class="n">bases</span><span class="p">)</span>

    <span class="c1"># only work on selected chromosomes</span>
    <span class="n">allowed_chromosomes</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([</span><span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">25</span><span class="p">)]</span> <span class="o">+</span> <span class="p">[</span><span class="s1">&#39;chr{}&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">25</span><span class="p">)]</span> <span class="o">+</span> \
                <span class="p">[</span><span class="s1">&#39;X&#39;</span><span class="p">,</span> <span class="s1">&#39;Y&#39;</span><span class="p">,</span> <span class="s1">&#39;M&#39;</span><span class="p">,</span> <span class="s1">&#39;chrX&#39;</span><span class="p">,</span> <span class="s1">&#39;chrY&#39;</span><span class="p">,</span> <span class="s1">&#39;chrM&#39;</span><span class="p">])</span>

    <span class="n">gene_size</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="nb">input</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="s1">&#39;rU&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">ifile</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">ifile</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;#&#39;</span><span class="p">):</span>
                <span class="k">continue</span>
            <span class="n">ls</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
            <span class="c1">#</span>
            <span class="c1"># ls[0]: seqname (chromosome name)</span>
            <span class="c1"># ls[1]: source (--ignored--)</span>
            <span class="c1"># ls[2]: feature (exon only)</span>
            <span class="c1"># ls[3]: start position (1-based)</span>
            <span class="c1"># ls[4]: end position (1-based)</span>
            <span class="c1"># ls[5]: score (--ignored---)</span>
            <span class="c1"># ls[6]: strand (--ignored--)</span>
            <span class="c1"># ls[7]: frame (--ignored--)</span>
            <span class="c1"># ls[8]: attributes, look for gene_id</span>
            <span class="c1">#</span>
            <span class="k">if</span> <span class="n">ls</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">allowwd_chromosomes</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">ls</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">5</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="k">if</span> <span class="n">ls</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="o">!=</span> <span class="s1">&#39;exon&#39;</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">gene_id</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">rsplit</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">strip</span><span class="p">(</span><span class="s1">&#39;&quot;&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">strip</span><span class="p">(</span><span class="s2">&quot;&#39;&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">ls</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">)</span> <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;gene_id&#39;</span><span class="p">)]</span>
            <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                <span class="n">sys</span><span class="o">.</span><span class="n">stderr</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;Failed to parse {}: {}&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">line</span><span class="p">,</span> <span class="n">e</span><span class="p">))</span>
                <span class="k">continue</span>
            <span class="c1">#</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">gene_id</span><span class="p">:</span>
                <span class="n">sys</span><span class="o">.</span><span class="n">stderr</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;No gene_id for record {}&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">line</span><span class="p">))</span>
            <span class="c1">#</span>
            <span class="c1"># use gene_id.chr as key</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">key</span> <span class="o">=</span> <span class="n">gene_id</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;.&#39;</span> <span class="o">+</span> <span class="n">ls</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                <span class="k">if</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">gene_size</span><span class="p">:</span>
                    <span class="k">if</span> <span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">ls</span><span class="p">[</span><span class="mi">3</span><span class="p">]),</span> <span class="nb">int</span><span class="p">(</span><span class="n">ls</span><span class="p">[</span><span class="mi">4</span><span class="p">]))</span> <span class="o">!=</span> <span class="n">gene_size</span><span class="p">[</span><span class="n">key</span><span class="p">][</span><span class="mi">3</span><span class="p">][</span><span class="o">-</span><span class="mi">1</span><span class="p">]:</span>
                        <span class="n">gene_size</span><span class="p">[</span><span class="n">key</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">gene_size</span><span class="p">[</span><span class="n">key</span><span class="p">][</span><span class="mi">1</span><span class="p">],</span> <span class="nb">int</span><span class="p">(</span><span class="n">ls</span><span class="p">[</span><span class="mi">3</span><span class="p">]))</span>
                        <span class="n">gene_size</span><span class="p">[</span><span class="n">key</span><span class="p">][</span><span class="mi">2</span><span class="p">]</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">gene_size</span><span class="p">[</span><span class="n">key</span><span class="p">][</span><span class="mi">2</span><span class="p">],</span> <span class="nb">int</span><span class="p">(</span><span class="n">ls</span><span class="p">[</span><span class="mi">4</span><span class="p">]))</span>
                        <span class="n">gene_size</span><span class="p">[</span><span class="n">key</span><span class="p">][</span><span class="mi">3</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="nb">int</span><span class="p">(</span><span class="n">ls</span><span class="p">[</span><span class="mi">3</span><span class="p">]),</span> <span class="nb">int</span><span class="p">(</span><span class="n">ls</span><span class="p">[</span><span class="mi">4</span><span class="p">])))</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">gene_size</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">ls</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="nb">int</span><span class="p">(</span><span class="n">ls</span><span class="p">[</span><span class="mi">3</span><span class="p">]),</span> <span class="nb">int</span><span class="p">(</span><span class="n">ls</span><span class="p">[</span><span class="mi">4</span><span class="p">]),</span> <span class="p">[(</span><span class="nb">int</span><span class="p">(</span><span class="n">ls</span><span class="p">[</span><span class="mi">3</span><span class="p">]),</span> <span class="nb">int</span><span class="p">(</span><span class="n">ls</span><span class="p">[</span><span class="mi">4</span><span class="p">]))]]</span>
            <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                <span class="n">sys</span><span class="o">.</span><span class="n">stderr</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;Failed to store key {}: {}&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">key</span><span class="p">,</span> <span class="n">e</span><span class="p">))</span>
                <span class="k">continue</span>
    <span class="c1">#</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">output</span><span class="err">!</span><span class="n">r</span><span class="p">},</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">ofile</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">gene_size</span><span class="o">.</span><span class="n">keys</span><span class="p">()):</span>
            <span class="n">val</span> <span class="o">=</span> <span class="n">gene_size</span><span class="p">[</span><span class="n">key</span><span class="p">]</span>
            <span class="c1">#</span>
            <span class="c1"># Remove chr from key for output</span>
            <span class="n">ofile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;{0}</span><span class="se">\t</span><span class="s1">{1}</span><span class="se">\t</span><span class="s1">{2}</span><span class="se">\t</span><span class="s1">{3}</span><span class="se">\t</span><span class="s1">{4}</span><span class="se">\n</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">val</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">key</span><span class="p">[:</span><span class="o">-</span><span class="nb">len</span><span class="p">(</span><span class="n">val</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">val</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">val</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span> 
                <span class="n">_size</span><span class="p">(</span><span class="n">val</span><span class="p">[</span><span class="mi">3</span><span class="p">])))</span>


<span class="p">[</span><span class="n">flagstat</span><span class="p">:</span> <span class="kp">provides</span><span class="o">=</span><span class="s1">&#39;{filename}.bam.flagstat&#39;</span><span class="p">]</span>
<span class="c1"># Alignment (samtools): Generating statistics of aligned reads.</span>
<span class="kn">input:</span>
    <span class="s1">&#39;${filename}.bam&#39;</span>

<span class="kn">run:</span>
    <span class="n">samtools</span> <span class="n">flagstat</span> <span class="err">$</span><span class="p">{</span><span class="nb">input</span><span class="p">}</span> <span class="o">&gt;</span> <span class="err">$</span><span class="p">{</span><span class="nb">input</span><span class="p">}</span><span class="o">.</span><span class="n">flagstat</span>

<span class="c1">## # format of flagstat file. </span>
<span class="c1">## # </span>
<span class="c1">## #2181882 + 0 in total (QC-passed reads + QC-failed reads)</span>
<span class="c1">## #0 + 0 duplicates</span>
<span class="c1">## #2181882 + 0 mapped (100.00%:nan%)</span>
<span class="c1">## #2181882 + 0 paired in sequencing</span>
<span class="c1">## #mapped_reads = ${eval([line.split(&#39;m&#39;)[0] for line in open(input[0] + &#39;.flagstat&#39;).read().split(&#39;\n&#39;) if &#39;mapped&#39; in line][0])}</span>
<span class="c1">## </span>

<span class="p">[</span><span class="n">mapped_reads</span><span class="p">:</span> <span class="kp">shared</span><span class="o">=</span><span class="s1">&#39;mapped_reads&#39;</span><span class="p">]</span>
<span class="kn">input:</span>
    <span class="s1">&#39;${accepted_hits}.flagstat&#39;</span>

<span class="kn">depends:</span>
    <span class="n">sos_variable</span><span class="p">(</span><span class="s1">&#39;accepted_hits&#39;</span><span class="p">)</span>

<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="nb">input</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="k">as</span> <span class="n">flgs</span><span class="p">:</span>
    <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">flgs</span><span class="p">:</span>
        <span class="k">if</span> <span class="s1">&#39;mapped&#39;</span> <span class="ow">in</span> <span class="n">line</span><span class="p">:</span>
            <span class="n">mapped_reads</span> <span class="o">=</span> <span class="nb">eval</span><span class="p">(</span><span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;m&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">])</span>
            <span class="k">break</span>


<span class="p">[</span><span class="n">human_hg19_620</span> <span class="p">(</span><span class="n">Gene</span> <span class="n">Count</span> <span class="n">RPKM</span><span class="p">)]</span>

<span class="kn">input:</span>
     <span class="s1">&#39;${count_out}/${samplename}-gene_counts.tsv&#39;</span>

<span class="kn">depends:</span> 
    <span class="n">genesize_txt</span><span class="p">,</span>
    <span class="n">sos_variable</span><span class="p">(</span><span class="s1">&#39;mapped_reads&#39;</span><span class="p">)</span>

<span class="kn">output:</span>
    <span class="s2">&quot;${count_out}/${samplename}-gene_counts_rpkm.xls&quot;</span>

<span class="kn">python:</span>
    <span class="kn">import</span> <span class="nn">sys</span>
    <span class="n">gene_size</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">ambiguous_size</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">depends</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="err">!</span><span class="n">r</span><span class="p">})</span> <span class="k">as</span> <span class="n">genesize</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">genesize</span><span class="p">:</span>
            <span class="nb">chr</span><span class="p">,</span> <span class="n">gene</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">stop</span><span class="p">,</span> <span class="n">size</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">gene_size</span> <span class="ow">and</span> <span class="n">gene_size</span><span class="p">[</span><span class="n">gene</span><span class="p">][</span><span class="mi">3</span><span class="p">]</span> <span class="o">!=</span> <span class="nb">int</span><span class="p">(</span><span class="n">size</span><span class="p">):</span>
                <span class="n">ambiguous_size</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span>
            <span class="n">gene_size</span><span class="p">[</span><span class="n">gene</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="nb">chr</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">stop</span><span class="p">,</span> <span class="nb">int</span><span class="p">(</span><span class="n">size</span><span class="p">))</span>
            <span class="n">gene_size</span><span class="p">[</span><span class="n">gene</span> <span class="o">+</span> <span class="s1">&#39;.&#39;</span> <span class="o">+</span> <span class="nb">chr</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="nb">chr</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">stop</span><span class="p">,</span> <span class="nb">int</span><span class="p">(</span><span class="n">size</span><span class="p">))</span>
    <span class="c1">#</span>
    <span class="k">print</span><span class="p">(</span><span class="s1">&#39;Size of {} genes imported.&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">gene_size</span><span class="p">)</span><span class="o">/</span><span class="mi">2</span><span class="p">))</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="nb">input</span><span class="err">!</span><span class="n">r</span><span class="p">})</span> <span class="k">as</span> <span class="n">counts</span><span class="p">,</span> <span class="nb">open</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">output</span><span class="err">!</span><span class="n">r</span><span class="p">},</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">output</span><span class="p">:</span>
        <span class="n">output</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;Chr</span><span class="se">\t</span><span class="s1">GeneID</span><span class="se">\t</span><span class="s1">Start</span><span class="se">\t</span><span class="s1">Stop</span><span class="se">\t</span><span class="s1">CodingLength</span><span class="se">\t</span><span class="s1">ReadCount</span><span class="se">\t</span><span class="s1">RPKM</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">counts</span><span class="p">:</span>
            <span class="c1"># &#39;__no_feature&#39;, &#39;__ambiguous&#39;, &#39;__too_low_aQual&#39;, &#39;not_aligned&#39;, &#39;__alignment_not_unique&#39;</span>
            <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;__&#39;</span><span class="p">):</span>
                <span class="k">continue</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">gene</span><span class="p">,</span> <span class="n">value</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
            <span class="k">except</span><span class="p">:</span>
                <span class="n">sys</span><span class="o">.</span><span class="n">stderr</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;Unprocess line: {}&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">line</span><span class="p">))</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">gene</span><span class="o">.</span><span class="n">strip</span><span class="p">():</span>
                <span class="k">continue</span>
            <span class="k">elif</span> <span class="n">gene</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">gene_size</span><span class="p">:</span>
                <span class="n">sys</span><span class="o">.</span><span class="n">stderr</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;Unknown size of gene {}&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">gene</span><span class="p">))</span>
                <span class="n">output</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;NA</span><span class="se">\t</span><span class="s1">{}</span><span class="se">\t</span><span class="s1">NA</span><span class="se">\t</span><span class="s1">NA</span><span class="se">\t</span><span class="s1">NA</span><span class="se">\t</span><span class="s1">{}</span><span class="se">\t</span><span class="s1">NA</span><span class="se">\n</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">value</span><span class="o">.</span><span class="n">strip</span><span class="p">()))</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">ambiguous_size</span><span class="p">:</span>
                    <span class="n">sys</span><span class="o">.</span><span class="n">stderr</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;{} appear in different chromosomes. Its RPKM value is based on the size of one of them&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">gene</span><span class="p">))</span>
                <span class="n">output</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;{}</span><span class="se">\t</span><span class="s1">{}</span><span class="se">\t</span><span class="s1">{}</span><span class="se">\t</span><span class="s1">{}</span><span class="se">\t</span><span class="s1">{}</span><span class="se">\t</span><span class="s1">{}</span><span class="se">\t</span><span class="s1">{:.2f}</span><span class="se">\n</span><span class="s1">&#39;</span>
                    <span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">gene_size</span><span class="p">[</span><span class="n">gene</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span>
                        <span class="c1"># output gene name if the genename was appended with chromosome name</span>
                        <span class="n">gene</span><span class="p">[:</span><span class="o">-</span><span class="nb">len</span><span class="p">(</span><span class="n">gene_size</span><span class="p">[</span><span class="n">gene</span><span class="p">][</span><span class="mi">0</span><span class="p">])</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="k">if</span> <span class="n">gene</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s1">&#39;.&#39;</span> <span class="o">+</span> <span class="n">gene_size</span><span class="p">[</span><span class="n">gene</span><span class="p">][</span><span class="mi">0</span><span class="p">])</span> <span class="k">else</span> <span class="n">gene</span><span class="p">,</span>
                        <span class="n">gene_size</span><span class="p">[</span><span class="n">gene</span><span class="p">][</span><span class="mi">1</span><span class="p">],</span>
                        <span class="n">gene_size</span><span class="p">[</span><span class="n">gene</span><span class="p">][</span><span class="mi">2</span><span class="p">],</span> <span class="n">gene_size</span><span class="p">[</span><span class="n">gene</span><span class="p">][</span><span class="mi">3</span><span class="p">],</span> <span class="n">value</span><span class="o">.</span><span class="n">strip</span><span class="p">(),</span>
                        <span class="mf">1.e9</span><span class="o">*</span><span class="nb">int</span><span class="p">(</span><span class="n">value</span><span class="p">)</span><span class="o">/</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">mapped_reads</span><span class="p">}</span> <span class="o">*</span> <span class="n">gene_size</span><span class="p">[</span><span class="n">gene</span><span class="p">][</span><span class="mi">3</span><span class="p">])))</span>

<span class="p">[</span><span class="n">human_hg19_630</span><span class="p">]</span>
<span class="c1">#Count: Running bedtools interectBed to get exon count.</span>
<span class="c1">#Count: Calculate RPKM for exon counts.</span>
<span class="kn">input:</span>
    <span class="n">sorted_bam</span>

<span class="kn">output:</span>
    <span class="s1">&#39;${count_out}/${samplename}-exon_counts.tsv&#39;</span>

<span class="kn">run:</span>
    <span class="n">coverageBed</span> <span class="o">-</span><span class="n">abam</span> <span class="err">$</span><span class="p">{</span><span class="nb">input</span><span class="p">}</span> <span class="o">-</span><span class="n">b</span> <span class="err">$</span><span class="p">{</span><span class="n">genes_exon_bed</span><span class="p">}</span>  \
        <span class="o">&gt;</span> <span class="err">$</span><span class="p">{</span><span class="n">count_out</span><span class="p">}</span><span class="o">/</span><span class="err">$</span><span class="p">{</span><span class="n">samplename</span><span class="p">}</span><span class="o">-</span><span class="n">exon_counts</span><span class="o">.</span><span class="n">tsv</span>

<span class="p">[</span><span class="n">human_hg19_631</span><span class="p">]</span>

<span class="kn">output:</span>
    <span class="s2">&quot;${count_out}/${samplename}-exon_counts_rpkm.xls&quot;</span>

<span class="kn">depends:</span>
    <span class="n">sos_variable</span><span class="p">(</span><span class="s1">&#39;mapped_reads&#39;</span><span class="p">)</span>

<span class="kn">python:</span>
    <span class="kn">import</span> <span class="nn">sys</span>
    <span class="kn">from</span> <span class="nn">collections</span> <span class="kn">import</span> <span class="n">defaultdict</span>

    <span class="n">exon_counts</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="nb">input</span><span class="err">!</span><span class="n">r</span><span class="p">})</span> <span class="k">as</span> <span class="n">counts</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">counts</span><span class="p">:</span>
            <span class="n">fields</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">fields</span><span class="p">)</span> <span class="o">==</span> <span class="mi">16</span><span class="p">:</span>
                <span class="n">exon_counts</span><span class="p">[</span><span class="nb">tuple</span><span class="p">(</span><span class="n">fields</span><span class="p">[</span><span class="mi">12</span><span class="p">:</span><span class="mi">16</span><span class="p">])]</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="k">elif</span> <span class="nb">len</span><span class="p">(</span><span class="n">fields</span><span class="p">)</span> <span class="o">==</span> <span class="mi">8</span><span class="p">:</span>
                <span class="k">if</span> <span class="nb">int</span><span class="p">(</span><span class="n">fields</span><span class="p">[</span><span class="mi">4</span><span class="p">])</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="n">exon_counts</span><span class="p">[</span><span class="nb">tuple</span><span class="p">(</span><span class="n">fields</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">4</span><span class="p">])]</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">fields</span><span class="p">[</span><span class="mi">4</span><span class="p">])</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">sys</span><span class="o">.</span><span class="n">stderr</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;Incorrect input line: {}&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">line</span><span class="p">))</span>
                <span class="k">continue</span>
    <span class="n">sys</span><span class="o">.</span><span class="n">stderr</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;Writing counts for {} exons&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">exon_counts</span><span class="p">)))</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">output</span><span class="err">!</span><span class="n">r</span><span class="p">},</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">output</span><span class="p">:</span>
        <span class="n">output</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;Chr</span><span class="se">\t</span><span class="s1">Start</span><span class="se">\t</span><span class="s1">Stop</span><span class="se">\t</span><span class="s1">Exon</span><span class="se">\t</span><span class="s1">ReadCount</span><span class="se">\t</span><span class="s1">RPKM</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">exon</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">exon_counts</span><span class="o">.</span><span class="n">keys</span><span class="p">(),</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="mi">3</span><span class="p">]):</span>
            <span class="n">count</span> <span class="o">=</span> <span class="n">exon_counts</span><span class="p">[</span><span class="n">exon</span><span class="p">]</span>
            <span class="c1"># the input files are in BED format with 0-based start position</span>
            <span class="c1"># we need to add 1 to the start position for consistency</span>
            <span class="n">output</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;{}</span><span class="se">\t</span><span class="s1">{}</span><span class="se">\t</span><span class="s1">{}</span><span class="se">\t</span><span class="s1">{}</span><span class="se">\t</span><span class="s1">{}</span><span class="se">\t</span><span class="s1">{:.2f}</span><span class="se">\n</span><span class="s1">&#39;</span>
                <span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">exon</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="nb">int</span><span class="p">(</span><span class="n">exon</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="o">+</span> <span class="mi">1</span><span class="p">,</span> <span class="n">exon</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span> <span class="n">exon</span><span class="p">[</span><span class="mi">3</span><span class="p">],</span> <span class="n">count</span><span class="p">,</span>
                        <span class="mf">1.e9</span><span class="o">*</span><span class="n">count</span><span class="o">/</span><span class="p">(</span><span class="err">$</span><span class="p">{</span><span class="n">mapped_reads</span><span class="p">}</span> <span class="o">*</span> <span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">exon</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span> <span class="o">-</span> <span class="nb">int</span><span class="p">(</span><span class="n">exon</span><span class="p">[</span><span class="mi">1</span><span class="p">])))))</span>



<span class="p">[</span><span class="n">update_blast</span><span class="p">:</span> <span class="kp">provides</span><span class="o">=</span><span class="s1">&#39;${ncbi_resource_dir}/blast/db/human_genomic.nal&#39;</span><span class="p">]</span>
<span class="c1">#Resource: Download NCBI Blast databases and link the directory</span>
<span class="c1"># to output dir for blast to work</span>
<span class="kn">depends:</span>
    <span class="n">executable</span><span class="p">(</span><span class="s1">&#39;update_blastdb.pl&#39;</span><span class="p">)</span>

<span class="kn">output:</span>
    <span class="s1">&#39;${ncbi_resource_dir}/blast/db/human_genomic.nal&#39;</span><span class="p">,</span>
    <span class="s1">&#39;${ncbi_resource_dir}/blast/db/nt.nal&#39;</span>

<span class="kn">task:</span>

<span class="kn">sh:</span> <span class="kp">workdir</span><span class="o">=</span><span class="s1">&#39;${ncbi_resource_dir}/blast/db&#39;</span>
    <span class="n">update_blastdb</span><span class="o">.</span><span class="n">pl</span> <span class="n">human_genomic</span> <span class="o">--</span><span class="n">decompress</span> <span class="o">||</span> <span class="n">true</span>
    <span class="n">update_blastdb</span><span class="o">.</span><span class="n">pl</span> <span class="n">nt</span> <span class="o">--</span><span class="n">decompress</span> <span class="o">||</span> <span class="n">true</span>

<span class="kn">sh:</span> <span class="kp">workdir</span><span class="o">=</span><span class="n">output_dir</span>
    <span class="p">[</span> <span class="o">-</span><span class="n">d</span> <span class="n">blast</span> <span class="p">]</span> <span class="o">||</span> <span class="n">ln</span> <span class="o">-</span><span class="n">s</span> <span class="err">$</span><span class="p">{</span><span class="n">ncbi_resource_dir</span><span class="p">}</span><span class="o">/</span><span class="n">blast</span><span class="o">/</span><span class="n">db</span> <span class="n">blast</span>


<span class="p">[</span><span class="n">before_tophat_fusion</span><span class="p">:</span> <span class="kp">provides</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;${output_dir}/refGene.txt&#39;</span><span class="p">,</span> <span class="s1">&#39;${output_dir}/ensGene.txt&#39;</span><span class="p">]]</span>
<span class="c1">#Fusion (prepare): Create an alias for tophat output directory because</span>
<span class="c1">#    tophat-fusion-post require the directory to have name tophat-XXX</span>
<span class="c1">#Fusion (prepare): Link refGene.txt and ensGene.txt to output directory to be used by tophat-fusion-post.</span>

<span class="kn">depends:</span>
    <span class="n">refgene_txt</span><span class="p">,</span> <span class="n">ensgene_txt</span>

<span class="kn">run:</span>    <span class="kp">workdir</span><span class="o">=</span><span class="n">output_dir</span>
    <span class="p">[</span> <span class="o">-</span><span class="n">f</span> <span class="n">refGene</span><span class="o">.</span><span class="n">txt</span> <span class="p">]</span> <span class="o">||</span> <span class="n">ln</span> <span class="o">-</span><span class="n">s</span> <span class="err">$</span><span class="p">{</span><span class="n">refgene_txt</span><span class="p">}</span> <span class="n">refGene</span><span class="o">.</span><span class="n">txt</span>
    <span class="p">[</span> <span class="o">-</span><span class="n">f</span> <span class="n">ensGene</span><span class="o">.</span><span class="n">txt</span> <span class="p">]</span> <span class="o">||</span> <span class="n">ln</span> <span class="o">-</span><span class="n">s</span> <span class="err">$</span><span class="p">{</span><span class="n">ensgene_txt</span><span class="p">}</span> <span class="n">ensGene</span><span class="o">.</span><span class="n">txt</span>

<span class="kp">stop_if</span><span class="p">(</span><span class="n">alignment_out</span><span class="o">.</span><span class="n">rstrip</span><span class="p">(</span><span class="s1">&#39;/&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">rsplit</span><span class="p">(</span><span class="s1">&#39;/&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;tophat_&#39;</span><span class="p">)</span> \
    <span class="ow">or</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isdir</span><span class="p">(</span><span class="n">output_dir</span> <span class="o">+</span> <span class="s1">&#39;/tophat_&#39;</span> <span class="o">+</span> <span class="n">alignment_out</span><span class="o">.</span><span class="n">rsplit</span><span class="p">(</span><span class="s2">&quot;/&quot;</span><span class="p">,</span><span class="mi">1</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]))</span>

<span class="kn">run:</span>    <span class="kp">workdir</span><span class="o">=</span><span class="n">output_dir</span>
    <span class="n">ln</span> <span class="o">-</span><span class="n">s</span> <span class="err">$</span><span class="p">{</span><span class="n">alignment_out</span><span class="o">.</span><span class="n">rstrip</span><span class="p">(</span><span class="s2">&quot;/&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">rsplit</span><span class="p">(</span><span class="s2">&quot;/&quot;</span><span class="p">,</span><span class="mi">1</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]}</span> <span class="err">$</span><span class="p">{</span><span class="s1">&#39;tophat_&#39;</span> <span class="o">+</span> <span class="n">alignment_out</span><span class="o">.</span><span class="n">rsplit</span><span class="p">(</span><span class="s2">&quot;/&quot;</span><span class="p">,</span><span class="mi">1</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]}</span>


<span class="p">[</span><span class="n">human_hg19_710</span><span class="p">]</span>
<span class="c1"># tophat-fusion for fusion detection</span>
<span class="kn">output:</span>
    <span class="s2">&quot;${fusion_out}/potential_fusion.txt&quot;</span>

<span class="kn">depends:</span>
    <span class="s1">&#39;${output_dir}/refGene.txt&#39;</span><span class="p">,</span>
    <span class="s1">&#39;${output_dir}/ensGene.txt&#39;</span><span class="p">,</span>
    <span class="s1">&#39;${ncbi_resource_dir}/blast/db/human_genomic.nal&#39;</span>

<span class="kn">run:</span> <span class="kp">workdir</span><span class="o">=</span><span class="n">output_dir</span>

    <span class="n">tophat</span><span class="o">-</span><span class="n">fusion</span><span class="o">-</span><span class="n">post</span> <span class="o">-</span><span class="n">p</span> <span class="mi">3</span>  <span class="o">-</span><span class="n">o</span> <span class="err">$</span><span class="p">{</span><span class="n">fusion_out</span><span class="o">.</span><span class="n">rstrip</span><span class="p">(</span><span class="s2">&quot;/&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">rsplit</span><span class="p">(</span><span class="s2">&quot;/&quot;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]}</span> \
        <span class="err">$</span><span class="p">{</span><span class="n">reference_dir</span><span class="p">}</span><span class="o">/</span><span class="n">BowtieIndex</span><span class="o">/</span><span class="n">genome</span>


<span class="p">[</span><span class="n">human_hg19_730</span><span class="p">]</span>
<span class="c1"># (Oncofuse): filter fusion candidates using Oncofuse</span>

<span class="kn">output:</span>
    <span class="s1">&#39;${fusion_out}/fusions_oncofuse.xls&#39;</span>

<span class="kn">run:</span>
    <span class="n">java</span> <span class="o">-</span><span class="n">Xmx1G</span> <span class="o">-</span><span class="n">jar</span> <span class="err">$</span><span class="p">{</span><span class="n">oncofuse_path</span><span class="p">}</span><span class="o">/</span><span class="n">Oncofuse</span><span class="o">.</span><span class="n">jar</span> \
        <span class="err">$</span><span class="p">{</span><span class="nb">input</span><span class="p">}</span> <span class="n">tophat</span><span class="o">-</span><span class="n">post</span> <span class="n">AVG</span> <span class="err">$</span><span class="p">{</span><span class="n">fusion_out</span><span class="p">}</span><span class="o">/</span><span class="n">fusions_oncofuse</span><span class="o">.</span><span class="n">xls</span>

<span class="p">[</span><span class="n">download_gatk</span><span class="p">:</span> <span class="kp">provides</span><span class="o">=</span><span class="n">gatk_resource_dir</span> <span class="o">+</span> <span class="s1">&#39;/dbsnp_138.hg19.vcf&#39;</span><span class="p">]</span>
<span class="kn">download:</span> <span class="n">dest_dir</span><span class="o">=</span><span class="n">gatk_resource_dir</span>
    <span class="err">$</span><span class="p">{</span><span class="n">gatk_url</span><span class="p">}</span><span class="o">/</span><span class="mi">1000</span><span class="n">G_omni2</span><span class="o">.</span><span class="mf">5.</span><span class="n">hg19</span><span class="o">.</span><span class="n">sites</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span>
    <span class="err">$</span><span class="p">{</span><span class="n">gatk_url</span><span class="p">}</span><span class="o">/</span><span class="mi">1000</span><span class="n">G_omni2</span><span class="o">.</span><span class="mf">5.</span><span class="n">hg19</span><span class="o">.</span><span class="n">sites</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span><span class="o">.</span><span class="n">md5</span>
    <span class="err">$</span><span class="p">{</span><span class="n">gatk_url</span><span class="p">}</span><span class="o">/</span><span class="mi">1000</span><span class="n">G_omni2</span><span class="o">.</span><span class="mf">5.</span><span class="n">hg19</span><span class="o">.</span><span class="n">sites</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">idx</span><span class="o">.</span><span class="n">gz</span>
    <span class="err">$</span><span class="p">{</span><span class="n">gatk_url</span><span class="p">}</span><span class="o">/</span><span class="mi">1000</span><span class="n">G_omni2</span><span class="o">.</span><span class="mf">5.</span><span class="n">hg19</span><span class="o">.</span><span class="n">sites</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">idx</span><span class="o">.</span><span class="n">gz</span><span class="o">.</span><span class="n">md5</span>
    <span class="err">$</span><span class="p">{</span><span class="n">gatk_url</span><span class="p">}</span><span class="o">/</span><span class="n">dbsnp_138</span><span class="o">.</span><span class="n">hg19</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span>
    <span class="err">$</span><span class="p">{</span><span class="n">gatk_url</span><span class="p">}</span><span class="o">/</span><span class="n">dbsnp_138</span><span class="o">.</span><span class="n">hg19</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span><span class="o">.</span><span class="n">md5</span>
    <span class="err">$</span><span class="p">{</span><span class="n">gatk_url</span><span class="p">}</span><span class="o">/</span><span class="n">dbsnp_138</span><span class="o">.</span><span class="n">hg19</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">idx</span><span class="o">.</span><span class="n">gz</span>
    <span class="err">$</span><span class="p">{</span><span class="n">gatk_url</span><span class="p">}</span><span class="o">/</span><span class="n">dbsnp_138</span><span class="o">.</span><span class="n">hg19</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">idx</span><span class="o">.</span><span class="n">gz</span><span class="o">.</span><span class="n">md5</span>
    <span class="err">$</span><span class="p">{</span><span class="n">gatk_url</span><span class="p">}</span><span class="o">/</span><span class="n">hapmap_3</span><span class="o">.</span><span class="mf">3.</span><span class="n">hg19</span><span class="o">.</span><span class="n">sites</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span>
    <span class="err">$</span><span class="p">{</span><span class="n">gatk_url</span><span class="p">}</span><span class="o">/</span><span class="n">hapmap_3</span><span class="o">.</span><span class="mf">3.</span><span class="n">hg19</span><span class="o">.</span><span class="n">sites</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">gz</span><span class="o">.</span><span class="n">md5</span>
    <span class="err">$</span><span class="p">{</span><span class="n">gatk_url</span><span class="p">}</span><span class="o">/</span><span class="n">hapmap_3</span><span class="o">.</span><span class="mf">3.</span><span class="n">hg19</span><span class="o">.</span><span class="n">sites</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">idx</span><span class="o">.</span><span class="n">gz</span>
    <span class="err">$</span><span class="p">{</span><span class="n">gatk_url</span><span class="p">}</span><span class="o">/</span><span class="n">hapmap_3</span><span class="o">.</span><span class="mf">3.</span><span class="n">hg19</span><span class="o">.</span><span class="n">sites</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">idx</span><span class="o">.</span><span class="n">gz</span><span class="o">.</span><span class="n">md5</span>

<span class="p">[</span><span class="n">human_hg19_800</span> <span class="p">(</span><span class="n">GATK</span> <span class="n">UnifiedGenotyper</span><span class="p">)]</span>
<span class="c1"># Use GATK UnifiedGenotyper caller to call variants</span>
<span class="kn">input:</span>
    <span class="n">sorted_unique</span>

<span class="kn">depends:</span>
    <span class="s1">&#39;${input}.bai&#39;</span><span class="p">,</span>
    <span class="n">sos_variable</span><span class="p">(</span><span class="s1">&#39;mapped_reads&#39;</span><span class="p">)</span>

<span class="c1"># An option -L can be used to limit the variant call to targeted regions</span>
<span class="c1"># -L $ref_dir/hg19.master.gene.bed </span>

<span class="kn">output:</span>
    <span class="s1">&#39;${variant_out}/${samplename}-gatk.vcf&#39;</span>

<span class="kn">run:</span>
    <span class="n">java</span> <span class="o">-</span><span class="n">Xmx16g</span> <span class="o">-</span><span class="n">Xms512m</span> <span class="o">-</span><span class="n">Djava</span><span class="o">.</span><span class="n">io</span><span class="o">.</span><span class="n">tmpdir</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="n">temp_dir</span><span class="p">}</span>  \
        <span class="o">-</span><span class="n">jar</span> <span class="err">$</span><span class="p">{</span><span class="n">gatk_path</span><span class="p">}</span><span class="o">/</span><span class="n">GenomeAnalysisTK</span><span class="o">.</span><span class="n">jar</span>   \
        <span class="o">-</span><span class="n">R</span> <span class="err">$</span><span class="p">{</span><span class="n">reference_dir</span><span class="p">}</span><span class="o">/</span><span class="n">WholeGenomeFasta</span><span class="o">/</span><span class="n">genome</span><span class="o">.</span><span class="n">fa</span>   \
        <span class="o">--</span><span class="n">filter_reads_with_N_cigar</span>   \
        <span class="o">-</span><span class="n">T</span> <span class="n">UnifiedGenotyper</span>   \
        <span class="o">-</span><span class="n">I</span> <span class="err">$</span><span class="p">{</span><span class="nb">input</span><span class="p">}</span>  \
        <span class="o">--</span><span class="n">out</span> <span class="err">$</span><span class="p">{</span><span class="n">variant_out</span><span class="p">}</span><span class="o">/</span><span class="err">$</span><span class="p">{</span><span class="n">samplename</span><span class="p">}</span><span class="o">-</span><span class="n">gatk</span><span class="o">.</span><span class="n">vcf</span>  

<span class="p">[</span><span class="n">human_hg19_810</span> <span class="p">(</span><span class="n">Recalibrate</span> <span class="n">Variants</span><span class="p">)]</span>
<span class="c1"># recalibrate and filter variants</span>
<span class="kn">output:</span>
    <span class="s1">&#39;${variant_out}/${samplename}-filter.vcf&#39;</span>

<span class="kn">depends:</span>
    <span class="s1">&#39;${reference_dir}/WholeGenomeFasta/genome.fa&#39;</span><span class="p">,</span>
    <span class="s1">&#39;${gatk_resource_dir}/hapmap_3.3.hg19.sites.vcf&#39;</span><span class="p">,</span>
    <span class="s1">&#39;${gatk_resource_dir}/1000G_omni2.5.hg19.sites.vcf&#39;</span><span class="p">,</span>
    <span class="s1">&#39;${gatk_resource_dir}/dbsnp_138.hg19.vcf&#39;</span>

<span class="kn">run:</span>
    <span class="n">java</span> <span class="o">-</span><span class="n">Xmx16g</span> <span class="o">-</span><span class="n">Xms512m</span>  \
        <span class="o">-</span><span class="n">Djava</span><span class="o">.</span><span class="n">io</span><span class="o">.</span><span class="n">tmpdir</span><span class="o">=</span><span class="err">$</span><span class="p">{</span><span class="n">temp_dir</span><span class="p">}</span>  \
        <span class="o">-</span><span class="n">jar</span> <span class="err">$</span><span class="p">{</span><span class="n">gatk_path</span><span class="p">}</span><span class="o">/</span><span class="n">GenomeAnalysisTK</span><span class="o">.</span><span class="n">jar</span>   \
        <span class="o">-</span><span class="n">R</span> <span class="err">$</span><span class="p">{</span><span class="n">reference_dir</span><span class="p">}</span><span class="o">/</span><span class="n">WholeGenomeFasta</span><span class="o">/</span><span class="n">genome</span><span class="o">.</span><span class="n">fa</span>   \
        <span class="o">-</span><span class="n">T</span> <span class="n">VariantRecalibrator</span>  \
        <span class="o">-</span><span class="n">mode</span> <span class="n">SNP</span>   \
        <span class="o">-</span><span class="nb">input</span> <span class="err">$</span><span class="p">{</span><span class="nb">input</span><span class="p">}</span>  \
        <span class="o">-</span><span class="n">resource</span><span class="p">:</span><span class="n">hapmap</span><span class="p">,</span><span class="n">known</span><span class="o">=</span><span class="n">false</span><span class="p">,</span><span class="n">training</span><span class="o">=</span><span class="n">true</span><span class="p">,</span><span class="n">truth</span><span class="o">=</span><span class="n">true</span><span class="p">,</span><span class="n">prior</span><span class="o">=</span><span class="mf">15.0</span> <span class="err">$</span><span class="p">{</span><span class="n">gatk_resource_dir</span><span class="p">}</span><span class="o">/</span><span class="n">hapmap_3</span><span class="o">.</span><span class="mf">3.</span><span class="n">hg19</span><span class="o">.</span><span class="n">sites</span><span class="o">.</span><span class="n">vcf</span>  \
        <span class="o">-</span><span class="n">resource</span><span class="p">:</span><span class="n">omni</span><span class="p">,</span><span class="n">known</span><span class="o">=</span><span class="n">false</span><span class="p">,</span><span class="n">training</span><span class="o">=</span><span class="n">true</span><span class="p">,</span><span class="n">truth</span><span class="o">=</span><span class="n">false</span><span class="p">,</span><span class="n">prior</span><span class="o">=</span><span class="mf">12.0</span>  <span class="err">$</span><span class="p">{</span><span class="n">gatk_resource_dir</span><span class="p">}</span><span class="o">/</span><span class="mi">1000</span><span class="n">G_omni2</span><span class="o">.</span><span class="mf">5.</span><span class="n">hg19</span><span class="o">.</span><span class="n">sites</span><span class="o">.</span><span class="n">vcf</span>  \
        <span class="o">-</span><span class="n">resource</span><span class="p">:</span><span class="n">dbsnp</span><span class="p">,</span><span class="n">known</span><span class="o">=</span><span class="n">true</span><span class="p">,</span><span class="n">training</span><span class="o">=</span><span class="n">false</span><span class="p">,</span><span class="n">truth</span><span class="o">=</span><span class="n">false</span><span class="p">,</span><span class="n">prior</span><span class="o">=</span><span class="mf">2.0</span>  <span class="err">$</span><span class="p">{</span><span class="n">gatk_resource_dir</span><span class="p">}</span><span class="o">/</span><span class="n">dbsnp_138</span><span class="o">.</span><span class="n">hg19</span><span class="o">.</span><span class="n">vcf</span>  \
        <span class="o">-</span><span class="n">an</span> <span class="n">ReadPosRankSum</span> <span class="o">-</span><span class="n">an</span> <span class="n">FS</span>   \
        <span class="o">-</span><span class="n">recalFile</span> <span class="err">$</span><span class="p">{</span><span class="n">variant_out</span><span class="p">}</span><span class="o">/</span><span class="err">$</span><span class="p">{</span><span class="n">samplename</span><span class="p">}</span><span class="o">.</span><span class="n">recal</span>   \
        <span class="o">-</span><span class="n">tranchesFile</span> <span class="err">$</span><span class="p">{</span><span class="n">variant_out</span><span class="p">}</span><span class="o">/</span><span class="err">$</span><span class="p">{</span><span class="n">samplename</span><span class="p">}</span><span class="o">.</span><span class="n">tranches</span>   \
        <span class="o">--</span><span class="n">maxGaussians</span> <span class="err">$</span><span class="p">{</span><span class="mi">4</span> <span class="k">if</span> <span class="n">mapped_reads</span> <span class="o">&gt;</span> <span class="mi">4000000</span> <span class="k">else</span> <span class="mi">2</span><span class="p">}</span> <span class="o">||</span> <span class="n">true</span>  \

<span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">getsize</span><span class="p">(</span><span class="s1">&#39;${variant_out}/${samplename}.recal&#39;</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
    <span class="kp">run</span><span class="p">(</span><span class="s1">r&#39;&#39;&#39;</span>
<span class="s1">        java -Xmx16g -Xms512m   \</span>
<span class="s1">            -Djava.io.tmpdir=${temp_dir}   \</span>
<span class="s1">            -jar ${gatk_path}/GenomeAnalysisTK.jar   \</span>
<span class="s1">            -R ${reference_dir}/WholeGenomeFasta/genome.fa  \</span>
<span class="s1">            -T ApplyRecalibration   \</span>
<span class="s1">            -mode SNP   \</span>
<span class="s1">            -input ${input}   \</span>
<span class="s1">            -recalFile ${variant_out}/${samplename}.recal  \</span>
<span class="s1">            -tranchesFile ${variant_out}/${samplename}.tranches  \</span>
<span class="s1">            -o ${variant_out}/${samplename}-filter.vcf   \</span>
<span class="s1">            --ts_filter_level 99.0</span>
<span class="s1">    &#39;&#39;&#39;</span><span class="p">)</span>
<span class="kn">else:</span>
    <span class="kp">run</span><span class="p">(</span><span class="s1">r&#39;&#39;&#39;</span>
<span class="s1">        java -Xmx16g -Xms512m  \</span>
<span class="s1">            -Djava.io.tmpdir=${temp_dir}  \</span>
<span class="s1">            -jar ${gatk_path}/GenomeAnalysisTK.jar  \</span>
<span class="s1">            -R ${reference_dir}/WholeGenomeFasta/genome.fa  \</span>
<span class="s1">            -V ${input}  \</span>
<span class="s1">            -l INFO -T VariantFiltration  \</span>
<span class="s1">            --filterExpression &quot;FS &gt; 20.0&quot; --filterName FSFilter  \</span>
<span class="s1">            --filterExpression &quot;ED &gt; 5&quot; --filterName EDFilter  \</span>
<span class="s1">            --filterExpression &quot;ReadPosRankSum &lt; -8.0&quot; --filterName RPRSFilter  \</span>
<span class="s1">            --filterExpression &quot;ReadPosRankSum &gt; -8.0&quot; --filterName RPRSFilter  \</span>
<span class="s1">            -o ${variant_out}/${samplename}-filter.vcf</span>
<span class="s1">    &#39;&#39;&#39;</span><span class="p">)</span>

<span class="p">[</span><span class="n">human_hg19_900</span><span class="p">]</span>
<span class="c1">#RSeQC: Generate JPEG versions of the RSeQC plots to be embedded in final report </span>
<span class="kn">input:</span>
    <span class="n">rseqc_pdf</span>

<span class="kn">output:</span>
    <span class="s1">&#39;${alignmentqc_out}/${samplename}.geneBodyCoverage.curves.jpg&#39;</span><span class="p">,</span>
    <span class="s1">&#39;${alignmentqc_out}/${samplename}.junctionSaturation_plot.jpg&#39;</span><span class="p">,</span>
    <span class="s1">&#39;${alignmentqc_out}/${samplename}.splice_events.jpg&#39;</span><span class="p">,</span>
    <span class="s1">&#39;${alignmentqc_out}/${samplename}.splice_junction.jpg&#39;</span><span class="p">,</span>
    <span class="s1">&#39;${alignmentqc_out}/${samplename}.inner_distance_plot.jpg&#39;</span><span class="p">,</span>
    <span class="s1">&#39;${alignmentqc_out}//${samplename}.DupRate_plot.jpg&#39;</span>

<span class="c1"># we need to wait the completion of geneBody_coverage</span>
<span class="kn">run:</span>
    <span class="n">cat</span> <span class="err">$</span><span class="p">{</span><span class="s1">&#39; &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">alignmentqc_out</span><span class="p">,</span> <span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="n">alignmentqc_out</span><span class="p">)</span> <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s1">&#39;.r&#39;</span><span class="p">)])}</span>  \
    <span class="o">|</span> <span class="n">sed</span> <span class="s1">&#39;s/pdf(\(.*\).pdf\(.\)/jpeg(</span><span class="se">\1</span><span class="s1">.jpg</span><span class="se">\2</span><span class="s1">, width=600, height=600/&#39;</span> <span class="o">|</span> <span class="n">R</span> <span class="o">--</span><span class="n">no</span><span class="o">-</span><span class="n">save</span>


<span class="p">[</span><span class="n">human_hg19_120</span><span class="p">:</span> <span class="kp">shared</span><span class="o">=</span><span class="s1">&#39;read_length&#39;</span><span class="p">]</span>

<span class="kn">input:</span>
   <span class="n">fastq_files</span>

<span class="kn">import</span> <span class="nn">gzip</span>
<span class="k">with</span> <span class="n">gzip</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="nb">input</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="s1">&#39;rb&#39;</span><span class="p">)</span> <span class="k">if</span> <span class="nb">input</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s1">&#39;.gz&#39;</span><span class="p">)</span> <span class="k">else</span> <span class="nb">open</span><span class="p">(</span><span class="nb">input</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="s1">&#39;rb&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fastq</span><span class="p">:</span>
    <span class="c1"># skip the first line</span>
    <span class="n">count</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">total_length</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">for</span> <span class="n">index</span><span class="p">,</span><span class="n">line</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">fastq</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">index</span> <span class="o">%</span> <span class="mi">4</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
            <span class="c1"># get the second line</span>
            <span class="n">total_length</span> <span class="o">+=</span> <span class="nb">len</span><span class="p">(</span><span class="n">line</span><span class="o">.</span><span class="n">decode</span><span class="p">()</span><span class="o">.</span><span class="n">strip</span><span class="p">())</span>
            <span class="n">count</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="k">if</span> <span class="n">count</span> <span class="o">==</span> <span class="mi">100</span><span class="p">:</span>
            <span class="k">break</span>
    <span class="n">read_length</span> <span class="o">=</span> <span class="n">total_length</span> <span class="o">/</span> <span class="n">count</span>



<span class="p">[</span><span class="n">alignment_summary</span><span class="p">:</span> <span class="kp">shared</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;total_reads&#39;</span><span class="p">,</span> <span class="s1">&#39;used_reads&#39;</span><span class="p">,</span> <span class="s1">&#39;mapped_reads_perc&#39;</span><span class="p">,</span> <span class="s1">&#39;uniq_mapped_reads&#39;</span><span class="p">,</span> <span class="s1">&#39;uniq_mapped_reads_perc&#39;</span><span class="p">,</span> <span class="s1">&#39;mapped_pairs&#39;</span><span class="p">,</span> <span class="s1">&#39;mapped_pairs_perc&#39;</span><span class="p">,</span> <span class="s1">&#39;concordant_mapped_pairs&#39;</span><span class="p">,</span> <span class="s1">&#39;concordant_mapped_pairs_perc&#39;</span><span class="p">,</span> <span class="s1">&#39;genome_mapped_count&#39;</span><span class="p">,</span> <span class="s1">&#39;genome_mapped_perc&#39;</span><span class="p">,</span> <span class="s1">&#39;junction_mapped_count&#39;</span><span class="p">,</span> <span class="s1">&#39;junction_mapped_perc&#39;</span><span class="p">)]</span>

<span class="kn">depends:</span>
    <span class="n">sos_variable</span><span class="p">(</span><span class="s1">&#39;accepted_hits&#39;</span><span class="p">)</span>

<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;${alignment_out}/prep_reads.info&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">prep_reads</span><span class="p">:</span>
    <span class="n">total_reads</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">used_reads</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">prep_reads</span><span class="p">:</span>
        <span class="k">if</span> <span class="s1">&#39;reads_in&#39;</span> <span class="ow">in</span> <span class="n">line</span><span class="p">:</span>
            <span class="n">total_reads</span> <span class="o">+=</span> <span class="nb">int</span><span class="p">(</span><span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;=&#39;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
        <span class="k">if</span> <span class="s1">&#39;reads_out&#39;</span> <span class="ow">in</span> <span class="n">line</span><span class="p">:</span>
            <span class="n">used_reads</span> <span class="o">+=</span> <span class="nb">int</span><span class="p">(</span><span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;=&#39;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
<span class="c1">#</span>
<span class="c1"># get mapped reads from align_summary.txt</span>
<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;${alignment_out}/align_summary.txt&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">align_summary</span><span class="p">:</span>
    <span class="n">uniq_mapped_reads</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">uniq_mapped_pairs</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">mapped_reads</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">mapped_pairs</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">concordant_mapped_pairs</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">line</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">align_summary</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">idx</span> <span class="ow">in</span> <span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">6</span><span class="p">):</span>
            <span class="c1">#  Mapped   :    770754 (77.1% of input)</span>
            <span class="n">mapped_reads</span> <span class="o">+=</span> <span class="nb">int</span><span class="p">(</span><span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;:&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;(&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">strip</span><span class="p">())</span>
        <span class="k">if</span> <span class="n">idx</span> <span class="o">==</span> <span class="mi">10</span><span class="p">:</span>
            <span class="c1"># Aligned pairs:</span>
            <span class="n">mapped_pairs</span> <span class="o">+=</span> <span class="nb">int</span><span class="p">(</span><span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;:&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">strip</span><span class="p">())</span>
        <span class="k">if</span> <span class="n">idx</span> <span class="ow">in</span> <span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">7</span><span class="p">):</span>
            <span class="n">uniq_mapped_reads</span> <span class="o">+=</span> <span class="nb">int</span><span class="p">(</span><span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;:&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;(&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">strip</span><span class="p">())</span>
        <span class="k">if</span> <span class="n">idx</span> <span class="o">==</span> <span class="mi">12</span><span class="p">:</span>
            <span class="n">concordant_mapped_pairs</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">()[</span><span class="mi">0</span><span class="p">])</span>
    <span class="n">uniq_mapped_reads</span> <span class="o">=</span> <span class="n">mapped_reads</span> <span class="o">-</span> <span class="n">uniq_mapped_reads</span>
    <span class="n">uniq_mapped_reads_perc</span> <span class="o">=</span> <span class="n">uniq_mapped_reads</span> <span class="o">*</span> <span class="mf">100.</span> <span class="o">/</span> <span class="n">total_reads</span>
    <span class="n">mapped_reads_perc</span> <span class="o">=</span> <span class="n">mapped_reads</span> <span class="o">*</span> <span class="mf">100.</span> <span class="o">/</span> <span class="n">total_reads</span>
    <span class="n">concordant_mapped_pairs</span> <span class="o">=</span> <span class="n">mapped_pairs</span> <span class="o">-</span> <span class="n">concordant_mapped_pairs</span>
    <span class="n">concordant_mapped_pairs_perc</span> <span class="o">=</span> <span class="n">concordant_mapped_pairs</span> <span class="o">*</span> <span class="mf">200.0</span> <span class="o">/</span> <span class="n">total_reads</span>
    <span class="n">mapped_pairs_perc</span> <span class="o">=</span> <span class="n">mapped_pairs</span> <span class="o">*</span> <span class="mf">200.0</span> <span class="o">/</span> <span class="n">total_reads</span>

<span class="c1"># get junction reads</span>
<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;${alignment_out}/junction.count&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">junction_count</span><span class="p">:</span>
    <span class="n">junction_mapped_count</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">junction_count</span><span class="o">.</span><span class="n">readline</span><span class="p">()</span><span class="o">.</span><span class="n">strip</span><span class="p">())</span>
    <span class="n">genome_mapped_count</span> <span class="o">=</span> <span class="n">mapped_reads</span> <span class="o">-</span> <span class="n">junction_mapped_count</span>
    <span class="n">junction_mapped_perc</span> <span class="o">=</span> <span class="n">junction_mapped_count</span> <span class="o">*</span> <span class="mf">100.</span> <span class="o">/</span> <span class="n">total_reads</span>
    <span class="n">genome_mapped_perc</span> <span class="o">=</span> <span class="n">genome_mapped_count</span> <span class="o">*</span> <span class="mf">100.</span> <span class="o">/</span> <span class="n">total_reads</span>

<span class="n">uniq_mapped_reads</span> <span class="o">=</span> <span class="n">mapped_reads</span> <span class="o">-</span> <span class="n">uniq_mapped_reads</span>
<span class="n">uniq_mapped_reads_perc</span> <span class="o">=</span> <span class="n">uniq_mapped_reads</span> <span class="o">*</span> <span class="mf">100.</span> <span class="o">/</span> <span class="n">total_reads</span>
<span class="n">mapped_reads_perc</span> <span class="o">=</span> <span class="n">mapped_reads</span> <span class="o">*</span> <span class="mf">100.</span> <span class="o">/</span> <span class="n">total_reads</span>
<span class="n">concordant_mapped_pairs</span> <span class="o">=</span> <span class="n">mapped_pairs</span> <span class="o">-</span> <span class="n">concordant_mapped_pairs</span>
<span class="n">concordant_mapped_pairs_perc</span> <span class="o">=</span> <span class="n">concordant_mapped_pairs</span> <span class="o">*</span> <span class="mf">200.0</span> <span class="o">/</span> <span class="n">total_reads</span>
<span class="n">mapped_pairs_perc</span> <span class="o">=</span> <span class="n">mapped_pairs</span> <span class="o">*</span> <span class="mf">200.0</span> <span class="o">/</span> <span class="n">total_reads</span>
<span class="n">genome_mapped_count</span> <span class="o">=</span> <span class="n">mapped_reads</span> <span class="o">-</span> <span class="n">junction_mapped_count</span>
<span class="n">junction_mapped_perc</span> <span class="o">=</span> <span class="n">junction_mapped_count</span> <span class="o">*</span> <span class="mf">100.</span> <span class="o">/</span> <span class="n">total_reads</span>
<span class="n">genome_mapped_perc</span> <span class="o">=</span> <span class="n">genome_mapped_count</span> <span class="o">*</span> <span class="mf">100.</span> <span class="o">/</span> <span class="n">total_reads</span>


<span class="p">[</span><span class="n">mean_depth</span><span class="p">:</span> <span class="kp">shared</span><span class="o">=</span><span class="s1">&#39;mean_depth&#39;</span><span class="p">]</span>

<span class="kn">input:</span>
    <span class="s1">&#39;${count_out}/${samplename}-exon_counts_rpkm.xls&#39;</span>

<span class="kn">depends:</span>
    <span class="n">sos_variable</span><span class="p">(</span><span class="s1">&#39;read_length&#39;</span><span class="p">)</span>

<span class="n">mean_depth</span> <span class="o">=</span> <span class="mi">0</span>
<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="nb">input</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="k">as</span> <span class="n">exons</span><span class="p">:</span>
    <span class="n">exons</span><span class="o">.</span><span class="n">readline</span><span class="p">()</span>
    <span class="n">total_reads</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">total_length</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">exons</span><span class="p">:</span>
        <span class="nb">chr</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="n">count</span><span class="p">,</span> <span class="n">rpkm</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
        <span class="k">if</span> <span class="nb">int</span><span class="p">(</span><span class="n">count</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">2</span><span class="p">:</span>
            <span class="k">continue</span>
        <span class="n">total_reads</span> <span class="o">+=</span> <span class="nb">int</span><span class="p">(</span><span class="n">count</span><span class="p">)</span>
        <span class="n">total_length</span> <span class="o">+=</span> <span class="nb">int</span><span class="p">(</span><span class="n">end</span><span class="p">)</span> <span class="o">-</span> <span class="nb">int</span><span class="p">(</span><span class="n">start</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span>
    <span class="n">mean_depth</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">total_reads</span><span class="p">)</span> <span class="o">*</span> <span class="nb">int</span><span class="p">(</span><span class="n">read_length</span><span class="p">)</span> <span class="o">/</span> <span class="n">total_length</span>


<span class="p">[</span><span class="n">get_counts</span><span class="p">:</span> <span class="kp">shared</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;num_genes&#39;</span><span class="p">,</span> <span class="s1">&#39;num_covered_genes&#39;</span><span class="p">,</span> <span class="s1">&#39;num_exons&#39;</span><span class="p">,</span> <span class="s1">&#39;num_variants&#39;</span><span class="p">,</span> <span class="s1">&#39;num_valid_variants&#39;</span><span class="p">,</span> <span class="s1">&#39;num_oncofuse_fusions&#39;</span><span class="p">)]</span>

<span class="kn">depends:</span>
    <span class="s1">&#39;${count_out}/${samplename}-gene_counts_rpkm.xls&#39;</span><span class="p">,</span>
    <span class="s1">&#39;${count_out}/${samplename}-exon_counts_rpkm.xls&#39;</span><span class="p">,</span>
    <span class="s1">&#39;${fusion_out}/fusions_oncofuse.xls&#39;</span><span class="p">,</span>
    <span class="s1">&#39;${variant_out}/${samplename}-filter.vcf&#39;</span>

<span class="k">def</span> <span class="nf">_lineCount</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="n">matching</span><span class="o">=</span><span class="bp">None</span><span class="p">):</span>
    <span class="n">count</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">count_no_comment</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">count_matching</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">ifile</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">ifile</span><span class="p">:</span>
            <span class="n">count</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;#&#39;</span><span class="p">):</span>
                <span class="n">count_no_comment</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="k">if</span> <span class="n">matching</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span><span class="p">:</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">()[</span><span class="n">matching</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="o">==</span> <span class="n">matching</span><span class="p">[</span><span class="mi">1</span><span class="p">]:</span>
                        <span class="n">count_matching</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="k">except</span><span class="p">:</span>
                    <span class="k">pass</span>
    <span class="k">return</span> <span class="n">count</span><span class="p">,</span> <span class="n">count_no_comment</span><span class="p">,</span> <span class="n">count_matching</span>

<span class="n">gene_count</span> <span class="o">=</span> <span class="n">_lineCount</span><span class="p">(</span><span class="s1">&#39;${count_out}/${samplename}-gene_counts_rpkm.xls&#39;</span><span class="p">,</span> <span class="n">matching</span><span class="o">=</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="s1">&#39;0&#39;</span><span class="p">))</span>
<span class="n">num_genes</span> <span class="o">=</span> <span class="n">gene_count</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">-</span> <span class="mi">1</span>
<span class="c1"># all lines - 0 count</span>
<span class="n">num_covered_genes</span> <span class="o">=</span> <span class="n">gene_count</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">-</span> <span class="n">gene_count</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="o">-</span> <span class="mi">1</span>
<span class="n">num_exons</span> <span class="o">=</span> <span class="n">_lineCount</span><span class="p">(</span><span class="s1">&#39;${count_out}/${samplename}-exon_counts_rpkm.xls&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">-</span> <span class="mi">1</span>

<span class="n">num_oncofuse_fusions</span> <span class="o">=</span> <span class="n">_lineCount</span><span class="p">(</span><span class="s1">&#39;${fusion_out}/fusions_oncofuse.xls&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">-</span> <span class="mi">1</span>
<span class="n">variant_count</span> <span class="o">=</span> <span class="n">_lineCount</span><span class="p">(</span><span class="s1">&#39;${variant_out}/${samplename}-filter.vcf&#39;</span><span class="p">,</span> <span class="n">matching</span><span class="o">=</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="s1">&#39;PASS&#39;</span><span class="p">))</span>

<span class="n">num_variants</span> <span class="o">=</span> <span class="n">variant_count</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
<span class="n">num_valid_variants</span> <span class="o">=</span> <span class="n">variant_count</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>

<span class="p">[</span><span class="n">human_hg19_910</span><span class="p">]</span>
<span class="c1">#Summary: Write summary.html to report all results.</span>

<span class="kn">output:</span>
    <span class="s1">&#39;${output_dir}/summary.html&#39;</span>

<span class="kn">depends:</span>
    <span class="p">[</span><span class="n">sos_variable</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="p">(</span>
        <span class="s1">&#39;read_length&#39;</span><span class="p">,</span>
        <span class="s1">&#39;mean_depth&#39;</span><span class="p">,</span>
        <span class="s1">&#39;mate_inner_dist&#39;</span><span class="p">,</span>
        <span class="s1">&#39;mate_std_dev&#39;</span><span class="p">,</span>
        <span class="s1">&#39;total_reads&#39;</span><span class="p">,</span>
        <span class="s1">&#39;used_reads&#39;</span><span class="p">,</span>
        <span class="s1">&#39;mapped_reads&#39;</span><span class="p">,</span>
        <span class="s1">&#39;mapped_reads_perc&#39;</span><span class="p">,</span>
        <span class="s1">&#39;uniq_mapped_reads&#39;</span><span class="p">,</span>
        <span class="s1">&#39;uniq_mapped_reads_perc&#39;</span><span class="p">,</span>
        <span class="s1">&#39;mapped_pairs&#39;</span><span class="p">,</span>
        <span class="s1">&#39;mapped_pairs_perc&#39;</span><span class="p">,</span>
        <span class="s1">&#39;concordant_mapped_pairs&#39;</span><span class="p">,</span>
        <span class="s1">&#39;concordant_mapped_pairs_perc&#39;</span><span class="p">,</span>
        <span class="s1">&#39;genome_mapped_count&#39;</span><span class="p">,</span>
        <span class="s1">&#39;genome_mapped_perc&#39;</span><span class="p">,</span>
        <span class="s1">&#39;junction_mapped_count&#39;</span><span class="p">,</span>
        <span class="s1">&#39;junction_mapped_perc&#39;</span><span class="p">,</span>
        <span class="s1">&#39;reference_genes&#39;</span><span class="p">,</span>
        <span class="s1">&#39;reference_genome&#39;</span><span class="p">,</span>
        <span class="s1">&#39;num_genes&#39;</span><span class="p">,</span>
        <span class="s1">&#39;num_covered_genes&#39;</span><span class="p">,</span>
        <span class="s1">&#39;num_exons&#39;</span><span class="p">,</span>
        <span class="s1">&#39;num_variants&#39;</span><span class="p">,</span>
        <span class="s1">&#39;num_valid_variants&#39;</span><span class="p">,</span>
        <span class="s1">&#39;num_oncofuse_fusions&#39;</span><span class="p">)]</span>

<span class="kn">parameter:</span> <span class="n">css</span> <span class="o">=</span> <span class="s1">&#39;default.css&#39;</span>

<span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">css</span><span class="p">):</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s1">&#39;{} does not exist. No CSS is used in HTML output&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">css</span><span class="p">))</span>
    <span class="n">css_style</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
<span class="kn">else:</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">css</span><span class="p">)</span> <span class="k">as</span> <span class="n">style</span><span class="p">:</span>
        <span class="n">css_style</span> <span class="o">=</span> <span class="n">style</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>

<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">output</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">summary</span><span class="p">:</span>
    <span class="n">summary</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">r&#39;&#39;&#39;</span>
<span class="s1">&lt;!DOCTYPE html PUBLIC &quot;-//W3C//DTD HTML 4.01 Transitional//EN&quot;&gt;</span>
<span class="s1">&lt;html&gt;</span>
<span class="s1">&lt;head&gt;</span>
<span class="s1">  &lt;meta content=&quot;text/html; charset=ISO-8859-1&quot;</span>
<span class="s1">    http-equiv=&quot;content-type&quot;&gt;</span>
<span class="s1">  &lt;title&gt;${samplename}&lt;/title&gt;</span>
<span class="s1"> &lt;style type=&quot;text/css&quot;&gt;</span>
<span class="s1"> ${css_style}</span>
<span class="s1"> &lt;/style&gt;</span>
<span class="s1">&lt;/head&gt;</span>
<span class="s1">&lt;body&gt;</span>
<span class="s1">&lt;h1&gt;RNA Seq analysis for sample ${samplename}&lt;/h1&gt;</span>
<span class="s1">&lt;p&gt;Generated by pipeline ${pipeline_name} ${pipeline_version}&lt;/p&gt;</span>

<span class="s1">&lt;h2&gt;Project description&lt;/h2&gt;</span>

<span class="s1">The goal of the analysis is to evaluate how well RNA Seq sequencing worked on the sample,</span>
<span class="s1">obtain gene and exon expression, fusion candidates, and variant calls from the sample.</span>

<span class="s1">&lt;h3&gt;Basic information&lt;/h3&gt;</span>

<span class="s1">&lt;table class=&quot;gridtable&quot;&gt;</span>
<span class="s1">&lt;tr&gt;</span>
<span class="s1">&lt;th&gt;Sample name&lt;/th&gt;&lt;td&gt;${samplename}&lt;/td&gt;</span>
<span class="s1">&lt;/tr&gt;</span>
<span class="s1">&lt;tr&gt;</span>
<span class="s1">&lt;th&gt;Type&lt;/th&gt;&lt;td&gt;Paired ends&lt;/td&gt;</span>
<span class="s1">&lt;/tr&gt;</span>
<span class="s1">&lt;tr&gt;</span>
<span class="s1">&lt;th&gt;Read length&lt;/th&gt;&lt;td&gt;${read_length}&lt;/td&gt;</span>
<span class="s1">&lt;/tr&gt;</span>
<span class="s1">&lt;tr&gt;</span>
<span class="s1">&lt;th&gt;Mean Depth of Coverage &lt;sup&gt;*&lt;/sup&gt;&lt;/th&gt;&lt;td&gt;${mean_depth:.1f} &lt;/td&gt;</span>
<span class="s1">&lt;/tr&gt;</span>
<span class="s1">&lt;tr&gt;</span>
<span class="s1">&lt;th&gt;Inner distance between paired reads&lt;sup&gt;**&lt;/sup&gt;&lt;/th&gt;&lt;td&gt;${mate_inner_dist} (stdev=${mate_std_dev})&lt;/td&gt;</span>
<span class="s1">&lt;/tr&gt;</span>
<span class="s1">&lt;/table&gt;</span>

<span class="s1">&lt;p&gt;* The mean depth of coverage is roughly estimated by number of mapped reads times read length and divided by the total length of exons. </span>
<span class="s1">A low depth of coverage might appear if the mapping rate is low.&lt;/p&gt;</span>
<span class="s1">&lt;p&gt;** Mean and standard deviation of inner distance is estimated by distances between 1M pairs of aligned reads.&lt;/p&gt;</span>

<span class="s1">&lt;h2&gt;Result summary&lt;/h2&gt;</span>

<span class="s1">&lt;h3&gt;Read QC -- FastQC report&lt;/h3&gt;</span>
<span class="s1">&lt;p&gt;&lt;a href=&quot;http://www.bioinformatics.babraham.ac.uk/projects/fastqc/&quot;&gt;FastQC&lt;/a&gt; is a quality control tool for high throughput sequence data. </span>
<span class="s1">Please check &lt;a href=&quot;http://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/&quot;&gt;here&lt;/a&gt; for details of each plot.</span>
<span class="s1">Note that it is normal for RNA Seq samples to have high duplication level and uneven distribution of Kmer content. Warnings </span>
<span class="s1">on items such GC content distribution are more alarming.&lt;/p&gt;</span>

<span class="s1">&lt;table class=&quot;gridtable&quot;&gt;</span>
<span class="s1">&lt;tr&gt;</span>
<span class="s1">&lt;tr&gt;&lt;th&gt;Per base sequence quality&lt;/th&gt;&lt;th&gt;Per base sequence content&lt;/th&gt;&lt;th&gt;Per sequence GC content&lt;/th&gt;&lt;/tr&gt;</span>
<span class="s1">&lt;tr&gt;</span>
<span class="s1">    &lt;td&gt;</span>
<span class="s1">    &lt;a href=&quot;../${readqc_out}/${samplename}_fastqc.html#M1&quot;&gt;</span>
<span class="s1">    &lt;IMG SRC=&quot;../${readqc_out}/per_base_quality.png&quot; WIDTH=300 ALT=&quot;Per base sequence quality&quot;&gt;</span>
<span class="s1">    &lt;/a&gt;&lt;/td&gt;</span>
<span class="s1">    &lt;td&gt;</span>
<span class="s1">     &lt;a href=&quot;../${readqc_out}/${samplename}_fastqc.html#M4&quot;&gt;</span>
<span class="s1">    &lt;IMG SRC=&quot;../${readqc_out}/per_base_sequence_content.png&quot; WIDTH=300 ALT=&quot;Per base sequence content&quot;&gt;</span>
<span class="s1">    &lt;/a&gt;&lt;/td&gt;</span>
<span class="s1">    &lt;td&gt;</span>
<span class="s1">    &lt;a href=&quot;../${readqc_out}/${samplename}_fastqc.html#M5&quot;&gt;</span>
<span class="s1">    &lt;IMG SRC=&quot;../${readqc_out}/per_sequence_gc_content.png&quot; WIDTH=300 ALT=&quot;Per sequence GC content&quot;&gt;</span>
<span class="s1">    &lt;/a&gt;&lt;/td&gt;</span>
<span class="s1">&lt;/tr&gt;</span>
<span class="s1">&lt;/table&gt;</span>

<span class="s1">&lt;p&gt;&lt;a href=&quot;../${readqc_out}/${samplename}_fastqc.html&quot;&gt;Complete FastQC reports&lt;/a&gt; &lt;/p&gt;</span>

<span class="s1">&lt;h3&gt;Alignment summary&lt;/h3&gt;</span>
<span class="s1">&lt;p&gt;Input reads are aligned to the transcriptome of ${reference_genes} genes using reference genome ${reference_genome}.</span>
<span class="s1">The mapping rate might appear higher if the reads are mapped to the whole genome.&lt;/p&gt;</span>

<span class="s1">&lt;table class=&quot;gridtable&quot;&gt;</span>
<span class="s1">&lt;tr&gt;</span>
<span class="s1">&lt;th&gt;Total reads&lt;/th&gt;</span>
<span class="s1">&lt;th&gt;Used reads&lt;/th&gt;</span>
<span class="s1">&lt;th&gt;Mapped reads&lt;/th&gt;</span>
<span class="s1">&lt;th&gt;Uniquely mapped reads&lt;/th&gt;</span>
<span class="s1">&lt;th&gt;Mapped pairs&lt;/th&gt;</span>
<span class="s1">&lt;th&gt;Concordant mapped pairs&lt;/th&gt;</span>
<span class="s1">&lt;th&gt;Mapped reads (genome)&lt;/th&gt;</span>
<span class="s1">&lt;th&gt;Mapped reads (junction)&lt;/th&gt;</span>
<span class="s1">&lt;/tr&gt;</span>
<span class="s1">&lt;tr&gt;</span>
<span class="s1">&lt;td&gt;${total_reads:,}&lt;/td&gt;</span>
<span class="s1">&lt;td&gt;${used_reads:,}&lt;/td&gt;</span>
<span class="s1">&lt;td&gt;${mapped_reads:,} (${mapped_reads_perc:.1f}%)&lt;/td&gt;</span>
<span class="s1">&lt;td&gt;${uniq_mapped_reads:,} (${uniq_mapped_reads_perc:.1f}%)&lt;/td&gt;</span>
<span class="s1">&lt;td&gt;${mapped_pairs:,} (${mapped_pairs_perc:.1f}%)&lt;/td&gt;</span>
<span class="s1">&lt;td&gt;${concordant_mapped_pairs:,} (${concordant_mapped_pairs_perc:.1f}%)&lt;/td&gt;</span>
<span class="s1">&lt;td&gt;${genome_mapped_count:,} (${genome_mapped_perc:.1f}%)&lt;/td&gt;</span>
<span class="s1">&lt;td&gt;${junction_mapped_count:,} (${junction_mapped_perc:.1f}%)&lt;/td&gt;</span>
<span class="s1">&lt;/tr&gt;</span>
<span class="s1">&lt;/table&gt;</span>

<span class="s1">&lt;h3&gt;Alignment QC -- RSeQC plots&lt;/h3&gt;</span>
<span class="s1">&lt;table class=&quot;gridtable&quot;&gt;</span>
<span class="s1">&lt;tr&gt;</span>
<span class="s1">&lt;tr&gt;&lt;th&gt;Junction saturation&lt;/th&gt;&lt;th&gt;Inner distance&lt;/th&gt;&lt;th&gt;Splicing junction&lt;/th&gt;&lt;/tr&gt;</span>
<span class="s1">&lt;tr&gt;</span>
<span class="s1">    &lt;td&gt;</span>
<span class="s1">    &lt;a href=&quot;../${alignmentqc_out}/${samplename}.junctionSaturation_plot.pdf&quot;&gt;</span>
<span class="s1">    &lt;IMG SRC=&quot;../${alignmentqc_out}/${samplename}.junctionSaturation_plot.jpg&quot; WIDTH=300 ALT=&quot;Junction saturation plot&quot;&gt;</span>
<span class="s1">    &lt;/a&gt;&lt;/td&gt;</span>
<span class="s1">    &lt;td&gt;</span>
<span class="s1">     &lt;a href=&quot;../${alignmentqc_out}/${samplename}.inner_distance_plot.pdf&quot;&gt;</span>
<span class="s1">    &lt;IMG SRC=&quot;../${alignmentqc_out}/${samplename}.inner_distance_plot.jpg&quot; WIDTH=300 ALT=&quot;Inner distance plot&quot;&gt;</span>
<span class="s1">    &lt;/a&gt;&lt;/td&gt;</span>
<span class="s1">    &lt;td&gt;</span>
<span class="s1">    &lt;a href=&quot;../${alignmentqc_out}/${samplename}.splice_junction.pdf&quot;&gt;</span>
<span class="s1">    &lt;IMG SRC=&quot;../${alignmentqc_out}/${samplename}.splice_junction.jpg&quot; WIDTH=300 ALT=&quot;Splicing junction plot&quot;&gt;</span>
<span class="s1">    &lt;/a&gt;&lt;/td&gt;</span>
<span class="s1">&lt;/tr&gt;</span>
<span class="s1">&lt;tr&gt;&lt;th&gt;Read duplication&lt;/th&gt;&lt;th&gt;Gene body coverage&lt;/th&gt;&lt;th&gt;Splicing events&lt;/th&gt;&lt;/tr&gt;</span>
<span class="s1">&lt;tr&gt;</span>
<span class="s1">    &lt;td&gt;</span>
<span class="s1">    &lt;a href=&quot;../${alignmentqc_out}/${samplename}.DupRate_plot.pdf&quot;&gt;</span>
<span class="s1">    &lt;IMG SRC=&quot;../${alignmentqc_out}/${samplename}.DupRate_plot.jpg&quot; WIDTH=300 ALT=&quot;Read duplication plot&quot;&gt;</span>
<span class="s1">    &lt;/a&gt;&lt;/td&gt;</span>
<span class="s1">    &lt;td&gt;</span>
<span class="s1">    &lt;a href=&quot;../${alignmentqc_out}/${samplename}.geneBodyCoverage.curves.pdf&quot;&gt;</span>
<span class="s1">    &lt;IMG SRC=&quot;../${alignmentqc_out}/${samplename}.geneBodyCoverage.curves.jpg&quot; WIDTH=300 ALT=&quot;Gene body coverage plot&quot;&gt;</span>
<span class="s1">    &lt;/a&gt;&lt;/td&gt;</span>
<span class="s1">    &lt;td&gt;</span>
<span class="s1">    &lt;a href=&quot;../${alignmentqc_out}/${samplename}.splice_events.pdf&quot;&gt;</span>
<span class="s1">    &lt;IMG SRC=&quot;../${alignmentqc_out}/${samplename}.splice_events.jpg&quot; WIDTH=300 ALT=&quot;Splicing events plot&quot;&gt;</span>
<span class="s1">    &lt;/a&gt;&lt;/td&gt;</span>
<span class="s1">    &lt;/tr&gt;</span>
<span class="s1">&lt;/table&gt;</span>

<span class="s1">&lt;h3&gt;Gene and exon counts&lt;/h3&gt;</span>

<span class="s1">&lt;ul&gt;</span>
<span class="s1">&lt;li&gt;&lt;a href=&quot;../${count_out}/${samplename}-gene_counts_rpkm.xls&quot;&gt;Gene counts with RPKM &lt;/a&gt; (${num_genes:,} genes, ${num_covered_genes:,} overlap with at least 1 read)&lt;/li&gt;</span>
<span class="s1">&lt;li&gt;&lt;a href=&quot;../${count_out}/${samplename}-exon_counts_rpkm.xls&quot;&gt;Exon counts with RPKM &lt;/a&gt; (${num_exons:,} exons overlapped with at least 1 read)&lt;/li&gt;</span>
<span class="s1">&lt;/ul&gt;</span>

<span class="s1">&lt;p&gt;The gene count was performed by &lt;a href=&quot;http://www-huber.embl.de/HTSeq/doc/count.html&quot;&gt;htseq-count&lt;/a&gt; under</span>
<span class="s1">the &lt;b&gt;union&lt;/b&gt; model. The exon count was performed by &lt;a href=&quot;https://code.google.com/p/bedtools/&quot;&gt;bedtools&lt;/a&gt;</span>
<span class="s1">which counts the number of reads that overlap with each exon. These files are tab-delimited files with .xls extension.&lt;/p&gt;</span>

<span class="s1">&lt;h3&gt;Fusion candidates&lt;/h3&gt;</span>
<span class="s1">The following are fusion candidates produced by tophat-fusion:</span>
<span class="s1">&lt;ul&gt;</span>
<span class="s1">&lt;li&gt;&lt;a href=&quot;../${fusion_out}/fusions_oncofuse.xls&quot;&gt;Oncofuse filtered fusion candidates&lt;/a&gt; (${num_oncofuse_fusions:,} fusions in tab-delimited file with .xls extension)&lt;/li&gt;</span>
<span class="s1">&lt;/ul&gt;</span>

<span class="s1">&lt;h3&gt;Variant calls&lt;/h3&gt;</span>
<span class="s1">&lt;ul&gt;</span>
<span class="s1">&lt;li&gt; &lt;a href=&quot;../${variant_out}/${samplename}-filter.vcf&quot;&gt;GATK detected SNV in VCF format&lt;/a&gt;(${num_variants:,} variants, ${num_valid_variants:,} passed basic filtering)&lt;/li&gt;</span>
<span class="s1">&lt;/ul&gt;</span>

<span class="s1">The variants are called using the GATK Unified Genotyper and include only Single Nucleotide Variants, no indel detection was performed.&lt;/p&gt;</span>
<span class="s1">&lt;p&gt;</span>
<span class="s1">&lt;p&gt;</span>
<span class="s1">&lt;/body&gt;</span>
<span class="s1">&lt;/html&gt;</span>
<span class="s1">&#39;&#39;&#39;</span><span class="p">)</span>

<span class="p">[</span><span class="n">human_hg19_1000</span><span class="p">]</span>
<span class="c1">#Summary: Compress all result files into a deliverable package.</span>
<span class="kn">input:</span>
    <span class="s1">&#39;${output_dir}/summary.html&#39;</span><span class="p">,</span>
    <span class="s1">&#39;${readqc_out}/${samplename}_fastqc.html&#39;</span><span class="p">,</span>
    <span class="s1">&#39;${readqc_out}/per_base_quality.png&#39;</span><span class="p">,</span>
    <span class="s1">&#39;${readqc_out}/per_base_sequence_content.png&#39;</span><span class="p">,</span>
    <span class="s1">&#39;${readqc_out}/per_sequence_gc_content.png&#39;</span><span class="p">,</span>
    <span class="s1">&#39;${alignmentqc_out}/${samplename}.geneBodyCoverage.curves.pdf&#39;</span><span class="p">,</span>
    <span class="s1">&#39;${alignmentqc_out}/${samplename}.junctionSaturation_plot.pdf&#39;</span><span class="p">,</span>
    <span class="s1">&#39;${alignmentqc_out}/${samplename}.splice_events.pdf&#39;</span><span class="p">,</span>
    <span class="s1">&#39;${alignmentqc_out}/${samplename}.splice_junction.pdf&#39;</span><span class="p">,</span>
    <span class="s1">&#39;${alignmentqc_out}/${samplename}.inner_distance_plot.pdf&#39;</span><span class="p">,</span>
    <span class="s1">&#39;${alignmentqc_out}//${samplename}.DupRate_plot.pdf&#39;</span><span class="p">,</span>
    <span class="s1">&#39;${alignmentqc_out}/${samplename}.geneBodyCoverage.curves.jpg&#39;</span><span class="p">,</span>
    <span class="s1">&#39;${alignmentqc_out}/${samplename}.junctionSaturation_plot.jpg&#39;</span><span class="p">,</span>
    <span class="s1">&#39;${alignmentqc_out}/${samplename}.splice_events.jpg&#39;</span><span class="p">,</span>
    <span class="s1">&#39;${alignmentqc_out}/${samplename}.splice_junction.jpg&#39;</span><span class="p">,</span>
    <span class="s1">&#39;${alignmentqc_out}/${samplename}.inner_distance_plot.jpg&#39;</span><span class="p">,</span>
    <span class="s1">&#39;${alignmentqc_out}/${samplename}.DupRate_plot.jpg&#39;</span><span class="p">,</span>
    <span class="s1">&#39;${count_out}/${samplename}-gene_counts_rpkm.xls&#39;</span><span class="p">,</span>
    <span class="s1">&#39;${count_out}/${samplename}-exon_counts_rpkm.xls&#39;</span><span class="p">,</span>
    <span class="s1">&#39;${fusion_out}/fusions_oncofuse.xls&#39;</span><span class="p">,</span>
    <span class="s1">&#39;${variant_out}/${samplename}-filter.vcf&#39;</span>

<span class="kn">output:</span>
    <span class="s1">&#39;${output_dir}/${samplename}.zip&#39;</span>

<span class="kn">run:</span>
    <span class="nb">zip</span> <span class="err">$</span><span class="p">{</span><span class="n">output_dir</span><span class="p">}</span><span class="o">/</span><span class="err">$</span><span class="p">{</span><span class="n">samplename</span><span class="p">}</span><span class="o">.</span><span class="n">zip</span> <span class="err">$</span><span class="p">{</span><span class="nb">input</span><span class="p">}</span>
</pre></div>
<h2 id="Workflow-to-install-required-tools">Workflow to install required tools<a class="anchor-link" href="#Workflow-to-install-required-tools">&#182;</a></h2><div class="highlight"><pre><span></span><span class="ch">#!/usr/bin/env sos-runner</span>
<span class="c1">#fileformat=SOS1.0</span>

<span class="c1">#</span>
<span class="c1"># Auxiliary steps to install required NGS tools if it is not avaible</span>
<span class="c1">#</span>

<span class="c1"># default installation directory. This directory is recognized by</span>
<span class="c1"># default by SoS so that commands installed would be immediately usable</span>
<span class="c1"># by sos workflows. You can use commands such as</span>
<span class="c1">#</span>
<span class="c1">#    sudo sos run install_ngs install_fastqc --install-dir /usr/local/bin/</span>
<span class="c1">#</span>
<span class="c1"># to install programs in a different directory.</span>

<span class="kn">parameter:</span> <span class="n">install_dir</span>  <span class="o">=</span> <span class="s1">&#39;~/.sos/bin&#39;</span>
<span class="kn">parameter:</span> <span class="n">download_dir</span> <span class="o">=</span> <span class="s1">&#39;~/.sos/bin/download&#39;</span>

<span class="p">[</span><span class="n">install_fastqc</span><span class="p">:</span> <span class="n">provides</span><span class="o">=</span><span class="n">executable</span><span class="p">(</span><span class="s1">&#39;fastqc&#39;</span><span class="p">)]</span>
<span class="n">ver</span> <span class="o">=</span> <span class="s1">&#39;v0.11.5&#39;</span>

<span class="kn">download:</span> <span class="n">dest_dir</span><span class="o">=</span><span class="n">download_dir</span>
    <span class="n">http</span><span class="p">:</span><span class="o">//</span><span class="n">www</span><span class="o">.</span><span class="n">bioinformatics</span><span class="o">.</span><span class="n">babraham</span><span class="o">.</span><span class="n">ac</span><span class="o">.</span><span class="n">uk</span><span class="o">/</span><span class="n">projects</span><span class="o">/</span><span class="n">fastqc</span><span class="o">/</span><span class="n">fastqc_</span><span class="err">$</span><span class="p">{</span><span class="n">ver</span><span class="p">}</span><span class="o">.</span><span class="n">zip</span>

<span class="n">warn_if</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isdir</span><span class="p">(</span><span class="s1">&#39;${install_dir!e}/FastQC_${ver}&#39;</span><span class="p">),</span>
    <span class="s2">&quot;Replacing FastQC version ${ver} under ${install_dir!e}/FastQC_${ver}.&quot;</span><span class="p">)</span>

<span class="n">warn_if</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="s1">&#39;${install_dir!e}/fastqc}&#39;</span><span class="p">),</span>
    <span class="s2">&quot;Replacing fastqc command ${install_dir!e}/fastqc.&quot;</span><span class="p">)</span>

<span class="kn">run:</span>    
    <span class="n">cd</span> <span class="err">$</span><span class="p">{</span><span class="n">install_dir</span><span class="err">!</span><span class="n">e</span><span class="p">}</span> <span class="o">&amp;&amp;</span> \
    <span class="n">rm</span> <span class="o">-</span><span class="n">rf</span> <span class="err">$</span><span class="p">{</span><span class="n">install_dir</span><span class="err">!</span><span class="n">e</span><span class="p">}</span><span class="o">/</span><span class="n">FastQC_</span><span class="err">$</span><span class="p">{</span><span class="n">ver</span><span class="p">}</span> <span class="o">&amp;&amp;</span> \
    <span class="n">rm</span> <span class="o">-</span><span class="n">f</span> <span class="err">$</span><span class="p">{</span><span class="n">install_dir</span><span class="err">!</span><span class="n">e</span><span class="p">}</span><span class="o">/</span><span class="n">fastqc</span> <span class="o">&amp;&amp;</span> \
    <span class="n">unzip</span> <span class="o">-</span><span class="n">o</span> <span class="err">$</span><span class="p">{</span><span class="n">download_dir</span><span class="err">!</span><span class="n">e</span><span class="p">}</span><span class="o">/</span><span class="n">fastqc_</span><span class="o">*.</span><span class="n">zip</span>  <span class="o">&amp;&amp;</span> \
    <span class="n">mv</span> <span class="err">$</span><span class="p">{</span><span class="n">install_dir</span><span class="err">!</span><span class="n">e</span><span class="p">}</span><span class="o">/</span><span class="n">FastQC</span> <span class="err">$</span><span class="p">{</span><span class="n">install_dir</span><span class="err">!</span><span class="n">e</span><span class="p">}</span><span class="o">/</span><span class="n">FastQC_</span><span class="err">$</span><span class="p">{</span><span class="n">ver</span><span class="p">}</span> <span class="o">&amp;&amp;</span> \
    <span class="n">chmod</span> <span class="mi">755</span> <span class="err">$</span><span class="p">{</span><span class="n">install_dir</span><span class="p">}</span><span class="o">/</span><span class="n">FastQC_</span><span class="err">$</span><span class="p">{</span><span class="n">ver</span><span class="p">}</span><span class="o">/</span><span class="n">fastqc</span> <span class="o">&amp;&amp;</span> \
    <span class="n">ln</span> <span class="o">-</span><span class="n">s</span> <span class="err">$</span><span class="p">{</span><span class="n">install_dir</span><span class="p">}</span><span class="o">/</span><span class="n">FastQC_</span><span class="err">$</span><span class="p">{</span><span class="n">ver</span><span class="p">}</span><span class="o">/</span><span class="n">fastqc</span> <span class="err">$</span><span class="p">{</span><span class="n">install_dir</span><span class="p">}</span>

<span class="p">[</span><span class="n">install_bowtie</span><span class="p">:</span> <span class="n">provides</span><span class="o">=</span><span class="n">executable</span><span class="p">(</span><span class="s1">&#39;bowtie&#39;</span><span class="p">)]</span>
<span class="kn">download:</span> <span class="n">dest_dir</span><span class="o">=</span><span class="n">download_dir</span>
    <span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">sourceforge</span><span class="o">.</span><span class="n">net</span><span class="o">/</span><span class="n">projects</span><span class="o">/</span><span class="n">bowtie</span><span class="o">-</span><span class="n">bio</span><span class="o">/</span><span class="n">files</span><span class="o">/</span><span class="n">bowtie</span><span class="o">/</span><span class="mf">1.1</span><span class="o">.</span><span class="mi">2</span><span class="o">/</span><span class="n">bowtie</span><span class="o">-</span><span class="mf">1.1</span><span class="o">.</span><span class="mi">2</span><span class="o">-</span><span class="n">macos</span><span class="o">-</span><span class="n">x86_64</span><span class="o">.</span><span class="n">zip</span>

<span class="kn">run:</span>    <span class="n">workdir</span><span class="o">=</span><span class="n">download_dir</span>
    <span class="n">unzip</span> <span class="n">bowtie</span><span class="o">-</span><span class="mf">1.1</span><span class="o">.</span><span class="mi">2</span><span class="o">-</span><span class="n">macos</span><span class="o">-</span><span class="n">x86_64</span><span class="o">.</span><span class="n">zip</span> <span class="o">&amp;&amp;</span> \
    <span class="n">rm</span> <span class="o">-</span><span class="n">rf</span> <span class="err">$</span><span class="p">{</span><span class="n">install_dir</span><span class="err">!</span><span class="n">e</span><span class="p">}</span><span class="o">/</span><span class="n">bowtie</span><span class="o">-</span><span class="mf">1.1</span><span class="o">.</span><span class="mi">2</span> <span class="o">&amp;&amp;</span> \
    <span class="n">mv</span> <span class="n">bowtie</span><span class="o">-</span><span class="mf">1.1</span><span class="o">.</span><span class="mi">2</span> <span class="err">$</span><span class="p">{</span><span class="n">install_dir</span><span class="err">!</span><span class="n">e</span><span class="p">}</span> <span class="o">&amp;&amp;</span> \
    <span class="n">ln</span> <span class="o">-</span><span class="n">s</span> <span class="o">-</span><span class="n">f</span> <span class="err">$</span><span class="p">{</span><span class="n">install_dir</span><span class="err">!</span><span class="n">e</span><span class="p">}</span><span class="o">/</span><span class="n">bowtie</span><span class="o">-</span><span class="mf">1.1</span><span class="o">.</span><span class="mi">2</span><span class="o">/*</span> <span class="err">$</span><span class="p">{</span><span class="n">install_dir</span><span class="err">!</span><span class="n">e</span><span class="p">}</span>

<span class="p">[</span><span class="n">install_bowtie2</span><span class="p">:</span> <span class="n">provides</span><span class="o">=</span><span class="n">executable</span><span class="p">(</span><span class="s1">&#39;bowtie2&#39;</span><span class="p">)]</span>
<span class="kn">download:</span> <span class="n">dest_dir</span><span class="o">=</span><span class="n">download_dir</span>
    <span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">sourceforge</span><span class="o">.</span><span class="n">net</span><span class="o">/</span><span class="n">projects</span><span class="o">/</span><span class="n">bowtie</span><span class="o">-</span><span class="n">bio</span><span class="o">/</span><span class="n">files</span><span class="o">/</span><span class="n">bowtie2</span><span class="o">/</span><span class="mf">2.2</span><span class="o">.</span><span class="mi">9</span><span class="o">/</span><span class="n">bowtie2</span><span class="o">-</span><span class="mf">2.2</span><span class="o">.</span><span class="mi">9</span><span class="o">-</span><span class="n">macos</span><span class="o">-</span><span class="n">x86_64</span><span class="o">.</span><span class="n">zip</span>

<span class="kn">run:</span>    <span class="n">workdir</span><span class="o">=</span><span class="n">download_dir</span>
    <span class="n">unzip</span> <span class="n">bowtie2</span><span class="o">-</span><span class="mf">2.2</span><span class="o">.</span><span class="mi">9</span><span class="o">-</span><span class="n">macos</span><span class="o">-</span><span class="n">x86_64</span><span class="o">.</span><span class="n">zip</span> <span class="o">&amp;&amp;</span> \
    <span class="n">rm</span> <span class="o">-</span><span class="n">rf</span> <span class="err">$</span><span class="p">{</span><span class="n">install_dir</span><span class="err">!</span><span class="n">e</span><span class="p">}</span><span class="o">/</span><span class="n">bowtie2</span><span class="o">-</span><span class="mf">2.2</span><span class="o">.</span><span class="mi">9</span> <span class="o">&amp;&amp;</span> \
    <span class="n">mv</span> <span class="n">bowtie2</span><span class="o">-</span><span class="mf">2.2</span><span class="o">.</span><span class="mi">9</span> <span class="err">$</span><span class="p">{</span><span class="n">install_dir</span><span class="err">!</span><span class="n">e</span><span class="p">}</span> <span class="o">&amp;&amp;</span> \
    <span class="n">ln</span> <span class="o">-</span><span class="n">s</span> <span class="o">-</span><span class="n">f</span> <span class="err">$</span><span class="p">{</span><span class="n">install_dir</span><span class="err">!</span><span class="n">e</span><span class="p">}</span><span class="o">/</span><span class="n">bowtie2</span><span class="o">-</span><span class="mf">2.2</span><span class="o">.</span><span class="mi">9</span><span class="o">/*</span> <span class="err">$</span><span class="p">{</span><span class="n">install_dir</span><span class="err">!</span><span class="n">e</span><span class="p">}</span>


<span class="p">[</span><span class="n">install_tophat2</span><span class="p">:</span> <span class="n">provides</span><span class="o">=</span><span class="n">executable</span><span class="p">(</span><span class="s1">&#39;tophat2&#39;</span><span class="p">)]</span>
<span class="c1"># install tophat 2, which requires bowtie2</span>
<span class="c1">#</span>
<span class="c1"># NOTE: The pre-compiled binary does not work well under mac so it is</span>
<span class="c1"># highly recommened that you compile from source. However, compiling from</span>
<span class="c1"># source requires xcode and boost, and the later is not among the easiest</span>
<span class="c1"># software package to install</span>

<span class="kn">depends:</span>  <span class="n">executable</span><span class="p">(</span><span class="s1">&#39;bowtie2&#39;</span><span class="p">)</span>

<span class="c1">#download: dest_dir=download_dir</span>
<span class="c1">#    https://ccb.jhu.edu/software/tophat/downloads/tophat-2.0.13.tar.gz</span>

<span class="c1">#run:    workdir=download_dir</span>
<span class="c1">#    tar zxf tophat-2.0.13.tar.gz &amp;&amp; \</span>
<span class="c1">#    cd tophat-2.0.13 &amp;&amp; \</span>
<span class="c1">#    ./configure --prefix=${os.path.dirname(install_dir)!e} &amp;&amp; \</span>
<span class="c1">#    make -B &amp;&amp; make install</span>

<span class="kn">download:</span> <span class="n">dest_dir</span><span class="o">=</span><span class="n">download_dir</span>
    <span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">ccb</span><span class="o">.</span><span class="n">jhu</span><span class="o">.</span><span class="n">edu</span><span class="o">/</span><span class="n">software</span><span class="o">/</span><span class="n">tophat</span><span class="o">/</span><span class="n">downloads</span><span class="o">/</span><span class="n">tophat</span><span class="o">-</span><span class="mf">2.0</span><span class="o">.</span><span class="mf">13.</span><span class="n">OSX_x86_64</span><span class="o">.</span><span class="n">tar</span><span class="o">.</span><span class="n">gz</span>

<span class="kn">run:</span>    <span class="n">workdir</span><span class="o">=</span><span class="n">download_dir</span>
    <span class="n">tar</span> <span class="n">zxf</span> <span class="n">tophat</span><span class="o">-</span><span class="mf">2.0</span><span class="o">.</span><span class="mf">13.</span><span class="n">OSX_x86_64</span><span class="o">.</span><span class="n">tar</span><span class="o">.</span><span class="n">gz</span> <span class="o">&amp;&amp;</span> \
    <span class="n">rm</span> <span class="o">-</span><span class="n">rf</span> <span class="err">$</span><span class="p">{</span><span class="n">install_dir</span><span class="err">!</span><span class="n">e</span><span class="p">}</span><span class="o">/</span><span class="n">tophat</span><span class="o">-</span><span class="mf">2.0</span><span class="o">.</span><span class="mf">13.</span><span class="n">OSX_x86_64</span> <span class="o">&amp;&amp;</span> \
    <span class="n">mv</span> <span class="n">tophat</span><span class="o">-</span><span class="mf">2.0</span><span class="o">.</span><span class="mf">13.</span><span class="n">OSX_x86_64</span> <span class="err">$</span><span class="p">{</span><span class="n">install_dir</span><span class="err">!</span><span class="n">e</span><span class="p">}</span> <span class="o">&amp;&amp;</span> \
    <span class="n">ln</span> <span class="o">-</span><span class="n">s</span> <span class="o">-</span><span class="n">f</span> <span class="err">$</span><span class="p">{</span><span class="n">install_dir</span><span class="err">!</span><span class="n">e</span><span class="p">}</span><span class="o">/</span><span class="n">tophat</span><span class="o">-</span><span class="mf">2.0</span><span class="o">.</span><span class="mf">13.</span><span class="n">OSX_x86_64</span><span class="o">/*</span> <span class="err">$</span><span class="p">{</span><span class="n">install_dir</span><span class="err">!</span><span class="n">e</span><span class="p">}</span>


<span class="p">[</span><span class="n">install_rseqc</span><span class="p">:</span> <span class="n">provides</span><span class="o">=</span><span class="n">executable</span><span class="p">(</span><span class="s1">&#39;inner_distance.py&#39;</span><span class="p">)]</span>

<span class="kn">run:</span>
    <span class="n">pip</span> <span class="n">install</span> <span class="n">RSeQC</span>
</pre></div>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span>
</pre></div>

</div>
</div>
</div>

</div>
    </div>
  </div>
</body>
</html>
